---
title: "01_Registration_to_Yao23_MsHyp_Subclasses"
output: html_document
date: "2024-08-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
require(colorout)
library(magrittr)
library(BiocParallel)
library(orthogene)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall
library(parallel)
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

## Load  the Yao spatial registration output from spatialLIBD::registration_wrapper applied to cell types at the subclass level in their hierarchical scheme, and the subclass level but with all of the VMH and ARC (there, ARH) types re-labeled singularly as VMH and ARH respectively, and from VMH and ARC superclusters
```{r}
yaoscl <- readRDS("../spatial_HYP/processed-data/10-Spatial Registration/02b-yaofilt_registrationwrapper.RDS")

yaosuper <- readRDS("../spatial_HYP/processed-data/10-Spatial Registration/02b-Yao23-VMHARH_supercluster_regwrapper.RDS")

yaoclps <- readRDS("../spatial_HYP/processed-data/10-Spatial Registration/02c-yaofilt_collapsedVMHandARHsubclasses_registrationwrapper.RDS")

### load the registration wrapper output generated the second time (i.e. after excluding sample specific clusters based on annotations made from the first marker detection run and cluster spot plot examinations)
hypb.retainedclusts.mk <- readRDS("processed-data/05_Banksy_M0lam0_res2_multisamp/03a-M0l0kg6_reproducibleClustersOnly_spLIBD_markers.RDS")
```

## get orthologous genes
```{r}
orthos <- as.data.table(unique(as.data.table(hypb.retainedclusts.mk$enrichment$ensembl)))
orthos <- convert_orthologs(gene_df = orthos,gene_input = "V1",gene_output = "columns",input_species = "human",output_species = "mouse",non121_strategy = "drop_both_species")

### subset the human data to the genes w mouse orthologs; rename t_stat clusters to the cluster ID for registration 
hypb.mks <- as.data.table(hypb.retainedclusts.mk$enrichment)
keepcols <- c("ensembl",grep(names(hypb.mks),pattern="^t_stat_",value=T))
hypb.mks <- hypb.mks[ensembl %in% orthos$input_gene,..keepcols]
    
setnames(hypb.mks,c(2:ncol(hypb.mks)),gsub(names(hypb.mks)[c(2:ncol(hypb.mks))],pattern="^t_stat_",replacement=""))


hypb.mks <- as.data.frame(hypb.mks)
rownames(hypb.mks) <- hypb.mks$ensembl
hypb.mks$ensembl <- NULL

## subset the Yao data to xenium genes with mouse orthologs, rename those t stat columns to clusters likewise
tmplist <- list(yaoscl,yaosuper,yaoclps)
tmplist <- lapply(X=tmplist,function(x){

    Y <- as.data.table(x$enrichment,keep.rownames=T)
    keepcols <- c("rn",grep(names(Y),pattern="^t_stat_",value=T))
    Y <- Y[,..keepcols]
    
    setnames(Y,c(2:ncol(Y)),gsub(names(Y)[c(2:ncol(Y))],pattern="^t_stat_",replacement=""))
    
    # subset to human genes w ensembl orthologs, rename rows to the human ortholog
    Y <- merge.data.table(Y,orthos,by.x="rn",by.y="ortholog_gene")

    Y <- as.data.frame(Y)
    rownames(Y) <- Y$input_gene
    Y$rn <- NULL
    Y$input_gene <- NULL
    return(as.data.frame(Y))
})


### adapting a pretty wild function i wrote in the spatial registration scripts for Visium (spatial_HYP/code/10-blah/02-blah)
m<- hypb.mks
abas <- tmplist
names(abas) <- c("Yao23_subclasses","Yao23_VMHARC_supertypes","Yao23_subclasses_VMHARHclps")

spatregs <- mcmapply(s=abas,z=names(abas),SIMPLIFY=FALSE,mc.cores=3,FUN = function(s,z){
    # iterator for top number of genes to use for correlation
    j<-seq(10,250,10)
    i<-1
    outlist <- list()
    for (i in c(1:(length(j)+1))){
        if(i<length(j)){
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- j[i]
        }
        else{
            # no value for top_n --> all genes used
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- nrow(cortab)
        }
        # initialize cormat
        nclaba <- ncol(s)
        nclhyp <- ncol(m)
        cormat <- as.data.frame(matrix(nrow=nclhyp,ncol=nclaba))
        k <- 1
        for (k in c(1:nclhyp)){
            rownames(cormat)[k] <- colnames(m)[k]
            l <- 1
            for (l in c(1:nclaba)){
                colnames(cormat)[l] <- colnames(s)[l]
                
                curhyp <- colnames(m)[k]
                curaba <- colnames(s)[l]
                setorderv(cortab,curhyp,-1)
                
                cormat[k,l] <- cor(cortab[c(1:ngene),..curhyp],cortab[c(1:ngene),..curaba],use="pairwise.complete.obs",method = "spearman")
            }
        }
        plttab <- as.data.table(cormat,keep.rownames=T)
        setnames(plttab,"rn","sphyp_domain")
        outlist[[i]] <- plttab
        names(outlist)[i] <- paste0("ABA23_HYP_",z,"_top_",ngene,"_sphypdomainmks")
        rm(curj,plttab,cortab)
    }
    return(outlist)
})

saveRDS(spatregs,"processed-data/10_Spatial_Registration/01-NonSampleSpecific_cluster_reg_to_Yao23_mouseHyp_subclasses_and_VMHARCsupertypes.RDS")
```

repro inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats     graphics  grDevices utils     methods   base     

other attached packages:
[1] parallelly_1.38.0   orthogene_1.10.0    BiocParallel_1.38.0 magrittr_2.0.3     
[5] colorout_1.3-0.2    data.table_1.15.4   rlang_1.1.4        

loaded via a namespace (and not attached):
 [1] gtable_0.3.5              babelgene_22.9            xfun_0.45                
 [4] ggplot2_3.5.1             htmlwidgets_1.6.4         rstatix_0.7.2            
 [7] lattice_0.22-6            bitops_1.0-7              vctrs_0.6.5              
[10] tools_4.4.1               generics_0.1.3            yulab.utils_0.1.4        
[13] tibble_3.2.1              fansi_1.0.6               pkgconfig_2.0.3          
[16] Matrix_1.7-0              homologene_1.4.68.19.3.27 ggplotify_0.1.2          
[19] lifecycle_1.0.4           compiler_4.4.1            treeio_1.28.0            
[22] munsell_0.5.1             codetools_0.2-20          carData_3.0-5            
[25] ggtree_3.12.0             gprofiler2_0.2.3          ggfun_0.1.5              
[28] htmltools_0.5.8.1         RCurl_1.98-1.14           yaml_2.3.9               
[31] lazyeval_0.2.2            plotly_4.10.4             pillar_1.9.0             
[34] car_3.1-2                 ggpubr_0.6.0              tidyr_1.3.1              
[37] cachem_1.1.0              grr_0.9.5                 abind_1.4-5              
[40] nlme_3.1-165              tidyselect_1.2.1          aplot_0.2.3              
[43] digest_0.6.36             dplyr_1.1.4               purrr_1.0.2              
[46] rprojroot_2.0.4           fastmap_1.2.0             grid_4.4.1               
[49] here_1.0.1                colorspace_2.1-0          cli_3.6.3                
[52] patchwork_1.2.0           utf8_1.2.4                broom_1.0.6              
[55] ape_5.8                   withr_3.0.0               scales_1.3.0             
[58] backports_1.5.0           httr_1.4.7                rmarkdown_2.27           
[61] ggsignif_0.6.4            memoise_2.0.1             evaluate_0.24.0          
[64] knitr_1.48                viridisLite_0.4.2         gridGraphics_0.5-1       
[67] Rcpp_1.0.13               glue_1.7.0                tidytree_0.4.6           
[70] rstudioapi_0.16.0         jsonlite_1.8.8            R6_2.5.1                 
[73] fs_1.6.4                 
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-08-20
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────────────
 ! package      * version        date (UTC) lib source
   abind          1.4-5          2016-07-21 [1] CRAN (R 4.4.0)
   ape            5.8            2024-04-11 [1] CRAN (R 4.4.0)
   aplot          0.2.3          2024-06-17 [1] CRAN (R 4.4.0)
   babelgene      22.9           2022-09-29 [1] CRAN (R 4.4.0)
   backports      1.5.0          2024-05-23 [1] CRAN (R 4.4.0)
   BiocParallel * 1.38.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   bitops         1.0-7          2021-04-24 [1] CRAN (R 4.4.0)
   broom          1.0.6          2024-05-17 [1] CRAN (R 4.4.0)
   cachem         1.1.0          2024-05-16 [1] CRAN (R 4.4.0)
   car            3.1-2          2023-03-30 [1] CRAN (R 4.4.0)
   carData        3.0-5          2022-01-06 [1] CRAN (R 4.4.0)
 P cli            3.6.3          2024-06-21 [2] CRAN (R 4.4.0)
   codetools      0.2-20         2024-03-31 [1] CRAN (R 4.4.1)
   colorout     * 1.3-0.2        2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace     2.1-0          2023-01-23 [1] CRAN (R 4.4.0)
   data.table   * 1.15.4         2024-03-30 [2] CRAN (R 4.4.0)
   digest         0.6.36         2024-06-23 [1] CRAN (R 4.4.0)
   dplyr          1.1.4          2023-11-17 [1] CRAN (R 4.4.0)
   evaluate       0.24.0         2024-06-10 [1] CRAN (R 4.4.0)
   fansi          1.0.6          2023-12-08 [1] CRAN (R 4.4.0)
   fastmap        1.2.0          2024-05-15 [1] CRAN (R 4.4.0)
   fs             1.6.4          2024-04-25 [1] CRAN (R 4.4.0)
   generics       0.1.3          2022-07-05 [1] CRAN (R 4.4.0)
   ggfun          0.1.5          2024-05-28 [1] CRAN (R 4.4.0)
   ggplot2        3.5.1          2024-04-23 [1] CRAN (R 4.4.0)
   ggplotify      0.1.2          2023-08-09 [1] CRAN (R 4.4.0)
   ggpubr         0.6.0          2023-02-10 [1] CRAN (R 4.4.0)
   ggsignif       0.6.4          2022-10-13 [1] CRAN (R 4.4.0)
   ggtree         3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   glue           1.7.0          2024-01-09 [1] CRAN (R 4.4.0)
   gprofiler2     0.2.3          2024-02-23 [1] CRAN (R 4.4.0)
   gridGraphics   0.5-1          2020-12-13 [1] CRAN (R 4.4.0)
   grr            0.9.5          2016-08-26 [1] CRAN (R 4.4.0)
   gtable         0.3.5          2024-04-22 [1] CRAN (R 4.4.0)
   here           1.0.1          2020-12-13 [1] CRAN (R 4.4.0)
   homologene     1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.4.0)
   htmltools      0.5.8.1        2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets    1.6.4          2023-12-06 [1] CRAN (R 4.4.0)
   httr           1.4.7          2023-08-15 [1] CRAN (R 4.4.0)
   jsonlite       1.8.8          2023-12-04 [1] CRAN (R 4.4.0)
   knitr          1.48           2024-07-07 [1] CRAN (R 4.4.1)
   lattice        0.22-6         2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval       0.2.2          2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle      1.0.4          2023-11-07 [1] CRAN (R 4.4.0)
   magrittr     * 2.0.3          2022-03-30 [1] CRAN (R 4.4.0)
   Matrix         1.7-0          2024-04-26 [1] CRAN (R 4.4.1)
   memoise        2.0.1          2021-11-26 [1] CRAN (R 4.4.0)
   munsell        0.5.1          2024-04-01 [1] CRAN (R 4.4.0)
   nlme           3.1-165        2024-06-06 [1] CRAN (R 4.4.0)
   orthogene    * 1.10.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   parallelly   * 1.38.0         2024-07-27 [1] CRAN (R 4.4.0)
   patchwork      1.2.0          2024-01-08 [1] CRAN (R 4.4.0)
   pillar         1.9.0          2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig      2.0.3          2019-09-22 [1] CRAN (R 4.4.0)
   plotly         4.10.4         2024-01-13 [1] CRAN (R 4.4.0)
   purrr          1.0.2          2023-08-10 [1] CRAN (R 4.4.0)
   R6             2.5.1          2021-08-19 [1] CRAN (R 4.4.0)
   Rcpp           1.0.13         2024-07-17 [1] CRAN (R 4.4.0)
   RCurl          1.98-1.14      2024-01-09 [1] CRAN (R 4.4.0)
 P rlang        * 1.1.4          2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown      2.27           2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot      2.0.4          2023-11-05 [1] CRAN (R 4.4.0)
   rstatix        0.7.2          2023-02-01 [1] CRAN (R 4.4.0)
   rstudioapi     0.16.0         2024-03-24 [1] CRAN (R 4.4.0)
   scales         1.3.0          2023-11-28 [1] CRAN (R 4.4.0)
   sessioninfo    1.2.2          2021-12-06 [1] CRAN (R 4.4.0)
   tibble         3.2.1          2023-03-20 [1] CRAN (R 4.4.0)
   tidyr          1.3.1          2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect     1.2.1          2024-03-11 [1] CRAN (R 4.4.0)
   tidytree       0.4.6          2023-12-12 [1] CRAN (R 4.4.0)
   treeio         1.28.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   utf8           1.2.4          2023-10-22 [1] CRAN (R 4.4.0)
   vctrs          0.6.5          2023-12-01 [1] CRAN (R 4.4.0)
   viridisLite    0.4.2          2023-05-02 [1] CRAN (R 4.4.0)
   withr          3.0.0          2024-01-16 [1] CRAN (R 4.4.0)
   xfun           0.45           2024-06-16 [1] CRAN (R 4.4.0)
   yaml           2.3.9          2024-07-05 [1] CRAN (R 4.4.0)
   yulab.utils    0.1.4          2024-01-28 [1] CRAN (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────
