---
title: "01_Registration_to_Yao23_MsHyp_Subclasses"
output: html_document
date: "2024-08-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(SingleCellExperiment)
library(SpatialExperiment)
library(spatialLIBD)
require(colorout)
library(magrittr)
library(BiocParallel)
library(orthogene)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

## Load  the Yao spatial registration output from spatialLIBD::registration_wrapper applied to cell types at the subclass level in their hierarchical scheme, and the subclass level but with all of the VMH and ARC (there, ARH) types re-labeled singularly as VMH and ARH respectively
```{r}
yaoscl <- readRDS("../spatial_HYP/processed-data/10-Spatial Registration/02b-yaofilt_registrationwrapper.RDS")

yaoclps <- readRDS("../spatial_HYP/processed-data/10-Spatial Registration/02c-yaofilt_collapsedVMHandARHsubclasses_registrationwrapper.RDS")

### load the registration wrapper output generated the second time (i.e. after excluding sample specific clusters based on annotations made from the first marker detection run and cluster spot plot examinations)
hypb.retainedclusts.mk <- readRDS("processed-data/05_Banksy_M0lam0_res2_multisamp/03a-M0l0kg6_reproducibleClustersOnly_spLIBD_markers.RDS")
```

## get orthologous genes
```{r}
orthos <- as.data.table(unique(as.data.table(hypb.retainedclusts.mk$enrichment$ensembl)))
orthos <- convert_orthologs(gene_df = orthos,gene_input = "V1",gene_output = "columns",input_species = "human",output_species = "mouse",non121_strategy = "drop_both_species")

### subset the human data to the genes w mouse orthologs; rename t_stat clusters to the cluster ID for registration 
hypb.mks <- as.data.table(hypb.retainedclusts.mk$enrichment)
keepcols <- c("ensembl",grep(names(hypb.mks),pattern="^t_stat_",value=T))
hypb.mks <- hypb.mks[ensembl %in% orthos$input_gene,..keepcols]
    
setnames(hypb.mks,c(2:ncol(hypb.mks)),gsub(names(hypb.mks)[c(2:ncol(hypb.mks))],pattern="^t_stat_",replacement=""))


hypb.mks <- as.data.frame(hypb.mks)
rownames(hypb.mks) <- hypb.mks$ensembl
hypb.mks$ensembl <- NULL

## subset the Yao data to xenium genes with mouse orthologs, rename those t stat columns to clusters likewise
tmplist <- list(yaoscl,yaoclps)
tmplist <- lapply(X=tmplist,function(x){

    Y <- as.data.table(x$enrichment,keep.rownames=T)
    keepcols <- c("rn",grep(names(Y),pattern="^t_stat_",value=T))
    Y <- Y[,..keepcols]
    
    setnames(Y,c(2:ncol(Y)),gsub(names(Y)[c(2:ncol(Y))],pattern="^t_stat_",replacement=""))
    
    # subset to human genes w ensembl orthologs, rename rows to the human ortholog
    Y <- merge.data.table(Y,orthos,by.x="rn",by.y="ortholog_gene")

    Y <- as.data.frame(Y)
    rownames(Y) <- Y$input_gene
    Y$rn <- NULL
    Y$input_gene <- NULL
    return(as.data.frame(Y))
})


### adapting a pretty wild function i wrote in the spatial registration scripts for Visium (spatial_HYP/code/10-blah/02-blah)
m<- hypb.mks
abas <- tmplist
names(abas) <- c("Yao23_subclasses","Yao23_subclasses_VMHARHclps")

spatregs <- mapply(abas,names(abas),SIMPLIFY=FALSE,FUN = function(s,z){
    # iterator for top number of genes to use for correlation
    j<-c(25,50,75,100,150,200,250)
    i<-1
    outlist <- list()
    for (i in c(1:(length(j)+1))){
        if(i<length(j)){
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- j[i]
        }
        else{
            # no value for top_n --> all genes used
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- nrow(cortab)
        }
        # initialize cormat
        nclaba <- ncol(s)
        nclhyp <- ncol(m)
        cormat <- as.data.frame(matrix(nrow=nclhyp,ncol=nclaba))
        k <- 1
        for (k in c(1:nclhyp)){
            rownames(cormat)[k] <- colnames(m)[k]
            l <- 1
            for (l in c(1:nclaba)){
                colnames(cormat)[l] <- colnames(s)[l]
                
                curhyp <- colnames(m)[k]
                curaba <- colnames(s)[l]
                setorderv(cortab,curhyp,-1)
                
                cormat[k,l] <- cor(cortab[c(1:ngene),..curhyp],cortab[c(1:ngene),..curaba],use="pairwise.complete.obs",method = "spearman")
            }
        }
        plttab <- as.data.table(cormat,keep.rownames=T)
        setnames(plttab,"rn","sphyp_domain")
        outlist[[i]] <- plttab
        names(outlist)[i] <- paste0("ABA23_HYP_",z,"_top_",ngene,"_sphypdomainmks")
        rm(curj,plttab,cortab)
    }
    return(outlist)
})

saveRDS(spatregs,"processed-data/10_Spatial_Registration/01-NonSampleSpecific_cluster_reg_to_Yao23_mouseHyp_subclasses.RDS")

```

repro inf
```{r}
sessionInfo()
sessioninfo::session_info()
```

R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] orthogene_1.10.0               magrittr_2.0.3                
 [3] Biostrings_2.72.1              XVector_0.44.0                
 [5] BiocParallel_1.38.0            parallelly_1.38.0             
 [7] colorout_1.3-0.2               spatialLIBD_1.16.2            
 [9] SpatialFeatureExperiment_1.6.1 SpatialExperiment_1.14.0      
[11] SingleCellExperiment_1.26.0    SummarizedExperiment_1.34.0   
[13] Biobase_2.64.0                 GenomicRanges_1.56.1          
[15] GenomeInfoDb_1.40.1            IRanges_2.38.1                
[17] S4Vectors_0.42.1               BiocGenerics_0.50.0           
[19] MatrixGenerics_1.16.0          matrixStats_1.3.0             
[21] gridExtra_2.3                  ggplot2_3.5.1                 
[23] data.table_1.15.4              rlang_1.1.4                   

loaded via a namespace (and not attached):
  [1] fs_1.6.4                  bitops_1.0-7              sf_1.0-16                
  [4] EBImage_4.46.0            httr_1.4.7                RColorBrewer_1.1-3       
  [7] doParallel_1.0.17         tools_4.4.1               backports_1.5.0          
 [10] utf8_1.2.4                R6_2.5.1                  DT_0.33                  
 [13] HDF5Array_1.32.0          lazyeval_0.2.2            rhdf5filters_1.16.0      
 [16] withr_3.0.0               sp_2.1-4                  cli_3.6.3                
 [19] sass_0.4.9                proxy_0.4-27              yulab.utils_0.1.4        
 [22] Rsamtools_2.20.0          R.utils_2.12.3            scater_1.32.0            
 [25] sessioninfo_1.2.2         attempt_0.3.1             maps_3.4.2               
 [28] limma_3.60.3              rstudioapi_0.16.0         RSQLite_2.3.7            
 [31] gridGraphics_0.5-1        generics_0.1.3            BiocIO_1.14.0            
 [34] car_3.1-2                 spdep_1.3-5               dplyr_1.1.4              
 [37] homologene_1.4.68.19.3.27 Matrix_1.7-0              ggbeeswarm_0.7.2         
 [40] fansi_1.0.6               abind_1.4-5               R.methodsS3_1.8.2        
 [43] terra_1.7-78              lifecycle_1.0.4           yaml_2.3.9               
 [46] edgeR_4.2.0               carData_3.0-5             rhdf5_2.48.0             
 [49] SparseArray_1.4.8         BiocFileCache_2.12.0      paletteer_1.6.0          
 [52] grid_4.4.1                blob_1.2.4                promises_1.3.0           
 [55] dqrng_0.4.1               ExperimentHub_2.12.0      crayon_1.5.3             
 [58] lattice_0.22-6            beachmat_2.20.0           cowplot_1.1.3            
 [61] KEGGREST_1.44.1           magick_2.8.3              zeallot_0.1.0            
 [64] pillar_1.9.0              knitr_1.48                rjson_0.2.21             
 [67] boot_1.3-30               codetools_0.2-20          wk_0.9.2                 
 [70] glue_1.7.0                ggfun_0.1.5               treeio_1.28.0            
 [73] vctrs_0.6.5               png_0.1-8                 spam_2.10-0              
 [76] gtable_0.3.5              rematch2_2.1.2            cachem_1.1.0             
 [79] xfun_0.45                 S4Arrays_1.4.1            mime_0.12                
 [82] DropletUtils_1.24.0       sfheaders_0.4.4           iterators_1.0.14         
 [85] units_0.8-5               fields_16.2               statmod_1.5.0            
 [88] nlme_3.1-165              ggtree_3.12.0             bit64_4.0.5              
 [91] filelock_1.0.3            rprojroot_2.0.4           bslib_0.7.0              
 [94] irlba_2.3.5.1             vipor_0.4.7               KernSmooth_2.23-24       
 [97] colorspace_2.1-0          spData_2.3.1              DBI_1.2.3                
[100] tidyselect_1.2.1          bit_4.0.5                 compiler_4.4.1           
[103] curl_5.2.1                BiocNeighbors_1.22.0      DelayedArray_0.30.1      
[106] plotly_4.10.4             rtracklayer_1.64.0        scales_1.3.0             
[109] classInt_0.4-10           rappdirs_0.3.3            tiff_0.1-12              
[112] digest_0.6.36             fftwtools_0.9-11          rmarkdown_2.27           
[115] benchmarkmeData_1.0.4     htmltools_0.5.8.1         pkgconfig_2.0.3          
[118] jpeg_0.1-10               sparseMatrixStats_1.16.0  dbplyr_2.5.0             
[121] fastmap_1.2.0             htmlwidgets_1.6.4         UCSC.utils_1.0.0         
[124] shiny_1.8.1.1             DelayedMatrixStats_1.26.0 jquerylib_0.1.4          
[127] jsonlite_1.8.8            config_0.3.2              R.oo_1.26.0              
[130] BiocSingular_1.20.0       RCurl_1.98-1.14           ggplotify_0.1.2          
[133] scuttle_1.14.0            GenomeInfoDbData_1.2.12   s2_1.1.6                 
[136] dotCall64_1.1-1           patchwork_1.2.0           Rhdf5lib_1.26.0          
[139] munsell_0.5.1             Rcpp_1.0.13               ape_5.8                  
[142] babelgene_22.9            viridis_0.6.5             zlibbioc_1.50.0          
[145] AnnotationHub_3.12.0      parallel_4.4.1            ggrepel_0.9.5            
[148] deldir_2.0-4              locfit_1.5-9.10           ggpubr_0.6.0             
[151] ggsignif_0.6.4            ScaledMatrix_1.12.0       gprofiler2_0.2.3         
[154] BiocVersion_3.19.1        XML_3.99-0.17             evaluate_0.24.0          
[157] golem_0.4.1               BiocManager_1.30.23       foreach_1.5.2            
[160] httpuv_1.6.15             grr_0.9.5                 tidyr_1.3.1              
[163] purrr_1.0.2               benchmarkme_1.0.8         rsvd_1.0.5               
[166] broom_1.0.6               xtable_1.8-4              restfulr_0.0.15          
[169] e1071_1.7-14              tidytree_0.4.6            rstatix_0.7.2            
[172] later_1.3.2               viridisLite_0.4.2         class_7.3-22             
[175] tibble_3.2.1              aplot_0.2.3               memoise_2.0.1            
[178] beeswarm_0.4.0            AnnotationDbi_1.66.0      GenomicAlignments_1.40.0 
[181] shinyWidgets_0.8.6        here_1.0.1               
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-08-04
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────────
 ! package                  * version        date (UTC) lib source
   abind                      1.4-5          2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi              1.66.0         2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub              3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   ape                        5.8            2024-04-11 [1] CRAN (R 4.4.0)
   aplot                      0.2.3          2024-06-17 [1] CRAN (R 4.4.0)
   attempt                    0.3.1          2020-05-03 [1] CRAN (R 4.4.0)
   babelgene                  22.9           2022-09-29 [1] CRAN (R 4.4.0)
   backports                  1.5.0          2024-05-23 [1] CRAN (R 4.4.0)
   beachmat                   2.20.0         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm                   0.4.0          2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme                1.0.8          2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData            1.0.4          2020-04-23 [1] CRAN (R 4.4.0)
   Biobase                  * 2.64.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache              2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics             * 0.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                     1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager                1.30.23        2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors              1.22.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel             * 1.38.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocSingular               1.20.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion                3.19.1         2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings               * 2.72.1         2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                        4.0.5          2022-11-15 [1] CRAN (R 4.4.0)
   bit64                      4.0.5          2020-08-30 [1] CRAN (R 4.4.0)
   bitops                     1.0-7          2021-04-24 [1] CRAN (R 4.4.0)
   blob                       1.2.4          2023-03-17 [1] CRAN (R 4.4.0)
   boot                       1.3-30         2024-02-26 [1] CRAN (R 4.4.1)
   broom                      1.0.6          2024-05-17 [1] CRAN (R 4.4.0)
   bslib                      0.7.0          2024-03-29 [1] CRAN (R 4.4.0)
   cachem                     1.1.0          2024-05-16 [1] CRAN (R 4.4.0)
   car                        3.1-2          2023-03-30 [1] CRAN (R 4.4.0)
   carData                    3.0-5          2022-01-06 [1] CRAN (R 4.4.0)
   class                      7.3-22         2023-05-03 [1] CRAN (R 4.4.1)
   classInt                   0.4-10         2023-09-05 [1] CRAN (R 4.4.0)
 P cli                        3.6.3          2024-06-21 [2] CRAN (R 4.4.0)
   codetools                  0.2-20         2024-03-31 [1] CRAN (R 4.4.1)
   colorout                 * 1.3-0.2        2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace                 2.1-0          2023-01-23 [1] CRAN (R 4.4.0)
   config                     0.3.2          2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                    1.1.3          2024-01-22 [1] CRAN (R 4.4.0)
   crayon                     1.5.3          2024-06-20 [1] CRAN (R 4.4.0)
   curl                       5.2.1          2024-03-01 [1] CRAN (R 4.4.0)
   data.table               * 1.15.4         2024-03-30 [2] CRAN (R 4.4.0)
   DBI                        1.2.3          2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                     2.5.0          2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray               0.30.1         2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats         1.26.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   deldir                     2.0-4          2024-02-28 [1] CRAN (R 4.4.0)
   digest                     0.6.36         2024-06-23 [1] CRAN (R 4.4.0)
   doParallel                 1.0.17         2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64                  1.1-1          2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                      1.1.4          2023-11-17 [1] CRAN (R 4.4.0)
   dqrng                      0.4.1          2024-05-28 [1] CRAN (R 4.4.0)
   DropletUtils               1.24.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   DT                         0.33           2024-04-04 [1] CRAN (R 4.4.0)
   e1071                      1.7-14         2023-12-06 [1] CRAN (R 4.4.0)
   EBImage                    4.46.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   edgeR                      4.2.0          2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate                   0.24.0         2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub              2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fansi                      1.0.6          2023-12-08 [1] CRAN (R 4.4.0)
   fastmap                    1.2.0          2024-05-15 [1] CRAN (R 4.4.0)
   fftwtools                  0.9-11         2021-03-01 [1] CRAN (R 4.4.0)
   fields                     16.2           2024-06-27 [1] CRAN (R 4.4.0)
   filelock                   1.0.3          2023-12-11 [1] CRAN (R 4.4.0)
   foreach                    1.5.2          2022-02-02 [1] CRAN (R 4.4.0)
   fs                         1.6.4          2024-04-25 [1] CRAN (R 4.4.0)
   generics                   0.1.3          2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb             * 1.40.1         2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData           1.2.12         2024-05-01 [1] Bioconductor
   GenomicAlignments          1.40.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges            * 1.56.1         2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm                 0.7.2          2023-04-29 [1] CRAN (R 4.4.0)
   ggfun                      0.1.5          2024-05-28 [1] CRAN (R 4.4.0)
   ggplot2                  * 3.5.1          2024-04-23 [1] CRAN (R 4.4.0)
   ggplotify                  0.1.2          2023-08-09 [1] CRAN (R 4.4.0)
   ggpubr                     0.6.0          2023-02-10 [1] CRAN (R 4.4.0)
   ggrepel                    0.9.5          2024-01-10 [1] CRAN (R 4.4.0)
   ggsignif                   0.6.4          2022-10-13 [1] CRAN (R 4.4.0)
   ggtree                     3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   glue                       1.7.0          2024-01-09 [1] CRAN (R 4.4.0)
   golem                      0.4.1          2023-06-05 [1] CRAN (R 4.4.0)
   gprofiler2                 0.2.3          2024-02-23 [1] CRAN (R 4.4.0)
   gridExtra                * 2.3            2017-09-09 [1] CRAN (R 4.4.0)
   gridGraphics               0.5-1          2020-12-13 [1] CRAN (R 4.4.0)
   grr                        0.9.5          2016-08-26 [1] CRAN (R 4.4.0)
   gtable                     0.3.5          2024-04-22 [1] CRAN (R 4.4.0)
   HDF5Array                  1.32.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   here                       1.0.1          2020-12-13 [1] CRAN (R 4.4.0)
   homologene                 1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.4.0)
   htmltools                  0.5.8.1        2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets                1.6.4          2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                     1.6.15         2024-03-26 [1] CRAN (R 4.4.0)
   httr                       1.4.7          2023-08-15 [1] CRAN (R 4.4.0)
   IRanges                  * 2.38.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                      2.3.5.1        2022-10-03 [1] CRAN (R 4.4.0)
   iterators                  1.0.14         2022-02-05 [1] CRAN (R 4.4.0)
   jpeg                       0.1-10         2022-11-29 [1] CRAN (R 4.4.0)
   jquerylib                  0.1.4          2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite                   1.8.8          2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST                   1.44.1         2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   KernSmooth                 2.23-24        2024-05-17 [1] CRAN (R 4.4.1)
   knitr                      1.48           2024-07-07 [1] CRAN (R 4.4.1)
   later                      1.3.2          2023-12-06 [1] CRAN (R 4.4.0)
   lattice                    0.22-6         2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval                   0.2.2          2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle                  1.0.4          2023-11-07 [1] CRAN (R 4.4.0)
   limma                      3.60.3         2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                     1.5-9.10       2024-06-24 [1] CRAN (R 4.4.0)
   magick                     2.8.3          2024-02-18 [1] CRAN (R 4.4.0)
   magrittr                 * 2.0.3          2022-03-30 [1] CRAN (R 4.4.0)
   maps                       3.4.2          2023-12-15 [1] CRAN (R 4.4.0)
   Matrix                     1.7-0          2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics           * 1.16.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats              * 1.3.0          2024-04-11 [1] CRAN (R 4.4.0)
   memoise                    2.0.1          2021-11-26 [1] CRAN (R 4.4.0)
   mime                       0.12           2021-09-28 [1] CRAN (R 4.4.0)
   munsell                    0.5.1          2024-04-01 [1] CRAN (R 4.4.0)
   nlme                       3.1-165        2024-06-06 [1] CRAN (R 4.4.0)
   orthogene                * 1.10.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   paletteer                  1.6.0          2024-01-21 [1] CRAN (R 4.4.0)
   parallelly               * 1.38.0         2024-07-27 [1] CRAN (R 4.4.0)
   patchwork                  1.2.0          2024-01-08 [1] CRAN (R 4.4.0)
   pillar                     1.9.0          2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig                  2.0.3          2019-09-22 [1] CRAN (R 4.4.0)
   plotly                     4.10.4         2024-01-13 [1] CRAN (R 4.4.0)
   png                        0.1-8          2022-11-29 [1] CRAN (R 4.4.0)
   promises                   1.3.0          2024-04-05 [1] CRAN (R 4.4.0)
   proxy                      0.4-27         2022-06-09 [1] CRAN (R 4.4.0)
   purrr                      1.0.2          2023-08-10 [1] CRAN (R 4.4.0)
   R.methodsS3                1.8.2          2022-06-13 [1] CRAN (R 4.4.0)
   R.oo                       1.26.0         2024-01-24 [1] CRAN (R 4.4.0)
   R.utils                    2.12.3         2023-11-18 [1] CRAN (R 4.4.0)
   R6                         2.5.1          2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs                   0.3.3          2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer               1.1-3          2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                       1.0.13         2024-07-17 [1] CRAN (R 4.4.0)
   RCurl                      1.98-1.14      2024-01-09 [1] CRAN (R 4.4.0)
   rematch2                   2.1.2          2020-05-01 [1] CRAN (R 4.4.0)
   restfulr                   0.0.15         2022-06-16 [1] CRAN (R 4.4.0)
   rhdf5                      2.48.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rhdf5filters               1.16.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Rhdf5lib                   1.26.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rjson                      0.2.21         2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                    * 1.1.4          2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown                  2.27           2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot                  2.0.4          2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools                  2.20.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                    2.3.7          2024-05-27 [1] CRAN (R 4.4.0)
   rstatix                    0.7.2          2023-02-01 [1] CRAN (R 4.4.0)
   rstudioapi                 0.16.0         2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                       1.0.5          2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer                1.64.0         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   s2                         1.1.6          2023-12-19 [1] CRAN (R 4.4.0)
   S4Arrays                   1.4.1          2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors                * 0.42.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                       0.4.9          2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix               1.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                     1.3.0          2023-11-28 [1] CRAN (R 4.4.0)
   scater                     1.32.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scuttle                    1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo                1.2.2          2021-12-06 [1] CRAN (R 4.4.0)
   sf                         1.0-16         2024-03-24 [1] CRAN (R 4.4.0)
   sfheaders                  0.4.4          2024-01-17 [1] CRAN (R 4.4.0)
   shiny                      1.8.1.1        2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets               0.8.6          2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment     * 1.26.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sp                         2.1-4          2024-04-30 [1] CRAN (R 4.4.0)
   spam                       2.10-0         2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray                1.4.8          2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats          1.16.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment        * 1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialFeatureExperiment * 1.6.1          2024-05-15 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD              * 1.16.2         2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   spData                     2.3.1          2024-05-31 [1] CRAN (R 4.4.0)
   spdep                      1.3-5          2024-06-10 [1] CRAN (R 4.4.0)
   statmod                    1.5.0          2023-01-06 [1] CRAN (R 4.4.0)
   SummarizedExperiment     * 1.34.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   terra                      1.7-78         2024-05-22 [1] CRAN (R 4.4.0)
   tibble                     3.2.1          2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                      1.3.1          2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect                 1.2.1          2024-03-11 [1] CRAN (R 4.4.0)
   tidytree                   0.4.6          2023-12-12 [1] CRAN (R 4.4.0)
   tiff                       0.1-12         2023-11-28 [1] CRAN (R 4.4.0)
   treeio                     1.28.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   UCSC.utils                 1.0.0          2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   units                      0.8-5          2023-11-28 [1] CRAN (R 4.4.0)
   utf8                       1.2.4          2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                      0.6.5          2023-12-01 [1] CRAN (R 4.4.0)
   vipor                      0.4.7          2023-12-18 [1] CRAN (R 4.4.0)
   viridis                    0.6.5          2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite                0.4.2          2023-05-02 [1] CRAN (R 4.4.0)
   withr                      3.0.0          2024-01-16 [1] CRAN (R 4.4.0)
   wk                         0.9.2          2024-07-09 [1] CRAN (R 4.4.0)
   xfun                       0.45           2024-06-16 [1] CRAN (R 4.4.0)
   XML                        3.99-0.17      2024-06-25 [1] CRAN (R 4.3.3)
   xtable                     1.8-4          2019-04-21 [1] CRAN (R 4.4.0)
   XVector                  * 0.44.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                       2.3.9          2024-07-05 [1] CRAN (R 4.4.0)
   yulab.utils                0.1.4          2024-01-28 [1] CRAN (R 4.4.0)
   zeallot                    0.1.0          2018-01-28 [1] CRAN (R 4.4.0)
   zlibbioc                   1.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────────
