---
title: "02-Spatial Registration vs Yao Subclasses w 50plus cells in ms HYP dissections"
author: "Bernard Mulvey"
date: "2024-08-01"
output: html_document
---
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(SingleCellExperiment)
library(SpatialExperiment)
library(spatialLIBD)
require(colorout)
library(magrittr)
library(BiocParallel)
library(orthogene)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc
library(parallel)
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
```

load SPE, load Yao data in SCE format
```{r}
hyp2 <- readRDS("processed-data/03-QC_filters/hypN10_umi210_gene126_chrm50_spotsweeped_lognorm_rotsNmirrors_072224.RDS")

bscl <- fread("processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl=="X4",cl:="ARC.1"]
bscl[cl=="X6",cl:="ARC.2"]
bscl[cl=="X7",cl:="VMH.1"]
bscl[cl=="X12",cl:="VMH.2"]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=c("VMH.1","VMH.2","ARC.1","ARC.2",paste0("X",c(1:3,5,8:11,13:15))))

### load yao SCE
yao <- readRDS("raw-data/ABA_Whole_Mouse_Brain_Yao23_HYP/bothdatasets_raw_asSCE.RDS")


### check metadata for spatialLIBD compatability (no leading digits)
unique(yao$subclass)
unique(yao$library_label.x)
### k, we should be good.
```

spatial registration wrapper: hyp2
```{r}
sphyp.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphyp.reg,"processed-data/10-Spatial Registration/02a-spHYP_registrationwrapper.RDS")

```

registration wrapper: yao data, min 50 cells (pre-filter to avoid dropped levels errors)
```{r}
# get cell types with a dataset-wide total > 50
cellfilt <- as.data.table(colData(yao),keep.rownames=T)
cellfilt1 <- cellfilt[,.N,by="subclass"]
cellfilt1 <- cellfilt1[N>=50]

cellfilt2 <- cellfilt[subclass %in% cellfilt1$subclass][,.N,by=c("library_label.x","subclass")]

## determine the max number of samples possible
length(unique(cellfilt2$library_label.x))
## 47 samples here, so get those where at >=20 samples have >= ? cells
cellfilt3 <- cellfilt2[N>=20,.N,by="subclass"][subclass %in% grep(subclass,pattern="ARH|VMH",value=T)]
### 20 cells would drop an ARC class (17 samples). try 15 cells.
cellfilt3 <- cellfilt2[N>=15,.N,by="subclass"][subclass %in% grep(subclass,pattern="ARH|VMH",value=T)]
### that gets us at least 22 samples for all of the VMH, ARC subclasses (6 total), which'll do. 53 clusters total
cellfilt3 <- cellfilt2[N>=15,.N,by="subclass"][N>15]

# colnames in the SCE to keep:
cellfilt <- cellfilt[subclass %in% cellfilt3$subclass,rn]
# subset SCE
yaofilt <- yao[,cellfilt]
# sanity check, 53 clusters expected:
length(unique(yaofilt$subclass))
# nice


## clean up
rm(cellfilt,cellfilt1,cellfilt2,cellfilt3,sphyp.reg,yao)
gc(full=T)
```

Yao filt spatialwrapper
```{r}
## only one column of gene metadata here--gene symbol
yaofilt.spwrap <- registration_wrapper(yaofilt,var_registration = "subclass",var_sample_id = "library_label.x",gene_ensembl = "gene_symbol",gene_name = "gene_symbol",min_ncells = 15)

saveRDS(yaofilt.spwrap,"processed-data/10-Spatial Registration/02b-yaofilt_registrationwrapper.RDS")

# RAM free-up
rm(yaofilt.spwrap)
gc(full=T)
```

## added 08-19-2024: Also run registration_wrapper at the level of supertypes from VMH and ARC on the background of the same overall dataset. filter out celltypes to be dropped to prevent registration_wrapper errors first
```{r}
# get cell types with a dataset-wide total > 50
cellfilt <- as.data.table(colData(yao),keep.rownames=T)
cellfilt1 <- cellfilt[,.N,by="supertype"]
cellfilt1 <- cellfilt1[N>=50]

cellfilt2 <- cellfilt[supertype %in% cellfilt1$supertype][,.N,by=c("library_label.x","supertype")]

## determine the max number of samples possible
length(unique(cellfilt2$library_label.x))
## 47 samples here, so get those where at >=10 samples have >= 10 cells (15 or 20 too stringent and drops some of these). the first N argument here is number of CELLS in a given library; the second is the number of libraries passing that.
cellfilt3 <- cellfilt2[N>=10,.N,by="supertype"][N>10]

## this gives us 173 supertypes. (17 are VMH/ARH of 22 supertypes that exist in their annotations.)



# colnames in the SCE to keep:
cellfilt <- cellfilt[supertype %in% cellfilt3$supertype,rn]
# subset SCE
yaofiltsuper <- yao[,cellfilt]
# sanity check, 173 clusters expected:
length(unique(yaofiltsuper$supertype))
# nice

## since we want to test markers of each VMH / ARH cluster against everything else, change supertype for NON-vmh/arh to "other" so that registration_wrapper doesn't think we care about those 150 clusters
yaofiltsuper$supertype <- as.character(yaofiltsuper$supertype)
colData(yaofiltsuper)[,"supertype"][!(yaofiltsuper$supertype %in% unique(grep(yaofiltsuper$supertype,pattern="VMH|ARH",value=T)))] <- "other"

# sanity check
# unique(yaofiltsuper$supertype) -> "other" and 17 VMH/ARH clusters

## clean up
rm(cellfilt,cellfilt1,cellfilt2,cellfilt3,yao)
gc(full=T)
```

```{r}
super1vall <- registration_wrapper(yaofiltsuper,var_registration = "supertype",var_sample_id = "library_label.x",gene_ensembl = "gene_symbol",gene_name = "gene_symbol",min_ncells = 10)
# 
# super1vall.m <- melt(as.data.table(super1vall[["enrichment"]]),id.vars=c("gene","ensembl"))
# super1vall.m[,ensembl:=NULL] # no ensembl ids with the source data
# 
# super1vall.m[,stat:=gsub(variable,pattern="^(.*)_x.*$|_other$",replacement="\\1")]
# super1vall.m[,clus:=gsub(variable,pattern="^.*_(x...._.*$)$",replacement="\\1")]
# super1vall.m[,clus:=gsub(clus,pattern="^(.*other.*)$",replacement="other")]
# 
# i<-1
# for (s in unique(super1vall.m$stat)){
#    tmp <- super1vall.m[stat==s]
#    setnames(tmp,"value",s)
#    tmp[,variable:=NULL]
#    tmp[,stat:=NULL]
#    if(s==unique(super1vall.m$stat)[1]){supervall.out <- copy(tmp)}
#    else{supervall.out <- merge.data.table(supervall.out,tmp,by=c("gene","clus"))}
# }

saveRDS(super1vall,"processed-data/10-Spatial Registration/02b-Yao23-VMHARH_supercluster_regwrapper.RDS")
```

### once more, but with VMH subclasses and ARC subclasses collapsed down to a single "VMH" and "ARC"
```{r}
tmpcd <- as.data.table(colData(yaofilt),keep.rownames=T)
tmpcd[subclass %in% grep(subclass,pattern="VMH",value=T),subclass:="VMHall"]
# careful here--there's also a subclass not labeled ARH but that has a gene symbol containing arh in the name!
unique(grep(tmpcd$subclass,pattern="ARH",value=T))
tmpcd[subclass %in% grep(subclass,pattern="x075|x120|x121|x114",value=T),subclass:="ARHall"]

tmpcd <- DataFrame(tmpcd)
rownames(tmpcd) <- tmpcd$rn
tmpcd$rn <- NULL
tmpcd <- tmpcd[colnames(yaofilt),]
colData(yaofilt) <- tmpcd
yaofilt$subclass <- factor(yaofilt$subclass)

# should now be 49 subclasses (53-4 ARC - 2 VMH + 1 joint vmh + 1 joint arc)
levels(yaofilt$subclass) # good

### now re-run spatial registration..
yaofilt.collapsed.spwrap <- registration_wrapper(yaofilt,var_registration = "subclass",var_sample_id = "library_label.x",gene_ensembl = "gene_symbol",gene_name = "gene_symbol",min_ncells = 15)


saveRDS(yaofilt.collapsed.spwrap,"processed-data/10-Spatial Registration/02c-yaofilt_collapsedVMHandARHsubclasses_registrationwrapper.RDS")

```

### for posterity, using the k20, k31, and k15 collapsed spatial domains:
    arc20= X1, X4
    vmh20= X2, X3
    arc31= X2, X4, X11, X22
    vmh31= X6, X21, X25, X26, X31

```{r}
rm(list=ls())
gc(full=T)

#
hyp2 <- readRDS("processed-data/03-QC_filters/hypN10_umi210_gene126_chrm50_spotsweeped_lognorm_rotsNmirrors_072224.RDS")

# k20
bscl <- fread("processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k20_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl=="X2",cl:="VMH20.1"]
bscl[cl=="X3",cl:="VMH20.2"]
bscl[cl=="X1",cl:="ARC20.1"]
bscl[cl=="X4",cl:="ARC20.2"]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=c("VMH20.1","VMH20.2","ARC20.1","ARC20.2",paste0("X",c(5:20))))


sphyp20.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphyp20.reg,"processed-data/10-Spatial Registration/02d-sphypk20_regwrapper.RDS")
```

# k31:
# arc31=c("X2","X4","X11","X22"),
# vmh31=c("X6","X21","X25","X26","X31")
```{r}
bscl <- fread("processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k31_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl=="X6",cl:="VMH31.1"]
bscl[cl=="X21",cl:="VMH31.2"]
bscl[cl=="X25",cl:="VMH31.3"]
bscl[cl=="X26",cl:="VMH31.4"]
bscl[cl=="X31",cl:="VMH31.5"]

bscl[cl=="X2",cl:="ARC31.1"]
bscl[cl=="X4",cl:="ARC31.2"]
bscl[cl=="X11",cl:="ARC31.3"]
bscl[cl=="X22",cl:="ARC31.4"]

bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

others <- as.numeric(gsub(unique(bscl$cl),pattern="^X(.*)",replacement="\\1"))
others <- others[!is.na(others)]

colLabels(hyp2) <- factor(bscl$cl,levels=c("VMH31.1","VMH31.2","VMH31.3","VMH31.4","VMH31.5","ARC31.1","ARC31.2","ARC31.3","ARC31.4",paste0("X",others)))

sphyp31.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphyp31.reg,"processed-data/10-Spatial Registration/02e-sphypk31_regwrapper.RDS")
```

# collapsed
```{r}
bscl <- fread("processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl %in% c("X4","X6"),cl:="ARC"]
bscl[cl %in% c("X7","X12"),cl:="VMH"]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=c("ARC","VMH",paste0("X",c(1:3,5,8:11,13:15))))

sphypcollapse.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphypcollapse.reg,"processed-data/10-Spatial Registration/02f-sphypcollapse_regwrapper.RDS")
```

Wipe env and load the outputs from spatialLIBD::registration_wrapper
```{r}
rm(list=ls())
gc(full=T)


abasub <- readRDS("processed-data/10-Spatial Registration/02b-yaofilt_registrationwrapper.RDS")

abasub.singlevmharc <- readRDS("processed-data/10-Spatial Registration/02c-yaofilt_collapsedVMHandARHsubclasses_registrationwrapper.RDS")

abasuper <- readRDS("processed-data/10-Spatial Registration/02b-Yao23-VMHARH_supercluster_regwrapper.RDS")

sphyp <- readRDS("processed-data/10-Spatial Registration/02a-spHYP_registrationwrapper.RDS")

sphyp20 <- readRDS("processed-data/10-Spatial Registration/02d-sphypk20_regwrapper.RDS")

sphyp31 <- readRDS("processed-data/10-Spatial Registration/02e-sphypk31_regwrapper.RDS")

sphypclps <- readRDS("processed-data/10-Spatial Registration/02f-sphypcollapse_regwrapper.RDS")
# 

orthos <- as.data.table(unique(as.data.table(sphyp$enrichment$ensembl)))
orthos <- convert_orthologs(gene_df = orthos,gene_input = "V1",gene_output = "columns",input_species = "human",output_species = "mouse",non121_strategy = "drop_both_species")

tmplist2 <- list(sphyp,sphyp20,sphyp31,sphypclps)
tmplist2 <- lapply(X=tmplist2,FUN=function(x){
    
    Y <- as.data.table(x$enrichment)
    keepcols <- c("ensembl",grep(names(Y),pattern="^t_stat_",value=T))
    Y <- Y[,..keepcols]
    
    setnames(Y,c(2:ncol(Y)),gsub(names(Y)[c(2:ncol(Y))],pattern="^t_stat_",replacement=""))


    Y <- as.data.frame(Y)
    rownames(Y) <- Y$ensembl
    Y$ensembl <- NULL
    return(as.data.frame(Y))
})

tmplist <- list(abasub,abasub.singlevmharc,abasuper)
tmplist <- lapply(X=tmplist,function(x){

    Y <- as.data.table(x$enrichment,keep.rownames=T)
    keepcols <- c("rn",grep(names(Y),pattern="^t_stat_",value=T))
    Y <- Y[,..keepcols]
    
    setnames(Y,c(2:ncol(Y)),gsub(names(Y)[c(2:ncol(Y))],pattern="^t_stat_",replacement=""))
    
    # get human ensembl orthologs
    Y <- merge.data.table(Y,orthos,by.x="rn",by.y="ortholog_gene")

    Y <- as.data.frame(Y)
    rownames(Y) <- Y$input_gene
    Y$rn <- NULL
    Y$input_gene <- NULL
    return(as.data.frame(Y))
})


abas <- list(tmplist[[1]],tmplist[[2]],tmplist[[3]],tmplist[[1]],tmplist[[2]],tmplist[[3]],tmplist[[1]],tmplist[[2]],tmplist[[3]],tmplist[[1]],tmplist[[2]],tmplist[[3]])
names(abas) <- rep(c("subclass","subclass_collapsed","VMHARC_supertypes"),4)

hyps <- list(tmplist2[[1]],tmplist2[[1]],tmplist2[[1]],tmplist2[[2]],tmplist2[[2]],tmplist2[[2]],tmplist2[[3]],tmplist2[[3]],tmplist2[[3]],tmplist2[[4]],tmplist2[[4]],tmplist2[[4]])
names(hyps) <- c(rep("sphypk15",3),rep("sphypk20",3),rep("sphypk31",3),rep("sphypk15_collapsed",3))
```

spatialLIBD layer_stat_cor won't cooperate so just do this manually
### correction 02/27/24: spearman, not pearson correlation
```{r}
outfilesuff <- as.list(paste0(names(abas),"_to_",names(hyps)))

spatregs <- mcmapply(m=hyps,s=abas,z=outfilesuff,SIMPLIFY = F,mc.cores=8,FUN = function(m,s,z){
    j<-c(25,50,75,100,250,500,750,1000)
    i<-1
    outlist <- list()
    for (i in c(1:9)){
        if(i<=8){
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- j[i]
        }
        else{
            # no value for top_n --> all genes used
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- nrow(cortab)
        }
        # initialize cormat
        nclaba <- ncol(s)
        nclhyp <- ncol(m)
        cormat <- as.data.frame(matrix(nrow=nclhyp,ncol=nclaba))
        k <- 1
        for (k in c(1:nclhyp)){
            rownames(cormat)[k] <- colnames(m)[k]
            l <- 1
            for (l in c(1:nclaba)){
                colnames(cormat)[l] <- colnames(s)[l]
                
                curhyp <- colnames(m)[k]
                curaba <- colnames(s)[l]
                setorderv(cortab,curhyp,-1)
                
                cormat[k,l] <- cor(cortab[c(1:ngene),..curhyp],cortab[c(1:ngene),..curaba],use="pairwise.complete.obs",method = "spearman")
            }
        }
        plttab <- as.data.table(cormat,keep.rownames=T)
        setnames(plttab,"rn","sphyp_domain")
        outlist[[i]] <- plttab
        names(outlist)[i] <- paste0("ABA23_HYP_",z,"_top_",ngene,"_sphypdomainmks")
        rm(curj,plttab,cortab)
    }
    return(outlist)
})

saveRDS(spatregs,"processed-data/10-Spatial Registration/02g-sphyp-k15-20-31-15clps_ABAms-47subclass-47withVMHARCclps-VMHARCsupertypes.RDS")
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] parallelly_1.38.0           orthogene_1.10.0           
 [3] BiocParallel_1.38.0         magrittr_2.0.3             
 [5] colorout_1.3-0.2            spatialLIBD_1.16.2         
 [7] SpatialExperiment_1.14.0    SingleCellExperiment_1.26.0
 [9] SummarizedExperiment_1.34.0 Biobase_2.64.0             
[11] GenomicRanges_1.56.1        MatrixGenerics_1.16.0      
[13] matrixStats_1.3.0           Biostrings_2.72.1          
[15] GenomeInfoDb_1.40.1         XVector_0.44.0             
[17] IRanges_2.38.1              S4Vectors_0.42.1           
[19] BiocGenerics_0.50.0         data.table_1.15.4          
[21] rlang_1.1.4                

loaded via a namespace (and not attached):
  [1] fs_1.6.4                  bitops_1.0-7              httr_1.4.7               
  [4] RColorBrewer_1.1-3        doParallel_1.0.17         Rgraphviz_2.48.0         
  [7] numDeriv_2016.8-1.1       tools_4.4.1               backports_1.5.0          
 [10] metafor_4.6-0             utf8_1.2.4                R6_2.5.1                 
 [13] DT_0.33                   lazyeval_0.2.2            withr_3.0.0              
 [16] prettyunits_1.2.0         gridExtra_2.3             cli_3.6.3                
 [19] sass_0.4.9                KEGGgraph_1.64.0          mvtnorm_1.2-5            
 [22] SQUAREM_2021.1            dreamlet_1.2.1            mixsqp_0.3-54            
 [25] Rsamtools_2.20.0          yulab.utils_0.1.4         zenith_1.6.0             
 [28] scater_1.32.0             sessioninfo_1.2.2         attempt_0.3.1            
 [31] invgamma_1.1              maps_3.4.2                limma_3.60.3             
 [34] rstudioapi_0.16.0         RSQLite_2.3.7             generics_0.1.3           
 [37] gridGraphics_0.5-1        BiocIO_1.14.0             gtools_3.9.5             
 [40] car_3.1-2                 dplyr_1.1.4               homologene_1.4.68.19.3.27
 [43] metadat_1.2-0             Matrix_1.7-0              ggbeeswarm_0.7.2         
 [46] fansi_1.0.6               abind_1.4-5               lifecycle_1.0.4          
 [49] yaml_2.3.9                edgeR_4.2.0               carData_3.0-5            
 [52] mathjaxr_1.6-0            gplots_3.1.3.1            SparseArray_1.4.8        
 [55] BiocFileCache_2.12.0      paletteer_1.6.0           grid_4.4.1               
 [58] blob_1.2.4                promises_1.3.0            ExperimentHub_2.12.0     
 [61] crayon_1.5.3              lattice_0.22-6            msigdbr_7.5.1            
 [64] beachmat_2.20.0           cowplot_1.1.3             annotate_1.82.0          
 [67] KEGGREST_1.44.1           magick_2.8.3              pillar_1.9.0             
 [70] knitr_1.48                boot_1.3-30               rjson_0.2.21             
 [73] corpcor_1.6.10            codetools_0.2-20          glue_1.7.0               
 [76] ggfun_0.1.5               vctrs_0.6.5               png_0.1-8                
 [79] treeio_1.28.0             spam_2.10-0               Rdpack_2.6               
 [82] gtable_0.3.5              assertthat_0.2.1          rematch2_2.1.2           
 [85] cachem_1.1.0              xfun_0.45                 rbibutils_2.2.16         
 [88] S4Arrays_1.4.1            mime_0.12                 Rfast_2.1.0              
 [91] iterators_1.0.14          fields_16.2               statmod_1.5.0            
 [94] nlme_3.1-165              pbkrtest_0.5.3            ggtree_3.12.0            
 [97] bit64_4.0.5               progress_1.2.3            EnvStats_2.8.1           
[100] filelock_1.0.3            rprojroot_2.0.4           bslib_0.7.0              
[103] irlba_2.3.5.1             KernSmooth_2.23-24        vipor_0.4.7              
[106] rmeta_3.0                 colorspace_2.1-0          DBI_1.2.3                
[109] tidyselect_1.2.1          bit_4.0.5                 compiler_4.4.1           
[112] curl_5.2.1                graph_1.82.0              BiocNeighbors_1.22.0     
[115] DelayedArray_0.30.1       plotly_4.10.4             rtracklayer_1.64.0       
[118] caTools_1.18.2            scales_1.3.0              remaCor_0.0.18           
[121] rappdirs_0.3.3            stringr_1.5.1             digest_0.6.36            
[124] minqa_1.2.7               variancePartition_1.34.0  rmarkdown_2.27           
[127] benchmarkmeData_1.0.4     aod_1.3.3                 RhpcBLASctl_0.23-42      
[130] htmltools_0.5.8.1         pkgconfig_2.0.3           lme4_1.1-35.5            
[133] sparseMatrixStats_1.16.0  mashr_0.2.79              dbplyr_2.5.0             
[136] fastmap_1.2.0             htmlwidgets_1.6.4         UCSC.utils_1.0.0         
[139] shiny_1.8.1.1             DelayedMatrixStats_1.26.0 jquerylib_0.1.4          
[142] jsonlite_1.8.8            config_0.3.2              BiocSingular_1.20.0      
[145] RCurl_1.98-1.14           scuttle_1.14.0            GenomeInfoDbData_1.2.12  
[148] ggplotify_0.1.2           dotCall64_1.1-1           patchwork_1.2.0          
[151] munsell_0.5.1             Rcpp_1.0.13               ape_5.8                  
[154] babelgene_22.9            viridis_0.6.5             RcppZiggurat_0.1.6       
[157] EnrichmentBrowser_2.34.1  stringi_1.8.4             zlibbioc_1.50.0          
[160] MASS_7.3-61               plyr_1.8.9                AnnotationHub_3.12.0     
[163] ggrepel_0.9.5             splines_4.4.1             hms_1.1.3                
[166] locfit_1.5-9.10           ggpubr_0.6.0              ggsignif_0.6.4           
[169] reshape2_1.4.4            ScaledMatrix_1.12.0       gprofiler2_0.2.3         
[172] BiocVersion_3.19.1        XML_3.99-0.17             evaluate_0.24.0          
[175] golem_0.4.1               RcppParallel_5.1.8        BiocManager_1.30.23      
[178] nloptr_2.1.1              foreach_1.5.2             httpuv_1.6.15            
[181] grr_0.9.5                 tidyr_1.3.1               purrr_1.0.2              
[184] scattermore_1.2           benchmarkme_1.0.8         ggplot2_3.5.1            
[187] ashr_2.2-63               rsvd_1.0.5                broom_1.0.6              
[190] xtable_1.8-4              restfulr_0.0.15           fANCOVA_0.6-1            
[193] tidytree_0.4.6            rstatix_0.7.2             later_1.3.2              
[196] viridisLite_0.4.2         truncnorm_1.0-9           tibble_3.2.1             
[199] lmerTest_3.1-3            aplot_0.2.3               memoise_2.0.1            
[202] beeswarm_0.4.0            AnnotationDbi_1.66.0      GenomicAlignments_1.40.0 
[205] shinyWidgets_0.8.6        GSEABase_1.66.0           here_1.0.1               
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-08-20
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────────────
 ! package              * version        date (UTC) lib source
   abind                  1.4-5          2016-07-21 [1] CRAN (R 4.4.0)
   annotate               1.82.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationDbi          1.66.0         2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub          3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   aod                    1.3.3          2023-12-13 [1] CRAN (R 4.4.0)
   ape                    5.8            2024-04-11 [1] CRAN (R 4.4.0)
   aplot                  0.2.3          2024-06-17 [1] CRAN (R 4.4.0)
   ashr                   2.2-63         2023-08-21 [1] CRAN (R 4.4.0)
   assertthat             0.2.1          2019-03-21 [1] CRAN (R 4.4.0)
   attempt                0.3.1          2020-05-03 [1] CRAN (R 4.4.0)
   babelgene              22.9           2022-09-29 [1] CRAN (R 4.4.0)
   backports              1.5.0          2024-05-23 [1] CRAN (R 4.4.0)
   beachmat               2.20.0         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm               0.4.0          2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme            1.0.8          2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData        1.0.4          2020-04-23 [1] CRAN (R 4.4.0)
   Biobase              * 2.64.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache          2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics         * 0.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                 1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager            1.30.23        2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors          1.22.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel         * 1.38.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocSingular           1.20.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion            3.19.1         2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings           * 2.72.1         2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                    4.0.5          2022-11-15 [1] CRAN (R 4.4.0)
   bit64                  4.0.5          2020-08-30 [1] CRAN (R 4.4.0)
   bitops                 1.0-7          2021-04-24 [1] CRAN (R 4.4.0)
   blob                   1.2.4          2023-03-17 [1] CRAN (R 4.4.0)
   boot                   1.3-30         2024-02-26 [1] CRAN (R 4.4.1)
   broom                  1.0.6          2024-05-17 [1] CRAN (R 4.4.0)
   bslib                  0.7.0          2024-03-29 [1] CRAN (R 4.4.0)
   cachem                 1.1.0          2024-05-16 [1] CRAN (R 4.4.0)
   car                    3.1-2          2023-03-30 [1] CRAN (R 4.4.0)
   carData                3.0-5          2022-01-06 [1] CRAN (R 4.4.0)
   caTools                1.18.2         2021-03-28 [1] CRAN (R 4.4.0)
 P cli                    3.6.3          2024-06-21 [2] CRAN (R 4.4.0)
   codetools              0.2-20         2024-03-31 [1] CRAN (R 4.4.1)
   colorout             * 1.3-0.2        2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace             2.1-0          2023-01-23 [1] CRAN (R 4.4.0)
   config                 0.3.2          2023-08-30 [1] CRAN (R 4.4.0)
   corpcor                1.6.10         2021-09-16 [1] CRAN (R 4.4.0)
   cowplot                1.1.3          2024-01-22 [1] CRAN (R 4.4.0)
   crayon                 1.5.3          2024-06-20 [1] CRAN (R 4.4.0)
   curl                   5.2.1          2024-03-01 [1] CRAN (R 4.4.0)
   data.table           * 1.15.4         2024-03-30 [2] CRAN (R 4.4.0)
   DBI                    1.2.3          2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                 2.5.0          2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray           0.30.1         2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats     1.26.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   digest                 0.6.36         2024-06-23 [1] CRAN (R 4.4.0)
   doParallel             1.0.17         2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64              1.1-1          2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                  1.1.4          2023-11-17 [1] CRAN (R 4.4.0)
   dreamlet               1.2.1          2024-06-09 [1] Bioconductor 3.19 (R 4.4.0)
   DT                     0.33           2024-04-04 [1] CRAN (R 4.4.0)
   edgeR                  4.2.0          2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   EnrichmentBrowser      2.34.1         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   EnvStats               2.8.1          2023-08-22 [1] CRAN (R 4.4.0)
   evaluate               0.24.0         2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub          2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fANCOVA                0.6-1          2020-11-13 [1] CRAN (R 4.4.0)
   fansi                  1.0.6          2023-12-08 [1] CRAN (R 4.4.0)
   fastmap                1.2.0          2024-05-15 [1] CRAN (R 4.4.0)
   fields                 16.2           2024-06-27 [1] CRAN (R 4.4.0)
   filelock               1.0.3          2023-12-11 [1] CRAN (R 4.4.0)
   foreach                1.5.2          2022-02-02 [1] CRAN (R 4.4.0)
   fs                     1.6.4          2024-04-25 [1] CRAN (R 4.4.0)
   generics               0.1.3          2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb         * 1.40.1         2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData       1.2.12         2024-05-01 [1] Bioconductor
   GenomicAlignments      1.40.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges        * 1.56.1         2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm             0.7.2          2023-04-29 [1] CRAN (R 4.4.0)
   ggfun                  0.1.5          2024-05-28 [1] CRAN (R 4.4.0)
   ggplot2                3.5.1          2024-04-23 [1] CRAN (R 4.4.0)
   ggplotify              0.1.2          2023-08-09 [1] CRAN (R 4.4.0)
   ggpubr                 0.6.0          2023-02-10 [1] CRAN (R 4.4.0)
   ggrepel                0.9.5          2024-01-10 [1] CRAN (R 4.4.0)
   ggsignif               0.6.4          2022-10-13 [1] CRAN (R 4.4.0)
   ggtree                 3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   glue                   1.7.0          2024-01-09 [1] CRAN (R 4.4.0)
   golem                  0.4.1          2023-06-05 [1] CRAN (R 4.4.0)
   gplots                 3.1.3.1        2024-02-02 [1] CRAN (R 4.4.0)
   gprofiler2             0.2.3          2024-02-23 [1] CRAN (R 4.4.0)
   graph                  1.82.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   gridExtra              2.3            2017-09-09 [1] CRAN (R 4.4.0)
   gridGraphics           0.5-1          2020-12-13 [1] CRAN (R 4.4.0)
   grr                    0.9.5          2016-08-26 [1] CRAN (R 4.4.0)
   GSEABase               1.66.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   gtable                 0.3.5          2024-04-22 [1] CRAN (R 4.4.0)
   gtools                 3.9.5          2023-11-20 [1] CRAN (R 4.4.0)
   here                   1.0.1          2020-12-13 [1] CRAN (R 4.4.0)
   hms                    1.1.3          2023-03-21 [1] CRAN (R 4.4.0)
   homologene             1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.4.0)
   htmltools              0.5.8.1        2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets            1.6.4          2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                 1.6.15         2024-03-26 [1] CRAN (R 4.4.0)
   httr                   1.4.7          2023-08-15 [1] CRAN (R 4.4.0)
   invgamma               1.1            2017-05-07 [1] CRAN (R 4.4.0)
   IRanges              * 2.38.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                  2.3.5.1        2022-10-03 [1] CRAN (R 4.4.0)
   iterators              1.0.14         2022-02-05 [1] CRAN (R 4.4.0)
   jquerylib              0.1.4          2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite               1.8.8          2023-12-04 [1] CRAN (R 4.4.0)
   KEGGgraph              1.64.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   KEGGREST               1.44.1         2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   KernSmooth             2.23-24        2024-05-17 [1] CRAN (R 4.4.1)
   knitr                  1.48           2024-07-07 [1] CRAN (R 4.4.1)
   later                  1.3.2          2023-12-06 [1] CRAN (R 4.4.0)
   lattice                0.22-6         2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval               0.2.2          2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle              1.0.4          2023-11-07 [1] CRAN (R 4.4.0)
   limma                  3.60.3         2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   lme4                   1.1-35.5       2024-07-03 [1] CRAN (R 4.4.1)
   lmerTest               3.1-3          2020-10-23 [1] CRAN (R 4.4.0)
   locfit                 1.5-9.10       2024-06-24 [1] CRAN (R 4.4.0)
   magick                 2.8.3          2024-02-18 [1] CRAN (R 4.4.0)
   magrittr             * 2.0.3          2022-03-30 [1] CRAN (R 4.4.0)
   maps                   3.4.2          2023-12-15 [1] CRAN (R 4.4.0)
   mashr                  0.2.79         2023-10-18 [1] CRAN (R 4.4.0)
   MASS                   7.3-61         2024-06-13 [1] CRAN (R 4.4.0)
   mathjaxr               1.6-0          2022-02-28 [1] CRAN (R 4.4.0)
   Matrix                 1.7-0          2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics       * 1.16.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats          * 1.3.0          2024-04-11 [1] CRAN (R 4.4.0)
   memoise                2.0.1          2021-11-26 [1] CRAN (R 4.4.0)
   metadat                1.2-0          2022-04-06 [1] CRAN (R 4.4.0)
   metafor                4.6-0          2024-03-28 [1] CRAN (R 4.4.0)
   mime                   0.12           2021-09-28 [1] CRAN (R 4.4.0)
   minqa                  1.2.7          2024-05-20 [1] CRAN (R 4.4.0)
   mixsqp                 0.3-54         2023-12-20 [1] CRAN (R 4.4.0)
   msigdbr                7.5.1          2022-03-30 [1] CRAN (R 4.4.0)
   munsell                0.5.1          2024-04-01 [1] CRAN (R 4.4.0)
   mvtnorm                1.2-5          2024-05-21 [1] CRAN (R 4.4.0)
   nlme                   3.1-165        2024-06-06 [1] CRAN (R 4.4.0)
   nloptr                 2.1.1          2024-06-25 [1] CRAN (R 4.4.0)
   numDeriv               2016.8-1.1     2019-06-06 [1] CRAN (R 4.4.0)
   orthogene            * 1.10.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   paletteer              1.6.0          2024-01-21 [1] CRAN (R 4.4.0)
   parallelly           * 1.38.0         2024-07-27 [1] CRAN (R 4.4.0)
   patchwork              1.2.0          2024-01-08 [1] CRAN (R 4.4.0)
   pbkrtest               0.5.3          2024-06-26 [1] CRAN (R 4.4.0)
   pillar                 1.9.0          2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig              2.0.3          2019-09-22 [1] CRAN (R 4.4.0)
   plotly                 4.10.4         2024-01-13 [1] CRAN (R 4.4.0)
   plyr                   1.8.9          2023-10-02 [1] CRAN (R 4.4.0)
   png                    0.1-8          2022-11-29 [1] CRAN (R 4.4.0)
   prettyunits            1.2.0          2023-09-24 [1] CRAN (R 4.4.0)
   progress               1.2.3          2023-12-06 [1] CRAN (R 4.4.0)
   promises               1.3.0          2024-04-05 [1] CRAN (R 4.4.0)
   purrr                  1.0.2          2023-08-10 [1] CRAN (R 4.4.0)
   R6                     2.5.1          2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs               0.3.3          2021-01-31 [1] CRAN (R 4.4.0)
   rbibutils              2.2.16         2023-10-25 [1] CRAN (R 4.4.0)
   RColorBrewer           1.1-3          2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                   1.0.13         2024-07-17 [1] CRAN (R 4.4.0)
   RcppParallel           5.1.8          2024-07-06 [1] CRAN (R 4.4.0)
   RcppZiggurat           0.1.6          2020-10-20 [1] CRAN (R 4.4.0)
   RCurl                  1.98-1.14      2024-01-09 [1] CRAN (R 4.4.0)
   Rdpack                 2.6            2023-11-08 [1] CRAN (R 4.4.0)
   remaCor                0.0.18         2024-02-08 [1] CRAN (R 4.4.0)
   rematch2               2.1.2          2020-05-01 [1] CRAN (R 4.4.0)
   reshape2               1.4.4          2020-04-09 [1] CRAN (R 4.4.0)
   restfulr               0.0.15         2022-06-16 [1] CRAN (R 4.4.0)
   Rfast                  2.1.0          2023-11-09 [1] CRAN (R 4.4.0)
   Rgraphviz              2.48.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RhpcBLASctl            0.23-42        2023-02-11 [1] CRAN (R 4.4.0)
   rjson                  0.2.21         2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                * 1.1.4          2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown              2.27           2024-05-17 [1] CRAN (R 4.4.0)
   rmeta                  3.0            2018-03-20 [1] CRAN (R 4.4.0)
   rprojroot              2.0.4          2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools              2.20.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                2.3.7          2024-05-27 [1] CRAN (R 4.4.0)
   rstatix                0.7.2          2023-02-01 [1] CRAN (R 4.4.0)
   rstudioapi             0.16.0         2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                   1.0.5          2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer            1.64.0         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   S4Arrays               1.4.1          2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors            * 0.42.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                   0.4.9          2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix           1.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                 1.3.0          2023-11-28 [1] CRAN (R 4.4.0)
   scater                 1.32.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scattermore            1.2            2023-06-12 [1] CRAN (R 4.4.0)
   scuttle                1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo            1.2.2          2021-12-06 [1] CRAN (R 4.4.0)
   shiny                  1.8.1.1        2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets           0.8.6          2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment * 1.26.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spam                   2.10-0         2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray            1.4.8          2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats      1.16.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment    * 1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD          * 1.16.2         2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   SQUAREM                2021.1         2021-01-13 [1] CRAN (R 4.4.0)
   statmod                1.5.0          2023-01-06 [1] CRAN (R 4.4.0)
   stringi                1.8.4          2024-05-06 [1] CRAN (R 4.4.0)
   stringr                1.5.1          2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment * 1.34.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   tibble                 3.2.1          2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                  1.3.1          2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect             1.2.1          2024-03-11 [1] CRAN (R 4.4.0)
   tidytree               0.4.6          2023-12-12 [1] CRAN (R 4.4.0)
   treeio                 1.28.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   truncnorm              1.0-9          2023-03-20 [1] CRAN (R 4.4.0)
   UCSC.utils             1.0.0          2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   utf8                   1.2.4          2023-10-22 [1] CRAN (R 4.4.0)
   variancePartition      1.34.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   vctrs                  0.6.5          2023-12-01 [1] CRAN (R 4.4.0)
   vipor                  0.4.7          2023-12-18 [1] CRAN (R 4.4.0)
   viridis                0.6.5          2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite            0.4.2          2023-05-02 [1] CRAN (R 4.4.0)
   withr                  3.0.0          2024-01-16 [1] CRAN (R 4.4.0)
   xfun                   0.45           2024-06-16 [1] CRAN (R 4.4.0)
   XML                    3.99-0.17      2024-06-25 [1] CRAN (R 4.3.3)
   xtable                 1.8-4          2019-04-21 [1] CRAN (R 4.4.0)
   XVector              * 0.44.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                   2.3.9          2024-07-05 [1] CRAN (R 4.4.0)
   yulab.utils            0.1.4          2024-01-28 [1] CRAN (R 4.4.0)
   zenith                 1.6.0          2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   zlibbioc               1.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────
