---
title: "02-Spatial Registration vs Yao Subclasses w 50plus cells in ms HYP dissections"
author: "Bernard Mulvey"
date: "2024-08-01"
output: html_document
---
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(SingleCellExperiment)
library(SpatialExperiment)
library(spatialLIBD)
require(colorout)
library(magrittr)
library(BiocParallel)
library(orthogene)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
```

load SPE, load Yao data in SCE format
```{r}
hyp2 <- readRDS("processed-data/03-QC_filters/hypN10_umi210_gene126_chrm50_spotsweeped_lognorm_rotsNmirrors_072224.RDS")

bscl <- fread("processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl=="X4",cl:="ARC.1"]
bscl[cl=="X6",cl:="ARC.2"]
bscl[cl=="X7",cl:="VMH.1"]
bscl[cl=="X12",cl:="VMH.2"]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=c("VMH.1","VMH.2","ARC.1","ARC.2",paste0("X",c(1:3,5,8:11,13:15))))

### load yao SCE
yao <- readRDS("raw-data/ABA_Whole_Mouse_Brain_Yao23_HYP/bothdatasets_raw_asSCE.RDS")


### check metadata for spatialLIBD compatability (no leading digits)
unique(yao$subclass)
unique(yao$library_label.x)
### k, we should be good.
```

spatial registration wrapper: hyp2
```{r}
sphyp.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphyp.reg,"processed-data/10-Spatial Registration/02a-spHYP_registrationwrapper.RDS")

```

registration wrapper: yao data, min 50 cells (pre-filter to avoid dropped levels errors)
```{r}
# get cell types with a dataset-wide total > 50
cellfilt <- as.data.table(colData(yao),keep.rownames=T)
cellfilt1 <- cellfilt[,.N,by="subclass"]
cellfilt1 <- cellfilt1[N>=50]

cellfilt2 <- cellfilt[subclass %in% cellfilt1$subclass][,.N,by=c("library_label.x","subclass")]

## determine the max number of samples possible
length(unique(cellfilt2$library_label.x))
## 47 samples here, so get those where at >=20 samples have >= ? cells
cellfilt3 <- cellfilt2[N>=20,.N,by="subclass"][subclass %in% grep(subclass,pattern="ARH|VMH",value=T)]
### 20 cells would drop an ARC class (17 samples). try 15 cells.
cellfilt3 <- cellfilt2[N>=15,.N,by="subclass"][subclass %in% grep(subclass,pattern="ARH|VMH",value=T)]
### that gets us at least 22 samples for all of the VMH, ARC subclasses (6 total), which'll do. 53 clusters total
cellfilt3 <- cellfilt2[N>=15,.N,by="subclass"][N>15]

# colnames in the SCE to keep:
cellfilt <- cellfilt[subclass %in% cellfilt3$subclass,rn]
# subset SCE
yaofilt <- yao[,cellfilt]
# sanity check, 53 clusters expected:
length(unique(yaofilt$subclass))
# nice


## clean up
rm(cellfilt,cellfilt1,cellfilt2,cellfilt3,sphyp.reg,yao)
gc(full=T)
```

Yao filt spatialwrapper
```{r}
## only one column of gene metadata here--gene symbol
yaofilt.spwrap <- registration_wrapper(yaofilt,var_registration = "subclass",var_sample_id = "library_label.x",gene_ensembl = "gene_symbol",gene_name = "gene_symbol",min_ncells = 15)

saveRDS(yaofilt.spwrap,"processed-data/10-Spatial Registration/02b-yaofilt_registrationwrapper.RDS")

# RAM free-up
rm(yaofilt.spwrap)
gc(full=T)
```

### once more, but with VMH subclasses and ARC subclasses collapsed down to a single "VMH" and "ARC"
```{r}
tmpcd <- as.data.table(colData(yaofilt),keep.rownames=T)
tmpcd[subclass %in% grep(subclass,pattern="VMH",value=T),subclass:="VMHall"]
# careful here--there's also a subclass not labeled ARH but that has a gene symbol containing arh in the name!
unique(grep(tmpcd$subclass,pattern="ARH",value=T))
tmpcd[subclass %in% grep(subclass,pattern="x075|x120|x121|x114",value=T),subclass:="ARHall"]

tmpcd <- DataFrame(tmpcd)
rownames(tmpcd) <- tmpcd$rn
tmpcd$rn <- NULL
tmpcd <- tmpcd[colnames(yaofilt),]
colData(yaofilt) <- tmpcd
yaofilt$subclass <- factor(yaofilt$subclass)

# should now be 49 subclasses (53-4 ARC - 2 VMH + 1 joint vmh + 1 joint arc)
levels(yaofilt$subclass) # good

### now re-run spatial registration..
yaofilt.collapsed.spwrap <- registration_wrapper(yaofilt,var_registration = "subclass",var_sample_id = "library_label.x",gene_ensembl = "gene_symbol",gene_name = "gene_symbol",min_ncells = 15)


saveRDS(yaofilt.collapsed.spwrap,"processed-data/10-Spatial Registration/02c-yaofilt_collapsedVMHandARHsubclasses_registrationwrapper.RDS")

```


### for posterity, using the k20, k31, and k15 collapsed spatial domains:
    arc20= X1, X4
    vmh20= X2, X3
    arc31= X2, X4, X11, X22
    vmh31= X6, X21, X25, X26, X31

```{r}
rm(list=ls())
gc(full=T)

#
hyp2 <- readRDS("processed-data/03-QC_filters/hypN10_umi210_gene126_chrm50_spotsweeped_lognorm_rotsNmirrors_072224.RDS")

# k20
bscl <- fread("processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k20_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl=="X2",cl:="VMH20.1"]
bscl[cl=="X3",cl:="VMH20.2"]
bscl[cl=="X1",cl:="ARC20.1"]
bscl[cl=="X4",cl:="ARC20.2"]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=c("VMH20.1","VMH20.2","ARC20.1","ARC20.2",paste0("X",c(5:20))))


sphyp20.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphyp20.reg,"processed-data/10-Spatial Registration/02d-sphypk20_regwrapper.RDS")
```

# k31:
# arc31=c("X2","X4","X11","X22"),
# vmh31=c("X6","X21","X25","X26","X31")
```{r}
bscl <- fread("processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k31_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl=="X6",cl:="VMH31.1"]
bscl[cl=="X21",cl:="VMH31.2"]
bscl[cl=="X25",cl:="VMH31.3"]
bscl[cl=="X26",cl:="VMH31.4"]
bscl[cl=="X31",cl:="VMH31.5"]

bscl[cl=="X2",cl:="ARC31.1"]
bscl[cl=="X4",cl:="ARC31.2"]
bscl[cl=="X11",cl:="ARC31.3"]
bscl[cl=="X22",cl:="ARC31.4"]

bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

others <- as.numeric(gsub(unique(bscl$cl),pattern="^X(.*)",replacement="\\1"))
others <- others[!is.na(others)]

colLabels(hyp2) <- factor(bscl$cl,levels=c("VMH31.1","VMH31.2","VMH31.3","VMH31.4","VMH31.5","ARC31.1","ARC31.2","ARC31.3","ARC31.4",paste0("X",others)))

sphyp31.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphyp31.reg,"processed-data/10-Spatial Registration/02e-sphypk31_regwrapper.RDS")
```

# collapsed
```{r}
bscl <- fread("processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl %in% c("X4","X6"),cl:="ARC"]
bscl[cl %in% c("X7","X12"),cl:="VMH"]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=c("ARC","VMH",paste0("X",c(1:3,5,8:11,13:15))))

sphypcollapse.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphypcollapse.reg,"processed-data/10-Spatial Registration/02f-sphypcollapse_regwrapper.RDS")
```

Wipe env and load the outputs from spatialLIBD::registration_wrapper
```{r}
rm(list=ls())
gc(full=T)


abasub <- readRDS("processed-data/10-Spatial Registration/02b-yaofilt_registrationwrapper.RDS")

abasub.singlevmharc <- readRDS("processed-data/10-Spatial Registration/02c-yaofilt_collapsedVMHandARHsubclasses_registrationwrapper.RDS")

sphyp <- readRDS("processed-data/10-Spatial Registration/02a-spHYP_registrationwrapper.RDS")

sphyp20 <- readRDS("data/10-Spatial Registration/02d-sphypk20_regwrapper.RDS")

sphyp31 <- readRDS("data/10-Spatial Registration/02e-sphypk31_regwrapper.RDS")

sphypclps <- readRDS("data/10-Spatial Registration/02f-sphypcollapse_regwrapper.RDS")
# 

orthos <- as.data.table(unique(as.data.table(sphyp$enrichment$ensembl)))
orthos <- convert_orthologs(gene_df = orthos,gene_input = "V1",gene_output = "columns",input_species = "human",output_species = "mouse",non121_strategy = "drop_both_species")

tmplist2 <- list(sphyp,sphyp20,sphyp31,sphypclps)
tmplist2 <- lapply(X=tmplist2,FUN=function(x){
    
    Y <- as.data.table(x$enrichment)
    keepcols <- c("ensembl",grep(names(Y),pattern="^t_stat_",value=T))
    Y <- Y[,..keepcols]
    
    setnames(Y,c(2:ncol(Y)),gsub(names(Y)[c(2:ncol(Y))],pattern="^t_stat_",replacement=""))


    Y <- as.data.frame(Y)
    rownames(Y) <- Y$ensembl
    Y$ensembl <- NULL
    return(as.data.frame(Y))
})

tmplist <- list(abasub,abasub.singlevmharc)
tmplist <- lapply(X=tmplist,function(x){

    Y <- as.data.table(x$enrichment,keep.rownames=T)
    keepcols <- c("rn",grep(names(Y),pattern="^t_stat_",value=T))
    Y <- Y[,..keepcols]
    
    setnames(Y,c(2:ncol(Y)),gsub(names(Y)[c(2:ncol(Y))],pattern="^t_stat_",replacement=""))
    
    # get human ensembl orthologs
    Y <- merge.data.table(Y,orthos,by.x="rn",by.y="ortholog_gene")

    Y <- as.data.frame(Y)
    rownames(Y) <- Y$input_gene
    Y$rn <- NULL
    Y$input_gene <- NULL
    return(as.data.frame(Y))
    
    
})


abas <- list(tmplist[[1]],tmplist[[2]],tmplist[[1]],tmplist[[2]],tmplist[[1]],tmplist[[2]],tmplist[[1]],tmplist[[2]])
names(abas) <- rep(c("subclass","subclass_collapsed"),4)

hyps <- list(tmplist2[[1]],tmplist2[[1]],tmplist2[[2]],tmplist2[[2]],tmplist2[[3]],tmplist2[[3]],tmplist2[[4]],tmplist2[[4]])
names(hyps) <- c(rep("sphypk15",2),rep("sphypk20",2),rep("sphypk31",2),rep("sphypk15_collapsed",2))
```

spatialLIBD layer_stat_cor won't cooperate so just do this manually
### correction 02/27/24: spearman, not pearson correlation
```{r}
outfilesuff <- as.list(paste0(names(abas),"_to_",names(hyps)))

sbp <- MulticoreParam(8)

spatregs <- bpmapply(m=hyps,s=abas,z=outfilesuff,SIMPLIFY = F,BPPARAM = sbp,FUN = function(m,s,z){
    j<-c(25,50,75,100,250,500,750,1000)
    i<-1
    outlist <- list()
    for (i in c(1:9)){
        if(i<=8){
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- j[i]
        }
        else{
            # no value for top_n --> all genes used
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- nrow(cortab)
        }
        # initialize cormat
        nclaba <- ncol(s)
        nclhyp <- ncol(m)
        cormat <- as.data.frame(matrix(nrow=nclhyp,ncol=nclaba))
        k <- 1
        for (k in c(1:nclhyp)){
            rownames(cormat)[k] <- colnames(m)[k]
            l <- 1
            for (l in c(1:nclaba)){
                colnames(cormat)[l] <- colnames(s)[l]
                
                curhyp <- colnames(m)[k]
                curaba <- colnames(s)[l]
                setorderv(cortab,curhyp,-1)
                
                cormat[k,l] <- cor(cortab[c(1:ngene),..curhyp],cortab[c(1:ngene),..curaba],use="pairwise.complete.obs",method = "spearman")
            }
        }
        plttab <- as.data.table(cormat,keep.rownames=T)
        setnames(plttab,"rn","sphyp_domain")
        outlist[[i]] <- plttab
        names(outlist)[i] <- paste0("ABA23_HYP_",z,"_top_",ngene,"_sphypdomainmks")
        rm(curj,plttab,cortab)
    }
    return(outlist)
})

saveRDS(spatregs,"processed-data/10-Spatial Registration/02g-sphyp-k15-20-31-15clps_ABAms-47subclass-47withVMHARCclps.RDS")
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] parallelly_1.38.0           default_1.0.0              
 [3] orthogene_1.10.0            BiocParallel_1.38.0        
 [5] magrittr_2.0.3              colorout_1.3-0.2           
 [7] spatialLIBD_1.16.2          SpatialExperiment_1.14.0   
 [9] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0
[11] Biobase_2.64.0              GenomicRanges_1.56.1       
[13] MatrixGenerics_1.16.0       matrixStats_1.3.0          
[15] Biostrings_2.72.1           GenomeInfoDb_1.40.1        
[17] XVector_0.44.0              IRanges_2.38.1             
[19] S4Vectors_0.42.1            BiocGenerics_0.50.0        
[21] data.table_1.15.4           rlang_1.1.4                

loaded via a namespace (and not attached):
  [1] later_1.3.2               BiocIO_1.14.0             ggplotify_0.1.2          
  [4] bitops_1.0-7              filelock_1.0.3            fields_16.2              
  [7] tibble_3.2.1              XML_3.99-0.17             lifecycle_1.0.4          
 [10] rstatix_0.7.2             edgeR_4.2.0               doParallel_1.0.17        
 [13] rprojroot_2.0.4           lattice_0.22-6            backports_1.5.0          
 [16] limma_3.60.3              plotly_4.10.4             sass_0.4.9               
 [19] rmarkdown_2.27            jquerylib_0.1.4           yaml_2.3.9               
 [22] httpuv_1.6.15             spam_2.10-0               sessioninfo_1.2.2        
 [25] cowplot_1.1.3             DBI_1.2.3                 RColorBrewer_1.1-3       
 [28] golem_0.4.1               maps_3.4.2                abind_1.4-5              
 [31] zlibbioc_1.50.0           purrr_1.0.2               RCurl_1.98-1.14          
 [34] yulab.utils_0.1.4         rappdirs_0.3.3            GenomeInfoDbData_1.2.12  
 [37] ggrepel_0.9.5             irlba_2.3.5.1             tidytree_0.4.6           
 [40] DelayedMatrixStats_1.26.0 codetools_0.2-20          DelayedArray_0.30.1      
 [43] DT_0.33                   scuttle_1.14.0            tidyselect_1.2.1         
 [46] aplot_0.2.3               UCSC.utils_1.0.0          ScaledMatrix_1.12.0      
 [49] viridis_0.6.5             shinyWidgets_0.8.6        BiocFileCache_2.12.0     
 [52] GenomicAlignments_1.40.0  jsonlite_1.8.8            BiocNeighbors_1.22.0     
 [55] scater_1.32.0             iterators_1.0.14          foreach_1.5.2            
 [58] tools_4.4.1               treeio_1.28.0             Rcpp_1.0.13              
 [61] glue_1.7.0                gridExtra_2.3             SparseArray_1.4.8        
 [64] xfun_0.45                 here_1.0.1                dplyr_1.1.4              
 [67] withr_3.0.0               BiocManager_1.30.23       fastmap_1.2.0            
 [70] fansi_1.0.6               digest_0.6.36             rsvd_1.0.5               
 [73] gridGraphics_0.5-1        R6_2.5.1                  mime_0.12                
 [76] colorspace_2.1-0          RSQLite_2.3.7             config_0.3.2             
 [79] utf8_1.2.4                tidyr_1.3.1               generics_0.1.3           
 [82] rtracklayer_1.64.0        httr_1.4.7                htmlwidgets_1.6.4        
 [85] S4Arrays_1.4.1            pkgconfig_2.0.3           gtable_0.3.5             
 [88] blob_1.2.4                htmltools_0.5.8.1         carData_3.0-5            
 [91] dotCall64_1.1-1           scales_1.3.0              png_0.1-8                
 [94] attempt_0.3.1             ggfun_0.1.5               knitr_1.48               
 [97] rstudioapi_0.16.0         rjson_0.2.21              nlme_3.1-165             
[100] curl_5.2.1                cachem_1.1.0              BiocVersion_3.19.1       
[103] parallel_4.4.1            vipor_0.4.7               AnnotationDbi_1.66.0     
[106] restfulr_0.0.15           pillar_1.9.0              grid_4.4.1               
[109] vctrs_0.6.5               ggpubr_0.6.0              promises_1.3.0           
[112] car_3.1-2                 BiocSingular_1.20.0       dbplyr_2.5.0             
[115] beachmat_2.20.0           xtable_1.8-4              beeswarm_0.4.0           
[118] paletteer_1.6.0           evaluate_0.24.0           magick_2.8.3             
[121] cli_3.6.3                 locfit_1.5-9.10           compiler_4.4.1           
[124] Rsamtools_2.20.0          crayon_1.5.3              grr_0.9.5                
[127] ggsignif_0.6.4            gprofiler2_0.2.3          rematch2_2.1.2           
[130] fs_1.6.4                  ggbeeswarm_0.7.2          viridisLite_0.4.2        
[133] babelgene_22.9            munsell_0.5.1             lazyeval_0.2.2           
[136] homologene_1.4.68.19.3.27 Matrix_1.7-0              ExperimentHub_2.12.0     
[139] benchmarkme_1.0.8         patchwork_1.2.0           sparseMatrixStats_1.16.0 
[142] bit64_4.0.5               ggplot2_3.5.1             KEGGREST_1.44.1          
[145] statmod_1.5.0             shiny_1.8.1.1             AnnotationHub_3.12.0     
[148] broom_1.0.6               memoise_2.0.1             bslib_0.7.0              
[151] ggtree_3.12.0             benchmarkmeData_1.0.4     bit_4.0.5                
[154] ape_5.8                  
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-08-01
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────────────
 ! package              * version        date (UTC) lib source
   abind                  1.4-5          2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi          1.66.0         2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub          3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   ape                    5.8            2024-04-11 [1] CRAN (R 4.4.0)
   aplot                  0.2.3          2024-06-17 [1] CRAN (R 4.4.0)
   attempt                0.3.1          2020-05-03 [1] CRAN (R 4.4.0)
   babelgene              22.9           2022-09-29 [1] CRAN (R 4.4.0)
   backports              1.5.0          2024-05-23 [1] CRAN (R 4.4.0)
   beachmat               2.20.0         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm               0.4.0          2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme            1.0.8          2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData        1.0.4          2020-04-23 [1] CRAN (R 4.4.0)
   Biobase              * 2.64.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache          2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics         * 0.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                 1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager            1.30.23        2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors          1.22.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel         * 1.38.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocSingular           1.20.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion            3.19.1         2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings           * 2.72.1         2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                    4.0.5          2022-11-15 [1] CRAN (R 4.4.0)
   bit64                  4.0.5          2020-08-30 [1] CRAN (R 4.4.0)
   bitops                 1.0-7          2021-04-24 [1] CRAN (R 4.4.0)
   blob                   1.2.4          2023-03-17 [1] CRAN (R 4.4.0)
   broom                  1.0.6          2024-05-17 [1] CRAN (R 4.4.0)
   bslib                  0.7.0          2024-03-29 [1] CRAN (R 4.4.0)
   cachem                 1.1.0          2024-05-16 [1] CRAN (R 4.4.0)
   car                    3.1-2          2023-03-30 [1] CRAN (R 4.4.0)
   carData                3.0-5          2022-01-06 [1] CRAN (R 4.4.0)
 P cli                    3.6.3          2024-06-21 [2] CRAN (R 4.4.0)
   codetools              0.2-20         2024-03-31 [1] CRAN (R 4.4.1)
   colorout             * 1.3-0.2        2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace             2.1-0          2023-01-23 [1] CRAN (R 4.4.0)
   config                 0.3.2          2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                1.1.3          2024-01-22 [1] CRAN (R 4.4.0)
   crayon                 1.5.3          2024-06-20 [1] CRAN (R 4.4.0)
   curl                   5.2.1          2024-03-01 [1] CRAN (R 4.4.0)
   data.table           * 1.15.4         2024-03-30 [2] CRAN (R 4.4.0)
   DBI                    1.2.3          2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                 2.5.0          2024-03-19 [1] CRAN (R 4.4.0)
   default              * 1.0.0          2017-08-07 [1] CRAN (R 4.4.0)
   DelayedArray           0.30.1         2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats     1.26.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   digest                 0.6.36         2024-06-23 [1] CRAN (R 4.4.0)
   doParallel             1.0.17         2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64              1.1-1          2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                  1.1.4          2023-11-17 [1] CRAN (R 4.4.0)
   DT                     0.33           2024-04-04 [1] CRAN (R 4.4.0)
   edgeR                  4.2.0          2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate               0.24.0         2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub          2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fansi                  1.0.6          2023-12-08 [1] CRAN (R 4.4.0)
   fastmap                1.2.0          2024-05-15 [1] CRAN (R 4.4.0)
   fields                 16.2           2024-06-27 [1] CRAN (R 4.4.0)
   filelock               1.0.3          2023-12-11 [1] CRAN (R 4.4.0)
   foreach                1.5.2          2022-02-02 [1] CRAN (R 4.4.0)
   fs                     1.6.4          2024-04-25 [1] CRAN (R 4.4.0)
   generics               0.1.3          2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb         * 1.40.1         2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData       1.2.12         2024-05-01 [1] Bioconductor
   GenomicAlignments      1.40.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges        * 1.56.1         2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm             0.7.2          2023-04-29 [1] CRAN (R 4.4.0)
   ggfun                  0.1.5          2024-05-28 [1] CRAN (R 4.4.0)
   ggplot2                3.5.1          2024-04-23 [1] CRAN (R 4.4.0)
   ggplotify              0.1.2          2023-08-09 [1] CRAN (R 4.4.0)
   ggpubr                 0.6.0          2023-02-10 [1] CRAN (R 4.4.0)
   ggrepel                0.9.5          2024-01-10 [1] CRAN (R 4.4.0)
   ggsignif               0.6.4          2022-10-13 [1] CRAN (R 4.4.0)
   ggtree                 3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   glue                   1.7.0          2024-01-09 [1] CRAN (R 4.4.0)
   golem                  0.4.1          2023-06-05 [1] CRAN (R 4.4.0)
   gprofiler2             0.2.3          2024-02-23 [1] CRAN (R 4.4.0)
   gridExtra              2.3            2017-09-09 [1] CRAN (R 4.4.0)
   gridGraphics           0.5-1          2020-12-13 [1] CRAN (R 4.4.0)
   grr                    0.9.5          2016-08-26 [1] CRAN (R 4.4.0)
   gtable                 0.3.5          2024-04-22 [1] CRAN (R 4.4.0)
   here                   1.0.1          2020-12-13 [1] CRAN (R 4.4.0)
   homologene             1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.4.0)
   htmltools              0.5.8.1        2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets            1.6.4          2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                 1.6.15         2024-03-26 [1] CRAN (R 4.4.0)
   httr                   1.4.7          2023-08-15 [1] CRAN (R 4.4.0)
   IRanges              * 2.38.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                  2.3.5.1        2022-10-03 [1] CRAN (R 4.4.0)
   iterators              1.0.14         2022-02-05 [1] CRAN (R 4.4.0)
   jquerylib              0.1.4          2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite               1.8.8          2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST               1.44.1         2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   knitr                  1.48           2024-07-07 [1] CRAN (R 4.4.1)
   later                  1.3.2          2023-12-06 [1] CRAN (R 4.4.0)
   lattice                0.22-6         2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval               0.2.2          2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle              1.0.4          2023-11-07 [1] CRAN (R 4.4.0)
   limma                  3.60.3         2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                 1.5-9.10       2024-06-24 [1] CRAN (R 4.4.0)
   magick                 2.8.3          2024-02-18 [1] CRAN (R 4.4.0)
   magrittr             * 2.0.3          2022-03-30 [1] CRAN (R 4.4.0)
   maps                   3.4.2          2023-12-15 [1] CRAN (R 4.4.0)
   Matrix                 1.7-0          2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics       * 1.16.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats          * 1.3.0          2024-04-11 [1] CRAN (R 4.4.0)
   memoise                2.0.1          2021-11-26 [1] CRAN (R 4.4.0)
   mime                   0.12           2021-09-28 [1] CRAN (R 4.4.0)
   munsell                0.5.1          2024-04-01 [1] CRAN (R 4.4.0)
   nlme                   3.1-165        2024-06-06 [1] CRAN (R 4.4.0)
   orthogene            * 1.10.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   paletteer              1.6.0          2024-01-21 [1] CRAN (R 4.4.0)
   parallelly           * 1.38.0         2024-07-27 [1] CRAN (R 4.4.0)
   patchwork              1.2.0          2024-01-08 [1] CRAN (R 4.4.0)
   pillar                 1.9.0          2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig              2.0.3          2019-09-22 [1] CRAN (R 4.4.0)
   plotly                 4.10.4         2024-01-13 [1] CRAN (R 4.4.0)
   png                    0.1-8          2022-11-29 [1] CRAN (R 4.4.0)
   promises               1.3.0          2024-04-05 [1] CRAN (R 4.4.0)
   purrr                  1.0.2          2023-08-10 [1] CRAN (R 4.4.0)
   R6                     2.5.1          2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs               0.3.3          2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer           1.1-3          2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                   1.0.13         2024-07-17 [1] CRAN (R 4.4.0)
   RCurl                  1.98-1.14      2024-01-09 [1] CRAN (R 4.4.0)
   rematch2               2.1.2          2020-05-01 [1] CRAN (R 4.4.0)
   restfulr               0.0.15         2022-06-16 [1] CRAN (R 4.4.0)
   rjson                  0.2.21         2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                * 1.1.4          2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown              2.27           2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot              2.0.4          2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools              2.20.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                2.3.7          2024-05-27 [1] CRAN (R 4.4.0)
   rstatix                0.7.2          2023-02-01 [1] CRAN (R 4.4.0)
   rstudioapi             0.16.0         2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                   1.0.5          2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer            1.64.0         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   S4Arrays               1.4.1          2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors            * 0.42.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                   0.4.9          2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix           1.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                 1.3.0          2023-11-28 [1] CRAN (R 4.4.0)
   scater                 1.32.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scuttle                1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo            1.2.2          2021-12-06 [1] CRAN (R 4.4.0)
   shiny                  1.8.1.1        2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets           0.8.6          2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment * 1.26.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spam                   2.10-0         2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray            1.4.8          2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats      1.16.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment    * 1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD          * 1.16.2         2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   statmod                1.5.0          2023-01-06 [1] CRAN (R 4.4.0)
   SummarizedExperiment * 1.34.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   tibble                 3.2.1          2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                  1.3.1          2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect             1.2.1          2024-03-11 [1] CRAN (R 4.4.0)
   tidytree               0.4.6          2023-12-12 [1] CRAN (R 4.4.0)
   treeio                 1.28.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   UCSC.utils             1.0.0          2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   utf8                   1.2.4          2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                  0.6.5          2023-12-01 [1] CRAN (R 4.4.0)
   vipor                  0.4.7          2023-12-18 [1] CRAN (R 4.4.0)
   viridis                0.6.5          2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite            0.4.2          2023-05-02 [1] CRAN (R 4.4.0)
   withr                  3.0.0          2024-01-16 [1] CRAN (R 4.4.0)
   xfun                   0.45           2024-06-16 [1] CRAN (R 4.4.0)
   XML                    3.99-0.17      2024-06-25 [1] CRAN (R 4.3.3)
   xtable                 1.8-4          2019-04-21 [1] CRAN (R 4.4.0)
   XVector              * 0.44.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                   2.3.9          2024-07-05 [1] CRAN (R 4.4.0)
   yulab.utils            0.1.4          2024-01-28 [1] CRAN (R 4.4.0)
   zlibbioc               1.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────
