---
title: "Fig 3 Panels"
author: "Bernard Mulvey"
date: "2024-02-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table) # Preferred data manipulation package
library(Biostrings) # General
library(ggrastr) # render plots with bajillions of points as raster so they can be manipulated in eg Illustrator
library(Cairo) # pdf device for ^ to properly save
library(ggplot2) # Dependency for several plotting functions
library(gridExtra)
library(SpatialExperiment) # Dependency
library(escheR) # spotplots with domain outlines
library(ggtext) # mixed fonts in ggplot elements
library(ggstance) # y axis dodge
require(colorout) # Utility for RStudio
ColorOut() 

# code reformatting in Rstudio
options("styler.addins_style_transformer" = "biocthis::bioc_style()") 

# set plotting defaults for ggplot
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 10), axis.title.x = element_text(size = 11), axis.text.y = element_text(size = 10), axis.title.y = element_text(size =11), plot.title = element_text(size = 12,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=9), legend.title = element_text(size=10,hjust=0.5)))
```

### Setup: load data, cluster assignments, and rename from generic to appropriate region (VMH/ARC/WM)
```{r}
hyp2 <- readRDS("data/03-QC_filters/hypN9_umi275_gene166_chrm50_lognorm_111723.RDS")

bscl <- fread("data/06-BayesSpace/02-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=paste0("X",c(1:15)))

### relabel as VMH/ARC/WM for spotplottings
tmpcd <- as.data.table(colData(hyp2),keep.rownames=T)
tmpcd[label %in% c("X11","X12"),label:="VMH"]
tmpcd[label %in% c("X3","X5"),label:="ARC"]
tmpcd[label %in% c("X1","X9"),label:="WM"]

tmpcd <- DataFrame(tmpcd)
rownames(tmpcd) <- tmpcd$rn
tmpcd <- tmpcd[colnames(hyp2),]
colData(hyp2) <- tmpcd
hyp2$label <- factor(hyp2$label,levels=c("VMH","ARC","WM",paste0("X",c(2,4,6:8,10,13:15))))

### also define cluster tables that only have VMH or ARC and otehrs as NA for escher plotting outlines specific to VMH/ARC
bscl.vmhonly <- as.data.table(bscl,keep.rownames=T)
bscl.arconly <- as.data.table(bscl,keep.rownames=T)
bscl.vmharc <- as.data.table(bscl,keep.rownames=T)

bscl.vmhonly <- bscl.vmhonly[cl %in% c("X11","X12"),cl2:="VMH"]
bscl.arconly <- bscl.arconly[cl %in% c("X3","X5"),cl2:="ARC"]
bscl.vmharc <- bscl.vmharc[cl %in% c("X11","X12"),cl2:="VMH"]
bscl.vmharc <- bscl.vmharc[cl %in% c("X3","X5"),cl2:="ARC"]

rm(tmpcd)
```

panels A-F: previously 1.75 height / 2 width; moving sample IDs to legend --> one line plot titles --> saves .18 inches of space / plot --> save w new height 1.57in.

### Panel A ANKRD34B (VMH) ###
```{r}
### extract sample to plot ###
ank <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(ank) <- rowData(ank)$gene_name
colData(ank)$`log counts` <- logcounts(ank)["ANKRD34B",]
anklab <- copy(bscl.vmharc)[rn %in% colnames(ank)]
anklab <- DataFrame(anklab)
rownames(anklab) <- anklab$rn

ank$Domain <- factor(anklab$cl2,levels=c("VMH","ARC"))
### make spotplots
p <- make_escheR(ank)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3A-ANKRD34B.pdf",height=1.57,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("*ANKRD34B* (Male)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0,-0.05,0,unit = "in"))
dev.off()

rm(ank,p,anklab)
```


### Panel B LAMP5 (VMH) ###
```{r}
### extract sample to plot ###
lamp <- hyp2[,hyp2$sample_id=="V12D05-350_D1"]
rownames(lamp) <- rowData(lamp)$gene_name
colData(lamp)$`log counts` <- logcounts(lamp)["LAMP5",]
lamplab <- copy(bscl.vmharc)[rn %in% colnames(lamp)]
lamplab <- DataFrame(lamplab)
rownames(lamplab) <- lamplab$rn

lamp$Domain <- factor(lamplab$cl2,levels=c("VMH","ARC"))
### make spotplots
p <- make_escheR(lamp)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3B-LAMP5.pdf",height=1.57,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("*LAMP5* (Female)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(lamp,lamplab,p)
```

### Panel C GABRE (ARC) ###
```{r}
### extract sample to plot ###
gabre <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(gabre) <- rowData(gabre)$gene_name
colData(gabre)$`log counts` <- logcounts(gabre)["GABRE",]
gabrelab <- copy(bscl.vmharc)[rn %in% colnames(gabre)]
gabrelab <- DataFrame(gabrelab)
rownames(gabrelab) <- gabrelab$rn

gabre$Domain <- factor(gabrelab$cl2,levels=c("VMH","ARC"))
### make spotplots
p <- make_escheR(gabre)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3C-GABRE.pdf",height=1.57,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("*GABRE* (Male)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(gabre,gabrelab,p)
```


### Panel D CYP26A1 ###
```{r}
### extract sample to plot ###
cyp <- hyp2[,hyp2$sample_id=="V12D05-350_D1"]
rownames(cyp) <- rowData(cyp)$gene_name
colData(cyp)$`log counts` <- logcounts(cyp)["CYP26A1",]
cyplab <- copy(bscl.vmharc)[rn %in% colnames(cyp)]
cyplab <- DataFrame(cyplab)
rownames(cyplab) <- cyplab$rn

cyp$Domain <- factor(cyplab$cl2,levels=c("VMH","ARC"))
### make spotplots
p <- make_escheR(cyp)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3D-CYP26A1.pdf",height=1.57,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("*CYP26A1* (Female)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(cyp,cyplab,p)
```

### ISH panels follow F ###
###########################

### intra-VMH SVG: NMU ###
### since we show both ARC domains in the other SVG panel (GAL), we will here too ###
```{r}
### extract sample to plot ###
nmu <- hyp2[,hyp2$sample_id=="V12D05-350_D1"]
rownames(nmu) <- rowData(nmu)$gene_name
colData(nmu)$`log counts` <- logcounts(nmu)["NMU",]
nmulab <- copy(as.data.table(bscl))[rn %in% colnames(nmu)]
nmulab[cl=="X3",cl:="ARC.1"]
nmulab[cl=="X5",cl:="ARC.2"]
nmulab[cl %in% c("X11","X12"),cl:="VMH"]
nmulab[!(cl %in% c("ARC.1","ARC.2","VMH")),cl:="NA"]
nmulab <- DataFrame(nmulab,row.names=nmulab$rn)

nmu$Domain <- factor(nmulab$cl,levels=c("VMH","ARC.1","ARC.2"))
### make spotplots
p <- make_escheR(nmu)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3X-NMU-VMHsvg.pdf",height=1.57,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf","#e949c1"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("*NMU* (Female)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(nmu,nmulab,p)
```

### intra-ARC SVG: GAL ### 
### show this one with both ARC boundaries illustrated in different colors to get across the non-domain-hewing expression pattern
```{r}
gal <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(gal) <- rowData(gal)$gene_name
colData(gal)$`log counts` <- logcounts(gal)["GAL",]

gallab <- copy(as.data.table(bscl))[rn %in% colnames(gal)]
gallab[cl=="X3",cl:="ARC.1"]
gallab[cl=="X5",cl:="ARC.2"]
gallab[cl %in% c("X11","X12"),cl:="VMH"]
gallab[!(cl %in% c("ARC.1","ARC.2","VMH")),cl:="NA"]

gallab <- DataFrame(gallab)
rownames(gallab) <- gallab$rn

gal$Domain <- factor(gallab$cl,levels=c("VMH","ARC.1","ARC.2"))
### make spotplots
p <- make_escheR(gal)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3X-GAL-ARCsvg.pdf",height=1.57,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf","#e949c1"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("*GAL* (Male)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(gal,gallab,p)

```

### for escher plots, extract two legends: one for the collapsed domains to be shared across all 4 panels and one for the split arc (SVG) domains and plot each once so they aren't taking up space. (plus, they're impossible to read as open circles so we can change that here too).
```{r}
### custom function to extrct legend from one plot to extend to all ###
extract_legend <- function(my_ggp) {
  step1 <- ggplot_gtable(ggplot_build(my_ggp))
  step2 <- which(sapply(step1$grobs, function(x) x$name) == "guide-box")
  step3 <- step1$grobs[[step2]]
  return(step3)
}

### run the first part of the code used to generate the marker panels but with the fill legend changed to be the domains factored in the same order
gabre <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(gabre) <- rowData(gabre)$gene_name
colData(gabre)$`log counts` <- logcounts(gabre)["GABRE",]
gabrelab <- copy(bscl.vmharc)[rn %in% colnames(gabre)]
gabrelab <- DataFrame(gabrelab)
rownames(gabrelab) <- gabrelab$rn

gabre$Domain <- factor(gabrelab$cl2,levels=c("VMH","ARC"))
### make a ggplot of nothing in particular, we just want the legend
p <- ggplot(as.data.table(colData(gabre))[!is.na(Domain)],aes(x=sum_umi,y=expr_chrM_ratio,color=Domain))+
    geom_point()+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9),plot.margin = margin(-0.3,0,-0.1,0,unit = "in"))

pl <- extract_legend(p)
dev.off()


# save in dimension ~equiv to the dimensions of a legend included in one panel
pdf("plots/Manuscript/3A-3D shared legend.pdf",height=0.6,width=0.45)
do.call("grid.arrange",list(pl))
dev.off()

rm(gabre,gabrelab)
```

## make a 3-domain ARC legend as well
```{r}
### run the first part of the code used to generate the marker panels but with the fill legend changed to be the domains factored in the same order
gal <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(gal) <- rowData(gal)$gene_name
colData(gal)$`log counts` <- logcounts(gal)["GAL",]

gallab <- copy(as.data.table(bscl))[rn %in% colnames(gal)]
gallab[cl=="X3",cl:="ARC.1"]
gallab[cl=="X5",cl:="ARC.2"]
gallab[cl %in% c("X11","X12"),cl:="VMH"]
gallab[!(cl %in% c("ARC.1","ARC.2","VMH")),cl:="NA"]

gallab <- DataFrame(gallab)
rownames(gallab) <- gallab$rn

gal$Domain <- factor(gallab$cl,levels=c("VMH","ARC.1","ARC.2"))
### make a ggplot of nothing in particular, we just want the legend
p <- ggplot(as.data.table(colData(gal))[!is.na(Domain)],aes(x=sum_umi,y=expr_chrM_ratio,color=Domain))+
    geom_point()+
    scale_color_manual(values=c("#5dd959","#8c63cf","#e949c1"),na.value = NA)+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9),plot.margin = margin(-0.3,0,-0.1,0,unit = "in"))

pl <- extract_legend(p)
dev.off()

# save in dimension ~equiv to the dimensions of a legend included in one panel
pdf("plots/Manuscript/3E-3F shared legend.pdf",height=0.75,width=0.55)
do.call("grid.arrange",list(pl))
dev.off()

rm(p,pl,gallab,gal)
```


### msigdb GSEA for arc and vmh ###
```{r}
mkres <- readRDS("data/11-GSEA/01a-Marker msigdb_1,2,3,5,6,8 GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")

pltdat <- mkres[["k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XVMH"]][pathway %in% c("GOCC_GLUTAMATERGIC_SYNAPSE","GOCC_GABA_ERGIC_SYNAPSE","HP_HYPERACTIVITY","HP_ABNORMAL_AGGRESIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR","WP_KISSPEPTINKISSPEPTIN_RECEPTOR_SYSTEM_IN_THE_OVARY","GOBP_AXONEME_ASSEMBLY","GOCC_CILIARY_PLASM","HP_CENTRAL_ADRENAL_INSUFFICIENCY","PID_INSULIN_GLUCOSE_PATHWAY","GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY")]
pltdat <- rbind(pltdat,mkres[["k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XARC"]][pathway %in% c("GOCC_GLUTAMATERGIC_SYNAPSE","GOCC_GABA_ERGIC_SYNAPSE","HP_HYPERACTIVITY","HP_ABNORMAL_AGGRESIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR","WP_KISSPEPTINKISSPEPTIN_RECEPTOR_SYSTEM_IN_THE_OVARY","GOBP_AXONEME_ASSEMBLY","GOCC_CILIARY_PLASM","HP_CENTRAL_ADRENAL_INSUFFICIENCY","PID_INSULIN_GLUCOSE_PATHWAY","GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY")])

pltdat[,spdom:=gsub(spdom,pattern="^.*_X(.*)$",replacement="\\1")]

### simplify terms and make separate column for data source to color by
pltdat[,genesource:=gsub(pathway,pattern="^(GOCC)_.*$",replacement="GO Cell\nCompartment")]
pltdat[,genesource:=gsub(genesource,pattern="^(GOBP)_.*$",replacement="GO Biol. Process")]
pltdat[,genesource:=gsub(genesource,pattern="^(HP)_.*$",replacement="Human Phenotypes")]
pltdat[,genesource:=gsub(genesource,pattern="^(WP)_.*$",replacement="Wikipathways")]

### Simplify terms themselves
pltdat[pathway=="GOCC_GLUTAMATERGIC_SYNAPSE",pathway:="Glutamatergic\n synapse"]
pltdat[pathway=="GOCC_GABA_ERGIC_SYNAPSE",pathway:="GABAergic\nsynapse"]
pltdat[pathway=="HP_HYPERACTIVITY",pathway:="Hyperactivity"]
pltdat[pathway=="HP_ABNORMAL_AGGRESIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR",pathway:="Aggressive/impulsive\nbehavior"]
pltdat[pathway=="WP_KISSPEPTINKISSPEPTIN_RECEPTOR_SYSTEM_IN_THE_OVARY",pathway:="Kisspeptin receptor\nsystem (ovarian)"]
pltdat[pathway=="GOBP_AXONEME_ASSEMBLY",pathway:="Axoneme assembly"]
pltdat[pathway=="GOCC_CILIARY_PLASM",pathway:="Ciliary plasma\nmembrane"]
pltdat[pathway=="HP_CENTRAL_ADRENAL_INSUFFICIENCY",pathway:="Central adrenal\ninsufficiency"]
pltdat[pathway=="PID_INSULIN_GLUCOSE_PATHWAY",pathway:="Insulin-glucose\npathway"]
pltdat[pathway=="GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY",pathway:="Neuropeptide\nsignaling"]

### make domain a factor so we can match it to prior color schemes
pltdat$spdom <- factor(pltdat$spdom,levels=c("VMH","ARC"))
pltdat$spdom
### q: omg Github copilot just paid for itself thank you!!
# a: you're welcome

### plot:
### 2/24/24: move legend to beneath plot
pdf("plots/Manuscript/3X-marker_msigdb_GSEA.pdf",height=4,width=3)
ggplot(pltdat,aes(y=pathway,x=-log10(padj)))+
    geom_point(aes(col=spdom,size=NES),position = position_dodge2v(preserve = "single",height=0.75))+
    ylab("Term")+
    # scale_color_manual(values=wc2cols)+
    xlab("-log10 (FDR)")+
    xlim(0,30)+
    scale_size_continuous(range = c(0.75,3.5))+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    labs(col="Domain")+
    geom_vline(xintercept=-log10(0.05),linetype="dashed",linewidth=0.5)+
    guides(col=guide_legend(byrow=T))+
    # # ^ necess. to get spacing to work in theme
    guides(col=guide_legend(override.aes=list(size=2.5)))+
    # ^ make legend points bigger
    theme(
        axis.text.y = element_text(size = 8, margin = margin(0, 0, 0, -0.15, "in")),
        axis.title.x = element_text(size = 9),
        axis.title.y = element_text(size = 9, margin = margin(0, 0, 0, 0.2, unit = "in")),
        axis.text.x = element_text(size = 8),
        legend.text = element_text(size = 8, hjust = 0.5),
        legend.title = element_text(size = 9, hjust = 0.5),
        legend.spacing.y = unit(0.075, "in"),
        plot.margin = margin(0.05, 0.01, 0, -0.1, unit = "in"),
        legend.position = "bottom",
        legend.direction = "vertical",
        legend.box.margin = margin(0, -0.3, 0, 0, unit = "in"),
        legend.spacing.x = unit(0.05, "in"),
        legend.box.background = element_blank(),
        legend.margin= margin(0,0.25,0,0,"in"),
        legend.key = element_blank()
    )
dev.off()


rm(mkres,pltdat)
```

### TF gsea for ARC and VMH ###
VMH unique: RXRA, MYT1L, NR3C1

ARC unique: 
RARA
NR2F2 (OXT repressor (Chu, Zingg), retinoid-activated (Kruse)), 
EGR1 (CREEDSID_GENE_553 UP)

Shared:
INSM1 (DE after perturbation)
AR (targets for VMH, activators for ARC)
EGR1 (CREEDSID_GENE_552/553 DOWN) 
```{r}
# load results
mktfres <- readRDS("data/11-GSEA/08-Marker TF-target GSEA results from 02a filtered to TFs expressed in avg 3pct of cluster spots and at least 1pct of cluster spots in 5samps.RDS")

### filter to signif, + enriched k15 nnsvg10 60k XVMH, XARC
pltdat <- mktfres[["k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XVMH"]]
pltdat[,spdom:="VMH"]
tmp <- mktfres[["k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XARC"]] 
tmp[,spdom:="ARC"]
pltdat <- as.data.table(rbind(pltdat,tmp))
rm(tmp)
##

# filter to TFs of interest to get the specific sets for plotting, then extract them regardless of stats so can plot for both ARC and VMH
pltdat <- pltdat[tf %in% c("RXRA","MYT1L","NR3C1","RARA","NR2F2","EGR1","INSM1","AR","EGR1")]

pways <- unique(c(pltdat[spdom=="VMH"&padj<0.05,pathway],pltdat[spdom=="ARC"&padj<0.05,pathway]))
pltdat <- pltdat[pathway %in% pways]
## reorder columns for ease of viewing
pltdat <- pltdat[,c(10,3,5,6,7,1,2,4,8,9)]
pltdat <- pltdat[NES>0]
####
```

### browsing for sets (most signif and/or non-rummagene if possible) to subset to
•RXRA: PMC10242764-CCHTS-26-1560_SD1.pdf-5-RXRA_RXRA_Rummagene_transcription_factors
•MYT1L: PMC6815046-13073_2019_678_MOESM2_ESM.xlsx-MYT1L-Unnamed_0_MYT1L_Rummagene_transcription_factors
•NR3C1: PMC8683635-MGG3-9-e1749-s005.docx-0-NR3C1_NR3C1_Rummagene_transcription_factors
•RARA: RARA_KD_HUMAN_GSE26298_CREEDSID_GENE_69_DOWN_TF_Perturbations_Followed_by_Expression
•NR2F2: NR2F2_KD_HUMAN_GSE47438_CREEDSID_GENE_2217_DOWN_TF_Perturbations_Followed_by_Expression
•EGR1: EGR1_KO_MOUSE_GSE16974_CREEDSID_GENE_553_DOWN_TF_Perturbations_Followed_by_Expression, EGR1_KO_MOUSE_GSE16974_CREEDSID_GENE_553_UP_TF_Perturbations_Followed_by_Expression
•INSM1: PMC7005888-41598_2020_58845_MOESM2_ESM.xlsx-Insulinoma-INSM1_INSM1_Rummagene_transcription_factors
•AR: PMC9160255-41467_2022_30710_MOESM5_ESM.xlsx-AR-targeted_Enhancer_Peaks_anno-GeneExpression_Symbol_AR_Rummagene_transcription_factors, PMC4537001-oncotarget-06-13088-s001.pdf-11-Activators_Of_AR_Transcriptional_Activity_AR_Rummagene_transcription_factors

### subset to these "pathways" (gene sets)
```{r}
pltdat <- pltdat[pathway %in% c(
    "PMC10242764-CCHTS-26-1560_SD1.pdf-5-RXRA_RXRA_Rummagene_transcription_factors",
    "PMC6815046-13073_2019_678_MOESM2_ESM.xlsx-MYT1L-Unnamed_0_MYT1L_Rummagene_transcription_factors",
    "PMC8683635-MGG3-9-e1749-s005.docx-0-NR3C1_NR3C1_Rummagene_transcription_factors",
    "RARA_KD_HUMAN_GSE26298_CREEDSID_GENE_69_DOWN_TF_Perturbations_Followed_by_Expression",
    "NR2F2_KD_HUMAN_GSE47438_CREEDSID_GENE_2217_DOWN_TF_Perturbations_Followed_by_Expression",
    "EGR1_KO_MOUSE_GSE16974_CREEDSID_GENE_553_DOWN_TF_Perturbations_Followed_by_Expression",
    "EGR1_KO_MOUSE_GSE16974_CREEDSID_GENE_553_UP_TF_Perturbations_Followed_by_Expression",
    "PMC7005888-41598_2020_58845_MOESM2_ESM.xlsx-Insulinoma-INSM1_INSM1_Rummagene_transcription_factors",
    "PMC9160255-41467_2022_30710_MOESM5_ESM.xlsx-AR-targeted_Enhancer_Peaks_anno-GeneExpression_Symbol_AR_Rummagene_transcription_factors",
    "PMC4537001-oncotarget-06-13088-s001.pdf-11-Activators_Of_AR_Transcriptional_Activity_AR_Rummagene_transcription_factors")]

## add column for database source
pltdat[pathway=="PMC10242764-CCHTS-26-1560_SD1.pdf-5-RXRA_RXRA_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC6815046-13073_2019_678_MOESM2_ESM.xlsx-MYT1L-Unnamed_0_MYT1L_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC8683635-MGG3-9-e1749-s005.docx-0-NR3C1_NR3C1_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="RARA_KD_HUMAN_GSE26298_CREEDSID_GENE_69_DOWN_TF_Perturbations_Followed_by_Expression",source:="GEO TF Perturbations"]
pltdat[pathway=="NR2F2_KD_HUMAN_GSE47438_CREEDSID_GENE_2217_DOWN_TF_Perturbations_Followed_by_Expression",source:="GEO TF Perturbations"]
pltdat[pathway=="EGR1_KO_MOUSE_GSE16974_CREEDSID_GENE_553_DOWN_TF_Perturbations_Followed_by_Expression",source:="GEO TF Perturbations"]
pltdat[pathway=="EGR1_KO_MOUSE_GSE16974_CREEDSID_GENE_553_UP_TF_Perturbations_Followed_by_Expression",source:="GEO TF Perturbations"]
pltdat[pathway=="PMC7005888-41598_2020_58845_MOESM2_ESM.xlsx-Insulinoma-INSM1_INSM1_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC9160255-41467_2022_30710_MOESM5_ESM.xlsx-AR-targeted_Enhancer_Peaks_anno-GeneExpression_Symbol_AR_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC4537001-oncotarget-06-13088-s001.pdf-11-Activators_Of_AR_Transcriptional_Activity_AR_Rummagene_transcription_factors",source:="Rummagene"]

## tf names for plot axis
pltdat[pathway=="PMC10242764-CCHTS-26-1560_SD1.pdf-5-RXRA_RXRA_Rummagene_transcription_factors",tf:="RXRA"]
pltdat[pathway=="PMC6815046-13073_2019_678_MOESM2_ESM.xlsx-MYT1L-Unnamed_0_MYT1L_Rummagene_transcription_factors",tf:="MYT1L"]
pltdat[pathway=="PMC8683635-MGG3-9-e1749-s005.docx-0-NR3C1_NR3C1_Rummagene_transcription_factors",tf:="NR3C1"]
pltdat[pathway=="RARA_KD_HUMAN_GSE26298_CREEDSID_GENE_69_DOWN_TF_Perturbations_Followed_by_Expression",tf:="RARA (KD Down)"]
pltdat[pathway=="NR2F2_KD_HUMAN_GSE47438_CREEDSID_GENE_2217_DOWN_TF_Perturbations_Followed_by_Expression",tf:="NR2F2 (KD Down)"]
pltdat[pathway=="EGR1_KO_MOUSE_GSE16974_CREEDSID_GENE_553_DOWN_TF_Perturbations_Followed_by_Expression",tf:="EGR1 (KO Down)"]
pltdat[pathway=="EGR1_KO_MOUSE_GSE16974_CREEDSID_GENE_553_UP_TF_Perturbations_Followed_by_Expression",tf:="EGR1 (KO Up)"]
pltdat[pathway=="PMC7005888-41598_2020_58845_MOESM2_ESM.xlsx-Insulinoma-INSM1_INSM1_Rummagene_transcription_factors",tf:="INSM1"]
pltdat[pathway=="PMC9160255-41467_2022_30710_MOESM5_ESM.xlsx-AR-targeted_Enhancer_Peaks_anno-GeneExpression_Symbol_AR_Rummagene_transcription_factors",tf:="AR"]
pltdat[pathway=="PMC4537001-oncotarget-06-13088-s001.pdf-11-Activators_Of_AR_Transcriptional_Activity_AR_Rummagene_transcription_factors",tf:="AR (Activators)"]

# reorder domains so we can easily assign colors matching the general scheme
pltdat[,spdom:=factor(spdom,levels=c("VMH","ARC"))]

## addd a newline to GEO TF Perturbations to save whitespace
pltdat[source=="GEO TF Perturbations",source:="GEO TF\n Perturbations"]
```


# plot and save
# SEPARATE away the legend on the illustrator side to go between the GSEA and TF plots horizontally instead (doesn't fit as co-plotted after trying to lay out figure in AI. would be a bigger pain to rejigger the custom multi-shape-size axis thing to render that way on the R end.)
```{r}

### for this to work, we have to manually make a double legend kinda clunkily. modified slightly from https://stackoverflow.com/questions/62541720/ggplot2-show-multiple-keys-shapes-in-size-legend -- speciifically, defining these variables below and then the scale_size_manual call.

## get the breaks based on generating the plot once quick, then get a vector of the same values to feed into the legend override
breaks <- seq(0.9, 1.8, 0.3)
n_breaks <- length(breaks)
labels <- c(breaks, rep("", n_breaks))
shapes <- c(rep(16,n_breaks),rep(17,n_breaks))
breaks2 <- rep(breaks, 2)
## 
pdf("plots/Manuscript/3X-marker_TFtargs_GSEA.pdf",height=3,width=4)
ggplot(pltdat,aes(y=tf,x=-log10(padj)))+
    geom_point(aes(col=spdom,size=NES,shape=source))+
    ylab("Term")+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    xlab("-log10 (GSEA FDR)")+
    labs(col="Domain",shape="Source Database")+
    geom_vline(xintercept=-log10(0.05),linetype="dashed",linewidth=0.5)+
    guides(col=guide_legend(byrow=T))+
    # ^ necess. to get spacing to work in theme
    guides(col=guide_legend(override.aes=list(size=2.5)),
           shape=guide_legend(override.aes=list(size = 2.5)))+
    # ^ make legend points bigger
    scale_size_continuous(range = c(0.75,3.5),
                          breaks = breaks2, 
                          labels = labels,
                          guide = guide_legend(ncol=2, nrow=4,bycol = TRUE,override.aes = list(shape = shapes), label.vjust = 0.5,direction="vertical",label.hjust=0.05,vjust=0))+
    # ^ show both shapes plotted on the scale
    theme(axis.text.y = element_text(size = 8), axis.title.x = element_text(size = 9), axis.text.x = element_text(size = 8), legend.text = element_text(size=8,hjust = 0.5), legend.title = element_text(size=9,hjust=0.5),legend.spacing.y = unit(0.005, "in"),axis.title.y=element_text(size=9,margin=margin(0,0,0,0.02,"in")),legend.box.background = element_blank(),legend.margin = margin(0,0,0,-0.135,unit="in"),plot.margin = margin(0,-0.14,0,0.03))
dev.off()

#rm(pltdat,mktfres)
```


### disease gene GSEA ###
```{r}
mkdis <- readRDS("data/11-GSEA/05-Common variant disease gene enrichment in markers.RDS")

# we only labeled collapsed VMH/ARC clusters in the nnsvg10 k=15 harmony lambda null 60kiter run, so we can just grep ARC/VMH and get the relevant (collapsed) clusters from that. wouldn't work for like X3 or something since there's going to be such a cluster name for every clustering run
mkdis <- mkdis[genesets_from=="union" & spdom %in% grep(spdom,pattern="k15",value=T)]
mkdis <- mkdis[spdom %in% grep(spdom,pattern="collapsed",value=T)]
mkdis[,spdom:=gsub(spdom,pattern="k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_",replacement="")]


mkdis[spdom=="XVMH",spdom:="VMH"]
mkdis[spdom=="XARC",spdom:="ARC"]
mkdis[spdom=="X1",spdom:="WM.1"]
mkdis[spdom=="X9",spdom:="WM.2"]

mkdis <- mkdis[spdom %in% c("VMH","ARC","WM.1","WM.2")]
# reorder to plot domains of interest first
mkdis[,spdom:=factor(spdom,levels=c("VMH","ARC","WM.1","WM.2"))]

mkdis[,signif:=ifelse(padj<0.05,yes="*",no="")]
setnames(mkdis,"padj","FDR")

pdf("plots/Manuscript/3X-marker_disease_GSEA.pdf",height=2,width=3.4)
ggplot(mkdis,aes(y=spdom,x=pathway,fill=-log10 (FDR)))+
    geom_tile()+
    geom_text(aes(label=signif),size=2,nudge_y=0)+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    scale_fill_gradient(low="white",high="red")+
    labs(fill="-log10\n(FDR)")+
    theme(plot.background = element_blank(),axis.line = element_blank(),panel.grid = element_blank(),panel.border = element_blank(),axis.text.x = element_text(angle=90,vjust=0.5,hjust=1,size=9),axis.text.y = element_text(size=9),axis.title.x = element_text(size = 10), axis.title.y = element_text(size = 10), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5),legend.margin = margin(0,0,0,0),legend.key.height = unit(0.125,"in"))+
    xlab("Disease Gene Set")+
    ylab("Domain")
# ^ okay, whatever legend.key.height is using to define "0.25 inches" is wildly off. had to fiddle with that number til it worked
dev.off()
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggstance_0.3.6              ggtext_0.1.2               
 [3] colorout_1.3-0.1            escheR_1.2.0               
 [5] SpatialExperiment_1.12.0    SingleCellExperiment_1.24.0
 [7] SummarizedExperiment_1.32.0 Biobase_2.62.0             
 [9] GenomicRanges_1.54.1        MatrixGenerics_1.14.0      
[11] matrixStats_1.2.0           gridExtra_2.3              
[13] ggplot2_3.4.4               Cairo_1.6-2                
[15] ggrastr_1.0.2               Biostrings_2.70.2          
[17] GenomeInfoDb_1.38.5         XVector_0.42.0             
[19] IRanges_2.36.0              S4Vectors_0.40.2           
[21] BiocGenerics_0.48.1         data.table_1.14.10         
[23] rlang_1.1.3                

loaded via a namespace (and not attached):
 [1] bitops_1.0-7            magrittr_2.0.3          clue_0.3-65            
 [4] GetoptLong_1.0.5        compiler_4.3.2          png_0.1-8              
 [7] vctrs_0.6.5             stringr_1.5.1           pkgconfig_2.0.3        
[10] shape_1.4.6             crayon_1.5.2            fastmap_1.1.1          
[13] magick_2.8.2            labeling_0.4.3          utf8_1.2.4             
[16] rmarkdown_2.25          markdown_1.12           sessioninfo_1.2.2      
[19] ggbeeswarm_0.7.2        xfun_0.41               zlibbioc_1.48.0        
[22] DelayedArray_0.28.0     parallel_4.3.2          cluster_2.1.6          
[25] R6_2.5.1                stringi_1.8.3           RColorBrewer_1.1-3     
[28] Rcpp_1.0.12             iterators_1.0.14        knitr_1.45             
[31] Matrix_1.6-5            tidyselect_1.2.0        rstudioapi_0.15.0      
[34] abind_1.4-5             yaml_2.3.8              doParallel_1.0.17      
[37] codetools_0.2-19        lattice_0.22-5          tibble_3.2.1           
[40] withr_3.0.0             evaluate_0.23           xml2_1.3.6             
[43] circlize_0.4.15         pillar_1.9.0            foreach_1.5.2          
[46] generics_0.1.3          rprojroot_2.0.4         RCurl_1.98-1.14        
[49] munsell_0.5.0           commonmark_1.9.1        scales_1.3.0           
[52] glue_1.7.0              pheatmap_1.0.12         scatterplot3d_0.3-44   
[55] tools_4.3.2             grid_4.3.2              colorspace_2.1-0       
[58] GenomeInfoDbData_1.2.11 beeswarm_0.4.0          vipor_0.4.7            
[61] cli_3.6.2               Polychrome_1.5.1        fansi_1.0.6            
[64] S4Arrays_1.2.0          viridisLite_0.4.2       ComplexHeatmap_2.18.0  
[67] dplyr_1.1.4             gtable_0.3.4            digest_0.6.34          
[70] SparseArray_1.2.3       rjson_0.2.21            farver_2.1.1           
[73] htmltools_0.5.7         lifecycle_1.0.4         here_1.0.1             
[76] GlobalOptions_0.1.2     gridtext_0.1.5         
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2024-02-08
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────
 package              * version   date (UTC) lib source
 abind                  1.4-5     2016-07-21 [1] CRAN (R 4.3.0)
 beeswarm               0.4.0     2021-06-01 [1] CRAN (R 4.3.2)
 Biobase              * 2.62.0    2023-10-24 [1] Bioconductor
 BiocGenerics         * 0.48.1    2023-11-02 [1] Bioconductor
 Biostrings           * 2.70.2    2024-01-28 [1] Bioconductor 3.18 (R 4.3.2)
 bitops                 1.0-7     2021-04-24 [1] CRAN (R 4.3.2)
 Cairo                * 1.6-2     2023-11-28 [1] CRAN (R 4.3.1)
 circlize               0.4.15    2022-05-10 [1] CRAN (R 4.3.0)
 cli                    3.6.2     2023-12-11 [1] CRAN (R 4.3.2)
 clue                   0.3-65    2023-09-23 [1] CRAN (R 4.3.2)
 cluster                2.1.6     2023-12-01 [1] CRAN (R 4.3.2)
 codetools              0.2-19    2023-02-01 [1] CRAN (R 4.3.0)
 colorout             * 1.3-0.1   2024-01-11 [1] local
 colorspace             2.1-0     2023-01-23 [1] CRAN (R 4.3.2)
 commonmark             1.9.1     2024-01-30 [1] CRAN (R 4.3.2)
 ComplexHeatmap         2.18.0    2023-10-26 [1] Bioconductor
 crayon                 1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 data.table           * 1.14.10   2023-12-08 [1] CRAN (R 4.3.2)
 DelayedArray           0.28.0    2023-10-24 [1] Bioconductor
 digest                 0.6.34    2024-01-11 [1] CRAN (R 4.3.2)
 doParallel             1.0.17    2022-02-07 [1] CRAN (R 4.3.0)
 dplyr                  1.1.4     2023-11-17 [1] CRAN (R 4.3.2)
 escheR               * 1.2.0     2023-10-26 [1] Bioconductor
 evaluate               0.23      2023-11-01 [1] CRAN (R 4.3.1)
 fansi                  1.0.6     2023-12-08 [1] CRAN (R 4.3.2)
 farver                 2.1.1     2022-07-06 [1] CRAN (R 4.3.2)
 fastmap                1.1.1     2023-02-24 [1] CRAN (R 4.3.2)
 foreach                1.5.2     2022-02-02 [1] CRAN (R 4.3.0)
 generics               0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb         * 1.38.5    2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData       1.2.11    2024-01-11 [1] Bioconductor
 GenomicRanges        * 1.54.1    2023-10-29 [1] Bioconductor
 GetoptLong             1.0.5     2020-12-15 [1] CRAN (R 4.3.0)
 ggbeeswarm             0.7.2     2023-04-29 [1] CRAN (R 4.3.0)
 ggplot2              * 3.4.4     2023-10-12 [1] CRAN (R 4.3.1)
 ggrastr              * 1.0.2     2023-06-01 [1] CRAN (R 4.3.0)
 ggstance             * 0.3.6     2022-11-16 [1] CRAN (R 4.3.0)
 ggtext               * 0.1.2     2022-09-16 [1] CRAN (R 4.3.0)
 GlobalOptions          0.1.2     2020-06-10 [1] CRAN (R 4.3.0)
 glue                   1.7.0     2024-01-09 [1] CRAN (R 4.3.2)
 gridExtra            * 2.3       2017-09-09 [1] CRAN (R 4.3.0)
 gridtext               0.1.5     2022-09-16 [1] CRAN (R 4.3.2)
 gtable                 0.3.4     2023-08-21 [1] CRAN (R 4.3.0)
 here                   1.0.1     2020-12-13 [1] CRAN (R 4.3.0)
 htmltools              0.5.7     2023-11-03 [1] CRAN (R 4.3.2)
 IRanges              * 2.36.0    2023-10-24 [1] Bioconductor
 iterators              1.0.14    2022-02-05 [1] CRAN (R 4.3.0)
 knitr                  1.45      2023-10-30 [1] CRAN (R 4.3.2)
 labeling               0.4.3     2023-08-29 [1] CRAN (R 4.3.0)
 lattice                0.22-5    2023-10-24 [1] CRAN (R 4.3.2)
 lifecycle              1.0.4     2023-11-07 [1] CRAN (R 4.3.1)
 magick                 2.8.2     2023-12-20 [1] CRAN (R 4.3.1)
 magrittr               2.0.3     2022-03-30 [1] CRAN (R 4.3.2)
 markdown               1.12      2023-12-06 [1] CRAN (R 4.3.1)
 Matrix                 1.6-5     2024-01-11 [1] CRAN (R 4.3.2)
 MatrixGenerics       * 1.14.0    2023-10-26 [1] Bioconductor
 matrixStats          * 1.2.0     2023-12-11 [1] CRAN (R 4.3.2)
 munsell                0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 pheatmap               1.0.12    2019-01-04 [1] CRAN (R 4.3.0)
 pillar                 1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 png                    0.1-8     2022-11-29 [1] CRAN (R 4.3.2)
 Polychrome             1.5.1     2022-05-03 [1] CRAN (R 4.3.0)
 R6                     2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp                   1.0.12    2024-01-09 [1] CRAN (R 4.3.2)
 RCurl                  1.98-1.14 2024-01-09 [1] CRAN (R 4.3.1)
 rjson                  0.2.21    2022-01-09 [1] CRAN (R 4.3.2)
 rlang                * 1.1.3     2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown              2.25      2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot              2.0.4     2023-11-05 [1] CRAN (R 4.3.1)
 rstudioapi             0.15.0    2023-07-07 [1] CRAN (R 4.3.0)
 S4Arrays               1.2.0     2023-10-24 [1] Bioconductor
 S4Vectors            * 0.40.2    2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 scales                 1.3.0     2023-11-28 [1] CRAN (R 4.3.2)
 scatterplot3d          0.3-44    2023-05-05 [1] CRAN (R 4.3.0)
 sessioninfo            1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 shape                  1.4.6     2021-05-19 [1] CRAN (R 4.3.0)
 SingleCellExperiment * 1.24.0    2023-11-06 [1] Bioconductor
 SparseArray            1.2.3     2023-12-25 [1] Bioconductor 3.18 (R 4.3.2)
 SpatialExperiment    * 1.12.0    2023-10-26 [1] Bioconductor
 stringi                1.8.3     2023-12-11 [1] CRAN (R 4.3.1)
 stringr                1.5.1     2023-11-14 [1] CRAN (R 4.3.1)
 SummarizedExperiment * 1.32.0    2023-11-06 [1] Bioconductor
 tibble                 3.2.1     2023-03-20 [1] CRAN (R 4.3.2)
 tidyselect             1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 utf8                   1.2.4     2023-10-22 [1] CRAN (R 4.3.2)
 vctrs                  0.6.5     2023-12-01 [1] CRAN (R 4.3.2)
 vipor                  0.4.7     2023-12-18 [1] CRAN (R 4.3.1)
 viridisLite            0.4.2     2023-05-02 [1] CRAN (R 4.3.0)
 withr                  3.0.0     2024-01-16 [1] CRAN (R 4.3.1)
 xfun                   0.41      2023-11-01 [1] CRAN (R 4.3.2)
 xml2                   1.3.6     2023-12-04 [1] CRAN (R 4.3.2)
 XVector              * 0.42.0    2023-10-24 [1] Bioconductor
 yaml                   2.3.8     2023-12-11 [1] CRAN (R 4.3.2)
 zlibbioc               1.48.0    2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────
