---
title: "01-Tabulate_nucCellUnassigned_Xscript_Props"
output: html_document
date: "2024-09-11"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(MoleculeExperiment)
library(SpatialExperiment)
library(SpatialFeatureExperiment)
## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel)
## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

### to get transcript-level data, we will use the slightly clunky but efficient MoleculeExperiment package to read in transcript detections, subset them by quality score >20 (qv--the raw transcripts-detected-by-location-table does not filter for this, though the tabulated counts per cell we've used up to now already are filtered thusly), and then tabulate the counts within the nuclear boundaries, within the cell boundaries, and those not assigned a cell. 

the key xenium output files being used behind the scenes here are in processed-data/01/[resegmented sample]/outs/transcripts.csv.gz and nucleus_boundaries.csv.gz .

### performance notes: request 2 more cores than specified in countMolecules below; peak combined RAM usage with 10 cores approx 80GB

## define tabulation function
```{r}
xenouts <- list.files(pattern="wnuclei",path="processed-data/01_xeniumranger1.7-resegment",full.names=T)
xenouts <- paste0(xenouts,"/outs")

xentabulator <- function(filepaths,cellid_filter=NULL,suffix="",mcc=1){
    xens <- lapply(xenouts,FUN=function(x){
        xendat <- MoleculeExperiment::readXenium(x,keepCols="all",addBoundaries=c("nucleus","cell"))
        
        ## here, we'll retain controls since compartment proportions might also be informative.
        
        ## filter by qv
        xendat@molecules$detected$outs <- lapply(xendat@molecules$detected$outs, FUN=function(g){return(g[which(g$qv>20),])})
        
        ## get sample name matched to the sample names we've been using; make cell_ids of this data match the ones that were generated in building the SPE (ie sample short id _ cell _id)
        sampn <- gsub(x,pattern=".*02_(.*)",replacement="\\1")
        sampn <- gsub(sampn,pattern="|.*03_(.*)",replacement="\\1")
        sampn <- gsub(sampn,pattern="^(.*)_reseg.*$",replacement="\\1")
        sampn <- gsub(sampn,pattern="^(.*)_xr17.*$",replacement="\\1")
        sampn <- paste0("X",sampn)
        
        ## if a cell id filter was provided, further filter by that
        if (!is.null(cellid_filter)){
            # disable DT multithreading during mclapply to prevent resource competition, reenabling multithread after (restore_after_fork=T)
            setDTthreads(1,restore_after_fork = TRUE)
            xendat@molecules$detected$outs <- mclapply(xendat@molecules$detected$outs,mc.cores=mcc,FUN=function(x){
                y <- as.data.table(x)
                # make cell_ids their sample specific equivalents for matching to filtered spe cell names
                y[cell_id!="UNASSIGNED",cell_id:=paste0(sampn,"_",cell_id)]
                
                y <- y[cell_id %in% c("UNASSIGNED",cellid_filter),]
                return(y)
            })
        }
        

        
        ## tabulate for each gene by compartment. this can be done with lapply using the fields in molecules$detected$outs$GENE$overlaps_nucleus and $cell_id
        genetabuls <- mapply(g=xendat@molecules$detected$outs,n=names(xendat@molecules$detected$outs),SIMPLIFY=FALSE,FUN=function(g,n){
            totx <- nrow(g)
            nucx <- nrow(as.data.table(g)[overlaps_nucleus==1])
            cellx <- nrow(as.data.table(g)[cell_id!="UNASSIGNED"])
            extranuc <- cellx-nucx
            noass <- totx-cellx
            return(data.table(samp=sampn,gene=n,total=totx,nucleus=nucx,cell=cellx,extranuc=extranuc,noass=noass))
        })
        
        genetabuls <- rbindlist(genetabuls)
        return(genetabuls)
    })
    
    # now we have a list with one table for all genes in a given sample, with a sample column already baked in. so we can just rbindlist those and we've got a samplewise-transcriptwise table.
    tabuls <- rbindlist(xens)
    
    # fix the fact that nontargeting probes previously called BLANK_nnnn were renamed to UnassignedCodeword at some point in xenium's development. The assigned numbers do not change, i.e. BLANK_0001 just became UnassignedCodeword_0001.
    tabuls[,gene:=gsub(gene,pattern="BLANK",replacement="UnassignedCodeword")]
    
    ## calculate proportions in each compartment (nuclear, extranuclear, unassigned) by sample
    tabuls[,propNuc:=nucleus/total]
    tabuls[,propExtranuc:=extranuc/total]
    tabuls[,propUnassigned:=noass/total]
    
    ### make a second table summing these metrics up by gene across all samples
    tabuls_dset <- tabuls[,lapply(.SD,sum),by=gene,.SDcols=c("total","nucleus","cell","extranuc","noass")]
    
    tabuls_dset[,propNuc:=nucleus/total]
    tabuls_dset[,propExtranuc:=extranuc/total]
    tabuls_dset[,propUnassigned:=noass/total]
    
    retlist <- list(tabuls,tabuls_dset)
    names(retlist) <- paste0(c("samplewise_","datasetWide_"),suffix)
    
    return(retlist)
}
```

# run for the original data only filter for transcript quality > 20, the 870870 cells clustered with banksy after QC, and the 819k cells retained after clustering and removing sample/donor specific clusters
```{r}
unfilttabuls <- xentabulator(xenouts,cellid_filter = NULL,suffix="unfilt",mcc=4)

## filter to cells passing QC (i.e., those input to banksy) and do the same. easiest way to get those cell/nuc ids is from the banksy output. sadly we need to reiterate backthru the transcript data loading and filter at that level as we're not returning the full transcript-level data for performance reasons.
bkcl <- fread("processed-data/05_Banksy_M0lam0_res2_multisamp/01-BanksyClusts_M0lam0_leiden_multisamp.txt")

qcfilttabuls <- xentabulator(xenouts[1:2],cellid_filter=bkcl$rn,suffix="qcfilt",mcc=4)


## filter to those cells from non-sample/donor specific clusters (i.e., data used for subsequent analyses). we can get the clusters to drop from the cluster annotations
annots <- fread("processed-data/05_Banksy_M0lam0_res2_multisamp/02-M0l0kg6_topClusMarkers_celltypes_annotated.txt")
annots <- unique(annots[,.(clus,bjm_annot)])

setnames(bkcl,2,"cl")
bkcl <- merge.data.table(bkcl,annots,by.x="cl",by.y="clus")
bkcl <- bkcl[!(bjm_annot %in% grep(bjm_annot,pattern="DISCARD",value=T))]

retaintabuls <- xentabulator(xenouts,cellid_filter=bkcl$rn,suffix="retained",mcc = 4)
```

## save the results
```{r}
reslist <- c(unfilttabuls,qcfilttabuls,retaintabuls)
saveRDS(reslist,"processed-data/12-Xscript_level_misc_VisVMHsvgs_nucCellUnassignedProps/01-Nuclear_Cellular_Extracellular_TranscriptsProps_preQC-postQC-retainedClusters.RDS")
```

reprod info
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.38.0            parallelly_1.38.0             
 [3] colorout_1.3-0.2               SpatialFeatureExperiment_1.6.1
 [5] SpatialExperiment_1.14.0       SingleCellExperiment_1.26.0   
 [7] SummarizedExperiment_1.34.0    Biobase_2.64.0                
 [9] GenomicRanges_1.56.1           GenomeInfoDb_1.40.1           
[11] IRanges_2.38.1                 S4Vectors_0.42.1              
[13] BiocGenerics_0.50.0            MatrixGenerics_1.16.0         
[15] matrixStats_1.3.0              MoleculeExperiment_1.4.1      
[17] data.table_1.15.4              rlang_1.1.4                   

loaded via a namespace (and not attached):
 [1] DBI_1.2.3                 bitops_1.0-7              deldir_2.0-4             
 [4] s2_1.1.6                  magrittr_2.0.3            e1071_1.7-14             
 [7] compiler_4.4.1            DelayedMatrixStats_1.26.0 png_0.1-8                
[10] sfheaders_0.4.4           fftwtools_0.9-11          vctrs_0.6.5              
[13] pkgconfig_2.0.3           wk_0.9.2                  crayon_1.5.3             
[16] fastmap_1.2.0             magick_2.8.3              XVector_0.44.0           
[19] scuttle_1.14.0            utf8_1.2.4                rmarkdown_2.27           
[22] sessioninfo_1.2.2         UCSC.utils_1.0.0          purrr_1.0.2              
[25] bit_4.0.5                 xfun_0.45                 zlibbioc_1.50.0          
[28] beachmat_2.20.0           jsonlite_1.8.8            rhdf5filters_1.16.0      
[31] DelayedArray_0.30.1       Rhdf5lib_1.26.0           jpeg_0.1-10              
[34] tiff_0.1-12               terra_1.7-78              R6_2.5.1                 
[37] limma_3.60.3              boot_1.3-30               Rcpp_1.0.13              
[40] knitr_1.48                R.utils_2.12.3            Matrix_1.7-0             
[43] tidyselect_1.2.1          rstudioapi_0.16.0         abind_1.4-5              
[46] yaml_2.3.9                EBImage_4.46.0            codetools_0.2-20         
[49] lattice_0.22-6            tibble_3.2.1              withr_3.0.0              
[52] evaluate_0.24.0           sf_1.0-16                 units_0.8-5              
[55] spData_2.3.1              proxy_0.4-27              pillar_1.9.0             
[58] KernSmooth_2.23-24        generics_0.1.3            rprojroot_2.0.4          
[61] sp_2.1-4                  RCurl_1.98-1.14           sparseMatrixStats_1.16.0 
[64] class_7.3-22              glue_1.7.0                tools_4.4.1              
[67] BiocNeighbors_1.22.0      locfit_1.5-9.10           rhdf5_2.48.0             
[70] grid_4.4.1                spdep_1.3-5               DropletUtils_1.24.0      
[73] edgeR_4.2.0               GenomeInfoDbData_1.2.12   HDF5Array_1.32.0         
[76] cli_3.6.3                 fansi_1.0.6               S4Arrays_1.4.1           
[79] dplyr_1.1.4               R.methodsS3_1.8.2         zeallot_0.1.0            
[82] digest_0.6.36             classInt_0.4-10           SparseArray_1.4.8        
[85] dqrng_0.4.1               rjson_0.2.21              htmlwidgets_1.6.4        
[88] htmltools_0.5.8.1         R.oo_1.26.0               lifecycle_1.0.4          
[91] httr_1.4.7                here_1.0.1                statmod_1.5.0            
[94] bit64_4.0.5              
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-09-11
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────────────
 ! package                  * version   date (UTC) lib source
   abind                      1.4-5     2016-07-21 [1] CRAN (R 4.4.0)
   beachmat                   2.20.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   Biobase                  * 2.64.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocGenerics             * 0.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocNeighbors              1.22.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel             * 1.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   bit                        4.0.5     2022-11-15 [1] CRAN (R 4.4.0)
   bit64                      4.0.5     2020-08-30 [1] CRAN (R 4.4.0)
   bitops                     1.0-7     2021-04-24 [1] CRAN (R 4.4.0)
   boot                       1.3-30    2024-02-26 [1] CRAN (R 4.4.1)
   class                      7.3-22    2023-05-03 [1] CRAN (R 4.4.1)
   classInt                   0.4-10    2023-09-05 [1] CRAN (R 4.4.0)
 P cli                        3.6.3     2024-06-21 [2] CRAN (R 4.4.0)
   codetools                  0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorout                 * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   crayon                     1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   data.table               * 1.15.4    2024-03-30 [2] CRAN (R 4.4.0)
   DBI                        1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   DelayedArray               0.30.1    2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats         1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   deldir                     2.0-4     2024-02-28 [1] CRAN (R 4.4.0)
   digest                     0.6.36    2024-06-23 [1] CRAN (R 4.4.0)
   dplyr                      1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   dqrng                      0.4.1     2024-05-28 [1] CRAN (R 4.4.0)
   DropletUtils               1.24.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   e1071                      1.7-14    2023-12-06 [1] CRAN (R 4.4.0)
   EBImage                    4.46.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   edgeR                      4.2.0     2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate                   0.24.0    2024-06-10 [1] CRAN (R 4.4.0)
   fansi                      1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   fastmap                    1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fftwtools                  0.9-11    2021-03-01 [1] CRAN (R 4.4.0)
   generics                   0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb             * 1.40.1    2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData           1.2.12    2024-05-01 [1] Bioconductor
   GenomicRanges            * 1.56.1    2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   glue                       1.7.0     2024-01-09 [1] CRAN (R 4.4.0)
   HDF5Array                  1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   here                       1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools                  0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets                1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httr                       1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   IRanges                  * 2.38.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   jpeg                       0.1-10    2022-11-29 [1] CRAN (R 4.4.0)
   jsonlite                   1.8.8     2023-12-04 [1] CRAN (R 4.4.0)
   KernSmooth                 2.23-24   2024-05-17 [1] CRAN (R 4.4.1)
   knitr                      1.48      2024-07-07 [1] CRAN (R 4.4.1)
   lattice                    0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lifecycle                  1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                      3.60.3    2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                     1.5-9.10  2024-06-24 [1] CRAN (R 4.4.0)
   magick                     2.8.3     2024-02-18 [1] CRAN (R 4.4.0)
   magrittr                   2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   Matrix                     1.7-0     2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics           * 1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats              * 1.3.0     2024-04-11 [1] CRAN (R 4.4.0)
   MoleculeExperiment       * 1.4.1     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   parallelly               * 1.38.0    2024-07-27 [1] CRAN (R 4.4.0)
   pillar                     1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig                  2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   png                        0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   proxy                      0.4-27    2022-06-09 [1] CRAN (R 4.4.0)
   purrr                      1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R.methodsS3                1.8.2     2022-06-13 [1] CRAN (R 4.4.0)
   R.oo                       1.26.0    2024-01-24 [1] CRAN (R 4.4.0)
   R.utils                    2.12.3    2023-11-18 [1] CRAN (R 4.4.0)
   R6                         2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   Rcpp                       1.0.13    2024-07-17 [1] CRAN (R 4.4.0)
   RCurl                      1.98-1.14 2024-01-09 [1] CRAN (R 4.4.0)
   rhdf5                      2.48.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rhdf5filters               1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Rhdf5lib                   1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rjson                      0.2.21    2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                    * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown                  2.27      2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot                  2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   rstudioapi                 0.16.0    2024-03-24 [1] CRAN (R 4.4.0)
   s2                         1.1.6     2023-12-19 [1] CRAN (R 4.4.0)
   S4Arrays                   1.4.1     2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors                * 0.42.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   scuttle                    1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo                1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   sf                         1.0-16    2024-03-24 [1] CRAN (R 4.4.0)
   sfheaders                  0.4.4     2024-01-17 [1] CRAN (R 4.4.0)
   SingleCellExperiment     * 1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sp                         2.1-4     2024-04-30 [1] CRAN (R 4.4.0)
   SparseArray                1.4.8     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats          1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment        * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialFeatureExperiment * 1.6.1     2024-05-15 [1] Bioconductor 3.19 (R 4.4.0)
   spData                     2.3.1     2024-05-31 [1] CRAN (R 4.4.0)
   spdep                      1.3-5     2024-06-10 [1] CRAN (R 4.4.0)
   statmod                    1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   SummarizedExperiment     * 1.34.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   terra                      1.7-78    2024-05-22 [1] CRAN (R 4.4.0)
   tibble                     3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyselect                 1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   tiff                       0.1-12    2023-11-28 [1] CRAN (R 4.4.0)
   UCSC.utils                 1.0.0     2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   units                      0.8-5     2023-11-28 [1] CRAN (R 4.4.0)
   utf8                       1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                      0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   withr                      3.0.0     2024-01-16 [1] CRAN (R 4.4.0)
   wk                         0.9.2     2024-07-09 [1] CRAN (R 4.4.0)
   xfun                       0.45      2024-06-16 [1] CRAN (R 4.4.0)
   XVector                    0.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                       2.3.9     2024-07-05 [1] CRAN (R 4.4.0)
   zeallot                    0.1.0     2018-01-28 [1] CRAN (R 4.4.0)
   zlibbioc                   1.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────
