---
title: "03-VMH ARC Sex DE Registration to Yao23 Mouse"
output: html_document
date: "2024-08-19"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(ggplot2)
library(gridExtra)
library(pheatmap)
require(colorout)
library(ggrepel)
library(ggtext)
library(Polychrome)
library(ggtext)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 10), axis.title.x = element_text(size = 11), axis.text.y = element_text(size = 10), axis.title.y = element_text(size=11), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=9), legend.title = element_text(size=10,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```


helper function to extract a legend and use as grob: from https://statisticsglobe.com/add-common-legend-to-combined-ggplot2-plots-in-r/#example-2-add-shared-legend-to-ggplot2-plots-using-gridextra-package 
```{r}
extract_legend <- function(my_ggp) {
  step1 <- ggplot_gtable(ggplot_build(my_ggp))
  step2 <- which(sapply(step1$grobs, function(x) x$name) == "guide-box")
  step3 <- step1$grobs[[step2]]
  return(step3)
}
```

### for VMH domain clusters - VMH subclasses / ARC domain - ARC subclasses
```{r}
deregs <- readRDS("processed-data/10_Spatial_Registration/03-Yao23-subclass-MvFsexDE-tstat_correls_to_celltypes-in-xenARCVMHdomains.RDS")

### plots: VMH subdomains to VMH clusters / ARCsubdomains to ARC clusters spatreg values as a function of n genes
plotdereg <- deregs[mouse_subclass %in% grep(mouse_subclass,pattern="VMH|ARH",value=T)]

### set up names for ease of plot rendering
setnames(plotdereg,"top_n_human","Top n Human Sex DEGs")
setnames(plotdereg,"mouse_subclass","Subclass")

### get arc, vmh domain data
arcplots <- plotdereg[human_domain_celltype %in% grep(human_domain_celltype,pattern="^ARC",value=T)&Subclass %in% grep(Subclass,pattern="ARH",value=T)]

vmhplots <- plotdereg[human_domain_celltype %in% grep(human_domain_celltype,pattern="^VMH",value=T)&Subclass %in% grep(Subclass,pattern="VMH",value=T)]
```

```{r}
### using the previously defined arc-type palette for consistency
arcplt <- list()

### generate one plot with a legend to extract and place beside all plots
arcplt[[1]] <- ggplot(arcplots[human_domain_celltype==unique(arcplots$human_domain_celltype)[1]],aes(x=`Top n Human Sex DEGs`,y=spearmancor,col=Subclass))+
    geom_line(linewidth=0.7,alpha=0.7)+
    ggtitle(gsub(unique(arcplots$human_domain_celltype)[1],
            pattern="^ARC_(.*)$",replacement="\\1"))+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.5,0.8)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))

### for all the other cell types, generate a plot without a legend to start w
otherarcs <- unique(arcplots[human_domain_celltype!=unique(arcplots$human_domain_celltype)[1],human_domain_celltype])

otherarcplts <- lapply(otherarcs,FUN=function(x){
    p <- ggplot(arcplots[human_domain_celltype==x],aes(x=`Top n Human Sex DEGs`,y=spearmancor,col=Subclass))+
    geom_line(linewidth=0.7,alpha=0.7)+
    guides(color="none")+
    ggtitle(gsub(x,pattern="^ARC_(.*)$",replacement="\\1"))+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.5,0.8)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))
    return(p)
})

arcplt <- c(arcplt,otherarcplts)
### extract legend to plot one time alongside the two plots:
leg.tmp <- extract_legend(arcplt[[1]])
arcplt[[1]] <- arcplt[[1]]+guides(color="none")
dev.off()

pdf("plots/10_Spatial_Registration/03-Sex DE correls of ARC domain cell types to Yao23 ARH subclasses.pdf",height=20,width=15)
do.call("grid.arrange",list(arrangeGrob(grobs=arcplt),leg.tmp,ncol=2,widths=c(7,1.5)))
dev.off()
```


##### and VMH
```{r}
vmhplt <- list()
vmhplt[[1]] <- ggplot(vmhplots[human_domain_celltype==unique(vmhplots$human_domain_celltype)[1]],aes(x=`Top n Human Sex DEGs`,y=spearmancor,col=Subclass))+
    geom_line(size=0.7,alpha=0.7)+
    ggtitle(gsub(unique(vmhplots$human_domain_celltype)[1],
                 pattern="^VMH_(.*)$",
                 replacement="\\1"))+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.7,0.8)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))

othervmhs <- unique(vmhplots[human_domain_celltype!=unique(vmhplots$human_domain_celltype)[1],human_domain_celltype])

othervmhplts <- lapply(othervmhs,FUN=function(x){
    p <- ggplot(vmhplots[human_domain_celltype==x],aes(x=`Top n Human Sex DEGs`,y=spearmancor,col=Subclass))+
    geom_line(linewidth=0.7,alpha=0.7)+
    guides(color="none")+
    ggtitle(gsub(x,pattern="^VMH_(.*)$",replacement="\\1"))+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.7,0.8)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))
    return(p)
})

vmhplt <- c(vmhplt,othervmhplts)
### extract legend to plot one time alongside the two plots:
leg.tmp <- extract_legend(vmhplt[[1]])
vmhplt[[1]] <- vmhplt[[1]]+guides(color="none")
dev.off()

pdf("plots/10_Spatial_Registration/03-Sex DE correls of VMH domain cell types to Yao23 VMH subclasses.pdf",height=20,width=15)
do.call("grid.arrange",list(arrangeGrob(grobs=vmhplt),leg.tmp,ncol=2,widths=c(7,1.5)))
dev.off()
```

######################

### for VMH domain clusters - VMH SUPERTYPES / ARC domain - ARC supertypes
```{r}
deregs <- readRDS("processed-data/10_Spatial_Registration/03-Yao23-VMHARCsupertype-MvFsexDE-tstat_correls_to_celltypes-in-xenARCVMHdomains.RDS")

### plots: VMH subdomains to VMH clusters / ARCsubdomains to ARC clusters spatreg values as a function of n genes
plotdereg <- deregs[mouse_supertype %in% grep(mouse_supertype,pattern="VMH|ARH",value=T)]

### set up names for ease of plot rendering
setnames(plotdereg,"top_n_human","Top n Human Sex DEGs")
setnames(plotdereg,"mouse_supertype","Supertype")

### get arc, vmh domain data
arcplots <- plotdereg[human_domain_celltype %in% grep(human_domain_celltype,pattern="^ARC",value=T)&Supertype %in% grep(Supertype,pattern="ARH",value=T)]

vmhplots <- plotdereg[human_domain_celltype %in% grep(human_domain_celltype,pattern="^VMH",value=T)&Supertype %in% grep(Supertype,pattern="VMH",value=T)]
```

```{r}
arcplt <- list()

### generate one plot with a legend to extract and place beside all plots
arcplt[[1]] <- ggplot(arcplots[human_domain_celltype==unique(arcplots$human_domain_celltype)[1]],aes(x=`Top n Human Sex DEGs`,y=spearmancor,col=Supertype))+
    geom_line(linewidth=0.7,alpha=0.7)+
    ggtitle(gsub(unique(arcplots$human_domain_celltype)[1],
            pattern="^ARC_(.*)$",replacement="\\1"))+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.8,0.9)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))

### for all the other cell types, generate a plot without a legend to start w
otherarcs <- unique(arcplots[human_domain_celltype!=unique(arcplots$human_domain_celltype)[1],human_domain_celltype])

otherarcplts <- lapply(otherarcs,FUN=function(x){
    p <- ggplot(arcplots[human_domain_celltype==x],aes(x=`Top n Human Sex DEGs`,y=spearmancor,col=Supertype))+
    geom_line(linewidth=0.7,alpha=0.7)+
    guides(color="none")+
    ggtitle(gsub(x,pattern="^ARC_(.*)$",replacement="\\1"))+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.8,0.9)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))
    return(p)
})

arcplt <- c(arcplt,otherarcplts)
### extract legend to plot one time alongside the two plots:
leg.tmp <- extract_legend(arcplt[[1]])
arcplt[[1]] <- arcplt[[1]]+guides(color="none")
dev.off()

pdf("plots/10_Spatial_Registration/03-Sex DE correls of ARC domain cell types to Yao23 ARH supertypes.pdf",height=20,width=15)
do.call("grid.arrange",list(arrangeGrob(grobs=arcplt),leg.tmp,ncol=2,widths=c(7,1.5)))
dev.off()
```


##### and VMH
```{r}
vmhplt <- list()

vmhplt[[1]] <- ggplot(vmhplots[human_domain_celltype==unique(vmhplots$human_domain_celltype)[1]],aes(x=`Top n Human Sex DEGs`,y=spearmancor,col=Supertype))+
    geom_line(size=0.7,alpha=0.7)+
    ggtitle(gsub(unique(vmhplots$human_domain_celltype)[1],
                 pattern="^VMH_(.*)$",
                 replacement="\\1"))+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.9,0.8)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))

othervmhs <- unique(vmhplots[human_domain_celltype!=unique(vmhplots$human_domain_celltype)[1],human_domain_celltype])

othervmhplts <- lapply(othervmhs,FUN=function(x){
    p <- ggplot(vmhplots[human_domain_celltype==x],aes(x=`Top n Human Sex DEGs`,y=spearmancor,col=Supertype))+
    geom_line(linewidth=0.7,alpha=0.7)+
    guides(color="none")+
    ggtitle(gsub(x,pattern="^VMH_(.*)$",replacement="\\1"))+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.9,0.8)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))
    return(p)
})

vmhplt <- c(vmhplt,othervmhplts)
### extract legend to plot one time alongside the two plots:
leg.tmp <- extract_legend(vmhplt[[1]])
vmhplt[[1]] <- vmhplt[[1]]+guides(color="none")
dev.off()

pdf("plots/10_Spatial_Registration/03-Sex DE correls of VMH domain cell types to Yao23 VMH supertypes.pdf",height=20,width=15)
do.call("grid.arrange",list(arrangeGrob(grobs=vmhplt),leg.tmp,ncol=2,widths=c(7,1.5)))
dev.off()
```

### reprod inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] Polychrome_1.5.1    ggtext_0.1.2        ggrepel_0.9.5       pheatmap_1.0.12    
 [5] gridExtra_2.3       ggplot2_3.5.1       Biostrings_2.72.1   GenomeInfoDb_1.40.1
 [9] XVector_0.44.0      IRanges_2.38.1      S4Vectors_0.42.1    BiocGenerics_0.50.0
[13] default_1.0.0       BiocParallel_1.38.0 gtools_3.9.5        biomaRt_2.60.1     
[17] orthogene_1.10.0    colorout_1.3-0.2    data.table_1.15.4   rlang_1.1.4        

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3        rstudioapi_0.16.0         jsonlite_1.8.8           
  [4] magrittr_2.0.3            farver_2.1.2              rmarkdown_2.27           
  [7] fs_1.6.4                  zlibbioc_1.50.0           vctrs_0.6.5              
 [10] memoise_2.0.1             RCurl_1.98-1.14           ggtree_3.12.0            
 [13] rstatix_0.7.2             htmltools_0.5.8.1         progress_1.2.3           
 [16] curl_5.2.1                broom_1.0.6               gridGraphics_0.5-1       
 [19] htmlwidgets_1.6.4         httr2_1.0.1               plotly_4.10.4            
 [22] cachem_1.1.0              commonmark_1.9.1          lifecycle_1.0.4          
 [25] pkgconfig_2.0.3           Matrix_1.7-0              R6_2.5.1                 
 [28] fastmap_1.2.0             GenomeInfoDbData_1.2.12   digest_0.6.36            
 [31] aplot_0.2.3               colorspace_2.1-0          patchwork_1.2.0          
 [34] AnnotationDbi_1.66.0      rprojroot_2.0.4           grr_0.9.5                
 [37] RSQLite_2.3.7             ggpubr_0.6.0              filelock_1.0.3           
 [40] labeling_0.4.3            fansi_1.0.6               httr_1.4.7               
 [43] abind_1.4-5               compiler_4.4.1            here_1.0.1               
 [46] bit64_4.0.5               withr_3.0.0               backports_1.5.0          
 [49] carData_3.0-5             DBI_1.2.3                 homologene_1.4.68.19.3.27
 [52] ggsignif_0.6.4            rappdirs_0.3.3            sessioninfo_1.2.2        
 [55] scatterplot3d_0.3-44      tools_4.4.1               ape_5.8                  
 [58] glue_1.7.0                nlme_3.1-165              gridtext_0.1.5           
 [61] grid_4.4.1                generics_0.1.3            gtable_0.3.5             
 [64] tidyr_1.3.1               hms_1.1.3                 xml2_1.3.6               
 [67] car_3.1-2                 utf8_1.2.4                pillar_1.9.0             
 [70] markdown_1.13             stringr_1.5.1             yulab.utils_0.1.4        
 [73] babelgene_22.9            dplyr_1.1.4               BiocFileCache_2.12.0     
 [76] treeio_1.28.0             lattice_0.22-6            bit_4.0.5                
 [79] tidyselect_1.2.1          knitr_1.48                xfun_0.45                
 [82] Biobase_2.64.0            stringi_1.8.4             UCSC.utils_1.0.0         
 [85] lazyeval_0.2.2            ggfun_0.1.5               yaml_2.3.9               
 [88] evaluate_0.24.0           codetools_0.2-20          tibble_3.2.1             
 [91] ggplotify_0.1.2           cli_3.6.3                 munsell_0.5.1            
 [94] Rcpp_1.0.13               dbplyr_2.5.0              gprofiler2_0.2.3         
 [97] png_0.1-8                 parallel_4.4.1            blob_1.2.4               
[100] prettyunits_1.2.0         bitops_1.0-7              viridisLite_0.4.2        
[103] tidytree_0.4.6            scales_1.3.0              purrr_1.0.2              
[106] crayon_1.5.3              KEGGREST_1.44.1          
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-08-19
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────────────
 ! package          * version        date (UTC) lib source
   abind              1.4-5          2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi      1.66.0         2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   ape                5.8            2024-04-11 [1] CRAN (R 4.4.0)
   aplot              0.2.3          2024-06-17 [1] CRAN (R 4.4.0)
   babelgene          22.9           2022-09-29 [1] CRAN (R 4.4.0)
   backports          1.5.0          2024-05-23 [1] CRAN (R 4.4.0)
   Biobase            2.64.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache      2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics     * 0.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel     * 1.38.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   biomaRt          * 2.60.1         2024-06-26 [1] Bioconductor 3.19 (R 4.4.1)
   Biostrings       * 2.72.1         2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                4.0.5          2022-11-15 [1] CRAN (R 4.4.0)
   bit64              4.0.5          2020-08-30 [1] CRAN (R 4.4.0)
   bitops             1.0-7          2021-04-24 [1] CRAN (R 4.4.0)
   blob               1.2.4          2023-03-17 [1] CRAN (R 4.4.0)
   broom              1.0.6          2024-05-17 [1] CRAN (R 4.4.0)
   cachem             1.1.0          2024-05-16 [1] CRAN (R 4.4.0)
   car                3.1-2          2023-03-30 [1] CRAN (R 4.4.0)
   carData            3.0-5          2022-01-06 [1] CRAN (R 4.4.0)
 P cli                3.6.3          2024-06-21 [2] CRAN (R 4.4.0)
   codetools          0.2-20         2024-03-31 [1] CRAN (R 4.4.1)
   colorout         * 1.3-0.2        2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace         2.1-0          2023-01-23 [1] CRAN (R 4.4.0)
   commonmark         1.9.1          2024-01-30 [1] CRAN (R 4.4.0)
   crayon             1.5.3          2024-06-20 [1] CRAN (R 4.4.0)
   curl               5.2.1          2024-03-01 [1] CRAN (R 4.4.0)
   data.table       * 1.15.4         2024-03-30 [2] CRAN (R 4.4.0)
   DBI                1.2.3          2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr             2.5.0          2024-03-19 [1] CRAN (R 4.4.0)
   default          * 1.0.0          2017-08-07 [1] CRAN (R 4.4.0)
   digest             0.6.36         2024-06-23 [1] CRAN (R 4.4.0)
   dplyr              1.1.4          2023-11-17 [1] CRAN (R 4.4.0)
   evaluate           0.24.0         2024-06-10 [1] CRAN (R 4.4.0)
   fansi              1.0.6          2023-12-08 [1] CRAN (R 4.4.0)
   farver             2.1.2          2024-05-13 [1] CRAN (R 4.4.0)
   fastmap            1.2.0          2024-05-15 [1] CRAN (R 4.4.0)
   filelock           1.0.3          2023-12-11 [1] CRAN (R 4.4.0)
   fs                 1.6.4          2024-04-25 [1] CRAN (R 4.4.0)
   generics           0.1.3          2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb     * 1.40.1         2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData   1.2.12         2024-05-01 [1] Bioconductor
   ggfun              0.1.5          2024-05-28 [1] CRAN (R 4.4.0)
   ggplot2          * 3.5.1          2024-04-23 [1] CRAN (R 4.4.0)
   ggplotify          0.1.2          2023-08-09 [1] CRAN (R 4.4.0)
   ggpubr             0.6.0          2023-02-10 [1] CRAN (R 4.4.0)
   ggrepel          * 0.9.5          2024-01-10 [1] CRAN (R 4.4.0)
   ggsignif           0.6.4          2022-10-13 [1] CRAN (R 4.4.0)
   ggtext           * 0.1.2          2022-09-16 [1] CRAN (R 4.4.0)
   ggtree             3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   glue               1.7.0          2024-01-09 [1] CRAN (R 4.4.0)
   gprofiler2         0.2.3          2024-02-23 [1] CRAN (R 4.4.0)
   gridExtra        * 2.3            2017-09-09 [1] CRAN (R 4.4.0)
   gridGraphics       0.5-1          2020-12-13 [1] CRAN (R 4.4.0)
   gridtext           0.1.5          2022-09-16 [1] CRAN (R 4.4.0)
   grr                0.9.5          2016-08-26 [1] CRAN (R 4.4.0)
   gtable             0.3.5          2024-04-22 [1] CRAN (R 4.4.0)
   gtools           * 3.9.5          2023-11-20 [1] CRAN (R 4.4.0)
   here               1.0.1          2020-12-13 [1] CRAN (R 4.4.0)
   hms                1.1.3          2023-03-21 [1] CRAN (R 4.4.0)
   homologene         1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.4.0)
   htmltools          0.5.8.1        2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets        1.6.4          2023-12-06 [1] CRAN (R 4.4.0)
   httr               1.4.7          2023-08-15 [1] CRAN (R 4.4.0)
   httr2              1.0.1          2024-04-01 [1] CRAN (R 4.4.0)
   IRanges          * 2.38.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   jsonlite           1.8.8          2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST           1.44.1         2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   knitr              1.48           2024-07-07 [1] CRAN (R 4.4.1)
   labeling           0.4.3          2023-08-29 [1] CRAN (R 4.4.0)
   lattice            0.22-6         2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval           0.2.2          2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle          1.0.4          2023-11-07 [1] CRAN (R 4.4.0)
   magrittr           2.0.3          2022-03-30 [1] CRAN (R 4.4.0)
   markdown           1.13           2024-06-04 [1] CRAN (R 4.4.0)
   Matrix             1.7-0          2024-04-26 [1] CRAN (R 4.4.1)
   memoise            2.0.1          2021-11-26 [1] CRAN (R 4.4.0)
   munsell            0.5.1          2024-04-01 [1] CRAN (R 4.4.0)
   nlme               3.1-165        2024-06-06 [1] CRAN (R 4.4.0)
   orthogene        * 1.10.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   patchwork          1.2.0          2024-01-08 [1] CRAN (R 4.4.0)
   pheatmap         * 1.0.12         2019-01-04 [1] CRAN (R 4.4.0)
   pillar             1.9.0          2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig          2.0.3          2019-09-22 [1] CRAN (R 4.4.0)
   plotly             4.10.4         2024-01-13 [1] CRAN (R 4.4.0)
   png                0.1-8          2022-11-29 [1] CRAN (R 4.4.0)
   Polychrome       * 1.5.1          2022-05-03 [1] CRAN (R 4.4.0)
   prettyunits        1.2.0          2023-09-24 [1] CRAN (R 4.4.0)
   progress           1.2.3          2023-12-06 [1] CRAN (R 4.4.0)
   purrr              1.0.2          2023-08-10 [1] CRAN (R 4.4.0)
   R6                 2.5.1          2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs           0.3.3          2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer       1.1-3          2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp               1.0.13         2024-07-17 [1] CRAN (R 4.4.0)
   RCurl              1.98-1.14      2024-01-09 [1] CRAN (R 4.4.0)
 P rlang            * 1.1.4          2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown          2.27           2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot          2.0.4          2023-11-05 [1] CRAN (R 4.4.0)
   RSQLite            2.3.7          2024-05-27 [1] CRAN (R 4.4.0)
   rstatix            0.7.2          2023-02-01 [1] CRAN (R 4.4.0)
   rstudioapi         0.16.0         2024-03-24 [1] CRAN (R 4.4.0)
   S4Vectors        * 0.42.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   scales             1.3.0          2023-11-28 [1] CRAN (R 4.4.0)
   scatterplot3d      0.3-44         2023-05-05 [1] CRAN (R 4.4.0)
   sessioninfo        1.2.2          2021-12-06 [1] CRAN (R 4.4.0)
   stringi            1.8.4          2024-05-06 [1] CRAN (R 4.4.0)
   stringr            1.5.1          2023-11-14 [1] CRAN (R 4.4.0)
   tibble             3.2.1          2023-03-20 [1] CRAN (R 4.4.0)
   tidyr              1.3.1          2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect         1.2.1          2024-03-11 [1] CRAN (R 4.4.0)
   tidytree           0.4.6          2023-12-12 [1] CRAN (R 4.4.0)
   treeio             1.28.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   UCSC.utils         1.0.0          2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   utf8               1.2.4          2023-10-22 [1] CRAN (R 4.4.0)
   vctrs              0.6.5          2023-12-01 [1] CRAN (R 4.4.0)
   viridisLite        0.4.2          2023-05-02 [1] CRAN (R 4.4.0)
   withr              3.0.0          2024-01-16 [1] CRAN (R 4.4.0)
   xfun               0.45           2024-06-16 [1] CRAN (R 4.4.0)
   xml2               1.3.6          2023-12-04 [1] CRAN (R 4.4.0)
   XVector          * 0.44.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml               2.3.9          2024-07-05 [1] CRAN (R 4.4.0)
   yulab.utils        0.1.4          2024-01-28 [1] CRAN (R 4.4.0)
   zlibbioc           1.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────
