---
title: "01_1-QC Filters"
author: "Bernie Mulvey"
date: "2023-05-22"
output: html_document
---

```{r setup, include=FALSE}
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(ggspavis)
library(spatialLIBD)
library(scater) # addPerCellQC
library(BiocParallel)
library(scran)
library(gridExtra)
```

## load pre-assembled SPE object of unfiltered data
```{r}
load("data/spe_raw_w-V13M13-362.Rdata")
hyp <- spe

rm(spe)
### for reference: initial dataset dimensions
dim(hyp)
head(colData(hyp))
tail(colData(hyp))
```


### filtering
# get rid of non-tissue spots after loupe alignment
```{r}
hyp2 <- hyp[,colData(hyp)$in_tissue]
### check number of spots lost
ncol(hyp)-ncol(hyp2)

## hold onto unfiltered object for now as comparator.
```

### % mitochondrial reads--values already added in by Ryan M.
```{r}
### plot
hist(colData(hyp)$expr_chrM_ratio)
# nothing over ~50%, so that's good. we saw in the early part of the dataset that the highest mitochondrial counts were in the VMH, so we don't want to filter too stringently on mitochondrial counts.
# there's one weird sample here with a handful of like 100% mito reads, so for color scaling purposes plot spots with % mito less than like 80.
plt <- spatialLIBD::vis_grid_gene(hyp[,hyp$expr_chrM_ratio<0.8],geneid="expr_chrM_ratio",pdf_file = F,return_plots = T,spatial = F,point_size = 1.5,auto_crop = T)
do.call("grid.arrange",plt)
# we can see pretty clearly that there is some biology to this -- regionally varies and NOT just at areas where we'd expect it from tissue damage (e.g., edges). so we should filter gently.
sum(colData(hyp)$expr_chrM_ratio > 0.8,na.rm=T)
# FILL IN NUMBER OF SPOTS
sum(colData(hyp)$expr_chrM_ratio > 0.5,na.rm=T)
# FILL IN NUMBER OF SPOTS

# let's go with 50% as the cutoff, considering how few spots cross that threshold and the biology of the signal.
hyp2 <- hyp2[,hyp2$expr_chrM_ratio<0.5]
```

### UMIs per spot
### standard threshold is 600 UMIs
### how many spots are under, say, 1000? let's plot the distribution of those.
```{r}
sum(hyp$sum_umi<1000)
### FILL IN NUMBER / PERCENT OF TOTAL SPOTS

### plot the distribution of UMIs/spot from spots < 1000 umi/spt
ggplot(as.data.frame(colData(hyp2[,hyp2$sum_umi<1000])), aes(x = sum_umi)) +
    geom_histogram(binwidth = 25)+
  geom_vline(xintercept = 600,col="red")
dev.off()

### the tail dropoff here is actually more around 250-275:
ggplot(as.data.frame(colData(hyp2[,hyp2$sum_umi<1000])), aes(x = sum_umi)) +
    geom_histogram(binwidth = 25)+
  geom_vline(xintercept = 275,col="red")
dev.off()
## which corresponds to FILL IN spots:
sum(colData(hyp)$sum_umi<275,na.rm=T)


## it looks like all of that is condensed together and practically all tissue edges with very few umis (definitely << 275). let's make a plot to confirm that:
hyp$under275 <- as.factor(ifelse(hyp$sum_umi<275),yes="under275",no="over275")
plt <- spatialLIBD::vis_grid_clus(tmp,"",pdf_file=NULL,return_plots = T,spatial = F,point_size = 1.5,auto_crop = T)
do.call("grid.arrange",plt)
dev.off()
# and those edge areas area indeed <275, so 275 the cutoff is.
```

### genes per spot (sum_gene)
# first, let's look at how the number of unique genes detected and number of umis detected relate to get a feel for how much umi space is going to highly expressed genes (i.e., multiple umis on one gene) 
```{r}
ggplot(as.data.table(colData(hyp)),aes(x=sum_gene,y=sum_umi)) +
    geom_point() +
    geom_abline(slope = 1,intercept = 0,col="red")
dev.off()

## that's an x^3 function across the full spectrum of values. let's look at the low end:
ggplot(as.data.table(colData(hyp[,hyp$sum_gene<1000])),aes(x=sum_gene,y=sum_umi)) +
    geom_point() +
    geom_abline(slope = 1.66,intercept = 0,col="red")
dev.off()

## that part is linear, with a slope of about 1.66, as overlaid, meaning the general expectation for the low end of unique gene totals is that there are about 1.66x as many umis, or for u umis, about u/1.66 genes.

## so 275/1.66 = 166 as a matching cutoff for unique genes; lets see how that looks on the histogram.
ggplot(as.data.table(colData(hyp[,hyp$sum_gene<1000])),aes(x=sum_gene)) +
    geom_histogram(binwidth=10) +
    geom_vline(xintercept = 166,col="red")
dev.off()

## wam bam thank you ma'am that's right off the cliff for sum_gene. let's make sure there's not any spatial patterns under 166 that would indicate biology worth keeping at an even more lenient threshold.

hyp$gene166 <- as.factor(ifelse(hyp$sum_gene<166,yes="under166",no="over166"))
plt <- spatialLIBD::vis_grid_clus(hyp,"gene166",pdf_file = NULL,spatial = F,return_plots = T,auto_crop = T,point_size = 1.5)
do.call("grid.arrange",plt)
dev.off()

### that's practically just the same spots with <275 umis, and a very small # of (colocalized) spots in other samples. so 166 minimum unique genes and 275 minimum UMIs.

### clear out the temporary columns used for labeling
hyp$under275 <- NULL
hyp$gene166 <- NULL
```

### apply mitochondria, UMI, unique gene cutoffs per above (<50%, > 275 umis, > 166 unique genes)
```{r}
ncol(hyp[,colData(hyp)$sum_umi>275&colData(hyp)$sum_gene>166&colData(hyp)$expr_chrM_ratio<0.5])/ncol(hyp)
# 98.66% spots retained for now!
hyp2 <- hyp[,colData(hyp)$sum_umi>275&colData(hyp)$sum_gene>166&colData(hyp)$expr_chrM_ratio<0.5]
rm(hyp)
gc(full=T)
```

### apply library-size normalized, LOG COUNT normalization AFTER filtering the data.
```{r}
hyp2 <- computeLibraryFactors(hyp2)
# peek
hist(sizeFactors(hyp2),breaks=100)
dev.off()
# append log counts using scater (don't use the SpatialExpt:: function here, that's for RETRIEVING an assay by the name of "logcounts" from spe objs)
hyp2 <- scater::logNormCounts(hyp2)
stopifnot(dim(counts(hyp2))==dim(logcounts(hyp2)))
saveRDS(hyp2,"data/01-QC_filters/hypN9_umi275_gene166_chrm50_lognorm_111723.RDS")

rm(hyp2)
gc(full=T)
```

#### session info pasted after call.
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] gridExtra_2.3               scran_1.30.0               
 [3] BiocParallel_1.36.0         scater_1.30.1              
 [5] scuttle_1.12.0              spatialLIBD_1.13.4         
 [7] ggspavis_1.8.0              ggplot2_3.4.4              
 [9] SpatialExperiment_1.12.0    SingleCellExperiment_1.24.0
[11] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[13] GenomicRanges_1.54.1        GenomeInfoDb_1.38.1        
[15] IRanges_2.36.0              S4Vectors_0.40.1           
[17] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
[19] matrixStats_1.1.0           data.table_1.14.8          
[21] rlang_1.1.2                

loaded via a namespace (and not attached):
  [1] later_1.3.1                   BiocIO_1.12.0                
  [3] bitops_1.0-7                  filelock_1.0.2               
  [5] fields_15.2                   tibble_3.2.1                 
  [7] XML_3.99-0.15                 lifecycle_1.0.4              
  [9] edgeR_4.0.1                   doParallel_1.0.17            
 [11] rprojroot_2.0.4               lattice_0.22-5               
 [13] magrittr_2.0.3                limma_3.58.1                 
 [15] plotly_4.10.3                 sass_0.4.7                   
 [17] rmarkdown_2.25                jquerylib_0.1.4              
 [19] yaml_2.3.7                    metapod_1.10.0               
 [21] httpuv_1.6.12                 spam_2.10-0                  
 [23] ggside_0.2.2                  sessioninfo_1.2.2            
 [25] cowplot_1.1.1                 DBI_1.1.3                    
 [27] RColorBrewer_1.1-3            golem_0.4.1                  
 [29] maps_3.4.1.1                  abind_1.4-5                  
 [31] zlibbioc_1.48.0               purrr_1.0.2                  
 [33] RCurl_1.98-1.13               rappdirs_0.3.3               
 [35] GenomeInfoDbData_1.2.11       ggrepel_0.9.4                
 [37] irlba_2.3.5.1                 dqrng_0.3.1                  
 [39] DelayedMatrixStats_1.24.0     codetools_0.2-19             
 [41] DelayedArray_0.28.0           DT_0.30                      
 [43] tidyselect_1.2.0              ScaledMatrix_1.10.0          
 [45] viridis_0.6.4                 shinyWidgets_0.8.0           
 [47] BiocFileCache_2.10.1          GenomicAlignments_1.38.0     
 [49] jsonlite_1.8.7                BiocNeighbors_1.20.0         
 [51] ellipsis_0.3.2                iterators_1.0.14             
 [53] foreach_1.5.2                 tools_4.3.1                  
 [55] Rcpp_1.0.11                   glue_1.6.2                   
 [57] SparseArray_1.2.2             xfun_0.41                    
 [59] here_1.0.1                    dplyr_1.1.4                  
 [61] withr_2.5.2                   BiocManager_1.30.22          
 [63] fastmap_1.1.1                 bluster_1.12.0               
 [65] fansi_1.0.5                   digest_0.6.33                
 [67] rsvd_1.0.5                    R6_2.5.1                     
 [69] mime_0.12                     colorspace_2.1-0             
 [71] RSQLite_2.3.3                 config_0.3.2                 
 [73] utf8_1.2.4                    tidyr_1.3.0                  
 [75] generics_0.1.3                rtracklayer_1.62.0           
 [77] httr_1.4.7                    htmlwidgets_1.6.2            
 [79] S4Arrays_1.2.0                pkgconfig_2.0.3              
 [81] gtable_0.3.4                  blob_1.2.4                   
 [83] XVector_0.42.0                htmltools_0.5.7              
 [85] dotCall64_1.1-0               scales_1.2.1                 
 [87] png_0.1-8                     attempt_0.3.1                
 [89] knitr_1.45                    rstudioapi_0.15.0            
 [91] rjson_0.2.21                  curl_5.1.0                   
 [93] cachem_1.0.8                  BiocVersion_3.18.1           
 [95] parallel_4.3.1                vipor_0.4.5                  
 [97] AnnotationDbi_1.64.1          restfulr_0.0.15              
 [99] pillar_1.9.0                  grid_4.3.1                   
[101] vctrs_0.6.4                   promises_1.2.1               
[103] BiocSingular_1.18.0           dbplyr_2.4.0                 
[105] beachmat_2.18.0               cluster_2.1.4                
[107] xtable_1.8-4                  beeswarm_0.4.0               
[109] paletteer_1.5.0               evaluate_0.23                
[111] magick_2.8.1                  cli_3.6.1                    
[113] locfit_1.5-9.8                compiler_4.3.1               
[115] Rsamtools_2.18.0              crayon_1.5.2                 
[117] rematch2_2.1.2                ggbeeswarm_0.7.2             
[119] viridisLite_0.4.2             munsell_0.5.0                
[121] Biostrings_2.70.1             lazyeval_0.2.2               
[123] Matrix_1.6-3                  ExperimentHub_2.10.0         
[125] benchmarkme_1.0.8             sparseMatrixStats_1.14.0     
[127] bit64_4.0.5                   KEGGREST_1.42.0              
[129] statmod_1.5.0                 shiny_1.8.0                  
[131] interactiveDisplayBase_1.40.0 AnnotationHub_3.10.0         
[133] igraph_1.5.1                  memoise_2.0.1                
[135] bslib_0.5.1                   benchmarkmeData_1.0.4        
[137] bit_4.0.5

##
─ Session info ────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.1 (2023-06-16)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2023-11-21
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────
 package                * version   date (UTC) lib source
 abind                    1.4-5     2016-07-21 [1] CRAN (R 4.3.0)
 AnnotationDbi            1.64.1    2023-11-02 [1] Bioconductor
 AnnotationHub            3.10.0    2023-10-24 [1] Bioconductor
 attempt                  0.3.1     2020-05-03 [1] CRAN (R 4.3.0)
 beachmat                 2.18.0    2023-10-24 [1] Bioconductor
 beeswarm                 0.4.0     2021-06-01 [1] CRAN (R 4.3.0)
 benchmarkme              1.0.8     2022-06-12 [1] CRAN (R 4.3.0)
 benchmarkmeData          1.0.4     2020-04-23 [1] CRAN (R 4.3.0)
 Biobase                * 2.62.0    2023-10-26 [1] Bioconductor
 BiocFileCache            2.10.1    2023-10-26 [1] Bioconductor
 BiocGenerics           * 0.48.1    2023-11-02 [1] Bioconductor
 BiocIO                   1.12.0    2023-10-24 [1] Bioconductor
 BiocManager              1.30.22   2023-08-08 [1] CRAN (R 4.3.0)
 BiocNeighbors            1.20.0    2023-10-24 [1] Bioconductor
 BiocParallel           * 1.36.0    2023-10-26 [1] Bioconductor
 BiocSingular             1.18.0    2023-10-24 [1] Bioconductor
 BiocVersion              3.18.1    2023-11-15 [1] Bioconductor
 Biostrings               2.70.1    2023-10-26 [1] Bioconductor
 bit                      4.0.5     2022-11-15 [1] CRAN (R 4.3.0)
 bit64                    4.0.5     2020-08-30 [1] CRAN (R 4.3.0)
 bitops                   1.0-7     2021-04-24 [1] CRAN (R 4.3.0)
 blob                     1.2.4     2023-03-17 [1] CRAN (R 4.3.0)
 bluster                  1.12.0    2023-10-24 [1] Bioconductor
 bslib                    0.5.1     2023-08-11 [1] CRAN (R 4.3.0)
 cachem                   1.0.8     2023-05-01 [1] CRAN (R 4.3.0)
 cli                      3.6.1     2023-03-23 [1] CRAN (R 4.3.0)
 cluster                  2.1.4     2022-08-22 [1] CRAN (R 4.3.0)
 codetools                0.2-19    2023-02-01 [1] CRAN (R 4.3.0)
 colorspace               2.1-0     2023-01-23 [1] CRAN (R 4.3.0)
 config                   0.3.2     2023-08-30 [1] CRAN (R 4.3.0)
 cowplot                  1.1.1     2020-12-30 [1] CRAN (R 4.3.0)
 crayon                   1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 curl                     5.1.0     2023-10-02 [1] CRAN (R 4.3.1)
 data.table             * 1.14.8    2023-02-17 [1] CRAN (R 4.3.0)
 DBI                      1.1.3     2022-06-18 [1] CRAN (R 4.3.0)
 dbplyr                   2.4.0     2023-10-26 [1] CRAN (R 4.3.1)
 DelayedArray             0.28.0    2023-11-06 [1] Bioconductor
 DelayedMatrixStats       1.24.0    2023-10-24 [1] Bioconductor
 digest                   0.6.33    2023-07-07 [1] CRAN (R 4.3.0)
 doParallel               1.0.17    2022-02-07 [1] CRAN (R 4.3.0)
 dotCall64                1.1-0     2023-10-17 [1] CRAN (R 4.3.1)
 dplyr                    1.1.4     2023-11-17 [1] CRAN (R 4.3.1)
 dqrng                    0.3.1     2023-08-30 [1] CRAN (R 4.3.0)
 DT                       0.30      2023-10-05 [1] CRAN (R 4.3.1)
 edgeR                    4.0.1     2023-10-29 [1] Bioconductor
 ellipsis                 0.3.2     2021-04-29 [1] CRAN (R 4.3.0)
 evaluate                 0.23      2023-11-01 [1] CRAN (R 4.3.1)
 ExperimentHub            2.10.0    2023-10-24 [1] Bioconductor
 fansi                    1.0.5     2023-10-08 [1] CRAN (R 4.3.1)
 fastmap                  1.1.1     2023-02-24 [1] CRAN (R 4.3.0)
 fields                   15.2      2023-08-17 [1] CRAN (R 4.3.0)
 filelock                 1.0.2     2018-10-05 [1] CRAN (R 4.3.0)
 foreach                  1.5.2     2022-02-02 [1] CRAN (R 4.3.0)
 generics                 0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb           * 1.38.1    2023-11-08 [1] Bioconductor
 GenomeInfoDbData         1.2.11    2023-11-10 [1] Bioconductor
 GenomicAlignments        1.38.0    2023-10-24 [1] Bioconductor
 GenomicRanges          * 1.54.1    2023-10-30 [1] Bioconductor
 ggbeeswarm               0.7.2     2023-04-29 [1] CRAN (R 4.3.0)
 ggplot2                * 3.4.4     2023-10-12 [1] CRAN (R 4.3.1)
 ggrepel                  0.9.4     2023-10-13 [1] CRAN (R 4.3.1)
 ggside                   0.2.2     2022-12-04 [1] CRAN (R 4.3.0)
 ggspavis               * 1.8.0     2023-10-24 [1] Bioconductor
 glue                     1.6.2     2022-02-24 [1] CRAN (R 4.3.0)
 golem                    0.4.1     2023-06-05 [1] CRAN (R 4.3.0)
 gridExtra              * 2.3       2017-09-09 [1] CRAN (R 4.3.0)
 gtable                   0.3.4     2023-08-21 [1] CRAN (R 4.3.0)
 here                     1.0.1     2020-12-13 [1] CRAN (R 4.3.0)
 htmltools                0.5.7     2023-11-03 [1] CRAN (R 4.3.1)
 htmlwidgets              1.6.2     2023-03-17 [1] CRAN (R 4.3.0)
 httpuv                   1.6.12    2023-10-23 [1] CRAN (R 4.3.1)
 httr                     1.4.7     2023-08-15 [1] CRAN (R 4.3.0)
 igraph                   1.5.1     2023-08-10 [1] CRAN (R 4.3.0)
 interactiveDisplayBase   1.40.0    2023-10-24 [1] Bioconductor
 IRanges                * 2.36.0    2023-10-26 [1] Bioconductor
 irlba                    2.3.5.1   2022-10-03 [1] CRAN (R 4.3.1)
 iterators                1.0.14    2022-02-05 [1] CRAN (R 4.3.0)
 jquerylib                0.1.4     2021-04-26 [1] CRAN (R 4.3.0)
 jsonlite                 1.8.7     2023-06-29 [1] CRAN (R 4.3.0)
 KEGGREST                 1.42.0    2023-10-26 [1] Bioconductor
 knitr                    1.45      2023-10-30 [1] CRAN (R 4.3.1)
 later                    1.3.1     2023-05-02 [1] CRAN (R 4.3.0)
 lattice                  0.22-5    2023-10-24 [1] CRAN (R 4.3.1)
 lazyeval                 0.2.2     2019-03-15 [1] CRAN (R 4.3.0)
 lifecycle                1.0.4     2023-11-07 [1] CRAN (R 4.3.1)
 limma                    3.58.1    2023-10-31 [1] Bioconductor
 locfit                   1.5-9.8   2023-06-11 [1] CRAN (R 4.3.0)
 magick                   2.8.1     2023-10-22 [1] CRAN (R 4.3.1)
 magrittr                 2.0.3     2022-03-30 [1] CRAN (R 4.3.0)
 maps                     3.4.1.1   2023-11-03 [1] CRAN (R 4.3.1)
 Matrix                   1.6-3     2023-11-14 [1] CRAN (R 4.3.1)
 MatrixGenerics         * 1.14.0    2023-10-26 [1] Bioconductor
 matrixStats            * 1.1.0     2023-11-07 [1] CRAN (R 4.3.1)
 memoise                  2.0.1     2021-11-26 [1] CRAN (R 4.3.0)
 metapod                  1.10.0    2023-10-24 [1] Bioconductor
 mime                     0.12      2021-09-28 [1] CRAN (R 4.3.0)
 munsell                  0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 paletteer                1.5.0     2022-10-19 [1] CRAN (R 4.3.0)
 pillar                   1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig                2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 plotly                   4.10.3    2023-10-21 [1] CRAN (R 4.3.1)
 png                      0.1-8     2022-11-29 [1] CRAN (R 4.3.0)
 promises                 1.2.1     2023-08-10 [1] CRAN (R 4.3.0)
 purrr                    1.0.2     2023-08-10 [1] CRAN (R 4.3.0)
 R6                       2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs                 0.3.3     2021-01-31 [1] CRAN (R 4.3.0)
 RColorBrewer             1.1-3     2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp                     1.0.11    2023-07-06 [1] CRAN (R 4.3.1)
 RCurl                    1.98-1.13 2023-11-02 [1] CRAN (R 4.3.1)
 rematch2                 2.1.2     2020-05-01 [1] CRAN (R 4.3.0)
 restfulr                 0.0.15    2022-06-16 [1] CRAN (R 4.3.0)
 rjson                    0.2.21    2022-01-09 [1] CRAN (R 4.3.0)
 rlang                  * 1.1.2     2023-11-04 [1] CRAN (R 4.3.1)
 rmarkdown                2.25      2023-09-18 [1] CRAN (R 4.3.1)
 rprojroot                2.0.4     2023-11-05 [1] CRAN (R 4.3.1)
 Rsamtools                2.18.0    2023-10-24 [1] Bioconductor
 RSQLite                  2.3.3     2023-11-04 [1] CRAN (R 4.3.1)
 rstudioapi               0.15.0    2023-07-07 [1] CRAN (R 4.3.0)
 rsvd                     1.0.5     2021-04-16 [1] CRAN (R 4.3.0)
 rtracklayer              1.62.0    2023-10-26 [1] Bioconductor
 S4Arrays                 1.2.0     2023-10-26 [1] Bioconductor
 S4Vectors              * 0.40.1    2023-10-26 [1] Bioconductor
 sass                     0.4.7     2023-07-15 [1] CRAN (R 4.3.0)
 ScaledMatrix             1.10.0    2023-10-24 [1] Bioconductor
 scales                   1.2.1     2022-08-20 [1] CRAN (R 4.3.0)
 scater                 * 1.30.1    2023-11-16 [1] Bioconductor
 scran                  * 1.30.0    2023-10-24 [1] Bioconductor
 scuttle                * 1.12.0    2023-10-24 [1] Bioconductor
 sessioninfo              1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 shiny                    1.8.0     2023-11-17 [1] CRAN (R 4.3.1)
 shinyWidgets             0.8.0     2023-08-30 [1] CRAN (R 4.3.0)
 SingleCellExperiment   * 1.24.0    2023-11-06 [1] Bioconductor
 spam                     2.10-0    2023-10-23 [1] CRAN (R 4.3.1)
 SparseArray              1.2.2     2023-11-08 [1] Bioconductor
 sparseMatrixStats        1.14.0    2023-10-24 [1] Bioconductor
 SpatialExperiment      * 1.12.0    2023-10-24 [1] Bioconductor
 spatialLIBD            * 1.13.4    2023-05-25 [1] Bioconductor
 statmod                  1.5.0     2023-01-06 [1] CRAN (R 4.3.0)
 SummarizedExperiment   * 1.32.0    2023-11-06 [1] Bioconductor
 tibble                   3.2.1     2023-03-20 [1] CRAN (R 4.3.0)
 tidyr                    1.3.0     2023-01-24 [1] CRAN (R 4.3.0)
 tidyselect               1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 utf8                     1.2.4     2023-10-22 [1] CRAN (R 4.3.1)
 vctrs                    0.6.4     2023-10-12 [1] CRAN (R 4.3.1)
 vipor                    0.4.5     2017-03-22 [1] CRAN (R 4.3.0)
 viridis                  0.6.4     2023-07-22 [1] CRAN (R 4.3.0)
 viridisLite              0.4.2     2023-05-02 [1] CRAN (R 4.3.0)
 withr                    2.5.2     2023-10-30 [1] CRAN (R 4.3.1)
 xfun                     0.41      2023-11-01 [1] CRAN (R 4.3.1)
 XML                      3.99-0.15 2023-11-02 [1] CRAN (R 4.3.1)
 xtable                   1.8-4     2019-04-21 [1] CRAN (R 4.3.0)
 XVector                  0.42.0    2023-10-26 [1] Bioconductor
 yaml                     2.3.7     2023-01-23 [1] CRAN (R 4.3.0)
 zlibbioc                 1.48.0    2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

───────────────────────────────────────────────────────────────────────────
