---
title: "06e-hmapVisVMH_pseudobulk_logCPM_spear_to_Kim19APmkSets"
author: "Bernie Mulvey"
date: "2025-04-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(dreamlet)
library(limma)
library(edgeR)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##


## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure.
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel)

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```


read in: mouse AP marker stats as calculated in 06a and 06b
```{r}
fig2mks <- fread("processed-data/12-Mouse comparisons/06a-Kim19msVMH_Fig2_APclustsOnly_smartseq_markerTstat.txt")
fig2mks[,gene:=NULL]

ss.apmk <- fread("processed-data/12-Mouse comparisons/06b-AllKim19_sFISH-matched-smartseq-clusts_mkTstats_by_peakAPloc.txt")
ss.apmk[,gene:=NULL]
```


load spatial data, convert to SPE;
obtained file from https://www.repository.cam.ac.uk/bitstreams/a47399a8-2735-4b65-b6cb-3cd1a48b08d7/download
```{r}
hmap <- readRDS("~/Desktop/humanHYPOMAP_spatial.rds")

## convert to SPE using code from https://github.com/drighelli/SpatialExperiment/issues/115 
seurat_to_spe <- function(seu, sample_id, img_id) {
    
    ## Convert to SCE
    sce <- Seurat::as.SingleCellExperiment(seu)
    
    ## Extract spatial coordinates
    spatialCoords <- as.matrix(
        seu@images[[img_id]]@coordinates[, c("imagecol", "imagerow")])
    
    ## Extract and process image data
    img <- SpatialExperiment::SpatialImage(
        x = as.raster(seu@images[[img_id]]@image))
    
    imgData <- DataFrame(
        sample_id = sample_id,
        image_id = img_id,
        data = I(list(img)),
        scaleFactor = seu@images[[img_id]]@scale.factors$lowres)
    
    # Convert to SpatialExperiment
    spe <- SpatialExperiment(
        assays = assays(sce),
        rowData = rowData(sce),
        colData = colData(sce),
        metadata = metadata(sce),
        reducedDims = reducedDims(sce),
        altExps = altExps(sce),
        sample_id = sample_id,
        spatialCoords = spatialCoords,
        imgData = imgData
    )
    # indicate all spots are on the tissue since the data is already processed
    spe$in_tissue <- 1
    spe$sample_id <- sample_id
    # Return Spatial Experiment object
    spe
}


## sample_ids and image_ids are permutations of the following:
## sampleid = hmap$captureArea (b2,a3,...)
## image id is a permutation of this (slice2B, slice3A..)


spat_metas <- list(c("b2","slice2B"),
                   c("a3","slice3A"),
                   c("b4","slice4B"),
                   c("a5","slice5A"),
                   c("b6","slice6B"),
                   c("a7","slice7A"),
                   c("b8","slice8B"),
                   c("b1_b","sliceB1B"),
                   c("c1_a","sliceC1A"))
names(spat_metas) <- unlist(lapply(spat_metas,function(x){x[2]}))

spe_ls <- lapply(spat_metas, function(m) {
    seurat_to_spe(seu = hmap[,hmap$captureArea==m[1]], sample_id = m[1], img_id = m[2])
})

spe <- Reduce(cbind, spe_ls)
rm(hmap,spat_metas,spe_ls)
gc(full=T)
```


## subset spe to VMH and shared 1-1 ortholog genes (specifically those we also had for comparing to vVMH--already the only ones in these tables). we'll probably have to do two separate sets since different numbers of genes pass marker analysis minima in 06a and 06b
```{r}
# vmhspots <- unique(c(
#    grep(spe$regional_clusters_named,pattern="VMH",value=T),grep(spe$regional_clusters_grouped,pattern="VMH",value=T))) 
# oh, they're all just called VMH in both cols.
# rm(vmhspots)

spe <- spe[,spe$regional_clusters_named=="VMH"|spe$regional_clusters_grouped=="VMH"]

# the hypomap spatial only has gene symbols, no ENSGs
# get genes that are in at least one of the 2 mouse marker analyses and hypomap spat VMH
orths <- fread("processed-data/12-Mouse comparisons/06a-1to1_orths_in_KimSmartseq_and_vVMH.txt")
keepg <- orths[ortholog_name %in% rownames(spe) & orths$ortholog_ensg %in% c(fig2mks$human_ensg,ss.apmk$human_ensg),.(ortholog_name,ortholog_ensg)]
spe <- spe[keepg$ortholog_name,]
```

pseudobulk by sample
```{r}
## check that the regional clusters labels, named and grouped, are equivalent for VMH (i.e., whether it matters which we pass as the label column for pbulking)
stopifnot(sum(spe$regional_clusters_named!=spe$regional_clusters_grouped)==0)

# run dreamlet only to pseudobulk each cluster or collapsed clusters 
ag <- dreamlet::aggregateToPseudoBulk(spe,sample_id="sample_id",cluster_id = "regional_clusters_named")

# extract colData to feed into DGEList as the "samples" data
agcd <- as.data.frame(colData(ag))

# get the aggregated counts
ag <- assays(ag)[[1]] # there is only 1 here
  
## make the DGEList:
X <- ag[,colSums(ag)>0]
s <- copy(agcd)[colnames(X),]

dge <- DGEList(counts=as.matrix(X),samples=s)
dge <- normLibSizes(dge,method = "TMM")
  
## find the smaller group for this cluster -- this is for consistency with vVMH pbulk filtering steps
smolgrp <- table(dge$samples$Sex)[which.min(table(dge$samples$Sex))]
  
## for the smallest group, determine low expression cutoff
rowmn.m <- as.numeric(rowMeans(cpm(dge[,(dge$samples$Sex==names(smolgrp)[1])],log=TRUE)))
  
## get histogram distribution with 10 bins at the low end of the logcpm space
h <- hist(rowmn.m[which(rowmn.m>min(rowmn.m)&rowmn.m<(4+min(rowmn.m)))],breaks=10,plot=FALSE)
  
## get the minimum in this low logcpm range and its corresponding bin (x axis value, i.e. log cpm)
minc <- h$breaks[which(h$counts==min(h$counts))]
  
## get the corresponding smallest-library count value for this minimum^
smollib <- min(dge$samples[(dge$samples$Sex==names(smolgrp)[1]),"lib.size"])
countthresh <- (minc^2)*(as.numeric(smollib)/10^6)
  
## and convert that smallest count to the logcpm value in said smallest library to use as a low-expression threshold
cpmthresh <- log2(countthresh/(as.numeric(smollib)*10^-6))
  
## filter for genes with at least thresh counts in at least (smaller sex group size) samples
keep <- rowSums(cpm(dge,log=T)>cpmthresh) >= as.numeric(smolgrp)
## 10k genes
  
## recalc lib factors after dropping low-expressed genes
dge <- dge[keep,,keep.lib.sizes=FALSE]
dge <- calcNormFactors(dge,method = "TMM")

### run voomLmFit and extract EList. they have one donor sampled twice here so we can specify that in the voomLmFit call, matching our modeling approach for vvmh
des <- model.matrix(~0+Sex, data = dge$samples)
# INITIALIZE the voomwithqualityweights model, equivalent to pseudobulkDGE(method="voom",qualities=T)
sexdif <- makeContrasts(SexMale-SexFemale,levels=des)

# run voomLmFit for the pseudobulked data, referring donor to duplicateCorrelation; using an adaptive span (number of genes, based on the number of genes in the dge) for smoothing the mean-variance trend
v.swt <- voomLmFit(dge,design = des,block = as.factor(dge$samples$donorID),adaptive.span = T,sample.weights = T)

## get the Elist
v.swt.elist <- as.data.table(v.swt$EList$E,keep.rownames=T)

## append orthology/ensg ids
v.swt.elist <- merge.data.table(v.swt.elist,orths[,.(ortholog_name,ortholog_ensg)],by.x="rn",by.y="ortholog_name")
v.swt.elist[,rn:=NULL]
```

cleanup
```{r}
rm(ag,agcd,des,h,s,keepg,sexdif,spe,v.swt,X,countthresh,cpmthresh,keep,minc,rowmn.m,smolgrp,smollib)
gc(full=T)
```

run correlation tests

run correls for each pb sample
```{r}
sampcor.list <- lapply(list(fig2mks,ss.apmk),FUN=function(m){
    
    samps <- names(v.swt.elist)[names(v.swt.elist)!="ortholog_ensg"]
    sampcors <- lapply(samps,FUN=function(s){
        
        pbnames <- c("ortholog_ensg",s)
        
        tmpd <- merge.data.table(v.swt.elist[,..pbnames],m,by.x="ortholog_ensg",by.y="human_ensg")
        sdnames <- names(tmpd)[!(names(tmpd) %in% c("ortholog_ensg",s))]
        
        ans <- tmpd[,lapply(.SD,FUN=function(x){cor(get(s),x,method="spearman")}),.SDcols=sdnames]
        ans[,samp:=s] 
        ans <- melt.data.table(ans,id.vars="samp")

        return(ans)
    })
    rettab <- rbindlist(sampcors)
    return(rettab)
})


# save
saveRDS(sampcor.list,"processed-data/12-Mouse comparisons/06e-hypomapVisVMH_pseudobulk_logCPM_spear_to_Kim19_APmkSets.RDS")
```

repinf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.4

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] parallelly_1.43.0           colorout_1.3-0.2            edgeR_4.4.2                
 [4] dreamlet_1.4.1              variancePartition_1.36.3    BiocParallel_1.40.0        
 [7] limma_3.62.2                ggplot2_3.5.1               SpatialExperiment_1.16.0   
[10] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0 Biobase_2.66.0             
[13] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3         IRanges_2.40.1             
[16] S4Vectors_0.44.0            BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[19] matrixStats_1.5.0           data.table_1.17.0           rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.1-0     bitops_1.0-9              httr_1.4.7               
  [4] RColorBrewer_1.1-3        Rgraphviz_2.50.0          numDeriv_2016.8-1.1      
  [7] sctransform_0.4.1         tools_4.4.3               backports_1.5.0          
 [10] R6_2.6.1                  metafor_4.8-0             uwot_0.2.3               
 [13] lazyeval_0.2.2            withr_3.0.2               sp_2.2-0                 
 [16] gridExtra_2.3             prettyunits_1.2.0         progressr_0.15.1         
 [19] cli_3.6.4                 spatstat.explore_3.4-2    fastDummies_1.7.5        
 [22] KEGGgraph_1.66.0          Seurat_5.2.1              spatstat.data_3.1-6      
 [25] SQUAREM_2021.1            mvtnorm_1.3-3             pbapply_1.7-2            
 [28] ggridges_0.5.6            mixsqp_0.3-54             zenith_1.8.0             
 [31] invgamma_1.1              rstudioapi_0.17.1         RSQLite_2.3.9            
 [34] generics_0.1.3            spatstat.random_3.3-3     gtools_3.9.5             
 [37] ica_1.0-3                 dplyr_1.1.4               Matrix_1.7-3             
 [40] metadat_1.4-0             ggbeeswarm_0.7.2          abind_1.4-8              
 [43] lifecycle_1.0.4           yaml_2.3.10               mathjaxr_1.6-0           
 [46] gplots_3.2.0              SparseArray_1.6.2         Rtsne_0.17               
 [49] grid_4.4.3                blob_1.2.4                promises_1.3.2           
 [52] crayon_1.5.3              miniUI_0.1.1.1            lattice_0.22-6           
 [55] msigdbr_10.0.1            cowplot_1.1.3             annotate_1.84.0          
 [58] KEGGREST_1.46.0           magick_2.8.6              pillar_1.10.1            
 [61] knitr_1.50                rjson_0.2.23              boot_1.3-31              
 [64] corpcor_1.6.10            future.apply_1.11.3       codetools_0.2-20         
 [67] glue_1.8.0                spatstat.univar_3.1-2     vctrs_0.6.5              
 [70] png_0.1-8                 spam_2.11-1               Rdpack_2.6.3             
 [73] gtable_0.3.6              assertthat_0.2.1          cachem_1.1.0             
 [76] zigg_0.0.2                xfun_0.51                 mime_0.13                
 [79] rbibutils_2.3             S4Arrays_1.6.0            Rfast_2.1.5.1            
 [82] reformulas_0.4.0          survival_3.8-3            iterators_1.0.14         
 [85] statmod_1.5.0             fitdistrplus_1.2-2        ROCR_1.0-11              
 [88] nlme_3.1-167              pbkrtest_0.5.3            bit64_4.6.0-1            
 [91] progress_1.2.3            EnvStats_3.0.0            RcppAnnoy_0.0.22         
 [94] rprojroot_2.0.4           irlba_2.3.5.1             vipor_0.4.7              
 [97] KernSmooth_2.23-26        colorspace_2.1-1          rmeta_3.0                
[100] DBI_1.2.3                 tidyselect_1.2.1          bit_4.6.0                
[103] compiler_4.4.3            graph_1.84.1              plotly_4.10.4            
[106] DelayedArray_0.32.0       scales_1.3.0              caTools_1.18.3           
[109] remaCor_0.0.18            lmtest_0.9-40             goftest_1.2-3            
[112] stringr_1.5.1             digest_0.6.37             spatstat.utils_3.1-3     
[115] minqa_1.2.8               rmarkdown_2.29            aod_1.3.3                
[118] XVector_0.46.0            RhpcBLASctl_0.23-42       htmltools_0.5.8.1        
[121] pkgconfig_2.0.3           lme4_1.1-36               sparseMatrixStats_1.18.0 
[124] mashr_0.2.79              fastmap_1.2.0             htmlwidgets_1.6.4        
[127] UCSC.utils_1.2.0          shiny_1.10.0              DelayedMatrixStats_1.28.1
[130] farver_2.1.2              zoo_1.8-13                jsonlite_1.9.1           
[133] RCurl_1.98-1.17           magrittr_2.0.3            GenomeInfoDbData_1.2.13  
[136] dotCall64_1.2             patchwork_1.3.0           munsell_0.5.1            
[139] Rcpp_1.0.14               reticulate_1.41.0.1       babelgene_22.9           
[142] EnrichmentBrowser_2.36.0  stringi_1.8.4             zlibbioc_1.52.0          
[145] MASS_7.3-65               plyr_1.8.9                listenv_0.9.1            
[148] ggrepel_0.9.6             deldir_2.0-4              Biostrings_2.74.1        
[151] splines_4.4.3             tensor_1.5                hms_1.1.3                
[154] locfit_1.5-9.12           igraph_2.1.4              spatstat.geom_3.3-6      
[157] RcppHNSW_0.6.0            reshape2_1.4.4            XML_3.99-0.18            
[160] evaluate_1.0.3            SeuratObject_5.0.2        RcppParallel_5.1.10      
[163] nloptr_2.2.1              httpuv_1.6.15             polyclip_1.10-7          
[166] RANN_2.6.2                tidyr_1.3.1               purrr_1.0.4              
[169] future_1.34.0             scattermore_1.2           ashr_2.2-63              
[172] broom_1.0.7               xtable_1.8-4              fANCOVA_0.6-1            
[175] RSpectra_0.16-2           later_1.4.1               viridisLite_0.4.2        
[178] truncnorm_1.0-9           tibble_3.2.1              lmerTest_3.1-3           
[181] memoise_2.0.1             beeswarm_0.4.0            AnnotationDbi_1.68.0     
[184] cluster_2.1.8.1           globals_0.16.3            GSEABase_1.68.0          
[187] here_1.0.1               
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sonoma 14.7.4
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-04-09
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ───────────────────────────────────────────────────────────────────────────
 ! package              * version    date (UTC) lib source
   abind                  1.4-8      2024-09-12 [1] CRAN (R 4.4.1)
   annotate               1.84.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   AnnotationDbi          1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   aod                    1.3.3      2023-12-13 [1] CRAN (R 4.4.0)
   ashr                   2.2-63     2023-08-21 [1] CRAN (R 4.4.0)
   assertthat             0.2.1      2019-03-21 [1] CRAN (R 4.4.0)
   babelgene              22.9       2022-09-29 [1] CRAN (R 4.4.0)
   backports              1.5.0      2024-05-23 [1] CRAN (R 4.4.0)
   beeswarm               0.4.0      2021-06-01 [1] CRAN (R 4.4.0)
   Biobase              * 2.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocGenerics         * 0.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocParallel         * 1.40.0     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
   Biostrings             2.74.1     2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
   bit                    4.6.0      2025-03-06 [1] CRAN (R 4.4.1)
   bit64                  4.6.0-1    2025-01-16 [1] CRAN (R 4.4.1)
   bitops                 1.0-9      2024-10-03 [1] CRAN (R 4.4.1)
   blob                   1.2.4      2023-03-17 [1] CRAN (R 4.4.0)
   boot                   1.3-31     2024-08-28 [1] CRAN (R 4.4.3)
   broom                  1.0.7      2024-09-26 [1] CRAN (R 4.4.1)
   cachem                 1.1.0      2024-05-16 [1] CRAN (R 4.4.0)
   caTools                1.18.3     2024-09-04 [1] CRAN (R 4.4.1)
 P cli                    3.6.4      2025-02-13 [2] CRAN (R 4.4.1)
   cluster                2.1.8.1    2025-03-12 [1] CRAN (R 4.4.1)
   codetools              0.2-20     2024-03-31 [1] CRAN (R 4.4.3)
   colorout             * 1.3-0.2    2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace             2.1-1      2024-07-26 [1] CRAN (R 4.4.0)
   corpcor                1.6.10     2021-09-16 [1] CRAN (R 4.4.0)
   cowplot                1.1.3      2024-01-22 [1] CRAN (R 4.4.0)
   crayon                 1.5.3      2024-06-20 [1] CRAN (R 4.4.0)
   data.table           * 1.17.0     2025-02-22 [1] CRAN (R 4.4.1)
   DBI                    1.2.3      2024-06-02 [1] CRAN (R 4.4.0)
   DelayedArray           0.32.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   DelayedMatrixStats     1.28.1     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
   deldir                 2.0-4      2024-02-28 [1] CRAN (R 4.4.0)
   digest                 0.6.37     2024-08-19 [1] CRAN (R 4.4.1)
   dotCall64              1.2        2024-10-04 [1] CRAN (R 4.4.1)
   dplyr                  1.1.4      2023-11-17 [1] CRAN (R 4.4.0)
   dreamlet             * 1.4.1      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   edgeR                * 4.4.2      2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
   EnrichmentBrowser      2.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   EnvStats               3.0.0      2024-08-24 [1] CRAN (R 4.4.1)
   evaluate               1.0.3      2025-01-10 [1] CRAN (R 4.4.1)
   fANCOVA                0.6-1      2020-11-13 [1] CRAN (R 4.4.0)
   farver                 2.1.2      2024-05-13 [1] CRAN (R 4.4.0)
   fastDummies            1.7.5      2025-01-20 [1] CRAN (R 4.4.1)
   fastmap                1.2.0      2024-05-15 [1] CRAN (R 4.4.0)
   fitdistrplus           1.2-2      2025-01-07 [1] CRAN (R 4.4.1)
   future                 1.34.0     2024-07-29 [1] CRAN (R 4.4.0)
   future.apply           1.11.3     2024-10-27 [1] CRAN (R 4.4.1)
   generics               0.1.3      2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb         * 1.42.3     2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
   GenomeInfoDbData       1.2.13     2024-12-12 [1] Bioconductor
   GenomicRanges        * 1.58.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   ggbeeswarm             0.7.2      2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2              * 3.5.1      2024-04-23 [1] CRAN (R 4.4.0)
   ggrepel                0.9.6      2024-09-07 [1] CRAN (R 4.4.1)
   ggridges               0.5.6      2024-01-23 [1] CRAN (R 4.4.0)
   globals                0.16.3     2024-03-08 [1] CRAN (R 4.4.0)
   glue                   1.8.0      2024-09-30 [1] CRAN (R 4.4.1)
   goftest                1.2-3      2021-10-07 [1] CRAN (R 4.4.0)
   gplots                 3.2.0      2024-10-05 [1] CRAN (R 4.4.1)
   graph                  1.84.1     2025-01-02 [1] Bioconductor 3.20 (R 4.4.2)
   gridExtra              2.3        2017-09-09 [1] CRAN (R 4.4.0)
   GSEABase               1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   gtable                 0.3.6      2024-10-25 [1] CRAN (R 4.4.1)
   gtools                 3.9.5      2023-11-20 [1] CRAN (R 4.4.0)
   here                   1.0.1      2020-12-13 [1] CRAN (R 4.4.0)
   hms                    1.1.3      2023-03-21 [1] CRAN (R 4.4.0)
   htmltools              0.5.8.1    2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets            1.6.4      2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                 1.6.15     2024-03-26 [1] CRAN (R 4.4.0)
   httr                   1.4.7      2023-08-15 [1] CRAN (R 4.4.0)
   ica                    1.0-3      2022-07-08 [1] CRAN (R 4.4.0)
   igraph                 2.1.4      2025-01-23 [1] CRAN (R 4.4.1)
   invgamma               1.1        2017-05-07 [1] CRAN (R 4.4.0)
   IRanges              * 2.40.1     2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
   irlba                  2.3.5.1    2022-10-03 [1] CRAN (R 4.4.0)
   iterators              1.0.14     2022-02-05 [1] CRAN (R 4.4.0)
   jsonlite               1.9.1      2025-03-03 [1] CRAN (R 4.4.1)
   KEGGgraph              1.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   KEGGREST               1.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   KernSmooth             2.23-26    2025-01-01 [1] CRAN (R 4.4.3)
   knitr                  1.50       2025-03-16 [1] CRAN (R 4.4.2)
   later                  1.4.1      2024-11-27 [1] CRAN (R 4.4.1)
   lattice                0.22-6     2024-03-20 [1] CRAN (R 4.4.3)
   lazyeval               0.2.2      2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle              1.0.4      2023-11-07 [1] CRAN (R 4.4.0)
   limma                * 3.62.2     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
   listenv                0.9.1      2024-01-29 [1] CRAN (R 4.4.0)
   lme4                   1.1-36     2025-01-11 [1] CRAN (R 4.4.1)
   lmerTest               3.1-3      2020-10-23 [1] CRAN (R 4.4.0)
   lmtest                 0.9-40     2022-03-21 [1] CRAN (R 4.4.0)
   locfit                 1.5-9.12   2025-03-05 [1] CRAN (R 4.4.1)
   magick                 2.8.6      2025-03-23 [1] CRAN (R 4.4.1)
   magrittr               2.0.3      2022-03-30 [1] CRAN (R 4.4.0)
   mashr                  0.2.79     2023-10-18 [1] CRAN (R 4.4.0)
   MASS                   7.3-65     2025-02-28 [1] CRAN (R 4.4.1)
   mathjaxr               1.6-0      2022-02-28 [1] CRAN (R 4.4.0)
   Matrix                 1.7-3      2025-03-11 [1] CRAN (R 4.4.1)
   MatrixGenerics       * 1.18.1     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
   matrixStats          * 1.5.0      2025-01-07 [1] CRAN (R 4.4.1)
   memoise                2.0.1      2021-11-26 [1] CRAN (R 4.4.0)
   metadat                1.4-0      2025-02-04 [1] CRAN (R 4.4.1)
   metafor                4.8-0      2025-01-28 [1] CRAN (R 4.4.1)
   mime                   0.13       2025-03-17 [1] CRAN (R 4.4.1)
   miniUI                 0.1.1.1    2018-05-18 [1] CRAN (R 4.4.0)
   minqa                  1.2.8      2024-08-17 [1] CRAN (R 4.4.0)
   mixsqp                 0.3-54     2023-12-20 [1] CRAN (R 4.4.0)
   msigdbr                10.0.1     2025-03-19 [1] CRAN (R 4.4.3)
   munsell                0.5.1      2024-04-01 [1] CRAN (R 4.4.0)
   mvtnorm                1.3-3      2025-01-10 [1] CRAN (R 4.4.1)
   nlme                   3.1-167    2025-01-27 [1] CRAN (R 4.4.3)
   nloptr                 2.2.1      2025-03-17 [1] CRAN (R 4.4.1)
   numDeriv               2016.8-1.1 2019-06-06 [1] CRAN (R 4.4.0)
   parallelly           * 1.43.0     2025-03-24 [1] CRAN (R 4.4.1)
   patchwork              1.3.0      2024-09-16 [1] CRAN (R 4.4.1)
   pbapply                1.7-2      2023-06-27 [1] CRAN (R 4.4.0)
   pbkrtest               0.5.3      2024-06-26 [1] CRAN (R 4.4.0)
   pillar                 1.10.1     2025-01-07 [1] CRAN (R 4.4.1)
   pkgconfig              2.0.3      2019-09-22 [1] CRAN (R 4.4.0)
   plotly                 4.10.4     2024-01-13 [1] CRAN (R 4.4.0)
   plyr                   1.8.9      2023-10-02 [1] CRAN (R 4.4.0)
   png                    0.1-8      2022-11-29 [1] CRAN (R 4.4.0)
   polyclip               1.10-7     2024-07-23 [1] CRAN (R 4.4.0)
   prettyunits            1.2.0      2023-09-24 [1] CRAN (R 4.4.0)
   progress               1.2.3      2023-12-06 [1] CRAN (R 4.4.0)
   progressr              0.15.1     2024-11-22 [1] CRAN (R 4.4.1)
   promises               1.3.2      2024-11-28 [1] CRAN (R 4.4.1)
   purrr                  1.0.4      2025-02-05 [1] CRAN (R 4.4.1)
   R6                     2.6.1      2025-02-15 [1] CRAN (R 4.4.1)
   RANN                   2.6.2      2024-08-25 [1] CRAN (R 4.4.1)
   rbibutils              2.3        2024-10-04 [1] CRAN (R 4.4.1)
   RColorBrewer           1.1-3      2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                   1.0.14     2025-01-12 [1] CRAN (R 4.4.1)
   RcppAnnoy              0.0.22     2024-01-23 [1] CRAN (R 4.4.0)
   RcppHNSW               0.6.0      2024-02-04 [1] CRAN (R 4.4.0)
   RcppParallel           5.1.10     2025-01-24 [1] CRAN (R 4.4.1)
   RCurl                  1.98-1.17  2025-03-22 [1] CRAN (R 4.4.1)
   Rdpack                 2.6.3      2025-03-16 [1] CRAN (R 4.4.2)
   reformulas             0.4.0      2024-11-03 [1] CRAN (R 4.4.1)
   remaCor                0.0.18     2024-02-08 [1] CRAN (R 4.4.0)
   reshape2               1.4.4      2020-04-09 [1] CRAN (R 4.4.0)
   reticulate             1.41.0.1   2025-03-09 [1] CRAN (R 4.4.1)
   Rfast                  2.1.5.1    2025-03-14 [1] CRAN (R 4.4.1)
   Rgraphviz              2.50.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   RhpcBLASctl            0.23-42    2023-02-11 [1] CRAN (R 4.4.0)
   rjson                  0.2.23     2024-09-16 [1] CRAN (R 4.4.1)
 P rlang                * 1.1.5      2025-01-17 [2] CRAN (R 4.4.1)
   rmarkdown              2.29       2024-11-04 [1] CRAN (R 4.4.1)
   rmeta                  3.0        2018-03-20 [1] CRAN (R 4.4.0)
   ROCR                   1.0-11     2020-05-02 [1] CRAN (R 4.4.0)
   rprojroot              2.0.4      2023-11-05 [1] CRAN (R 4.4.0)
   RSpectra               0.16-2     2024-07-18 [1] CRAN (R 4.4.0)
   RSQLite                2.3.9      2024-12-03 [1] CRAN (R 4.4.1)
   rstudioapi             0.17.1     2024-10-22 [1] CRAN (R 4.4.1)
   Rtsne                  0.17       2023-12-07 [1] CRAN (R 4.4.0)
   S4Arrays               1.6.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   S4Vectors            * 0.44.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   scales                 1.3.0      2023-11-28 [1] CRAN (R 4.4.0)
   scattermore            1.2        2023-06-12 [1] CRAN (R 4.4.0)
   sctransform            0.4.1      2023-10-19 [1] CRAN (R 4.4.0)
   sessioninfo            1.2.3      2025-02-05 [1] CRAN (R 4.4.1)
   Seurat                 5.2.1      2025-01-24 [1] CRAN (R 4.4.1)
   SeuratObject           5.0.2      2024-05-08 [1] CRAN (R 4.4.0)
   shiny                  1.10.0     2024-12-14 [1] CRAN (R 4.4.1)
   SingleCellExperiment * 1.28.1     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
   sp                     2.2-0      2025-02-01 [1] CRAN (R 4.4.1)
   spam                   2.11-1     2025-01-20 [1] CRAN (R 4.4.1)
   SparseArray            1.6.2      2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
   sparseMatrixStats      1.18.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   SpatialExperiment    * 1.16.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   spatstat.data          3.1-6      2025-03-17 [1] CRAN (R 4.4.2)
   spatstat.explore       3.4-2      2025-03-21 [1] CRAN (R 4.4.1)
   spatstat.geom          3.3-6      2025-03-18 [1] CRAN (R 4.4.1)
   spatstat.random        3.3-3      2025-03-19 [1] CRAN (R 4.4.1)
   spatstat.sparse        3.1-0      2024-06-21 [1] CRAN (R 4.4.0)
   spatstat.univar        3.1-2      2025-03-05 [1] CRAN (R 4.4.1)
   spatstat.utils         3.1-3      2025-03-15 [1] CRAN (R 4.4.1)
   SQUAREM                2021.1     2021-01-13 [1] CRAN (R 4.4.0)
   statmod                1.5.0      2023-01-06 [1] CRAN (R 4.4.0)
   stringi                1.8.4      2024-05-06 [1] CRAN (R 4.4.0)
   stringr                1.5.1      2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment * 1.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   survival               3.8-3      2024-12-17 [1] CRAN (R 4.4.3)
   tensor                 1.5        2012-05-05 [1] CRAN (R 4.4.0)
   tibble                 3.2.1      2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                  1.3.1      2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect             1.2.1      2024-03-11 [1] CRAN (R 4.4.0)
   truncnorm              1.0-9      2023-03-20 [1] CRAN (R 4.4.0)
   UCSC.utils             1.2.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   uwot                   0.2.3      2025-02-24 [1] CRAN (R 4.4.1)
   variancePartition    * 1.36.3     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
   vctrs                  0.6.5      2023-12-01 [1] CRAN (R 4.4.0)
   vipor                  0.4.7      2023-12-18 [1] CRAN (R 4.4.0)
   viridisLite            0.4.2      2023-05-02 [1] CRAN (R 4.4.0)
   withr                  3.0.2      2024-10-28 [1] CRAN (R 4.4.1)
   xfun                   0.51       2025-02-19 [1] CRAN (R 4.4.1)
   XML                    3.99-0.18  2025-01-01 [1] CRAN (R 4.4.1)
   xtable                 1.8-4      2019-04-21 [1] CRAN (R 4.4.0)
   XVector                0.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   yaml                   2.3.10     2024-07-26 [1] CRAN (R 4.4.0)
   zenith                 1.8.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   zigg                   0.0.2      2025-02-07 [1] CRAN (R 4.4.1)
   zlibbioc               1.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   zoo                    1.8-13     2025-02-22 [1] CRAN (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────────────
