---
title: "02c_raw_spe_Jul2024"
output: html_document
date: "2024-07-16"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

# not executed from JHPCE (summer '24 downtime), but if running from JHPCE:
# setwd('/dcs04/lieber/marmaypag/spatialHYP_LIBD4195/spatial_HYP/')

library(data.table)
library(Biostrings)
library(SpatialExperiment)
library(spatialLIBD)
library(rtracklayer)
library(sessioninfo)
library(here)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure.
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")

```

```{r}

# setwd('/dcs04/lieber/marmaypag/spatialHYP_LIBD4195/spatial_HYP/')


## Define some info for the samples -- innacessible during JHPCE downtime, but key components are in raw-data/demos
# load(here::here("code", "REDCap", "REDCap_HYP.rda"))

demos <- fread("raw-data/demos_by_sampleid.txt")
demos <- demos[sample_id=="V13Y24-346_C1"]
demos[,slide:="V13Y24-346"]
demos[,array:="C1"]
demos[,species:="Human"]
demos[,replicate:=1]
demos[,sample_path:="processed-data/01_spaceranger/V13Y24-346_C1/outs"]


## Build basic SPE
Sys.time()

# 2024-07-16 12:29:17 CDT

### reference gtf not available w/ jhpce down, retrieved from https://www.10xgenomics.com/support/software/cell-ranger/latest/release-notes/cr-reference-release-notes#2020-a
# from JHPCE, argument would be: 
# reference_gtf = file.path("/dcs04/lieber/lcolladotor/annotationFiles_LIBD001/10x/refdata-gex-GRCh38-2020-A/","genes", "genes.gtf")

spe <- read10xVisiumWrapper(
    demos$sample_path,
    demos$sample_id,
    type = "sparse",
    data = "raw",
    images = c("lowres", "hires", "detected", "aligned"),
    load = TRUE,
    reference_gtf = "~/Desktop/refdata-gex-GRCh38-2020-A/genes/genes.gtf"
)
 
# 2024-07-16 12:31:05.830659 SpatialExperiment::read10xVisium: reading basic data from SpaceRanger
# 2024-07-16 12:31:08.845009 read10xVisiumAnalysis: reading analysis output from SpaceRanger
# 2024-07-16 12:31:08.919168 add10xVisiumAnalysis: adding analysis output from SpaceRanger
# 2024-07-16 12:31:09.040988 rtracklayer::import: reading the reference GTF file
# 2024-07-16 12:31:30.480865 adding gene information to the SPE object
# 2024-07-16 12:31:30.500885 adding information used by spatialLIBD

## Add the study design info
add_design <- function(spe) {
    new_col <- merge(colData(spe), demos)
    ## Fix order
    new_col <- new_col[match(spe$key, new_col$key), ]
    stopifnot(identical(new_col$key, spe$key))
    rownames(new_col) <- rownames(colData(spe))
    colData(spe) <-
        new_col[, -which(colnames(new_col) == "sample_path")]
    return(spe)
}
spe <- add_design(spe)

# make  colnames unique (spe$key)
colnames(spe) <- spe$key

# Not applicable: Read in cell counts and segmentation results
# segmentations_list <-
#   lapply(demos$sample_id, function(sampleid) {
#     file <-
#       here(
#         "processed-data",
#         "01_spaceranger",
#         "spaceranger_230511",
#         sampleid,
#         "outs",
#         "spatial",
#         "tissue_spot_counts.csv"
#       )
#     if (!file.exists(file)) {
#       return(NULL)
#     }
#     x <- read.csv(file)
#     x$key <- paste0(x$barcode, "_", sampleid)
#     return(x)
#   })
# 
# ## Merge them (once the these files are done, this could be replaced by an rbind)
# segmentations <-
#   Reduce(function(...) {
#     merge(..., all = TRUE)
#   }, segmentations_list[lengths(segmentations_list) > 0])
# 
# ## Add the information
# segmentation_match <- match(spe$key, segmentations$key)
# segmentation_info <-
#   segmentations[segmentation_match, -which(
#     colnames(segmentations) %in% c("barcode", "tissue", "row", "col", "imagerow", "imagecol", "key")
#   )]
# colData(spe) <- cbind(colData(spe), segmentation_info)

## For visualizing later with spatialLIBD
spe$overlaps_tissue <-
  factor(ifelse(spe$in_tissue, "in", "out"))

## Save raw SPE. we will filter it down to append to the existing SPE.
spe_raw <- spe
save(spe_raw, file = here::here("processed-data", "02_build_spe", "spe_raw_V13Y24-346_C1_only.RData"))

## clean up
rm(spe,add_design,demos)
```


### drop empty spots but not any genes (we will subset to genes previously included in the spe)
### bind the new sample raw spe to the raw spe for the rest of the dataset (which we can't rebuild from scratch with JHPCE down)
```{r}
rm(list=ls())
#

load("processed-data/02_build_spe/spe_raw_V13Y24-346_C1_only.RData")
spev13y24 <- copy(spe_raw)


## Remove spots without counts
if (any(colSums(counts(spev13y24)) == 0)) {
  message("removing spots without counts for spe")
  spev13y24 <- spev13y24[, -which(colSums(counts(spev13y24)) == 0)]
  dim(spev13y24)
}
# no spots removed?

## Now drop the spots outside the tissue
spev13y24 <- spev13y24[, spev13y24$in_tissue]
dim(spev13y24)
## 36601 4062

## clean up in case spe/spe_raw gets replaced by loading the other raw SPE
rm(spe_raw)
gc(full=T)

## load the existing raw SPE, drop its countless and nontissue spots
load("processed-data/02_build_spe/spe_raw_w_V13M13-362.RData")
if (any(colSums(counts(spe_raw)) == 0)) {
  message("removing spots without counts for spe")
  spe_raw <- spe_raw[, -which(colSums(counts(spe_raw)) == 0)]
  dim(spe_raw)
}
# 30297 44830

spe_raw <- spe_raw[,spe_raw$in_tissue]
dim(spe_raw)
# 30297 41579

### check that the new sample doesn't contain any genes not in the other samples
detecnew <- rownames(spev13y24)[which(rowSums(counts(spev13y24)) > 0)]

stopifnot(all(detecnew %in% rownames(spe_raw)))
## not true

addnlgenes <- detecnew[!(detecnew %in% rownames(spe_raw))]


### make a mock counts matrix for the old spe containing the new genes
oldspe.tmpcount <- as.matrix(counts(spe_raw))[1:length(addnlgenes),]
rownames(oldspe.tmpcount) <- addnlgenes
oldspe.tmpcount <- as.data.table(oldspe.tmpcount,keep.rownames = T)
set(oldspe.tmpcount,j=c(2:ncol(oldspe.tmpcount)),value = 0)
oldspe.tmpcount <- as.data.frame(oldspe.tmpcount)
rownames(oldspe.tmpcount) <- oldspe.tmpcount$rn
oldspe.tmpcount$rn <- NULL

oldspe.tmpcount <- as.matrix(oldspe.tmpcount)
oldspe.tmpcount <- as(oldspe.tmpcount,"dgCMatrix")
oldspe.tmpcount <- oldspe.tmpcount[,colnames(spe_raw)]

oldspe.ct <- rbind(counts(spe_raw),oldspe.tmpcount)

## get the additional rows of rowdata for these genes to build out the rowdata as well

tmprd <- as.data.table(rowData(spev13y24),keep.rownames=T)[rn %in% addnlgenes]
oldrd <- as.data.table(rowData(spe_raw),keep.rownames=T)
allrd <- rbind(tmprd,oldrd)
allrd <- DataFrame(allrd,row.names=allrd$rn)
allrd$rn <- NULL
allrd <- allrd[rownames(oldspe.ct),]
## now, rejigger spev13y24 to have the prior spe's genes plus its 101 specific ones
spev13y24 <- spev13y24[rownames(oldspe.ct),]
dim(spev13y24)
# 30398 4062

### get the colData frames to be the same (we can drop a bunch of 10x garbage cols this way too)
keepcols <- grep(colnames(colData(spe_raw)),pattern="10x|mask",value=T,invert=T)
stopifnot(all(keepcols %in% colnames(colData(spev13y24))))

colData(spev13y24) <- colData(spev13y24)[,keepcols]
colData(spe_raw) <- colData(spe_raw)[,keepcols]

newcd <- rbind(colData(spe_raw),colData(spev13y24))

## make a DataFrame object containing the imgData from both assays in the same order we're concat'ing everything else
newimdat <- rbind(imgData(spe_raw),imgData(spev13y24))

# spat coords, whose rownames don't actually match the spe colnames to begin with so we can fix that here. not that the discrepancy has affected anything downstream before AFAIK.
oldspat <- as.data.frame(cbind(colnames(spe_raw),spatialCoords(spe_raw)))
oldspat <- DataFrame(oldspat,row.names=oldspat$V1)
oldspat$V1 <- NULL
oldspat <- oldspat[colnames(spe_raw),]

newspat <- as.data.frame(cbind(colnames(spev13y24),spatialCoords(spev13y24)))
newspat <- DataFrame(newspat,row.names=newspat$V1)
newspat$V1 <- NULL
newspat <- newspat[colnames(spev13y24),]

allspat <- rbind(oldspat,newspat)
allspat$pxl_col_in_fullres <- as.numeric(allspat$pxl_col_in_fullres)
allspat$pxl_row_in_fullres <- as.numeric(allspat$pxl_row_in_fullres)
allspat <- as.matrix(allspat)

# concatenate count matrices. we already ordered the new sample's rows as (the old spe + new-sample-unique genes) so this is ready to go
newct <- cbind(oldspe.ct,counts(spev13y24))

# sanity checks
stopifnot(nrow(allrd)==nrow(newct) & all(rownames(allrd)==rownames(newct)) & nrow(newcd)==ncol(newct) & all(rownames(newcd)==colnames(newct)) & nrow(allspat)==ncol(newct))
```

# build a new SPE from the reconstructed parts above
```{r}
spe_raw2 <- SpatialExperiment(
  assays = list(counts = cbind(oldspe.ct,counts(spev13y24))),
  rowData = allrd,
  colData = newcd,
  imgData = newimdat,
  spatialCoords = allspat)

dim(spe_raw2)
# 30398 45641
```

# append additional sample metadata (Sex, PMI, etc)
```{r}
demos <- fread("raw-data/demos.txt")
demos <- demos[BrNum %in% colData(spe_raw2)$brnum]
demos <- unique(demos)
setnames(demos,"BrNum","brnum")
tmp <- merge.data.table(demos,as.data.table(colData(spe_raw2)),by="brnum")
# k. also fix "Best Rin PFC" and "BMI (calculated)" to not have spaces
setnames(tmp,c("Best RIN PFC","BMI (calculated)"),c("best_rin_pfc","bmi"))

# ^ has no rownames because it's a data table; change to DataFrame and make rownames = key in the merged table. reorder to match the column order in spe_raw2.
tmp <- DataFrame(tmp,row.names = tmp$key)
tmp <- tmp[colnames(spe_raw2),]
stopifnot(sum(rownames(tmp)==colnames(spe_raw2))==ncol(spe_raw2))

### add the demographic'd and cleaned colData to the SPE
colData(spe_raw2) <- tmp

rm(tmp,demos)
```

## save the new spe (which isn't QUITE raw--zero count genes and out of tissue spots are gone.)
```{r}
save(spe_raw2,file="processed-data/02_build_spe/spe_w_v13y24-346_Jul24.RData")
```

## Reproducibility information
```{r}
Sys.time()
proc.time()
options(width = 120)
session_info()
```
> Sys.time()
[1] "2024-07-16 13:55:10 CDT"
> proc.time()
    user   system  elapsed 
 154.333    6.989 1237.654 
> options(width = 120)
> session_info()
─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-07-16
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
 ! package              * version   date (UTC) lib source
   abind                  1.4-5     2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi          1.66.0    2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub          3.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   attempt                0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
   beachmat               2.20.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm               0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme            1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData        1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
   Biobase              * 2.64.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocFileCache          2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics         * 0.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager            1.30.23   2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors          1.22.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel         * 1.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocSingular           1.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion            3.19.1    2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings           * 2.72.1    2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                    4.0.5     2022-11-15 [1] CRAN (R 4.4.0)
   bit64                  4.0.5     2020-08-30 [1] CRAN (R 4.4.0)
   bitops                 1.0-7     2021-04-24 [1] CRAN (R 4.4.0)
   blob                   1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   bslib                  0.7.0     2024-03-29 [1] CRAN (R 4.4.0)
   cachem                 1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
 P cli                    3.6.3     2024-06-21 [2] CRAN (R 4.4.0)
   codetools              0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorout             * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace             2.1-0     2023-01-23 [1] CRAN (R 4.4.0)
   config                 0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
   crayon                 1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   curl                   5.2.1     2024-03-01 [1] CRAN (R 4.4.0)
   data.table           * 1.15.4    2024-03-30 [2] CRAN (R 4.4.0)
   DBI                    1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                 2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray           0.30.1    2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats     1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   digest                 0.6.36    2024-06-23 [1] CRAN (R 4.4.0)
   doParallel             1.0.17    2022-02-07 [1] CRAN (R 4.4.0)
   dotCall64              1.1-1     2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                  1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   DT                     0.33      2024-04-04 [1] CRAN (R 4.4.0)
   edgeR                  4.2.0     2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate               0.24.0    2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub          2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fansi                  1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   fastmap                1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fields                 16.2      2024-06-27 [1] CRAN (R 4.4.0)
   filelock               1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   foreach                1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
   generics               0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb         * 1.40.1    2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData       1.2.12    2024-05-01 [1] Bioconductor
   GenomicAlignments      1.40.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges        * 1.56.1    2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm             0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2                3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrepel                0.9.5     2024-01-10 [1] CRAN (R 4.4.0)
   glue                   1.7.0     2024-01-09 [1] CRAN (R 4.4.0)
   golem                  0.4.1     2023-06-05 [1] CRAN (R 4.4.0)
   gridExtra              2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gtable                 0.3.5     2024-04-22 [1] CRAN (R 4.4.0)
   here                 * 1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools              0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets            1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                 1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
   httr                   1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   IRanges              * 2.38.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                  2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   iterators              1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
   jquerylib              0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite               1.8.8     2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST               1.44.1    2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   knitr                  1.48      2024-07-07 [1] CRAN (R 4.4.1)
   later                  1.3.2     2023-12-06 [1] CRAN (R 4.4.0)
   lattice                0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval               0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle              1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                  3.60.3    2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                 1.5-9.10  2024-06-24 [1] CRAN (R 4.4.0)
   magick                 2.8.3     2024-02-18 [1] CRAN (R 4.4.0)
   magrittr               2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   maps                   3.4.2     2023-12-15 [1] CRAN (R 4.4.0)
   Matrix                 1.7-0     2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics       * 1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats          * 1.3.0     2024-04-11 [1] CRAN (R 4.4.0)
   memoise                2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   mime                   0.12      2021-09-28 [1] CRAN (R 4.4.0)
   munsell                0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   paletteer              1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
   parallelly           * 1.37.1    2024-02-29 [1] CRAN (R 4.4.0)
   pillar                 1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   plotly                 4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
   png                    0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   promises               1.3.0     2024-04-05 [1] CRAN (R 4.4.0)
   purrr                  1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R6                     2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs               0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                   1.0.12    2024-01-09 [1] CRAN (R 4.4.0)
   RCurl                  1.98-1.14 2024-01-09 [1] CRAN (R 4.4.0)
   rematch2               2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
   restfulr               0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
   rjson                  0.2.21    2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown              2.27      2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot              2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools              2.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                2.3.7     2024-05-27 [1] CRAN (R 4.4.0)
   rstudioapi             0.16.0    2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                   1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer          * 1.64.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   S4Arrays               1.4.1     2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors            * 0.42.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                   0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix           1.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                 1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater                 1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scuttle                1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo          * 1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   shiny                  1.8.1.1   2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets           0.8.6     2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment * 1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spam                   2.10-0    2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray            1.4.8     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats      1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment    * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD          * 1.16.2    2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   statmod                1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   SummarizedExperiment * 1.34.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   tibble                 3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                  1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect             1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   UCSC.utils             1.0.0     2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   utf8                   1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                  0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                  0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite            0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   xfun                   0.45      2024-06-16 [1] CRAN (R 4.4.0)
   XML                    3.99-0.17 2024-06-25 [1] CRAN (R 4.3.3)
   xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
   XVector              * 0.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                   2.3.9     2024-07-05 [1] CRAN (R 4.4.0)
   zlibbioc               1.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
