---
title: "SF-Xen sex DE schematic"
author: "Bernie Mulvey"
date: "2024-11-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10, fig.width = 7, include = FALSE)
knitr::opts_chunk$set(fig.width = 7, fig.height = 10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table) # Preferred data manipulation package
library(ggplot2) # Dependency for several plotting functions
library(ggtext) # more character types in ggplots
library(ggrastr) # avoid making raster images one can't even open
library(SpatialExperiment) # Visium data framework
library(SpatialFeatureExperiment) # Xenium data framework
library(escheR) # alternative spotplotting function, at least for visium
library(sf) # define the polygonal boundaries of xenium domains

require(colorout) # Utility for RStudio


## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 9), axis.title.x = element_text(size = 8), axis.text.y = element_text(size = 8), axis.title.y = element_text(size =9), plot.title = element_markdown(size = 11,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```


#### XENIUM setup ####
```{r}
hypx <- readRDS("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-sfe-in_staggered-coords_genetarg-only_NONlog-norm.RDS")

bksmooth <- fread("xenium_HYP/processed-data/07_Domains-subdomains_by_knnSmoothing_Banksytypes/03b-ARCVMHdomains_2stepsmooth_VMH-k10-0.2-VMH-k200-0.2_ARC-k50-0.1_ARC-k500-0.5.txt") ## xenium domain assignments after smoothing, by xenium cell

## but drop discarded cell clusters, too
bkcl <- fread("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-BanksyClusts_M0lam0_leiden_multisamp.txt")
setnames(bkcl,2,"cl")

bkanno <- fread("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/02-M0l0kg6_topClusMarkers_celltypes_annotated.txt")

## drop sample-specific clusters, append cluster annotations 
bkanno <- unique(bkanno[,.(clus,bjm_annot)])
bkanno <- bkanno[bjm_annot!="DISCARD"&bjm_annot!="VMH_4_DISCARD"]

bkcl <- merge.data.table(bkcl,bkanno,by.x="cl",by.y="clus")
## drop excluded clusters from xenium spe
hypx <- hypx[,colnames(hypx) %in% bkcl$rn]

## append cluster annots to xenium spe
bkcl <- DataFrame(bkcl,row.names=bkcl$rn)[colnames(hypx),]
bkcl$rn <- NULL
hypx$banksyclus <- bkcl$bjm_annot

## assign the domain labels to the xenium cells
bksmooth[dualVMHARC4=="other",dualVMHARC4:="Other"]
bksmooth <- DataFrame(bksmooth,row.names=bksmooth$rn)[colnames(hypx),]
hypx$dom <- bksmooth$dualVMHARC4
```

### ## create supercluster labels to group related cell types together and use the previously saved + loaded xenium palette for cell groups for the next panel.
```{r}
bkcl <- as.data.table(bkcl,keep.rownames=T)
bkcl[,supercl:="ifthisisplottedigoofed"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="VMH",value=T),supercl:="VMH (4)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="ARC",value=T),supercl:="ARC (5)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="Tanyc|Alar",value=T),supercl:="Tanycytes,<br>Portal Vasc. (4)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="Oligo",value=T),supercl:="Oligo (4)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="Astro",value=T),supercl:="Astro (2)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="Microg",value=T),supercl:="Microglia (3)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="Supraopt",value=T),supercl:="SON (2)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="Unsure_AVP",value=T),supercl:="Non-SON<br>*AVP*+*OXT*+ (1)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="Macrop|Periph|Vascul|Endothel",value=T),supercl:="Vascular, Periph.<br>Immune (5)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="GABA",value=T),supercl:="Other GABAergic (2)"]
bkcl[bjm_annot %in% grep(bjm_annot,pattern="Periven",value=T),supercl:="PeriVN<br>(Excitatory) (1)"]
### the parenthetical numbers add up to 33, so we're good.


### lastly, in the case of the xenium spotplot of all cell groups, we need to put them in a specific order for them to fit within the plot space alongside 1225B: VMH, ARC, SON, Astro, Oligo, Microglia, PeriVN, Non-SON AVP+OXT+, Tanycytes n portal, Other GABA, Vasc n periph
levs <- c("VMH (4)","ARC (5)","SON (2)","Astro (2)","Oligo (4)","Microglia (3)","PeriVN<br>(Excitatory) (1)","Non-SON<br>*AVP*+*OXT*+ (1)","Tanycytes,<br>Portal Vasc. (4)","Other GABAergic (2)","Vascular, Periph.<br>Immune (5)")
stopifnot(all(levs %in% unique(bkcl$supercl)))

bkcl$supercl <- factor(bkcl$supercl,levels=c("VMH (4)","ARC (5)","SON (2)","Astro (2)","Oligo (4)","Microglia (3)","PeriVN<br>(Excitatory) (1)","Non-SON<br>*AVP*+*OXT*+ (1)","Tanycytes,<br>Portal Vasc. (4)","Other GABAergic (2)","Vascular, Periph.<br>Immune (5)"))
```


The colors slotted in 1 and 3 are also picked to have the same luminance so that gene expression coloration fill in escheR grayscale doesn't look falsely darker or lighter in one spot type vs. the other 
```{r}
pals <- readRDS("manuscript_plot_code/domain_and_xencluster_palettes_CHECKPALNAMESBEFOREUS.RDS")

pal <- pals[[1]][1:2]

mscriptids <- fread("standardized_sampleids_for_plotting.txt")
```

```{r}
pane <- hypx[,hypx$sample_id=="X99_1225B"]
# 
## attach the coordinate info to colData so we can make an sf object out of this for hull extraction. see well-annotated code from DE benchmarking of domain boundaries in xenium_HYP/code/07/03c-DE benchmarks...
tmpcoldata <- cbind(colData(pane),spatialCoords(pane))
panjv <- st_as_sf(as.data.frame(tmpcoldata[tmpcoldata$dom=="VMH",]),coords = c("sdimx","sdimy"))
panjv <- st_convex_hull(st_union(panjv))

panja <- st_as_sf(as.data.frame(tmpcoldata[tmpcoldata$dom=="ARC",]),coords=c("sdimx","sdimy"))
panja <- st_concave_hull(st_union(panja),ratio = 0.1)

## create the data table we will use for overlaying the boundary polygons
dompoly <- cbind(panjv[[1]][[1]],rep("VMH",nrow(panjv[[1]][[1]])))
dompoly <- rbind(dompoly,cbind(panja[[1]][[1]],rep("ARC",nrow(panja[[1]][[1]]))))
dompoly <- as.data.table(dompoly)
setnames(dompoly,c("xpol","ypol","dompol"))
dompoly[,xpol:=as.numeric(xpol)]
dompoly[,ypol:=as.numeric(ypol)]

rm(panja,panjv)
```

xenium "spot"plot with superclusters labeled. and XL boundaries on the two domains. then additional plots of just those domains. fix the palette names to match the tweaked names above (done to keep legend from taking up too much width)
```{r}
## make palette names match tweaks from previous block
xenpal <- pals$xen_cellgroups
names(xenpal)[names(xenpal)=="PeriVN (Excitatory) (1)"] <- "PeriVN<br>(Excitatory) (1)"
names(xenpal)[names(xenpal)=="Non-SON AVP+OXT+ (1)"] <- "Non-SON<br>*AVP*+*OXT*+ (1)"
names(xenpal)[names(xenpal)=="Tanycytes, Portal Vasc. (4)"] <- "Tanycytes,<br>Portal Vasc. (4)"
names(xenpal)[names(xenpal)=="Vascular and Peripheral Immune (5)"] <- "Vascular, Periph.<br>Immune (5)"
names(xenpal)[names(xenpal)=="Other GABAergic Neurons (2)"] <- "Other GABAergic (2)"

stopifnot(all(names(xenpal) %in% unique(bkcl$supercl)))

## assign superclusters to SPE
hypx$supercl <- DataFrame(bkcl,row.names=bkcl$rn)[colnames(hypx),]$supercl
```

plot whole sample
```{r}
# get sample
panealt <- hypx[,hypx$sample_id=="X99_1225B"]

pane <- make_escheR(panealt,y_reverse=FALSE)
pane <- pane |> add_fill("supercl",size=0.3,point_size = 0.3)

## escheR doesn't seem to pass on the alpha argument in add_fill, so manually append it to the pane object if desired
pane$layers[[1]][["aes_params"]][length(pane$layers[[1]][["aes_params"]])+1] <- 0.25
names(pane[["layers"]][[1]][["aes_params"]])[length(pane[["layers"]][[1]][["aes_params"]])] <- "alpha"

pdf("manuscript_plots/Supplemental/SF-Xen sex DE analysis schematic/pt1.pdf",height=2.4,width=2.4)
pane+
    guides(fill="none")+
    scale_fill_manual(values=xenpal)+
    theme(panel.grid = element_blank(), axis.text.x = element_blank(), axis.text.y = element_blank(), axis.title.x = element_blank(), axis.title.y = element_blank(), axis.ticks = element_blank()) +
    geom_path(
        data = dompoly,
        aes(
            x = xpol,
            y = ypol,
            group = dompol,
            col = dompol
        ),
        linewidth = 1
    ) +
    scale_color_manual(values = pal, na.value = NA) + guides(color = "none")
dev.off()


```

```{r}
panealt <- hypx[,hypx$sample_id=="X99_1225B"&hypx$dom=="VMH"]

pane <- make_escheR(panealt,y_reverse=FALSE)
pane <- pane |> add_fill("supercl",size=0.15,point_size = 0.15)

## escheR doesn't seem to pass on the alpha argument in add_fill, so manually append it to the pane object if desired
pane$layers[[1]][["aes_params"]][length(pane$layers[[1]][["aes_params"]])+1] <- 0.25
names(pane[["layers"]][[1]][["aes_params"]])[length(pane[["layers"]][[1]][["aes_params"]])] <- "alpha"

pdf("manuscript_plots/Supplemental/SF-Xen sex DE analysis schematic/schematic pt2.pdf" ,height=1.1,width=1.1)
pane+
    guides(fill="none")+
    scale_fill_manual(values=xenpal)+
    theme(panel.grid = element_blank(), axis.text.x = element_blank(), axis.text.y = element_blank(), axis.title.x = element_blank(), axis.title.y = element_blank(), axis.ticks = element_blank()) +
    geom_path(
        data = dompoly[dompol=="VMH"],
        aes(
            x = xpol,
            y = ypol,
            group = dompol,
            col = dompol
        ),
        linewidth = 0.5
    ) +
    scale_color_manual(values = pal, na.value = NA) + guides(color = "none")
dev.off()

#### ARC
panealt <- hypx[,hypx$sample_id=="X99_1225B"&hypx$dom=="ARC"]

pane <- make_escheR(panealt,y_reverse=FALSE)
pane <- pane |> add_fill("supercl",size=0.15,point_size = 0.15)

## escheR doesn't seem to pass on the alpha argument in add_fill, so manually append it to the pane object if desired
pane$layers[[1]][["aes_params"]][length(pane$layers[[1]][["aes_params"]])+1] <- 0.25
names(pane[["layers"]][[1]][["aes_params"]])[length(pane[["layers"]][[1]][["aes_params"]])] <- "alpha"

pdf("manuscript_plots/Supplemental/SF-Xen sex DE analysis schematic/pt3.pdf",height=1.1,width=1.1)
pane+
    guides(fill="none")+
    scale_fill_manual(values=xenpal)+
    theme(panel.grid = element_blank(), axis.text.x = element_blank(), axis.text.y = element_blank(), axis.title.x = element_blank(), axis.title.y = element_blank(), axis.ticks = element_blank()) +
    geom_path(
        data = dompoly[dompol=="ARC"],
        aes(
            x = xpol,
            y = ypol,
            group = dompol,
            col = dompol
        ),
        linewidth = 0.5
    ) +
    scale_color_manual(values = pal, na.value = NA) + guides(color = "none")
dev.off()
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] sf_1.0-16                      colorout_1.3-0.2              
 [3] Polychrome_1.5.1               viridis_0.6.5                 
 [5] viridisLite_0.4.2              escheR_1.4.0                  
 [7] spatialLIBD_1.16.2             SpatialFeatureExperiment_1.6.1
 [9] SpatialExperiment_1.14.0       SingleCellExperiment_1.26.0   
[11] SummarizedExperiment_1.34.0    Biobase_2.64.0                
[13] GenomicRanges_1.56.1           GenomeInfoDb_1.40.1           
[15] IRanges_2.38.1                 S4Vectors_0.42.1              
[17] BiocGenerics_0.50.0            MatrixGenerics_1.16.0         
[19] matrixStats_1.3.0              ggrastr_1.0.2                 
[21] ggtext_0.1.2                   ggplot2_3.5.1                 
[23] data.table_1.15.4              rlang_1.1.4                   

loaded via a namespace (and not attached):
  [1] later_1.3.2               BiocIO_1.14.0             fields_16.2              
  [4] bitops_1.0-7              filelock_1.0.3            tibble_3.2.1             
  [7] R.oo_1.26.0               XML_3.99-0.17             lifecycle_1.0.4          
 [10] edgeR_4.2.0               doParallel_1.0.17         rprojroot_2.0.4          
 [13] lattice_0.22-6            magrittr_2.0.3            sass_0.4.9               
 [16] limma_3.60.3              plotly_4.10.4             rmarkdown_2.27           
 [19] jquerylib_0.1.4           yaml_2.3.9                httpuv_1.6.15            
 [22] spam_2.10-0               sessioninfo_1.2.2         sp_2.1-4                 
 [25] cowplot_1.1.3             DBI_1.2.3                 RColorBrewer_1.1-3       
 [28] golem_0.4.1               maps_3.4.2                abind_1.4-5              
 [31] zlibbioc_1.50.0           purrr_1.0.2               R.utils_2.12.3           
 [34] RCurl_1.98-1.14           rappdirs_0.3.3            GenomeInfoDbData_1.2.12  
 [37] ggrepel_0.9.5             irlba_2.3.5.1             terra_1.7-78             
 [40] units_0.8-5               dqrng_0.4.1               commonmark_1.9.1         
 [43] DelayedMatrixStats_1.26.0 codetools_0.2-20          DropletUtils_1.24.0      
 [46] DelayedArray_0.30.1       DT_0.33                   scuttle_1.14.0           
 [49] xml2_1.3.6                tidyselect_1.2.1          farver_2.1.2             
 [52] UCSC.utils_1.0.0          ScaledMatrix_1.12.0       shinyWidgets_0.8.6       
 [55] BiocFileCache_2.12.0      GenomicAlignments_1.40.0  jsonlite_1.8.8           
 [58] BiocNeighbors_1.22.0      e1071_1.7-14              scater_1.32.0            
 [61] iterators_1.0.14          foreach_1.5.2             tools_4.4.1              
 [64] Rcpp_1.0.13               glue_1.7.0                gridExtra_2.3            
 [67] SparseArray_1.4.8         xfun_0.45                 here_1.0.1               
 [70] EBImage_4.46.0            dplyr_1.1.4               HDF5Array_1.32.0         
 [73] withr_3.0.0               BiocManager_1.30.23       fastmap_1.2.0            
 [76] boot_1.3-30               rhdf5filters_1.16.0       fansi_1.0.6              
 [79] spData_2.3.1              rsvd_1.0.5                digest_0.6.36            
 [82] R6_2.5.1                  mime_0.12                 colorspace_2.1-0         
 [85] wk_0.9.2                  markdown_1.13             jpeg_0.1-10              
 [88] RSQLite_2.3.7             R.methodsS3_1.8.2         config_0.3.2             
 [91] utf8_1.2.4                tidyr_1.3.1               generics_0.1.3           
 [94] rtracklayer_1.64.0        class_7.3-22              httr_1.4.7               
 [97] htmlwidgets_1.6.4         S4Arrays_1.4.1            scatterplot3d_0.3-44     
[100] spdep_1.3-5               pkgconfig_2.0.3           gtable_0.3.5             
[103] blob_1.2.4                XVector_0.44.0            htmltools_0.5.8.1        
[106] dotCall64_1.1-1           fftwtools_0.9-11          scales_1.3.0             
[109] attempt_0.3.1             png_0.1-8                 knitr_1.48               
[112] rstudioapi_0.16.0         rjson_0.2.21              curl_5.2.1               
[115] proxy_0.4-27              cachem_1.1.0              rhdf5_2.48.0             
[118] stringr_1.5.1             BiocVersion_3.19.1        KernSmooth_2.23-24       
[121] parallel_4.4.1            vipor_0.4.7               AnnotationDbi_1.66.0     
[124] restfulr_0.0.15           s2_1.1.6                  pillar_1.9.0             
[127] grid_4.4.1                vctrs_0.6.5               promises_1.3.0           
[130] BiocSingular_1.20.0       dbplyr_2.5.0              beachmat_2.20.0          
[133] xtable_1.8-4              sfheaders_0.4.4           beeswarm_0.4.0           
[136] paletteer_1.6.0           evaluate_0.24.0           zeallot_0.1.0            
[139] magick_2.8.3              Rsamtools_2.20.0          cli_3.6.3                
[142] locfit_1.5-9.10           compiler_4.4.1            crayon_1.5.3             
[145] labeling_0.4.3            classInt_0.4-10           rematch2_2.1.2           
[148] ggbeeswarm_0.7.2          stringi_1.8.4             deldir_2.0-4             
[151] BiocParallel_1.38.0       munsell_0.5.1             Biostrings_2.72.1        
[154] lazyeval_0.2.2            tiff_0.1-12               Matrix_1.7-0             
[157] ExperimentHub_2.12.0      benchmarkme_1.0.8         sparseMatrixStats_1.16.0 
[160] bit64_4.0.5               Rhdf5lib_1.26.0           KEGGREST_1.44.1          
[163] statmod_1.5.0             shiny_1.8.1.1             AnnotationHub_3.12.0     
[166] gridtext_0.1.5            memoise_2.0.1             bslib_0.7.0              
[169] benchmarkmeData_1.0.4     bit_4.0.5                
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2024-11-20
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────────────
 ! package                  * version   date (UTC) lib source
   abind                      1.4-5     2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi              1.66.0    2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub              3.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   attempt                    0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
   beachmat                   2.20.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm                   0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme                1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData            1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
   Biobase                  * 2.64.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics             * 0.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                     1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager                1.30.23   2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors              1.22.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel               1.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocSingular               1.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion                3.19.1    2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings                 2.72.1    2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                        4.0.5     2022-11-15 [1] CRAN (R 4.4.0)
   bit64                      4.0.5     2020-08-30 [1] CRAN (R 4.4.0)
   bitops                     1.0-7     2021-04-24 [1] CRAN (R 4.4.0)
   blob                       1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   boot                       1.3-30    2024-02-26 [1] CRAN (R 4.4.1)
   bslib                      0.7.0     2024-03-29 [1] CRAN (R 4.4.0)
   cachem                     1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
   class                      7.3-22    2023-05-03 [1] CRAN (R 4.4.1)
   classInt                   0.4-10    2023-09-05 [1] CRAN (R 4.4.0)
 P cli                        3.6.3     2024-06-21 [2] CRAN (R 4.4.0)
   codetools                  0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorout                 * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace                 2.1-0     2023-01-23 [1] CRAN (R 4.4.0)
   commonmark                 1.9.1     2024-01-30 [1] CRAN (R 4.4.0)
   config                     0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                    1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
   crayon                     1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   curl                       5.2.1     2024-03-01 [1] CRAN (R 4.4.0)
   data.table               * 1.15.4    2024-03-30 [2] CRAN (R 4.4.0)
   DBI                        1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                     2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray               0.30.1    2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats         1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   deldir                     2.0-4     2024-02-28 [1] CRAN (R 4.4.0)
   digest                     0.6.36    2024-06-23 [1] CRAN (R 4.4.0)
   doParallel                 1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64                  1.1-1     2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                      1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   dqrng                      0.4.1     2024-05-28 [1] CRAN (R 4.4.0)
   DropletUtils               1.24.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   DT                         0.33      2024-04-04 [1] CRAN (R 4.4.0)
   e1071                      1.7-14    2023-12-06 [1] CRAN (R 4.4.0)
   EBImage                    4.46.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   edgeR                      4.2.0     2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   escheR                   * 1.4.0     2024-05-16 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate                   0.24.0    2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fansi                      1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   farver                     2.1.2     2024-05-13 [1] CRAN (R 4.4.0)
   fastmap                    1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fftwtools                  0.9-11    2021-03-01 [1] CRAN (R 4.4.0)
   fields                     16.2      2024-06-27 [1] CRAN (R 4.4.0)
   filelock                   1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   foreach                    1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
   generics                   0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb             * 1.40.1    2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData           1.2.12    2024-05-01 [1] Bioconductor
   GenomicAlignments          1.40.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges            * 1.56.1    2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm                 0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2                  * 3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrastr                  * 1.0.2     2023-06-01 [1] CRAN (R 4.4.0)
   ggrepel                    0.9.5     2024-01-10 [1] CRAN (R 4.4.0)
   ggtext                   * 0.1.2     2022-09-16 [1] CRAN (R 4.4.0)
   glue                       1.7.0     2024-01-09 [1] CRAN (R 4.4.0)
   golem                      0.4.1     2023-06-05 [1] CRAN (R 4.4.0)
   gridExtra                  2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gridtext                   0.1.5     2022-09-16 [1] CRAN (R 4.4.0)
   gtable                     0.3.5     2024-04-22 [1] CRAN (R 4.4.0)
   HDF5Array                  1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   here                       1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools                  0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets                1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                     1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
   httr                       1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   IRanges                  * 2.38.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                      2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   iterators                  1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
   jpeg                       0.1-10    2022-11-29 [1] CRAN (R 4.4.0)
   jquerylib                  0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite                   1.8.8     2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST                   1.44.1    2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   KernSmooth                 2.23-24   2024-05-17 [1] CRAN (R 4.4.1)
   knitr                      1.48      2024-07-07 [1] CRAN (R 4.4.1)
   labeling                   0.4.3     2023-08-29 [1] CRAN (R 4.4.0)
   later                      1.3.2     2023-12-06 [1] CRAN (R 4.4.0)
   lattice                    0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval                   0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle                  1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                      3.60.3    2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                     1.5-9.10  2024-06-24 [1] CRAN (R 4.4.0)
   magick                     2.8.3     2024-02-18 [1] CRAN (R 4.4.0)
   magrittr                   2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   maps                       3.4.2     2023-12-15 [1] CRAN (R 4.4.0)
   markdown                   1.13      2024-06-04 [1] CRAN (R 4.4.0)
   Matrix                     1.7-0     2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics           * 1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats              * 1.3.0     2024-04-11 [1] CRAN (R 4.4.0)
   memoise                    2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   mime                       0.12      2021-09-28 [1] CRAN (R 4.4.0)
   munsell                    0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   paletteer                  1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
   pillar                     1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig                  2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   plotly                     4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
   png                        0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   Polychrome               * 1.5.1     2022-05-03 [1] CRAN (R 4.4.0)
   promises                   1.3.0     2024-04-05 [1] CRAN (R 4.4.0)
   proxy                      0.4-27    2022-06-09 [1] CRAN (R 4.4.0)
   purrr                      1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R.methodsS3                1.8.2     2022-06-13 [1] CRAN (R 4.4.0)
   R.oo                       1.26.0    2024-01-24 [1] CRAN (R 4.4.0)
   R.utils                    2.12.3    2023-11-18 [1] CRAN (R 4.4.0)
   R6                         2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs                   0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer               1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                       1.0.13    2024-07-17 [1] CRAN (R 4.4.0)
   RCurl                      1.98-1.14 2024-01-09 [1] CRAN (R 4.4.0)
   rematch2                   2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
   restfulr                   0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
   rhdf5                      2.48.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rhdf5filters               1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Rhdf5lib                   1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rjson                      0.2.21    2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                    * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown                  2.27      2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot                  2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools                  2.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                    2.3.7     2024-05-27 [1] CRAN (R 4.4.0)
   rstudioapi                 0.16.0    2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                       1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer                1.64.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   s2                         1.1.6     2023-12-19 [1] CRAN (R 4.4.0)
   S4Arrays                   1.4.1     2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors                * 0.42.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                       0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix               1.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                     1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater                     1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scatterplot3d              0.3-44    2023-05-05 [1] CRAN (R 4.4.0)
   scuttle                    1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo                1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   sf                       * 1.0-16    2024-03-24 [1] CRAN (R 4.4.0)
   sfheaders                  0.4.4     2024-01-17 [1] CRAN (R 4.4.0)
   shiny                      1.8.1.1   2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets               0.8.6     2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment     * 1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sp                         2.1-4     2024-04-30 [1] CRAN (R 4.4.0)
   spam                       2.10-0    2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray                1.4.8     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats          1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment        * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialFeatureExperiment * 1.6.1     2024-05-15 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD              * 1.16.2    2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   spData                     2.3.1     2024-05-31 [1] CRAN (R 4.4.0)
   spdep                      1.3-5     2024-06-10 [1] CRAN (R 4.4.0)
   statmod                    1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   stringi                    1.8.4     2024-05-06 [1] CRAN (R 4.4.0)
   stringr                    1.5.1     2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment     * 1.34.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   terra                      1.7-78    2024-05-22 [1] CRAN (R 4.4.0)
   tibble                     3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                      1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect                 1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   tiff                       0.1-12    2023-11-28 [1] CRAN (R 4.4.0)
   UCSC.utils                 1.0.0     2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   units                      0.8-5     2023-11-28 [1] CRAN (R 4.4.0)
   utf8                       1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                      0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                      0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                  * 0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite              * 0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   withr                      3.0.0     2024-01-16 [1] CRAN (R 4.4.0)
   wk                         0.9.2     2024-07-09 [1] CRAN (R 4.4.0)
   xfun                       0.45      2024-06-16 [1] CRAN (R 4.4.0)
   XML                        3.99-0.17 2024-06-25 [1] CRAN (R 4.3.3)
   xml2                       1.3.6     2023-12-04 [1] CRAN (R 4.4.0)
   xtable                     1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
   XVector                    0.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                       2.3.9     2024-07-05 [1] CRAN (R 4.4.0)
   zeallot                    0.1.0     2018-01-28 [1] CRAN (R 4.4.0)
   zlibbioc                   1.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────