---
title: "ST-Xenium_metadata_sheet"
author: "Bernie Mulvey"
date: "2024-12-06"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(SpatialFeatureExperiment)
library(spatialLIBD)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

### Description of the supplemental table is as follows: 
"per-sample readout statistics from the Xeniumranger 1.7-resegmented results. Mean, median, and standard deviation are presented for i) total counts per cell (including non-targeting/control probes), ii. total transcript counts per cell, and iii. number of cells. Each value is given for the Xenium data i. as initially imported into R (only the default xeniumranger filter of transcripts with a “quality score” of ≥ 20 having been applied), in the initially clustered data (with 41 clusters), and in the final data (only cells in the 33 non-sample-specific clusters)."

## read initial unfiltered SPE and tabulate things
```{r}
xeninit <- readRDS("xenium_HYP/processed-data/03_make_SPE-SFE/01_sfe_raw.RDS")

xenfilt1 <- readRDS("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-sfe-in_staggered-coords_genetarg-only_NONlog-norm.RDS")
```


Filter to the final (ie analyzed), 33-cluster xenium dataset:
```{r}
xenfin <- xenfilt1

bscl <- fread("Xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-BanksyClusts_M0lam0_leiden_multisamp.txt")
setnames(bscl,2,"cl")

annot <- fread("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/02-M0l0kg6_topClusMarkers_celltypes_annotated.txt")
annot <- unique(annot[,.(clus,bjm_annot)])
annot <- annot[bjm_annot %in% grep(bjm_annot,pattern="DISCARD",value=T,invert=T)]
stopifnot(nrow(annot)==33)

bscl <- merge.data.table(bscl,annot,by.x="cl",by.y="clus")
stopifnot(nrow(bscl)<ncol(xenfilt1)|sum(bscl$rn %in% colnames(xenfilt1))<800000) # should be 819k or so

xenfin <- xenfilt1[,colnames(xenfilt1) %in% bscl$rn]
stopifnot(ncol(xenfin)==nrow(bscl))
```

Divy up each of these into a list by sample id
```{r}
sids <- unique(xeninit$sample_id)

rawsamps <- lapply(sids,function(x){xeninit[,xeninit$sample_id==x]})
filtsamps <- lapply(sids,function(x){xenfilt1[,xenfilt1$sample_id==x]})
finsamps <- lapply(sids,function(x){xenfin[,xenfin$sample_id==x]})

names(rawsamps) <- sids
names(filtsamps) <- sids
names(finsamps) <- sids

### clear out RAM by removing the undivied-up spes
rm(xenfin,xeninit,xenfilt1)
gc(full=T)
```

iterate through each sample at each stage and get
mean, median, SD total counts
mean median sd gene counts
number of cells

```{r}
sampres <- mapply(raw=rawsamps,f=filtsamps,fin=finsamps,SIMPLIFY=FALSE,FUN=function(raw,f,fin){
    rawdt <- as.data.table(colData(raw))
    fdt <- as.data.table(colData(f))
    findt <- as.data.table(colData(fin))
    
    res <- as.data.frame(matrix(ncol=22,nrow=1))
    colnames(res) <- c("samp",
                       "raw_mean_tot",
                       "raw_med_tot",
                       "raw_sd_tot",
                       "raw_mean_gene",
                       "raw_med_gene",
                       "raw_sd_gene",
                       "raw_ncells",
                       
                       "filt_mean_tot",
                       "filt_med_tot",
                       "filt_sd_tot",
                       "filt_mean_gene",
                       "filt_med_gene",
                       "filt_sd_gene",
                       "filt_ncells",
                       
                       "fin_mean_tot",
                       "fin_med_tot",
                       "fin_sd_tot",
                       "fin_mean_gene",
                       "fin_med_gene",
                       "fin_sd_gene",
                       "fin_ncells")
    
    res$samp <- rawdt$sample_id[1]
    
    res$raw_mean_tot <- mean(rawdt$total_counts,na.rm=T)
    res$raw_med_tot <- median(rawdt$total_counts,na.rm=T)
    res$raw_sd_tot <- sd(rawdt$total_counts,na.rm=T)
    res$raw_mean_gene <- mean(rawdt$transcript_counts,na.rm=T)
    res$raw_med_gene <- median(rawdt$transcript_counts,na.rm=T)
    res$raw_sd_gene <- sd(rawdt$transcript_counts,na.rm=T)
    res$raw_ncells <- nrow(rawdt)
    
    res$filt_mean_tot <- mean(fdt$total_counts,na.rm=T)
    res$filt_med_tot <- median(fdt$total_counts,na.rm=T)
    res$filt_sd_tot <- sd(fdt$total_counts,na.rm=T)
    res$filt_mean_gene <- mean(fdt$transcript_counts,na.rm=T)
    res$filt_med_gene <- median(fdt$transcript_counts,na.rm=T)
    res$filt_sd_gene <- sd(fdt$transcript_counts,na.rm=T)
    res$filt_ncells <- nrow(fdt)
    
    res$fin_mean_tot <- mean(findt$total_counts,na.rm=T)
    res$fin_med_tot <- median(findt$total_counts,na.rm=T)
    res$fin_sd_tot <- sd(findt$total_counts,na.rm=T)
    res$fin_mean_gene <- mean(findt$transcript_counts,na.rm=T)
    res$fin_med_gene <- median(findt$transcript_counts,na.rm=T)
    res$fin_sd_gene <- sd(findt$transcript_counts,na.rm=T)
    res$fin_ncells <- nrow(findt)
    
    res <- as.data.table(res)
    return(res)
})

restab <- rbindlist(sampres)

fwrite(restab,"manuscript_plots/Supplemental/ST-Xenium_samplemetadata.txt",sep='\t',quote=F)
```

reprod inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.38.0            parallelly_1.38.0             
 [3] colorout_1.3-0.2               spatialLIBD_1.16.2            
 [5] SpatialFeatureExperiment_1.6.1 SpatialExperiment_1.14.0      
 [7] SingleCellExperiment_1.26.0    SummarizedExperiment_1.34.0   
 [9] Biobase_2.64.0                 GenomicRanges_1.56.1          
[11] GenomeInfoDb_1.40.1            IRanges_2.38.1                
[13] S4Vectors_0.42.1               BiocGenerics_0.50.0           
[15] MatrixGenerics_1.16.0          matrixStats_1.3.0             
[17] data.table_1.15.4              rlang_1.1.4                   

loaded via a namespace (and not attached):
  [1] later_1.3.2               BiocIO_1.14.0             bitops_1.0-7             
  [4] filelock_1.0.3            fields_16.2               tibble_3.2.1             
  [7] R.oo_1.26.0               XML_3.99-0.17             lifecycle_1.0.4          
 [10] sf_1.0-16                 edgeR_4.2.0               doParallel_1.0.17        
 [13] rprojroot_2.0.4           lattice_0.22-6            magrittr_2.0.3           
 [16] sass_0.4.9                limma_3.60.3              plotly_4.10.4            
 [19] rmarkdown_2.27            jquerylib_0.1.4           yaml_2.3.9               
 [22] httpuv_1.6.15             spam_2.10-0               sessioninfo_1.2.2        
 [25] sp_2.1-4                  cowplot_1.1.3             DBI_1.2.3                
 [28] RColorBrewer_1.1-3        golem_0.4.1               maps_3.4.2               
 [31] abind_1.4-5               zlibbioc_1.50.0           purrr_1.0.2              
 [34] R.utils_2.12.3            RCurl_1.98-1.14           rappdirs_0.3.3           
 [37] GenomeInfoDbData_1.2.12   ggrepel_0.9.5             irlba_2.3.5.1            
 [40] terra_1.7-78              units_0.8-5               dqrng_0.4.1              
 [43] DelayedMatrixStats_1.26.0 codetools_0.2-20          DropletUtils_1.24.0      
 [46] DelayedArray_0.30.1       DT_0.33                   scuttle_1.14.0           
 [49] tidyselect_1.2.1          UCSC.utils_1.0.0          viridis_0.6.5            
 [52] ScaledMatrix_1.12.0       shinyWidgets_0.8.6        BiocFileCache_2.12.0     
 [55] GenomicAlignments_1.40.0  jsonlite_1.8.8            BiocNeighbors_1.22.0     
 [58] e1071_1.7-14              scater_1.32.0             iterators_1.0.14         
 [61] foreach_1.5.2             tools_4.4.1               Rcpp_1.0.13              
 [64] glue_1.7.0                gridExtra_2.3             SparseArray_1.4.8        
 [67] xfun_0.45                 here_1.0.1                EBImage_4.46.0           
 [70] dplyr_1.1.4               HDF5Array_1.32.0          BiocManager_1.30.23      
 [73] fastmap_1.2.0             boot_1.3-30               rhdf5filters_1.16.0      
 [76] fansi_1.0.6               spData_2.3.1              rsvd_1.0.5               
 [79] digest_0.6.36             R6_2.5.1                  mime_0.12                
 [82] colorspace_2.1-0          wk_0.9.2                  jpeg_0.1-10              
 [85] RSQLite_2.3.7             R.methodsS3_1.8.2         config_0.3.2             
 [88] utf8_1.2.4                tidyr_1.3.1               generics_0.1.3           
 [91] rtracklayer_1.64.0        class_7.3-22              httr_1.4.7               
 [94] htmlwidgets_1.6.4         S4Arrays_1.4.1            spdep_1.3-5              
 [97] pkgconfig_2.0.3           gtable_0.3.5              rsconnect_1.3.1          
[100] blob_1.2.4                XVector_0.44.0            htmltools_0.5.8.1        
[103] dotCall64_1.1-1           fftwtools_0.9-11          scales_1.3.0             
[106] png_0.1-8                 attempt_0.3.1             knitr_1.48               
[109] rstudioapi_0.16.0         rjson_0.2.21              curl_5.2.1               
[112] proxy_0.4-27              cachem_1.1.0              rhdf5_2.48.0             
[115] BiocVersion_3.19.1        KernSmooth_2.23-24        vipor_0.4.7              
[118] parallel_4.4.1            AnnotationDbi_1.66.0      restfulr_0.0.15          
[121] s2_1.1.6                  pillar_1.9.0              grid_4.4.1               
[124] vctrs_0.6.5               promises_1.3.0            BiocSingular_1.20.0      
[127] dbplyr_2.5.0              beachmat_2.20.0           xtable_1.8-4             
[130] sfheaders_0.4.4           beeswarm_0.4.0            paletteer_1.6.0          
[133] evaluate_0.24.0           zeallot_0.1.0             magick_2.8.3             
[136] Rsamtools_2.20.0          cli_3.6.3                 locfit_1.5-9.10          
[139] compiler_4.4.1            crayon_1.5.3              classInt_0.4-10          
[142] rematch2_2.1.2            ggbeeswarm_0.7.2          viridisLite_0.4.2        
[145] deldir_2.0-4              munsell_0.5.1             Biostrings_2.72.1        
[148] lazyeval_0.2.2            tiff_0.1-12               Matrix_1.7-0             
[151] ExperimentHub_2.12.0      benchmarkme_1.0.8         sparseMatrixStats_1.16.0 
[154] bit64_4.0.5               ggplot2_3.5.1             Rhdf5lib_1.26.0          
[157] KEGGREST_1.44.1           statmod_1.5.0             shiny_1.8.1.1            
[160] AnnotationHub_3.12.0      memoise_2.0.1             bslib_0.7.0              
[163] benchmarkmeData_1.0.4     bit_4.0.5                
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-12-05
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────────────
 ! package                  * version   date (UTC) lib source
   abind                      1.4-5     2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi              1.66.0    2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub              3.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   attempt                    0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
   beachmat                   2.20.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm                   0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme                1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData            1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
   Biobase                  * 2.64.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics             * 0.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                     1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager                1.30.23   2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors              1.22.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel             * 1.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocSingular               1.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion                3.19.1    2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings                 2.72.1    2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                        4.0.5     2022-11-15 [1] CRAN (R 4.4.0)
   bit64                      4.0.5     2020-08-30 [1] CRAN (R 4.4.0)
   bitops                     1.0-7     2021-04-24 [1] CRAN (R 4.4.0)
   blob                       1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   boot                       1.3-30    2024-02-26 [1] CRAN (R 4.4.1)
   bslib                      0.7.0     2024-03-29 [1] CRAN (R 4.4.0)
   cachem                     1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
   class                      7.3-22    2023-05-03 [1] CRAN (R 4.4.1)
   classInt                   0.4-10    2023-09-05 [1] CRAN (R 4.4.0)
 P cli                        3.6.3     2024-06-21 [2] CRAN (R 4.4.0)
   codetools                  0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorout                 * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace                 2.1-0     2023-01-23 [1] CRAN (R 4.4.0)
   config                     0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                    1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
   crayon                     1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   curl                       5.2.1     2024-03-01 [1] CRAN (R 4.4.0)
   data.table               * 1.15.4    2024-03-30 [2] CRAN (R 4.4.0)
   DBI                        1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                     2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray               0.30.1    2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats         1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   deldir                     2.0-4     2024-02-28 [1] CRAN (R 4.4.0)
   digest                     0.6.36    2024-06-23 [1] CRAN (R 4.4.0)
   doParallel                 1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64                  1.1-1     2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                      1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   dqrng                      0.4.1     2024-05-28 [1] CRAN (R 4.4.0)
   DropletUtils               1.24.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   DT                         0.33      2024-04-04 [1] CRAN (R 4.4.0)
   e1071                      1.7-14    2023-12-06 [1] CRAN (R 4.4.0)
   EBImage                    4.46.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   edgeR                      4.2.0     2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate                   0.24.0    2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fansi                      1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   fastmap                    1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fftwtools                  0.9-11    2021-03-01 [1] CRAN (R 4.4.0)
   fields                     16.2      2024-06-27 [1] CRAN (R 4.4.0)
   filelock                   1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   foreach                    1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
   generics                   0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb             * 1.40.1    2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData           1.2.12    2024-05-01 [1] Bioconductor
   GenomicAlignments          1.40.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges            * 1.56.1    2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm                 0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2                    3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrepel                    0.9.5     2024-01-10 [1] CRAN (R 4.4.0)
   glue                       1.7.0     2024-01-09 [1] CRAN (R 4.4.0)
   golem                      0.4.1     2023-06-05 [1] CRAN (R 4.4.0)
   gridExtra                  2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gtable                     0.3.5     2024-04-22 [1] CRAN (R 4.4.0)
   HDF5Array                  1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   here                       1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools                  0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets                1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                     1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
   httr                       1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   IRanges                  * 2.38.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                      2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   iterators                  1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
   jpeg                       0.1-10    2022-11-29 [1] CRAN (R 4.4.0)
   jquerylib                  0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite                   1.8.8     2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST                   1.44.1    2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   KernSmooth                 2.23-24   2024-05-17 [1] CRAN (R 4.4.1)
   knitr                      1.48      2024-07-07 [1] CRAN (R 4.4.1)
   later                      1.3.2     2023-12-06 [1] CRAN (R 4.4.0)
   lattice                    0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval                   0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle                  1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                      3.60.3    2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                     1.5-9.10  2024-06-24 [1] CRAN (R 4.4.0)
   magick                     2.8.3     2024-02-18 [1] CRAN (R 4.4.0)
   magrittr                   2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   maps                       3.4.2     2023-12-15 [1] CRAN (R 4.4.0)
   Matrix                     1.7-0     2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics           * 1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats              * 1.3.0     2024-04-11 [1] CRAN (R 4.4.0)
   memoise                    2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   mime                       0.12      2021-09-28 [1] CRAN (R 4.4.0)
   munsell                    0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   paletteer                  1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
   parallelly               * 1.38.0    2024-07-27 [1] CRAN (R 4.4.0)
   pillar                     1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig                  2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   plotly                     4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
   png                        0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   promises                   1.3.0     2024-04-05 [1] CRAN (R 4.4.0)
   proxy                      0.4-27    2022-06-09 [1] CRAN (R 4.4.0)
   purrr                      1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R.methodsS3                1.8.2     2022-06-13 [1] CRAN (R 4.4.0)
   R.oo                       1.26.0    2024-01-24 [1] CRAN (R 4.4.0)
   R.utils                    2.12.3    2023-11-18 [1] CRAN (R 4.4.0)
   R6                         2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs                   0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer               1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                       1.0.13    2024-07-17 [1] CRAN (R 4.4.0)
   RCurl                      1.98-1.14 2024-01-09 [1] CRAN (R 4.4.0)
   rematch2                   2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
   restfulr                   0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
   rhdf5                      2.48.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rhdf5filters               1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Rhdf5lib                   1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rjson                      0.2.21    2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                    * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown                  2.27      2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot                  2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools                  2.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rsconnect                  1.3.1     2024-06-04 [1] CRAN (R 4.4.0)
   RSQLite                    2.3.7     2024-05-27 [1] CRAN (R 4.4.0)
   rstudioapi                 0.16.0    2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                       1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer                1.64.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   s2                         1.1.6     2023-12-19 [1] CRAN (R 4.4.0)
   S4Arrays                   1.4.1     2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors                * 0.42.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                       0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix               1.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                     1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater                     1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scuttle                    1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo                1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   sf                         1.0-16    2024-03-24 [1] CRAN (R 4.4.0)
   sfheaders                  0.4.4     2024-01-17 [1] CRAN (R 4.4.0)
   shiny                      1.8.1.1   2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets               0.8.6     2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment     * 1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sp                         2.1-4     2024-04-30 [1] CRAN (R 4.4.0)
   spam                       2.10-0    2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray                1.4.8     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats          1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment        * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialFeatureExperiment * 1.6.1     2024-05-15 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD              * 1.16.2    2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   spData                     2.3.1     2024-05-31 [1] CRAN (R 4.4.0)
   spdep                      1.3-5     2024-06-10 [1] CRAN (R 4.4.0)
   statmod                    1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   SummarizedExperiment     * 1.34.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   terra                      1.7-78    2024-05-22 [1] CRAN (R 4.4.0)
   tibble                     3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                      1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect                 1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   tiff                       0.1-12    2023-11-28 [1] CRAN (R 4.4.0)
   UCSC.utils                 1.0.0     2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   units                      0.8-5     2023-11-28 [1] CRAN (R 4.4.0)
   utf8                       1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                      0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                      0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                    0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite                0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   wk                         0.9.2     2024-07-09 [1] CRAN (R 4.4.0)
   xfun                       0.45      2024-06-16 [1] CRAN (R 4.4.0)
   XML                        3.99-0.17 2024-06-25 [1] CRAN (R 4.3.3)
   xtable                     1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
   XVector                    0.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                       2.3.9     2024-07-05 [1] CRAN (R 4.4.0)
   zeallot                    0.1.0     2018-01-28 [1] CRAN (R 4.4.0)
   zlibbioc                   1.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────