---
title: "04-Sex DE TF-Target GSEA"
author: "Bernard Mulvey"
date: "2024-10-31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(clusterProfiler) ## appears necessary for parallelization to work
library(fgsea)

require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc
library(parallel)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(parallelly)
options(bphost="localhost")
library(BiocParallel)
```

load voomLmFit sex DE stats, which were only run for the four clusterings of interest to begin with

```{r}
sexde <- readRDS("processed-data/09-Sex DE/01-voomLmFit_svg10-svg20-hvg20_Hmnydflt-mnn30-HmnylmbNA_BS-15-20-31-15VMHARCclpsd.RDS")

sexde <- sexde[grep(names(sexde),pattern="HARMONYlmbna_nnsvg10")]

# check:
# names(sexde)
# [1] "BSpace_k15_HARMONYlmbna_nnsvg10"          
# [2] "BSpace_k20_HARMONYlmbna_nnsvg10"          
# [3] "BSpace_k31_HARMONYlmbna_nnsvg10"          
# [4] "BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH"

```

### Enrichr TF-target databases gene sets retrieval
```{r}
tftargsets <- readRDS("processed-data/11-GSEA/00a-enrichrTFsets_forfgseainput.RDS")
```

# sex de gsea for TF-target sets
# note that we need SYMBOLS ("gene" column), not ens IDs in this case. the TF-target lists are only available as symbols
```{r}
### drop mito genes and chr X, Y genes
sexde <- lapply(sexde,FUN=function(x){x[!(chromosome_name %in% grep(chromosome_name,pattern="MT|X|Y",value=T))]})

### get named vectors of marker t-stats for each cluster in each result set (max 15+20+31+13)=79; probably less for clusters that dropped due to poor representation
sexdelist <- list()
i<-1
for (i in c(1:length(sexde))){
    sdoms <- unique(sexde[[i]]$assay)
    setnames(sexde[[i]],"t","tstat")
    j<-1
    for (j in c(1:length(sdoms))){
        tmp <- copy(sexde[[i]][assay==sdoms[j]])[,.(gene_name,tstat)]
        setorderv(tmp,"tstat",-1)
        ## get the stat column as a vector and give it the gene ids as names for fgsea
        ## note that since data.table (invisibly) uses list() for column data, its cleaner to just make the working table a data frame for this last bit
        tmp <- as.data.frame(tmp)
        v <- tmp$tstat
        names(v) <- tmp$gene_name
        sexdelist[[length(sexdelist)+1]] <- v
        names(sexdelist)[length(sexdelist)] <- paste0(names(sexde)[i],"_",sdoms[j])
        rm(v,tmp)
    }
    rm(j,sdoms)
}
rm(i)
```

### run fgsea
### looping through each domain (each run is very fast with 10 cpus in biocparallel, so just serially run 10cpu GSEA for each cluster)

```{r}
sexderes <- mapply(X=sexdelist,Y=names(sexdelist),SIMPLIFY=FALSE,FUN=function(X,Y){
   register(MulticoreParam())
   
   # get male-up enriched results
   z1<-fgseaMultilevel(pathways=tftargsets,stats = X,minSize = 15,maxSize = 500,eps=0.0,nPermSimple = 50000,nproc=10,scoreType = "pos")
   # coonvert leadingedge to one long string (it's returned as a list which is not compatible with data.table)
   z1$leadingEdge <- as.character(z1$leadingEdge)
   z1$leadingEdge <- gsub(z1$leadingEdge,pattern="^c\\((.*)\\)$",replacement="\\1")
   z1$leadingEdge <- gsub(z1$leadingEdge,pattern='"',replacement='')
   # drop newlines from leadingEdge
   z1$leadingEdge <- gsub(z1$leadingEdge,pattern=", \n",replacement=", ")
   z1 <- as.data.table(z1)
   z1[,spdom:=Y]
   z1[,dir:="male_up"]
   
   # get female-up enriched results
   register(MulticoreParam())
   z2<-fgseaMultilevel(pathways=tftargsets,stats = X,minSize = 15,maxSize = 500,eps=0.0,nPermSimple = 50000,nproc=10,scoreType = "neg")
   # coonvert leadingedge to one long string (it's returned as a list which is not compatible with data.table)
   z2$leadingEdge <- as.character(z2$leadingEdge)
   z2$leadingEdge <- gsub(z2$leadingEdge,pattern="^c\\((.*)\\)$",replacement="\\1")
   z2$leadingEdge <- gsub(z2$leadingEdge,pattern='"',replacement='')
   # drop newlines from leadingEdge
   z2$leadingEdge <- gsub(z2$leadingEdge,pattern=", \n",replacement=", ")
   z2 <- as.data.table(z2)
   z2[,spdom:=Y]
   z2[,dir:="female_up"]
   
   z<-rbind(z1,z2)
   return(z)
})
names(sexderes) <- names(sexdelist)

saveRDS(sexderes,"processed-data/11-GSEA/04a-Sex DE TF-target GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")

### collapse pathways-can't run this this part is WAY WAY slower than the GSEA itself

sexderes2 <- mcmapply(X=sexderes,Y=sexdelist,SIMPLIFY = FALSE, mc.cores=10,FUN=function(X,Y){
    tmpres <- fgsea::collapsePathways(fgseaRes = X,stats=Y,pathways = tftargsets)
    return(tmpres)
})
names(sexderes2) <- names(sexderes)

saveRDS(sexderes2,"processed-data/11-GSEA/04b-Sex DE TF-target fGSEA-collapsePathways output.RDS")

rm(sexdelist)
gc(full=T)

### subset to main enrichments
sexderes3 <- mapply(X=sexderes,Y=sexderes2,SIMPLIFY = FALSE,FUN=function(X,Y){
    X[pathway %in% Y[[1]]]
})

saveRDS(sexderes3,"processed-data/11-GSEA/04c-Sex DE TF-target GSEA clpspath mainPathways-only.RDS")
rm(sexderes,sexderes2,sexderes3)
gc(full=T)
```

### PT 2: unsigned (absolute value) de stats, one tailed test for enrichment in general sex de

get UNSIGNED de t stats
```{r}
### drop mito and sex chr genes
sexde <- lapply(sexde,FUN=function(x){
    x[!(chromosome_name %in% grep(chromosome_name,pattern="MT|X|Y",value=T))]})

### get named vectors of marker t-stats for each cluster in each result set (15+20+31+13)=79 (if no clusters dropped out of sex de analysis)
sexdelist.unsign <- list()
i<-1
for (i in c(1:length(sexde))){
    sdoms <- unique(sexde[[i]]$assay)
    j<-1
    for (j in c(1:length(sdoms))){
        tmp <- copy(sexde[[i]][assay==sdoms[j]])[,.(gene_name,abs(tstat))]
        setnames(tmp,2,"tstat")
        setorderv(tmp,"tstat",-1)
        ## get the stat column as a vector and give it the gene ids as names for fgsea
        ## note that since data.table (invisibly) uses list() for column data, its cleaner to just make the working table a data frame for this last bit
        tmp <- as.data.frame(tmp)
        v <- tmp$tstat
        names(v) <- tmp$gene_name
        sexdelist.unsign[[length(sexdelist.unsign)+1]] <- v
        names(sexdelist.unsign)[length(sexdelist.unsign)] <- paste0(names(sexde)[i],"_",sdoms[j])
        rm(v,tmp)
    }
    rm(j,sdoms)
}
rm(i)
```


```{r}
sexderes.unsign <- mapply(X=sexdelist.unsign,Y=names(sexdelist.unsign),SIMPLIFY=FALSE,FUN=function(X,Y){
   register(MulticoreParam())
   z<-fgseaMultilevel(pathways=tftargsets,stats = X,minSize = 15,maxSize = 500,eps=0.0,nPermSimple = 50000,nproc=10,scoreType = "pos")
   # coonvert leadingedge to one long string (it's returned as a list which is not compatible with data.table)
   z$leadingEdge <- as.character(z$leadingEdge)
   z$leadingEdge <- gsub(z$leadingEdge,pattern="^c\\((.*)\\)$",replacement="\\1")
   z$leadingEdge <- gsub(z$leadingEdge,pattern='"',replacement='')
   # remove newlines from leading edge
   z$leadingEdge <- gsub(z$leadingEdge,pattern=", \n",replacement=", ")
   z <- as.data.table(z)
   z[,spdom:=Y]
   return(z)
})
names(sexderes.unsign) <- names(sexdelist.unsign)

saveRDS(sexderes.unsign,"processed-data/11-GSEA/04d-Unsigned Sex DE TF-target GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")


# clear some memory for the next step which is more intensive

gc(full=T)
### collapse pathways-this part is WAY WAY slower than the GSEA itself

sexderes.unsign2 <- mcmapply(X=sexderes.unsign,Y=sexdelist.unsign,SIMPLIFY = FALSE, mc.cores=7,FUN=function(X,Y){
    tmpres <- fgsea::collapsePathways(fgseaRes = X,stats=Y,pathways = tftargsets)
    return(tmpres)
})
names(sexderes.unsign2) <- names(sexderes.unsign)

saveRDS(sexderes.unsign2,"processed-data/11-GSEA/04e-Unsigned Sex DE TF-target fGSEA-collapsePathways output.RDS")

rm(sexde)
gc(full=T)

### subset to main enrichments
sexderes.unsign3 <- mapply(X=sexderes.unsign,Y=sexderes.unsign2,SIMPLIFY = FALSE,FUN=function(X,Y){
    X[pathway %in% Y[[1]]]
})

saveRDS(sexderes.unsign3,"processed-data/11-GSEA/04f-Unsigned Sex DE TF-target GSEA clpspath mainPathways-only.RDS")
```


```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] BiocParallel_1.38.0    parallelly_1.38.0      default_1.0.0         
[4] colorout_1.3-0.2       fgsea_1.30.0           clusterProfiler_4.12.1
[7] data.table_1.15.4      rlang_1.1.4           

loaded via a namespace (and not attached):
  [1] DBI_1.2.3               gson_0.1.0              shadowtext_0.1.3       
  [4] gridExtra_2.3           magrittr_2.0.3          DOSE_3.30.1            
  [7] compiler_4.4.1          RSQLite_2.3.7           png_0.1-8              
 [10] vctrs_0.6.5             reshape2_1.4.4          stringr_1.5.1          
 [13] pkgconfig_2.0.3         crayon_1.5.3            fastmap_1.2.0          
 [16] XVector_0.44.0          ggraph_2.2.1            utf8_1.2.4             
 [19] HDO.db_0.99.1           rmarkdown_2.27          enrichplot_1.24.0      
 [22] UCSC.utils_1.0.0        purrr_1.0.2             bit_4.0.5              
 [25] xfun_0.45               zlibbioc_1.50.0         cachem_1.1.0           
 [28] aplot_0.2.3             GenomeInfoDb_1.40.1     jsonlite_1.8.8         
 [31] blob_1.2.4              tweenr_2.0.3            R6_2.5.1               
 [34] stringi_1.8.4           RColorBrewer_1.1-3      GOSemSim_2.30.0        
 [37] Rcpp_1.0.13             knitr_1.48              IRanges_2.38.1         
 [40] Matrix_1.7-0            splines_4.4.1           igraph_2.0.3           
 [43] tidyselect_1.2.1        qvalue_2.36.0           rstudioapi_0.16.0      
 [46] yaml_2.3.9              viridis_0.6.5           codetools_0.2-20       
 [49] lattice_0.22-6          tibble_3.2.1            plyr_1.8.9             
 [52] treeio_1.28.0           Biobase_2.64.0          withr_3.0.0            
 [55] KEGGREST_1.44.1         evaluate_0.24.0         gridGraphics_0.5-1     
 [58] scatterpie_0.2.3        polyclip_1.10-6         Biostrings_2.72.1      
 [61] ggtree_3.12.0           pillar_1.9.0            stats4_4.4.1           
 [64] ggfun_0.1.5             generics_0.1.3          rprojroot_2.0.4        
 [67] S4Vectors_0.42.1        ggplot2_3.5.1           tidytree_0.4.6         
 [70] munsell_0.5.1           scales_1.3.0            glue_1.7.0             
 [73] lazyeval_0.2.2          tools_4.4.1             fs_1.6.4               
 [76] graphlayouts_1.1.1      fastmatch_1.1-4         tidygraph_1.3.1        
 [79] cowplot_1.1.3           grid_4.4.1              ape_5.8                
 [82] tidyr_1.3.1             AnnotationDbi_1.66.0    colorspace_2.1-0       
 [85] nlme_3.1-165            patchwork_1.2.0         GenomeInfoDbData_1.2.12
 [88] ggforce_0.4.2           cli_3.6.3               fansi_1.0.6            
 [91] viridisLite_0.4.2       dplyr_1.1.4             gtable_0.3.5           
 [94] yulab.utils_0.1.4       digest_0.6.36           BiocGenerics_0.50.0    
 [97] ggplotify_0.1.2         ggrepel_0.9.5           farver_2.1.2           
[100] memoise_2.0.1           htmltools_0.5.8.1       lifecycle_1.0.4        
[103] httr_1.4.7              here_1.0.1              GO.db_3.19.1           
[106] bit64_4.0.5             MASS_7.3-61            
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-11-01
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────────────
 ! package          * version date (UTC) lib source
   AnnotationDbi      1.66.0  2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   ape                5.8     2024-04-11 [1] CRAN (R 4.4.0)
   aplot              0.2.3   2024-06-17 [1] CRAN (R 4.4.0)
   Biobase            2.64.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocGenerics       0.50.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel     * 1.38.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings         2.72.1  2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                4.0.5   2022-11-15 [1] CRAN (R 4.4.0)
   bit64              4.0.5   2020-08-30 [1] CRAN (R 4.4.0)
   blob               1.2.4   2023-03-17 [1] CRAN (R 4.4.0)
   cachem             1.1.0   2024-05-16 [1] CRAN (R 4.4.0)
 P cli                3.6.3   2024-06-21 [2] CRAN (R 4.4.0)
   clusterProfiler  * 4.12.1  2024-07-21 [1] Bioconductor 3.19 (R 4.4.1)
   codetools          0.2-20  2024-03-31 [1] CRAN (R 4.4.1)
   colorout         * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace         2.1-0   2023-01-23 [1] CRAN (R 4.4.0)
   cowplot            1.1.3   2024-01-22 [1] CRAN (R 4.4.0)
   crayon             1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
   data.table       * 1.15.4  2024-03-30 [2] CRAN (R 4.4.0)
   DBI                1.2.3   2024-06-02 [1] CRAN (R 4.4.0)
   default          * 1.0.0   2017-08-07 [1] CRAN (R 4.4.0)
   digest             0.6.36  2024-06-23 [1] CRAN (R 4.4.0)
   DOSE               3.30.1  2024-05-16 [1] Bioconductor 3.19 (R 4.4.0)
   dplyr              1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
   enrichplot         1.24.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate           0.24.0  2024-06-10 [1] CRAN (R 4.4.0)
   fansi              1.0.6   2023-12-08 [1] CRAN (R 4.4.0)
   farver             2.1.2   2024-05-13 [1] CRAN (R 4.4.0)
   fastmap            1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
   fastmatch          1.1-4   2023-08-18 [1] CRAN (R 4.4.0)
   fgsea            * 1.30.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   fs                 1.6.4   2024-04-25 [1] CRAN (R 4.4.0)
   generics           0.1.3   2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb       1.40.1  2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData   1.2.12  2024-05-01 [1] Bioconductor
   ggforce            0.4.2   2024-02-19 [1] CRAN (R 4.4.0)
   ggfun              0.1.5   2024-05-28 [1] CRAN (R 4.4.0)
   ggplot2            3.5.1   2024-04-23 [1] CRAN (R 4.4.0)
   ggplotify          0.1.2   2023-08-09 [1] CRAN (R 4.4.0)
   ggraph             2.2.1   2024-03-07 [1] CRAN (R 4.4.0)
   ggrepel            0.9.5   2024-01-10 [1] CRAN (R 4.4.0)
   ggtree             3.12.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   glue               1.7.0   2024-01-09 [1] CRAN (R 4.4.0)
   GO.db              3.19.1  2024-05-01 [1] Bioconductor
   GOSemSim           2.30.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   graphlayouts       1.1.1   2024-03-09 [1] CRAN (R 4.4.0)
   gridExtra          2.3     2017-09-09 [1] CRAN (R 4.4.0)
   gridGraphics       0.5-1   2020-12-13 [1] CRAN (R 4.4.0)
   gson               0.1.0   2023-03-07 [1] CRAN (R 4.4.0)
   gtable             0.3.5   2024-04-22 [1] CRAN (R 4.4.0)
   HDO.db             0.99.1  2024-05-01 [1] Bioconductor
   here               1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
   htmltools          0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
   httr               1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
   igraph             2.0.3   2024-03-13 [1] CRAN (R 4.4.0)
   IRanges            2.38.1  2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   jsonlite           1.8.8   2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST           1.44.1  2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   knitr              1.48    2024-07-07 [1] CRAN (R 4.4.1)
   lattice            0.22-6  2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval           0.2.2   2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle          1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
   magrittr           2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
   MASS               7.3-61  2024-06-13 [1] CRAN (R 4.4.0)
   Matrix             1.7-0   2024-04-26 [1] CRAN (R 4.4.1)
   memoise            2.0.1   2021-11-26 [1] CRAN (R 4.4.0)
   munsell            0.5.1   2024-04-01 [1] CRAN (R 4.4.0)
   nlme               3.1-165 2024-06-06 [1] CRAN (R 4.4.0)
   parallelly       * 1.38.0  2024-07-27 [1] CRAN (R 4.4.0)
   patchwork          1.2.0   2024-01-08 [1] CRAN (R 4.4.0)
   pillar             1.9.0   2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig          2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
   plyr               1.8.9   2023-10-02 [1] CRAN (R 4.4.0)
   png                0.1-8   2022-11-29 [1] CRAN (R 4.4.0)
   polyclip           1.10-6  2023-09-27 [1] CRAN (R 4.4.0)
   purrr              1.0.2   2023-08-10 [1] CRAN (R 4.4.0)
   qvalue             2.36.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   R6                 2.5.1   2021-08-19 [1] CRAN (R 4.4.0)
   RColorBrewer       1.1-3   2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp               1.0.13  2024-07-17 [1] CRAN (R 4.4.0)
   reshape2           1.4.4   2020-04-09 [1] CRAN (R 4.4.0)
 P rlang            * 1.1.4   2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown          2.27    2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot          2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
   RSQLite            2.3.7   2024-05-27 [1] CRAN (R 4.4.0)
   rstudioapi         0.16.0  2024-03-24 [1] CRAN (R 4.4.0)
   S4Vectors          0.42.1  2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   scales             1.3.0   2023-11-28 [1] CRAN (R 4.4.0)
   scatterpie         0.2.3   2024-06-05 [1] CRAN (R 4.4.0)
   sessioninfo        1.2.2   2021-12-06 [1] CRAN (R 4.4.0)
   shadowtext         0.1.3   2024-01-19 [1] CRAN (R 4.4.0)
   stringi            1.8.4   2024-05-06 [1] CRAN (R 4.4.0)
   stringr            1.5.1   2023-11-14 [1] CRAN (R 4.4.0)
   tibble             3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
   tidygraph          1.3.1   2024-01-30 [1] CRAN (R 4.4.0)
   tidyr              1.3.1   2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect         1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
   tidytree           0.4.6   2023-12-12 [1] CRAN (R 4.4.0)
   treeio             1.28.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   tweenr             2.0.3   2024-02-26 [1] CRAN (R 4.4.0)
   UCSC.utils         1.0.0   2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   utf8               1.2.4   2023-10-22 [1] CRAN (R 4.4.0)
   vctrs              0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
   viridis            0.6.5   2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite        0.4.2   2023-05-02 [1] CRAN (R 4.4.0)
   withr              3.0.0   2024-01-16 [1] CRAN (R 4.4.0)
   xfun               0.45    2024-06-16 [1] CRAN (R 4.4.0)
   XVector            0.44.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml               2.3.9   2024-07-05 [1] CRAN (R 4.4.0)
   yulab.utils        0.1.4   2024-01-28 [1] CRAN (R 4.4.0)
   zlibbioc           1.50.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────
