---
title: "06-Sex DE-Common Variant Risk Gene GSEA"
author: "Bernard Mulvey"
date: "2024-07-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(clusterProfiler) ## appears necessary for parallelization to work
library(fgsea)

require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc
library(parallel)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(parallelly)
options(bphost="localhost")
library(BiocParallel)
```

### load aggregated disease-gene lists and reformat
```{r}
setDTthreads(1,restore_after_fork=FALSE)

gwas <- fread("raw-data/Genesets from GWAS pubs, TWAS, and disgenet/gwassets.txt",header=T)
twas <- fread("raw-data/Genesets from GWAS pubs, TWAS, and disgenet/twassets.txt")
dgn <- fread("raw-data/Genesets from GWAS pubs, TWAS, and disgenet/disgenet_curated_genes.txt")
setnames(dgn,"gene_symbol","symbol")
un <- fread("raw-data/Genesets from GWAS pubs, TWAS, and disgenet/nonredundant_union_set_genesymbols.txt")

dissets <- list(gwas,twas,dgn,un)
names(dissets) <- c("gwas","twas","dgn","union")

# make a list object of identifiers per gene set
# result is a list of lists by modality
dissets <- lapply(dissets,FUN=function(x){
    dissets2 <- copy(x)
    q <- split(dissets2$symbol,dissets2$dx)
    d <- lapply(q,FUN=function(x){x[!duplicated(x)]})
    return(d)
})

rm(gwas,twas,dgn,un)
gc(full=T)
```


### sex DEG setup
```{r}
sexde <- readRDS("processed-data/09-Sex DE/01-voomLmFit_svg10-svg20-hvg20_Hmnydflt-mnn30-HmnylmbNA_BS-15-20-31-15VMHARCclpsd.RDS")

### subset to harmony lambda=NULL nnsvg10 runs (k15 k20 k31 and k15 with vmh and arc collapsed to single domains)
sexde <- sexde[grep(names(sexde),pattern="HARMONYlmbna_nnsvg10")]

### drop mito genes and sex chromosomal genes (i.e. autosomal analysis) and any duplicated symbols
sexde <- lapply(sexde,FUN=function(x){
    x[!(chromosome_name %in% grep(chromosome_name,pattern="MT|X|Y",value=T))]})

### get named vectors of marker t-stats for each cluster in each result set (15+20+31+13)=79. we need symbols since not all of the dis-gene data sources provide ens identifiers
sexdelist <- list()
i<-1
for (i in c(1:length(sexde))){
    sdoms <- unique(sexde[[i]]$assay)
    setnames(sexde[[i]],"t","tstat")
    j<-1
    for (j in c(1:length(sdoms))){
        tmp <- copy(sexde[[i]][assay==sdoms[j]])[,.(gene_name,tstat)]
        ### juuuust in case duplicate genes are somehow leaking through
        tmp <- tmp[!duplicated(gene_name)]
        setorderv(tmp,"tstat",-1)
        ## get the stat column as a vector and give it the gene ids as names for fgsea
        ## note that since data.table (invisibly) uses list() for column data, its cleaner to just make the working table a data frame for this last bit
        tmp <- as.data.frame(tmp)
        v <- tmp$tstat
        names(v) <- tmp$gene_name
        sexdelist[[length(sexdelist)+1]] <- v
        names(sexdelist)[length(sexdelist)] <- paste0(names(sexde)[i],"_",sdoms[j])
        rm(v,tmp)
    }
    rm(j,sdoms)
}
rm(i,sexde)
```

```{r}
### loop through each domain and modality-disease gene set (each run is very fast with 10 cpus in biocparallel, so just serially run 10cpu GSEA for each cluster)
dissets.res <- list()

i<-1
for (i in c(1:length(dissets))){
    cursets <- copy(dissets[[i]])
    sexderes <- mapply(X=sexdelist,Y=names(sexdelist),SIMPLIFY=FALSE,FUN=function(X,Y){
        # max size for dis gene sets is BMI at 8100-some, so set max size to 8200 
        register(MulticoreParam())
        z<-fgseaMultilevel(pathways=cursets,stats = X,minSize = 15,maxSize = 8200,eps=0.0,nPermSimple = 50000,nproc=8)
        # coonvert leadingedge to one long string (it's returned as a list which is not compatible with data.table)
        z$leadingEdge <- as.character(z$leadingEdge)
        z$leadingEdge <- gsub(z$leadingEdge,pattern="^c\\((.*)\\)$",replacement="\\1")
        z$leadingEdge <- gsub(z$leadingEdge,pattern='"',replacement='')
        z <- as.data.table(z)
        z[,spdom:=Y]
        return(z)
    })
    dissets.res[[i]] <- sexderes
    names(dissets.res)[i] <- names(dissets)[i]
    rm(cursets,sexderes)
}
rm(i)
```


#### concatenate
```{r}
i <-1
for (i in c(1:length(dissets.res))){
    j<-1
    for (j in c(1:length(dissets.res[[i]]))){
        if(i==1 & j==1){
            allsexdedisres <- copy(dissets.res[[i]][[j]])
            allsexdedisres[,genesets_from:=names(dissets.res)[i]]
        }
        else{
            tmp <- copy(dissets.res[[i]][[j]])
            tmp[,genesets_from:=names(dissets.res)[i]]
            allsexdedisres <- rbind(allsexdedisres,tmp)
            rm(tmp)
        }
    }
}

saveRDS(allsexdedisres,"processed-data/11-GSEA/06-Common variant disease gene enrichment in sex DEGs.RDS")
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
─ Session info ───────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-07-28
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────────
 ! package          * version date (UTC) lib source
   BiocGenerics     * 0.50.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager        1.30.23 2024-05-04 [1] CRAN (R 4.4.0)
   BiocParallel     * 1.38.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings       * 2.72.1  2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   callr              3.7.6   2024-03-25 [1] CRAN (R 4.4.0)
 P cli                3.6.3   2024-06-21 [2] CRAN (R 4.4.0)
   codetools          0.2-20  2024-03-31 [1] CRAN (R 4.4.1)
   colorout         * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace         2.1-0   2023-01-23 [1] CRAN (R 4.4.0)
   cowplot            1.1.3   2024-01-22 [1] CRAN (R 4.4.0)
   crayon             1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
   curl               5.2.1   2024-03-01 [1] CRAN (R 4.4.0)
   data.table       * 1.15.4  2024-03-30 [2] CRAN (R 4.4.0)
   default          * 1.0.0   2017-08-07 [1] CRAN (R 4.4.0)
   desc               1.4.3   2023-12-10 [1] CRAN (R 4.4.0)
   digest             0.6.36  2024-06-23 [1] CRAN (R 4.4.0)
   dplyr              1.1.4   2023-11-17 [1] CRAN (R 4.4.0)
   evaluate           0.24.0  2024-06-10 [1] CRAN (R 4.4.0)
   fansi              1.0.6   2023-12-08 [1] CRAN (R 4.4.0)
   fastmap            1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
   fastmatch          1.1-4   2023-08-18 [1] CRAN (R 4.4.0)
   fgsea            * 1.30.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   generics           0.1.3   2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb     * 1.40.1  2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData   1.2.12  2024-05-01 [1] Bioconductor
   ggplot2            3.5.1   2024-04-23 [1] CRAN (R 4.4.0)
   glue               1.7.0   2024-01-09 [1] CRAN (R 4.4.0)
   gtable             0.3.5   2024-04-22 [1] CRAN (R 4.4.0)
   here               1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
   htmltools          0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
   httr               1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
   IRanges          * 2.38.1  2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   jsonlite           1.8.8   2023-12-04 [1] CRAN (R 4.4.0)
   knitr              1.48    2024-07-07 [1] CRAN (R 4.4.1)
   lattice            0.22-6  2024-03-20 [1] CRAN (R 4.4.1)
   lifecycle          1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
   magrittr           2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
   Matrix             1.7-0   2024-04-26 [1] CRAN (R 4.4.1)
   munsell            0.5.1   2024-04-01 [1] CRAN (R 4.4.0)
   parallelly       * 1.38.0  2024-07-27 [1] CRAN (R 4.4.0)
   pillar             1.9.0   2023-03-22 [1] CRAN (R 4.4.0)
   pkgbuild           1.4.4   2024-03-17 [1] CRAN (R 4.4.0)
   pkgconfig          2.0.3   2019-09-22 [1] CRAN (R 4.4.0)
   processx           3.8.4   2024-03-16 [1] CRAN (R 4.4.0)
   ps                 1.7.7   2024-07-02 [1] CRAN (R 4.4.0)
   R6                 2.5.1   2021-08-19 [1] CRAN (R 4.4.0)
   Rcpp               1.0.13  2024-07-17 [1] CRAN (R 4.4.0)
   remotes            2.5.0   2024-03-17 [1] CRAN (R 4.4.0)
 P rlang            * 1.1.4   2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown          2.27    2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot          2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
   rstudioapi         0.16.0  2024-03-24 [1] CRAN (R 4.4.0)
   S4Vectors        * 0.42.1  2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   scales             1.3.0   2023-11-28 [1] CRAN (R 4.4.0)
   sessioninfo        1.2.2   2021-12-06 [1] CRAN (R 4.4.0)
   tibble             3.2.1   2023-03-20 [1] CRAN (R 4.4.0)
   tidyselect         1.2.1   2024-03-11 [1] CRAN (R 4.4.0)
   UCSC.utils         1.0.0   2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   utf8               1.2.4   2023-10-22 [1] CRAN (R 4.4.0)
   vctrs              0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
   xfun               0.45    2024-06-16 [1] CRAN (R 4.4.0)
   XVector          * 0.44.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml               2.3.9   2024-07-05 [1] CRAN (R 4.4.0)
   zlibbioc           1.50.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────────
