---
title: "11-Marker GSEA supplements output"
author: "Bernard Mulvey"
date: "2025-03-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

```


### read in msigdb pre and post "collapsePathways" results
```{r}
msigdbfull <- readRDS("processed-data/11-GSEA/01a-Marker msigdb_1,2,3,5,6,8 GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")

msigdbcollapsed <- readRDS("processed-data/11-GSEA/01c-Marker mSigDB_1,2,3,5,6,8 GSEA clpspath mainPathways-only.RDS")

msigbindpatterns <- c("k15_HARMONYlmbna_nnsvg10_60kiter_collapsed","k15_HARMONYlmbna_nnsvg10_60kiter_X","k20_HARMONYlmbna_nnsvg10_60kiter","k31_HARMONYlmbna_nnsvg10_60kiter")

clustres <- list()
i<-1
for (i in c(1:length(msigbindpatterns))){
    curpatt <- msigbindpatterns[i]
    subfull <- msigdbfull[grep(names(msigdbfull),pattern=curpatt)]
    subclps <- msigdbcollapsed[grep(names(msigdbcollapsed),pattern=curpatt)]
    
    j<-1
    for (j in c(1:length(subfull))){
        tmpres <- subfull[[j]]
        tmpres[,spdom:=gsub(spdom,pattern="^.*_(X.*)$",replacement="\\1")]
        
        tmpres[,mainPathway:="No"]
        tmpres[pathway %in% msigdbcollapsed[[names(subfull)[j]]]$pathway,mainPathway:="Yes"]
        
        if(j==1){catres <- copy(tmpres)}
        else{catres <- rbind(catres,tmpres)}
        rm(tmpres)
    }
    clustres[[i]] <- catres
    rm(subfull,subclps,curpatt,j)
}
rm(i,catres)

### fix VMH/ARC cluster labels for the collapsed case which doesn't really work with this (no leading X)
clustres[[1]][spdom=="XVMH",spdom:="VMH"]
clustres[[1]][spdom=="XARC",spdom:="ARC"]

### assign VMH.1,2 ARC1,2 labels to the corresponding k15 results
clustres[[2]][spdom=="X4",spdom:="ARC.1"]
clustres[[2]][spdom=="X6",spdom:="ARC.2"]
clustres[[2]][spdom=="X7",spdom:="VMH.1"]
clustres[[2]][spdom=="X12",spdom:="VMH.2"]

### add the k20/k31 ARC and VMH cluster identifiers as used elsewhere
clustres[[3]][spdom=="X2",spdom:="VMH20.1"]
clustres[[3]][spdom=="X3",spdom:="VMH20.2"]
clustres[[3]][spdom=="X1",spdom:="ARC20.1"]
clustres[[3]][spdom=="X4",spdom:="ARC20.2"]

clustres[[4]][spdom=="X6",spdom:="VMH31.1"]
clustres[[4]][spdom=="X21",spdom:="VMH31.2"]
clustres[[4]][spdom=="X25",spdom:="VMH31.3"]
clustres[[4]][spdom=="X26",spdom:="VMH31.4"]
clustres[[4]][spdom=="X31",spdom:="VMH31.5"]
clustres[[4]][spdom=="X2",spdom:="ARC31.1"]
clustres[[4]][spdom=="X4",spdom:="ARC31.2"]
clustres[[4]][spdom=="X11",spdom:="ARC31.3"]
clustres[[4]][spdom=="X22",spdom:="ARC31.4"]

## check these are okay
lapply(clustres,FUN=function(x){unique(x$spdom)})
clustres[[1]][,spdom:=gsub(spdom,pattern="XX",replacement="X")]
## 

names(clustres) <- c("BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH","BSpace_k15_HARMONYlmbna_nnsvg10","BSpace_k20_HARMONYlmbna_nnsvg10","BSpace_k31_HARMONYlmbna_nnsvg10")

clustres <- lapply(clustres,function(x){
    # subset to enrichments >0, negative enrichments don't mean anything here (as opposed to sex DE results)
    x <- x[NES>0]
    setnames(x,c(1:10),c("Pathway","P.Value","Adj.P","log2err","Enrichment Score","Normalized Enrichment Score","Geneset Size","Leading Edge Genes","Visium Domain","Main Pathway after fgsea::collapsePathways"))
    x <- x[,c(9,1,10,3,6,5,7,4,2,8)]
    return(x)
})

saveRDS(clustres,"local_supptab_assembly/Supp Data-Domain msigdb GSEA results.RDS")

rm(clustres,msigdbcollapsed,msigdbfull,msigbindpatterns)
```


### make a table of the TF-target sets tested for GSEA
### need to re-create this from the fgsea-ready list object saved for running analyses. code that created that is in other repo scripts as cited below
```{r}
library(Biostrings)
library(stringr)

tftargsets <- list.files("processed-data/11-GSEA/00-enrichr TF-target sets from dbs and lit mining DLed 120623",pattern=".txt",full.names=T)

sourcename <- gsub(tftargsets,pattern="^.*DLed 120623/(.*).txt$",replacement="\\1")

# read each file as a single column table of longass lines, split at the first \t to get set name and set members separated, then expand them out to a one-to-one table to feed into split
tftargsets <- mapply(X=tftargsets,Y=sourcename,SIMPLIFY=F,FUN=function(X,Y){
  tmp <- fread(X,sep = '\n')
  tmp[,setmembers:=apply(.SD,MARGIN=1,FUN=function(x){stringi::stri_split_fixed(x,pattern="\t",n=2)[[1]][2]}),.SDcols=1]
  tmp[,setname:=apply(.SD,MARGIN=1,FUN=function(x){stringi::stri_split_fixed(x,pattern="\t",n=2)[[1]][1]}),.SDcols=1]
  tmp[,setmembers:=gsub(setmembers,pattern="\t",replacement=";")]
  tmp[,setmembers:=gsub("^;","",setmembers)]
  tmp[,setname:=gsub(setname,pattern=" ",replacement="_")]
  tmp[,sourceset:=gsub(Y,pattern=" ",replacement="_")]
  return(tmp[,c(3,2,4)])
})

# for the TF-LOF_Expression_from_GEO we need to take an extra step, since these also have scores (DE values?) with them
tftargmanual <- fread("processed-data/11-GSEA/00-enrichr TF-target sets from dbs and lit mining DLed 120623/TF-LOF_Expression_from_GEO.txt",sep="\n")
setnames(tftargmanual,1,"origstr")
tftargmanual[,setname:=apply(.SD,MARGIN=1,FUN=function(x){str_split_fixed(x,pattern="\t",n=2)[1]}),.SDcols=1]
tftargmanual[,setname:=gsub(" ","_",setname)]
tftargmanual[,setmembers:=apply(.SD,MARGIN=1,FUN=function(x){str_split_fixed(x,pattern="\t",n=2)[2]}),.SDcols=1]
tftargmanual[,setmembers2 := gsub("\\d","",setmembers)]
tftargmanual[,setmembers2 := gsub("\\.","",setmembers2)]
tftargmanual[,setmembers2:=gsub(",\t",";",setmembers2)]
tftargmanual[,setmembers2:=gsub("\t","",setmembers2)]
tftargmanual[,setmembers2:=gsub(" ","",setmembers2)]
tftargmanual[,setmembers:=setmembers2]
tftargmanual[,setmembers2:=NULL]
tftargmanual[,sourceset:="TF-LOF_Expression_from_GEO"]

### replace the original TF-LOF_Expression_from_GEO with the new one
tftargsets[[4]] <- copy(tftargmanual)[,c(3,2,4)]

# clean up
rm(tftargmanual)

# concatenate

i<-1
for (i in c(1:length(tftargsets))){
    if(i==1){tftargtab <- copy(tftargsets[[i]])}
    else{tftargtab <- rbind(tftargtab,tftargsets[[i]])}
}

### add a column corresponding to the TF symbol itself. this comes from code in 11-GSEA/08 and 09 ....rmd

y <- tftargtab[!(sourceset=="Rummagene_transcription_factors")]
y[,tf:=apply(.SD,MARGIN = 1,FUN=function(h){strsplit(h,"_",1)[[1]][1]}),.SDcols="setname"]

### rummagene TFs are named differently:
z <- tftargtab[sourceset=="Rummagene_transcription_factors"]
# reverse the output of a strsplit of a strsplit o.0 to get the string immediately preceding "_Rummagene_.."
z[,tf:=apply(.SD,MARGIN=1,FUN=function(p){
    rev.default(strsplit(p,"_",-1)[[1]])[1]
}),.SDcols="setname"]
fix <- rbind(z,y)
# make all upper case
fix[,tf:=toupper(tf)]


### finally, change the TF set names to match those in the result tables by appending the source
fix[,setname:=paste0(setname,"_",sourceset)]

names(fix)
fix <- fix[,c(4,1,3,2)]

### hang onto this objectfor merging in this info into TF GSEA results next


rm(y,z,tftargsets,tftargtab,sourcename)
```

### write out marker-TF target results; append the TF symbol; append whether that TF is among those that passed the domain-expression threshold
```{r}
tffull <- readRDS("processed-data/11-GSEA/02a-Marker TF-target GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")

tfcollapsed <- readRDS("processed-data/11-GSEA/02c-Marker TF-target GSEA clpspath mainPathways-only.RDS")

tfxprd <- readRDS("processed-data/11-GSEA/08-Marker TF-target GSEA results from 02a filtered to TFs expressed in avg 5pct of cluster spots and at least 1pct of cluster spots in 5samps.RDS")

tfbindpatterns <- c("k15_HARMONYlmbna_nnsvg10_60kiter_collapsed","k15_HARMONYlmbna_nnsvg10_60kiter_X","k20_HARMONYlmbna_nnsvg10_60kiter","k31_HARMONYlmbna_nnsvg10_60kiter")

clustres <- list()
i<-1
for (i in c(1:length(tfbindpatterns))){
    curpatt <- tfbindpatterns[i]
    subfull <- tffull[grep(names(tffull),pattern=curpatt)]
    subclps <- tfcollapsed[grep(names(tfcollapsed),pattern=curpatt)]
    subxpr <- tfxprd[grep(names(tfcollapsed),pattern=curpatt)]
    j<-1
    for (j in c(1:length(subfull))){
        tmpres <- subfull[[j]]
        tmpres[,spdom:=gsub(spdom,pattern="^.*_(X.*)$",replacement="\\1")]
        
        tmpres[,mainPathway:="No"]
        tmpres[pathway %in% tfcollapsed[[names(subfull)[j]]]$pathway,mainPathway:="Yes"]
        
        tmpres[,Expressed_TF:="No"]
        tmpres[pathway %in% subxpr[[names(subfull)[j]]]$pathway,Expressed_TF:="Yes"]

        
        if(j==1){catres <- copy(tmpres)}
        else{catres <- rbind(catres,tmpres)}
        rm(tmpres)
    }
    clustres[[i]] <- catres
    rm(subfull,subclps,curpatt,j,subxpr)
}
rm(i,catres)

### fix VMH/ARC cluster labels for the collapsed case which doesn't really work with this (no leading X)
### fix VMH/ARC cluster labels for the collapsed case which doesn't really work with this (no leading X)
clustres[[1]][spdom=="XVMH",spdom:="VMH"]
clustres[[1]][spdom=="XARC",spdom:="ARC"]

### assign VMH.1,2 ARC1,2 labels to the corresponding k15 results
clustres[[2]][spdom=="X4",spdom:="ARC.1"]
clustres[[2]][spdom=="X6",spdom:="ARC.2"]
clustres[[2]][spdom=="X7",spdom:="VMH.1"]
clustres[[2]][spdom=="X12",spdom:="VMH.2"]

### add the k20/k31 ARC and VMH cluster identifiers as used elsewhere
clustres[[3]][spdom=="X2",spdom:="VMH20.1"]
clustres[[3]][spdom=="X3",spdom:="VMH20.2"]
clustres[[3]][spdom=="X1",spdom:="ARC20.1"]
clustres[[3]][spdom=="X4",spdom:="ARC20.2"]

clustres[[4]][spdom=="X6",spdom:="VMH31.1"]
clustres[[4]][spdom=="X21",spdom:="VMH31.2"]
clustres[[4]][spdom=="X25",spdom:="VMH31.3"]
clustres[[4]][spdom=="X26",spdom:="VMH31.4"]
clustres[[4]][spdom=="X31",spdom:="VMH31.5"]
clustres[[4]][spdom=="X2",spdom:="ARC31.1"]
clustres[[4]][spdom=="X4",spdom:="ARC31.2"]
clustres[[4]][spdom=="X11",spdom:="ARC31.3"]
clustres[[4]][spdom=="X22",spdom:="ARC31.4"]


### name the result list
names(clustres) <- c("BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH","BSpace_k15_HARMONYlmbna_nnsvg10","BSpace_k20_HARMONYlmbna_nnsvg10","BSpace_k31_HARMONYlmbna_nnsvg10")

clustres <- lapply(clustres,function(x){
    z <- merge.data.table(fix[,.(tf,setname)],x,by.x="setname",by.y="pathway",all.y=T)
    ## subset to positive enrichments, since negative ones don't mean anything here (unlike sex DE)
    z <- z[NES>0]
        
    setnames(z,c("TF Target Set","TF Symbol","P.Value","Adj.P","log2err","Enrichment Score","Normalized Enrichment Score","Geneset Size","Leading Edge Genes","Visium Domain","Main Geneset after fgsea::collapsePathways","Expressed TF"))
    z <- z[,c(2,10:12,4,7,6,5,1,8,3,9)]
    return(z)
})

saveRDS(clustres,"local_supptab_assembly/Supp Data-Domain_Marker_TFtarget_GSEA_results_posenrichonly.RDS")

rm(clustres,tfcollapsed,tffull,tfbindpatterns,fix)
```

### write out marker - disease GSEA results
```{r}
mkdisres <- readRDS("processed-data/11-GSEA/05-Common variant disease gene enrichment in markers.RDS")

spdompatt <- c("k15_HARMONYlmbna_nnsvg10_60kiter_collapsed","k15_HARMONYlmbna_nnsvg10_60kiter_X","k20_HARMONYlmbna_nnsvg10_60kiter","k31_HARMONYlmbna_nnsvg10_60kiter")

i<-1
for (i in c(1:4)){
    ## special condition for the second since its weird
    if(i==2){
        mkdisres[spdom %in% grep(spdom, pattern=spdompatt[i],value=T),Domain:=gsub(spdom,pattern="k15_HARMONYlmbna_nnsvg10_60kiter_(X.*)",replacement="\\1")]
    }
    
    else{
        mkdisres[spdom %in% grep(spdom, pattern=spdompatt[i],value=T),Domain:=gsub(spdom,pattern=paste0(spdompatt[i],"_(.*)"),replacement="\\1")]
    }
    mkdisres[spdom %in% grep(spdom,pattern=spdompatt[i],value=T),`BayesSpace Run`:=gsub(spdom,pattern="^(k.*)_HARMONY.*",replacement="\\1")]
}

### correct for the collapsed run
mkdisres[spdom %in% grep(spdom,pattern=spdompatt[1],value=T),`BayesSpace Run`:="k15-ARC and VMH Domains Collapsed After"]

### fix labels for collapsed, add labels for k15 singles
mkdisres[`BayesSpace Run`=="k15-ARC and VMH Domains Collapsed After"&Domain=="XARC",Domain:="ARC"]
mkdisres[`BayesSpace Run`=="k15-ARC and VMH Domains Collapsed After"&Domain=="XVMH",Domain:="VMH"]

mkdisres[`BayesSpace Run`=="k15"&Domain=="X4",Domain:="ARC.1"]
mkdisres[`BayesSpace Run`=="k15"&Domain=="X6",Domain:="ARC.2"]
mkdisres[`BayesSpace Run`=="k15"&Domain=="X7",Domain:="VMH.1"]
mkdisres[`BayesSpace Run`=="k15"&Domain=="X12",Domain:="VMH.2"]

### add the k20/k31 ARC and VMH cluster identifiers as used elsewhere
mkdisres[`BayesSpace Run`=="k20"&Domain=="X2",Domain:="VMH20.1"]
mkdisres[`BayesSpace Run`=="k20"&Domain=="X3",Domain:="VMH20.2"]
mkdisres[`BayesSpace Run`=="k20"&Domain=="X1",Domain:="ARC20.1"]
mkdisres[`BayesSpace Run`=="k20"&Domain=="X4",Domain:="ARC20.2"]

mkdisres[`BayesSpace Run`=="k31"&Domain=="X6",Domain:="VMH31.1"]
mkdisres[`BayesSpace Run`=="k31"&Domain=="X21",Domain:="VMH31.2"]
mkdisres[`BayesSpace Run`=="k31"&Domain=="X25",Domain:="VMH31.3"]
mkdisres[`BayesSpace Run`=="k31"&Domain=="X26",Domain:="VMH31.4"]
mkdisres[`BayesSpace Run`=="k31"&Domain=="X31",Domain:="VMH31.5"]
mkdisres[`BayesSpace Run`=="k31"&Domain=="X2",Domain:="ARC31.1"]
mkdisres[`BayesSpace Run`=="k31"&Domain=="X4",Domain:="ARC31.2"]
mkdisres[`BayesSpace Run`=="k31"&Domain=="X11",Domain:="ARC31.3"]
mkdisres[`BayesSpace Run`=="k31"&Domain=="X22",Domain:="ARC31.4"]

mkdisres[,spdom:=NULL]

names(mkdisres)
setnames(mkdisres,c("Diagnosis or Trait","P.Value","P.Adj","log2err","Enrichment Score","Normalized Enrichment Score","Geneset Size","Leading Edge Genes","Data or Publication Type from Which Geneset Derives","Domain","BayesSpace Run"))

mkdisres <- mkdisres[,c(1,9,11,10,3,6,5,4,7,2,8)]
names(mkdisres)

fwrite(mkdisres,"local_supptab_assembly/marker-disgene-gsea.txt",sep='\t',quote=F)
```

## 031525 -- there was an errant line break in the text file that was preventing excel from seeing some rows when filtering and splitting out by gene source when making supp tabs. the source .txt file and onedrive submission 3 copy of ST11 (sheet A, where this issue presented) have been fixed.

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] stringr_1.5.1       Biostrings_2.72.1   GenomeInfoDb_1.40.1 XVector_0.44.0     
 [5] IRanges_2.38.1      S4Vectors_0.42.1    BiocGenerics_0.50.0 default_1.0.0      
 [9] colorout_1.3-0.2    data.table_1.15.4   rlang_1.1.4        

loaded via a namespace (and not attached):
 [1] vctrs_0.6.5             crayon_1.5.3            httr_1.4.7             
 [4] cli_3.6.3               knitr_1.48              xfun_0.45              
 [7] stringi_1.8.4           UCSC.utils_1.0.0        jsonlite_1.8.8         
[10] glue_1.7.0              rprojroot_2.0.4         htmltools_0.5.8.1      
[13] rmarkdown_2.27          evaluate_0.24.0         fastmap_1.2.0          
[16] lifecycle_1.0.4         yaml_2.3.9              compiler_4.4.1         
[19] here_1.0.1              rstudioapi_0.16.0       digest_0.6.36          
[22] R6_2.5.1                GenomeInfoDbData_1.2.12 magrittr_2.0.3         
[25] tools_4.4.1             zlibbioc_1.50.0        
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-08-06
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────────
 ! package          * version date (UTC) lib source
   BiocGenerics     * 0.50.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings       * 2.72.1  2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
 P cli                3.6.3   2024-06-21 [2] CRAN (R 4.4.0)
   colorout         * 1.3-0.2 2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   crayon             1.5.3   2024-06-20 [1] CRAN (R 4.4.0)
   data.table       * 1.15.4  2024-03-30 [2] CRAN (R 4.4.0)
   default          * 1.0.0   2017-08-07 [1] CRAN (R 4.4.0)
   digest             0.6.36  2024-06-23 [1] CRAN (R 4.4.0)
   evaluate           0.24.0  2024-06-10 [1] CRAN (R 4.4.0)
   fastmap            1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
   GenomeInfoDb     * 1.40.1  2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData   1.2.12  2024-05-01 [1] Bioconductor
   glue               1.7.0   2024-01-09 [1] CRAN (R 4.4.0)
   here               1.0.1   2020-12-13 [1] CRAN (R 4.4.0)
   htmltools          0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
   httr               1.4.7   2023-08-15 [1] CRAN (R 4.4.0)
   IRanges          * 2.38.1  2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   jsonlite           1.8.8   2023-12-04 [1] CRAN (R 4.4.0)
   knitr              1.48    2024-07-07 [1] CRAN (R 4.4.1)
   lifecycle          1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
   magrittr           2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
   R6                 2.5.1   2021-08-19 [1] CRAN (R 4.4.0)
 P rlang            * 1.1.4   2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown          2.27    2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot          2.0.4   2023-11-05 [1] CRAN (R 4.4.0)
   rstudioapi         0.16.0  2024-03-24 [1] CRAN (R 4.4.0)
   S4Vectors        * 0.42.1  2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sessioninfo        1.2.2   2021-12-06 [1] CRAN (R 4.4.0)
   stringi            1.8.4   2024-05-06 [1] CRAN (R 4.4.0)
   stringr          * 1.5.1   2023-11-14 [1] CRAN (R 4.4.0)
   UCSC.utils         1.0.0   2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   vctrs              0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
   xfun               0.45    2024-06-16 [1] CRAN (R 4.4.0)
   XVector          * 0.44.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml               2.3.9   2024-07-05 [1] CRAN (R 4.4.0)
   zlibbioc           1.50.0  2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────────
