---
title: "Fig 6-Spat Reg and Yao23 Spat Reg Supp Tabs"
author: "Bernard Mulvey"
date: "2023-12-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(ggplot2)
library(gridExtra)
library(pheatmap)
require(colorout)
library(ggrepel)
library(ggtext)
library(Polychrome)
library(ggtext)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 10), axis.title.x = element_text(size = 11), axis.text.y = element_text(size = 10), axis.title.y = element_text(size=11), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=9), legend.title = element_text(size=10,hjust=0.5)))
```


### 6A: scCoco results in heatmap format
```{r}
cocores <- readRDS("data/10-Spatial Registration/01-scCoco query of top 150 markers per k15-20-31-collapsedK15 domain in ABA ms hyp CCF areas.RDS")

cocores <- cocores[[4]]
cocores <- as.data.table(cocores[["scores_per_target_level_region_all"]])

setnames(cocores,c(3:15),gsub(names(cocores)[3:15],pattern="k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_(X.*)",replacement="\\1"))
setnames(cocores,c("XVMH","XARC"),c("VMH","ARC"))
cocores <- cocores[,.(topnode,topname,VMH,ARC,X1, X2, X4, X6, X7, X8, X9, X10, X13, X14, X15)]

keepres <- c("")
i <- 3
for (i in c(3:15)){
    curvar <- names(cocores)[i]
    setorderv(cocores,curvar,-1)
    keepres <- c(keepres,cocores[c(1:3),topname])
}
keepres <- unique(keepres)[keepres!=""]
rm(i,curvar)

cocores <- melt(cocores[topname %in% keepres,],id.vars=c("topnode","topname"),variable.name="region",value.name="score")
setnames(cocores,"score","scCoco Score (top\n150 Visium Markers)")
cocores[topname=="Periventricular hypothalamic nucleus, posterior part",topname:="Posterior periventricular nu."]

pdf("plots/Manuscript/5A-scCoco registration to adult mouse ISH of hyp.pdf",width=7,height=4)
ggplot(cocores,aes(x=region,y=topname,fill=`scCoco Score (top\n150 Visium Markers)`))+
    geom_tile()+
    scale_fill_gradient2(low="white",mid="yellow",high="red",midpoint=0.5)+
    theme(axis.text.x = element_text(size = 9,angle = 90, hjust = 1))+
    scale_x_discrete(expand = c(0,0))+
    scale_y_discrete(expand=c(0,0))+
    ylab("Allen Mouse Atlas Region Name")+
    xlab("Visium Domain")+
    theme(axis.title.x = element_text(size = 10), axis.text.y = element_text(size = 9), axis.title.y = element_text(size=11),legend.title = element_text(size=10,hjust=0.5),legend.text=element_text(size=9))
dev.off()
```

# 5B: Spatial registration to Yao '23
```{r}
spatregs <- readRDS("data/10-Spatial Registration/02g-sphyp-k15-20-31-15clps_ABAms-47subclass-47withVMHARCclps.RDS")

jointrestabs <- list()
i<-1
for (i in c(1:length(spatregs))){
    j <- 1
    for (j in c(1:length(spatregs[[i]]))){
        newdat <- copy(spatregs[[i]][[j]])
        newdat[,ngenes:=as.numeric(gsub(names(spatregs[[i]])[j],pattern="^ABA.*_top_(.*)_sphyp.*$",replacement="\\1"))]
        newdat[,comparison:=gsub(names(spatregs[[i]])[j],pattern="^(.*)_top.*$",replacement="\\1")]
        if(j==1){
            curres <- copy(newdat)
        }
        else{curres <- rbind(curres,newdat)}
        rm(newdat)
    }
    jointrestabs[[i]] <- copy(curres)
    names(jointrestabs)[i] <- unique(curres$comparison)
    rm(j,curres)
}

### save these as Supp Tabs since we made them
# i<-1
# for (i in c(1:8)){
#     fwrite(jointrestabs[[i]],paste0("local_supptab_assembly/Yao23 Spatreg sheets/",names(jointrestabs)[i],".txt"),sep='\t',quote=F)
# }


### plots: VMH subdomains to VMH clusters / ARCsubdomains to ARC clusters spatreg values as a function of n genes
plotspatreg <- jointrestabs[["ABA23_HYP_subclass_to_sphypk15"]]
keepcols <- c("ngenes","sphyp_domain",grep(colnames(plotspatreg),pattern="VMH|ARH",value=T))
plotspatreg <- plotspatreg[sphyp_domain %in% c("VMH.1","VMH.2","ARC.1","ARC.2"),..keepcols]
rm(keepcols)

arccols <- c("ngenes","sphyp_domain",grep(colnames(plotspatreg),pattern="ARH",value=T))
arcplots <- plotspatreg[sphyp_domain %in% c("ARC.1","ARC.2"),..arccols]

vmhcols <- c("ngenes","sphyp_domain",grep(colnames(plotspatreg),pattern="VMH",value=T))
vmhplots <- plotspatreg[sphyp_domain %in% c("VMH.1","VMH.2"),..vmhcols]

### melt so that we can have colored lines for the subclasses vs. ea spatial domain --> 4x plots with 2-4 subclasses ea.
arcplots <- melt(arcplots,id.vars=c("ngenes","sphyp_domain"),variable.name="Subclass",value.name="Spearman Cor. (marker t stats)")
setnames(arcplots,"ngenes","Top n Human Domain Markers")
#arcplots <- arcplots[`Top n Human Domain Markers`<600]

vmhplots <- melt(vmhplots,id.vars=c("ngenes","sphyp_domain"),variable.name="Subclass",value.name="Spearman Cor. (marker t stats)")
setnames(vmhplots,"ngenes","Top n Human Domain Markers")
#vmhplots <- vmhplots[`Top n Human Domain Markers`<600]
```
## 5B continued: make a palette for 4 line colors that aren't egregiously colorblind-unfriendly. keep this for 5C.
```{r}
set.seed(82229)
pal <- Polychrome::createPalette(9,M = 100000,range = c(40,60),seedcolors = "#000000")
Polychrome::swatch(pal)

# k fine
dev.off()
pal <- pal[c(2:4,7)]
```

helper function to extract a legend and use as grob: from https://statisticsglobe.com/add-common-legend-to-combined-ggplot2-plots-in-r/#example-2-add-shared-legend-to-ggplot2-plots-using-gridextra-package 
```{r}
extract_legend <- function(my_ggp) {
  step1 <- ggplot_gtable(ggplot_build(my_ggp))
  step2 <- which(sapply(step1$grobs, function(x) x$name) == "guide-box")
  step3 <- step1$grobs[[step2]]
  return(step3)
}
```
## 5B continued: save plots
```{r}
arcpal <- copy(pal)
names(arcpal) <- unique(arcplots$Subclass)
### do not regenerate this for the sex de plot--otherwise they'll be in a different order
arcplt <- list()
arcplt[[1]] <- ggplot(arcplots[sphyp_domain=="ARC.1"],aes(x=log10(`Top n Human Domain Markers`),y=`Spearman Cor. (marker t stats)`,col=Subclass))+
    geom_line(size=0.7,alpha=0.7)+
    scale_color_manual(values=arcpal)+
    guides(color="none")+
    ggtitle("ARC.1")+
    ylim(-0.25,0.5)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_text(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))

arcplt[[2]] <- ggplot(arcplots[sphyp_domain=="ARC.2"],aes(x=log10(`Top n Human Domain Markers`),y=`Spearman Cor. (marker t stats)`,col=Subclass))+
    geom_line(size=0.7,alpha=0.7)+
    scale_color_manual(values=arcpal)+
    ggtitle("ARC.2")+
    ylim(-0.25,0.5)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_blank(),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10),legend.background = element_blank())

### extract legend to plot one time alongside the two plots:
leg.tmp <- extract_legend(arcplt[[2]])
arcplt[[2]] <- arcplt[[2]]+guides(color="none")
dev.off()

pdf("plots/Manuscript/5Bi-ARC.1, ARC.2 to ARH subclasses.pdf",height=3,width=3.5)
do.call("grid.arrange",list(arrangeGrob(grobs=arcplt,ncol=2,widths = c(1.875,1.75)),leg.tmp,nrow=2,heights=c(12,8)))
dev.off()



##### and VMH
vmhpal <- copy(pal[c(1,3)])
names(vmhpal) <- unique(vmhplots$Subclass)
vmhplt <- list()
vmhplt[[1]] <- ggplot(vmhplots[sphyp_domain=="VMH.1"],aes(x=log10(`Top n Human Domain Markers`),y=`Spearman Cor. (marker t stats)`,col=Subclass))+
    geom_line(size=0.7,alpha=0.7)+
    scale_color_manual(values=vmhpal)+
    guides(color="none")+
    ggtitle("VMH.1")+
    ylim(0.0,0.6)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_text(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))

vmhplt[[2]] <- ggplot(vmhplots[sphyp_domain=="VMH.2"],aes(x=log10(`Top n Human Domain Markers`),y=`Spearman Cor. (marker t stats)`,col=Subclass))+
    geom_line(size=0.7,alpha=0.7)+
    scale_color_manual(values=vmhpal)+
    ggtitle("VMH.2")+
    ylim(0.0,0.6)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_blank(),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10),legend.background = element_blank())

### extract legend to plot one time alongside the two plots:
leg.tmp <- extract_legend(vmhplt[[2]])
vmhplt[[2]] <- vmhplt[[2]]+guides(color="none")
dev.off()

pdf("plots/Manuscript/5Bii-VMH.1, VMH.2 to VMH subclasses.pdf",height=3,width=3.5)
do.call("grid.arrange",list(arrangeGrob(grobs=vmhplt,ncol=2,widths=c(1.875,1.75)),leg.tmp,nrow=2,heights=c(12,8)))
dev.off()
```

### 5C: plots as above, for VMH domain - VMH clusters / ARC domain - ARC clusters
```{r}
deregs <- readRDS("data/12-Mouse comparisons/03-Yao23-subclass-MvFsexDE-tstat_correls_to_k15-20-31-15collapsed_bytopNhumanDEGs.RDS")

plotdereg <- deregs[[1]]
plotdereg[human_domain=="X3",human_domain:="ARC.1"]
plotdereg[human_domain=="X5",human_domain:="ARC.2"]
plotdereg[human_domain=="X11",human_domain:="VMH.1"]
plotdereg[human_domain=="X12",human_domain:="VMH.2"]

### plots: VMH subdomains to VMH clusters / ARCsubdomains to ARC clusters spatreg values as a function of n genes
plotdereg <- plotdereg[mouse_subclass %in% grep(mouse_subclass,pattern="VMH|ARH",value=T)]

### set up names for ease of plot rendering
setnames(plotdereg,"top_n_human","Top n Human Sex DEGs")
setnames(plotdereg,"mouse_subclass","Subclass")

### get arc, vmh domain data
arcplots <- plotdereg[human_domain %in% c("ARC.1","ARC.2")&Subclass %in% grep(Subclass,pattern="ARH",value=T)]

vmhplots <- plotdereg[human_domain %in% c("VMH.1","VMH.2")&Subclass %in% grep(Subclass,pattern="VMH",value=T)]
```

## 5C continued: save plots
```{r}
### using the previously defined arc-type palette for consistency
arcplots[,Subclass:=factor(Subclass,levels=names(arcpal))]
arcplt <- list()

arcplt[[1]] <- ggplot(arcplots[human_domain=="ARC.1"],aes(x=log10(`Top n Human Sex DEGs`),y=spearmancor,col=Subclass))+
    geom_line(linewidth=0.7,alpha=0.7)+
    scale_color_manual(values=arcpal)+
    guides(color="none")+
    ggtitle("ARC.1")+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.6,0.1)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))

arcplt[[2]] <- ggplot(arcplots[human_domain=="ARC.2"],aes(x=log10(`Top n Human Sex DEGs`),y=spearmancor,col=Subclass))+
    geom_line(linewidth=0.7,alpha=0.7)+
    scale_color_manual(values=arcpal)+
    ggtitle("ARC.2")+
    ylim(-0.6,0.1)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_blank(),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10),legend.background = element_blank())

### extract legend to plot one time alongside the two plots:
leg.tmp <- extract_legend(arcplt[[2]])
arcplt[[2]] <- arcplt[[2]]+guides(color="none")
dev.off()

pdf("plots/Manuscript/5Ci-Sex DE correls-ARC.1, ARC.2 to ARH subclasses.pdf",height=3,width=3.5)
do.call("grid.arrange",list(arrangeGrob(grobs=arcplt,ncol=2,widths=c(2.025,1.75)),leg.tmp,nrow=2,heights=c(12,8)))
dev.off()



##### and VMH
vmhpal <- copy(pal[c(1,3)])
names(vmhpal) <- unique(vmhplots$Subclass)
vmhplt <- list()
vmhplt[[1]] <- ggplot(vmhplots[human_domain=="VMH.1"],aes(x=log10(`Top n Human Sex DEGs`),y=spearmancor,col=Subclass))+
    geom_line(size=0.7,alpha=0.7)+
    scale_color_manual(values=vmhpal)+
    guides(color="none")+
    ggtitle("VMH.1")+
    ylab("Spearman Correlation (Human-<br>Mouse DE *t* stats)")+
    ylim(-0.6,0.1)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_markdown(size=9),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10))

vmhplt[[2]] <- ggplot(vmhplots[human_domain=="VMH.2"],aes(x=log10(`Top n Human Sex DEGs`),y=spearmancor,col=Subclass))+
    geom_line(size=0.7,alpha=0.7)+
    scale_color_manual(values=vmhpal)+
    ggtitle("VMH.2")+
    ylim(-0.6,0.1)+
    theme(axis.title.x = element_blank(), axis.text.y = element_text(size = 8),axis.text.x=element_text(size=8,angle=90,vjust=1,hjust=0.5), axis.title.y = element_blank(),legend.title = element_text(size=8,hjust=0.5),legend.text=element_text(size=8),plot.title = element_text(size=10),legend.background = element_blank())

### extract legend to plot one time alongside the two plots:
leg.tmp <- extract_legend(vmhplt[[2]])
vmhplt[[2]] <- vmhplt[[2]]+guides(color="none")
dev.off()

pdf("plots/Manuscript/5Cii-Sex DE correls-VMH.1, VMH.2 to VMH subclasses.pdf",height=3,width=3.5)
do.call("grid.arrange",list(arrangeGrob(grobs=vmhplt,ncol=2,widths=c(2.05,1.75)),leg.tmp,nrow=2,heights=c(12,8)))
dev.off()
```


### old 5C: scatter plot per Visium cluster, with each dot representing a Yao23 subclass, using all genes
```{r}
# correls <- readRDS("data/12-Mouse comparisons/03-Yao23-subclass-MvFsexDE-tstat_correls_to_k15-20-31-15collapsed.RDS")
# 
# correls[[1]][human_domain=="X11",human_domain:="VMH.1"]
# correls[[1]][human_domain=="X12",human_domain:="VMH.2"]
# correls[[1]][human_domain=="X3",human_domain:="ARC.1"]
# correls[[1]][human_domain=="X5",human_domain:="ARC.2"]
# correls[[1]][human_domain=="X1",human_domain:="WM.1"]
# correls[[1]][human_domain=="X9",human_domain:="WM.2"]
# 
# 
# correls <- correls[[1]][human_domain %in% c("VMH.1","VMH.2","ARC.1","ARC.2","WM.1","WM.2")]
# ### make cell type names more point-label friendly for the cell types that will get labeled: 214, 047, 054, 088, 096, 250, 16
# correls[mouse_subclass == "x214_LGv_SPFp_SPFm_Gata3_Gaba", 
#         mouse_subclass := "x214_LGv_SPFp_\nSPFm_Gata3_Gaba"]
# 
# correls[mouse_subclass == "x080_GPi_Skor1_Gaba_Glut", 
#         mouse_subclass := "x080_GPi_Skor1_\nGaba_Glut"]
# 
# correls[mouse_subclass == "x116_PVpo_VMPO_MPN_Hmx2_Gaba", 
#         mouse_subclass := "x116_PVpo_VMPO_\nMPN_Hmx2_Gaba"]
# 
# ### assign VMH color to the VMH subclasses and ARC to ARC
# correls[mouse_subclass %in% grep(mouse_subclass,pattern="VMH",value=T),label:="VMH"]
# correls[mouse_subclass %in% grep(mouse_subclass,pattern="ARH",value=T),label:="ARC"]
# correls[!(label %in% c("ARC","VMH")),label:="Other"]
# 
# pal <- c("VMH"="#5dd959","ARC"="#8c63cf","Other"="#000000")
# 
# ### need to change something about the invisible bounds around the ggrepel text overexpanding the x-axis?
# 
# ## loop through the six human domains and write the corresponding plot
# i <- 1
# for (i in c(1:length(unique(correls$human_domain)))) {
#     curdat <-
#         copy(correls[human_domain == unique(correls$human_domain)[i]])
#     setorderv(curdat, c("human_domain", "spearmancor"))
#     curdat[, mouse_subclass := factor(mouse_subclass, levels = as.character(curdat$mouse_subclass))] # factor levels should now match the row order of the data table
#     curdat[, subcl_label := as.character(NA)]
#     curdat[c(1, nrow(curdat)), subcl_label := mouse_subclass]
#     
#     ## create a numerical x coordinate system to work with ggrepel more easily (since we're not using axis labels, doesn't matter what's actually composing the x axis udner the hood)
#     curdat[, xcoord := c(1:nrow(curdat))]
#     
#     plts <- ggplot(curdat, aes(x = xcoord, y = spearmancor)) +
#         geom_point(aes(col=label)) +
#         scale_color_manual(values=pal) +
#         ylim(-0.5, 0.5) +
#         guides(col="none") +
#         geom_text_repel(
#             aes(label = ifelse(is.na(subcl_label), NA, subcl_label)),
#             nudge_x = -35 * sign(curdat$spearmancor),
#             nudge_y = 0.1 * sign(curdat$spearmancor),
#             size = 2,
#             box.padding = 0,
#             point.padding = 0,
#             min.segment.length = 0
#         ) +
#         xlim(1, nrow(curdat)) +
#         # can just place axis titles centered along all these panels in illustrator.
#         #xlab("Yao '23 Mouse HYP Subclass")+
#         #ylab("Spearman Correlation (Mouse\nvs Human Sex DE t)")+
#         ggtitle(unique(correls$human_domain)[i]) +
#         theme(
#             axis.text.x = element_blank(),
#             axis.text.y = element_text(size = 9),
#             axis.title.x = element_blank(),
#             axis.title.y = element_blank(),
#             plot.title = element_text(size = 11),
#             axis.ticks.length.x = unit(0, "pt"),
#             panel.grid.major.x = element_blank(),
#             panel.grid.minor.x = element_blank()
#         )
#     
#     ### save
#     pdf(
#         paste0(
#             "plots/Manuscript/5C-YaoSexDE_spearcorrels_",
#             unique(correls$human_domain)[i],
#             ".pdf"
#         ),
#         height = 2.3,
#         width = 2.3
#     )
#     print(plts)
#     dev.off()
#     
#     ### extract a legend from one of the plots
#     
#     if (i == length(unique(correls$human_domain))) {
#         plts <- ggplot(curdat, aes(x = xcoord, y = spearmancor)) +
#                 geom_point(aes(col=label)) +
#                 scale_color_manual(values=pal) +
#                 ylim(-0.5, 0.5) +
#                 labs(col="Mouse\nSubclass")
#                 theme(legend.title=element_text(size=9,hjust=0.5),legend.text=element_text(size=8),legend.margin = margin(0,0,0,0,unit="in"))
#         
#         leg <- extract_legend(plts)
#         dev.off()
#         
#         pdf(
#             "plots/Manuscript/5C-YaoSexDE_spearcorrels_sharedlegend.pdf",
#             height = 1.4,
#             width = 1
#         )
#         do.call("grid.arrange",list(leg))
#         dev.off()
#     }
#     ####
#     
#     rm(plts,curdat)
# }
```


### OLD 6B (Spearman) and 6C or Supp Fig (Pearson): sex DE logFC correlations between mouse genes sex DE for each subclass in HYP dissections from Yao 23 and human orthologs in each domain
```{r}
# correls <- readRDS("data/12-Mouse comparisons/03-Yao23-subclass-MvFsexDE-tstat_correls_to_k15-20-31-15collapsed.RDS")
# 
# ## 
# correls[[1]][human_domain=="X11",human_domain:="VMH.1"]
# correls[[1]][human_domain=="X12",human_domain:="VMH.2"]
# correls[[1]][human_domain=="X3",human_domain:="ARC.1"]
# correls[[1]][human_domain=="X5",human_domain:="ARC.2"]
# correls[[1]][human_domain=="X1",human_domain:="WM.1"]
# correls[[1]][human_domain=="X9",human_domain:="WM.2"]
# 
# correls <- correls[human_domain %in% c("VMH.1","VMH.2","ARC.1","ARC.2","WM.1","WM.2")]
# 
# pccs <- dcast(correls, mouse_subclass ~ human_domain, value.var = "pcc")
# spears <- dcast(correls, mouse_subclass ~ human_domain, value.var="spearmancor")
# 
# pccs <- as.data.frame(pccs)
# rownames(pccs) <- pccs$mouse_subclass
# pccs$mouse_subclass <- NULL
# 
# spears <- as.data.frame(spears)
# rownames(spears) <- spears$mouse_subclass
# spears$mouse_subclass <- NULL
# 
# pdf("plots/Manuscript/6C-Yaosubclass_sexDE_vs_visium_Pearsoncors.pdf",height=6,width=3.5)
# pheatmap(as.matrix(pccs),treeheight_col=0,treeheight_row = 0)
# dev.off()
# 
# 
# pdf("plots/Manuscript/6B-Yaosubclass_sexDE_vs_visium_spearmancors.pdf",height=10,width=7)
# pheatmap(as.matrix(spears),treeheight_col=0,treeheight_row = 0)
# dev.off()
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6.4

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] default_1.0.0       Polychrome_1.5.1    ggtext_0.1.2       
 [4] ggrepel_0.9.5       colorout_1.3-0.1    pheatmap_1.0.12    
 [7] gridExtra_2.3       ggplot2_3.4.4       Biostrings_2.70.2  
[10] GenomeInfoDb_1.38.5 XVector_0.42.0      IRanges_2.36.0     
[13] S4Vectors_0.40.2    BiocGenerics_0.48.1 data.table_1.14.10 
[16] rlang_1.1.3        

loaded via a namespace (and not attached):
  [1] later_1.3.2                   BiocIO_1.12.0                
  [3] bitops_1.0-7                  filelock_1.0.3               
  [5] fields_15.2                   tibble_3.2.1                 
  [7] XML_3.99-0.16.1               lifecycle_1.0.4              
  [9] edgeR_4.0.14                  doParallel_1.0.17            
 [11] rprojroot_2.0.4               lattice_0.22-5               
 [13] magrittr_2.0.3                limma_3.58.1                 
 [15] plotly_4.10.4                 sass_0.4.8                   
 [17] rmarkdown_2.25                jquerylib_0.1.4              
 [19] yaml_2.3.8                    httpuv_1.6.13                
 [21] spam_2.10-0                   sessioninfo_1.2.2            
 [23] cowplot_1.1.3                 DBI_1.2.1                    
 [25] RColorBrewer_1.1-3            golem_0.4.1                  
 [27] maps_3.4.2                    abind_1.4-5                  
 [29] zlibbioc_1.48.0               GenomicRanges_1.54.1         
 [31] purrr_1.0.2                   RCurl_1.98-1.14              
 [33] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      
 [35] irlba_2.3.5.1                 commonmark_1.9.1             
 [37] DelayedMatrixStats_1.24.0     codetools_0.2-19             
 [39] DelayedArray_0.28.0           xml2_1.3.6                   
 [41] DT_0.31                       scuttle_1.12.0               
 [43] tidyselect_1.2.0              farver_2.1.1                 
 [45] ScaledMatrix_1.10.0           viridis_0.6.5                
 [47] shinyWidgets_0.8.1            matrixStats_1.2.0            
 [49] BiocFileCache_2.10.1          GenomicAlignments_1.38.2     
 [51] jsonlite_1.8.8                BiocNeighbors_1.20.2         
 [53] ellipsis_0.3.2                scater_1.30.1                
 [55] iterators_1.0.14              foreach_1.5.2                
 [57] tools_4.3.2                   Rcpp_1.0.12                  
 [59] glue_1.7.0                    SparseArray_1.2.3            
 [61] xfun_0.41                     here_1.0.1                   
 [63] MatrixGenerics_1.14.0         dplyr_1.1.4                  
 [65] withr_3.0.0                   BiocManager_1.30.22          
 [67] fastmap_1.1.1                 fansi_1.0.6                  
 [69] digest_0.6.34                 rsvd_1.0.5                   
 [71] R6_2.5.1                      mime_0.12                    
 [73] colorspace_2.1-0              markdown_1.12                
 [75] RSQLite_2.3.5                 config_0.3.2                 
 [77] utf8_1.2.4                    tidyr_1.3.1                  
 [79] generics_0.1.3                rtracklayer_1.62.0           
 [81] httr_1.4.7                    htmlwidgets_1.6.4            
 [83] S4Arrays_1.2.0                scatterplot3d_0.3-44         
 [85] pkgconfig_2.0.3               gtable_0.3.4                 
 [87] blob_1.2.4                    SingleCellExperiment_1.24.0  
 [89] htmltools_0.5.7               dotCall64_1.1-1              
 [91] scales_1.3.0                  Biobase_2.62.0               
 [93] png_0.1-8                     attempt_0.3.1                
 [95] SpatialExperiment_1.12.0      knitr_1.45                   
 [97] rstudioapi_0.15.0             rjson_0.2.21                 
 [99] curl_5.2.0                    cachem_1.0.8                 
[101] stringr_1.5.1                 BiocVersion_3.18.1           
[103] parallel_4.3.2                vipor_0.4.7                  
[105] AnnotationDbi_1.64.1          restfulr_0.0.15              
[107] pillar_1.9.0                  grid_4.3.2                   
[109] vctrs_0.6.5                   promises_1.2.1               
[111] BiocSingular_1.18.0           dbplyr_2.4.0                 
[113] beachmat_2.18.0               xtable_1.8-4                 
[115] beeswarm_0.4.0                paletteer_1.6.0              
[117] evaluate_0.23                 magick_2.8.2                 
[119] cli_3.6.2                     locfit_1.5-9.8               
[121] compiler_4.3.2                Rsamtools_2.18.0             
[123] crayon_1.5.2                  labeling_0.4.3               
[125] rematch2_2.1.2                ggbeeswarm_0.7.2             
[127] stringi_1.8.3                 viridisLite_0.4.2            
[129] BiocParallel_1.36.0           munsell_0.5.0                
[131] lazyeval_0.2.2                Matrix_1.6-5                 
[133] ExperimentHub_2.10.0          benchmarkme_1.0.8            
[135] sparseMatrixStats_1.14.0      bit64_4.0.5                  
[137] KEGGREST_1.42.0               statmod_1.5.0                
[139] shiny_1.8.0                   SummarizedExperiment_1.32.0  
[141] interactiveDisplayBase_1.40.0 AnnotationHub_3.10.0         
[143] gridtext_0.1.5                memoise_2.0.1                
[145] bslib_0.6.1                   benchmarkmeData_1.0.4        
[147] bit_4.0.5                     spatialLIBD_1.14.1           
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6.4
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2024-02-19
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────
 package                * version     date (UTC) lib source
 abind                    1.4-5       2016-07-21 [1] CRAN (R 4.3.0)
 AnnotationDbi            1.64.1      2023-11-02 [1] Bioconductor
 AnnotationHub            3.10.0      2023-10-24 [1] Bioconductor
 attempt                  0.3.1       2020-05-03 [1] CRAN (R 4.3.0)
 beachmat                 2.18.0      2023-10-24 [1] Bioconductor
 beeswarm                 0.4.0       2021-06-01 [1] CRAN (R 4.3.2)
 benchmarkme              1.0.8       2022-06-12 [1] CRAN (R 4.3.0)
 benchmarkmeData          1.0.4       2020-04-23 [1] CRAN (R 4.3.0)
 Biobase                  2.62.0      2023-10-24 [1] Bioconductor
 BiocFileCache            2.10.1      2023-10-26 [1] Bioconductor
 BiocGenerics           * 0.48.1      2023-11-02 [1] Bioconductor
 BiocIO                   1.12.0      2023-10-26 [1] Bioconductor
 BiocManager              1.30.22     2023-08-08 [1] CRAN (R 4.3.0)
 BiocNeighbors            1.20.2      2024-01-07 [1] Bioconductor 3.18 (R 4.3.2)
 BiocParallel             1.36.0      2023-10-26 [1] Bioconductor
 BiocSingular             1.18.0      2023-10-24 [1] Bioconductor
 BiocVersion              3.18.1      2023-11-18 [1] Bioconductor 3.18 (R 4.3.2)
 Biostrings             * 2.70.2      2024-01-28 [1] Bioconductor 3.18 (R 4.3.2)
 bit                      4.0.5       2022-11-15 [1] CRAN (R 4.3.2)
 bit64                    4.0.5       2020-08-30 [1] CRAN (R 4.3.2)
 bitops                   1.0-7       2021-04-24 [1] CRAN (R 4.3.2)
 blob                     1.2.4       2023-03-17 [1] CRAN (R 4.3.0)
 bslib                    0.6.1       2023-11-28 [1] CRAN (R 4.3.1)
 cachem                   1.0.8       2023-05-01 [1] CRAN (R 4.3.2)
 cli                      3.6.2       2023-12-11 [1] CRAN (R 4.3.2)
 codetools                0.2-19      2023-02-01 [1] CRAN (R 4.3.0)
 colorout               * 1.3-0.1     2024-01-11 [1] local
 colorspace               2.1-0       2023-01-23 [1] CRAN (R 4.3.2)
 commonmark               1.9.1       2024-01-30 [1] CRAN (R 4.3.2)
 config                   0.3.2       2023-08-30 [1] CRAN (R 4.3.0)
 cowplot                  1.1.3       2024-01-22 [1] CRAN (R 4.3.1)
 crayon                   1.5.2       2022-09-29 [1] CRAN (R 4.3.0)
 curl                     5.2.0       2023-12-08 [1] CRAN (R 4.3.2)
 data.table             * 1.14.10     2023-12-08 [1] CRAN (R 4.3.2)
 DBI                      1.2.1       2024-01-12 [1] CRAN (R 4.3.2)
 dbplyr                   2.4.0       2023-10-26 [1] CRAN (R 4.3.1)
 default                * 1.0.0       2017-08-07 [1] CRAN (R 4.3.0)
 DelayedArray             0.28.0      2023-10-24 [1] Bioconductor
 DelayedMatrixStats       1.24.0      2023-11-06 [1] Bioconductor
 digest                   0.6.34      2024-01-11 [1] CRAN (R 4.3.2)
 doParallel               1.0.17      2022-02-07 [1] CRAN (R 4.3.0)
 dotCall64                1.1-1       2023-11-28 [1] CRAN (R 4.3.2)
 dplyr                    1.1.4       2023-11-17 [1] CRAN (R 4.3.2)
 DT                       0.31        2023-12-09 [1] CRAN (R 4.3.1)
 edgeR                    4.0.14      2024-01-29 [1] Bioconductor 3.18 (R 4.3.2)
 ellipsis                 0.3.2       2021-04-29 [1] CRAN (R 4.3.2)
 evaluate                 0.23        2023-11-01 [1] CRAN (R 4.3.1)
 ExperimentHub            2.10.0      2023-10-26 [1] Bioconductor
 fansi                    1.0.6       2023-12-08 [1] CRAN (R 4.3.2)
 farver                   2.1.1       2022-07-06 [1] CRAN (R 4.3.2)
 fastmap                  1.1.1       2023-02-24 [1] CRAN (R 4.3.2)
 fields                   15.2        2023-08-17 [1] CRAN (R 4.3.2)
 filelock                 1.0.3       2023-12-11 [1] CRAN (R 4.3.2)
 foreach                  1.5.2       2022-02-02 [1] CRAN (R 4.3.0)
 generics                 0.1.3       2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb           * 1.38.5      2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData         1.2.11      2024-01-11 [1] Bioconductor
 GenomicAlignments        1.38.2      2024-01-16 [1] Bioconductor 3.18 (R 4.3.2)
 GenomicRanges            1.54.1      2023-10-29 [1] Bioconductor
 ggbeeswarm               0.7.2       2023-04-29 [1] CRAN (R 4.3.0)
 ggplot2                * 3.4.4       2023-10-12 [1] CRAN (R 4.3.1)
 ggrepel                * 0.9.5       2024-01-10 [1] CRAN (R 4.3.2)
 ggtext                 * 0.1.2       2022-09-16 [1] CRAN (R 4.3.0)
 glue                     1.7.0       2024-01-09 [1] CRAN (R 4.3.2)
 golem                    0.4.1       2023-06-05 [1] CRAN (R 4.3.0)
 gridExtra              * 2.3         2017-09-09 [1] CRAN (R 4.3.0)
 gridtext                 0.1.5       2022-09-16 [1] CRAN (R 4.3.2)
 gtable                   0.3.4       2023-08-21 [1] CRAN (R 4.3.0)
 here                     1.0.1       2020-12-13 [1] CRAN (R 4.3.0)
 htmltools                0.5.7       2023-11-03 [1] CRAN (R 4.3.2)
 htmlwidgets              1.6.4       2023-12-06 [1] CRAN (R 4.3.1)
 httpuv                   1.6.13      2023-12-06 [1] CRAN (R 4.3.1)
 httr                     1.4.7       2023-08-15 [1] CRAN (R 4.3.0)
 interactiveDisplayBase   1.40.0      2023-10-26 [1] Bioconductor
 IRanges                * 2.36.0      2023-10-24 [1] Bioconductor
 irlba                    2.3.5.1     2022-10-03 [1] CRAN (R 4.3.2)
 iterators                1.0.14      2022-02-05 [1] CRAN (R 4.3.0)
 jquerylib                0.1.4       2021-04-26 [1] CRAN (R 4.3.0)
 jsonlite                 1.8.8       2023-12-04 [1] CRAN (R 4.3.2)
 KEGGREST                 1.42.0      2023-10-26 [1] Bioconductor
 knitr                    1.45        2023-10-30 [1] CRAN (R 4.3.2)
 labeling                 0.4.3       2023-08-29 [1] CRAN (R 4.3.0)
 later                    1.3.2       2023-12-06 [1] CRAN (R 4.3.2)
 lattice                  0.22-5      2023-10-24 [1] CRAN (R 4.3.2)
 lazyeval                 0.2.2       2019-03-15 [1] CRAN (R 4.3.2)
 lifecycle                1.0.4       2023-11-07 [1] CRAN (R 4.3.1)
 limma                    3.58.1      2023-10-31 [1] Bioconductor
 locfit                   1.5-9.8     2023-06-11 [1] CRAN (R 4.3.2)
 magick                   2.8.2       2023-12-20 [1] CRAN (R 4.3.1)
 magrittr                 2.0.3       2022-03-30 [1] CRAN (R 4.3.2)
 maps                     3.4.2       2023-12-15 [1] CRAN (R 4.3.1)
 markdown                 1.12        2023-12-06 [1] CRAN (R 4.3.1)
 Matrix                   1.6-5       2024-01-11 [1] CRAN (R 4.3.2)
 MatrixGenerics           1.14.0      2023-10-26 [1] Bioconductor
 matrixStats              1.2.0       2023-12-11 [1] CRAN (R 4.3.2)
 memoise                  2.0.1       2021-11-26 [1] CRAN (R 4.3.0)
 mime                     0.12        2021-09-28 [1] CRAN (R 4.3.2)
 munsell                  0.5.0       2018-06-12 [1] CRAN (R 4.3.0)
 paletteer                1.6.0       2024-01-21 [1] CRAN (R 4.3.1)
 pheatmap               * 1.0.12      2019-01-04 [1] CRAN (R 4.3.0)
 pillar                   1.9.0       2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig                2.0.3       2019-09-22 [1] CRAN (R 4.3.0)
 plotly                   4.10.4      2024-01-13 [1] CRAN (R 4.3.1)
 png                      0.1-8       2022-11-29 [1] CRAN (R 4.3.2)
 Polychrome             * 1.5.1       2022-05-03 [1] CRAN (R 4.3.0)
 promises                 1.2.1       2023-08-10 [1] CRAN (R 4.3.2)
 purrr                    1.0.2       2023-08-10 [1] CRAN (R 4.3.2)
 R6                       2.5.1       2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs                 0.3.3       2021-01-31 [1] CRAN (R 4.3.2)
 RColorBrewer             1.1-3       2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp                     1.0.12      2024-01-09 [1] CRAN (R 4.3.2)
 RCurl                    1.98-1.14   2024-01-09 [1] CRAN (R 4.3.1)
 rematch2                 2.1.2       2020-05-01 [1] CRAN (R 4.3.0)
 restfulr                 0.0.15      2022-06-16 [1] CRAN (R 4.3.2)
 rjson                    0.2.21      2022-01-09 [1] CRAN (R 4.3.2)
 rlang                  * 1.1.3       2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown                2.25        2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot                2.0.4       2023-11-05 [1] CRAN (R 4.3.1)
 Rsamtools                2.18.0      2023-10-24 [1] Bioconductor
 RSQLite                  2.3.5       2024-01-21 [1] CRAN (R 4.3.2)
 rstudioapi               0.15.0      2023-07-07 [1] CRAN (R 4.3.0)
 rsvd                     1.0.5       2021-04-16 [1] CRAN (R 4.3.0)
 rtracklayer              1.62.0      2023-10-26 [1] Bioconductor
 S4Arrays                 1.2.0       2023-10-24 [1] Bioconductor
 S4Vectors              * 0.40.2      2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 sass                     0.4.8       2023-12-06 [1] CRAN (R 4.3.1)
 ScaledMatrix             1.10.0      2023-11-06 [1] Bioconductor
 scales                   1.3.0       2023-11-28 [1] CRAN (R 4.3.2)
 scater                   1.30.1      2023-11-16 [1] Bioconductor
 scatterplot3d            0.3-44      2023-05-05 [1] CRAN (R 4.3.0)
 scuttle                  1.12.0      2023-10-24 [1] Bioconductor
 sessioninfo              1.2.2       2021-12-06 [1] CRAN (R 4.3.0)
 shiny                    1.8.0       2023-11-17 [1] CRAN (R 4.3.1)
 shinyWidgets             0.8.1       2024-01-10 [1] CRAN (R 4.3.1)
 SingleCellExperiment     1.24.0      2023-11-06 [1] Bioconductor
 spam                     2.10-0      2023-10-23 [1] CRAN (R 4.3.2)
 SparseArray              1.2.3       2023-12-25 [1] Bioconductor 3.18 (R 4.3.2)
 sparseMatrixStats        1.14.0      2023-10-24 [1] Bioconductor
 SpatialExperiment        1.12.0      2023-10-26 [1] Bioconductor
 spatialLIBD              1.14.1      2023-11-30 [1] Bioconductor 3.18 (R 4.3.2)
 statmod                  1.5.0       2023-01-06 [1] CRAN (R 4.3.2)
 stringi                  1.8.3       2023-12-11 [1] CRAN (R 4.3.1)
 stringr                  1.5.1       2023-11-14 [1] CRAN (R 4.3.1)
 SummarizedExperiment     1.32.0      2023-11-06 [1] Bioconductor
 tibble                   3.2.1       2023-03-20 [1] CRAN (R 4.3.2)
 tidyr                    1.3.1       2024-01-24 [1] CRAN (R 4.3.1)
 tidyselect               1.2.0       2022-10-10 [1] CRAN (R 4.3.0)
 utf8                     1.2.4       2023-10-22 [1] CRAN (R 4.3.2)
 vctrs                    0.6.5       2023-12-01 [1] CRAN (R 4.3.2)
 vipor                    0.4.7       2023-12-18 [1] CRAN (R 4.3.1)
 viridis                  0.6.5       2024-01-29 [1] CRAN (R 4.3.2)
 viridisLite              0.4.2       2023-05-02 [1] CRAN (R 4.3.0)
 withr                    3.0.0       2024-01-16 [1] CRAN (R 4.3.1)
 xfun                     0.41        2023-11-01 [1] CRAN (R 4.3.2)
 XML                      3.99-0.16.1 2024-01-22 [1] CRAN (R 4.3.1)
 xml2                     1.3.6       2023-12-04 [1] CRAN (R 4.3.2)
 xtable                   1.8-4       2019-04-21 [1] CRAN (R 4.3.0)
 XVector                * 0.42.0      2023-10-24 [1] Bioconductor
 yaml                     2.3.8       2023-12-11 [1] CRAN (R 4.3.2)
 zlibbioc                 1.48.0      2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────
