---
title: "Fig 5-Sex DEG Enrichments"
author: "Bernard Mulvey"
date: "2023-12-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(ggplot2)
library(gridExtra)
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))
```

define a consistent sex•region color palette to use across plots
```{r}
sexregpal <- c("VMH Fem. Up"="#eb69ff","VMH Male Up"="#5d186b","ARC Fem. Up"="#69ff63","ARC Male Up"="#215220")
```

Sex DE GSEA
```{r}
sexreslist <- readRDS("data/11-GSEA/03a-Sex DE msigdb_1,2,3,5,6,8 GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")

vmhs <- sexreslist[["BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_VMH"]][pathway %in% c("HP_ABNORMAL_AGGRESSIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR", "GOMF_CHROMATIN_BINDING","HP_ABNORMALITY_OF_ACID_BASE_HOMEOSTASIS", "YAO_TEMPORAL_RESPONSE_TO_PROGESTERONE_CLUSTER_17","GABRIELY_MIR21_TARGETS", "MIR30A_3P","MCLACHLAN_DENTAL_CARIES_UP", "FAN_OVARY_CL6_PUTATIVE_EARLY_ATRETIC_FOLLICLE_THECAL_CELL_2")]

arcs <- sexreslist[["BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_ARC"]][pathway %in% c("HP_ABNORMAL_AGGRESSIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR", "GOMF_CHROMATIN_BINDING","HP_ABNORMALITY_OF_ACID_BASE_HOMEOSTASIS", "YAO_TEMPORAL_RESPONSE_TO_PROGESTERONE_CLUSTER_17","GABRIELY_MIR21_TARGETS", "MIR30A_3P","MCLACHLAN_DENTAL_CARIES_UP", "FAN_OVARY_CL6_PUTATIVE_EARLY_ATRETIC_FOLLICLE_THECAL_CELL_2")]

pltdat <- rbind(vmhs,arcs)
pltdat[,spdom:=gsub(spdom,pattern="BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_(.*)",replacement="\\1")]
pltdat[spdom=="VMH",sexreg:=ifelse(NES>0,yes="VMH Male Up",no="VMH Fem. Up")]
pltdat[spdom=="ARC",sexreg:=ifelse(NES>0,yes="ARC Male Up",no="ARC Fem. Up")]
rm(vmhs,arcs)
### extract pathway source db

i<-1
for (i in c(1:nrow(pltdat))){
    if(i==1){gsrc <- unlist(strsplit(pltdat[i,pathway],"_"))[1]}
    else{gsrc <- c(gsrc,
                    unlist(strsplit(pltdat[i,pathway],"_"))[1])}
}

## tidy gene sources; will note publications for published sets in legend.
pltdat$genesource <- gsrc
pltdat[genesource %in% c("FAN","GABRIELY","MCLACHLAN","YAO"),genesource:="MSigDB Literature-Curated"]
pltdat[genesource=="GOMF",genesource:="GO Mol. Fxn"]
pltdat[genesource=="HP",genesource:="Human Phenotypes"]
pltdat[genesource=="MIR30A",genesource:="MSigDB miRNA Targets"]


rm(i,gsrc)
### simplify pathways
pltdat[pathway=="HP_ABNORMAL_AGGRESSIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR",pathway:="Aggressive/impulsive behavior"]
pltdat[pathway=="GOMF_CHROMATIN_BINDING",pathway:="Chromatin binding"]
pltdat[pathway=="HP_ABNORMALITY_OF_ACID_BASE_HOMEOSTASIS",pathway:="Acid-base abnormality"]
pltdat[pathway=="YAO_TEMPORAL_RESPONSE_TO_PROGESTERONE_CLUSTER_17",pathway:="Temporal progesterone response"]
pltdat[pathway=="GABRIELY_MIR21_TARGETS",pathway:="miR-21 targets"]
pltdat[pathway=="MIR30A_3P",pathway:="miR-30a_3P targets"]
pltdat[pathway=="MCLACHLAN_DENTAL_CARIES_UP",pathway:="Dental caries"]
pltdat[pathway=="FAN_OVARY_CL6_PUTATIVE_EARLY_ATRETIC_FOLLICLE_THECAL_CELL_2",pathway:="Thecal cell (early atretic ovarian follicle)"]

### flip sign on negative NES scores
pltdat[NES<0,NES:=-1*NES]

pdf("plots/Manuscript/5A as 4K-Sex DEG msigdb GSEA.pdf",height=4,width=3)
ggplot(pltdat,aes(x=pathway,y=-log10(padj)))+
    geom_point(aes(col=sexreg,size=NES))+#,shape=genesource))+
    xlab("Term")+
    #Amplify the scaling by NES by 2x for better visualization
    scale_radius(range = c(1.25*min(pltdat$NES),1.25*max(pltdat$NES)),trans = "identity",guide="none")+
    scale_color_manual(values=sexregpal)+
    ylab("-log10 (FDR)")+
    labs(col="DE Gene Set")+
    geom_hline(yintercept=-log10(0.05),linetype="dashed")+
    #scale_shape_manual(values=c("MSigDB Literature-Curated"=16,"GO Mol. Fxn"=15,"Human Phenotypes"=18,"MSigDB miRNA Targets"=17))+
    guides(col=guide_legend(byrow=T))+
    # ^ necess. to get spacing to work in theme
    guides(col=guide_legend(override.aes=list(size=2.5)),
           #shape=guide_legend(override.aes=list(size = 2.5)),
           )+
    # ^ make legend points bigger
    theme(axis.text.y = element_text(size = 9),axis.title.y=element_text(size=10), axis.title.x = element_text(size = 10,margin=margin(-0.15,0,0,0,unit="in")), axis.text.x = element_text(size = 9,angle=90,hjust=1,vjust=0.5,margin=margin(0,0.1,0,0.1,"in")), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5,margin=margin(0,0,-0.1,0,"in")),legend.spacing.y = unit(0.125, "in"),legend.margin = margin(0,0,0,-0.1,unit = "in"))
dev.off()

rm(pltdat,sexreslist)
```

5B-sex de GSEA - TFs

## VMH-male: PMC10307746-12672_2023_738_MOESM12_ESM.docx-0-Supplementary_Table_S3_The_Top_100_MTF1-correlated_Genes_That_Obtained_From_The_GEPIA2_0_Database_MTF1_Rummagene_transcription_factors, 	
PMC7162552-pcbi.1007800.s006.txt--TFEA_TFE3_Rummagene_transcription_factors,SOX5_OE_FETALCORTEX_HUMAN_GSE89057_RNASEQ_UP_TF_Perturbations_Followed_by_Expression, PMC4342183-dddt-9-1103.pdf-33-ZNF121_ZNF121_Rummagene_transcription_factors
## VMH-female: GATA3_OE_MOUSE_GSE12999_CREEDSID_GENE_2531_DOWN_TF_Perturbations_Followed_by_Expression, ESR1_MUT_HUMAN_GSE1045_CREEDSID_GENE_1193_DOWN_TF_Perturbations_Followed_by_Expression, PMC8554124-DataSheet4.XLSX-17-HOXC10_HOXC10_Rummagene_transcription_factors, KLF4_human_TRRUST_Transcription_Factors_2019
## ARC-male: PMC4122406-pone.0104158.s010.pdf-87-ZNF286_ZNF286A_Rummagene_transcription_factors, PMC3921124-pone.0087610.s009.pdf-1-ESR1_ESR1_Rummagene_transcription_factors, 	
PMC3470785-NIHMS377161-supplement-12.pdf-669-ZNF598_ZNF598_Rummagene_transcription_factors
## ARC-female: RUNX2_mouse_TRRUST_Transcription_Factors_2019, 	
PMC7339813-Supplementary_Data2.docx-37-PRRX1_PRRX1_Rummagene_transcription_factors, PMC9357817-6934744.f1.docx-3-AR_AR_Rummagene_transcription_factors


```{r}
sexreslist <- readRDS("data/11-GSEA/04a-Sex DE TF-target GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")

vmhs <- sexreslist[["BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_VMH"]][pathway %in% c("PMC10307746-12672_2023_738_MOESM12_ESM.docx-0-Supplementary_Table_S3_The_Top_100_MTF1-correlated_Genes_That_Obtained_From_The_GEPIA2_0_Database_MTF1_Rummagene_transcription_factors"," 	
PMC7162552-pcbi.1007800.s006.txt--TFEA_TFE3_Rummagene_transcription_factors","SOX5_OE_FETALCORTEX_HUMAN_GSE89057_RNASEQ_UP_TF_Perturbations_Followed_by_Expression",
#"PMC4342183-dddt-9-1103.pdf-33-ZNF121_ZNF121_Rummagene_transcription_factors",
"GATA3_OE_MOUSE_GSE12999_CREEDSID_GENE_2531_DOWN_TF_Perturbations_Followed_by_Expression","ESR1_MUT_HUMAN_GSE1045_CREEDSID_GENE_1193_DOWN_TF_Perturbations_Followed_by_Expression","PMC8554124-DataSheet4.XLSX-17-HOXC10_HOXC10_Rummagene_transcription_factors","KLF4_human_TRRUST_Transcription_Factors_2019",
#"PMC4122406-pone.0104158.s010.pdf-87-ZNF286_ZNF286A_Rummagene_transcription_factors",
"PMC3921124-pone.0087610.s009.pdf-1-ESR1_ESR1_Rummagene_transcription_factors",
#"PMC3470785-NIHMS377161-supplement-12.pdf-669-ZNF598_ZNF598_Rummagene_transcription_factors",
"RUNX2_mouse_TRRUST_Transcription_Factors_2019","PMC7339813-Supplementary_Data2.docx-37-PRRX1_PRRX1_Rummagene_transcription_factors","PMC9357817-6934744.f1.docx-3-AR_AR_Rummagene_transcription_factors")]

arcs <- sexreslist[["BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_ARC"]][pathway %in% c("PMC10307746-12672_2023_738_MOESM12_ESM.docx-0-Supplementary_Table_S3_The_Top_100_MTF1-correlated_Genes_That_Obtained_From_The_GEPIA2_0_Database_MTF1_Rummagene_transcription_factors"," 	
PMC7162552-pcbi.1007800.s006.txt--TFEA_TFE3_Rummagene_transcription_factors","SOX5_OE_FETALCORTEX_HUMAN_GSE89057_RNASEQ_UP_TF_Perturbations_Followed_by_Expression",
#"PMC4342183-dddt-9-1103.pdf-33-ZNF121_ZNF121_Rummagene_transcription_factors",
"GATA3_OE_MOUSE_GSE12999_CREEDSID_GENE_2531_DOWN_TF_Perturbations_Followed_by_Expression","ESR1_MUT_HUMAN_GSE1045_CREEDSID_GENE_1193_DOWN_TF_Perturbations_Followed_by_Expression","PMC8554124-DataSheet4.XLSX-17-HOXC10_HOXC10_Rummagene_transcription_factors","KLF4_human_TRRUST_Transcription_Factors_2019",
#"PMC4122406-pone.0104158.s010.pdf-87-ZNF286_ZNF286A_Rummagene_transcription_factors",
"PMC3921124-pone.0087610.s009.pdf-1-ESR1_ESR1_Rummagene_transcription_factors",
#"PMC3470785-NIHMS377161-supplement-12.pdf-669-ZNF598_ZNF598_Rummagene_transcription_factors",
"RUNX2_mouse_TRRUST_Transcription_Factors_2019","PMC7339813-Supplementary_Data2.docx-37-PRRX1_PRRX1_Rummagene_transcription_factors","PMC9357817-6934744.f1.docx-3-AR_AR_Rummagene_transcription_factors")]

pltdat <- rbind(vmhs,arcs)
pltdat[,spdom:=gsub(spdom,pattern="BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_(.*)",replacement="\\1")]
pltdat[spdom=="VMH",sexreg:=ifelse(NES>0,yes="VMH Male Up",no="VMH Fem. Up")]
pltdat[spdom=="ARC",sexreg:=ifelse(NES>0,yes="ARC Male Up",no="ARC Fem. Up")]
rm(vmhs,arcs)

## flip all NESes to be positive
pltdat[NES<0,NES:=-1*NES]

## tidy gene sources
pltdat[pathway=="PMC10307746-12672_2023_738_MOESM12_ESM.docx-0-Supplementary_Table_S3_The_Top_100_MTF1-correlated_Genes_That_Obtained_From_The_GEPIA2_0_Database_MTF1_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC7162552-pcbi.1007800.s006.txt--TFEA_TFE3_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="SOX5_OE_FETALCORTEX_HUMAN_GSE89057_RNASEQ_UP_TF_Perturbations_Followed_by_Expression",source:="GEO TF\nPerturbations"]
pltdat[pathway=="PMC4342183-dddt-9-1103.pdf-33-ZNF121_ZNF121_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="GATA3_OE_MOUSE_GSE12999_CREEDSID_GENE_2531_DOWN_TF_Perturbations_Followed_by_Expression",source:="GEO TF\nPerturbations"]
pltdat[pathway=="ESR1_MUT_HUMAN_GSE1045_CREEDSID_GENE_1193_DOWN_TF_Perturbations_Followed_by_Expression",source:="GEO TF\nPerturbations"]
pltdat[pathway=="PMC8554124-DataSheet4.XLSX-17-HOXC10_HOXC10_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="KLF4_human_TRRUST_Transcription_Factors_2019",source:="TRRUST"]
pltdat[pathway=="PMC4122406-pone.0104158.s010.pdf-87-ZNF286_ZNF286A_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC3921124-pone.0087610.s009.pdf-1-ESR1_ESR1_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC3470785-NIHMS377161-supplement-12.pdf-669-ZNF598_ZNF598_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="RUNX2_mouse_TRRUST_Transcription_Factors_2019",source:="TRRUST"]
pltdat[pathway=="PMC7339813-Supplementary_Data2.docx-37-PRRX1_PRRX1_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC9357817-6934744.f1.docx-3-AR_AR_Rummagene_transcription_factors",source:="Rummagene"]

### tidy gene names
pltdat[pathway=="PMC10307746-12672_2023_738_MOESM12_ESM.docx-0-Supplementary_Table_S3_The_Top_100_MTF1-correlated_Genes_That_Obtained_From_The_GEPIA2_0_Database_MTF1_Rummagene_transcription_factors",pathway:="MTF1"]
pltdat[pathway=="PMC7162552-pcbi.1007800.s006.txt--TFEA_TFE3_Rummagene_transcription_factors",pathway:="TFE3"]
pltdat[pathway=="SOX5_OE_FETALCORTEX_HUMAN_GSE89057_RNASEQ_UP_TF_Perturbations_Followed_by_Expression",pathway:="SOX5 (OE -> Up)"]
pltdat[pathway=="PMC4342183-dddt-9-1103.pdf-33-ZNF121_ZNF121_Rummagene_transcription_factors",pathway:="ZNF121"]
pltdat[pathway=="GATA3_OE_MOUSE_GSE12999_CREEDSID_GENE_2531_DOWN_TF_Perturbations_Followed_by_Expression",pathway:="GATA3 (OE -> Down)"]
pltdat[pathway=="ESR1_MUT_HUMAN_GSE1045_CREEDSID_GENE_1193_DOWN_TF_Perturbations_Followed_by_Expression",pathway:="ESR1 (Mut -> Down)"]
pltdat[pathway=="PMC8554124-DataSheet4.XLSX-17-HOXC10_HOXC10_Rummagene_transcription_factors",pathway:="HOXC10"]
pltdat[pathway=="KLF4_human_TRRUST_Transcription_Factors_2019",pathway:="KLF4"]
pltdat[pathway=="PMC4122406-pone.0104158.s010.pdf-87-ZNF286_ZNF286A_Rummagene_transcription_factors",pathway:="ZNF286"]
pltdat[pathway=="PMC3921124-pone.0087610.s009.pdf-1-ESR1_ESR1_Rummagene_transcription_factors",pathway:="ESR1"]
pltdat[pathway=="PMC3470785-NIHMS377161-supplement-12.pdf-669-ZNF598_ZNF598_Rummagene_transcription_factors",pathway:="ZNF598"]
pltdat[pathway=="RUNX2_mouse_TRRUST_Transcription_Factors_2019",pathway:="RUNX2"]
pltdat[pathway=="PMC7339813-Supplementary_Data2.docx-37-PRRX1_PRRX1_Rummagene_transcription_factors",pathway:="PRRX1"]
pltdat[pathway=="PMC9357817-6934744.f1.docx-3-AR_AR_Rummagene_transcription_factors",pathway:="AR"]

### make plot
pdf("plots/Manuscript/5B as 4L-Sex DEG TF GSEA.pdf",height=1.5,width=4.5)
ggplot(pltdat,aes(y=pathway,x=-log10(padj),size=NES))+
    geom_point(aes(col=sexreg,size=NES,shape=source),alpha=0.7)+
    ylab("Term")+
    scale_color_manual(values=sexregpal,guide="none")+
    # scale point size for better visibility by doubling the max range of size increase
    scale_radius(range = c(min(pltdat$NES),1.5*max(pltdat$NES)),trans = "identity",guide="none")+
    xlab("-log10 (FDR)")+
    labs(shape="Source Database",col="DE Gene Set")+
    geom_vline(xintercept=-log10(0.05),linetype="dashed")+
    scale_shape_manual(values=c("Rummagene"=16,"TRRUST"=15,"GEO TF\nPerturbations"=18))+
    guides(col="none")+
    guides(size="none")+
    # ^ necess. to get spacing to work in theme
    guides(shape=guide_legend(override.aes=list(size = 2)))+
    # ^ make legend points bigger
    theme(axis.text.y = element_text(size = 9),axis.title.y=element_text(size=10), axis.title.x = element_text(size = 10), axis.text.x = element_text(size = 9), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5),legend.spacing.y = unit(0.005, "in"),legend.margin = margin(t = 0.1,0,0,-0.1,unit="in"))
dev.off()

rm(pltdat,sexreslist)

```

5C-disease enrichments
```{r}
sexdis <- readRDS("data/11-GSEA/06-Common variant disease gene enrichment in sex DEGs.RDS")

sexdis <- sexdis[spdom %in% grep(spdom,pattern="clpsd",value=T)]
sexdis <- sexdis[genesets_from=="union"]

sexdis[,spdom:=gsub(spdom,pattern="^.*_clpsdVMH_(.*)$",replacement="\\1")]
sexdis[,spdom:=gsub(spdom,pattern="X1$",replacement="WM.1")]
sexdis[,spdom:=gsub(spdom,pattern="X9$",replacement="WM.2")]
sexdis[,spdom:=factor(spdom,levels=c("ARC","VMH","WM.1","WM.2",paste0("X",c(2,4,6:8,10,13:15))))]

setnames(sexdis,"padj","FDR")

sexdis[,signif:=ifelse(FDR<0.05,yes="*",no="")]



pdf("plots/Manuscript/5C as 4M-sex DE_disease_GSEA.pdf",height=2.5,width=5)
ggplot(sexdis,aes(x=pathway,y=spdom,fill=ES))+
    geom_tile()+
    geom_text(aes(label=signif),size=2)+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    scale_fill_gradient2()+
    theme(plot.background = element_blank(),axis.line = element_blank(),panel.grid = element_blank(),panel.border = element_blank(),axis.text.x = element_text(angle=90,vjust=0.5,hjust=1,size=9),axis.text.y = element_text(size=9),axis.title.x = element_text(size = 10), axis.title.y = element_text(size = 10), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5),legend.spacing.y = unit(0.125, "in"),legend.margin = margin(0,0,0,0))+
    xlab("Disease Gene Set")+
    ylab("Domain")

dev.off()
```

5D-fDR-sig sex DE enrichments in disease genes by hypergeometric test
### not including for now
```{r}
resdat <- fread("data/11-GSEA/07-hypergeometric enrichment for FDR-sig sex DEGs per k15 cluster in neuropsych GWAS-TWAS-litmined union sets.txt")

resdat <- resdat[clus_set=="BSpace_k15_HARMONYlmbna_nnsvg10"]
resdat <- resdat[spdomain_sex %in% c(paste0(c("X11_","X12_","X3_","X5_","X1_","X9_"),"MaleUp"),paste0(c("X11_","X12_","X3_","X5_","X1_","X9_"),"FemUp"))]
#resdat <- resdat[label %in% unique(resdat[fdr<0.2,label])]

resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="^(.*)_FemUp",replacement=paste0("\\1"," Female Up"))]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="^(.*)_MaleUp",replacement=paste0("\\1"," Male Up"))]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X1 ",replacement="WM.1 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X9 ",replacement="WM.2 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X3 ",replacement="ARC.1 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X5 ",replacement="ARC.2 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X11 ",replacement="VMH.1 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X12 ",replacement="VMH.2 ")]

# make column labels to order by so that each cluster is paired X1male, X1 female, etc
i<-1
colord <- c()
pltcl <- c("VMH.1 ","VMH.2 ","ARC.1 ","ARC.2 ","WM.1 ","WM.2 ")
for (i in c(1:6)){
    colord <- c(colord,paste0(pltcl[i],c("Male Up","Female Up")))
}

pltdat <- dcast(resdat,formula = label~spdomain_sex,value.var = "pval")
pltdat <- as.data.frame(pltdat)
rownames(pltdat) <- pltdat$label
pltdat$label <- NULL
pltdat <- -log10(pltdat)

pdf("plots/Manuscript/5D-FDR sig sex DEG enrichments in dis genesets.pdf",height=3,width=3.5)
pheatmap::pheatmap(pltdat,cluster_cols = F,treeheight_row = 0,fontsize = 9,fontsize_row = 10)
dev.off()
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] BiocParallel_1.36.0 parallelly_1.36.0   default_1.0.0      
 [4] colorout_1.3-0.1    gridExtra_2.3       ggplot2_3.4.4      
 [7] Biostrings_2.70.1   GenomeInfoDb_1.38.5 XVector_0.42.0     
[10] IRanges_2.36.0      S4Vectors_0.40.2    BiocGenerics_0.48.1
[13] data.table_1.14.10  rlang_1.1.3        

loaded via a namespace (and not attached):
 [1] utf8_1.2.4              generics_0.1.3          bitops_1.0-7           
 [4] digest_0.6.34           magrittr_2.0.3          evaluate_0.23          
 [7] grid_4.3.2              RColorBrewer_1.1-3      fastmap_1.1.1          
[10] rprojroot_2.0.4         sessioninfo_1.2.2       fansi_1.0.6            
[13] scales_1.3.0            codetools_0.2-19        cli_3.6.2              
[16] crayon_1.5.2            munsell_0.5.0           withr_3.0.0            
[19] yaml_2.3.8              tools_4.3.2             parallel_4.3.2         
[22] dplyr_1.1.4             colorspace_2.1-0        GenomeInfoDbData_1.2.11
[25] here_1.0.1              vctrs_0.6.5             R6_2.5.1               
[28] lifecycle_1.0.4         zlibbioc_1.48.0         pkgconfig_2.0.3        
[31] pillar_1.9.0            gtable_0.3.4            glue_1.7.0             
[34] xfun_0.41               tibble_3.2.1            tidyselect_1.2.0       
[37] rstudioapi_0.15.0       knitr_1.45              farver_2.1.1           
[40] htmltools_0.5.7         labeling_0.4.3          rmarkdown_2.25         
[43] pheatmap_1.0.12         compiler_4.3.2          RCurl_1.98-1.14        
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-01-23
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────
 package          * version   date (UTC) lib source
 BiocGenerics     * 0.48.1    2023-11-02 [1] Bioconductor
 BiocParallel     * 1.36.0    2023-10-26 [1] Bioconductor
 Biostrings       * 2.70.1    2023-10-25 [1] Bioconductor
 bitops             1.0-7     2021-04-24 [1] CRAN (R 4.3.2)
 cli                3.6.2     2023-12-11 [1] CRAN (R 4.3.2)
 codetools          0.2-19    2023-02-01 [1] CRAN (R 4.3.0)
 colorout         * 1.3-0.1   2024-01-11 [1] local
 colorspace         2.1-0     2023-01-23 [1] CRAN (R 4.3.2)
 crayon             1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 data.table       * 1.14.10   2023-12-08 [1] CRAN (R 4.3.2)
 default          * 1.0.0     2017-08-07 [1] CRAN (R 4.3.0)
 digest             0.6.34    2024-01-11 [1] CRAN (R 4.3.2)
 dplyr              1.1.4     2023-11-17 [1] CRAN (R 4.3.2)
 evaluate           0.23      2023-11-01 [1] CRAN (R 4.3.1)
 fansi              1.0.6     2023-12-08 [1] CRAN (R 4.3.2)
 farver             2.1.1     2022-07-06 [1] CRAN (R 4.3.2)
 fastmap            1.1.1     2023-02-24 [1] CRAN (R 4.3.2)
 generics           0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb     * 1.38.5    2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData   1.2.11    2024-01-11 [1] Bioconductor
 ggplot2          * 3.4.4     2023-10-12 [1] CRAN (R 4.3.1)
 glue               1.7.0     2024-01-09 [1] CRAN (R 4.3.2)
 gridExtra        * 2.3       2017-09-09 [1] CRAN (R 4.3.0)
 gtable             0.3.4     2023-08-21 [1] CRAN (R 4.3.0)
 here               1.0.1     2020-12-13 [1] CRAN (R 4.3.0)
 htmltools          0.5.7     2023-11-03 [1] CRAN (R 4.3.2)
 IRanges          * 2.36.0    2023-10-24 [1] Bioconductor
 knitr              1.45      2023-10-30 [1] CRAN (R 4.3.2)
 labeling           0.4.3     2023-08-29 [1] CRAN (R 4.3.0)
 lifecycle          1.0.4     2023-11-07 [1] CRAN (R 4.3.1)
 magrittr           2.0.3     2022-03-30 [1] CRAN (R 4.3.2)
 munsell            0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 parallelly       * 1.36.0    2023-05-26 [1] CRAN (R 4.3.0)
 pheatmap           1.0.12    2019-01-04 [1] CRAN (R 4.3.0)
 pillar             1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig          2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 R6                 2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 RColorBrewer       1.1-3     2022-04-03 [1] CRAN (R 4.3.0)
 RCurl              1.98-1.14 2024-01-09 [1] CRAN (R 4.3.1)
 rlang            * 1.1.3     2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown          2.25      2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot          2.0.4     2023-11-05 [1] CRAN (R 4.3.1)
 rstudioapi         0.15.0    2023-07-07 [1] CRAN (R 4.3.0)
 S4Vectors        * 0.40.2    2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 scales             1.3.0     2023-11-28 [1] CRAN (R 4.3.2)
 sessioninfo        1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 tibble             3.2.1     2023-03-20 [1] CRAN (R 4.3.2)
 tidyselect         1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 utf8               1.2.4     2023-10-22 [1] CRAN (R 4.3.2)
 vctrs              0.6.5     2023-12-01 [1] CRAN (R 4.3.2)
 withr              3.0.0     2024-01-16 [1] CRAN (R 4.3.1)
 xfun               0.41      2023-11-01 [1] CRAN (R 4.3.2)
 XVector          * 0.42.0    2023-10-24 [1] Bioconductor
 yaml               2.3.8     2023-12-11 [1] CRAN (R 4.3.2)
 zlibbioc           1.48.0    2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

───────────────────────────────────────────────────────────────────────────
