---
title: "Fig 4-Sex DE"
author: "Bernard Mulvey"
date: "2023-12-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table) # Preferred data manipulation package
library(Biostrings) # General
library(ggrastr) # render plots with bajillions of points as raster so they can be manipulated in eg Illustrator
library(Cairo) # pdf device for ^ to properly save
library(ggplot2) # Dependency for several plotting functions
library(gridExtra)
library(SpatialExperiment) # Dependency
library(escheR) # spotplots with domain outlines
library(ggrepel) # for plotting labels on a volcano cleanly
require(colorout) # Utility for RStudio
library(ggtext) # mixed font styles in ggplots
ColorOut() 

# code reformatting in Rstudio
options(styler.addins_style_transformer = "biocthis::bioc_style()") 

# set plotting defaults for ggplot
theme_set(theme_bw() + theme(axis.text.x = element_text(size = 10), axis.title.x = element_text(size = 11), axis.text.y = element_text(size = 10), axis.title.y = element_text(size = 11), plot.title = element_text(size = 12, hjust = 0.5), strip.text = element_text(size = 18), legend.text = element_text(size = 9), legend.title = element_text(size = 10, hjust = 0.5)))
```

### Setup: load data, cluster assignments, and rename from generic to appropriate region (VMH/ARC/WM)
```{r}
hyp2 <- readRDS("data/03-QC_filters/hypN9_umi275_gene166_chrm50_lognorm_111723.RDS")

bscl <- fread("data/06-BayesSpace/02-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=paste0("X",c(1:15)))

### relabel as VMH/ARC/WM for spotplottings
tmpcd <- as.data.table(colData(hyp2),keep.rownames=T)
tmpcd[label %in% c("X11","X12"),label:="VMH"]
tmpcd[label %in% c("X3","X5"),label:="ARC"]
tmpcd[label %in% c("X1","X9"),label:="WM"]

tmpcd <- DataFrame(tmpcd)
rownames(tmpcd) <- tmpcd$rn
tmpcd <- tmpcd[colnames(hyp2),]
colData(hyp2) <- tmpcd
hyp2$label <- factor(hyp2$label,levels=c("VMH","ARC","WM",paste0("X",c(2,4,6:8,10,13:15))))

### also define cluster tables that only have VMH or ARC and otehrs as NA for escher plotting outlines specific to VMH/ARC
bscl.vmhonly <- as.data.table(bscl,keep.rownames=T)
bscl.arconly <- as.data.table(bscl,keep.rownames=T)
bscl.vmharc <- as.data.table(bscl,keep.rownames=T)

bscl.vmhonly <- bscl.vmhonly[cl %in% c("X11","X12"),cl2:="VMH"]
bscl.arconly <- bscl.arconly[cl %in% c("X3","X5"),cl2:="ARC"]
bscl.vmharc <- bscl.vmharc[cl %in% c("X11","X12"),cl2:="VMH"]
bscl.vmharc <- bscl.vmharc[cl %in% c("X3","X5"),cl2:="ARC"]

rm(tmpcd)
```


4A: Sex DE volcanos for VMH
```{r}
sexderes <- fread("data/09-Sex DE/01c-dreamlet-ranfx_nnsvg10-HmnylmbNA-BS-15-VMHARCclpsd.txt")

sexderes <- sexderes[!(chromosome_name %in% c("chrX","chrY","chrMT"))]

vmhde <- sexderes[assay=="VMH"]
vmhde[,genelabel:=as.character(NA)]
vmhde[gene_name %in% c("CACNA1D", "ANKRD17", "PTGR1", "HACE1"),genelabel:=gene_name]

vmhvol <- ggplot(vmhde, aes(y = -log10(P.Value), x = logFC)) +
    geom_point(size = 0.4,
               col = "#5dd959") +
    geom_hline(yintercept = -log10(0.05),
        linetype = "dashed",
        color = "black",
        linewidth=0.25) +
    geom_hline(yintercept = -log10(max(sexderes[adj.P.Val<0.05&assay=="VMH",P.Value])),
        linetype = "dashed",
        color = "red",
        linewidth=0.25) +
    geom_text_repel(aes(y = -log10(P.Value),
            x = logFC,
            label = ifelse(is.na(genelabel), "", gene_name)),
        nudge_y = ifelse(vmhde$genelabel %in% c("PTGR1","CACNA1D"), yes = 1.2, no = -0.4),
        nudge_x = ifelse(vmhde$genelabel%in%c("HACE1","CACNA1D"),yes=sign(vmhde$logFC)*1,no=sign(vmhde$logFC)*0.3),
        min.segment.length = 0,
        size = 2,
        segment.size = 0.25) +
    xlab("VMH logFC\n(Male vs. Female)") +
    ylab("-log10 (p-value)") +
    xlim(-2.25, 2.25) +
    theme(axis.title.x = element_text(size = 10),
        axis.text.x = element_text(size = 9),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 9))

pdf("plots/Manuscript/4A-VMH Sex DE Volcano.pdf",height=2.5,width=2.3)
vmhvol
dev.off()

rm(vmhvol)
```

4B: ARC Volcano
```{r}
arcde <- sexderes[assay=="ARC"]
arcde[,genelabel:=as.character(NA)]
arcde[gene_name %in% c("PENK", "IGFLR1", "IGFBP6","CAPS2"),genelabel:=gene_name]

arcvol <- ggplot(arcde, aes(y = -log10(P.Value), x = logFC)) +
    geom_point(size = 0.4,
               col = "#8c63cf") +
    geom_hline(yintercept = -log10(0.05),
        linetype = "dashed",
        color = "black",
        linewidth=0.25) +
    geom_hline(yintercept = -log10(max(sexderes[adj.P.Val<0.05&assay=="ARC",P.Value])),
        linetype = "dashed",
        color = "red",
        linewidth=0.25) +
    geom_text_repel(aes(y = -log10(P.Value),
                        x = logFC,
                        label = ifelse(is.na(genelabel), "", gene_name)),
        nudge_y = 4,
        nudge_x = 2*sign(arcde$logFC)/(3*abs(arcde$logFC)),
        min.segment.length = 0,
        size = 2,
        segment.size = 0.25) +
    xlab("ARC logFC\n(Male vs. Female)") +
    ylab("-log10 (p-value)") +
    xlim(-3, 3) +
    theme(axis.title.x = element_text(size = 10),
        axis.text.x = element_text(size = 9),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 9))
    
pdf("plots/Manuscript/4B-ARC Sex DE Volcano.pdf",height=2.5,width=2.3)
arcvol
dev.off()
rm(arcvol)
```

4C: ARC-VMH sex DE t stat correls
```{r}
# make primary plotting table
setnames(vmhde,"t","tvmh")
setnames(arcde,"t","tarc")

arcvmh <- merge(vmhde[,.(ID,tvmh)],arcde[,.(ID,tarc)],by="ID",all.x=T,all.y=T)

## get genes significant in one, both, and neither
setnames(arcde,"adj.P.Val","arcadjp")
setnames(vmhde,"adj.P.Val","vmhadjp")
setnames(arcde,"logFC","arclogfc")
setnames(vmhde,"logFC","vmhlogfc")
arcvmh2 <- merge(vmhde[,.(ID,vmhadjp,vmhlogfc)],arcde[,.(ID,arcadjp,arclogfc)],by="ID",all.x=T,all.y=T)

arcvmh2[vmhadjp<0.05&arcadjp<0.05,de:="Both"]
arcvmh2[vmhadjp<0.05&(arcadjp>=0.05|is.na(arcadjp)),de:="VMH"]
arcvmh2[(vmhadjp>=0.05|is.na(vmhadjp))&arcadjp<0.05,de:="ARC"]
arcvmh2[(vmhadjp>=0.05|is.na(vmhadjp))&(arcadjp>=0.05|is.na(arcadjp)),de:="Neither"]

## add that information for point coloring
arcvmh <- merge(arcvmh,arcvmh2[,.(ID,de)],by="ID",all=T)
arcvmh[,de:=factor(de,levels=c("VMH","ARC","Both","Neither"))]

# make a palette for these four colors
pal <- c("Both"="#157e82","VMH"="#5dd959","ARC"="#8c63cf","Neither"="#808080")
# get pearson correlation to include in plot
cor(arcvmh$tvmh,arcvmh$tarc,method="pearson")

arcvmhcor <- ggplot(arcvmh, aes(x = tvmh, y = tarc)) +
    geom_point(aes(col = de), size = ifelse(arcvmh$de %in% c("Both", "VMH", "ARC"), 0.65, 0.2),alpha=0.7) +
    scale_color_manual(values = pal) +
    scale_fill_manual(values=pal) +
    guides(fill="none")+
    geom_abline(
        slope = 1,
        intercept = 0,
        linetype = "dashed",
        col = "black",
        linewidth = 0.25
    ) +
    geom_text(aes(x = -5.5, y = 12), label = "Pearson Coef=0.379", size =
                  2.5) +
    xlim(-9, 12) +
    ylim(-9, 12) +
    labs(color = "Sex DE\nDomains") +
    xlab("VMH Sex DE *t*<br>(Male vs. Female)") +
    ylab("ARC Sex DE *t*<br>(Male vs. Female)") +
    theme(
        axis.title.x = element_markdown(size = 10),
        axis.text.x = element_text(size = 9),
        axis.title.y = element_markdown(size = 10),
        axis.text.y = element_text(size = 9),
        legend.title = element_text(size = 9, hjust = 0.5),
        legend.position = c(.89, .23),
        legend.box.background = element_blank(),
        legend.key.size=unit(0.125,"in"),
        legend.key = element_blank(),
        legend.background = element_blank(),
        legend.text=element_text(size=8,
                                 margin=margin(0,0.05,0,-0.05,"in"),
                                 vjust=0.5))

pdf("plots/Manuscript/4C-ARC, VMH DE correls.pdf",height=2.5,width=3)
arcvmhcor
dev.off()


rm(arcvmhcor,vmhvol,arcvol,plts,arcvmh)
```

## Fig 4Di and Dii/Ei + Eii: sex DEG spotplots
D-E: PTGR1 in VMH (2E-9)
F-G: PGR in ARC (FDR 0.015)
```{r}
ptgrM <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(ptgrM) <- rowData(ptgrM)$gene_name
colData(ptgrM)$`log counts` <- logcounts(ptgrM)["PTGR1",]
ptgrlab <- copy(bscl.vmharc)[rn %in% colnames(ptgrM)]
ptgrlab <- DataFrame(ptgrlab)
rownames(ptgrlab) <- ptgrlab$rn


ptgrM$Domain <- factor(ptgrlab$cl2,levels=c("VMH","ARC"))
ptgrM <- ptgrM[,ptgrM$Domain %in% c("VMH","ARC")]
### make spotplots
p <- make_escheR(ptgrM)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/4Di-PTGR1_VMHsexDE_male.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    scale_fill_gradient(low="white",high="black",limits=c(0,3))+
    ggtitle("*PTGR1* (V12Y31-<br>080_A1 - Male)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
    #theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9),plot.margin = margin(-0.3,0,-0.1,0,unit = "in"))
dev.off()

rm(ptgrM,p,ptgrlab)

### 
ptgrF <- hyp2[,hyp2$sample_id=="V13M13-362_D1"]
rownames(ptgrF) <- rowData(ptgrF)$gene_name
colData(ptgrF)$`log counts` <- logcounts(ptgrF)["PTGR1",]
ptgrlab <- copy(bscl.vmharc)[rn %in% colnames(ptgrF)]
ptgrlab <- DataFrame(ptgrlab)
rownames(ptgrlab) <- ptgrlab$rn

ptgrF$Domain <- factor(ptgrlab$cl2,levels=c("VMH","ARC"))
ptgrF <- ptgrF[,ptgrF$Domain %in% c("VMH","ARC")]
### make spotplots
p <- make_escheR(ptgrF)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/4Dii-PTGR1_VMHsexDE_female.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    scale_fill_gradient(low="white",high="black",limits=c(0,3))+
    ggtitle("*PTGR1* (V13M13-<br>362_D1 - Female)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(ptgrF.p,ptgrF,p,ptgrlab)
```


# F/G: PGR
```{r}
pgrM <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(pgrM) <- rowData(pgrM)$gene_name
colData(pgrM)$`log counts` <- logcounts(pgrM)["PGR",]
pgrlab <- copy(bscl.vmharc)[rn %in% colnames(pgrM)]
pgrlab <- DataFrame(pgrlab)
rownames(pgrlab) <- pgrlab$rn


pgrM$Domain <- factor(pgrlab$cl2,levels=c("VMH","ARC"))
pgrM <- pgrM[,pgrM$Domain %in% c("VMH","ARC")]
### make spotplots
p <- make_escheR(pgrM)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/4Ei-PGR_ARCDE_male.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    scale_fill_gradient(low="white",high="black",limits=c(0,2.5))+
    ggtitle("*PGR* (V12Y31-<br>080_A1 - Male)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
    #theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9),plot.margin = margin(-0.3,0,-0.1,0,unit = "in"))
dev.off()

rm(pgrM,pgrlab,ptgrlab)

### 
pgrF <- hyp2[,hyp2$sample_id=="V12D05-348_D1"]
rownames(pgrF) <- rowData(pgrF)$gene_name
colData(pgrF)$`log counts` <- logcounts(pgrF)["PGR",]
pgrlab <- copy(bscl.vmharc)[rn %in% colnames(pgrF)]
pgrlab <- DataFrame(pgrlab)
rownames(pgrlab) <- pgrlab$rn

pgrF$Domain <- factor(pgrlab$cl2,levels=c("VMH","ARC"))
pgrF <- pgrF[,pgrF$Domain %in% c("VMH","ARC")]
### make spotplots
p <- make_escheR(pgrF)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/4Eii-PGR_ARCDE_female.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    scale_fill_gradient(low="white",high="black",limits=c(0,2.5))+
    ggtitle("*PGR* (V12D05-<br>348_D1 - Female)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(pgrF,pgrlab,p,ptgrlab)
```

Shared legend: 4D thru 4G
```{r}
### custom function to extrct legend from one plot to extend to all ###
extract_legend <- function(my_ggp) {
  step1 <- ggplot_gtable(ggplot_build(my_ggp))
  step2 <- which(sapply(step1$grobs, function(x) x$name) == "guide-box")
  step3 <- step1$grobs[[step2]]
  return(step3)
}

### run the first part of the code used to generate the marker panels but with the fill legend changed to be the domains factored in the same order
gabre <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(gabre) <- rowData(gabre)$gene_name
colData(gabre)$`log counts` <- logcounts(gabre)["GABRE",]
gabrelab <- copy(bscl.vmharc)[rn %in% colnames(gabre)]
gabrelab <- DataFrame(gabrelab)
rownames(gabrelab) <- gabrelab$rn

gabre$Domain <- factor(gabrelab$cl2,levels=c("VMH","ARC"))
### make a ggplot of nothing in particular, we just want the legend
p <- ggplot(as.data.table(colData(gabre))[!is.na(Domain)],aes(x=sum_umi,y=expr_chrM_ratio,color=Domain))+
    geom_point()+
    scale_color_manual(values=c("#5dd959","#8c63cf"),na.value = NA)+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9),plot.margin = margin(-0.3,0,-0.1,0,unit = "in"))

pl <- extract_legend(p)
dev.off()


# save in dimension ~equiv to the dimensions of a legend included in one panel
pdf("plots/Manuscript/4D-E shared legend.pdf",height=0.6,width=0.45)
do.call("grid.arrange",list(pl))
dev.off()
```

```{r}
sessionInfo()
sessioninfo::session_info()
```

R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggtext_0.1.2                colorout_1.3-0.1           
 [3] ggrepel_0.9.5               escheR_1.2.0               
 [5] SpatialExperiment_1.12.0    SingleCellExperiment_1.24.0
 [7] SummarizedExperiment_1.32.0 Biobase_2.62.0             
 [9] GenomicRanges_1.54.1        MatrixGenerics_1.14.0      
[11] matrixStats_1.2.0           gridExtra_2.3              
[13] ggplot2_3.4.4               Cairo_1.6-2                
[15] ggrastr_1.0.2               Biostrings_2.70.2          
[17] GenomeInfoDb_1.38.5         XVector_0.42.0             
[19] IRanges_2.36.0              S4Vectors_0.40.2           
[21] BiocGenerics_0.48.1         data.table_1.14.10         
[23] rlang_1.1.3                

loaded via a namespace (and not attached):
  [1] later_1.3.2                   BiocIO_1.12.0                
  [3] bitops_1.0-7                  filelock_1.0.3               
  [5] fields_15.2                   tibble_3.2.1                 
  [7] XML_3.99-0.16.1               lifecycle_1.0.4              
  [9] edgeR_4.0.14                  doParallel_1.0.17            
 [11] rprojroot_2.0.4               lattice_0.22-5               
 [13] magrittr_2.0.3                limma_3.58.1                 
 [15] plotly_4.10.4                 sass_0.4.8                   
 [17] rmarkdown_2.25                jquerylib_0.1.4              
 [19] yaml_2.3.8                    httpuv_1.6.13                
 [21] spam_2.10-0                   sessioninfo_1.2.2            
 [23] cowplot_1.1.3                 DBI_1.2.1                    
 [25] RColorBrewer_1.1-3            golem_0.4.1                  
 [27] maps_3.4.2                    abind_1.4-5                  
 [29] zlibbioc_1.48.0               purrr_1.0.2                  
 [31] RCurl_1.98-1.14               rappdirs_0.3.3               
 [33] GenomeInfoDbData_1.2.11       irlba_2.3.5.1                
 [35] pheatmap_1.0.12               commonmark_1.9.1             
 [37] DelayedMatrixStats_1.24.0     codetools_0.2-19             
 [39] DelayedArray_0.28.0           xml2_1.3.6                   
 [41] DT_0.31                       scuttle_1.12.0               
 [43] tidyselect_1.2.0              farver_2.1.1                 
 [45] ScaledMatrix_1.10.0           viridis_0.6.5                
 [47] shinyWidgets_0.8.1            BiocFileCache_2.10.1         
 [49] GenomicAlignments_1.38.2      jsonlite_1.8.8               
 [51] BiocNeighbors_1.20.2          ellipsis_0.3.2               
 [53] scater_1.30.1                 iterators_1.0.14             
 [55] foreach_1.5.2                 tools_4.3.2                  
 [57] Rcpp_1.0.12                   glue_1.7.0                   
 [59] SparseArray_1.2.3             xfun_0.41                    
 [61] here_1.0.1                    dplyr_1.1.4                  
 [63] withr_3.0.0                   BiocManager_1.30.22          
 [65] fastmap_1.1.1                 fansi_1.0.6                  
 [67] digest_0.6.34                 rsvd_1.0.5                   
 [69] R6_2.5.1                      mime_0.12                    
 [71] colorspace_2.1-0              markdown_1.12                
 [73] RSQLite_2.3.5                 config_0.3.2                 
 [75] utf8_1.2.4                    tidyr_1.3.1                  
 [77] generics_0.1.3                rtracklayer_1.62.0           
 [79] httr_1.4.7                    htmlwidgets_1.6.4            
 [81] S4Arrays_1.2.0                pkgconfig_2.0.3              
 [83] gtable_0.3.4                  blob_1.2.4                   
 [85] htmltools_0.5.7               dotCall64_1.1-1              
 [87] scales_1.3.0                  png_0.1-8                    
 [89] attempt_0.3.1                 knitr_1.45                   
 [91] rstudioapi_0.15.0             rjson_0.2.21                 
 [93] curl_5.2.0                    cachem_1.0.8                 
 [95] stringr_1.5.1                 BiocVersion_3.18.1           
 [97] parallel_4.3.2                vipor_0.4.7                  
 [99] AnnotationDbi_1.64.1          restfulr_0.0.15              
[101] pillar_1.9.0                  grid_4.3.2                   
[103] vctrs_0.6.5                   promises_1.2.1               
[105] BiocSingular_1.18.0           dbplyr_2.4.0                 
[107] beachmat_2.18.0               xtable_1.8-4                 
[109] beeswarm_0.4.0                paletteer_1.6.0              
[111] evaluate_0.23                 magick_2.8.2                 
[113] cli_3.6.2                     locfit_1.5-9.8               
[115] compiler_4.3.2                Rsamtools_2.18.0             
[117] crayon_1.5.2                  labeling_0.4.3               
[119] rematch2_2.1.2                ggbeeswarm_0.7.2             
[121] stringi_1.8.3                 viridisLite_0.4.2            
[123] BiocParallel_1.36.0           munsell_0.5.0                
[125] lazyeval_0.2.2                Matrix_1.6-5                 
[127] ExperimentHub_2.10.0          benchmarkme_1.0.8            
[129] sparseMatrixStats_1.14.0      bit64_4.0.5                  
[131] KEGGREST_1.42.0               statmod_1.5.0                
[133] shiny_1.8.0                   interactiveDisplayBase_1.40.0
[135] AnnotationHub_3.10.0          gridtext_0.1.5               
[137] memoise_2.0.1                 bslib_0.6.1                  
[139] benchmarkmeData_1.0.4         bit_4.0.5                    
[141] spatialLIBD_1.14.1           
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2024-02-10
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────
 package                * version     date (UTC) lib source
 abind                    1.4-5       2016-07-21 [1] CRAN (R 4.3.0)
 AnnotationDbi            1.64.1      2023-11-02 [1] Bioconductor
 AnnotationHub            3.10.0      2023-10-24 [1] Bioconductor
 attempt                  0.3.1       2020-05-03 [1] CRAN (R 4.3.0)
 beachmat                 2.18.0      2023-10-24 [1] Bioconductor
 beeswarm                 0.4.0       2021-06-01 [1] CRAN (R 4.3.2)
 benchmarkme              1.0.8       2022-06-12 [1] CRAN (R 4.3.0)
 benchmarkmeData          1.0.4       2020-04-23 [1] CRAN (R 4.3.0)
 Biobase                * 2.62.0      2023-10-24 [1] Bioconductor
 BiocFileCache            2.10.1      2023-10-26 [1] Bioconductor
 BiocGenerics           * 0.48.1      2023-11-02 [1] Bioconductor
 BiocIO                   1.12.0      2023-10-26 [1] Bioconductor
 BiocManager              1.30.22     2023-08-08 [1] CRAN (R 4.3.0)
 BiocNeighbors            1.20.2      2024-01-07 [1] Bioconductor 3.18 (R 4.3.2)
 BiocParallel             1.36.0      2023-10-26 [1] Bioconductor
 BiocSingular             1.18.0      2023-10-24 [1] Bioconductor
 BiocVersion              3.18.1      2023-11-18 [1] Bioconductor 3.18 (R 4.3.2)
 Biostrings             * 2.70.2      2024-01-28 [1] Bioconductor 3.18 (R 4.3.2)
 bit                      4.0.5       2022-11-15 [1] CRAN (R 4.3.2)
 bit64                    4.0.5       2020-08-30 [1] CRAN (R 4.3.2)
 bitops                   1.0-7       2021-04-24 [1] CRAN (R 4.3.2)
 blob                     1.2.4       2023-03-17 [1] CRAN (R 4.3.0)
 bslib                    0.6.1       2023-11-28 [1] CRAN (R 4.3.1)
 cachem                   1.0.8       2023-05-01 [1] CRAN (R 4.3.2)
 Cairo                  * 1.6-2       2023-11-28 [1] CRAN (R 4.3.1)
 cli                      3.6.2       2023-12-11 [1] CRAN (R 4.3.2)
 codetools                0.2-19      2023-02-01 [1] CRAN (R 4.3.0)
 colorout               * 1.3-0.1     2024-01-11 [1] local
 colorspace               2.1-0       2023-01-23 [1] CRAN (R 4.3.2)
 commonmark               1.9.1       2024-01-30 [1] CRAN (R 4.3.2)
 config                   0.3.2       2023-08-30 [1] CRAN (R 4.3.0)
 cowplot                  1.1.3       2024-01-22 [1] CRAN (R 4.3.1)
 crayon                   1.5.2       2022-09-29 [1] CRAN (R 4.3.0)
 curl                     5.2.0       2023-12-08 [1] CRAN (R 4.3.2)
 data.table             * 1.14.10     2023-12-08 [1] CRAN (R 4.3.2)
 DBI                      1.2.1       2024-01-12 [1] CRAN (R 4.3.2)
 dbplyr                   2.4.0       2023-10-26 [1] CRAN (R 4.3.1)
 DelayedArray             0.28.0      2023-10-24 [1] Bioconductor
 DelayedMatrixStats       1.24.0      2023-11-06 [1] Bioconductor
 digest                   0.6.34      2024-01-11 [1] CRAN (R 4.3.2)
 doParallel               1.0.17      2022-02-07 [1] CRAN (R 4.3.0)
 dotCall64                1.1-1       2023-11-28 [1] CRAN (R 4.3.2)
 dplyr                    1.1.4       2023-11-17 [1] CRAN (R 4.3.2)
 DT                       0.31        2023-12-09 [1] CRAN (R 4.3.1)
 edgeR                    4.0.14      2024-01-29 [1] Bioconductor 3.18 (R 4.3.2)
 ellipsis                 0.3.2       2021-04-29 [1] CRAN (R 4.3.2)
 escheR                 * 1.2.0       2023-10-26 [1] Bioconductor
 evaluate                 0.23        2023-11-01 [1] CRAN (R 4.3.1)
 ExperimentHub            2.10.0      2023-10-26 [1] Bioconductor
 fansi                    1.0.6       2023-12-08 [1] CRAN (R 4.3.2)
 farver                   2.1.1       2022-07-06 [1] CRAN (R 4.3.2)
 fastmap                  1.1.1       2023-02-24 [1] CRAN (R 4.3.2)
 fields                   15.2        2023-08-17 [1] CRAN (R 4.3.2)
 filelock                 1.0.3       2023-12-11 [1] CRAN (R 4.3.2)
 foreach                  1.5.2       2022-02-02 [1] CRAN (R 4.3.0)
 generics                 0.1.3       2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb           * 1.38.5      2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData         1.2.11      2024-01-11 [1] Bioconductor
 GenomicAlignments        1.38.2      2024-01-16 [1] Bioconductor 3.18 (R 4.3.2)
 GenomicRanges          * 1.54.1      2023-10-29 [1] Bioconductor
 ggbeeswarm               0.7.2       2023-04-29 [1] CRAN (R 4.3.0)
 ggplot2                * 3.4.4       2023-10-12 [1] CRAN (R 4.3.1)
 ggrastr                * 1.0.2       2023-06-01 [1] CRAN (R 4.3.0)
 ggrepel                * 0.9.5       2024-01-10 [1] CRAN (R 4.3.2)
 ggtext                 * 0.1.2       2022-09-16 [1] CRAN (R 4.3.0)
 glue                     1.7.0       2024-01-09 [1] CRAN (R 4.3.2)
 golem                    0.4.1       2023-06-05 [1] CRAN (R 4.3.0)
 gridExtra              * 2.3         2017-09-09 [1] CRAN (R 4.3.0)
 gridtext                 0.1.5       2022-09-16 [1] CRAN (R 4.3.2)
 gtable                   0.3.4       2023-08-21 [1] CRAN (R 4.3.0)
 here                     1.0.1       2020-12-13 [1] CRAN (R 4.3.0)
 htmltools                0.5.7       2023-11-03 [1] CRAN (R 4.3.2)
 htmlwidgets              1.6.4       2023-12-06 [1] CRAN (R 4.3.1)
 httpuv                   1.6.13      2023-12-06 [1] CRAN (R 4.3.1)
 httr                     1.4.7       2023-08-15 [1] CRAN (R 4.3.0)
 interactiveDisplayBase   1.40.0      2023-10-26 [1] Bioconductor
 IRanges                * 2.36.0      2023-10-24 [1] Bioconductor
 irlba                    2.3.5.1     2022-10-03 [1] CRAN (R 4.3.2)
 iterators                1.0.14      2022-02-05 [1] CRAN (R 4.3.0)
 jquerylib                0.1.4       2021-04-26 [1] CRAN (R 4.3.0)
 jsonlite                 1.8.8       2023-12-04 [1] CRAN (R 4.3.2)
 KEGGREST                 1.42.0      2023-10-26 [1] Bioconductor
 knitr                    1.45        2023-10-30 [1] CRAN (R 4.3.2)
 labeling                 0.4.3       2023-08-29 [1] CRAN (R 4.3.0)
 later                    1.3.2       2023-12-06 [1] CRAN (R 4.3.2)
 lattice                  0.22-5      2023-10-24 [1] CRAN (R 4.3.2)
 lazyeval                 0.2.2       2019-03-15 [1] CRAN (R 4.3.2)
 lifecycle                1.0.4       2023-11-07 [1] CRAN (R 4.3.1)
 limma                    3.58.1      2023-10-31 [1] Bioconductor
 locfit                   1.5-9.8     2023-06-11 [1] CRAN (R 4.3.2)
 magick                   2.8.2       2023-12-20 [1] CRAN (R 4.3.1)
 magrittr                 2.0.3       2022-03-30 [1] CRAN (R 4.3.2)
 maps                     3.4.2       2023-12-15 [1] CRAN (R 4.3.1)
 markdown                 1.12        2023-12-06 [1] CRAN (R 4.3.1)
 Matrix                   1.6-5       2024-01-11 [1] CRAN (R 4.3.2)
 MatrixGenerics         * 1.14.0      2023-10-26 [1] Bioconductor
 matrixStats            * 1.2.0       2023-12-11 [1] CRAN (R 4.3.2)
 memoise                  2.0.1       2021-11-26 [1] CRAN (R 4.3.0)
 mime                     0.12        2021-09-28 [1] CRAN (R 4.3.2)
 munsell                  0.5.0       2018-06-12 [1] CRAN (R 4.3.0)
 paletteer                1.6.0       2024-01-21 [1] CRAN (R 4.3.1)
 pheatmap                 1.0.12      2019-01-04 [1] CRAN (R 4.3.0)
 pillar                   1.9.0       2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig                2.0.3       2019-09-22 [1] CRAN (R 4.3.0)
 plotly                   4.10.4      2024-01-13 [1] CRAN (R 4.3.1)
 png                      0.1-8       2022-11-29 [1] CRAN (R 4.3.2)
 promises                 1.2.1       2023-08-10 [1] CRAN (R 4.3.2)
 purrr                    1.0.2       2023-08-10 [1] CRAN (R 4.3.2)
 R6                       2.5.1       2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs                 0.3.3       2021-01-31 [1] CRAN (R 4.3.2)
 RColorBrewer             1.1-3       2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp                     1.0.12      2024-01-09 [1] CRAN (R 4.3.2)
 RCurl                    1.98-1.14   2024-01-09 [1] CRAN (R 4.3.1)
 rematch2                 2.1.2       2020-05-01 [1] CRAN (R 4.3.0)
 restfulr                 0.0.15      2022-06-16 [1] CRAN (R 4.3.2)
 rjson                    0.2.21      2022-01-09 [1] CRAN (R 4.3.2)
 rlang                  * 1.1.3       2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown                2.25        2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot                2.0.4       2023-11-05 [1] CRAN (R 4.3.1)
 Rsamtools                2.18.0      2023-10-24 [1] Bioconductor
 RSQLite                  2.3.5       2024-01-21 [1] CRAN (R 4.3.2)
 rstudioapi               0.15.0      2023-07-07 [1] CRAN (R 4.3.0)
 rsvd                     1.0.5       2021-04-16 [1] CRAN (R 4.3.0)
 rtracklayer              1.62.0      2023-10-26 [1] Bioconductor
 S4Arrays                 1.2.0       2023-10-24 [1] Bioconductor
 S4Vectors              * 0.40.2      2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 sass                     0.4.8       2023-12-06 [1] CRAN (R 4.3.1)
 ScaledMatrix             1.10.0      2023-11-06 [1] Bioconductor
 scales                   1.3.0       2023-11-28 [1] CRAN (R 4.3.2)
 scater                   1.30.1      2023-11-16 [1] Bioconductor
 scuttle                  1.12.0      2023-10-24 [1] Bioconductor
 sessioninfo              1.2.2       2021-12-06 [1] CRAN (R 4.3.0)
 shiny                    1.8.0       2023-11-17 [1] CRAN (R 4.3.1)
 shinyWidgets             0.8.1       2024-01-10 [1] CRAN (R 4.3.1)
 SingleCellExperiment   * 1.24.0      2023-11-06 [1] Bioconductor
 spam                     2.10-0      2023-10-23 [1] CRAN (R 4.3.2)
 SparseArray              1.2.3       2023-12-25 [1] Bioconductor 3.18 (R 4.3.2)
 sparseMatrixStats        1.14.0      2023-10-24 [1] Bioconductor
 SpatialExperiment      * 1.12.0      2023-10-26 [1] Bioconductor
 spatialLIBD              1.14.1      2023-11-30 [1] Bioconductor 3.18 (R 4.3.2)
 statmod                  1.5.0       2023-01-06 [1] CRAN (R 4.3.2)
 stringi                  1.8.3       2023-12-11 [1] CRAN (R 4.3.1)
 stringr                  1.5.1       2023-11-14 [1] CRAN (R 4.3.1)
 SummarizedExperiment   * 1.32.0      2023-11-06 [1] Bioconductor
 tibble                   3.2.1       2023-03-20 [1] CRAN (R 4.3.2)
 tidyr                    1.3.1       2024-01-24 [1] CRAN (R 4.3.1)
 tidyselect               1.2.0       2022-10-10 [1] CRAN (R 4.3.0)
 utf8                     1.2.4       2023-10-22 [1] CRAN (R 4.3.2)
 vctrs                    0.6.5       2023-12-01 [1] CRAN (R 4.3.2)
 vipor                    0.4.7       2023-12-18 [1] CRAN (R 4.3.1)
 viridis                  0.6.5       2024-01-29 [1] CRAN (R 4.3.2)
 viridisLite              0.4.2       2023-05-02 [1] CRAN (R 4.3.0)
 withr                    3.0.0       2024-01-16 [1] CRAN (R 4.3.1)
 xfun                     0.41        2023-11-01 [1] CRAN (R 4.3.2)
 XML                      3.99-0.16.1 2024-01-22 [1] CRAN (R 4.3.1)
 xml2                     1.3.6       2023-12-04 [1] CRAN (R 4.3.2)
 xtable                   1.8-4       2019-04-21 [1] CRAN (R 4.3.0)
 XVector                * 0.42.0      2023-10-24 [1] Bioconductor
 yaml                     2.3.8       2023-12-11 [1] CRAN (R 4.3.2)
 zlibbioc                 1.48.0      2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

────────────────────────────────────────────────────────────────────────
