---
title: "Fig 6-Spat Reg"
author: "Bernard Mulvey"
date: "2023-12-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(ggplot2)
library(gridExtra)
library(pheatmap)
require(colorout)
library(ggrepel)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 10), axis.title.x = element_text(size = 11), axis.text.y = element_text(size = 10), axis.title.y = element_text(size=11), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=9), legend.title = element_text(size=10,hjust=0.5)))
```


### 6A: scCoco results in heatmap format
```{r}
cocores <- readRDS("data/10-Spatial Registration/01-scCoco query of top 150 markers per k15-20-31-collapsedK15 domain in ABA ms hyp CCF areas.RDS")

cocores <- cocores[[4]]
cocores <- as.data.table(cocores[["scores_per_target_level_region_all"]])

setnames(cocores,c(3:15),gsub(names(cocores)[3:15],pattern="k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_(X.*)",replacement="\\1"))
setnames(cocores,c("XVMH","XARC"),c("VMH","ARC"))
cocores <- cocores[,.(topnode,topname,VMH,ARC,X1, X2, X4, X6, X7, X8, X9, X10, X13, X14, X15)]

keepres <- c("")
i <- 3
for (i in c(3:15)){
    curvar <- names(cocores)[i]
    setorderv(cocores,curvar,-1)
    keepres <- c(keepres,cocores[c(1:3),topname])
}
keepres <- unique(keepres)[keepres!=""]
rm(i,curvar)

cocores <- melt(cocores[topname %in% keepres,],id.vars=c("topnode","topname"),variable.name="region",value.name="score")
setnames(cocores,"score","scCoco Score (top\n150 Visium Markers)")
cocores[topname=="Periventricular hypothalamic nucleus, posterior part",topname:="Posterior periventricular nu."]

pdf("plots/Manuscript/6A-scCoco registration to adult mouse ISH of hyp.pdf",width=7,height=4)
ggplot(cocores,aes(x=region,y=topname,fill=`scCoco Score (top\n150 Visium Markers)`))+
    geom_tile()+
    scale_fill_gradient2(low="white",mid="yellow",high="red",midpoint=0.5)+
    theme(axis.text.x = element_text(size = 9,angle = 90, hjust = 1))+
    scale_x_discrete(expand = c(0,0))+
    scale_y_discrete(expand=c(0,0))+
    ylab("Allen Mouse Atlas Region Name")+
    xlab("Visium Cluster")+
    theme(axis.title.x = element_text(size = 10), axis.text.y = element_text(size = 9), axis.title.y = element_text(size=11),legend.title = element_text(size=10,hjust=0.5),legend.text=element_text(size=9))
dev.off()
```

### 6B (Spearman) and 6C or Supp Fig (Pearson): sex DE logFC correlations between mouse genes sex DE for each subclass in HYP dissections from Yao 23 and human orthologs in each domain
```{r}
correls <- fread("data/12-Mouse comparisons/03-Yao23-subclass-MvFsexDE_logFCcorrels_to_singleSpDom-MvF.txt")

## 
correls[human_domain=="X11",human_domain:="VMH.1"]
correls[human_domain=="X12",human_domain:="VMH.2"]
correls[human_domain=="X3",human_domain:="ARC.1"]
correls[human_domain=="X5",human_domain:="ARC.2"]
correls[human_domain=="X1",human_domain:="WM.1"]
correls[human_domain=="X9",human_domain:="WM.2"]

correls <- correls[human_domain %in% c("VMH.1","VMH.2","ARC.1","ARC.2","WM.1","WM.2")]

pccs <- dcast(correls, mouse_subclass ~ human_domain, value.var = "pcc")
spears <- dcast(correls, mouse_subclass ~ human_domain, value.var="spearmancor")

pccs <- as.data.frame(pccs)
rownames(pccs) <- pccs$mouse_subclass
pccs$mouse_subclass <- NULL

spears <- as.data.frame(spears)
rownames(spears) <- spears$mouse_subclass
spears$mouse_subclass <- NULL

pdf("plots/Manuscript/6C-Yaosubclass_sexDE_vs_visium_Pearsoncors.pdf",height=6,width=3.5)
pheatmap(as.matrix(pccs),treeheight_col=0,treeheight_row = 0)
dev.off()


pdf("plots/Manuscript/6B-Yaosubclass_sexDE_vs_visium_spearmancors.pdf",height=10,width=7)
pheatmap(as.matrix(spears),treeheight_col=0,treeheight_row = 0)
dev.off()
```

### 6B alt: scatter plot per Visium cluster, with each dot representing a Yao23 subclass
```{r}
### make cell type names more point-label friendly for the cell types that will get labeled: 214, 047, 054, 088, 096, 250, 16
correls[mouse_subclass == "x214_LGv_SPFp_SPFm_Gata3_Gaba", mouse_subclass :=
            "x214_LGv_SPFp_\nSPFm_Gata3_Gaba"]
correls[mouse_subclass == "x047_NDB_SI_MA_STRv_Lhx8_Gaba", mouse_subclass :=
            "x047_NDB_SI_MA_\nSTRv_Lhx8_Gaba"]
correls[mouse_subclass == "x096_SI_MA_NDB_Six3_Meis2_Gaba", mouse_subclass :=
            "x096_SI_MA_NDB_\nSix3_Meis2_Gaba"]
correls[mouse_subclass == "x250_SNc_VTA_RAmb_Foxa1_Dopa", mouse_subclass :=
            "x250_SNc_VTA_\nRAmb_Foxa1_Dopa"]
correls[mouse_subclass == "x116_PVpo_VMPO_MPN_Hmx2_Gaba", mouse_subclass :=
            "x116_PVpo_VMPO_\nMPN_Hmx2_Gaba"]



### need to change something about the invisible bounds around the ggrepel text overexpanding the x-axis?

## loop through the six human domains and write the corresponding plot
i <- 1
for (i in c(1:length(unique(correls$human_domain)))) {
    curdat <-
        copy(correls[human_domain == unique(correls$human_domain)[i]])
    setorderv(curdat, c("human_domain", "spearmancor"))
    curdat[, mouse_subclass := factor(mouse_subclass, levels = as.character(curdat$mouse_subclass))] # factor levels should now match the row order of the data table
    curdat[, subcl_label := as.character(NA)]
    curdat[c(1, nrow(curdat)), subcl_label := mouse_subclass]
    
    ## create a numerical x coordinate system to work with ggrepel more easily (since we're not using axis labels, doesn't matter what's actually composing the x axis udner the hood)
    curdat[, xcoord := c(1:nrow(curdat))]
    
    plts <- ggplot(curdat, aes(x = xcoord, y = spearmancor)) +
        geom_point() +
        ylim(-0.5, 0.5) +
        geom_text_repel(
            aes(label = ifelse(is.na(subcl_label), NA, subcl_label)),
            nudge_x = -35 * sign(curdat$spearmancor),
            nudge_y = 0.1 * sign(curdat$spearmancor),
            size = 2,
            box.padding = 0,
            point.padding = 0,
            min.segment.length = 0
        ) +
        xlim(1, nrow(curdat)) +
        # can just place axis titles centered along all these panels in illustrator.
        #xlab("Yao '23 Mouse HYP Subclass")+
        #ylab("Spearman Correlation (Mouse\nvs Human Sex DE logFC)")+
        ggtitle(unique(correls$human_domain)[i]) +
        theme(
            axis.text.x = element_blank(),
            axis.text.y = element_text(size = 9),
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            plot.title = element_text(size = 11),
            axis.ticks.length.x = unit(0, "pt"),
            panel.grid.major.x = element_blank(),
            panel.grid.minor.x = element_blank()
        )
    
    ### save
    pdf(
        paste0(
            "plots/Manuscript/6B_alt-YaoSexDEcorrels_",
            unique(correls$human_domain)[i],
            ".pdf"
        ),
        height = 2.3,
        width = 2.3
    )
    print(plts)
    dev.off()
    rm(plts,curdat)
}
```

```{r}
sessionInfo()
sessioninfo::session_info()
```

R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggrepel_0.9.5       pheatmap_1.0.12     BiocParallel_1.36.0
 [4] parallelly_1.36.0   default_1.0.0       colorout_1.3-0.1   
 [7] gridExtra_2.3       ggplot2_3.4.4       Biostrings_2.70.1  
[10] GenomeInfoDb_1.38.5 XVector_0.42.0      IRanges_2.36.0     
[13] S4Vectors_0.40.2    BiocGenerics_0.48.1 data.table_1.14.10 
[16] rlang_1.1.3        

loaded via a namespace (and not attached):
 [1] utf8_1.2.4              generics_0.1.3          bitops_1.0-7           
 [4] digest_0.6.34           magrittr_2.0.3          evaluate_0.23          
 [7] grid_4.3.2              RColorBrewer_1.1-3      fastmap_1.1.1          
[10] rprojroot_2.0.4         sessioninfo_1.2.2       limma_3.58.1           
[13] fansi_1.0.6             scales_1.3.0            codetools_0.2-19       
[16] cli_3.6.2               crayon_1.5.2            munsell_0.5.0          
[19] withr_3.0.0             yaml_2.3.8              tools_4.3.2            
[22] parallel_4.3.2          dplyr_1.1.4             colorspace_2.1-0       
[25] GenomeInfoDbData_1.2.11 here_1.0.1              vctrs_0.6.5            
[28] R6_2.5.1                lifecycle_1.0.4         zlibbioc_1.48.0        
[31] pkgconfig_2.0.3         pillar_1.9.0            gtable_0.3.4           
[34] Rcpp_1.0.12             glue_1.7.0              statmod_1.5.0          
[37] xfun_0.41               tibble_3.2.1            tidyselect_1.2.0       
[40] rstudioapi_0.15.0       knitr_1.45              farver_2.1.1           
[43] htmltools_0.5.7         labeling_0.4.3          rmarkdown_2.25         
[46] compiler_4.3.2          RCurl_1.98-1.14        
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-01-24
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────
 package          * version   date (UTC) lib source
 BiocGenerics     * 0.48.1    2023-11-02 [1] Bioconductor
 BiocParallel     * 1.36.0    2023-10-26 [1] Bioconductor
 Biostrings       * 2.70.1    2023-10-25 [1] Bioconductor
 bitops             1.0-7     2021-04-24 [1] CRAN (R 4.3.2)
 cli                3.6.2     2023-12-11 [1] CRAN (R 4.3.2)
 codetools          0.2-19    2023-02-01 [1] CRAN (R 4.3.0)
 colorout         * 1.3-0.1   2024-01-11 [1] local
 colorspace         2.1-0     2023-01-23 [1] CRAN (R 4.3.2)
 crayon             1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 data.table       * 1.14.10   2023-12-08 [1] CRAN (R 4.3.2)
 default          * 1.0.0     2017-08-07 [1] CRAN (R 4.3.0)
 digest             0.6.34    2024-01-11 [1] CRAN (R 4.3.2)
 dplyr              1.1.4     2023-11-17 [1] CRAN (R 4.3.2)
 evaluate           0.23      2023-11-01 [1] CRAN (R 4.3.1)
 fansi              1.0.6     2023-12-08 [1] CRAN (R 4.3.2)
 farver             2.1.1     2022-07-06 [1] CRAN (R 4.3.2)
 fastmap            1.1.1     2023-02-24 [1] CRAN (R 4.3.2)
 generics           0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb     * 1.38.5    2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData   1.2.11    2024-01-11 [1] Bioconductor
 ggplot2          * 3.4.4     2023-10-12 [1] CRAN (R 4.3.1)
 ggrepel          * 0.9.5     2024-01-10 [1] CRAN (R 4.3.2)
 glue               1.7.0     2024-01-09 [1] CRAN (R 4.3.2)
 gridExtra        * 2.3       2017-09-09 [1] CRAN (R 4.3.0)
 gtable             0.3.4     2023-08-21 [1] CRAN (R 4.3.0)
 here               1.0.1     2020-12-13 [1] CRAN (R 4.3.0)
 htmltools          0.5.7     2023-11-03 [1] CRAN (R 4.3.2)
 IRanges          * 2.36.0    2023-10-24 [1] Bioconductor
 knitr              1.45      2023-10-30 [1] CRAN (R 4.3.2)
 labeling           0.4.3     2023-08-29 [1] CRAN (R 4.3.0)
 lifecycle          1.0.4     2023-11-07 [1] CRAN (R 4.3.1)
 limma              3.58.1    2023-10-31 [1] Bioconductor
 magrittr           2.0.3     2022-03-30 [1] CRAN (R 4.3.2)
 munsell            0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 parallelly       * 1.36.0    2023-05-26 [1] CRAN (R 4.3.0)
 pheatmap         * 1.0.12    2019-01-04 [1] CRAN (R 4.3.0)
 pillar             1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig          2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 R6                 2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 RColorBrewer       1.1-3     2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp               1.0.12    2024-01-09 [1] CRAN (R 4.3.2)
 RCurl              1.98-1.14 2024-01-09 [1] CRAN (R 4.3.1)
 rlang            * 1.1.3     2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown          2.25      2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot          2.0.4     2023-11-05 [1] CRAN (R 4.3.1)
 rstudioapi         0.15.0    2023-07-07 [1] CRAN (R 4.3.0)
 S4Vectors        * 0.40.2    2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 scales             1.3.0     2023-11-28 [1] CRAN (R 4.3.2)
 sessioninfo        1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 statmod            1.5.0     2023-01-06 [1] CRAN (R 4.3.2)
 tibble             3.2.1     2023-03-20 [1] CRAN (R 4.3.2)
 tidyselect         1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 utf8               1.2.4     2023-10-22 [1] CRAN (R 4.3.2)
 vctrs              0.6.5     2023-12-01 [1] CRAN (R 4.3.2)
 withr              3.0.0     2024-01-16 [1] CRAN (R 4.3.1)
 xfun               0.41      2023-11-01 [1] CRAN (R 4.3.2)
 XVector          * 0.42.0    2023-10-24 [1] Bioconductor
 yaml               2.3.8     2023-12-11 [1] CRAN (R 4.3.2)
 zlibbioc           1.48.0    2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

───────────────────────────────────────────────────────────────────────────
