---
title: "Fig 3 Panels"
author: "Bernard Mulvey"
date: "2023-12-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table) # Preferred data manipulation package
library(Biostrings) # General
library(ggrastr) # render plots with bajillions of points as raster so they can be manipulated in eg Illustrator
library(Cairo) # pdf device for ^ to properly save
library(ggplot2) # Dependency for several plotting functions
library(gridExtra)
library(SpatialExperiment) # Dependency
library(escheR) # spotplots with domain outlines
require(colorout) # Utility for RStudio
ColorOut() 

# code reformatting in Rstudio
options("styler.addins_style_transformer" = "biocthis::bioc_style()") 

# set plotting defaults for ggplot
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 10), axis.title.x = element_text(size = 11), axis.text.y = element_text(size = 10), axis.title.y = element_text(size =11), plot.title = element_text(size = 12,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=9), legend.title = element_text(size=10,hjust=0.5)))
```

### Setup: load data, cluster assignments, and rename from generic to appropriate region (VMH/ARC/WM)
```{r}
hyp2 <- readRDS("data/03-QC_filters/hypN9_umi275_gene166_chrm50_lognorm_111723.RDS")

bscl <- fread("data/06-BayesSpace/02-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=paste0("X",c(1:15)))

### relabel as VMH/ARC/WM for spotplottings
tmpcd <- as.data.table(colData(hyp2),keep.rownames=T)
tmpcd[label %in% c("X11","X12"),label:="VMH"]
tmpcd[label %in% c("X3","X5"),label:="ARC"]
tmpcd[label %in% c("X1","X9"),label:="WM"]

tmpcd <- DataFrame(tmpcd)
rownames(tmpcd) <- tmpcd$rn
tmpcd <- tmpcd[colnames(hyp2),]
colData(hyp2) <- tmpcd
hyp2$label <- factor(hyp2$label,levels=c("VMH","ARC","WM",paste0("X",c(2,4,6:8,10,13:15))))

### also define cluster tables that only have VMH or ARC and otehrs as NA for escher plotting outlines specific to VMH/ARC
bscl.vmhonly <- as.data.table(bscl,keep.rownames=T)
bscl.arconly <- as.data.table(bscl,keep.rownames=T)
bscl.vmharc <- as.data.table(bscl,keep.rownames=T)

bscl.vmhonly <- bscl.vmhonly[cl %in% c("X11","X12"),cl2:="VMH"]
bscl.arconly <- bscl.arconly[cl %in% c("X3","X5"),cl2:="ARC"]
bscl.vmharc <- bscl.vmharc[cl %in% c("X11","X12"),cl2:="VMH"]
bscl.vmharc <- bscl.vmharc[cl %in% c("X3","X5"),cl2:="ARC"]

rm(tmpcd)
```

### Panel A ANKRD34B (VMH) ###
```{r}
### extract sample to plot ###
ank <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(ank) <- rowData(ank)$gene_name
colData(ank)$`log counts` <- logcounts(ank)["ANKRD34B",]
anklab <- copy(bscl.vmharc)[rn %in% colnames(ank)]
anklab <- DataFrame(anklab)
rownames(anklab) <- anklab$rn

ank$Domain <- factor(anklab$cl2,levels=c("VMH","ARC"))
### make spotplots
p <- make_escheR(ank)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3A-ANKRD34B.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("purple","darkgreen"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("ANKRD34B (V12Y31-\n080_A1 - Male)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0,-0.05,0,unit = "in"))
dev.off()

rm(ank,p,anklab)
```


### Panel B LAMP5 (VMH) ###
```{r}
### extract sample to plot ###
lamp <- hyp2[,hyp2$sample_id=="V12D05-350_D1"]
rownames(lamp) <- rowData(lamp)$gene_name
colData(lamp)$`log counts` <- logcounts(lamp)["LAMP5",]
lamplab <- copy(bscl.vmharc)[rn %in% colnames(lamp)]
lamplab <- DataFrame(lamplab)
rownames(lamplab) <- lamplab$rn

lamp$Domain <- factor(lamplab$cl2,levels=c("VMH","ARC"))
### make spotplots
p <- make_escheR(lamp)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3B-LAMP5.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("purple","darkgreen"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("LAMP5 (V12D05-\n350_D1 - Female)")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(lamp,lamplab,p)
```

### Panel C GABRE (ARC) ###
```{r}
### extract sample to plot ###
gabre <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(gabre) <- rowData(gabre)$gene_name
colData(gabre)$`log counts` <- logcounts(gabre)["GABRE",]
gabrelab <- copy(bscl.vmharc)[rn %in% colnames(gabre)]
gabrelab <- DataFrame(gabrelab)
rownames(gabrelab) <- gabrelab$rn

gabre$Domain <- factor(gabrelab$cl2,levels=c("VMH","ARC"))
### make spotplots
p <- make_escheR(gabre)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3C-GABRE.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("purple","darkgreen"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("GABRE")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(gabre,gabrelab,p)
```


### Panel D CYP26A1 ###
```{r}
### extract sample to plot ###
cyp <- hyp2[,hyp2$sample_id=="V12D05-350_D1"]
rownames(cyp) <- rowData(cyp)$gene_name
colData(cyp)$`log counts` <- logcounts(cyp)["CYP26A1",]
cyplab <- copy(bscl.vmharc)[rn %in% colnames(cyp)]
cyplab <- DataFrame(cyplab)
rownames(cyplab) <- cyplab$rn

cyp$Domain <- factor(cyplab$cl2,levels=c("VMH","ARC"))
### make spotplots
p <- make_escheR(cyp)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3D-CYP26A1.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("purple","darkgreen"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("CYP26A1")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(cyp,cyplab,p)
```

### ISH panels follow D ###
###########################

### intra-VMH SVG: NMU ###
### since we show both ARC domains in the other SVG panel (GAL), we will here too ###
```{r}
### extract sample to plot ###
nmu <- hyp2[,hyp2$sample_id=="V12D05-350_D1"]
rownames(nmu) <- rowData(nmu)$gene_name
colData(nmu)$`log counts` <- logcounts(nmu)["NMU",]
nmulab <- copy(as.data.table(bscl))[rn %in% colnames(nmu)]
nmulab[cl=="X3",cl:="ARC.1"]
nmulab[cl=="X5",cl:="ARC.2"]
nmulab[cl %in% c("X11","X12"),cl:="VMH"]
nmulab[!(cl %in% c("ARC.1","ARC.2","VMH")),cl:="NA"]
nmulab <- DataFrame(nmulab,row.names=nmulab$rn)

nmu$Domain <- factor(nmulab$cl,levels=c("VMH","ARC.1","ARC.2"))
### make spotplots
p <- make_escheR(nmu)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3X-NMU-VMHsvg.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("purple","darkgreen","green"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("NMU")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(nmu,nmulab,p)
```

### intra-ARC SVG: GAL ### 
### show this one with both ARC boundaries illustrated in different colors to get across the non-domain-hewing expression pattern
```{r}
gal <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(gal) <- rowData(gal)$gene_name
colData(gal)$`log counts` <- logcounts(gal)["GAL",]

gallab <- copy(as.data.table(bscl))[rn %in% colnames(gal)]
gallab[cl=="X3",cl:="ARC.1"]
gallab[cl=="X5",cl:="ARC.2"]
gallab[cl %in% c("X11","X12"),cl:="VMH"]
gallab[!(cl %in% c("ARC.1","ARC.2","VMH")),cl:="NA"]

gallab <- DataFrame(gallab)
rownames(gallab) <- gallab$rn

gal$Domain <- factor(gallab$cl,levels=c("VMH","ARC.1","ARC.2"))
### make spotplots
p <- make_escheR(gal)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.6)
p <- p |> add_ground(var = "Domain",stroke=0.15,point_size = 0.6)

pdf("plots/Manuscript/3X-GAL-ARCsvg.pdf",height=1.75,width=2)
p+
    scale_color_manual(values=c("purple","darkgreen","green"),na.value = NA)+
    scale_fill_gradient(low="white",high="black")+
    ggtitle("GAL")+
    guides(color="none")+
    labs(fill="log\ncounts")+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9,hjust=0.5),plot.margin = margin(-0.05,0.03,-0.05,0,unit = "in"))
dev.off()

rm(gal,gallab,p)

```

### for escher plots, extract two legends: one for the collapsed domains to be shared across all 4 panels and one for the split arc (SVG) domains and plot each once so they aren't taking up space. (plus, they're impossible to read as open circles so we can change that here too).
```{r}
### custom function to extrct legend from one plot to extend to all ###
extract_legend <- function(my_ggp) {
  step1 <- ggplot_gtable(ggplot_build(my_ggp))
  step2 <- which(sapply(step1$grobs, function(x) x$name) == "guide-box")
  step3 <- step1$grobs[[step2]]
  return(step3)
}

### run the first part of the code used to generate the marker panels but with the fill legend changed to be the domains factored in the same order
gabre <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(gabre) <- rowData(gabre)$gene_name
colData(gabre)$`log counts` <- logcounts(gabre)["GABRE",]
gabrelab <- copy(bscl.vmharc)[rn %in% colnames(gabre)]
gabrelab <- DataFrame(gabrelab)
rownames(gabrelab) <- gabrelab$rn

gabre$Domain <- factor(gabrelab$cl2,levels=c("VMH","ARC"))
### make a ggplot of nothing in particular, we just want the legend
p <- ggplot(as.data.table(colData(gabre))[!is.na(Domain)],aes(x=sum_umi,y=expr_chrM_ratio,color=Domain))+
    geom_point()+
    scale_color_manual(values=c("purple","darkgreen"),na.value = NA)+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9),plot.margin = margin(-0.3,0,-0.1,0,unit = "in"))

pl <- extract_legend(p)
dev.off()


# save in dimension ~equiv to the dimensions of a legend included in one panel
pdf("plots/Manuscript/3A-3D shared legend.pdf",height=0.6,width=0.45)
do.call("grid.arrange",list(pl))
dev.off()

rm(gabre,gabrelab)
```

## make a 3-domain ARC legend as well
```{r}
### run the first part of the code used to generate the marker panels but with the fill legend changed to be the domains factored in the same order
gal <- hyp2[,hyp2$sample_id=="V12Y31-080_A1"]
rownames(gal) <- rowData(gal)$gene_name
colData(gal)$`log counts` <- logcounts(gal)["GAL",]

gallab <- copy(as.data.table(bscl))[rn %in% colnames(gal)]
gallab[cl=="X3",cl:="ARC.1"]
gallab[cl=="X5",cl:="ARC.2"]
gallab[cl %in% c("X11","X12"),cl:="VMH"]
gallab[!(cl %in% c("ARC.1","ARC.2","VMH")),cl:="NA"]

gallab <- DataFrame(gallab)
rownames(gallab) <- gallab$rn

gal$Domain <- factor(gallab$cl,levels=c("VMH","ARC.1","ARC.2"))
### make a ggplot of nothing in particular, we just want the legend
p <- ggplot(as.data.table(colData(gal))[!is.na(Domain)],aes(x=sum_umi,y=expr_chrM_ratio,color=Domain))+
    geom_point()+
    scale_color_manual(values=c("purple","darkgreen","green"),na.value = NA)+
    theme(plot.title.position = "plot",plot.title = element_text(size=10,hjust=0.5),legend.text = element_text(size=8),legend.key.size = unit(0.125,"in"),legend.title=element_text(size=9),plot.margin = margin(-0.3,0,-0.1,0,unit = "in"))

pl <- extract_legend(p)
dev.off()

# save in dimension ~equiv to the dimensions of a legend included in one panel
pdf("plots/Manuscript/3E-3F shared legend.pdf",height=0.75,width=0.55)
do.call("grid.arrange",list(pl))
dev.off()

rm(p,pl,gallab,gal)
```


### msigdb GSEA for arc and vmh ###
```{r}
mkres <- readRDS("data/11-GSEA/01c-Marker mSigDB_1,2,3,5,6,8 GSEA clpspath mainPathways-only.RDS")

pltdat <- mkres[["k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XVMH"]][pathway %in% c("GOCC_GLUTAMATERGIC_SYNAPSE","GOCC_GABA_ERGIC_SYNAPSE","HP_HYPERACTIVITY","HP_ABNORMAL_AGGRESIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR","WP_KISSPEPTINKISSPEPTIN_RECEPTOR_SYSTEM_IN_THE_OVARY","GOBP_AXONEME_ASSEMBLY","GOCC_CILIARY_PLASM","HP_CENTRAL_ADRENAL_INSUFFICIENCY","DESCARTES_FETAL_ADRENAL_SYMPATHOBLASTS","GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY")]
pltdat <- rbind(pltdat,mkres[["k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XARC"]][pathway %in% c("GOCC_GLUTAMATERGIC_SYNAPSE","GOCC_GABA_ERGIC_SYNAPSE","HP_HYPERACTIVITY","HP_ABNORMAL_AGGRESIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR","WP_KISSPEPTINKISSPEPTIN_RECEPTOR_SYSTEM_IN_THE_OVARY","GOBP_AXONEME_ASSEMBLY","GOCC_CILIARY_PLASM","HP_CENTRAL_ADRENAL_INSUFFICIENCY","DESCARTES_FETAL_ADRENAL_SYMPATHOBLASTS","GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY")])

pltdat[,spdom:=gsub(spdom,pattern="^.*_X(.*)$",replacement="\\1")]

### simplify terms and make separate column for data source to color by
pltdat[,genesource:=gsub(pathway,pattern="^(GOCC)_.*$",replacement="GO Cell\nCompartment")]
pltdat[,genesource:=gsub(genesource,pattern="^(GOBP)_.*$",replacement="GO Biol. Process")]
pltdat[,genesource:=gsub(genesource,pattern="^(HP)_.*$",replacement="Human Phenotypes")]
pltdat[,genesource:=gsub(genesource,pattern="^(WP)_.*$",replacement="Wikipathways")]
pltdat[,genesource:=gsub(genesource,pattern="^(DESCARTES)_.*$",replacement="Descartes scRNA-seq\n(Human)")]

### Simplify terms themselves
pltdat[pathway=="GOCC_GLUTAMATERGIC_SYNAPSE",pathway:="Glutamatergic synapse"]
pltdat[pathway=="GOCC_GABA_ERGIC_SYNAPSE",pathway:="GABAergic synapse"]
pltdat[pathway=="HP_HYPERACTIVITY",pathway:="Hyperactivity"]
pltdat[pathway=="HP_ABNORMAL_AGGRESIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR",pathway:="Aggressive/impulsive behavior"]
pltdat[pathway=="WP_KISSPEPTINKISSPEPTIN_RECEPTOR_SYSTEM_IN_THE_OVARY",pathway:="Kisspeptin receptor system (ovarian)"]
pltdat[pathway=="GOBP_AXONEME_ASSEMBLY",pathway:="Axoneme assembly"]
pltdat[pathway=="GOCC_CILIARY_PLASM",pathway:="Ciliary plasma membrane"]
pltdat[pathway=="HP_CENTRAL_ADRENAL_INSUFFICIENCY",pathway:="Central adrenal insufficiency"]
pltdat[pathway=="DESCARTES_FETAL_ADRENAL_SYMPATHOBLASTS",pathway:="Fetal adrenal sympathoblasts"]
pltdat[pathway=="GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY",pathway:="Neuropeptide signaling"]

### make domain a factor so we can match it to prior color schemes
pltdat$spdom <- factor(pltdat$spdom,levels=c("VMH","ARC"))
pltdat$spdom
### q: omg Github copilot just paid for itself thank you!!
# a: you're welcome

### plot:
pdf("plots/Manuscript/3X-marker_msigdb_GSEA.pdf",height=3.5,width=3)
ggplot(pltdat,aes(x=pathway,y=-log10(padj)))+
    geom_point(aes(col=spdom,size=NES))+
    xlab("Term")+
    scale_size(guide="none")+
    # scale_color_manual(values=wc2cols)+
    ylab("-log10\nFDR")+
    scale_color_manual(values=c("purple","darkgreen"),na.value = NA)+
    labs(col="Domain")+
    geom_hline(yintercept=-log10(0.05),linetype="dashed")+
    # guides(col=guide_legend(byrow=T))+
    # # ^ necess. to get spacing to work in theme
    # guides(col=guide_legend(override.aes=list(size=2.5)))+
    guides(col="none")+
    # ^ make legend points bigger
    theme(axis.text.y = element_text(size = 10), axis.title.y=element_text(size=11), axis.title.x = element_text(size = 11), axis.text.x = element_text(size = 10.5,angle=-90,hjust=0,vjust=0.5), legend.text = element_text(size=9), legend.title = element_text(size=10,hjust=0.5),legend.spacing.y = unit(0.125, "in"))
dev.off()



rm(mkres,pltdat)
```

### TF gsea for ARC and VMH ###
### VMH hits of interest: NR5A1_mouse_TRRUST_Transcription_Factors_2019 , PMC10242764-CCHTS-26-1560_SD1.pdf-5-RXRA_RXRA_Rummagene_transcription_factors , PMC8143797-elife-62857-fig4-data2.xlsx-CAMTA1_ChIP_RNAseq_merged-ENSEMBL_CAMTA1_Rummagene_transcription_factors
### ARC: PMC4537001-oncotarget-06-13088-s001.pdf-11-Activators_Of_AR_Transcriptional_Activity_AR_Rummagene_transcription_factors , PMC8093144-Table_3.docx-0-Table-3_Presumptive_AR_Cofactors_In_AR_Pull-Down_Proteins_4_AR_Rummagene_transcription_factors , RCOR3_Transcription_Factor_PPIs
```{r}
mktfres <- readRDS("data/11-GSEA/02a-Marker TF-target GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")


pltdat <- mktfres[["k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XVMH"]][pathway %in% c("NR5A1_mouse_TRRUST_Transcription_Factors_2019" , "PMC10242764-CCHTS-26-1560_SD1.pdf-5-RXRA_RXRA_Rummagene_transcription_factors" , "PMC8143797-elife-62857-fig4-data2.xlsx-CAMTA1_ChIP_RNAseq_merged-ENSEMBL_CAMTA1_Rummagene_transcription_factors")]

pltdat <- rbind(pltdat,mktfres[["k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XARC"]][pathway %in% c("PMC4537001-oncotarget-06-13088-s001.pdf-11-Activators_Of_AR_Transcriptional_Activity_AR_Rummagene_transcription_factors" , "PMC8093144-Table_3.docx-0-Table-3_Presumptive_AR_Cofactors_In_AR_Pull-Down_Proteins_4_AR_Rummagene_transcription_factors" , "INSM1_DEFICIENCY_MOUSE_GSE46139_CREEDSID_GENE_299_DOWN_TF_Perturbations_Followed_by_Expression")])

pltdat[spdom %in% grep(spdom,pattern="VMH",value=T),spdom:="VMH"]
pltdat[spdom %in% grep(spdom,pattern="ARC",value=T),spdom:="ARC"]

## clean names and data sources
pltdat[pathway=="NR5A1_mouse_TRRUST_Transcription_Factors_2019",source:="TRRUST"]
pltdat[pathway=="NR5A1_mouse_TRRUST_Transcription_Factors_2019",pathway:="NR5A1"]
pltdat[pathway=="PMC10242764-CCHTS-26-1560_SD1.pdf-5-RXRA_RXRA_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC10242764-CCHTS-26-1560_SD1.pdf-5-RXRA_RXRA_Rummagene_transcription_factors",pathway:="RXRA"]
pltdat[pathway=="PMC8143797-elife-62857-fig4-data2.xlsx-CAMTA1_ChIP_RNAseq_merged-ENSEMBL_CAMTA1_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC8143797-elife-62857-fig4-data2.xlsx-CAMTA1_ChIP_RNAseq_merged-ENSEMBL_CAMTA1_Rummagene_transcription_factors",pathway:="CAMTA1"]

pltdat[pathway=="PMC4537001-oncotarget-06-13088-s001.pdf-11-Activators_Of_AR_Transcriptional_Activity_AR_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC4537001-oncotarget-06-13088-s001.pdf-11-Activators_Of_AR_Transcriptional_Activity_AR_Rummagene_transcription_factors",pathway:="AR (Activators)"]
pltdat[pathway=="PMC8093144-Table_3.docx-0-Table-3_Presumptive_AR_Cofactors_In_AR_Pull-Down_Proteins_4_AR_Rummagene_transcription_factors",source:="Rummagene"]
pltdat[pathway=="PMC8093144-Table_3.docx-0-Table-3_Presumptive_AR_Cofactors_In_AR_Pull-Down_Proteins_4_AR_Rummagene_transcription_factors",pathway:="AR (Cofactors)"]
pltdat[pathway=="INSM1_DEFICIENCY_MOUSE_GSE46139_CREEDSID_GENE_299_DOWN_TF_Perturbations_Followed_by_Expression",source:="GEO TF Perturbations"]
pltdat[pathway=="INSM1_DEFICIENCY_MOUSE_GSE46139_CREEDSID_GENE_299_DOWN_TF_Perturbations_Followed_by_Expression",pathway:="INSM1 (KO DEGs)"]

# reorder domains so we can easily assign colors matching the general scheme
pltdat[,spdom:=factor(spdom,levels=c("VMH","ARC"))]

## addd a newline to GEO TF Perturbations to save whitespace
pltdat[source=="GEO TF Perturbations",source:="GEO TF\n Perturbations"]

pdf("plots/Manuscript/3X-marker_TFtargs_GSEA.pdf",height=2,width=4)
ggplot(pltdat,aes(y=pathway,x=-log10(padj)))+
    geom_point(aes(col=spdom,size=NES,shape=source))+
    ylab("Term")+
    scale_size(guide="none")+
    scale_color_manual(values=c("purple","darkgreen"),na.value = NA)+
    xlab("-log10 (GSEA FDR)")+
    labs(col="Domain",shape="Source Database")+
    geom_vline(xintercept=-log10(0.05),linetype="dashed")+
    guides(col=guide_legend(byrow=T))+
    # ^ necess. to get spacing to work in theme
    guides(col=guide_legend(override.aes=list(size=2.5)),
           shape=guide_legend(override.aes=list(size = 2.5)))+
    # ^ make legend points bigger
    theme(axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 11), axis.text.x = element_text(size = 10), legend.text = element_text(size=9,hjust = 0.5), legend.title = element_text(size=10,hjust=0.5),legend.spacing.y = unit(0.005, "in"),axis.title.y=element_text(size=11))
dev.off()

rm(pltdat,mktfres)
```

### disease gene GSEA ###
```{r}
mkdis <- readRDS("data/11-GSEA/05-Common variant disease gene enrichment in markers.RDS")

# we only labeled collapsed VMH/ARC clusters in the nnsvg10 k=15 harmony lambda null 60kiter run, so we can just grep ARC/VMH and get the relevant (collapsed) clusters from that. wouldn't work for like X3 or something since there's going to be such a cluster name for every clustering run
mkdis <- mkdis[genesets_from=="union" & spdom %in% grep(spdom,pattern="k15",value=T)]
mkdis <- mkdis[spdom %in% grep(spdom,pattern="collapsed",value=T)]
mkdis[,spdom:=gsub(spdom,pattern="k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_",replacement="")]


mkdis[spdom=="XVMH",spdom:="VMH"]
mkdis[spdom=="XARC",spdom:="ARC"]
mkdis[spdom=="X1",spdom:="WM.1"]
mkdis[spdom=="X9",spdom:="WM.2"]

mkdis <- mkdis[spdom %in% c("VMH","ARC","WM.1","WM.2")]
# reorder to plot domains of interest first
mkdis[,spdom:=factor(spdom,levels=c("VMH","ARC","WM.1","WM.2"))]

mkdis[,signif:=ifelse(padj<0.05,yes="*",no="")]
setnames(mkdis,"padj","FDR")

pdf("plots/Manuscript/3X-marker_disease_GSEA.pdf",height=2,width=3.4)
ggplot(mkdis,aes(y=spdom,x=pathway,fill=-log10 (FDR)))+
    geom_tile()+
    geom_text(aes(label=signif),size=2,nudge_y=0)+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    scale_fill_gradient(low="white",high="red")+
    labs(fill="-log10\n(FDR)")+
    theme(plot.background = element_blank(),axis.line = element_blank(),panel.grid = element_blank(),panel.border = element_blank(),axis.text.x = element_text(angle=90,vjust=0.5,hjust=1,size=10),axis.text.y = element_text(size=10),axis.title.x = element_text(size = 11), axis.title.y = element_text(size = 11), legend.text = element_text(size=10), legend.title = element_text(size=10,hjust=0.5),legend.margin = margin(0,0,0,0),legend.key.height = unit(0.125,"in"))+
    xlab("Disease Gene Set")+
    ylab("Domain")
# ^ okay, whatever legend.key.height is using to define "0.25 inches" is wildly off. had to fiddle with that number til it worked
dev.off()
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] escheR_1.2.0                colorout_1.3-0.1           
 [3] scater_1.30.1               scuttle_1.12.0             
 [5] spatialLIBD_1.14.1          SpatialExperiment_1.12.0   
 [7] SingleCellExperiment_1.24.0 SummarizedExperiment_1.32.0
 [9] Biobase_2.62.0              GenomicRanges_1.54.1       
[11] MatrixGenerics_1.14.0       matrixStats_1.2.0          
[13] gridExtra_2.3               ggplot2_3.4.4              
[15] Cairo_1.6-2                 ggrastr_1.0.2              
[17] Biostrings_2.70.1           GenomeInfoDb_1.38.5        
[19] XVector_0.42.0              IRanges_2.36.0             
[21] S4Vectors_0.40.2            BiocGenerics_0.48.1        
[23] data.table_1.14.10          rlang_1.1.3                

loaded via a namespace (and not attached):
  [1] later_1.3.2                   BiocIO_1.12.0                
  [3] bitops_1.0-7                  filelock_1.0.3               
  [5] fields_15.2                   tibble_3.2.1                 
  [7] XML_3.99-0.16                 lifecycle_1.0.4              
  [9] edgeR_4.0.11                  doParallel_1.0.17            
 [11] rprojroot_2.0.4               lattice_0.22-5               
 [13] magrittr_2.0.3                rmarkdown_2.25               
 [15] limma_3.58.1                  plotly_4.10.4                
 [17] sass_0.4.8                    jquerylib_0.1.4              
 [19] yaml_2.3.8                    httpuv_1.6.13                
 [21] spam_2.10-0                   sessioninfo_1.2.2            
 [23] cowplot_1.1.2                 DBI_1.2.1                    
 [25] RColorBrewer_1.1-3            golem_0.4.1                  
 [27] maps_3.4.2                    abind_1.4-5                  
 [29] zlibbioc_1.48.0               purrr_1.0.2                  
 [31] RCurl_1.98-1.14               rappdirs_0.3.3               
 [33] GenomeInfoDbData_1.2.11       ggrepel_0.9.5                
 [35] irlba_2.3.5.1                 pheatmap_1.0.12              
 [37] DelayedMatrixStats_1.24.0     codetools_0.2-19             
 [39] DelayedArray_0.28.0           DT_0.31                      
 [41] tidyselect_1.2.0              farver_2.1.1                 
 [43] ScaledMatrix_1.10.0           viridis_0.6.4                
 [45] shinyWidgets_0.8.1            BiocFileCache_2.10.1         
 [47] GenomicAlignments_1.38.2      jsonlite_1.8.8               
 [49] BiocNeighbors_1.20.2          ellipsis_0.3.2               
 [51] iterators_1.0.14              foreach_1.5.2                
 [53] tools_4.3.2                   Rcpp_1.0.12                  
 [55] glue_1.7.0                    SparseArray_1.2.3            
 [57] xfun_0.41                     here_1.0.1                   
 [59] dplyr_1.1.4                   withr_3.0.0                  
 [61] BiocManager_1.30.22           fastmap_1.1.1                
 [63] fansi_1.0.6                   digest_0.6.34                
 [65] rsvd_1.0.5                    R6_2.5.1                     
 [67] mime_0.12                     colorspace_2.1-0             
 [69] RSQLite_2.3.5                 config_0.3.2                 
 [71] utf8_1.2.4                    tidyr_1.3.0                  
 [73] generics_0.1.3                rtracklayer_1.62.0           
 [75] httr_1.4.7                    htmlwidgets_1.6.4            
 [77] S4Arrays_1.2.0                scatterplot3d_0.3-44         
 [79] pkgconfig_2.0.3               gtable_0.3.4                 
 [81] blob_1.2.4                    htmltools_0.5.7              
 [83] dotCall64_1.1-1               scales_1.3.0                 
 [85] png_0.1-8                     attempt_0.3.1                
 [87] knitr_1.45                    rstudioapi_0.15.0            
 [89] rjson_0.2.21                  curl_5.2.0                   
 [91] cachem_1.0.8                  Polychrome_1.5.1             
 [93] BiocVersion_3.18.1            parallel_4.3.2               
 [95] vipor_0.4.7                   AnnotationDbi_1.64.1         
 [97] restfulr_0.0.15               pillar_1.9.0                 
 [99] grid_4.3.2                    vctrs_0.6.5                  
[101] promises_1.2.1                BiocSingular_1.18.0          
[103] dbplyr_2.4.0                  beachmat_2.18.0              
[105] xtable_1.8-4                  beeswarm_0.4.0               
[107] paletteer_1.5.0               evaluate_0.23                
[109] magick_2.8.2                  cli_3.6.2                    
[111] locfit_1.5-9.8                compiler_4.3.2               
[113] Rsamtools_2.18.0              crayon_1.5.2                 
[115] labeling_0.4.3                rematch2_2.1.2               
[117] ggbeeswarm_0.7.2              viridisLite_0.4.2            
[119] BiocParallel_1.36.0           munsell_0.5.0                
[121] lazyeval_0.2.2                Matrix_1.6-5                 
[123] ExperimentHub_2.10.0          benchmarkme_1.0.8            
[125] sparseMatrixStats_1.14.0      bit64_4.0.5                  
[127] KEGGREST_1.42.0               statmod_1.5.0                
[129] shiny_1.8.0                   interactiveDisplayBase_1.40.0
[131] AnnotationHub_3.10.0          memoise_2.0.1                
[133] bslib_0.6.1                   benchmarkmeData_1.0.4        
[135] bit_4.0.5                    
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-01-21
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────
 package                * version   date (UTC) lib source
 abind                    1.4-5     2016-07-21 [1] CRAN (R 4.3.0)
 AnnotationDbi            1.64.1    2023-11-02 [1] Bioconductor
 AnnotationHub            3.10.0    2023-10-24 [1] Bioconductor
 attempt                  0.3.1     2020-05-03 [1] CRAN (R 4.3.0)
 beachmat                 2.18.0    2023-10-24 [1] Bioconductor
 beeswarm                 0.4.0     2021-06-01 [1] CRAN (R 4.3.2)
 benchmarkme              1.0.8     2022-06-12 [1] CRAN (R 4.3.0)
 benchmarkmeData          1.0.4     2020-04-23 [1] CRAN (R 4.3.0)
 Biobase                * 2.62.0    2023-10-24 [1] Bioconductor
 BiocFileCache            2.10.1    2023-10-26 [1] Bioconductor
 BiocGenerics           * 0.48.1    2023-11-02 [1] Bioconductor
 BiocIO                   1.12.0    2023-10-26 [1] Bioconductor
 BiocManager              1.30.22   2023-08-08 [1] CRAN (R 4.3.0)
 BiocNeighbors            1.20.2    2024-01-07 [1] Bioconductor 3.18 (R 4.3.2)
 BiocParallel             1.36.0    2023-10-26 [1] Bioconductor
 BiocSingular             1.18.0    2023-10-24 [1] Bioconductor
 BiocVersion              3.18.1    2023-11-18 [1] Bioconductor 3.18 (R 4.3.2)
 Biostrings             * 2.70.1    2023-10-25 [1] Bioconductor
 bit                      4.0.5     2022-11-15 [1] CRAN (R 4.3.2)
 bit64                    4.0.5     2020-08-30 [1] CRAN (R 4.3.2)
 bitops                   1.0-7     2021-04-24 [1] CRAN (R 4.3.2)
 blob                     1.2.4     2023-03-17 [1] CRAN (R 4.3.0)
 bslib                    0.6.1     2023-11-28 [1] CRAN (R 4.3.1)
 cachem                   1.0.8     2023-05-01 [1] CRAN (R 4.3.2)
 Cairo                  * 1.6-2     2023-11-28 [1] CRAN (R 4.3.1)
 cli                      3.6.2     2023-12-11 [1] CRAN (R 4.3.2)
 codetools                0.2-19    2023-02-01 [1] CRAN (R 4.3.0)
 colorout               * 1.3-0.1   2024-01-11 [1] local
 colorspace               2.1-0     2023-01-23 [1] CRAN (R 4.3.2)
 config                   0.3.2     2023-08-30 [1] CRAN (R 4.3.0)
 cowplot                  1.1.2     2023-12-15 [1] CRAN (R 4.3.2)
 crayon                   1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 curl                     5.2.0     2023-12-08 [1] CRAN (R 4.3.2)
 data.table             * 1.14.10   2023-12-08 [1] CRAN (R 4.3.2)
 DBI                      1.2.1     2024-01-12 [1] CRAN (R 4.3.2)
 dbplyr                   2.4.0     2023-10-26 [1] CRAN (R 4.3.1)
 DelayedArray             0.28.0    2023-10-24 [1] Bioconductor
 DelayedMatrixStats       1.24.0    2023-11-06 [1] Bioconductor
 digest                   0.6.34    2024-01-11 [1] CRAN (R 4.3.2)
 doParallel               1.0.17    2022-02-07 [1] CRAN (R 4.3.0)
 dotCall64                1.1-1     2023-11-28 [1] CRAN (R 4.3.2)
 dplyr                    1.1.4     2023-11-17 [1] CRAN (R 4.3.2)
 DT                       0.31      2023-12-09 [1] CRAN (R 4.3.1)
 edgeR                    4.0.11    2024-01-19 [1] Bioconductor 3.18 (R 4.3.2)
 ellipsis                 0.3.2     2021-04-29 [1] CRAN (R 4.3.2)
 escheR                 * 1.2.0     2023-10-26 [1] Bioconductor
 evaluate                 0.23      2023-11-01 [1] CRAN (R 4.3.1)
 ExperimentHub            2.10.0    2023-10-26 [1] Bioconductor
 fansi                    1.0.6     2023-12-08 [1] CRAN (R 4.3.2)
 farver                   2.1.1     2022-07-06 [1] CRAN (R 4.3.2)
 fastmap                  1.1.1     2023-02-24 [1] CRAN (R 4.3.2)
 fields                   15.2      2023-08-17 [1] CRAN (R 4.3.2)
 filelock                 1.0.3     2023-12-11 [1] CRAN (R 4.3.2)
 foreach                  1.5.2     2022-02-02 [1] CRAN (R 4.3.0)
 generics                 0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb           * 1.38.5    2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData         1.2.11    2024-01-11 [1] Bioconductor
 GenomicAlignments        1.38.2    2024-01-16 [1] Bioconductor 3.18 (R 4.3.2)
 GenomicRanges          * 1.54.1    2023-10-29 [1] Bioconductor
 ggbeeswarm               0.7.2     2023-04-29 [1] CRAN (R 4.3.0)
 ggplot2                * 3.4.4     2023-10-12 [1] CRAN (R 4.3.1)
 ggrastr                * 1.0.2     2023-06-01 [1] CRAN (R 4.3.0)
 ggrepel                  0.9.5     2024-01-10 [1] CRAN (R 4.3.2)
 glue                     1.7.0     2024-01-09 [1] CRAN (R 4.3.2)
 golem                    0.4.1     2023-06-05 [1] CRAN (R 4.3.0)
 gridExtra              * 2.3       2017-09-09 [1] CRAN (R 4.3.0)
 gtable                   0.3.4     2023-08-21 [1] CRAN (R 4.3.0)
 here                     1.0.1     2020-12-13 [1] CRAN (R 4.3.0)
 htmltools                0.5.7     2023-11-03 [1] CRAN (R 4.3.2)
 htmlwidgets              1.6.4     2023-12-06 [1] CRAN (R 4.3.1)
 httpuv                   1.6.13    2023-12-06 [1] CRAN (R 4.3.1)
 httr                     1.4.7     2023-08-15 [1] CRAN (R 4.3.0)
 interactiveDisplayBase   1.40.0    2023-10-26 [1] Bioconductor
 IRanges                * 2.36.0    2023-10-24 [1] Bioconductor
 irlba                    2.3.5.1   2022-10-03 [1] CRAN (R 4.3.2)
 iterators                1.0.14    2022-02-05 [1] CRAN (R 4.3.0)
 jquerylib                0.1.4     2021-04-26 [1] CRAN (R 4.3.0)
 jsonlite                 1.8.8     2023-12-04 [1] CRAN (R 4.3.2)
 KEGGREST                 1.42.0    2023-10-26 [1] Bioconductor
 knitr                    1.45      2023-10-30 [1] CRAN (R 4.3.2)
 labeling                 0.4.3     2023-08-29 [1] CRAN (R 4.3.0)
 later                    1.3.2     2023-12-06 [1] CRAN (R 4.3.2)
 lattice                  0.22-5    2023-10-24 [1] CRAN (R 4.3.2)
 lazyeval                 0.2.2     2019-03-15 [1] CRAN (R 4.3.2)
 lifecycle                1.0.4     2023-11-07 [1] CRAN (R 4.3.1)
 limma                    3.58.1    2023-10-31 [1] Bioconductor
 locfit                   1.5-9.8   2023-06-11 [1] CRAN (R 4.3.2)
 magick                   2.8.2     2023-12-20 [1] CRAN (R 4.3.1)
 magrittr                 2.0.3     2022-03-30 [1] CRAN (R 4.3.2)
 maps                     3.4.2     2023-12-15 [1] CRAN (R 4.3.1)
 Matrix                   1.6-5     2024-01-11 [1] CRAN (R 4.3.2)
 MatrixGenerics         * 1.14.0    2023-10-26 [1] Bioconductor
 matrixStats            * 1.2.0     2023-12-11 [1] CRAN (R 4.3.2)
 memoise                  2.0.1     2021-11-26 [1] CRAN (R 4.3.0)
 mime                     0.12      2021-09-28 [1] CRAN (R 4.3.2)
 munsell                  0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 paletteer                1.5.0     2022-10-19 [1] CRAN (R 4.3.0)
 pheatmap                 1.0.12    2019-01-04 [1] CRAN (R 4.3.0)
 pillar                   1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig                2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 plotly                   4.10.4    2024-01-13 [1] CRAN (R 4.3.1)
 png                      0.1-8     2022-11-29 [1] CRAN (R 4.3.2)
 Polychrome               1.5.1     2022-05-03 [1] CRAN (R 4.3.0)
 promises                 1.2.1     2023-08-10 [1] CRAN (R 4.3.2)
 purrr                    1.0.2     2023-08-10 [1] CRAN (R 4.3.2)
 R6                       2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs                 0.3.3     2021-01-31 [1] CRAN (R 4.3.2)
 RColorBrewer             1.1-3     2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp                     1.0.12    2024-01-09 [1] CRAN (R 4.3.2)
 RCurl                    1.98-1.14 2024-01-09 [1] CRAN (R 4.3.1)
 rematch2                 2.1.2     2020-05-01 [1] CRAN (R 4.3.0)
 restfulr                 0.0.15    2022-06-16 [1] CRAN (R 4.3.2)
 rjson                    0.2.21    2022-01-09 [1] CRAN (R 4.3.2)
 rlang                  * 1.1.3     2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown                2.25      2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot                2.0.4     2023-11-05 [1] CRAN (R 4.3.1)
 Rsamtools                2.18.0    2023-10-24 [1] Bioconductor
 RSQLite                  2.3.5     2024-01-21 [1] CRAN (R 4.3.2)
 rstudioapi               0.15.0    2023-07-07 [1] CRAN (R 4.3.0)
 rsvd                     1.0.5     2021-04-16 [1] CRAN (R 4.3.0)
 rtracklayer              1.62.0    2023-10-26 [1] Bioconductor
 S4Arrays                 1.2.0     2023-10-24 [1] Bioconductor
 S4Vectors              * 0.40.2    2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 sass                     0.4.8     2023-12-06 [1] CRAN (R 4.3.1)
 ScaledMatrix             1.10.0    2023-11-06 [1] Bioconductor
 scales                   1.3.0     2023-11-28 [1] CRAN (R 4.3.2)
 scater                 * 1.30.1    2023-11-16 [1] Bioconductor
 scatterplot3d            0.3-44    2023-05-05 [1] CRAN (R 4.3.0)
 scuttle                * 1.12.0    2023-10-24 [1] Bioconductor
 sessioninfo              1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 shiny                    1.8.0     2023-11-17 [1] CRAN (R 4.3.1)
 shinyWidgets             0.8.1     2024-01-10 [1] CRAN (R 4.3.1)
 SingleCellExperiment   * 1.24.0    2023-11-06 [1] Bioconductor
 spam                     2.10-0    2023-10-23 [1] CRAN (R 4.3.2)
 SparseArray              1.2.3     2023-12-25 [1] Bioconductor 3.18 (R 4.3.2)
 sparseMatrixStats        1.14.0    2023-10-24 [1] Bioconductor
 SpatialExperiment      * 1.12.0    2023-10-26 [1] Bioconductor
 spatialLIBD            * 1.14.1    2023-11-30 [1] Bioconductor 3.18 (R 4.3.2)
 statmod                  1.5.0     2023-01-06 [1] CRAN (R 4.3.2)
 SummarizedExperiment   * 1.32.0    2023-11-06 [1] Bioconductor
 tibble                   3.2.1     2023-03-20 [1] CRAN (R 4.3.2)
 tidyr                    1.3.0     2023-01-24 [1] CRAN (R 4.3.2)
 tidyselect               1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 utf8                     1.2.4     2023-10-22 [1] CRAN (R 4.3.2)
 vctrs                    0.6.5     2023-12-01 [1] CRAN (R 4.3.2)
 vipor                    0.4.7     2023-12-18 [1] CRAN (R 4.3.1)
 viridis                  0.6.4     2023-07-22 [1] CRAN (R 4.3.0)
 viridisLite              0.4.2     2023-05-02 [1] CRAN (R 4.3.0)
 withr                    3.0.0     2024-01-16 [1] CRAN (R 4.3.1)
 xfun                     0.41      2023-11-01 [1] CRAN (R 4.3.2)
 XML                      3.99-0.16 2023-11-29 [1] CRAN (R 4.3.1)
 xtable                   1.8-4     2019-04-21 [1] CRAN (R 4.3.0)
 XVector                * 0.42.0    2023-10-24 [1] Bioconductor
 yaml                     2.3.8     2023-12-11 [1] CRAN (R 4.3.2)
 zlibbioc                 1.48.0    2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

───────────────────────────────────────────────────────────────────────────
