---
title: "Fig 5-Sex DEG Enrichments"
author: "Bernard Mulvey"
date: "2023-12-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(ggplot2)
library(gridExtra)
require(colorout)
library(ggstance) # for y axis dodge
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))
```

define a consistent sex•region color palette to use across plots
```{r}
sexregpal <- c("VMH Fem. Up"="#af45c4","VMH Male Up"="#59103e","ARC Fem. Up"="#1fde72","ARC Male Up"="#053b05")
```

Sex DE GSEA
```{r}
sexreslist <- readRDS("data/11-GSEA/03a-Sex DE msigdb_1,2,3,5,6,8 GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")

vmhs <- sexreslist[["BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_VMH"]][pathway %in% c("HP_ABNORMAL_AGGRESSIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR", "GOMF_CHROMATIN_BINDING","HP_ABNORMALITY_OF_ACID_BASE_HOMEOSTASIS", "YAO_TEMPORAL_RESPONSE_TO_PROGESTERONE_CLUSTER_17","GABRIELY_MIR21_TARGETS", "MIR30A_3P","MCLACHLAN_DENTAL_CARIES_UP", "FAN_OVARY_CL6_PUTATIVE_EARLY_ATRETIC_FOLLICLE_THECAL_CELL_2")]

arcs <- sexreslist[["BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_ARC"]][pathway %in% c("HP_ABNORMAL_AGGRESSIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR", "GOMF_CHROMATIN_BINDING","HP_ABNORMALITY_OF_ACID_BASE_HOMEOSTASIS", "YAO_TEMPORAL_RESPONSE_TO_PROGESTERONE_CLUSTER_17","GABRIELY_MIR21_TARGETS", "MIR30A_3P","MCLACHLAN_DENTAL_CARIES_UP", "FAN_OVARY_CL6_PUTATIVE_EARLY_ATRETIC_FOLLICLE_THECAL_CELL_2")]

pltdat <- rbind(vmhs,arcs)
pltdat[,spdom:=gsub(spdom,pattern="BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_(.*)",replacement="\\1")]
pltdat[spdom=="VMH",sexreg:=ifelse(NES>0,yes="VMH Male Up",no="VMH Fem. Up")]
pltdat[spdom=="ARC",sexreg:=ifelse(NES>0,yes="ARC Male Up",no="ARC Fem. Up")]
rm(vmhs,arcs)
### extract pathway source db

i<-1
for (i in c(1:nrow(pltdat))){
    if(i==1){gsrc <- unlist(strsplit(pltdat[i,pathway],"_"))[1]}
    else{gsrc <- c(gsrc,
                    unlist(strsplit(pltdat[i,pathway],"_"))[1])}
}

## tidy gene sources; will note publications for published sets in legend.
pltdat$genesource <- gsrc
pltdat[genesource %in% c("FAN","GABRIELY","MCLACHLAN","YAO"),genesource:="MSigDB Literature-Curated"]
pltdat[genesource=="GOMF",genesource:="GO Mol. Fxn"]
pltdat[genesource=="HP",genesource:="Human Phenotypes"]
pltdat[genesource=="MIR30A",genesource:="MSigDB miRNA Targets"]


rm(i,gsrc)
### simplify pathways
pltdat[pathway=="HP_ABNORMAL_AGGRESSIVE_IMPULSIVE_OR_VIOLENT_BEHAVIOR",pathway:="Aggressive/impulsive behavior"]
pltdat[pathway=="GOMF_CHROMATIN_BINDING",pathway:="Chromatin binding"]
pltdat[pathway=="HP_ABNORMALITY_OF_ACID_BASE_HOMEOSTASIS",pathway:="Acid-base abnormality"]
pltdat[pathway=="YAO_TEMPORAL_RESPONSE_TO_PROGESTERONE_CLUSTER_17",pathway:="Temporal progesterone response"]
pltdat[pathway=="GABRIELY_MIR21_TARGETS",pathway:="miR-21 targets"]
pltdat[pathway=="MIR30A_3P",pathway:="miR-30a_3P targets"]
pltdat[pathway=="MCLACHLAN_DENTAL_CARIES_UP",pathway:="Dental caries"]
pltdat[pathway=="FAN_OVARY_CL6_PUTATIVE_EARLY_ATRETIC_FOLLICLE_THECAL_CELL_2",pathway:="Thecal cell (early atretic ovarian follicle)"]

### flip sign on negative NES scores
pltdat[NES<0,NES:=-1*NES]

pdf("plots/Manuscript/5A as 4K-Sex DEG msigdb GSEA.pdf",height=4,width=3)
ggplot(pltdat,aes(x=pathway,y=-log10(padj)))+
    geom_point(aes(col=sexreg,size=NES))+#,shape=genesource))+
    xlab("Term")+
    #Amplify the scaling by NES by 2x for better visualization
    scale_radius(range = c(1.25*min(pltdat$NES),1.25*max(pltdat$NES)),trans = "identity",guide="none")+
    scale_color_manual(values=sexregpal)+
    ylab("-log10 (FDR)")+
    labs(col="DE Gene Set")+
    geom_hline(yintercept=-log10(0.05),linetype="dashed")+
    #scale_shape_manual(values=c("MSigDB Literature-Curated"=16,"GO Mol. Fxn"=15,"Human Phenotypes"=18,"MSigDB miRNA Targets"=17))+
    guides(col=guide_legend(byrow=T))+
    # ^ necess. to get spacing to work in theme
    guides(col=guide_legend(override.aes=list(size=2.5)),
           #shape=guide_legend(override.aes=list(size = 2.5)),
           )+
    # ^ make legend points bigger
    theme(axis.text.y = element_text(size = 9),axis.title.y=element_text(size=10), axis.title.x = element_text(size = 10,margin=margin(-0.15,0,0,0,unit="in")), axis.text.x = element_text(size = 9,angle=90,hjust=1,vjust=0.5,margin=margin(0,0.1,0,0.1,"in")), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5,margin=margin(0,0,-0.1,0,"in")),legend.spacing.y = unit(0.125, "in"),legend.margin = margin(0,0,0,-0.1,unit = "in"))
dev.off()

rm(pltdat,sexreslist)
```

5B-sex de GSEA - TFs

VMH UNIQUE Male:
MYT1L
TCF20 (rare LOF NDD gene Vetrini '19)

VMH UNIQUE Fem:
PPARA
PPARG

VMH shared-ish
ESR1 (KD-down, female; ESR1 (targets, interactors male)
AR (KD-bidir DE, female; targets male)

ARC unique male:
LIN28B (growth phenos)
ASH1L (ID/ASD in mice https://www.nature.com/articles/s42003-021-02282-z )

ARC unique female:
VDR (KO-bidir DE)
ESR1 (KO-bidir DE)

ARC shared: 
STAT3 (required for hypothalamic response to leptin Buettner 06)        


```{r}
sexreslist <- readRDS("data/11-GSEA/09-Sex DE TF-target GSEA results from 04a filtered to TFs expressed in avg 1pct of cluster spots and at least .25pct of cluster spots in 3samps.RDS")

### subset to TFs planned to plot, then get the corresponding gene set names we want to use for plotting those stats
pltdat <- sexreslist[["BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_VMH"]]
pltdat[,spdom:="VMH"]
tmp <- sexreslist[["BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_ARC"]]
tmp[,spdom:="ARC"]
pltdat <- rbind(pltdat,tmp)
rm(tmp)
### tfs to plot
pltdat <- pltdat[tf %in% c("MYT1L","TCF20","PPARA","PPARG","ESR1","AR","LIN28B","ASH1L","VDR","STAT3","ZNF142")]

pways <- unique(c(pltdat[spdom=="VMH"&padj<0.05,pathway],pltdat[spdom=="ARC"&padj<0.05,pathway]))
pltdat <- pltdat[pathway %in% pways]
## reorder columns for ease of viewing
pltdat <- pltdat[,c(10,3,5,6,7,1,2,4,8,9)]
```

browse for particular sets to extract and plot
•MYT1L: PMC6815046-13073_2019_678_MOESM4_ESM.xlsx-MYT1L-Unnamed_0_MYT1L_Rummagene_transcription_factors
•TCF20: PMC6815046-13073_2019_678_MOESM3_ESM.xlsx-TCF20-Unnamed_0_TCF20_Rummagene_transcription_factors
•PPARA: PPARA_KO_MOUSE_GSE6864_CREEDSID_GENE_1372_DOWN_TF_Perturbations_Followed_by_Expression, PPARA_KO_MOUSE_GSE6864_CREEDSID_GENE_1373_UP_TF_Perturbations_Followed_by_Expression
•PPARG: PPARG_OE_MOUSE_GSE2192_CREEDSID_GENE_861_UP_TF_Perturbations_Followed_by_Expression, PPARG_OE_MOUSE_GSE2192_CREEDSID_GENE_865_DOWN_TF_Perturbations_Followed_by_Expression
•ESR1: ESR1_KD_HUMAN_GSE26298_CREEDSID_GENE_479_DOWN_TF_Perturbations_Followed_by_Expression, 	
PMC7162552-pcbi.1007800.s007.txt--ESR1_Interactors_ESR1_Rummagene_transcription_factors, PMC5998130-41467_2018_4619_MOESM10_ESM.xlsx-OHT_ESR1-Unnamed_2_ESR1_Rummagene_transcription_factors
•AR: AR_35413990_ChIP-Seq_Human_ARposCRProstateCancer_ChEA_2022, AR_KD_HUMAN_GSE22483_CREEDSID_GENE_2194_DOWN_TF_Perturbations_Followed_by_Expression, AR_KD_HUMAN_GSE22483_CREEDSID_GENE_2194_UP_TF_Perturbations_Followed_by_Expression
•LIN28B: PMC7335061-41467_2020_17150_MOESM4_ESM.xlsx-Table_S3-LIN28B_Correlates_LIN28B_Rummagene_transcription_factors
•PMC6815046-13073_2019_678_MOESM2_ESM.xlsx-ASH1L-Unnamed_0_ASH1L_Rummagene_transcription_factors
•VDR_KO_MOUSE_GSE2421_CREEDSID_GENE_1153_UP_TF_Perturbations_Followed_by_Expression, VDR_KO_MOUSE_GSE2421_CREEDSID_GENE_1153_DOWN_TF_Perturbations_Followed_by_Expression
•ESR1_KO_MOUSE_GSE53812_CREEDSID_GENE_2898_UP_TF_Perturbations_Followed_by_Expression, ESR1_KO_MOUSE_GSE53812_CREEDSID_GENE_2898_DOWN_TF_Perturbations_Followed_by_Expression
•STAT3: PMC7801680-41598_2020_79740_MOESM9_ESM.xlsx-STAT3_Target_Genes_In_AD1-STAT3_Target_Genes_Predicted_By_CHIP-seq_Overlap_With_Genes_In_AD_M1_STAT3_Rummagene_transcription_factors, STAT3_human_TRRUST_Transcription_Factors_2019
•ZNF142: ZNF142_SIRNA_MCF7_HUMAN_GSE79586_RNASEQ_UP_TF_Perturbations_Followed_by_Expression, ZNF142_SIRNA_MCF7_HUMAN_GSE79586_RNASEQ_DOWN_TF_Perturbations_Followed_by_Expression

```{r}
pltdat <- pltdat[pathway %in% c(
    "PMC6815046-13073_2019_678_MOESM4_ESM.xlsx-MYT1L-Unnamed_0_MYT1L_Rummagene_transcription_factors",
    "PMC6815046-13073_2019_678_MOESM3_ESM.xlsx-TCF20-Unnamed_0_TCF20_Rummagene_transcription_factors",
    "PPARA_KO_MOUSE_GSE6864_CREEDSID_GENE_1372_DOWN_TF_Perturbations_Followed_by_Expression",
    "PPARA_KO_MOUSE_GSE6864_CREEDSID_GENE_1373_UP_TF_Perturbations_Followed_by_Expression",
    "PPARG_OE_MOUSE_GSE2192_CREEDSID_GENE_861_UP_TF_Perturbations_Followed_by_Expression",
    "PPARG_OE_MOUSE_GSE2192_CREEDSID_GENE_865_DOWN_TF_Perturbations_Followed_by_Expression",
    "ESR1_KD_HUMAN_GSE26298_CREEDSID_GENE_479_DOWN_TF_Perturbations_Followed_by_Expression",
    "PMC7162552-pcbi.1007800.s007.txt--ESR1_Interactors_ESR1_Rummagene_transcription_factors",
    "PMC5998130-41467_2018_4619_MOESM10_ESM.xlsx-OHT_ESR1-Unnamed_2_ESR1_Rummagene_transcription_factors",
    "AR_35413990_ChIP-Seq_Human_ARposCRProstateCancer_ChEA_2022",
    "AR_KD_HUMAN_GSE22483_CREEDSID_GENE_2194_DOWN_TF_Perturbations_Followed_by_Expression",
    "AR_KD_HUMAN_GSE22483_CREEDSID_GENE_2194_UP_TF_Perturbations_Followed_by_Expression",
    "PMC7335061-41467_2020_17150_MOESM4_ESM.xlsx-Table_S3-LIN28B_Correlates_LIN28B_Rummagene_transcription_factors",
    "PMC6815046-13073_2019_678_MOESM2_ESM.xlsx-ASH1L-Unnamed_0_ASH1L_Rummagene_transcription_factors",
    "VDR_KO_MOUSE_GSE2421_CREEDSID_GENE_1153_UP_TF_Perturbations_Followed_by_Expression",
    "VDR_KO_MOUSE_GSE2421_CREEDSID_GENE_1153_DOWN_TF_Perturbations_Followed_by_Expression",
    "ESR1_KO_MOUSE_GSE53812_CREEDSID_GENE_2898_UP_TF_Perturbations_Followed_by_Expression",
    "ESR1_KO_MOUSE_GSE53812_CREEDSID_GENE_2898_DOWN_TF_Perturbations_Followed_by_Expression",
    "STAT3_human_TRRUST_Transcription_Factors_2019",
    "PMC7801680-41598_2020_79740_MOESM9_ESM.xlsx-STAT3_Target_Genes_In_AD1-STAT3_Target_Genes_Predicted_By_CHIP-seq_Overlap_With_Genes_In_AD_M1_STAT3_Rummagene_transcription_factors",
    "ZNF142_SIRNA_MCF7_HUMAN_GSE79586_RNASEQ_UP_TF_Perturbations_Followed_by_Expression",
    "ZNF142_SIRNA_MCF7_HUMAN_GSE79586_RNASEQ_DOWN_TF_Perturbations_Followed_by_Expression"
    ),]

## data sourcesr
pltdat[pathway %in% grep(pathway,pattern="Rummagene",value=T),source:="Rummagene"]
pltdat[pathway %in% grep(pathway,pattern="TRRUST",value=T),source:="TRRUST"]
pltdat[pathway %in% grep(pathway,pattern="CREEDSID|_GSE",value=T),source:="GEO TF\nPerturbations"]
pltdat[pathway %in% grep(pathway,pattern="ChEA",value=T),source:="ChEA 2022"]

## TF specifics where applicable
pltdat[pathway %in% c("PPARA_KO_MOUSE_GSE6864_CREEDSID_GENE_1372_DOWN_TF_Perturbations_Followed_by_Expression",
    "PPARA_KO_MOUSE_GSE6864_CREEDSID_GENE_1373_UP_TF_Perturbations_Followed_by_Expression"),tf:="PPARA (KO DE)"]
pltdat[pathway %in% c("PPARG_OE_MOUSE_GSE2192_CREEDSID_GENE_861_UP_TF_Perturbations_Followed_by_Expression",
    "PPARG_OE_MOUSE_GSE2192_CREEDSID_GENE_865_DOWN_TF_Perturbations_Followed_by_Expression"),tf:="PPARG (OE DE)"]
pltdat[pathway=="ESR1_KD_HUMAN_GSE26298_CREEDSID_GENE_479_DOWN_TF_Perturbations_Followed_by_Expression",tf:="ESR1 (KD Down)"]
pltdat[pathway=="PMC7162552-pcbi.1007800.s007.txt--ESR1_Interactors_ESR1_Rummagene_transcription_factors",tf:="ESR1 (Interactors)"]
pltdat[pathway %in% c("AR_KD_HUMAN_GSE22483_CREEDSID_GENE_2194_DOWN_TF_Perturbations_Followed_by_Expression",
    "AR_KD_HUMAN_GSE22483_CREEDSID_GENE_2194_UP_TF_Perturbations_Followed_by_Expression"),tf:="AR (KD DE)"]
pltdat[pathway %in% c("VDR_KO_MOUSE_GSE2421_CREEDSID_GENE_1153_UP_TF_Perturbations_Followed_by_Expression",
    "VDR_KO_MOUSE_GSE2421_CREEDSID_GENE_1153_DOWN_TF_Perturbations_Followed_by_Expression"),tf:="VDR (KO DE)"]
pltdat[pathway %in% c(
    "ESR1_KO_MOUSE_GSE53812_CREEDSID_GENE_2898_UP_TF_Perturbations_Followed_by_Expression",
     "ESR1_KO_MOUSE_GSE53812_CREEDSID_GENE_2898_DOWN_TF_Perturbations_Followed_by_Expression"),tf:="ESR1 (KO DE)"]
pltdat[pathway=="ZNF142_SIRNA_MCF7_HUMAN_GSE79586_RNASEQ_UP_TF_Perturbations_Followed_by_Expression",tf:="ZNF142 (siRNA Up)"]
pltdat[pathway=="ZNF142_SIRNA_MCF7_HUMAN_GSE79586_RNASEQ_DOWN_TF_Perturbations_Followed_by_Expression",tf:="ZNF142 (siRNA Down)"]

### group by sex•region, flip all NES to positive
pltdat[NES<0&spdom=="ARC",sexreg:="ARC Fem. Up"]
pltdat[NES>0&spdom=="ARC",sexreg:="ARC Male Up"]
pltdat[NES<0&spdom=="VMH",sexreg:="VMH Fem. Up"]
pltdat[NES>0&spdom=="VMH",sexreg:="VMH Male Up"]
pltdat[,NES:=abs(NES)]
```


```{r}
### for this to work, we have to manually make a double legend kinda clunkily. modified slightly from https://stackoverflow.com/questions/62541720/ggplot2-show-multiple-keys-shapes-in-size-legend -- speciifically, defining these variables below and then the scale_size_manual call.

## get the breaks based on generating the plot once quick, then get a vector of the same values to feed into the legend override
breaks <- seq(1.5,2.5,0.5)
n_breaks <- length(breaks)
labels <- c(c(rep("",n_breaks),breaks,rep("",2*n_breaks)))
shapes <- c(rep(16,n_breaks),rep(17,n_breaks),rep(18,n_breaks),rep(15,n_breaks))
breaks2 <- rep(breaks, 4)

### make plot
pdf("plots/Manuscript/5B as 4M-Sex DEG TF GSEA.pdf",height=4,width=5)

ggplot(pltdat,aes(y=tf,x=-log10(padj),size=NES))+
    geom_point(aes(col=sexreg,size=NES,shape=source,y=tf),alpha=0.7,position = position_dodge2v(preserve = "single",height=0.75))+
    ylab("Term")+
    scale_color_manual(values=sexregpal)+
    xlab("-log10 (FDR)")+
    labs(shape="Source Database",col="DE Gene Set")+
    geom_vline(xintercept=-log10(0.05),linetype="dashed",linewidth=0.125,color="darkred")+
    scale_shape_manual(values=c("Rummagene"=16,"TRRUST"=15,"GEO TF\nPerturbations"=18,"ChEA 2022"=17))+
    #add minor gridlines between TFs to contain jittered points
    geom_hline(yintercept = seq(0.5,length(unique(pltdat$tf))-0.5),color="grey",linewidth=0.125)+
    #guides(size="none")+
    # ^ necess. to get spacing to work in theme
    guides(shape=guide_legend(override.aes=list(size = 2)))+
    # ^ make legend points bigger
    scale_size_continuous(range = c(0.75,3.5),breaks = breaks2, labels = labels, guide = guide_legend(ncol=4, nrow=3,bycol = TRUE,override.aes = list(shape = shapes), label.vjust = 0.5,label.hjust=0.5,direction="vertical"))+
    # ^ make size legend with all point types
    theme(axis.text.y = element_text(size = 9),axis.title.y=element_text(size=10), axis.title.x = element_text(size = 10), axis.text.x = element_text(size = 9), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5),legend.spacing.y = unit(0.005, "in"),legend.margin = margin(t = 0.1,0,0,-0.1,unit="in"),panel.grid.major.y = element_line(color = "#595959",linewidth = 0.25),panel.grid.minor.y = element_line(color = "#595959",linewidth = 0.25),panel.grid.major.x=element_line(color="#bfbfbf",linewidth=0.25),panel.grid.minor.x=element_line(color="#bfbfbf",linewidth=0.25),axis.ticks.x = element_line(linewidth=0.25,color="#bfbfbf"),axis.ticks.y=element_line(color = "#595959",linewidth = 0.25),legend.spacing.x = unit(-0.015,"in"))

dev.off()

# rm(pltdat,sexreslist)

```

5C-disease enrichments
```{r}
sexdis <- readRDS("data/11-GSEA/06-Common variant disease gene enrichment in sex DEGs.RDS")

sexdis <- sexdis[spdom %in% grep(spdom,pattern="clpsd",value=T)]
sexdis <- sexdis[genesets_from=="union"]

sexdis[,spdom:=gsub(spdom,pattern="^.*_clpsdVMH_(.*)$",replacement="\\1")]
sexdis[,spdom:=gsub(spdom,pattern="X1$",replacement="WM.1")]
sexdis[,spdom:=gsub(spdom,pattern="X9$",replacement="WM.2")]
sexdis[,spdom:=factor(spdom,levels=c("ARC","VMH","WM.1","WM.2",paste0("X",c(2,4,6:8,10,13:15))))]

setnames(sexdis,"padj","FDR")

sexdis[,signif:=ifelse(FDR<0.05,yes="*",no="")]



pdf("plots/Manuscript/5C as 4M-sex DE_disease_GSEA.pdf",height=2.5,width=5)
ggplot(sexdis,aes(x=pathway,y=spdom,fill=NES))+
    geom_tile()+
    geom_text(aes(label=signif),size=2)+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    scale_fill_gradient2()+
    theme(plot.background = element_blank(),axis.line = element_blank(),panel.grid = element_blank(),panel.border = element_blank(),axis.text.x = element_text(angle=90,vjust=0.5,hjust=1,size=9),axis.text.y = element_text(size=9),axis.title.x = element_text(size = 10), axis.title.y = element_text(size = 10), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5),legend.spacing.y = unit(0.125, "in"),legend.margin = margin(0,0,0,0))+
    xlab("Disease Gene Set")+
    ylab("Domain")

dev.off()
```

5D-fDR-sig sex DE enrichments in disease genes by hypergeometric test
### not including for now
```{r}
resdat <- fread("data/11-GSEA/07-hypergeometric enrichment for FDR-sig sex DEGs per k15 cluster in neuropsych GWAS-TWAS-litmined union sets.txt")

resdat <- resdat[clus_set=="BSpace_k15_HARMONYlmbna_nnsvg10"]
resdat <- resdat[spdomain_sex %in% c(paste0(c("X11_","X12_","X3_","X5_","X1_","X9_"),"MaleUp"),paste0(c("X11_","X12_","X3_","X5_","X1_","X9_"),"FemUp"))]
#resdat <- resdat[label %in% unique(resdat[fdr<0.2,label])]

resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="^(.*)_FemUp",replacement=paste0("\\1"," Female Up"))]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="^(.*)_MaleUp",replacement=paste0("\\1"," Male Up"))]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X1 ",replacement="WM.1 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X9 ",replacement="WM.2 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X3 ",replacement="ARC.1 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X5 ",replacement="ARC.2 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X11 ",replacement="VMH.1 ")]
resdat[,spdomain_sex:=gsub(spdomain_sex,pattern="X12 ",replacement="VMH.2 ")]

# make column labels to order by so that each cluster is paired X1male, X1 female, etc
i<-1
colord <- c()
pltcl <- c("VMH.1 ","VMH.2 ","ARC.1 ","ARC.2 ","WM.1 ","WM.2 ")
for (i in c(1:6)){
    colord <- c(colord,paste0(pltcl[i],c("Male Up","Female Up")))
}

pltdat <- dcast(resdat,formula = label~spdomain_sex,value.var = "pval")
pltdat <- as.data.frame(pltdat)
rownames(pltdat) <- pltdat$label
pltdat$label <- NULL
pltdat <- -log10(pltdat)

pdf("plots/Manuscript/5D-FDR sig sex DEG enrichments in dis genesets.pdf",height=3,width=3.5)
pheatmap::pheatmap(pltdat,cluster_cols = F,treeheight_row = 0,fontsize = 9,fontsize_row = 10)
dev.off()
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] BiocParallel_1.36.0 parallelly_1.36.0   default_1.0.0      
 [4] colorout_1.3-0.1    gridExtra_2.3       ggplot2_3.4.4      
 [7] Biostrings_2.70.1   GenomeInfoDb_1.38.5 XVector_0.42.0     
[10] IRanges_2.36.0      S4Vectors_0.40.2    BiocGenerics_0.48.1
[13] data.table_1.14.10  rlang_1.1.3        

loaded via a namespace (and not attached):
 [1] utf8_1.2.4              generics_0.1.3          bitops_1.0-7           
 [4] digest_0.6.34           magrittr_2.0.3          evaluate_0.23          
 [7] grid_4.3.2              RColorBrewer_1.1-3      fastmap_1.1.1          
[10] rprojroot_2.0.4         sessioninfo_1.2.2       fansi_1.0.6            
[13] scales_1.3.0            codetools_0.2-19        cli_3.6.2              
[16] crayon_1.5.2            munsell_0.5.0           withr_3.0.0            
[19] yaml_2.3.8              tools_4.3.2             parallel_4.3.2         
[22] dplyr_1.1.4             colorspace_2.1-0        GenomeInfoDbData_1.2.11
[25] here_1.0.1              vctrs_0.6.5             R6_2.5.1               
[28] lifecycle_1.0.4         zlibbioc_1.48.0         pkgconfig_2.0.3        
[31] pillar_1.9.0            gtable_0.3.4            glue_1.7.0             
[34] xfun_0.41               tibble_3.2.1            tidyselect_1.2.0       
[37] rstudioapi_0.15.0       knitr_1.45              farver_2.1.1           
[40] htmltools_0.5.7         labeling_0.4.3          rmarkdown_2.25         
[43] pheatmap_1.0.12         compiler_4.3.2          RCurl_1.98-1.14        
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-01-23
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────
 package          * version   date (UTC) lib source
 BiocGenerics     * 0.48.1    2023-11-02 [1] Bioconductor
 BiocParallel     * 1.36.0    2023-10-26 [1] Bioconductor
 Biostrings       * 2.70.1    2023-10-25 [1] Bioconductor
 bitops             1.0-7     2021-04-24 [1] CRAN (R 4.3.2)
 cli                3.6.2     2023-12-11 [1] CRAN (R 4.3.2)
 codetools          0.2-19    2023-02-01 [1] CRAN (R 4.3.0)
 colorout         * 1.3-0.1   2024-01-11 [1] local
 colorspace         2.1-0     2023-01-23 [1] CRAN (R 4.3.2)
 crayon             1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 data.table       * 1.14.10   2023-12-08 [1] CRAN (R 4.3.2)
 default          * 1.0.0     2017-08-07 [1] CRAN (R 4.3.0)
 digest             0.6.34    2024-01-11 [1] CRAN (R 4.3.2)
 dplyr              1.1.4     2023-11-17 [1] CRAN (R 4.3.2)
 evaluate           0.23      2023-11-01 [1] CRAN (R 4.3.1)
 fansi              1.0.6     2023-12-08 [1] CRAN (R 4.3.2)
 farver             2.1.1     2022-07-06 [1] CRAN (R 4.3.2)
 fastmap            1.1.1     2023-02-24 [1] CRAN (R 4.3.2)
 generics           0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb     * 1.38.5    2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData   1.2.11    2024-01-11 [1] Bioconductor
 ggplot2          * 3.4.4     2023-10-12 [1] CRAN (R 4.3.1)
 glue               1.7.0     2024-01-09 [1] CRAN (R 4.3.2)
 gridExtra        * 2.3       2017-09-09 [1] CRAN (R 4.3.0)
 gtable             0.3.4     2023-08-21 [1] CRAN (R 4.3.0)
 here               1.0.1     2020-12-13 [1] CRAN (R 4.3.0)
 htmltools          0.5.7     2023-11-03 [1] CRAN (R 4.3.2)
 IRanges          * 2.36.0    2023-10-24 [1] Bioconductor
 knitr              1.45      2023-10-30 [1] CRAN (R 4.3.2)
 labeling           0.4.3     2023-08-29 [1] CRAN (R 4.3.0)
 lifecycle          1.0.4     2023-11-07 [1] CRAN (R 4.3.1)
 magrittr           2.0.3     2022-03-30 [1] CRAN (R 4.3.2)
 munsell            0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 parallelly       * 1.36.0    2023-05-26 [1] CRAN (R 4.3.0)
 pheatmap           1.0.12    2019-01-04 [1] CRAN (R 4.3.0)
 pillar             1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig          2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 R6                 2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 RColorBrewer       1.1-3     2022-04-03 [1] CRAN (R 4.3.0)
 RCurl              1.98-1.14 2024-01-09 [1] CRAN (R 4.3.1)
 rlang            * 1.1.3     2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown          2.25      2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot          2.0.4     2023-11-05 [1] CRAN (R 4.3.1)
 rstudioapi         0.15.0    2023-07-07 [1] CRAN (R 4.3.0)
 S4Vectors        * 0.40.2    2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 scales             1.3.0     2023-11-28 [1] CRAN (R 4.3.2)
 sessioninfo        1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 tibble             3.2.1     2023-03-20 [1] CRAN (R 4.3.2)
 tidyselect         1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 utf8               1.2.4     2023-10-22 [1] CRAN (R 4.3.2)
 vctrs              0.6.5     2023-12-01 [1] CRAN (R 4.3.2)
 withr              3.0.0     2024-01-16 [1] CRAN (R 4.3.1)
 xfun               0.41      2023-11-01 [1] CRAN (R 4.3.2)
 XVector          * 0.42.0    2023-10-24 [1] Bioconductor
 yaml               2.3.8     2023-12-11 [1] CRAN (R 4.3.2)
 zlibbioc           1.48.0    2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

───────────────────────────────────────────────────────────────────────────
