---
title: "01_build_SPE"
output: html_document
date: "2024-06-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Voyager)
library(patchwork)
library(SpatialFeatureExperiment)
library(SingleCellExperiment)
library(SpatialExperiment)
library(ggplot2)
library(gridExtra)
library(stringr)
library(scuttle)
library(BiocSingular)
library(scater)
library(Matrix)
library(DropletUtils)
library(vroom)
library(sf)
library(here)
library(biomaRt)

library(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc on M processor macs
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
Sys.setenv("OBJC_DISABLE_INITIALIZE_FORK_SAFETY"="YES")

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))
```

### Define Xenium data importer to spatialfeatureexperiment format (adapted from a function courtesy of Cindy Fang--gets an SFE together from just 35 MB of xenium output files per sample!)

### 06-19-24: add sample-specific steps while in SPE format to rotate/mirror get ARC in bottom left for all samples.
Goal: ARC on bottom left side.
1735a (X97_reg1): 90 degrees clockwise
1735b (X97_reg2): 90 degrees clockwise
6197 (X86_reg1): 180 degrees; horizontal mirror
5993a (X86_reg2): 90 degrees counter-clockwise; horizontal mirror
5993b (X86_reg3): 90 degrees counter-clockwise; horizontal mirror
6588 (X97_reg3): 90 counter-clockwise; horizontal mirror
1225a (X99_1225A): 90 counter-clockwise
1225b (X99_1225B): 90 counter-clockwise
5459a (X36_5459A): 90 clockwise
5459b (X36_5459B): 90 counter-clockwise
8667 (X36_8667C): 90 deg clockwise
8741c (X99_8741C): horizontal mirror
8741d (X99_8741D): 180 degrees; horizontal mirror


```{r}
# Reference: https://github.com/pachterlab/SFEData/blob/main/inst/scripts/make-data.R

# Read in Xenium data and create an SFE object
# give this function a different name than readXenium, which is imported
# from SpatialFeatureExperiment

## add in an argument + steps to give the cell ids in metadata/geometry tables the same unique IDs they have in the main tables (e.g. counts/colnames(spe)) 
readXenium_manual <- function(dir_name, samp_altid){

    counts_path <- here(dir_name, "cell_feature_matrix.h5")
    cell_info_path <- here(dir_name, "cells.csv.gz")
    cell_poly_path <- here(dir_name, "cell_boundaries.csv.gz")
    nuc_poly_path <- here(dir_name, "nucleus_boundaries.csv.gz")
    # 
    # Read in the data
    sce <- read10xCounts(counts_path)
    counts(sce) <- as(realize(counts(sce)), "dgCMatrix")
    cell_info <- vroom(cell_info_path)
    
    print("Read in files")
    
    cell_poly <- vroom(cell_poly_path)
    nuc_poly <- vroom(nuc_poly_path)
    
    names(cell_poly)[1] <- "ID"
    names(nuc_poly)[1] <- "ID"
    
    cell_poly$ID <- paste0(samp_altid,"_",cell_poly$ID)
    nuc_poly$ID <- paste0(samp_altid,"_",nuc_poly$ID)
    # change df to sf object for cell/nuc images
    cells_sf <- df2sf(cell_poly, c("vertex_x", "vertex_y"), geometryType = "POLYGON")
    nuc_sf <- df2sf(nuc_poly, c("vertex_x", "vertex_y"), geometryType = "POLYGON")
    
    # QC check
    all(st_is_valid(cells_sf))
    all(st_is_valid(nuc_sf))
    
    # get rid if invalid cells/nucs
    ind_invalid <- !st_is_valid(nuc_sf)
    nuc_sf[ind_invalid,] <- nngeo::st_remove_holes(st_buffer(nuc_sf[ind_invalid,], 0))
    
    # add cell info
    colData(sce) <- cbind(colData(sce), cell_info)
    print(cell_info)
    
    # make spatialExperiment with unique colnames
    #  drop zero count cells from the SPE
    spe <- toSpatialExperiment(sce, spatialCoordsNames = c("x_centroid", "y_centroid")) 
    spe <- spe[,colSums(counts(spe))>0]
    
    # add the sample-level ID to the one-sample SPE in processing
    spe$sample_id <- samp_altid
    spe$Sample <- samp_altid
    
    # use this^ and the cell_id to give the cell a globally unique ID
    # which we can then use as the spe/sfe colnames
    spe$key <- paste0(spe$Sample,"_",spe$cell_id)
    colnames(spe) <- spe$key
    
    ### IMPORTANT: for whatever reason, "BLANK" codewords were renamed
    ### to "UnassignedCodeword"s on 10xs end somewhere between the two
    ### runs, (with the same identifying numbers though) so we need to
    ### fix these to match across all parts of the data. Since the
    ### newer convention is "UnassignedCodeword", we'll use that.
    rownames(spe) <- gsub(rownames(spe),pattern="BLANK",replacement="UnassignedCodeword")
    
    # but, these don't automatically go into the ID or Symbol columns of rowData(spe), only its rownames, so copy it over to both. (Plus, for the non-targeting/etc probes, the Symbol is the same as the ID so we need to carry it over there too.) they also didn't assign ENSGs to the probes we had to urge them to put on the custom panel despite them not being part of their library of avail--those are given the ID "AVP_gene" and "OXT_gene".
    rowData(spe)$ID <- rownames(spe)
    rowData(spe)$Symbol[grep(rowData(spe)$ID,pattern="ENSG|AVP|OXT",invert=T)] <- rownames(spe)[grep(rowData(spe)$ID,pattern="ENSG|AVP|OXT",invert=T)]
    
    # lastly, the SFE will only place the segmentations in the 
    # cellSeg/nucSeg slots if the dimensions are exactly the same.
    # since we've given the cells unique identifiers in both the seg
    # tables and the predecessor SPE object, we can get these subsetted 
    # to be the same cells before proceeding to SFE
    # (and this should also circumvent weird issues with segmentations
    # getting placed in int_metadata in some cases, but in
    # int_metadata$annotGeometries (which are supposed to correspond to
    # non-cell, non-nucleus, non-gene geometry annots) in most)
    dim(spe)
    nuc_sf <- nuc_sf[which(nuc_sf$ID %in% spe$key),]
    cells_sf <- cells_sf[which(cells_sf$ID %in% spe$key),]
    
    sfe <- toSpatialFeatureExperiment(spe)
    
    # add segmented cells/nuc to spatial object
    # since we've given the cells unique identifiers in both the seg
    # tables and the predecessor SPE object, these should all match up
    # (and this should also circumvent weird issues with segmentations
    # getting placed in int_metadata in some cases, but in int_metadata$annotGeometries (which are supposed to correspond to non-cell, non-nucleus, non-gene geometry annots) in most)
    cellSeg(sfe, withDimnames = T) <- cells_sf
    nucSeg(sfe, withDimnames = T) <- nuc_sf
    
    
    # Add some QC Metrics
    colData(sfe)$nCounts <- colSums(counts(sfe))
    colData(sfe)$nGenes <- colSums(counts(sfe) > 0)
    
    is_blank <- str_detect(rownames(sfe), "^BLANK_")
    is_neg <- str_detect(rownames(sfe), "^NegControlProbe")
    is_neg2 <- str_detect(rownames(sfe), "^NegControlCodeword")
    is_anti <- str_detect(rownames(sfe), "^antisense")
    is_depr <- str_detect(rownames(sfe), "^DeprecatedCodeword")
    is_unassigned <- str_detect(rownames(sfe), "^Unassigned")
    
    is_any_neg <- is_blank | is_neg | is_neg2 | is_anti | is_depr | is_unassigned
    rowData(sfe)$is_neg <- is_any_neg
    
    n_panel <- nrow(sfe) - sum(is_any_neg)
    #print(n_panel)
    
    # normalize counts after QC
    colData(sfe)$nCounts_normed <- sfe$nCounts/n_panel
    colData(sfe)$nGenes_normed <- sfe$nGenes/n_panel
    colData(sfe)$prop_nuc <- sfe$nucleus_area / sfe$cell_area
    
    # add QC columns
    sfe <- addPerCellQCMetrics(sfe, subsets = list(blank = is_blank,
    negProbe = is_neg,
    negCodeword = is_neg2,
    anti = is_anti,
    depr = is_depr,
    unassigned = is_unassigned,
    any_neg = is_any_neg))

    # add sample-specific metrics re: features; give these sample-specific colnames in rowData so we can bind samples together
    rowData(sfe)$means <- rowMeans(counts(sfe))
    rowData(sfe)$vars <- rowVars(counts(sfe))
    rowData(sfe)$cv2 <- rowData(sfe)$vars/rowData(sfe)$means^2
    
    colnames(rowData(sfe))[colnames(rowData(sfe))=="means"] <- paste0("means_",samp_altid)
    colnames(rowData(sfe))[colnames(rowData(sfe))=="vars"] <- paste0("vars_",samp_altid)
    colnames(rowData(sfe))[colnames(rowData(sfe))=="cv2"] <- paste0("cv2_",samp_altid)
    
    # Add cell ids and make gene names unique
    
    rownames(sfe) <- uniquifyFeatureNames(ID=rownames(sfe),  names=rowData(sfe)$Symbol)

    # based on the sample alt identifier, rotate/mirror the ARC to be in the bottom left corner per above comments.
    if(samp_altid %in% c("X97_reg1","X97_reg2")){
        sfe <- SpatialFeatureExperiment::rotate(sfe, degrees = 90)
    }
    else if(samp_altid %in% c("X86_reg1")){
        sfe <- SpatialFeatureExperiment::rotate(sfe, degrees = 180)
        sfe <- SpatialFeatureExperiment::mirror(sfe,direction = "horizontal")
    }
    else if(samp_altid %in% c("X86_reg2","X86_reg3","X97_reg3")){
        sfe <- SpatialFeatureExperiment::rotate(sfe, degrees = -90)
        sfe <- SpatialFeatureExperiment::mirror(sfe,direction = "horizontal")
    }
    else if(samp_altid %in% c("X99_1225A","X99_1225B","X36_5459B")){
        sfe <- SpatialFeatureExperiment::rotate(sfe, degrees = -90)
    }
    else if(samp_altid %in% c("X36_5459A","X36_8667C")){
        sfe <- SpatialFeatureExperiment::rotate(sfe, degrees = 90)
    }
    else if(samp_altid %in% c("X99_8741C")){
        sfe <- SpatialFeatureExperiment::mirror(sfe,direction = "horizontal")
    }
    else if(samp_altid %in% c("X99_8741D")){
        sfe <- SpatialFeatureExperiment::rotate(sfe, degrees = 180)
        sfe <- SpatialFeatureExperiment::mirror(sfe,direction = "horizontal")
    }
    
    return(sfe)
}
```

### load combined cell-nucleus resegmentations and original segmentations as separate SpatialFeatureExperiment objects to start with
```{r}
resegs <- list.files("processed-data/01_xeniumranger1.7-resegment",pattern="03|02",full.names = T)
resegsampname <- gsub(resegs,pattern="^.*1.7-resegment/0._(.*)_reseg_wnuclei$",replacement="\\1")
## drop the "xr17" from the 2024 slide sample short names
resegsampname <- gsub(resegsampname,pattern="_xr17",replacement="")
## tack a leading nonnumeric character onto the short names to avoid problems later (some fxns/packages will choke on leading numerics))
resegsampname <- paste0("X",resegsampname)
resegs <- paste0(resegs,"/outs")

### iterate through resegmentations and add/bind them to an SPE
i<-1
for (i in c(1:length(resegs))){
    if(i==1){
        resegsamps <- readXenium_manual(resegs[i],resegsampname[i])

    }
    else{
        tmpresegsamp <- readXenium_manual(resegs[i],resegsampname[i])
        resegsamps <- cbind(resegsamps,tmpresegsamp)
        
        rm(tmpresegsamp)
    }
}
# cleanup
rm(i)
gc(full=T)
```

## append donor metadata to colData
```{r}
demos <- fread("raw-data/demos_and_xeniumrun_metadata.tsv")
tmpcd <- as.data.table(colData(resegsamps),keep.rownames=T)

tmpcd <- merge.data.table(tmpcd,demos,by.x="Sample",by.y="sample_name_SFEonly")
tmpcd <- as.data.frame(tmpcd)
rownames(tmpcd) <- tmpcd$rn
tmpcd$rn <- NULL
tmpcd <- tmpcd[colnames(resegsamps),]
colData(resegsamps) <- DataFrame(tmpcd)

rm(tmpcd,demos)
```

### add panel information (brain base panel / HYP custom) and ensembl identifiers for genes without
```{r}
basepan <- fread("raw-data/humanbrain_basepanel_genes.txt")
tmprd <- as.data.table(rowData(resegsamps),keep.rownames=T)
tmprd[Symbol %in% basepan$Gene,panel:="Human Brain Base Panel"]
### The only other targeting probes are then HYP panel genes, which already have an ENSEMBL identifier (except 'AVP_gene' and 'OXT_gene'
tmprd[ID %in% grep(ID,pattern="ENSG|OXT|AVP",value=T) & !(Symbol %in% basepan$Gene),panel:="Custom Human HYP Targets"]
```
### add ENSG identifiers for OXT, AVP using the ENS110 reference that 10x's 2024 reference is built on.

### 10x 2024 human genome reference: ens110 as ref build, but with ens109 X/Y sequences used to avoid mapping problems between pseudoautosomal region being unmasked on both (leading to ambiguous mapping of PAR genes). 

see:
https://www.10xgenomics.com/support/software/cell-ranger/downloads/cr-ref-build-steps 
for more. quoth their code for building the gtf for cellranger/spaceranger:

# Since Ensembl 110, the PAR locus genes are included on chrY as copies of chrX
# Using the GRCh38.p13 assembly hard masks these regions on chrY, but removing the chrY PAR genes is still desirable so they do not end up as extra entries in the output.
# [building this gtf] excludes all PAR_Y genes, including XGY2.
# The non-coding gene XGY2 straddles the PAR1 boundary on chrY, and is homologous to XG on chrX. GRCh38-2024-A excludes XGY2, but includes SRY and ENSG00000286130, which are in an intron of XGY2, and RPS4Y1, which overlaps XGY2.
```{r}
mart <- useMart(biomart="ENSEMBL_MART_ENSEMBL",host="https://jul2023.archive.ensembl.org")
mart <- useDataset("hsapiens_gene_ensembl",mart)
genedat <- getBM(attributes=c("ensembl_gene_id","external_gene_name","hgnc_symbol"),mart = mart)
genedat <- as.data.table(genedat)
## add ENSG ids using this^ as a reference table. (confirmed there is one ENSG ID per gene symbol for these two.)
tmprd[ID=="AVP_gene",ID:=genedat[external_gene_name=="AVP",ensembl_gene_id]]
tmprd[ID=="OXT_gene",ID:=genedat[external_gene_name=="OXT",ensembl_gene_id]]
## fix the gene symbols in the temp rowData
tmprd[Symbol %in% c("AVP_gene","OXT_gene"),Symbol:=gsub(Symbol,pattern="_gene",replacement="")]
## fix the rownames in the temp rowData
tmprd[Symbol=="AVP",rn:="AVP"]
tmprd[Symbol=="OXT",rn:="OXT"]
## to put the modified rowData in place of the existing one, we'll have to fix the AVP_gene/OXT_gene rownames on the existing SFE so we can reorder it by the same rownames. easy enough.
rownames(resegsamps)[which(rownames(resegsamps)=="AVP_gene")] <- "AVP"
rownames(resegsamps)[which(rownames(resegsamps)=="OXT_gene")] <- "OXT"

## coerce temp rowData to DF --> reorder according to SFE --> place 
tmprd <- DataFrame(tmprd)
rownames(tmprd) <- tmprd$rn
tmprd$rn <- NULL
tmprd <- tmprd[rownames(resegsamps),]
rowData(resegsamps) <- tmprd

# cleanup
rm(tmprd,basepan,genedat,mart)
```

save core Xenium data as RDS
```{r}
saveRDS(resegsamps,"processed-data/03_make_SPE-SFE/01_sfe_raw.RDS")
```

since the downstream work occurs mostly in an SPE after filtering and coordinate-staggering (part of banksy set-up), save the original cell id-coordinate table so we can plot transcript-level data in the future. transcript-level data is not retained in the SPE/SFE and thus later imports will need the original, not modified coordinates.
```{r}
origcoords <- as.data.table(cbind(colnames(xenshift),spatialCoords(xenshift)))
fwrite(origcoords,"processed-data/03_make_SPE-SFE/01_sfe_raw_cellcentroidcoords.txt",sep='\t',quote=F)
```

### Reproducibility info
```{r}
sessionInfo()
sessioninfo::session_info()
```

R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] BiocParallel_1.38.0            parallelly_1.37.1             
 [3] default_1.0.0                  colorout_1.3-0.2              
 [5] biomaRt_2.60.0                 here_1.0.1                    
 [7] sf_1.0-16                      vroom_1.6.5                   
 [9] DropletUtils_1.24.0            Matrix_1.7-0                  
[11] scater_1.32.0                  BiocSingular_1.20.0           
[13] scuttle_1.14.0                 stringr_1.5.1                 
[15] gridExtra_2.3                  ggplot2_3.5.1                 
[17] SpatialExperiment_1.14.0       SingleCellExperiment_1.26.0   
[19] SummarizedExperiment_1.34.0    Biobase_2.64.0                
[21] GenomicRanges_1.56.1           GenomeInfoDb_1.40.1           
[23] IRanges_2.38.0                 S4Vectors_0.42.0              
[25] BiocGenerics_0.50.0            MatrixGenerics_1.16.0         
[27] matrixStats_1.3.0              patchwork_1.2.0               
[29] Voyager_1.6.0                  SpatialFeatureExperiment_1.6.1
[31] data.table_1.15.4              rlang_1.1.4                   

loaded via a namespace (and not attached):
  [1] rstudioapi_0.16.0         jsonlite_1.8.8           
  [3] wk_0.9.1                  magrittr_2.0.3           
  [5] ggbeeswarm_0.7.2          magick_2.8.3             
  [7] rmarkdown_2.27            zlibbioc_1.50.0          
  [9] vctrs_0.6.5               spdep_1.3-5              
 [11] memoise_2.0.1             DelayedMatrixStats_1.26.0
 [13] RCurl_1.98-1.14           terra_1.7-78             
 [15] progress_1.2.3            htmltools_0.5.8.1        
 [17] S4Arrays_1.4.1            curl_5.2.1               
 [19] BiocNeighbors_1.22.0      Rhdf5lib_1.26.0          
 [21] s2_1.1.6                  SparseArray_1.4.8        
 [23] rhdf5_2.48.0              spData_2.3.1             
 [25] KernSmooth_2.23-24        htmlwidgets_1.6.4        
 [27] httr2_1.0.1               cachem_1.1.0             
 [29] igraph_2.0.3              lifecycle_1.0.4          
 [31] pkgconfig_2.0.3           rsvd_1.0.5               
 [33] R6_2.5.1                  fastmap_1.2.0            
 [35] GenomeInfoDbData_1.2.12   digest_0.6.35            
 [37] colorspace_2.1-0          ggnewscale_0.4.10        
 [39] AnnotationDbi_1.66.0      rprojroot_2.0.4          
 [41] dqrng_0.4.1               RSpectra_0.16-1          
 [43] irlba_2.3.5.1             nngeo_0.4.8              
 [45] RSQLite_2.3.7             beachmat_2.20.0          
 [47] filelock_1.0.3            fansi_1.0.6              
 [49] httr_1.4.7                abind_1.4-5              
 [51] compiler_4.4.1            proxy_0.4-27             
 [53] bit64_4.0.5               withr_3.0.0              
 [55] tiff_0.1-12               viridis_0.6.5            
 [57] DBI_1.2.3                 HDF5Array_1.32.0         
 [59] R.utils_2.12.3            rappdirs_0.3.3           
 [61] DelayedArray_0.30.1       rjson_0.2.21             
 [63] classInt_0.4-10           bluster_1.14.0           
 [65] tools_4.4.1               units_0.8-5              
 [67] vipor_0.4.7               beeswarm_0.4.0           
 [69] R.oo_1.26.0               glue_1.7.0               
 [71] EBImage_4.46.0            rhdf5filters_1.16.0      
 [73] grid_4.4.1                cluster_2.1.6            
 [75] generics_0.1.3            memuse_4.2-3             
 [77] gtable_0.3.5              tzdb_0.4.0               
 [79] R.methodsS3_1.8.2         class_7.3-22             
 [81] hms_1.1.3                 xml2_1.3.6               
 [83] ScaledMatrix_1.12.0       sp_2.1-4                 
 [85] utf8_1.2.4                XVector_0.44.0           
 [87] ggrepel_0.9.5             pillar_1.9.0             
 [89] limma_3.60.2              dplyr_1.1.4              
 [91] BiocFileCache_2.12.0      lattice_0.22-6           
 [93] bit_4.0.5                 deldir_2.0-4             
 [95] tidyselect_1.2.1          locfit_1.5-9.9           
 [97] Biostrings_2.72.1         sfheaders_0.4.4          
 [99] knitr_1.47                edgeR_4.2.0              
[101] xfun_0.45                 statmod_1.5.0            
[103] stringi_1.8.4             UCSC.utils_1.0.0         
[105] fftwtools_0.9-11          yaml_2.3.8               
[107] boot_1.3-30               evaluate_0.24.0          
[109] codetools_0.2-20          tibble_3.2.1             
[111] cli_3.6.2                 munsell_0.5.1            
[113] Rcpp_1.0.12               dbplyr_2.5.0             
[115] zeallot_0.1.0             png_0.1-8                
[117] parallel_4.4.1            blob_1.2.4               
[119] prettyunits_1.2.0         jpeg_0.1-10              
[121] sparseMatrixStats_1.16.0  bitops_1.0-7             
[123] viridisLite_0.4.2         scales_1.3.0             
[125] e1071_1.7-14              purrr_1.0.2              
[127] crayon_1.5.2              scico_1.5.0              
[129] KEGGREST_1.44.0          
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-06-19
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────
 ! package                  * version   date (UTC) lib source
   abind                      1.4-5     2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi              1.66.0    2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   beachmat                   2.20.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm                   0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   Biobase                  * 2.64.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocFileCache              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics             * 0.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocNeighbors              1.22.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel             * 1.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocSingular             * 1.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   biomaRt                  * 2.60.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings                 2.72.1    2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                        4.0.5     2022-11-15 [1] CRAN (R 4.4.0)
   bit64                      4.0.5     2020-08-30 [1] CRAN (R 4.4.0)
   bitops                     1.0-7     2021-04-24 [1] CRAN (R 4.4.0)
   blob                       1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   bluster                    1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   boot                       1.3-30    2024-02-26 [1] CRAN (R 4.4.1)
   cachem                     1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
   class                      7.3-22    2023-05-03 [1] CRAN (R 4.4.1)
   classInt                   0.4-10    2023-09-05 [1] CRAN (R 4.4.0)
 P cli                        3.6.2     2023-12-11 [2] CRAN (R 4.4.0)
   cluster                    2.1.6     2023-12-01 [1] CRAN (R 4.4.1)
   codetools                  0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorout                 * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace                 2.1-0     2023-01-23 [1] CRAN (R 4.4.0)
   crayon                     1.5.2     2022-09-29 [1] CRAN (R 4.4.0)
   curl                       5.2.1     2024-03-01 [1] CRAN (R 4.4.0)
   data.table               * 1.15.4    2024-03-30 [2] CRAN (R 4.4.0)
   DBI                        1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                     2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   default                  * 1.0.0     2017-08-07 [1] CRAN (R 4.4.0)
   DelayedArray               0.30.1    2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats         1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   deldir                     2.0-4     2024-02-28 [1] CRAN (R 4.4.0)
   digest                     0.6.35    2024-03-11 [1] CRAN (R 4.4.0)
   dplyr                      1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   dqrng                      0.4.1     2024-05-28 [1] CRAN (R 4.4.0)
   DropletUtils             * 1.24.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   e1071                      1.7-14    2023-12-06 [1] CRAN (R 4.4.0)
   EBImage                    4.46.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   edgeR                      4.2.0     2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate                   0.24.0    2024-06-10 [1] CRAN (R 4.4.0)
   fansi                      1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   fastmap                    1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fftwtools                  0.9-11    2021-03-01 [1] CRAN (R 4.4.0)
   filelock                   1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   generics                   0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb             * 1.40.1    2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData           1.2.12    2024-05-01 [1] Bioconductor
   GenomicRanges            * 1.56.1    2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm                 0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggnewscale                 0.4.10    2024-02-08 [1] CRAN (R 4.4.0)
   ggplot2                  * 3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrepel                    0.9.5     2024-01-10 [1] CRAN (R 4.4.0)
   glue                       1.7.0     2024-01-09 [1] CRAN (R 4.4.0)
   gridExtra                * 2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gtable                     0.3.5     2024-04-22 [1] CRAN (R 4.4.0)
   HDF5Array                  1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   here                     * 1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   hms                        1.1.3     2023-03-21 [1] CRAN (R 4.4.0)
   htmltools                  0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets                1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httr                       1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   httr2                      1.0.1     2024-04-01 [1] CRAN (R 4.4.0)
   igraph                     2.0.3     2024-03-13 [1] CRAN (R 4.4.0)
   IRanges                  * 2.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   irlba                      2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   jpeg                       0.1-10    2022-11-29 [1] CRAN (R 4.4.0)
   jsonlite                   1.8.8     2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST                   1.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   KernSmooth                 2.23-24   2024-05-17 [1] CRAN (R 4.4.1)
   knitr                      1.47      2024-05-29 [1] CRAN (R 4.4.0)
   lattice                    0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lifecycle                  1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                      3.60.2    2024-05-19 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                     1.5-9.9   2024-03-01 [1] CRAN (R 4.4.0)
   magick                     2.8.3     2024-02-18 [1] CRAN (R 4.4.0)
   magrittr                   2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   Matrix                   * 1.7-0     2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics           * 1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats              * 1.3.0     2024-04-11 [1] CRAN (R 4.4.0)
   memoise                    2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   memuse                     4.2-3     2023-01-24 [1] CRAN (R 4.4.0)
   munsell                    0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   nngeo                      0.4.8     2024-04-17 [1] CRAN (R 4.4.0)
   parallelly               * 1.37.1    2024-02-29 [1] CRAN (R 4.4.0)
   patchwork                * 1.2.0     2024-01-08 [1] CRAN (R 4.4.0)
   pillar                     1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig                  2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   png                        0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   prettyunits                1.2.0     2023-09-24 [1] CRAN (R 4.4.0)
   progress                   1.2.3     2023-12-06 [1] CRAN (R 4.4.0)
   proxy                      0.4-27    2022-06-09 [1] CRAN (R 4.4.0)
   purrr                      1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R.methodsS3                1.8.2     2022-06-13 [1] CRAN (R 4.4.0)
   R.oo                       1.26.0    2024-01-24 [1] CRAN (R 4.4.0)
   R.utils                    2.12.3    2023-11-18 [1] CRAN (R 4.4.0)
   R6                         2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs                   0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   Rcpp                       1.0.12    2024-01-09 [1] CRAN (R 4.4.0)
   RCurl                      1.98-1.14 2024-01-09 [1] CRAN (R 4.4.0)
   rhdf5                      2.48.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rhdf5filters               1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Rhdf5lib                   1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rjson                      0.2.21    2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                    * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown                  2.27      2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot                  2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   RSpectra                   0.16-1    2022-04-24 [1] CRAN (R 4.4.0)
   RSQLite                    2.3.7     2024-05-27 [1] CRAN (R 4.4.0)
   rstudioapi                 0.16.0    2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                       1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   s2                         1.1.6     2023-12-19 [1] CRAN (R 4.4.0)
   S4Arrays                   1.4.1     2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors                * 0.42.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   ScaledMatrix               1.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                     1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater                   * 1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scico                      1.5.0     2023-08-14 [1] CRAN (R 4.4.0)
   scuttle                  * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo                1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   sf                       * 1.0-16    2024-03-24 [1] CRAN (R 4.4.0)
   sfheaders                  0.4.4     2024-01-17 [1] CRAN (R 4.4.0)
   SingleCellExperiment     * 1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sp                         2.1-4     2024-04-30 [1] CRAN (R 4.4.0)
   SparseArray                1.4.8     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats          1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment        * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialFeatureExperiment * 1.6.1     2024-05-15 [1] Bioconductor 3.19 (R 4.4.0)
   spData                     2.3.1     2024-05-31 [1] CRAN (R 4.4.0)
   spdep                      1.3-5     2024-06-10 [1] CRAN (R 4.4.0)
   statmod                    1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   stringi                    1.8.4     2024-05-06 [1] CRAN (R 4.4.0)
   stringr                  * 1.5.1     2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment     * 1.34.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   terra                      1.7-78    2024-05-22 [1] CRAN (R 4.4.0)
   tibble                     3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyselect                 1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   tiff                       0.1-12    2023-11-28 [1] CRAN (R 4.4.0)
   tzdb                       0.4.0     2023-05-12 [1] CRAN (R 4.4.0)
   UCSC.utils                 1.0.0     2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   units                      0.8-5     2023-11-28 [1] CRAN (R 4.4.0)
   utf8                       1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                      0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                      0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                    0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite                0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   Voyager                  * 1.6.0     2024-05-30 [2] Bioconductor 3.19 (R 4.4.0)
   vroom                    * 1.6.5     2023-12-05 [1] CRAN (R 4.4.0)
   withr                      3.0.0     2024-01-16 [1] CRAN (R 4.4.0)
   wk                         0.9.1     2023-11-29 [1] CRAN (R 4.4.0)
   xfun                       0.45      2024-06-16 [1] CRAN (R 4.4.0)
   xml2                       1.3.6     2023-12-04 [1] CRAN (R 4.4.0)
   XVector                    0.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                       2.3.8     2023-12-11 [1] CRAN (R 4.4.0)
   zeallot                    0.1.0     2018-01-28 [1] CRAN (R 4.4.0)
   zlibbioc                   1.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────
