---
title: "01-QC Filters"
author: "Bernie Mulvey"
date: "2024-07-19"
output: html_document
---

```{r setup, include=FALSE}
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(ggspavis)
library(spatialLIBD)
library(scater) # addPerCellQC
library(BiocParallel)
library(scran)
library(gridExtra)
library(SpotSweeper)
```

## load pre-assembled SPE object of mostly unfiltered data (has already been cut down to tissue-overlapping spots and genes with at least one count in at least one spot dataset-wide)
```{r}
load("processed-data/02_build_spe/spe_w_v13y24-346_Jul24.RData")
hyp <- spe_raw2
rm(spe_raw2)
gc(full=T)
### for reference: initial dataset dimensions
dim(hyp)
head(colData(hyp))
tail(colData(hyp))
```

### filtering
# double check removal of non-tissue spots from loupe alignment
```{r}
hyp2 <- hyp[,colData(hyp)$in_tissue]
### check number of spots lost
ncol(hyp)-ncol(hyp2)
# 0 -- good, already removed them all.
## nonetheless, hold onto unfiltered object for now as comparator.
```


### apply rotations/mirrors to have all samples oriented medial=left, ventral=bottom for ease of reading the rest of these plots. ### WARNING : DISTORTS SOME SAMPLES (fiduc. frame and tissue end up misaligned and/or cut off, somehow) WHEN PLOTTING IN SPATIAL LIBD--need to set auto_crop=FALSE; unsure why. looks like maybe if the fiducial frame is not an equal distance from all four image boundaries it doesn't work. as long as we don't make figure panels from slibd doesn't matter. (not very space-efficient to use uncropped spatial images)

V12D05-348_C1: 90 CW
V12D05-348_D1: no ∆
V12D05-350_C1: 90 CW
V12D05-350_D1: no ∆
V12D07-075_A1: 270 CW
V12D07-075_D1: 180 + mirror
V12Y31-080_A1: 90 CW
V13M13-362_A1: mirror
V13M13-362_D1: 270 CW
V13Y24-346_C1: horiz mirror

```{r}
hyp3 <- hyp2[,hyp2$sample_id %in% c("V12D05-350_D1","V12D05-348_D1")]

tmp1 <- rotateObject(hyp2,sample_id="V12D05-348_C1",degrees=90)
tmp2 <- rotateObject(hyp2,sample_id="V12D05-350_C1",degrees=90)

tmp3 <- rotateObject(hyp2,sample_id="V12D07-075_A1",degrees=270)

tmp4 <- rotateObject(hyp2,sample_id="V12D07-075_D1",degrees=180)
tmp4 <- mirrorObject(tmp4,sample_id="V12D07-075_D1",axis="v")

tmp5 <- rotateObject(hyp2,sample_id="V12Y31-080_A1",degrees=90)

tmp6 <- mirrorObject(hyp2,sample_id="V13M13-362_A1",axis="v")

tmp7 <- rotateObject(hyp2,sample_id="V13M13-362_D1",degrees=270)

tmp8 <- mirrorObject(hyp2,sample_id="V13Y24-346_C1",axis="v")
## rearrange to match original for posterity
hyp4 <- cbind(hyp3,tmp1,tmp2,tmp3,tmp4,tmp5,tmp6,tmp7,tmp8)
identical(dim(hyp2),dim(hyp4))
hyp4 <- hyp4[rownames(hyp2),colnames(hyp2)]
identical(dim(hyp2),dim(hyp4))

# plot check
rownames(hyp4) <- rowData(hyp4)$gene_name
plts <- vis_grid_gene(hyp4,"FEZF1",pdf_file = NULL,spatial = T,return_plots = T,auto_crop = F,point_size = 0.7,alpha=0.3,image_id = "lowres",assayname = "counts")
do.call("grid.arrange",plts)
dev.off()

# rownames back to gene identifiers
rownames(hyp4) <- rowData(hyp4)$gene_id

hyp2 <- copy(hyp4)
rm(tmp1,tmp2,tmp3,tmp4,tmp5,tmp6,tmp7,tmp8,hyp3,hyp4,plts)
gc(full=T)
```

### spotsweeper local outlier detections: % mitochondrial reads, UMIs, # genes -- using spotsweeper for local outlier detection. don't log-norm mito %age values per spotsweeper docs
```{r}
hyp2 <- SpotSweeper::localOutliers(spe=hyp2,metric = "expr_chrM_ratio",direction = "higher",log=FALSE)
# outlier calls are in colData()$expr_chrM_ratio_outliers, z scores used for determining that in colData()$expr_chrM_ratio_z

sum(hyp2$expr_chrM_ratio_outliers)
# 28

hyp2 <- SpotSweeper::localOutliers(hyp2,metric = "sum_umi",direction = "lower",log=TRUE)

sum(hyp2$sum_umi_outliers)
# 139

hyp2 <- SpotSweeper::localOutliers(hyp2,metric="sum_gene",direction="lower",log=TRUE)

sum(hyp2$sum_gene_outliers)
# 162

## total number of spots to be dropped due to outlier status: 162+139+28 = 329
```

### spotsweeper for tissue edge/hangnail cleanup. this uses raw mito counts AND mito %age

### however, there's a minor bug in SpotSweeper 1.0.2 that prevents this function that prevents return of results, letting Mike know but editing the function here to make it work--striking the reducedDims(spe) <- reudcedDims(spe.temp line)
```{r}
findArtifactsFix <- function (spe, mito_percent = "expr_chrM_ratio", mito_sum = "expr_chrM", 
    samples = "sample_id", n_rings = 5, log = TRUE, name = "artifact", 
    var_output = TRUE,BPPARAM=SerialParam()) 
{
    if (!("SpatialExperiment" %in% class(spe))) {
        stop("Input data must be a SpatialExperiment object.")
    }
    if (!all(mito_percent %in% colnames(colData(spe)))) {
        stop("mito_percent must be present in colData.")
    }
    if (!mito_sum %in% colnames(colData(spe))) {
        stop("mito_sum must be present in colData.")
    }
    if (!samples %in% colnames(colData(spe))) {
        stop("Samples column must be present in colData.")
    }
    if (!is.numeric(n_rings) || n_rings <= 0 || n_rings != round(n_rings)) {
        stop("'n_rings' must be a positive integer.")
    }
    unique_sample_ids <- unique(colData(spe)[[samples]])
    
    ## split spe into a list of SPEs by sample id for memory efficiency (esp if parallelizing)
    spelist <- lapply(unique_sample_ids, FUN=function(s){spe[,colData(spe)[[samples]]==s]})
    names(spelist) <- unique_sample_ids
    
    columnData_list <- bplapply(spelist,BPPARAM=BPPARAM,FUN=function(sampspe) {
        var_matrix <- vapply(seq_len(n_rings), function(i) {
            n_neighbors <- 3 * i * (i + 1)
            tmp.name <- paste0("k", n_neighbors)
            sampspe <<- localVariance(sampspe, metric = mito_percent, 
                n_neighbors = n_neighbors, name = tmp.name, 
                log = log)
            colData(sampspe)[[tmp.name]]
        }, numeric(length(sampspe[[1]])))
        var_matrix <- cbind(var_matrix, colData(sampspe)[[mito_percent]], 
            colData(sampspe)[[mito_sum]])
        var_df <- data.frame(var_matrix)
        pc <- prcomp(var_df, center = TRUE, scale = TRUE)
        rownames(pc$x) <- colnames(sampspe)
        reducedDim(sampspe, "PCA_artifacts") <- pc$x
        clus <- kmeans(pc$x, centers = 2, nstart = 25)
        sampspe$Kmeans <- clus$cluster
        
        retrievek <- 3*n_rings*(n_rings+1)
        clus1_mean <- mean(colData(sampspe)[[paste0("k", retrievek)]][sampspe$Kmeans == 
            1])
        clus2_mean <- mean(colData(sampspe)[[paste0("k", retrievek)]][sampspe$Kmeans == 
            2])
        artifact_clus <- which.min(c(clus1_mean, clus2_mean))
        sampspe$artifact <- FALSE
        sampspe$artifact[sampspe$Kmeans == artifact_clus] <- TRUE
        sampspe$Kmeans <- NULL
        return(colData(sampspe))
    })
    columnData_aggregated <- do.call(rbind, columnData_list)
    columnData_aggregated <- DataFrame(columnData_aggregated)
    columnData_aggregated <- columnData_aggregated[colnames(spe),]
    colData(spe) <- columnData_aggregated
    # reducedDims(spe) <- reducedDims(spe.temp)
    return(spe)
}
```


## run the modified findArtifacts to pick up artifacts (which should just be tissue edges and tears/etc)
```{r}
hyp2 <- findArtifactsFix(hyp2,mito_percent = "expr_chrM_ratio",mito_sum = "expr_chrM",samples = "sample_id",BPPARAM=MulticoreParam(5))

sum(hyp2$artifact)
# 26364 spots (like 60% of the data -- and since this uses k=2 kmeans clustering, and picks one of the two clusters as the artifact cluster, this is impractical even if it were picking the wrong cluster as the outlier set. so we won't use the findArtifacts filtering for now.

# so, remove the columns that it added in colData()
hyp2$artifact <- NULL
hyp2$k18 <- NULL
hyp2$k36 <- NULL
hyp2$k60 <- NULL
hyp2$k90 <- NULL
hyp2$k6 <- NULL
hyp2$expr_chrM_ratio_log <- NULL
```


### re: % mitochondrial reads -- do we need a manual threshold for high-end values?
```{r}
### plot
hist(colData(hyp)$expr_chrM_ratio)
# nothing over ~50%, so that's good. we saw in the early part of the dataset that the highest mitochondrial counts were in the VMH, so we don't want to filter too stringently on mitochondrial counts.

# there's one weird sample here with a handful of like 50+% mito reads, so for color scaling purposes plot spots with % mito less than like 80.
# we need to specify the "counts" assay only because that's where it's going to look for this column first
plt <- spatialLIBD::vis_grid_gene(hyp2[,hyp2$expr_chrM_ratio<0.8],geneid = "expr_chrM_ratio",pdf_file = F,return_plots = T,spatial = F,point_size = 0.5,auto_crop = F,assayname = "counts")
do.call("grid.arrange",plt)
# we can see pretty clearly that there is some biology to this -- regionally varies and NOT just at areas where we'd expect it from tissue damage (e.g., edges). so we should filter gently.
sum(colData(hyp2)$expr_chrM_ratio > 0.8,na.rm=T)
# 1 spot
sum(colData(hyp2)$expr_chrM_ratio > 0.5,na.rm=T)
# 10 spots

## how many of these WERENT local outliers for mito proportion?
as.data.table(colData(hyp2))[expr_chrM_ratio_outliers==TRUE&expr_chrM_ratio>0.5,.N]
# only 1 of the 10 over 50% were also an outlier.

# so let's go with 50% as the extreme tentative cutoff, considering how few spots cross that threshold and the biology of the signal. we'll let spotsweeper take care of anything else.

## additional spots to be dropped: 9 
```


### UMIs per spot
### standard threshold is 600 UMIs
### how many spots are under, say, 1000? let's plot the distribution of those.
```{r}
sum(hyp$sum_umi<1000)
### 3750 spots; (of 45641) = 8% of spots

### plot the distribution of UMIs/spot from spots < 1000 umi/spt
ggplot(as.data.frame(colData(hyp2[,hyp2$sum_umi<1000])), aes(x = sum_umi)) +
    geom_histogram(binwidth = 25)+
  geom_vline(xintercept = 600,col="red")
dev.off()

### the tail dropoff here is actually more around 210:
ggplot(as.data.frame(colData(hyp2[,hyp2$sum_umi<1000])), aes(x = sum_umi)) +
    geom_histogram(binwidth = 25)+
  geom_vline(xintercept = 210,col="red")
dev.off()
## which corresponds to...
sum(colData(hyp2)$sum_umi<210,na.rm=T)
## 438 spots, including some not detected by spotsweeper
sum(colData(hyp2)$sum_umi<210&hyp2$sum_umi_outliers==FALSE&hyp2$sum_umi_outliers==FALSE)
## 384 to be exact

## let's make a plot to see what we're removing in this case (supposing we've already dropped local outliers from spotsweeper)
hyp2$under210 <- as.factor(ifelse(hyp2$sum_umi<210,yes="under210",no="over210"))
plt <- spatialLIBD::vis_grid_clus(hyp2[,!(hyp2$sum_gene_outliers)&!(hyp2$sum_umi_outliers)],"under210",pdf_file=NULL,return_plots = T,spatial = F,point_size = 0.5,auto_crop = F)
plt <- lapply(plt,FUN=function(x){x+theme(title = element_text(size=4))})
do.call("grid.arrange",plt)

## it looks like many of these spots with under 210 umis are themselves tissue edges (which is convenient, since spotsweeper isn't going to cut it for this dataset). possibly also a bit of white matter in some samples.
dev.off()
#so 210 the cutoff is.
```

### genes per spot (sum_gene)
# first, let's look at how the number of unique genes detected and number of umis detected relate to get a feel for how much umi space is going to highly expressed genes (i.e., multiple umis on one gene) 
```{r}
ggplot(as.data.table(colData(hyp2[,hyp2$sum_gene_outliers==F&hyp2$sum_umi_outliers==F])),aes(x=sum_gene,y=sum_umi)) +
    geom_point() +
    geom_abline(slope = 1,intercept = 0,col="red")
dev.off()

## that's an x^3 function across the full spectrum of values. let's look at the low end:
ggplot(as.data.table(colData(hyp[,hyp$sum_gene<1000])),aes(x=sum_gene,y=sum_umi)) +
    geom_point() +
    geom_abline(slope = 1.66,intercept = 0,col="red")
dev.off()

## that part is linear, with a slope of about 1.66, as overlaid, meaning the general expectation for the low end of unique gene totals is that there are about 1.66x as many umis, or for u umis, about u/1.66 genes.

## so 210/1.66 = 126 as a matching cutoff for unique genes; lets see how that looks on the histogram.
ggplot(as.data.table(colData(hyp[,hyp$sum_gene<1000])),aes(x=sum_gene)) +
    geom_histogram(binwidth=10) +
    geom_vline(xintercept = 126,col="red")
dev.off()

## wam bam thank you ma'am that's right off the cliff for sum_gene. let's make sure there's not any spatial patterns under 126 that would indicate biology worth keeping at an even more lenient threshold.

hyp$gene126 <- as.factor(ifelse(hyp$sum_gene<126,yes="under126",no="over126"))
plt <- spatialLIBD::vis_grid_clus(hyp,"gene126",pdf_file = NULL,spatial = F,return_plots = T,auto_crop = T,point_size = 0.67)
do.call("grid.arrange",plt)
dev.off()

### that's practically just the same spots with <210 umis, and a very small # of interior spots in other samples. so 126 minimum unique genes and 210 minimum UMIs.

### clear out the temporary columns used for labeling
hyp2$under210 <- NULL
hyp2$gene166 <- NULL
```

### apply mitochondria, UMI, unique gene cutoffs per above (<50%, >= 210 umis, > >= 126 unique genes)
```{r}
ncol(hyp2[,hyp2$in_tissue==T & hyp2$sum_umi>=210 & hyp2$sum_gene>=126 & hyp2$expr_chrM_ratio<=0.5 & hyp2$sum_umi_outliers==F & hyp2$sum_gene_outliers==F & hyp2$expr_chrM_ratio_outliers==F])/ncol(hyp[,hyp$in_tissue==T])
# 98.757% spots retained for now!
hyp2 <- hyp2[,hyp2$in_tissue==T & hyp2$sum_umi>=210 & hyp2$sum_gene>=126 & hyp2$expr_chrM_ratio<=0.5 & hyp2$sum_umi_outliers==F & hyp2$sum_gene_outliers==F & hyp2$expr_chrM_ratio_outliers==F]

dim(hyp2)[2]
# number of post-filter spots: 45074

# remove genes for which there is no longer data
no_expr <- which(rowSums(counts(hyp2)) == 0)
hyp2 <- hyp2[-no_expr, ]
dim(hyp2)[1]
# 30361 genes post-filtering

rm(hyp)
gc(full=T)
```

### apply library-size normalized, LOG COUNT normalization AFTER filtering the data.
```{r}
hyp2 <- computeLibraryFactors(hyp2)
# peek
hist(sizeFactors(hyp2),breaks=100)
dev.off()
# append log counts using scater (don't use the SpatialExpt:: function here, that's for RETRIEVING an assay by the name of "logcounts" from spe objs)
hyp2 <- scater::logNormCounts(hyp2)
```

```{r}
saveRDS(hyp2,"processed-data/03-QC_filters/hypN10_umi210_gene126_chrm50_spotsweeped_lognorm_rotsNmirrors_071924.RDS")

rm(list=ls())
gc(full=T)
```


#### session info pasted after call.
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] SpotSweeper_1.0.2           gridExtra_2.3              
 [3] scran_1.32.0                BiocParallel_1.38.0        
 [5] scater_1.32.0               scuttle_1.14.0             
 [7] spatialLIBD_1.16.2          ggspavis_1.10.0            
 [9] ggplot2_3.5.1               SpatialExperiment_1.14.0   
[11] SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0
[13] Biobase_2.64.0              GenomicRanges_1.56.1       
[15] GenomeInfoDb_1.40.1         IRanges_2.38.1             
[17] S4Vectors_0.42.1            BiocGenerics_0.50.0        
[19] MatrixGenerics_1.16.0       matrixStats_1.3.0          
[21] data.table_1.15.4           rlang_1.1.4                

loaded via a namespace (and not attached):
  [1] later_1.3.2               BiocIO_1.14.0             bitops_1.0-7             
  [4] filelock_1.0.3            fields_16.2               tibble_3.2.1             
  [7] XML_3.99-0.17             lifecycle_1.0.4           edgeR_4.2.0              
 [10] doParallel_1.0.17         rprojroot_2.0.4           lattice_0.22-6           
 [13] MASS_7.3-61               magrittr_2.0.3            sass_0.4.9               
 [16] limma_3.60.3              plotly_4.10.4             rmarkdown_2.27           
 [19] jquerylib_0.1.4           yaml_2.3.9                metapod_1.12.0           
 [22] httpuv_1.6.15             spam_2.10-0               ggside_0.3.1             
 [25] sessioninfo_1.2.2         cowplot_1.1.3             DBI_1.2.3                
 [28] RColorBrewer_1.1-3        golem_0.4.1               maps_3.4.2               
 [31] abind_1.4-5               zlibbioc_1.50.0           purrr_1.0.2              
 [34] RCurl_1.98-1.14           rappdirs_0.3.3            GenomeInfoDbData_1.2.12  
 [37] ggrepel_0.9.5             irlba_2.3.5.1             terra_1.7-78             
 [40] dqrng_0.4.1               DelayedMatrixStats_1.26.0 codetools_0.2-20         
 [43] DelayedArray_0.30.1       DT_0.33                   tidyselect_1.2.1         
 [46] farver_2.1.2              UCSC.utils_1.0.0          ScaledMatrix_1.12.0      
 [49] viridis_0.6.5             shinyWidgets_0.8.6        BiocFileCache_2.12.0     
 [52] GenomicAlignments_1.40.0  jsonlite_1.8.8            escheR_1.4.0             
 [55] BiocNeighbors_1.22.0      iterators_1.0.14          foreach_1.5.2            
 [58] tools_4.4.1               Rcpp_1.0.13               glue_1.7.0               
 [61] SparseArray_1.4.8         xfun_0.45                 here_1.0.1               
 [64] dplyr_1.1.4               withr_3.0.0               BiocManager_1.30.23      
 [67] fastmap_1.2.0             bluster_1.14.0            fansi_1.0.6              
 [70] digest_0.6.36             rsvd_1.0.5                R6_2.5.1                 
 [73] mime_0.12                 colorspace_2.1-0          RSQLite_2.3.7            
 [76] config_0.3.2              utf8_1.2.4                tidyr_1.3.1              
 [79] generics_0.1.3            rtracklayer_1.64.0        httr_1.4.7               
 [82] htmlwidgets_1.6.4         S4Arrays_1.4.1            pkgconfig_2.0.3          
 [85] gtable_0.3.5              blob_1.2.4                XVector_0.44.0           
 [88] htmltools_0.5.8.1         dotCall64_1.1-1           scales_1.3.0             
 [91] png_0.1-8                 attempt_0.3.1             knitr_1.48               
 [94] rstudioapi_0.16.0         rjson_0.2.21              curl_5.2.1               
 [97] cachem_1.1.0              BiocVersion_3.19.1        parallel_4.4.1           
[100] vipor_0.4.7               AnnotationDbi_1.66.0      restfulr_0.0.15          
[103] pillar_1.9.0              grid_4.4.1                vctrs_0.6.5              
[106] promises_1.3.0            BiocSingular_1.20.0       dbplyr_2.5.0             
[109] beachmat_2.20.0           cluster_2.1.6             xtable_1.8-4             
[112] beeswarm_0.4.0            paletteer_1.6.0           evaluate_0.24.0          
[115] magick_2.8.3              cli_3.6.3                 locfit_1.5-9.10          
[118] compiler_4.4.1            Rsamtools_2.20.0          crayon_1.5.3             
[121] labeling_0.4.3            rematch2_2.1.2            spatialEco_2.0-2         
[124] ggbeeswarm_0.7.2          viridisLite_0.4.2         munsell_0.5.1            
[127] Biostrings_2.72.1         lazyeval_0.2.2            Matrix_1.7-0             
[130] ExperimentHub_2.12.0      benchmarkme_1.0.8         sparseMatrixStats_1.16.0 
[133] bit64_4.0.5               KEGGREST_1.44.1           statmod_1.5.0            
[136] shiny_1.8.1.1             AnnotationHub_3.12.0      igraph_2.0.3             
[139] memoise_2.0.1             bslib_0.7.0               benchmarkmeData_1.0.4    
[142] bit_4.0.5                
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-07-19
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────────────
 ! package              * version   date (UTC) lib source
   abind                  1.4-5     2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi          1.66.0    2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub          3.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   attempt                0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
   beachmat               2.20.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm               0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme            1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData        1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
   Biobase              * 2.64.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocFileCache          2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics         * 0.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager            1.30.23   2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors          1.22.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel         * 1.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocSingular           1.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion            3.19.1    2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings             2.72.1    2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                    4.0.5     2022-11-15 [1] CRAN (R 4.4.0)
   bit64                  4.0.5     2020-08-30 [1] CRAN (R 4.4.0)
   bitops                 1.0-7     2021-04-24 [1] CRAN (R 4.4.0)
   blob                   1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   bluster                1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   bslib                  0.7.0     2024-03-29 [1] CRAN (R 4.4.0)
   cachem                 1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
 P cli                    3.6.3     2024-06-21 [2] CRAN (R 4.4.0)
   cluster                2.1.6     2023-12-01 [1] CRAN (R 4.4.1)
   codetools              0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorspace             2.1-0     2023-01-23 [1] CRAN (R 4.4.0)
   config                 0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
   crayon                 1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   curl                   5.2.1     2024-03-01 [1] CRAN (R 4.4.0)
   data.table           * 1.15.4    2024-03-30 [2] CRAN (R 4.4.0)
   DBI                    1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                 2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray           0.30.1    2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats     1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   digest                 0.6.36    2024-06-23 [1] CRAN (R 4.4.0)
   doParallel             1.0.17    2022-02-07 [1] CRAN (R 4.4.0)
   dotCall64              1.1-1     2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                  1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   dqrng                  0.4.1     2024-05-28 [1] CRAN (R 4.4.0)
   DT                     0.33      2024-04-04 [1] CRAN (R 4.4.0)
   edgeR                  4.2.0     2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   escheR                 1.4.0     2024-05-16 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate               0.24.0    2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub          2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fansi                  1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   farver                 2.1.2     2024-05-13 [1] CRAN (R 4.4.0)
   fastmap                1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fields                 16.2      2024-06-27 [1] CRAN (R 4.4.0)
   filelock               1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   foreach                1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
   generics               0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb         * 1.40.1    2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData       1.2.12    2024-05-01 [1] Bioconductor
   GenomicAlignments      1.40.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges        * 1.56.1    2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm             0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2              * 3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrepel                0.9.5     2024-01-10 [1] CRAN (R 4.4.0)
   ggside                 0.3.1     2024-03-01 [1] CRAN (R 4.4.0)
   ggspavis             * 1.10.0    2024-05-16 [1] Bioconductor 3.19 (R 4.4.0)
   glue                   1.7.0     2024-01-09 [1] CRAN (R 4.4.0)
   golem                  0.4.1     2023-06-05 [1] CRAN (R 4.4.0)
   gridExtra            * 2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gtable                 0.3.5     2024-04-22 [1] CRAN (R 4.4.0)
   here                   1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools              0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets            1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                 1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
   httr                   1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   igraph                 2.0.3     2024-03-13 [1] CRAN (R 4.4.0)
   IRanges              * 2.38.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                  2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   iterators              1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
   jquerylib              0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite               1.8.8     2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST               1.44.1    2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   knitr                  1.48      2024-07-07 [1] CRAN (R 4.4.1)
   labeling               0.4.3     2023-08-29 [1] CRAN (R 4.4.0)
   later                  1.3.2     2023-12-06 [1] CRAN (R 4.4.0)
   lattice                0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval               0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle              1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                  3.60.3    2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                 1.5-9.10  2024-06-24 [1] CRAN (R 4.4.0)
   magick                 2.8.3     2024-02-18 [1] CRAN (R 4.4.0)
   magrittr               2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   maps                   3.4.2     2023-12-15 [1] CRAN (R 4.4.0)
   MASS                   7.3-61    2024-06-13 [1] CRAN (R 4.4.0)
   Matrix                 1.7-0     2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics       * 1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats          * 1.3.0     2024-04-11 [1] CRAN (R 4.4.0)
   memoise                2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   metapod                1.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   mime                   0.12      2021-09-28 [1] CRAN (R 4.4.0)
   munsell                0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   paletteer              1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
   pillar                 1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   plotly                 4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
   png                    0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   promises               1.3.0     2024-04-05 [1] CRAN (R 4.4.0)
   purrr                  1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R6                     2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs               0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                   1.0.13    2024-07-17 [1] CRAN (R 4.4.1)
   RCurl                  1.98-1.14 2024-01-09 [1] CRAN (R 4.4.0)
   rematch2               2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
   restfulr               0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
   rjson                  0.2.21    2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown              2.27      2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot              2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools              2.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                2.3.7     2024-05-27 [1] CRAN (R 4.4.0)
   rstudioapi             0.16.0    2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                   1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer            1.64.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   S4Arrays               1.4.1     2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors            * 0.42.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                   0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix           1.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                 1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater               * 1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scran                * 1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scuttle              * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo            1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   shiny                  1.8.1.1   2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets           0.8.6     2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment * 1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spam                   2.10-0    2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray            1.4.8     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats      1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spatialEco             2.0-2     2023-11-17 [1] CRAN (R 4.4.0)
   SpatialExperiment    * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD          * 1.16.2    2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   SpotSweeper          * 1.0.2     2024-07-14 [1] Bioconductor 3.19 (R 4.4.1)
   statmod                1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   SummarizedExperiment * 1.34.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   terra                  1.7-78    2024-05-22 [1] CRAN (R 4.4.0)
   tibble                 3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                  1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect             1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   UCSC.utils             1.0.0     2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   utf8                   1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                  0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                  0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite            0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   withr                  3.0.0     2024-01-16 [1] CRAN (R 4.4.0)
   xfun                   0.45      2024-06-16 [1] CRAN (R 4.4.0)
   XML                    3.99-0.17 2024-06-25 [1] CRAN (R 4.3.3)
   xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
   XVector                0.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                   2.3.9     2024-07-05 [1] CRAN (R 4.4.0)
   zlibbioc               1.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────
