---
title: "Fig 6-Sex DE and Disease"
output: html_document
date: "2024-09-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(SpatialExperiment)
library(escheR)
library(ggplot2)
library(gridExtra)
library(ggtext)
library(Cairo)
library(ggrastr)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

read GSEA results of markers, sex DEGs in GWAS-derived gene sets; subset to sets from the union of GWAS TWAS and disgenet
```{r}
mkgs <- readRDS("spatial_HYP/processed-data/11-GSEA/05-Common variant disease gene enrichment in markers.RDS")
mkgs <- mkgs[genesets_from=="union"]

sxgs <- readRDS("spatial_HYP/processed-data/11-GSEA/06-Common variant disease gene enrichment in sex DEGs.RDS")
sxgs <- sxgs[genesets_from=="union"]
```


Panel A: domain-level, markers
```{r}
dmk <- mkgs[spdom %in% c("k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XARC","k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XVMH")]

dmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XARC",spdom:="ARC"]
dmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XVMH",spdom:="VMH"]

## reorder traits of interest into broad groups (not that they will be explicitly labeled as such, but still)

dmk[,pathway:=factor(pathway,levels=c("Lefthanded","BMI","Anorexia","Panic","PTSD","GAD","OCD","TS","ADHD","NDD","ASD","OUD","AUD","Cannabis","SCZ","BIP","MDD","AD"))]

cairo_pdf("manuscript_plots/Fig6/6A-Marker_Dis_GSEA.pdf",height=4,width=2.75)
ggplot(dmk,aes(x=spdom,y=pathway,fill=NES))+
    geom_tile(linewidth = 0.25,colour = "black")+
    geom_text(label=ifelse(dmk$padj<0.05,"*",""),size = 5,nudge_y = -0.2)+
    geom_text(label=ifelse(dmk$padj>0.05&dmk$pval<0.05,"•",""),size = 4)+
    scale_fill_viridis_c(begin=0.2,end=1,option="E")+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    labs(fill="Normalized\nEnrichment\nScore")+
    xlab("Visium Domain")+
    ylab("Disease / Trait\nGene Set") +
    theme(plot.background = element_blank(),axis.text.x = element_text(size=9,angle=90,vjust=0.5),axis.text.y=element_text(size=9),axis.title.x=element_text(size=10,hjust=0.5),axis.title.y=element_text(size=10,vjust=0.5,margin=margin(0,-0.15,0,0,"in")),legend.title = element_text(size=9,hjust=0.5),legend.text=element_text(size=8),legend.margin=margin(0,0,0,-0.1,"in"))
dev.off()

rm(dmk)
```

# panel B: VMH/ARC k15 subclusters, markers
```{r}
cmk <- mkgs[spdom %in% c("k15_HARMONYlmbna_nnsvg10_60kiter_X4","k15_HARMONYlmbna_nnsvg10_60kiter_X6","k15_HARMONYlmbna_nnsvg10_60kiter_X7","k15_HARMONYlmbna_nnsvg10_60kiter_X12")]

cmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_X4",spdom:="ARC.1"]
cmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_X6",spdom:="ARC.2"]
cmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_X7",spdom:="VMH.1"]
cmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_X12",spdom:="VMH.2"]

cmk[,pathway:=factor(pathway,levels=c("Lefthanded","BMI","Anorexia","Panic","PTSD","GAD","OCD","TS","ADHD","NDD","ASD","OUD","AUD","Cannabis","SCZ","BIP","MDD","AD"))]

cairo_pdf("manuscript_plots/Fig6/6B-SubclusMarker_Dis_GSEA.pdf",height=4,width=2.75)
ggplot(cmk,aes(x=spdom,y=pathway,fill=NES))+
    geom_tile(linewidth = 0.25,colour = "black")+
    geom_text(label=ifelse(cmk$padj<0.05,"*",""),size = 4,nudge_y = -0.2)+
    geom_text(label=ifelse(cmk$padj>0.05&cmk$pval<0.05,"•",""),size = 3)+
    scale_fill_viridis_c(begin=0.2,option = "E")+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    labs(fill="Normalized\nEnrichment\nScore")+
    xlab("Visium Subcluster")+
    ylab("Disease / Trait\nGene Set") +
    theme(plot.background = element_blank(),axis.text.x = element_text(size=9,angle=90,vjust=0.5),axis.text.y=element_text(size=9),axis.title.x=element_text(size=10,hjust=0.5),axis.title.y=element_text(size=10,vjust=0.5,margin=margin(0,-0.15,0,0,"in")),legend.title = element_text(size=9,hjust=0.5),legend.text=element_text(size=8),legend.margin=margin(0,0,0,-0.1,"in"))
dev.off()








```     

panel C: sex DE, collapsed domains
```{r}
dsx <- sxgs[spdom %in% c("BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_ARC","BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_VMH")]

dsx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_ARC",spdom:="ARC"]
dsx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_VMH",spdom:="VMH"]

## reorder traits of interest into broad groups (not that they will be explicitly labeled as such, but still)
dsx[,pathway:=factor(pathway,levels=c("Lefthanded","BMI","Anorexia","Panic","PTSD","GAD","OCD","TS","ADHD","NDD","ASD","OUD","AUD","Cannabis","SCZ","BIP","MDD","AD"))]

## make plot; use a rearranged version of the wesanderson:: palette "Darjeeling2" (colors 3,1,4,2)
scalefact <- max(dsx$NES)-min(dsx$NES)
zeropt <- -min(dsx$NES)/scalefact
zero.to.zeropt.mid <- zeropt/2
zeropt.to.one.mid <- zeropt*1.5

cairo_pdf("manuscript_plots/Fig6/6C-DomainSexDE_Dis_GSEA.pdf",height=4,width=2.75)
ggplot(dsx,aes(x=spdom,y=pathway,fill=NES))+
    geom_tile(linewidth = 0.25,colour = "black")+
    geom_text(label=ifelse(dsx$padj<0.05,"*",""),size = 5,nudge_y = -0.2)+
    geom_text(label=ifelse(dsx$padj>0.05&dsx$pval<0.05,"•",""),size = 4)+
    #scale_fill_viridis_c(begin=0.15,end=0.85,option="H",direction = -1)+
    scale_fill_gradientn(colors=c("#D69C4E","#ECCBAE","#ABDDDE","#046C9A"),values = c(0,zero.to.zeropt.mid,zeropt.to.one.mid,1))+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    labs(fill="Normalized\nEnrichment\nScore\n(>0: Male-Up,\n<0: Female-Up)")+
    xlab("Visium Domain")+
    ylab("Disease / Trait\nGene Set") +
    theme(plot.background = element_blank(),axis.text.x = element_text(size=9,angle=90,vjust=0.5),axis.text.y=element_text(size=9),axis.title.x=element_text(size=10,hjust=0.5),axis.title.y=element_text(size=10,vjust=0.5,margin=margin(0,-0.15,0,0,"in")),legend.title = element_text(size=9,hjust=0.5),legend.text=element_text(size=8),legend.margin=margin(0,0,0,-0.1,"in"))
dev.off()

rm(dsx)
```

same for subclusts
```{r}
csx <- sxgs[spdom %in% c("BSpace_k15_HARMONYlmbna_nnsvg10_X4","BSpace_k15_HARMONYlmbna_nnsvg10_X6","BSpace_k15_HARMONYlmbna_nnsvg10_X7","BSpace_k15_HARMONYlmbna_nnsvg10_X12")]

csx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_X4",spdom:="ARC.1"]
csx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_X6",spdom:="ARC.2"]
csx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_X7",spdom:="VMH.1"]
csx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_X12",spdom:="VMH.2"]

csx[,pathway:=factor(pathway,levels=c("Lefthanded","BMI","Anorexia","Panic","PTSD","GAD","OCD","TS","ADHD","NDD","ASD","OUD","AUD","Cannabis","SCZ","BIP","MDD","AD"))]

## make plot; use a rearranged version of the wesanderson:: palette "Darjeeling2" (colors 3,1,4,2)
scalefact <- max(csx$NES)-min(csx$NES)
zeropt <- -min(csx$NES)/scalefact
zero.to.zeropt.mid <- zeropt/2
zeropt.to.one.mid <- zeropt*1.5

cairo_pdf("manuscript_plots/Fig6/6D-SubclusSexDE_Dis_GSEA.pdf",height=4,width=3.25)
ggplot(csx,aes(x=spdom,y=pathway,fill=NES))+
    geom_tile(linewidth = 0.25,colour = "black")+
    geom_text(label=ifelse(csx$padj<0.05,"*",""),size = 4,nudge_y = -0.2)+
    geom_text(label=ifelse(csx$padj>0.05&csx$pval<0.05,"•",""),size = 3)+
    scale_fill_gradientn(colors=c("#D69C4E","#ECCBAE","#ABDDDE","#046C9A"),values = c(0,zero.to.zeropt.mid,zeropt.to.one.mid,1))+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    labs(fill="Normalized\nEnrichment\nScore\n(>0: Male-Up,\n<0: Female-Up)")+
    xlab("Visium Domain")+
    ylab("Disease / Trait\nGene Set") +
    theme(plot.background = element_blank(),axis.text.x = element_text(size=9,angle=90,vjust=0.5),axis.text.y=element_text(size=9),axis.title.x=element_text(size=10,hjust=0.5),axis.title.y=element_text(size=10,vjust=0.5,margin=margin(0,-0.15,0,0,"in")),legend.title = element_text(size=9,hjust=0.5),legend.text=element_text(size=8),legend.margin=margin(0,0,0,-0.1,"in"))
dev.off()
```


## looking at the overlap between leadingEdge genes driving male ASD enrichment in ARC with genes on Xenium, MYT1L, an obesity+ASD syndrome, is nominally sex DE in ARC_4 and much more weakly so in ARC_5. let's take a look at this warlock

```{r}
hypx <- readRDS("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-sfe-in_staggered-coords_genetarg-only_NONlog-norm.RDS")


bkcl <- fread("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-BanksyClusts_M0lam0_leiden_multisamp.txt")
bkanno <- fread("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/02-M0l0kg6_topClusMarkers_celltypes_annotated.txt")
fineanno <- fread("manuscript_plot_code/xARCxVMH_cluster_detailedlabs_andplotformatnames.txt")
fineanno[plotclusname=="*POMC*-Oligo",plotclusname:="*POMC*-Oligo Mixed"]

smoove <- fread("xenium_HYP/processed-data/07_Domains-subdomains_by_knnSmoothing_Banksytypes/03b-ARCVMHdomains_2stepsmooth_VMH-k10-0.2-VMH-k200-0.2_ARC-k50-0.1_ARC-k500-0.5.txt")

setnames(bkcl,2,"cl")
bkanno <- unique(bkanno[,.(clus,bjm_annot)])

# drop DISCARDS
bkanno <- bkanno[!(bjm_annot %in% grep(bjm_annot,pattern="DISCARD",value=T))]
bkcl <- merge.data.table(bkcl,bkanno,by.x="cl",by.y="clus")
bkcl <- merge.data.table(bkcl,fineanno[,.(bjm_annot,dom.clus,plotclusname)],by="bjm_annot",all.x=T)
bkcl[!is.na(plotclusname),bjm_annot:=plotclusname]

# add clusters and smoothed domains
hypx$dom <- DataFrame(smoove[,.(rn,dualVMHARC4)],row.names=smoove$rn)[colnames(hypx),"dualVMHARC4"]

hypx$bkcl <- DataFrame(bkcl,row.names=bkcl$rn)[colnames(hypx),"bjm_annot"]

```

load cell type palettes
```{r}
pals <- readRDS("manuscript_plot_code/domain_and_xencluster_palettes_CHECKPALNAMESBEFOREUS.RDS")

xenarcpal <- pals[[3]]
names(xenarcpal)[5] <- "*TAC3*-*ESR1*"


stopifnot(names(xenarcpal) %in% unique(bkcl[dom.clus %in% grep(dom.clus,pattern="ARC",value=T),plotclusname]))
```

### LEFT OFF HERE ### 

### plot MYT1L in two male, two female samples showing ARC_4 and ARC_5 (TAC3-ESR1 and POMC-Oligo Mixed). append manuscript sample ids
```{r}
mytx <- hypx["MYT1L"]
mytx <- mytx[,mytx$dom=="ARC"]
mytx <- mytx[,mytx$bkcl %in% c("*TAC3*-*ESR1*","*POMC*-Oligo Mixed")]

# add manuscript identifiers
mscriptid <- fread("standardized_sampleids_for_plotting.txt")
tmpcd <- as.data.table(colData(mytx),keep.rownames=T)
tmpcd <- merge.data.table(tmpcd,mscriptid[,.(sample_id,manuscript_id)],all.x=T,by="sample_id")
tmpcd <- DataFrame(tmpcd,row.names=tmpcd$rn)[colnames(mytx),]
tmpcd$rn <- NULL
colData(mytx) <- tmpcd
mytx$sample_id <- mytx$manuscript_id

mytxa <- mytx[,mytx$sample_id %in% c("x1735C_M","x1225B_M","x8741C_F","x6588C_F")]
mytxa$`log counts` <- logcounts(mytxa)["MYT1L",]

mytx.m1 <- mytxa[,mytxa$sample_id=="x1735C_M"]

mytx.m2 <- mytxa[,mytxa$sample_id=="x1225B_M"]

mytx.f1 <- mytxa[,mytxa$sample_id=="x8741C_F"]

mytx.f2 <-  mytxa[,mytxa$sample_id=="x6588C_F"]

mytxs <- list(mytx.m1,mytx.m2,mytx.f1,mytx.f2)

mytxplt <- lapply(mytxs,FUN=function(m){
    # m <- m[,m$bkcl=="*TAC3*-*ESR1*"]
    colData(m)$`log counts` <- logcounts(m)["MYT1L",]
    p <- as.data.table(cbind(spatialCoords(m),m$`log counts`,m$bkcl))
    setnames(p,c("sdimx","sdimy","log counts","bkcl"))
    p[,sdimx:=as.numeric(sdimx)]
    p[,sdimy:=as.numeric(sdimy)]
    p[,`log counts`:=as.numeric(`log counts`)]
    #p <- p |> add_ground("bkcl",stroke = 0.35,point_size = 0.875)
    #p <- p |> add_fill("log counts",size=0.875,point_size = 0.875)
    p2 <- ggplot(p,aes(x=bkcl,y=`log counts`,col=bkcl,fill=`log counts`))+
        #gbeeswarm::geom_beeswarm(size=1.25,stroke =0.45,shape=21,corral.width = 0.1,)+
        rasterize(geom_quasirandom(dodge.width=1.5,size=0.67,stroke =0.225,shape=21),dpi = 600,dev="cairo_png")+
        scale_color_manual(values=xenarcpal)+
        scale_fill_gradient(low="white",high="black",limits=c(min(mytxa$`log counts`),max(mytxa$`log counts`)))+
        ggtitle(paste0("*MYT1L* ",unique(m$sample_id)))+
        labs(fill="log counts",color="xARC Cluster")+
        theme(plot.title.position = "plot",plot.title = element_markdown(size=10,hjust=0.5),legend.text = element_text(size=7),legend.key.size = ggplot2::unit(0.125,"in") ,legend.title=element_text(size=7.5,hjust=0.5),axis.text.x=element_blank(),axis.text.y=element_blank(),axis.title.x=element_blank(),axis.title.y=element_blank(),axis.ticks.x = element_blank(),axis.ticks.y=element_blank())
    return(p2)
})

### helper function to extract a legend and pipe back to a diff plot: from https://statisticsglobe.com/add-common-legend-to-combined-ggplot2-plots-in-r/#example-2-add-shared-legend-to-ggplot2-plots-using-gridextra-package 
extract_legend <- function(my_ggp) {
  step1 <- ggplot_gtable(ggplot_build(my_ggp))
  step2 <- which(sapply(step1$grobs, function(x) x$name) == "guide-box")
  step3 <- step1$grobs[[step2]]
  return(step3)
}
#####

leg.tmp <- extract_legend(mytxplt[[1]])
dev.off()

mytxplt<- lapply(mytxplt,FUN=function(x){x+guides(fill="none")+guides(color="none")})
```

```{r}
pdf("manuscript_plots/Fig6/6E-MYT1L beeswarms.pdf",height=4,width=8)
do.call("grid.arrange",list(arrangeGrob(grobs=mytxplt,nrow=1),leg.tmp,ncol=2,widths=c(6.75,1.25)))
dev.off()




```
