---
title: "Fig 5-Sex DE and Disease"
output: html_document
date: "2024-09-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(gridExtra)
library(ggtext)
library(Cairo)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

read GSEA results of markers, sex DEGs in GWAS-derived gene sets; subset to sets from the union of GWAS TWAS and disgenet
```{r}
mkgs <- readRDS("spatial_HYP/processed-data/11-GSEA/05-Common variant disease gene enrichment in markers.RDS")
mkgs <- mkgs[genesets_from=="union"]

sxgs <- readRDS("spatial_HYP/processed-data/11-GSEA/06-Common variant disease gene enrichment in sex DEGs.RDS")
sxgs <- sxgs[genesets_from=="union"]
```


Panel A: domain-level, markers
```{r}
dmk <- mkgs[spdom %in% c("k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XARC","k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XVMH")]

dmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XARC",spdom:="ARC"]
dmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_collapsed_XVMH",spdom:="VMH"]

## reorder traits of interest into broad groups (not that they will be explicitly labeled as such, but still)

dmk[,pathway:=factor(pathway,levels=c("Lefthanded","BMI","Anorexia","Panic","PTSD","GAD","OCD","TS","ADHD","NDD","ASD","OUD","AUD","Cannabis","SCZ","BIP","MDD","AD"))]

cairo_pdf("manuscript_plots/Fig5/5A-Marker_Dis_GSEA.pdf",height=4,width=2.75)
ggplot(dmk,aes(x=spdom,y=pathway,fill=NES))+
    geom_tile(linewidth = 0.25,colour = "black")+
    geom_text(label=ifelse(dmk$padj<0.05,"*",""),size = 5,nudge_y = -0.2)+
    geom_text(label=ifelse(dmk$padj>0.05&dmk$pval<0.05,"•",""),size = 4)+
    scale_fill_viridis_c(begin=0.2,end=1,option="E")+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    labs(fill="Normalized\nEnrichment\nScore")+
    xlab("Visium Domain")+
    ylab("Disease / Trait\nGene Set") +
    theme(plot.background = element_blank(),axis.text.x = element_text(size=9,angle=90,vjust=0.5),axis.text.y=element_text(size=9),axis.title.x=element_text(size=10,hjust=0.5),axis.title.y=element_text(size=10,vjust=0.5,margin=margin(0,-0.15,0,0,"in")),legend.title = element_text(size=9,hjust=0.5),legend.text=element_text(size=8),legend.margin=margin(0,0,0,-0.1,"in"))
dev.off()

rm(dmk)
```

# panel B: VMH/ARC k15 subclusters, markers
```{r}
cmk <- mkgs[spdom %in% c("k15_HARMONYlmbna_nnsvg10_60kiter_X4","k15_HARMONYlmbna_nnsvg10_60kiter_X6","k15_HARMONYlmbna_nnsvg10_60kiter_X7","k15_HARMONYlmbna_nnsvg10_60kiter_X12")]

cmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_X4",spdom:="ARC.1"]
cmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_X6",spdom:="ARC.2"]
cmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_X7",spdom:="VMH.1"]
cmk[spdom=="k15_HARMONYlmbna_nnsvg10_60kiter_X12",spdom:="VMH.2"]

cmk[,pathway:=factor(pathway,levels=c("Lefthanded","BMI","Anorexia","Panic","PTSD","GAD","OCD","TS","ADHD","NDD","ASD","OUD","AUD","Cannabis","SCZ","BIP","MDD","AD"))]

cairo_pdf("manuscript_plots/Fig5/5B-SubclusMarker_Dis_GSEA.pdf",height=4,width=2.75)
ggplot(cmk,aes(x=spdom,y=pathway,fill=NES))+
    geom_tile(linewidth = 0.25,colour = "black")+
    geom_text(label=ifelse(cmk$padj<0.05,"*",""),size = 4,nudge_y = -0.2)+
    geom_text(label=ifelse(cmk$padj>0.05&cmk$pval<0.05,"•",""),size = 3)+
    scale_fill_viridis_c(begin=0.2,option = "E")+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    labs(fill="Normalized\nEnrichment\nScore")+
    xlab("Visium Subcluster")+
    ylab("Disease / Trait\nGene Set") +
    theme(plot.background = element_blank(),axis.text.x = element_text(size=9,angle=90,vjust=0.5),axis.text.y=element_text(size=9),axis.title.x=element_text(size=10,hjust=0.5),axis.title.y=element_text(size=10,vjust=0.5,margin=margin(0,-0.15,0,0,"in")),legend.title = element_text(size=9,hjust=0.5),legend.text=element_text(size=8),legend.margin=margin(0,0,0,-0.1,"in"))
dev.off()








```     

panel C: sex DE, collapsed domains
```{r}
dsx <- sxgs[spdom %in% c("BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_ARC","BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_VMH")]

dsx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_ARC",spdom:="ARC"]
dsx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_VMH",spdom:="VMH"]

## reorder traits of interest into broad groups (not that they will be explicitly labeled as such, but still)
dsx[,pathway:=factor(pathway,levels=c("Lefthanded","BMI","Anorexia","Panic","PTSD","GAD","OCD","TS","ADHD","NDD","ASD","OUD","AUD","Cannabis","SCZ","BIP","MDD","AD"))]

## make plot; use a rearranged version of the wesanderson:: palette "Darjeeling2" (colors 3,1,4,2)
scalefact <- max(dsx$NES)-min(dsx$NES)
zeropt <- -min(dsx$NES)/scalefact
zero.to.zeropt.mid <- zeropt/2
zeropt.to.one.mid <- zeropt*1.5

cairo_pdf("manuscript_plots/Fig5/5C-DomainSexDE_Dis_GSEA.pdf",height=4,width=2.75)
ggplot(dsx,aes(x=spdom,y=pathway,fill=NES))+
    geom_tile(linewidth = 0.25,colour = "black")+
    geom_text(label=ifelse(dsx$padj<0.05,"*",""),size = 5,nudge_y = -0.2)+
    geom_text(label=ifelse(dsx$padj>0.05&dsx$pval<0.05,"•",""),size = 4)+
    #scale_fill_viridis_c(begin=0.15,end=0.85,option="H",direction = -1)+
    scale_fill_gradientn(colors=c("#D69C4E","#ECCBAE","#ABDDDE","#046C9A"),values = c(0,zero.to.zeropt.mid,zeropt.to.one.mid,1))+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    labs(fill="Normalized\nEnrichment\nScore\n(>0: Male-Up,\n<0: Female-Up)")+
    xlab("Visium Domain")+
    ylab("Disease / Trait\nGene Set") +
    theme(plot.background = element_blank(),axis.text.x = element_text(size=9,angle=90,vjust=0.5),axis.text.y=element_text(size=9),axis.title.x=element_text(size=10,hjust=0.5),axis.title.y=element_text(size=10,vjust=0.5,margin=margin(0,-0.15,0,0,"in")),legend.title = element_text(size=9,hjust=0.5),legend.text=element_text(size=8),legend.margin=margin(0,0,0,-0.1,"in"))
dev.off()

rm(dsx)
```

same for subclusts
```{r}
csx <- sxgs[spdom %in% c("BSpace_k15_HARMONYlmbna_nnsvg10_X4","BSpace_k15_HARMONYlmbna_nnsvg10_X6","BSpace_k15_HARMONYlmbna_nnsvg10_X7","BSpace_k15_HARMONYlmbna_nnsvg10_X12")]

csx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_X4",spdom:="ARC.1"]
csx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_X6",spdom:="ARC.2"]
csx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_X7",spdom:="VMH.1"]
csx[spdom=="BSpace_k15_HARMONYlmbna_nnsvg10_X12",spdom:="VMH.2"]

csx[,pathway:=factor(pathway,levels=c("Lefthanded","BMI","Anorexia","Panic","PTSD","GAD","OCD","TS","ADHD","NDD","ASD","OUD","AUD","Cannabis","SCZ","BIP","MDD","AD"))]

## make plot; use a rearranged version of the wesanderson:: palette "Darjeeling2" (colors 3,1,4,2)
scalefact <- max(csx$NES)-min(csx$NES)
zeropt <- -min(csx$NES)/scalefact
zero.to.zeropt.mid <- zeropt/2
zeropt.to.one.mid <- zeropt*1.5

cairo_pdf("manuscript_plots/Fig5/5D-SubclusSexDE_Dis_GSEA.pdf",height=4,width=3.25)
ggplot(csx,aes(x=spdom,y=pathway,fill=NES))+
    geom_tile(linewidth = 0.25,colour = "black")+
    geom_text(label=ifelse(csx$padj<0.05,"*",""),size = 4,nudge_y = -0.2)+
    geom_text(label=ifelse(csx$padj>0.05&csx$pval<0.05,"•",""),size = 3)+
    scale_fill_gradientn(colors=c("#D69C4E","#ECCBAE","#ABDDDE","#046C9A"),values = c(0,zero.to.zeropt.mid,zeropt.to.one.mid,1))+
    scale_x_discrete(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0)) +
    labs(fill="Normalized\nEnrichment\nScore\n(>0: Male-Up,\n<0: Female-Up)")+
    xlab("Visium Domain")+
    ylab("Disease / Trait\nGene Set") +
    theme(plot.background = element_blank(),axis.text.x = element_text(size=9,angle=90,vjust=0.5),axis.text.y=element_text(size=9),axis.title.x=element_text(size=10,hjust=0.5),axis.title.y=element_text(size=10,vjust=0.5,margin=margin(0,-0.15,0,0,"in")),legend.title = element_text(size=9,hjust=0.5),legend.text=element_text(size=8),legend.margin=margin(0,0,0,-0.1,"in"))
dev.off()
```
