---
title: "Fig1_Overview"
author: "Bernard Mulvey"
date: "2024-08-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10, fig.width = 7, include = FALSE)
knitr::opts_chunk$set(fig.width = 7, fig.height = 10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table) # Preferred data manipulation package
library(ggplot2) # Dependency for several plotting functions
library(ggtext) # more character types in ggplots
library(ggrastr) # avoid making raster images one can't even open
library(SpatialExperiment) # Visium data framework
library(SpatialFeatureExperiment) # Xenium data framework
library(spatialLIBD) # one option for plotting cluster assignments, but breaks when the in-tissue portion of a visium area is very un-square.
library(escheR) # alternative spotplotting function, at least for visium
library(viridis) # palettes
library(Polychrome) # better palettes
require(colorout) # Utility for RStudio
library(sf) # define the polygonal boundaries of xenium domains
ColorOut()

# code reformatting in Rstudio
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

# set plotting defaults for ggplot
theme_set(theme_bw() + theme(axis.text.x = element_text(size = 8), axis.title.x = element_text(size = 9), axis.text.y = element_text(size = 8), axis.title.y = element_text(size = 9), plot.title = element_blank(), strip.text = element_text(size = 10), legend.text = element_text(size = 8), legend.title = element_text(size = 8, hjust = 0.5)))
```

## load visium, xenium experiments, cluster labels
```{r}
#### VISIUM ####
hyp2 <- readRDS("spatial_HYP/processed-data/03-QC_filters/hypN10_umi210_gene126_chrm50_spotsweeped_lognorm_rotsNmirrors_072224.RDS")

bscl <- fread("spatial_HYP/processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt") # main visium clustering (bs=bayesspace)

setnames(bscl,2,"cl")
bscl[,cl:=paste0("Vis",cl)]
bscl[cl=="Vis7",cl:="VMH.1"]
bscl[cl=="Vis12",cl:="VMH.2"]
bscl[cl=="Vis4",cl:="ARC.1"]
bscl[cl=="Vis6",cl:="ARC.2"]
# WM (optic tract, OT): Clusters 3 and 10
bscl[cl=="Vis3",cl:="OT.1"]
bscl[cl=="Vis10",cl:="OT.2"]

bscl[,dom:=cl]
bscl[dom %in% c("VMH.1","VMH.2"),dom:="VMH"]
bscl[dom %in% c("ARC.1","ARC.2"),dom:="ARC"]
bscl[dom %in% c("OT.1","OT.2"),dom:="OT"]
bscl[!(dom %in% c("VMH","ARC","OT")),dom:="Other"]

bscl<-DataFrame(bscl,row.names=bscl$rn)[colnames(hyp2),]
hyp2$k15dom <- bscl$cl
hyp2$`Visium Domain` <- factor(bscl$dom,levels=c("ARC","VMH","OT","Other"))

#### XENIUM ####
hypx <- readRDS("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-sfe-in_staggered-coords_genetarg-only_NONlog-norm.RDS")

bksmooth <- fread("xenium_HYP/processed-data/07_Domains-subdomains_by_knnSmoothing_Banksytypes/03b-ARCVMHdomains_2stepsmooth_VMH-k10-0.2-VMH-k200-0.2_ARC-k50-0.1_ARC-k500-0.5.txt") ## xenium domain assignments after smoothing, by xenium cell

bksmooth[dualVMHARC4=="other",dualVMHARC4:="Other"]
bksmooth <- DataFrame(bksmooth,row.names=bksmooth$rn)[colnames(hypx),]
hypx$dom <- bksmooth$dualVMHARC4
```

### previously made a working 15-color palette, so we can use this for both Xen/Vis VMH/ARC at subdomain/cluster levels (max 8 VM+ARC combined, in xenium) and for overarching domains (2 colors + a gray for other)

that prior palette, with slots 1-2 for ARC, 3-4 for VMH, and 5-6 for WM or other, was: "#269422", "#169482", "#4a2881", "#6A0D53", "#3D4047", "#A9A7A9", "#F70022", "#551CFD", "#F59B26", "#00B3FD", "#FF0DF7", "#D193FE", "#D1CA1C", "#FB479F", "#AD260D"
The colors slotted in 1 and 3 are also picked to have the same luminance so that gene expression coloration fill in escheR grayscale doesn't look falsely darker or lighter in one spot type vs. the other 
```{r}
pal <- c("#8c63cf","#5dd959","#F59B26","#A9A7A9")
names(pal) <- c("VMH","ARC","OT","Other")

### also save the prior palette for quick loadup
fullpal <- c("#8c63cf", "#169482", ,"#5dd959" "#6A0D53", "#3D4047", "#A9A7A9", "#F70022", "#551CFD", "#F59B26", "#00B3FD", "#FF0DF7", "#D193FE", "#D1CA1C", "#FB479F", "#AD260D")
saveRDS(fullpal,"manuscript_plots/15color_palette.RDS")
rm(fullpal)
####
```

###### TIP: LEGEND ORDER ######
### i would've been up shit creek without this handy post https://stackoverflow.com/questions/76397512/ggplot-ordering-legends-with-guides-changes-continuous-legend-to-discrete
#################################

Panels A and B are experimental overviews/schematics.
Panel C is an exemplary Visium sample with pertinent domain labels.
Our exemplary sample here is V12D05-350_C1 (xen rep Br8741,X99_8741C looks better) 

## (another option: V13Y24-346_C1 (Br1225, xenium rep X99_1225A looks better)
```{r}
panc <- hyp2[,hyp2$sample_id=="V12D05-350_C1"]

p <- make_escheR(panc)
p <- p |> add_fill("Visium Domain",size=0.25,point_size = 0.6)

pdf("manuscript_plots/Fig1/1C-V12D05_350_C1_Visdomains.pdf",height=2,width=2)
p+
    scale_fill_manual(values=pal,na.value = NA)+
    ggtitle("V12D05-350_C1 (Visium)")+
    guides(color="none",fill=guide_legend(override.aes=list(size=1.5)))+
    labs(fill="Visium\nDomain")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=9,hjust=0.5),legend.text = element_text(size=7),legend.key.size = ggplot2::unit(0.075,"in"),legend.title=element_text(size=7.5,hjust=0.5),plot.margin = margin(0,0.05,0,-0.05,unit = "in"),legend.margin = margin(0,0,0,-0.14,"in"))
dev.off()

rm(panc,p)
```

### Panel D: VMH marker in Visium sample (NR5A1)

```{r}
## set visium rownames  to gene symbols
rownames(hyp2) <- rowData(hyp2)$gene_name

pand <- hyp2["NR5A1",hyp2$sample_id=="V12D05-350_C1"]
colData(pand)$`log counts` <- as.numeric(logcounts(pand)["NR5A1",])

p <- make_escheR(pand)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.5)
p <- p |> add_ground(var = "Visium Domain",stroke=0.15,point_size = 0.5)

pdf("manuscript_plots/Fig1/1D-V12D05_350_C1_NR5A1.pdf",height=2,width=2)
p+
    scale_color_manual(values=pal,na.value = NA)+
    scale_fill_continuous("log\ncounts",low= "#FFFFFF",high="#000000")+
    ggtitle("*NR5A1* (Visium)")+
    guides(color=guide_legend(override.aes=list(size=1.5,stroke=1)))+
    labs(color="Visium\nDomain")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=9,hjust=0.5),legend.text = element_text(size=7),legend.key.size = ggplot2::unit(0.075,"in"),legend.title=element_text(size=7.5,hjust=0.5),plot.margin = margin(0,0.05,0,-0.05,unit = "in"),legend.margin = margin(0,0,0,-0.14,"in"))
dev.off()

rm(p,pand)
```


# panels E is a representative ARC marker in our Visium sample : POMC
```{r}
## we need to set rownames of the visium object to gene symbols instead of ensembl identifiers for ease here
rownames(hyp2) <- rowData(hyp2)$gene_name

## for escheR, we put the assay value into the colData for it to use
pane <- hyp2["POMC",hyp2$sample_id=="V12D05-350_C1"]
colData(pane)$`log counts` <- as.numeric(logcounts(pane)["POMC",])

p <- make_escheR(pane)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.5)
p <- p |> add_ground(var = "Visium Domain",stroke=0.15,point_size = 0.5)

pdf("manuscript_plots/Fig1/1E-V12D05_350_C1_POMC.pdf",height=2,width=2)
p+
    scale_color_manual(values=pal,na.value = NA)+
    scale_fill_continuous("log\ncounts",low= "#FFFFFF",high="#000000")+
    #scale_fill_grey()+
    ggtitle("*POMC* (Visium)")+
    guides(color=guide_legend(override.aes=list(size=1.5,stroke=1)))+
    labs(color="Visium\nDomain")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=9,hjust=0.5),legend.text = element_text(size=7),legend.key.size = ggplot2::unit(0.075,"in"),legend.title=element_text(size=7.5,hjust=0.5),plot.margin = margin(0,0.05,0,-0.05,unit = "in"),legend.margin = margin(0,0,0,-0.14,"in"))
dev.off()

rm(p,pane)
```

Panel F is Visium expression of a WM marker, KLK6. (Our options for those are somewhat limited on the xenium side).
```{r}
## for escheR, we put the assay value into the colData for it to use
panf <- hyp2["KLK6",hyp2$sample_id=="V12D05-350_C1"]
colData(panf)$`log counts` <- as.numeric(logcounts(panf)["KLK6",])

p <- make_escheR(panf)
p <- p |> add_fill("log counts",size=0.25,point_size = 0.5)
p <- p |> add_ground(var = "Visium Domain",stroke=0.15,point_size = 0.5)

pdf("manuscript_plots/Fig1/1F-V12D05_350_C1_KLK6.pdf",height=2,width=2)
p+
    scale_color_manual(values=pal,na.value = NA)+
    scale_fill_continuous("log\ncounts",low= "#FFFFFF",high="#000000")+
    #scale_fill_grey()+
    ggtitle("*KLK6* (Visium)")+
    guides(color=guide_legend(override.aes=list(size=1.5,stroke=1)))+
    labs(color="Visium\nDomain")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=9,hjust=0.5),legend.text = element_text(size=7),legend.key.size = ggplot2::unit(0.075,"in"),legend.title=element_text(size=7.5,hjust=0.5),plot.margin = margin(0,0.05,0,-0.05,unit = "in"),legend.margin = margin(0,0,0,-0.14,"in"))
dev.off()

rm(p,panf)
```

Panel G is the xenium domains in the same donor (Br8741 , sample X99_8741C).

### Panel G: xenium counterpart to C (X99_8741C)
```{r}
pang <- hypx[,hypx$sample_id=="X99_8741C"]

# y_reverse is an argument to unexported escheR internals that we don't want set to TRUE, since we have oriented all the tissue the way we want it already.
p <- make_escheR(pang,y_reverse=FALSE)
p <- p |> add_fill("dom",size=0.05,point_size = 0.2)

pdf("manuscript_plots/Fig1/1G-X99_8741C_Xendomains.pdf",height=2,width=2)
p+
    scale_fill_manual(values=pal,na.value = NA)+
    ggtitle("X99_8741C (Xenium)")+
    guides(color="none",fill=guide_legend(override.aes=list(size=1.5)))+
    labs(fill="Xenium\nDomain")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=9,hjust=0.5,margin=margin(0.15,0,-0.15,0,"in")),legend.text = element_text(size=7),legend.key.size = ggplot2::unit(0.125,"in"),legend.title=element_text(size=7.5,hjust=0.5),plot.margin = margin(-0.07,-0.1,-0.07,-0.1,unit = "in"),legend.margin=margin(0,0,0,-0.15,"in"))
dev.off()

rm(pang,p)

```


### xenium panels: no WM "domain" defined, so change the domain color palette to exclude WM. also remove the "other" category, since we only have VMH/ARC outlines colored. (expression is filled with grayscale.)
```{r}
pal <- c("#8c63cf","#5dd959")
names(pal) <- c("VMH","ARC")
```

### The xenium expression panels (H, I,J) are going to be tricky -- there's so many spots we can't really use the line-color / fill heuristic. we'll have to add an outline around the domain instead.

### H I and J are the same genes in the xenium counterpart, X99_8741C
```{r}
panh <- hypx["NR5A1",hypx$sample_id=="X99_8741C"]
# 
## attach the coordinate info to colData so we can make an sf object out of this for hull extraction. see well-annotated code from DE benchmarking of domain boundaries in xenium_HYP/code/07/03c-DE benchmarks...
tmpcoldata <- cbind(colData(panh),spatialCoords(panh))
panhv <- st_as_sf(as.data.frame(tmpcoldata[tmpcoldata$dom=="VMH",]),coords = c("sdimx","sdimy"))
panhv <- st_convex_hull(st_union(panhv))

panha <- st_as_sf(as.data.frame(tmpcoldata[tmpcoldata$dom=="ARC",]),coords=c("sdimx","sdimy"))
panha <- st_concave_hull(st_union(panha),ratio = 0.1)

## create the data table we will use for overlaying the boundary polygons
dompoly <- cbind(panhv[[1]][[1]],rep("VMH",nrow(panhv[[1]][[1]])))
dompoly <- rbind(dompoly,cbind(panha[[1]][[1]],rep("ARC",nrow(panha[[1]][[1]]))))
dompoly <- as.data.table(dompoly)
setnames(dompoly,c("xpol","ypol","dompol"))
dompoly[,xpol:=as.numeric(xpol)]
dompoly[,ypol:=as.numeric(ypol)]

rm(panha,panhv)

# ## finally, append the counts for our gene of interest (NR5A1)
colData(panh)$`log counts` <- as.numeric(logcounts(panh)["NR5A1",])

p <- make_escheR(panh,y_reverse=FALSE)
p <- p |> add_fill("log counts",size=0.125,point_size = 0.125)

pdf("manuscript_plots/Fig1/1H-X99_8741C_NR5A1.pdf",height=2,width=2)
p+
    scale_fill_continuous("log\ncounts",low= "#FFFFFF",high="#000000")+
    geom_path(data=dompoly,aes(x=xpol,y=ypol,group=dompol,col=dompol),linewidth = 0.25)+
    scale_color_manual(values=pal,na.value = NA)+
    ggtitle("*NR5A1* (Xenium)")+
    guides(color="none")+
    labs(col="Xenium\nDomain")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=9,hjust=0.5,margin=margin(0.15,0,-0.15,0,"in")),legend.text = element_text(size=7),legend.key.size = ggplot2::unit(0.125,"in"),legend.title=element_text(size=7.5,hjust=0.5),plot.margin = margin(-0.07,-0.1,-0.07,-0.1,unit = "in"),legend.margin=margin(0,0,0,-0.15,"in"))
dev.off()

rm(panh,p,tmpcoldata)

```

I: POMC in X99_8741C
```{r}
pani <- hypx["POMC",hypx$sample_id=="X99_8741C"]
colData(pani)$`log counts` <- as.numeric(logcounts(pani)["POMC",])

p <- make_escheR(pani,y_reverse=FALSE)
p <- p |> add_fill("log counts",size=0.125,point_size = 0.125)

pdf("manuscript_plots/Fig1/1I-X99_8741C_POMC.pdf",height=2,width=2)
p+
    scale_fill_continuous("log\ncounts",low= "#FFFFFF",high="#000000")+
    geom_path(data=dompoly,aes(x=xpol,y=ypol,group=dompol,col=dompol),linewidth = 0.25)+
    scale_color_manual(values=pal,na.value = NA)+
    ggtitle("*POMC* (Xenium)")+
    guides(color="none")+
    labs(col="Xenium\nDomain")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=9,hjust=0.5,margin=margin(0.15,0,-0.15,0,"in")),legend.text = element_text(size=7),legend.key.size = ggplot2::unit(0.125,"in"),legend.title=element_text(size=7.5,hjust=0.5),plot.margin = margin(-0.07,-0.1,-0.07,-0.1,unit = "in"),legend.margin=margin(0,0,0,-0.15,"in"))
dev.off()

rm(pani,p)
```

1J: KLK6 in xenium
```{r}
panj <- hypx["KLK6",hypx$sample_id=="X99_8741C"]
colData(panj)$`log counts` <- as.numeric(logcounts(panj)["KLK6",])

p <- make_escheR(panj,y_reverse=FALSE)
p <- p |> add_fill("log counts",size=0.125,point_size = 0.175)

pdf("manuscript_plots/Fig1/1J-X99_8741C_KLK6.pdf",height=2,width=2)
p+
    scale_fill_continuous("log\ncounts",low= "#FFFFFF",high="#000000")+
    geom_path(data=dompoly,aes(x=xpol,y=ypol,group=dompol,col=dompol),linewidth = 0.25)+
    scale_color_manual(values=pal,na.value = NA)+
    ggtitle("*KLK6* (Xenium)")+
    guides(color="none")+
    labs(col="Xenium\nDomain")+
    theme(plot.title.position = "plot",plot.title = element_markdown(size=9,hjust=0.5,margin=margin(0.15,0,-0.15,0,"in")),legend.text = element_text(size=7),legend.key.size = ggplot2::unit(0.125,"in"),legend.title=element_text(size=7.5,hjust=0.5),plot.margin = margin(-0.07,-0.1,-0.07,-0.1,unit = "in"),legend.margin=margin(0,0,0,-0.15,"in"))
dev.off()

rm(panj,p)
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] sf_1.0-16                      colorout_1.3-0.2              
 [3] Polychrome_1.5.1               viridis_0.6.5                 
 [5] viridisLite_0.4.2              escheR_1.4.0                  
 [7] spatialLIBD_1.16.2             SpatialFeatureExperiment_1.6.1
 [9] SpatialExperiment_1.14.0       SingleCellExperiment_1.26.0   
[11] SummarizedExperiment_1.34.0    Biobase_2.64.0                
[13] GenomicRanges_1.56.1           GenomeInfoDb_1.40.1           
[15] IRanges_2.38.1                 S4Vectors_0.42.1              
[17] BiocGenerics_0.50.0            MatrixGenerics_1.16.0         
[19] matrixStats_1.3.0              ggrastr_1.0.2                 
[21] ggtext_0.1.2                   ggplot2_3.5.1                 
[23] data.table_1.15.4              rlang_1.1.4                   

loaded via a namespace (and not attached):
  [1] later_1.3.2               BiocIO_1.14.0             fields_16.2              
  [4] bitops_1.0-7              filelock_1.0.3            tibble_3.2.1             
  [7] R.oo_1.26.0               XML_3.99-0.17             lifecycle_1.0.4          
 [10] edgeR_4.2.0               doParallel_1.0.17         rprojroot_2.0.4          
 [13] lattice_0.22-6            magrittr_2.0.3            rmarkdown_2.27           
 [16] sass_0.4.9                limma_3.60.3              plotly_4.10.4            
 [19] jquerylib_0.1.4           yaml_2.3.9                httpuv_1.6.15            
 [22] spam_2.10-0               sessioninfo_1.2.2         sp_2.1-4                 
 [25] cowplot_1.1.3             DBI_1.2.3                 RColorBrewer_1.1-3       
 [28] golem_0.4.1               maps_3.4.2                abind_1.4-5              
 [31] zlibbioc_1.50.0           purrr_1.0.2               R.utils_2.12.3           
 [34] RCurl_1.98-1.14           rappdirs_0.3.3            GenomeInfoDbData_1.2.12  
 [37] ggrepel_0.9.5             irlba_2.3.5.1             terra_1.7-78             
 [40] units_0.8-5               dqrng_0.4.1               commonmark_1.9.1         
 [43] DelayedMatrixStats_1.26.0 codetools_0.2-20          DropletUtils_1.24.0      
 [46] DelayedArray_0.30.1       DT_0.33                   scuttle_1.14.0           
 [49] xml2_1.3.6                tidyselect_1.2.1          farver_2.1.2             
 [52] UCSC.utils_1.0.0          ScaledMatrix_1.12.0       shinyWidgets_0.8.6       
 [55] BiocFileCache_2.12.0      GenomicAlignments_1.40.0  jsonlite_1.8.8           
 [58] BiocNeighbors_1.22.0      e1071_1.7-14              scater_1.32.0            
 [61] iterators_1.0.14          foreach_1.5.2             tools_4.4.1              
 [64] Rcpp_1.0.13               glue_1.7.0                gridExtra_2.3            
 [67] SparseArray_1.4.8         xfun_0.45                 here_1.0.1               
 [70] EBImage_4.46.0            dplyr_1.1.4               HDF5Array_1.32.0         
 [73] withr_3.0.0               BiocManager_1.30.23       fastmap_1.2.0            
 [76] boot_1.3-30               rhdf5filters_1.16.0       fansi_1.0.6              
 [79] spData_2.3.1              rsvd_1.0.5                digest_0.6.36            
 [82] R6_2.5.1                  mime_0.12                 colorspace_2.1-0         
 [85] wk_0.9.2                  markdown_1.13             jpeg_0.1-10              
 [88] RSQLite_2.3.7             R.methodsS3_1.8.2         config_0.3.2             
 [91] utf8_1.2.4                tidyr_1.3.1               generics_0.1.3           
 [94] rtracklayer_1.64.0        class_7.3-22              httr_1.4.7               
 [97] htmlwidgets_1.6.4         S4Arrays_1.4.1            scatterplot3d_0.3-44     
[100] spdep_1.3-5               pkgconfig_2.0.3           gtable_0.3.5             
[103] blob_1.2.4                XVector_0.44.0            htmltools_0.5.8.1        
[106] dotCall64_1.1-1           fftwtools_0.9-11          scales_1.3.0             
[109] attempt_0.3.1             png_0.1-8                 knitr_1.48               
[112] rstudioapi_0.16.0         rjson_0.2.21              curl_5.2.1               
[115] proxy_0.4-27              cachem_1.1.0              rhdf5_2.48.0             
[118] stringr_1.5.1             BiocVersion_3.19.1        KernSmooth_2.23-24       
[121] parallel_4.4.1            vipor_0.4.7               AnnotationDbi_1.66.0     
[124] restfulr_0.0.15           s2_1.1.6                  pillar_1.9.0             
[127] grid_4.4.1                vctrs_0.6.5               promises_1.3.0           
[130] BiocSingular_1.20.0       dbplyr_2.5.0              beachmat_2.20.0          
[133] xtable_1.8-4              sfheaders_0.4.4           beeswarm_0.4.0           
[136] paletteer_1.6.0           evaluate_0.24.0           zeallot_0.1.0            
[139] magick_2.8.3              Rsamtools_2.20.0          cli_3.6.3                
[142] locfit_1.5-9.10           compiler_4.4.1            crayon_1.5.3             
[145] labeling_0.4.3            classInt_0.4-10           rematch2_2.1.2           
[148] ggbeeswarm_0.7.2          stringi_1.8.4             deldir_2.0-4             
[151] BiocParallel_1.38.0       munsell_0.5.1             Biostrings_2.72.1        
[154] lazyeval_0.2.2            tiff_0.1-12               Matrix_1.7-0             
[157] ExperimentHub_2.12.0      benchmarkme_1.0.8         sparseMatrixStats_1.16.0 
[160] bit64_4.0.5               Rhdf5lib_1.26.0           KEGGREST_1.44.1          
[163] statmod_1.5.0             shiny_1.8.1.1             AnnotationHub_3.12.0     
[166] gridtext_0.1.5            memoise_2.0.1             bslib_0.7.0              
[169] benchmarkmeData_1.0.4     bit_4.0.5                
> sessioninfo::session_info()
─ Session info ──────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-08-02
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────────────
 ! package                  * version   date (UTC) lib source
   abind                      1.4-5     2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi              1.66.0    2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub              3.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   attempt                    0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
   beachmat                   2.20.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm                   0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme                1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData            1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
   Biobase                  * 2.64.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics             * 0.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                     1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager                1.30.23   2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors              1.22.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel               1.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocSingular               1.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion                3.19.1    2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings                 2.72.1    2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                        4.0.5     2022-11-15 [1] CRAN (R 4.4.0)
   bit64                      4.0.5     2020-08-30 [1] CRAN (R 4.4.0)
   bitops                     1.0-7     2021-04-24 [1] CRAN (R 4.4.0)
   blob                       1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   boot                       1.3-30    2024-02-26 [1] CRAN (R 4.4.1)
   bslib                      0.7.0     2024-03-29 [1] CRAN (R 4.4.0)
   cachem                     1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
   class                      7.3-22    2023-05-03 [1] CRAN (R 4.4.1)
   classInt                   0.4-10    2023-09-05 [1] CRAN (R 4.4.0)
 P cli                        3.6.3     2024-06-21 [2] CRAN (R 4.4.0)
   codetools                  0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorout                 * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace                 2.1-0     2023-01-23 [1] CRAN (R 4.4.0)
   commonmark                 1.9.1     2024-01-30 [1] CRAN (R 4.4.0)
   config                     0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                    1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
   crayon                     1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   curl                       5.2.1     2024-03-01 [1] CRAN (R 4.4.0)
   data.table               * 1.15.4    2024-03-30 [2] CRAN (R 4.4.0)
   DBI                        1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                     2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray               0.30.1    2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats         1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   deldir                     2.0-4     2024-02-28 [1] CRAN (R 4.4.0)
   digest                     0.6.36    2024-06-23 [1] CRAN (R 4.4.0)
   doParallel                 1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64                  1.1-1     2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                      1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   dqrng                      0.4.1     2024-05-28 [1] CRAN (R 4.4.0)
   DropletUtils               1.24.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   DT                         0.33      2024-04-04 [1] CRAN (R 4.4.0)
   e1071                      1.7-14    2023-12-06 [1] CRAN (R 4.4.0)
   EBImage                    4.46.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   edgeR                      4.2.0     2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   escheR                   * 1.4.0     2024-05-16 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate                   0.24.0    2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fansi                      1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   farver                     2.1.2     2024-05-13 [1] CRAN (R 4.4.0)
   fastmap                    1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fftwtools                  0.9-11    2021-03-01 [1] CRAN (R 4.4.0)
   fields                     16.2      2024-06-27 [1] CRAN (R 4.4.0)
   filelock                   1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   foreach                    1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
   generics                   0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb             * 1.40.1    2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData           1.2.12    2024-05-01 [1] Bioconductor
   GenomicAlignments          1.40.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges            * 1.56.1    2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm                 0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2                  * 3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrastr                  * 1.0.2     2023-06-01 [1] CRAN (R 4.4.0)
   ggrepel                    0.9.5     2024-01-10 [1] CRAN (R 4.4.0)
   ggtext                   * 0.1.2     2022-09-16 [1] CRAN (R 4.4.0)
   glue                       1.7.0     2024-01-09 [1] CRAN (R 4.4.0)
   golem                      0.4.1     2023-06-05 [1] CRAN (R 4.4.0)
   gridExtra                  2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gridtext                   0.1.5     2022-09-16 [1] CRAN (R 4.4.0)
   gtable                     0.3.5     2024-04-22 [1] CRAN (R 4.4.0)
   HDF5Array                  1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   here                       1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools                  0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets                1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                     1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
   httr                       1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   IRanges                  * 2.38.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                      2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   iterators                  1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
   jpeg                       0.1-10    2022-11-29 [1] CRAN (R 4.4.0)
   jquerylib                  0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite                   1.8.8     2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST                   1.44.1    2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   KernSmooth                 2.23-24   2024-05-17 [1] CRAN (R 4.4.1)
   knitr                      1.48      2024-07-07 [1] CRAN (R 4.4.1)
   labeling                   0.4.3     2023-08-29 [1] CRAN (R 4.4.0)
   later                      1.3.2     2023-12-06 [1] CRAN (R 4.4.0)
   lattice                    0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval                   0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle                  1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                      3.60.3    2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                     1.5-9.10  2024-06-24 [1] CRAN (R 4.4.0)
   magick                     2.8.3     2024-02-18 [1] CRAN (R 4.4.0)
   magrittr                   2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   maps                       3.4.2     2023-12-15 [1] CRAN (R 4.4.0)
   markdown                   1.13      2024-06-04 [1] CRAN (R 4.4.0)
   Matrix                     1.7-0     2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics           * 1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats              * 1.3.0     2024-04-11 [1] CRAN (R 4.4.0)
   memoise                    2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   mime                       0.12      2021-09-28 [1] CRAN (R 4.4.0)
   munsell                    0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   paletteer                  1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
   pillar                     1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig                  2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   plotly                     4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
   png                        0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   Polychrome               * 1.5.1     2022-05-03 [1] CRAN (R 4.4.0)
   promises                   1.3.0     2024-04-05 [1] CRAN (R 4.4.0)
   proxy                      0.4-27    2022-06-09 [1] CRAN (R 4.4.0)
   purrr                      1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R.methodsS3                1.8.2     2022-06-13 [1] CRAN (R 4.4.0)
   R.oo                       1.26.0    2024-01-24 [1] CRAN (R 4.4.0)
   R.utils                    2.12.3    2023-11-18 [1] CRAN (R 4.4.0)
   R6                         2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs                   0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer               1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                       1.0.13    2024-07-17 [1] CRAN (R 4.4.0)
   RCurl                      1.98-1.14 2024-01-09 [1] CRAN (R 4.4.0)
   rematch2                   2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
   restfulr                   0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
   rhdf5                      2.48.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rhdf5filters               1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Rhdf5lib                   1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rjson                      0.2.21    2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                    * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown                  2.27      2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot                  2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools                  2.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                    2.3.7     2024-05-27 [1] CRAN (R 4.4.0)
   rstudioapi                 0.16.0    2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                       1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer                1.64.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   s2                         1.1.6     2023-12-19 [1] CRAN (R 4.4.0)
   S4Arrays                   1.4.1     2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors                * 0.42.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                       0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix               1.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                     1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater                     1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scatterplot3d              0.3-44    2023-05-05 [1] CRAN (R 4.4.0)
   scuttle                    1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo                1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   sf                       * 1.0-16    2024-03-24 [1] CRAN (R 4.4.0)
   sfheaders                  0.4.4     2024-01-17 [1] CRAN (R 4.4.0)
   shiny                      1.8.1.1   2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets               0.8.6     2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment     * 1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sp                         2.1-4     2024-04-30 [1] CRAN (R 4.4.0)
   spam                       2.10-0    2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray                1.4.8     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats          1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment        * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialFeatureExperiment * 1.6.1     2024-05-15 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD              * 1.16.2    2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   spData                     2.3.1     2024-05-31 [1] CRAN (R 4.4.0)
   spdep                      1.3-5     2024-06-10 [1] CRAN (R 4.4.0)
   statmod                    1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   stringi                    1.8.4     2024-05-06 [1] CRAN (R 4.4.0)
   stringr                    1.5.1     2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment     * 1.34.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   terra                      1.7-78    2024-05-22 [1] CRAN (R 4.4.0)
   tibble                     3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                      1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect                 1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   tiff                       0.1-12    2023-11-28 [1] CRAN (R 4.4.0)
   UCSC.utils                 1.0.0     2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   units                      0.8-5     2023-11-28 [1] CRAN (R 4.4.0)
   utf8                       1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                      0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                      0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                  * 0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite              * 0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   withr                      3.0.0     2024-01-16 [1] CRAN (R 4.4.0)
   wk                         0.9.2     2024-07-09 [1] CRAN (R 4.4.0)
   xfun                       0.45      2024-06-16 [1] CRAN (R 4.4.0)
   XML                        3.99-0.17 2024-06-25 [1] CRAN (R 4.3.3)
   xml2                       1.3.6     2023-12-04 [1] CRAN (R 4.4.0)
   xtable                     1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
   XVector                    0.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                       2.3.9     2024-07-05 [1] CRAN (R 4.4.0)
   zeallot                    0.1.0     2018-01-28 [1] CRAN (R 4.4.0)
   zlibbioc                   1.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

─────────────────────────────────────────────────────────────────────────────────