---
title: "Fig4_Attributing_SexDE_to_celltypes"
output: html_document
date: "2024-08-29"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table) # Preferred data manipulation package
library(ggplot2) # Dependency for several plotting functions
library(ggtext) # more character types in ggplots
library(ggrepel) # text labels in volcano plots
library(ggrastr) # avoid making raster images one can't even open
library(SpatialExperiment) # Visium data framework
library(SpatialFeatureExperiment) # Xenium data framework
library(spatialLIBD) # one option for plotting cluster assignments, but breaks when the in-tissue portion of a visium area is very un-square.
library(escheR) # alternative spotplotting function, at least for visium
library(viridis) # palettes
library(Polychrome) # better palettes
library(ggstance) # for y axis dodge
library(sf) # define the polygonal boundaries of xenium domains
library(ggrepel) # for most of the ARC cell type palette https://github.com/moldach/vapoRwave 

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw() + theme(axis.text.x = element_text(size = 8), axis.title.x = element_text(size = 9), axis.text.y = element_text(size = 8), axis.title.y = element_text(size = 9), plot.title = element_blank(), strip.text = element_text(size = 10), legend.text = element_text(size = 8), legend.title = element_text(size = 8, hjust = 0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```



#### VISIUM setup ####
```{r}
hyp2 <- readRDS("spatial_HYP/processed-data/03-QC_filters/hypN10_umi210_gene126_chrm50_spotsweeped_lognorm_rotsNmirrors_072224.RDS")

bscl <- fread("spatial_HYP/processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt") # main visium clustering (bs=bayesspace)

setnames(bscl,2,"cl")
bscl[,cl:=paste0("Vis",cl)]
bscl[cl=="Vis7",cl:="VMH.1"]
bscl[cl=="Vis12",cl:="VMH.2"]
bscl[cl=="Vis4",cl:="ARC.1"]
bscl[cl=="Vis6",cl:="ARC.2"]

bscl[,dom:=cl]
bscl[dom %in% c("VMH.1","VMH.2"),dom:="VMH"]
bscl[dom %in% c("ARC.1","ARC.2"),dom:="ARC"]
bscl[!(dom %in% c("VMH","ARC")),dom:="Other"]

bscl<-DataFrame(bscl,row.names=bscl$rn)[colnames(hyp2),]
hyp2$k15dom <- bscl$cl
hyp2$`Visium Domain` <- factor(bscl$dom,levels=c("ARC","VMH","Other"))
```

### XENIUM setup ###
```{r}
hypx <- readRDS("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-sfe-in_staggered-coords_genetarg-only_NONlog-norm.RDS")

bksmooth <- fread("xenium_HYP/processed-data/07_Domains-subdomains_by_knnSmoothing_Banksytypes/03b-ARCVMHdomains_2stepsmooth_VMH-k10-0.2-VMH-k200-0.2_ARC-k50-0.1_ARC-k500-0.5.txt") ## xenium domain assignments after smoothing, by xenium cell

## but drop discarded cell clusters, too
bkcl <- fread("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-BanksyClusts_M0lam0_leiden_multisamp.txt")
setnames(bkcl,2,"cl")

bkanno <- fread("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/02-M0l0kg6_topClusMarkers_celltypes_annotated.txt")

## drop sample-specific clusters, append cluster annotations 
bkanno <- unique(bkanno[,.(clus,bjm_annot)])
bkanno <- bkanno[bjm_annot!="DISCARD"&bjm_annot!="VMH_4_DISCARD"]

## convert ARC_1..etc to more descriptive cluster IDs defined based on ARC/VMH specific cluster marker analyses
arcvmhanno <- fread("xenium_HYP/processed-data/07_Domains-subdomains_by_knnSmoothing_Banksytypes/04e_ARC_and_VMH_cluster_identities_based_on_cellgroup_1vall.txt")
## simplify these by removing VMH/ARC and underscores; asterisk-ize so the gene names  can be italiczed in ggplot elements; remove vmh and arc from the names but add a temporary column here so we know they're VMH and ARC for mapping onto the de result sets
arcvmhanno[,subclus_domain:=gsub(subclus_annot,pattern="^(VMH)_.*$",replacement="\\1")]
arcvmhanno[,subclus_annot:=gsub(subclus_annot,pattern="VMH_",replacement="")]

arcvmhanno[subclus_domain!="VMH",subclus_domain:=gsub(subclus_annot,pattern="^(ARC)_.*$",replacement="\\1")]
arcvmhanno[,subclus_annot:=gsub(subclus_annot,pattern="ARC_",replacement="")]

arcvmhanno[,subclus_annot:=gsub(subclus_annot,pattern="_",replacement=" ")]
arcvmhanno[subclus_annot=="SLC17A6-CRHR2",subclus_annot:="*SLC17A6*-*CRHR2*"]
arcvmhanno[subclus_annot=="FEZF1-LAMP5-NRGN",subclus_annot:="*FEZF1*-*LAMP5*-*NRGN*"]
arcvmhanno[subclus_annot=="HCRT-SLC6A3-TAC1",subclus_annot:="*HCRT*-*SLC6A3*-*TAC1*"]
arcvmhanno[subclus_annot=="AGRP",subclus_annot:="*AGRP*"]
arcvmhanno[subclus_annot=="TAC3-AR-ESR1",subclus_annot:="*TAC3*-*AR*-*ESR1*"]
arcvmhanno[subclus_annot=="POMC-Oligo-Mixed",subclus_annot:="*POMC*-Oligo"]
arcvmhanno[subclus_annot=="Excit-SLC17A7-Glial-Mixed",subclus_annot:="*SLC17A7*-Glia"]
arcvmhanno[subclus_annot=="GAL-TRH-GHRH",subclus_annot:="*GAL*-*TRH*-*GHRH*"]

## append these to the other annots and replace bjm_annot with the ARC/VMH cell type labels where applicable
bkanno <- merge.data.table(bkanno,arcvmhanno,by="bjm_annot",all.x=T)
bkanno[!is.na(subclus_annot),bjm_annot:=subclus_annot]

## this drops rows that were assigned one of the discarded types (ie no longer in the cell type annotations)
bkcl <- merge.data.table(bkcl,bkanno,by.x="cl",by.y="clus")
## and now we can drop excluded clusters from xenium spe
hypx <- hypx[,colnames(hypx) %in% bkcl$rn]

## append cluster annots to xenium spe
bkcl <- DataFrame(bkcl,row.names=bkcl$rn)[colnames(hypx),]
bkcl$rn <- NULL
hypx$banksyclus <- bkcl$bjm_annot

## assign the domain labels to the xenium cells
bksmooth[dualVMHARC4=="other",dualVMHARC4:="Other"]
bksmooth <- DataFrame(bksmooth,row.names=bksmooth$rn)[colnames(hypx),]
hypx$dom <- bksmooth$dualVMHARC4
```

### sex DE results
```{r}
visde <- fread("spatial_HYP/processed-data/09-Sex DE/01c-voomLmFit_nnsvg10-HmnylmbNA-BS-15-VMHARCclpsd.txt")
### exclude sex chromosomal, mito genes (need to append this info)
genelut <- fread("spatial_HYP/raw-data/USEFOR10X_ensg_lut_hg38_fromEns98_111123.txt")
genelut <- unique(genelut[,.(ensembl_gene_id,chromosome_name)])
setnames(genelut,c("ens","chr"))
visde <- merge.data.table(visde,genelut,by.x="gene_id",by.y="ens")
visde <- visde[!(chr %in% c("MT","X","Y"))]

### load by cell type and full-domains xenium results from 4 ARC-cell-type-smoothing domain
xenarcde <- fread("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/02a-4ARCtypesmoothed_withinARC_celltypeSexDE.txt")
xenvmhde <- fread("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/02b-4ARCtypesmoothed_withinVMH_celltypeSexDE.txt")

xendomde <- fread("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/04-4typeARC-and-dualassignVMH_domainwise_sexDE.txt")

### for correlation plot, merge xenium domain and visium arc/vmh stats
vistmp <- copy(visde[assay %in% c("VMH","ARC")])
setnames(vistmp,"logFC","logFC_Visium")

setnames(xendomde,"logFC_MvF","logFC_Xenium")

xendomde <- merge.data.table(xendomde[,.(domain,gene_name,logFC_Xenium)],vistmp[,.(assay,gene_name,logFC_Visium)],by.x=c("domain","gene_name"),by.y=c("assay","gene_name"))

## clean up
rm(genelut,vistmp)

# also load pseudobulk ELists used in the DE analysis for plotting DE from pseudobulk results
arcelist <- readRDS("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/02c-pseudobulkARC_voomLmFitELists.RDS")
vmhelist <- readRDS("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/02d-pseudobulkVMH_voomLmFitELists.RDS")

## subset the pseudobulk tables to the ARC clusters for the arc domain and VMH for vmh
arcelist <- arcelist[bkanno[subclus_domain=="ARC",clus]]
vmhelist <- vmhelist[bkanno[subclus_domain=="VMH",clus]]

## change the pseudobulk entry names to be the more granularly annotated cell types
names(vmhelist) <-  bkanno[match(names(vmhelist),bkanno$clus),bjm_annot]
names(arcelist) <-  bkanno[match(names(arcelist),bkanno$clus),bjm_annot]
```

cell types palettes
```{r}
vmhpal <- c("#8c63cf", "lightgray", "#6A0D53", "#AD260D")
names(vmhpal) <- c("*SLC17A6*-*CRHR2*","Microglia","*FEZF1*-*LAMP5*-*NRGN*","*HCRT*-*SLC6A3*-*TAC1*")

arcpal <- c("#5dd959","#296656","#09979B","#A997AB","#D0B")
names(arcpal) <-  c("*AGRP*","*POMC*-Oligo","*GAL*-*TRH*-*GHRH*","*SLC17A7*-Glia","*TAC3*-*AR*-*ESR1*")

```

## palette set up (shared)
```{r}
pal <- c("#8c63cf","#5dd959","#A9A7A9")
names(pal) <- c("VMH","ARC","Other")
```

## panel A: xenium volcano, VMH types in VMH domain
```{r}
xenvmhdeplt <- xenvmhde[cluster_annotated %in% grep(cluster_annotated,pattern="VMH",value=T)]

xenvmhdeplt <- merge.data.table(xenvmhdeplt,arcvmhanno,by.x="cluster_annotated",by.y="bjm_annot")

xenvmhdeplt[gene_name %in% c("PTGDS","SHANK3","TAC1"),label:=gene_name]

p <- ggplot(xenvmhdeplt, aes(y = -log10(P.Value), x = logFC_MvF,col=subclus_annot)) +
    geom_point(size = 0.15) +
    scale_color_manual(values=vmhpal) +
    geom_hline(yintercept = -log10(0.05),
        linetype = "dashed",
        color = "black",
        linewidth=0.25) +
    geom_hline(yintercept = -log10(max(xenvmhdeplt[FDR<0.05,P.Value])),
        linetype = "dashed",
        color = "red",
        linewidth=0.25) +
    geom_text_repel(aes(y = -log10(P.Value),
            x = logFC_MvF,
            label = ifelse(is.na(label), "", label),size=),
        min.segment.length = 0,
        size = 0.9,
        show.legend=FALSE,
        nudge_x=1.25*sign(xenvmhdeplt$logFC_MvF),
        seed=10403,
        max.iter=100000,
        max.overlaps=10000,
        max.time=3,
        segment.size=0.125)+
    xlab("logFC (M:F)")+
    ylab("-log10 (P-value)")+
    labs(color="VMH Cell Type")+
    theme(axis.text=element_text(size=7),axis.title=element_text(size=7.5),legend.key.height = ggplot2::unit(0.075,"in"),legend.key.width=ggplot2::unit(0.04,"in"),legend.position.inside = c(0.21,0.9),legend.position = "inside",legend.text=element_markdown(size=5,margin=margin(0,0,0,0,"in")),legend.box=element_blank(),legend.background=element_blank(),legend.title = element_text(size=6,vjust=0.5,margin=margin(0,0,0.01,0,"in")))

# rasterize
p <- rasterize(p,layers="Points",dpi=450)

# save
pdf("manuscript_plots/Fig4/4A-XeniumVMHcelltypes_sexDE.pdf",height=2.5,width=2.3)
p
dev.off()

# clean up
rm(xenvmhdeplt,p)

```

Panel B: ARC sex DE volcano by cell type.
```{r}
xenarcdeplt <- xenarcde[cluster_annotated %in% grep(cluster_annotated,pattern="ARC",value=T)]

xenarcdeplt[gene_name %in% c("SHANK3","DCC","RXRG"),label:=gene_name]

## replace names with informative cell type labels for plot, and thus in palette
xenarcdeplt <- merge.data.table(xenarcdeplt,arcvmhanno,by.x="cluster_annotated",by.y="bjm_annot")

p <- ggplot(xenarcdeplt, aes(y = -log10(P.Value), x = logFC_MvF,col=subclus_annot)) +
    geom_point(size = 0.075) +
    scale_color_manual(values=arcpal) +
    geom_hline(yintercept = -log10(0.05),
        linetype = "dashed",
        color = "black",
        linewidth=0.25) +
    geom_hline(yintercept = -log10(max(xenarcdeplt[FDR<0.05,P.Value])),
        linetype = "dashed",
        color = "red",
        linewidth=0.25) +
    geom_text_repel(aes(y = -log10(P.Value),
            x = logFC_MvF,
            label = ifelse(is.na(label), "", label)),
        min.segment.length = 0,
        size = 0.9,
        show.legend=FALSE,
        nudge_x=1.25*sign(xenarcdeplt$logFC_MvF),
        seed=10403,
        max.iter=100000,
        max.overlaps=10000,
        max.time=3,
        segment.size=0.125)+
    xlab("logFC (M:F)")+
    ylab("-log10 (P-value)")+
    labs(color="ARC Cell Type")+
    theme(axis.text=element_text(size=7),axis.title=element_text(size=7.5),legend.key.height = ggplot2::unit(0.075,"in"),legend.key.width=ggplot2::unit(0.04,"in"),legend.position.inside = c(0.1675,0.88),legend.position = "inside",legend.text=element_markdown(size=5,margin=margin(0,0,0,0,"in")),legend.box=element_blank(),legend.background=element_blank(),legend.title = element_text(size=6,vjust=0.5,margin=margin(0,0,0.01,0,"in")))

# rasterize
p <- rasterize(p,layers="Points",dpi=450)

# save
pdf("manuscript_plots/Fig4/4B-XeniumARCcelltypes_sexDE.pdf",height=2.5,width=2.3)
p
dev.off()
```


Panel C: Show a DEG per VMH cluster, shown for all VMH xenium clusters in a violin plot datawide (at the level of pseudobulked samples). VMH_1: MC4R, VMH_2 (and 3): TAC1, VMH_3: TMEM233
```{r}
vmhe <- rbindlist(vmhelist,idcol = "celltype",fill=T)
# removing leading "elist." from colnames
setnames(vmhe,gsub(names(vmhe),pattern="^elist\\.",replacement=""))

# melt for genes of interest
vmhe.m <- melt(vmhe[rn %in% c("MC4R","TAC1","TMEM233")],id.vars=c("rn","celltype"))

# append sex info
vmhe.m <- merge.data.table(vmhe.m,unique(as.data.table(colData(hypx))[,.(sample_id,Sex)]),by.x="variable",by.y="sample_id")
vmhe.m[,Sex:=gsub(Sex,pattern="^(.).*$",replacement="\\1")]

## tidy up for plotting -- also do this for the palette or they won't match up
vmhe.m[,celltype:=gsub(celltype,pattern="_",replacement=" ")]
names(vmhpal) <- gsub(names(vmhpal),pattern="_",replacement=" ")
setnames(vmhe.m,"celltype","Xenium Cluster")

# make the gene names markdown'ed so they appear in italics
vmhe.m[,rn:=paste0("*",rn,"*")]

pdf("manuscript_plots/Fig4/4C-VMH_DEGs_boxplot.pdf",height=3,width=3.5)
ggplot(vmhe.m,aes(x=`Xenium Cluster`,y=value,color=`Xenium Cluster`,fill = Sex))+
    geom_boxplot(outliers = FALSE,linewidth = 0.25)+
    geom_point(size=0.35,position=position_jitterdodge(jitter.width = 0.15))+
    scale_fill_manual(values=c("#FFFFFF","#000000"))+
    scale_color_manual(values=vmhpal)+
    guides(color="none")+
    ylab("Pseudobulk log CPM")+
    facet_wrap(~rn,scales = "free_y")+
    theme(strip.text = element_markdown(size = 8,margin=margin(0.1,0,0,0,unit="in")),strip.background = element_blank(),axis.text.x=element_markdown(angle=90,size=7,hjust=1,vjust=0.5),axis.title.x=element_text(size=8),axis.title.y=element_text(size=8),axis.text.y=element_text(size=7),plot.margin = margin(-0.065,-0.05,0,0.02,unit="in"),legend.background = element_blank(),legend.box=element_blank(),legend.margin=margin(0,0.1,0,-0.075,unit="in"))
dev.off()


# clean up
rm(vmhe.m,vmhe)
```

## for panel D, we want to do the same for three ARC genes of interest. For ARC_4 we have the pick of the litter--we'll go with CDH13, which has known associations to weight. For ARC_3, RXRG is pretty high up there signif wise. And ARC_1 has signif CYP26A1 DE.  
```{r}
arce <- rbindlist(arcelist,idcol = "celltype",fill=T)
# removing leading "elist." from colnames
setnames(arce,gsub(names(arce),pattern="^elist\\.",replacement=""))

# melt for genes of interest
arce.m <- melt(arce[rn %in% c("CDH13","RXRG","CYP26A1")],id.vars=c("rn","celltype"))

# append sex info
arce.m <- merge.data.table(arce.m,unique(as.data.table(colData(hypx))[,.(sample_id,Sex)]),by.x="variable",by.y="sample_id")
arce.m[,Sex:=gsub(Sex,pattern="^(.).*$",replacement="\\1")]

## tidy up for plotting -- also do this for the palette or they won't match up
arce.m[,celltype:=gsub(celltype,pattern="_",replacement=" ")]
names(arcpal) <- gsub(names(arcpal),pattern="_",replacement=" ")
setnames(arce.m,"celltype","Xenium Cluster")

# make the gene names markdown'ed so they appear in italics
arce.m[,rn:=paste0("*",rn,"*")]

pdf("manuscript_plots/Fig4/4D-ARC_DEGs_boxplot.pdf",height=3,width=4)
ggplot(arce.m,aes(x=`Xenium Cluster`,y=value,color=`Xenium Cluster`,fill = Sex))+
    geom_boxplot(outliers = FALSE,linewidth = 0.25)+
    geom_point(size=0.35,position=position_jitterdodge(jitter.width = 0.15))+
    scale_fill_manual(values=c("#FFFFFF","#000000"))+
    scale_color_manual(values=arcpal)+
    guides(color="none")+
    ylab("Pseudobulk log CPM")+
    facet_wrap(~rn,scales = "free_y")+
    theme(strip.text = element_markdown(size = 8,margin=margin(0.1,0,0,0,unit="in")),strip.background = element_blank(),axis.text.x=element_markdown(angle=90,size=7,hjust=1,vjust=0.5),axis.title.x=element_text(size=8),axis.title.y=element_text(size=8),axis.text.y=element_text(size=7),plot.margin = margin(-0.065,-0.05,0,0.02,unit="in"),legend.background = element_blank(),legend.box=element_blank(),legend.margin=margin(0,0.1,0,-0.075,unit="in"))
dev.off()


# clean up
rm(arce.m,arce)
```

repro inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.38.0            parallelly_1.38.0             
 [3] ggrepel_0.9.5                  sf_1.0-16                     
 [5] colorout_1.3-0.2               ggstance_0.3.7                
 [7] Polychrome_1.5.1               viridis_0.6.5                 
 [9] viridisLite_0.4.2              escheR_1.4.0                  
[11] spatialLIBD_1.16.2             SpatialFeatureExperiment_1.6.1
[13] SpatialExperiment_1.14.0       SingleCellExperiment_1.26.0   
[15] SummarizedExperiment_1.34.0    Biobase_2.64.0                
[17] GenomicRanges_1.56.1           GenomeInfoDb_1.40.1           
[19] IRanges_2.38.1                 S4Vectors_0.42.1              
[21] BiocGenerics_0.50.0            MatrixGenerics_1.16.0         
[23] matrixStats_1.3.0              ggrastr_1.0.2                 
[25] ggtext_0.1.2                   ggplot2_3.5.1                 
[27] data.table_1.15.4              rlang_1.1.4                   

loaded via a namespace (and not attached):
  [1] bitops_1.0-7              EBImage_4.46.0            httr_1.4.7               
  [4] RColorBrewer_1.1-3        doParallel_1.0.17         tools_4.4.1              
  [7] utf8_1.2.4                R6_2.5.1                  DT_0.33                  
 [10] HDF5Array_1.32.0          lazyeval_0.2.2            rhdf5filters_1.16.0      
 [13] withr_3.0.0               sp_2.1-4                  gridExtra_2.3            
 [16] cli_3.6.3                 Cairo_1.6-2               labeling_0.4.3           
 [19] sass_0.4.9                proxy_0.4-27              Rsamtools_2.20.0         
 [22] commonmark_1.9.1          R.utils_2.12.3            scater_1.32.0            
 [25] sessioninfo_1.2.2         attempt_0.3.1             maps_3.4.2               
 [28] limma_3.60.3              rstudioapi_0.16.0         RSQLite_2.3.7            
 [31] generics_0.1.3            BiocIO_1.14.0             spdep_1.3-5              
 [34] dplyr_1.1.4               Matrix_1.7-0              ggbeeswarm_0.7.2         
 [37] fansi_1.0.6               abind_1.4-5               R.methodsS3_1.8.2        
 [40] terra_1.7-78              lifecycle_1.0.4           scatterplot3d_0.3-44     
 [43] yaml_2.3.9                edgeR_4.2.0               rhdf5_2.48.0             
 [46] SparseArray_1.4.8         BiocFileCache_2.12.0      paletteer_1.6.0          
 [49] grid_4.4.1                blob_1.2.4                promises_1.3.0           
 [52] dqrng_0.4.1               ExperimentHub_2.12.0      crayon_1.5.3             
 [55] lattice_0.22-6            beachmat_2.20.0           cowplot_1.1.3            
 [58] KEGGREST_1.44.1           magick_2.8.3              zeallot_0.1.0            
 [61] pillar_1.9.0              knitr_1.48                rjson_0.2.21             
 [64] boot_1.3-30               codetools_0.2-20          wk_0.9.2                 
 [67] glue_1.7.0                vctrs_0.6.5               png_0.1-8                
 [70] spam_2.10-0               gtable_0.3.5              rematch2_2.1.2           
 [73] cachem_1.1.0              xfun_0.45                 S4Arrays_1.4.1           
 [76] mime_0.12                 DropletUtils_1.24.0       sfheaders_0.4.4          
 [79] iterators_1.0.14          units_0.8-5               fields_16.2              
 [82] statmod_1.5.0             bit64_4.0.5               filelock_1.0.3           
 [85] rprojroot_2.0.4           bslib_0.7.0               irlba_2.3.5.1            
 [88] vipor_0.4.7               KernSmooth_2.23-24        colorspace_2.1-0         
 [91] spData_2.3.1              DBI_1.2.3                 tidyselect_1.2.1         
 [94] bit_4.0.5                 compiler_4.4.1            curl_5.2.1               
 [97] BiocNeighbors_1.22.0      xml2_1.3.6                DelayedArray_0.30.1      
[100] plotly_4.10.4             rtracklayer_1.64.0        scales_1.3.0             
[103] classInt_0.4-10           rappdirs_0.3.3            tiff_0.1-12              
[106] stringr_1.5.1             digest_0.6.36             fftwtools_0.9-11         
[109] rmarkdown_2.27            benchmarkmeData_1.0.4     XVector_0.44.0           
[112] htmltools_0.5.8.1         pkgconfig_2.0.3           jpeg_0.1-10              
[115] sparseMatrixStats_1.16.0  dbplyr_2.5.0              fastmap_1.2.0            
[118] htmlwidgets_1.6.4         UCSC.utils_1.0.0          shiny_1.8.1.1            
[121] DelayedMatrixStats_1.26.0 farver_2.1.2              jquerylib_0.1.4          
[124] jsonlite_1.8.8            config_0.3.2              R.oo_1.26.0              
[127] BiocSingular_1.20.0       RCurl_1.98-1.14           magrittr_2.0.3           
[130] scuttle_1.14.0            GenomeInfoDbData_1.2.12   s2_1.1.6                 
[133] dotCall64_1.1-1           Rhdf5lib_1.26.0           munsell_0.5.1            
[136] Rcpp_1.0.13               stringi_1.8.4             zlibbioc_1.50.0          
[139] AnnotationHub_3.12.0      parallel_4.4.1            deldir_2.0-4             
[142] Biostrings_2.72.1         gridtext_0.1.5            locfit_1.5-9.10          
[145] markdown_1.13             ScaledMatrix_1.12.0       BiocVersion_3.19.1       
[148] XML_3.99-0.17             evaluate_0.24.0           golem_0.4.1              
[151] BiocManager_1.30.23       foreach_1.5.2             httpuv_1.6.15            
[154] tidyr_1.3.1               purrr_1.0.2               benchmarkme_1.0.8        
[157] rsvd_1.0.5                xtable_1.8-4              restfulr_0.0.15          
[160] e1071_1.7-14              later_1.3.2               class_7.3-22             
[163] tibble_3.2.1              memoise_2.0.1             beeswarm_0.4.0           
[166] AnnotationDbi_1.66.0      GenomicAlignments_1.40.0  shinyWidgets_0.8.6       
[169] here_1.0.1               
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-08-17
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────────────
 ! package                  * version   date (UTC) lib source
   abind                      1.4-5     2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi              1.66.0    2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub              3.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   attempt                    0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
   beachmat                   2.20.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm                   0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme                1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData            1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
   Biobase                  * 2.64.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics             * 0.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                     1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager                1.30.23   2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors              1.22.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel             * 1.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocSingular               1.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion                3.19.1    2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings                 2.72.1    2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                        4.0.5     2022-11-15 [1] CRAN (R 4.4.0)
   bit64                      4.0.5     2020-08-30 [1] CRAN (R 4.4.0)
   bitops                     1.0-7     2021-04-24 [1] CRAN (R 4.4.0)
   blob                       1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   boot                       1.3-30    2024-02-26 [1] CRAN (R 4.4.1)
   bslib                      0.7.0     2024-03-29 [1] CRAN (R 4.4.0)
   cachem                     1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
   Cairo                      1.6-2     2023-11-28 [1] CRAN (R 4.4.0)
   class                      7.3-22    2023-05-03 [1] CRAN (R 4.4.1)
   classInt                   0.4-10    2023-09-05 [1] CRAN (R 4.4.0)
 P cli                        3.6.3     2024-06-21 [2] CRAN (R 4.4.0)
   codetools                  0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorout                 * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace                 2.1-0     2023-01-23 [1] CRAN (R 4.4.0)
   commonmark                 1.9.1     2024-01-30 [1] CRAN (R 4.4.0)
   config                     0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                    1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
   crayon                     1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   curl                       5.2.1     2024-03-01 [1] CRAN (R 4.4.0)
   data.table               * 1.15.4    2024-03-30 [2] CRAN (R 4.4.0)
   DBI                        1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                     2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray               0.30.1    2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats         1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   deldir                     2.0-4     2024-02-28 [1] CRAN (R 4.4.0)
   digest                     0.6.36    2024-06-23 [1] CRAN (R 4.4.0)
   doParallel                 1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64                  1.1-1     2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                      1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   dqrng                      0.4.1     2024-05-28 [1] CRAN (R 4.4.0)
   DropletUtils               1.24.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   DT                         0.33      2024-04-04 [1] CRAN (R 4.4.0)
   e1071                      1.7-14    2023-12-06 [1] CRAN (R 4.4.0)
   EBImage                    4.46.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   edgeR                      4.2.0     2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   escheR                   * 1.4.0     2024-05-16 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate                   0.24.0    2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fansi                      1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   farver                     2.1.2     2024-05-13 [1] CRAN (R 4.4.0)
   fastmap                    1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fftwtools                  0.9-11    2021-03-01 [1] CRAN (R 4.4.0)
   fields                     16.2      2024-06-27 [1] CRAN (R 4.4.0)
   filelock                   1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   foreach                    1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
   generics                   0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb             * 1.40.1    2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData           1.2.12    2024-05-01 [1] Bioconductor
   GenomicAlignments          1.40.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges            * 1.56.1    2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm                 0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2                  * 3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrastr                  * 1.0.2     2023-06-01 [1] CRAN (R 4.4.0)
   ggrepel                  * 0.9.5     2024-01-10 [1] CRAN (R 4.4.0)
   ggstance                 * 0.3.7     2024-04-05 [1] CRAN (R 4.4.0)
   ggtext                   * 0.1.2     2022-09-16 [1] CRAN (R 4.4.0)
   glue                       1.7.0     2024-01-09 [1] CRAN (R 4.4.0)
   golem                      0.4.1     2023-06-05 [1] CRAN (R 4.4.0)
   gridExtra                  2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gridtext                   0.1.5     2022-09-16 [1] CRAN (R 4.4.0)
   gtable                     0.3.5     2024-04-22 [1] CRAN (R 4.4.0)
   HDF5Array                  1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   here                       1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools                  0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets                1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                     1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
   httr                       1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   IRanges                  * 2.38.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                      2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   iterators                  1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
   jpeg                       0.1-10    2022-11-29 [1] CRAN (R 4.4.0)
   jquerylib                  0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite                   1.8.8     2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST                   1.44.1    2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   KernSmooth                 2.23-24   2024-05-17 [1] CRAN (R 4.4.1)
   knitr                      1.48      2024-07-07 [1] CRAN (R 4.4.1)
   labeling                   0.4.3     2023-08-29 [1] CRAN (R 4.4.0)
   later                      1.3.2     2023-12-06 [1] CRAN (R 4.4.0)
   lattice                    0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval                   0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle                  1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                      3.60.3    2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                     1.5-9.10  2024-06-24 [1] CRAN (R 4.4.0)
   magick                     2.8.3     2024-02-18 [1] CRAN (R 4.4.0)
   magrittr                   2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   maps                       3.4.2     2023-12-15 [1] CRAN (R 4.4.0)
   markdown                   1.13      2024-06-04 [1] CRAN (R 4.4.0)
   Matrix                     1.7-0     2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics           * 1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats              * 1.3.0     2024-04-11 [1] CRAN (R 4.4.0)
   memoise                    2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   mime                       0.12      2021-09-28 [1] CRAN (R 4.4.0)
   munsell                    0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   paletteer                  1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
   parallelly               * 1.38.0    2024-07-27 [1] CRAN (R 4.4.0)
   pillar                     1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig                  2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   plotly                     4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
   png                        0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   Polychrome               * 1.5.1     2022-05-03 [1] CRAN (R 4.4.0)
   promises                   1.3.0     2024-04-05 [1] CRAN (R 4.4.0)
   proxy                      0.4-27    2022-06-09 [1] CRAN (R 4.4.0)
   purrr                      1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R.methodsS3                1.8.2     2022-06-13 [1] CRAN (R 4.4.0)
   R.oo                       1.26.0    2024-01-24 [1] CRAN (R 4.4.0)
   R.utils                    2.12.3    2023-11-18 [1] CRAN (R 4.4.0)
   R6                         2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs                   0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer               1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                       1.0.13    2024-07-17 [1] CRAN (R 4.4.0)
   RCurl                      1.98-1.14 2024-01-09 [1] CRAN (R 4.4.0)
   rematch2                   2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
   restfulr                   0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
   rhdf5                      2.48.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rhdf5filters               1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Rhdf5lib                   1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rjson                      0.2.21    2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                    * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown                  2.27      2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot                  2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools                  2.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                    2.3.7     2024-05-27 [1] CRAN (R 4.4.0)
   rstudioapi                 0.16.0    2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                       1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer                1.64.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   s2                         1.1.6     2023-12-19 [1] CRAN (R 4.4.0)
   S4Arrays                   1.4.1     2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors                * 0.42.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                       0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix               1.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                     1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater                     1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scatterplot3d              0.3-44    2023-05-05 [1] CRAN (R 4.4.0)
   scuttle                    1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo                1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   sf                       * 1.0-16    2024-03-24 [1] CRAN (R 4.4.0)
   sfheaders                  0.4.4     2024-01-17 [1] CRAN (R 4.4.0)
   shiny                      1.8.1.1   2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets               0.8.6     2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment     * 1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sp                         2.1-4     2024-04-30 [1] CRAN (R 4.4.0)
   spam                       2.10-0    2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray                1.4.8     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats          1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment        * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialFeatureExperiment * 1.6.1     2024-05-15 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD              * 1.16.2    2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   spData                     2.3.1     2024-05-31 [1] CRAN (R 4.4.0)
   spdep                      1.3-5     2024-06-10 [1] CRAN (R 4.4.0)
   statmod                    1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   stringi                    1.8.4     2024-05-06 [1] CRAN (R 4.4.0)
   stringr                    1.5.1     2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment     * 1.34.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   terra                      1.7-78    2024-05-22 [1] CRAN (R 4.4.0)
   tibble                     3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                      1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect                 1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   tiff                       0.1-12    2023-11-28 [1] CRAN (R 4.4.0)
   UCSC.utils                 1.0.0     2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   units                      0.8-5     2023-11-28 [1] CRAN (R 4.4.0)
   utf8                       1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                      0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                      0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                  * 0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite              * 0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   withr                      3.0.0     2024-01-16 [1] CRAN (R 4.4.0)
   wk                         0.9.2     2024-07-09 [1] CRAN (R 4.4.0)
   xfun                       0.45      2024-06-16 [1] CRAN (R 4.4.0)
   XML                        3.99-0.17 2024-06-25 [1] CRAN (R 4.3.3)
   xml2                       1.3.6     2023-12-04 [1] CRAN (R 4.4.0)
   xtable                     1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
   XVector                    0.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                       2.3.9     2024-07-05 [1] CRAN (R 4.4.0)
   zeallot                    0.1.0     2018-01-28 [1] CRAN (R 4.4.0)
   zlibbioc                   1.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────