---
title: "05-Marker-Common Variant Risk Gene GSEA"
author: "Bernard Mulvey"
date: "2023-12-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
require(colorout)
library(fgsea)
library(parallel)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
```

### load aggregated disease-gene lists and reformat
```{r}
gwas <- fread("raw-data/Genesets from GWAS pubs, TWAS, and disgenet/gwassets.txt")
twas <- fread("raw-data/Genesets from GWAS pubs, TWAS, and disgenet/twassets.txt")
dgn <- fread("raw-data/Genesets from GWAS pubs, TWAS, and disgenet/disgenet_curated_genes.txt")
setnames(dgn,"gene_symbol","symbol")
un <- fread("raw-data/Genesets from GWAS pubs, TWAS, and disgenet/nonredundant_union_set_genesymbols.txt")

dissets <- list(gwas,twas,dgn,un)
names(dissets) <- c("gwas","twas","dgn","union")

# make a list object of identifiers per gene set
# result is a list of lists by modality
dissets <- lapply(dissets,FUN=function(x){
    split(x$symbol,x$dx)
})

rm(gwas,twas,dgn,un)
gc(full=T)
```

### marker setup
```{r}
mks <- readRDS("data/07-Marker_genes/01b-BSpace_allruns_spatialLIBD-regwrap-enrichment_full.RDS")

mks <- mks[grep(names(mks),pattern="HARMONYlmbna_nnsvg10_60kiter")]
# check:
# names(mks)
# [1] "k15_HARMONYlmbna_nnsvg10_60kiter"          
# [2] "k20_HARMONYlmbna_nnsvg10_60kiter"          
# [3] "k31_HARMONYlmbna_nnsvg10_60kiter"          
# [4] "k15_HARMONYlmbna_nnsvg10_60kiter_collapsed"

### drop mito genes
mks <- lapply(mks,FUN=function(x){x[!(gene %in% grep(gene,pattern="MT-",value=T))]})

### get named vectors of marker t-stats for each cluster in each result set (15+20+31+13)=79. we need symbols since not all of the dis-gene data sources provide ens identifiers
mklist <- list()
i<-1
for (i in c(1:length(mks))){
    sdoms <- grep(names(mks[[i]]),pattern="logFC_",value=T)
    sdoms <- gsub(sdoms,pattern="^logFC_(.*)$",replacement="\\1")
    j<-1
    for (j in c(1:length(sdoms))){
        getcol <- paste0("t_stat_",sdoms[j])
        tmp <- copy(mks[[i]])[,c("gene",..getcol)]
        setorderv(tmp,getcol,-1)
        ## get the stat column as a vector and give it the gene ids as names for fgsea
        ## note that since data.table (invisibly) uses list() for column data, we have to call unlist() to get a bonafide vector below
        setnames(tmp,getcol,"stat")
        tmp <- as.data.frame(tmp)
        v <- tmp$stat
        names(v) <- tmp$gene
        mklist[[length(mklist)+1]] <- v
        names(mklist)[length(mklist)] <- paste0(names(mks)[i],"_",sdoms[j])
        rm(getcol,v,tmp)
    }
    rm(j,sdoms)
}
rm(i,mks)
```

```{r}
### loop through each domain and modality-disease gene set (each run is very fast with 10 cpus in biocparallel, so just serially run 10cpu GSEA for each cluster)
dissets.res <- list()
sbp <- MulticoreParam(10)
register(sbp)

i<-1
for (i in c(1:length(dissets))){
    cursets <- copy(dissets[[i]])
    mkres <- mapply(X=mklist,Y=names(mklist),SIMPLIFY=FALSE,FUN=function(X,Y){
        z<-as.data.table(fgseaMultilevel(pathways=cursets,stats = X,minSize = 15,maxSize = 500,eps=0.0,BPPARAM = sbp,nPermSimple = 50000))
        z[,spdom:=Y]
        return(z)
    })
    dissets.res[[i]] <- mkres
    names(dissets.res)[i] <- names(dissets)[i]
    rm(cursets)
}
rm(i)
```

#### concatenate
```{r}
i <-1 
for (i in c(1:length(dissets.res))){
    j<-1
    for (j in c(1:length(dissets.res[[i]]))){
        if(i==1 & j==1){
            allmkdisres <- copy(dissets.res[[i]][[j]])
            allmkdisres[,genesets_from:=names(dissets.res)[i]]
        }
        else{
            tmp <- copy(dissets.res[[i]][[j]])
            tmp[,genesets_from:=names(dissets.res)[i]]
            allmkdisres <- rbind(allmkdisres,tmp)
            rm(tmp)
        }
    }
}

saveRDS(allmkdisres,"data/11-GSEA/05-Common variant disease gene enrichment in markers.RDS")
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] BiocParallel_1.36.0 parallelly_1.36.0   default_1.0.0      
 [4] fgsea_1.28.0        colorout_1.3-0      Biostrings_2.70.1  
 [7] GenomeInfoDb_1.38.2 XVector_0.42.0      IRanges_2.36.0     
[10] S4Vectors_0.40.2    BiocGenerics_0.48.1 data.table_1.14.10 
[13] rlang_1.1.2        

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.0        viridisLite_0.4.2       dplyr_1.1.4            
 [4] farver_2.1.1            bitops_1.0-7            fastmap_1.1.1          
 [7] RCurl_1.98-1.13         tweenr_2.0.2            promises_1.2.1         
[10] digest_0.6.33           mime_0.12               lifecycle_1.0.4        
[13] ellipsis_0.3.2          magrittr_2.0.3          compiler_4.3.1         
[16] tools_4.3.1             igraph_1.6.0            utf8_1.2.4             
[19] yaml_2.3.8              knitr_1.45              htmlwidgets_1.6.4      
[22] here_1.0.1              plyr_1.8.9              xml2_1.3.6             
[25] rsconnect_1.2.0         babelgene_22.9          withr_2.5.2            
[28] purrr_1.0.2             grid_4.3.1              polyclip_1.10-6        
[31] fansi_1.0.6             xtable_1.8-4            colorspace_2.1-0       
[34] ggplot2_3.4.4           scales_1.3.0            MASS_7.3-60            
[37] cli_3.6.2               hypeR_2.0.0             rmarkdown_2.25         
[40] crayon_1.5.2            generics_0.1.3          rstudioapi_0.15.0      
[43] httr_1.4.7              reshape2_1.4.4          sessioninfo_1.2.2      
[46] visNetwork_2.1.2        ggforce_0.4.1           stringr_1.5.1          
[49] zlibbioc_1.48.0         rvest_1.0.3             msigdbr_7.5.1          
[52] vctrs_0.6.5             Matrix_1.6-4            webshot_0.5.5          
[55] jsonlite_1.8.8          systemfonts_1.0.5       glue_1.6.2             
[58] codetools_0.2-19        cowplot_1.1.2           stringi_1.8.3          
[61] gtable_0.3.4            later_1.3.2             munsell_0.5.0          
[64] tibble_3.2.1            pillar_1.9.0            htmltools_0.5.7        
[67] reactable_0.4.4         GenomeInfoDbData_1.2.11 R6_2.5.1               
[70] rprojroot_2.0.4         lattice_0.22-5          evaluate_0.23          
[73] shiny_1.8.0             kableExtra_1.3.4        openxlsx_4.2.5.2       
[76] httpuv_1.6.13           Rcpp_1.0.11             zip_2.3.0              
[79] fastmatch_1.1-4         svglite_2.1.3           xfun_0.41              
[82] pkgconfig_2.0.3        
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.1 (2023-06-16)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2023-12-23
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────
 package          * version   date (UTC) lib source
 babelgene          22.9      2022-09-29 [1] CRAN (R 4.3.0)
 BiocGenerics     * 0.48.1    2023-11-02 [1] Bioconductor
 BiocParallel     * 1.36.0    2023-10-26 [1] Bioconductor
 Biostrings       * 2.70.1    2023-10-26 [1] Bioconductor
 bitops             1.0-7     2021-04-24 [1] CRAN (R 4.3.0)
 cli                3.6.2     2023-12-11 [1] CRAN (R 4.3.1)
 codetools          0.2-19    2023-02-01 [1] CRAN (R 4.3.0)
 colorout         * 1.3-0     2023-11-01 [1] local (/Users/bmulvey/Desktop/colorout)
 colorspace         2.1-0     2023-01-23 [1] CRAN (R 4.3.0)
 cowplot            1.1.2     2023-12-15 [1] CRAN (R 4.3.1)
 crayon             1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 data.table       * 1.14.10   2023-12-08 [1] CRAN (R 4.3.1)
 default          * 1.0.0     2017-08-07 [1] CRAN (R 4.3.0)
 digest             0.6.33    2023-07-07 [1] CRAN (R 4.3.0)
 dplyr              1.1.4     2023-11-17 [1] CRAN (R 4.3.1)
 ellipsis           0.3.2     2021-04-29 [1] CRAN (R 4.3.0)
 evaluate           0.23      2023-11-01 [1] CRAN (R 4.3.1)
 fansi              1.0.6     2023-12-08 [1] CRAN (R 4.3.1)
 farver             2.1.1     2022-07-06 [1] CRAN (R 4.3.0)
 fastmap            1.1.1     2023-02-24 [1] CRAN (R 4.3.0)
 fastmatch          1.1-4     2023-08-18 [1] CRAN (R 4.3.0)
 fgsea            * 1.28.0    2023-10-24 [1] Bioconductor
 generics           0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb     * 1.38.2    2023-12-13 [1] Bioconductor 3.18 (R 4.3.1)
 GenomeInfoDbData   1.2.11    2023-11-10 [1] Bioconductor
 ggforce            0.4.1     2022-10-04 [1] CRAN (R 4.3.0)
 ggplot2            3.4.4     2023-10-12 [1] CRAN (R 4.3.1)
 glue               1.6.2     2022-02-24 [1] CRAN (R 4.3.0)
 gtable             0.3.4     2023-08-21 [1] CRAN (R 4.3.0)
 here               1.0.1     2020-12-13 [1] CRAN (R 4.3.0)
 htmltools          0.5.7     2023-11-03 [1] CRAN (R 4.3.1)
 htmlwidgets        1.6.4     2023-12-06 [1] CRAN (R 4.3.1)
 httpuv             1.6.13    2023-12-06 [1] CRAN (R 4.3.1)
 httr               1.4.7     2023-08-15 [1] CRAN (R 4.3.0)
 hypeR              2.0.0     2023-10-24 [1] Bioconductor
 igraph             1.6.0     2023-12-11 [1] CRAN (R 4.3.1)
 IRanges          * 2.36.0    2023-10-26 [1] Bioconductor
 jsonlite           1.8.8     2023-12-04 [1] CRAN (R 4.3.1)
 kableExtra         1.3.4     2021-02-20 [1] CRAN (R 4.3.0)
 knitr              1.45      2023-10-30 [1] CRAN (R 4.3.1)
 later              1.3.2     2023-12-06 [1] CRAN (R 4.3.1)
 lattice            0.22-5    2023-10-24 [1] CRAN (R 4.3.1)
 lifecycle          1.0.4     2023-11-07 [1] CRAN (R 4.3.1)
 magrittr           2.0.3     2022-03-30 [1] CRAN (R 4.3.0)
 MASS               7.3-60    2023-05-04 [1] CRAN (R 4.3.0)
 Matrix             1.6-4     2023-11-30 [1] CRAN (R 4.3.1)
 mime               0.12      2021-09-28 [1] CRAN (R 4.3.0)
 msigdbr            7.5.1     2022-03-30 [1] CRAN (R 4.3.0)
 munsell            0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 openxlsx           4.2.5.2   2023-02-06 [1] CRAN (R 4.3.0)
 parallelly       * 1.36.0    2023-05-26 [1] CRAN (R 4.3.0)
 pillar             1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig          2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 plyr               1.8.9     2023-10-02 [1] CRAN (R 4.3.1)
 polyclip           1.10-6    2023-09-27 [1] CRAN (R 4.3.1)
 promises           1.2.1     2023-08-10 [1] CRAN (R 4.3.0)
 purrr              1.0.2     2023-08-10 [1] CRAN (R 4.3.0)
 R6                 2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 Rcpp               1.0.11    2023-07-06 [1] CRAN (R 4.3.1)
 RCurl              1.98-1.13 2023-11-02 [1] CRAN (R 4.3.1)
 reactable          0.4.4     2023-03-12 [1] CRAN (R 4.3.0)
 reshape2           1.4.4     2020-04-09 [1] CRAN (R 4.3.0)
 rlang            * 1.1.2     2023-11-04 [1] CRAN (R 4.3.1)
 rmarkdown          2.25      2023-09-18 [1] CRAN (R 4.3.1)
 rprojroot          2.0.4     2023-11-05 [1] CRAN (R 4.3.1)
 rsconnect          1.2.0     2023-12-15 [1] CRAN (R 4.3.1)
 rstudioapi         0.15.0    2023-07-07 [1] CRAN (R 4.3.0)
 rvest              1.0.3     2022-08-19 [1] CRAN (R 4.3.0)
 S4Vectors        * 0.40.2    2023-11-25 [1] Bioconductor 3.18 (R 4.3.2)
 scales             1.3.0     2023-11-28 [1] CRAN (R 4.3.1)
 sessioninfo        1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 shiny              1.8.0     2023-11-17 [1] CRAN (R 4.3.1)
 stringi            1.8.3     2023-12-11 [1] CRAN (R 4.3.1)
 stringr            1.5.1     2023-11-14 [1] CRAN (R 4.3.1)
 svglite            2.1.3     2023-12-08 [1] CRAN (R 4.3.1)
 systemfonts        1.0.5     2023-10-09 [1] CRAN (R 4.3.1)
 tibble             3.2.1     2023-03-20 [1] CRAN (R 4.3.0)
 tidyselect         1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 tweenr             2.0.2     2022-09-06 [1] CRAN (R 4.3.0)
 utf8               1.2.4     2023-10-22 [1] CRAN (R 4.3.1)
 vctrs              0.6.5     2023-12-01 [1] CRAN (R 4.3.1)
 viridisLite        0.4.2     2023-05-02 [1] CRAN (R 4.3.0)
 visNetwork         2.1.2     2022-09-29 [1] CRAN (R 4.3.0)
 webshot            0.5.5     2023-06-26 [1] CRAN (R 4.3.0)
 withr              2.5.2     2023-10-30 [1] CRAN (R 4.3.1)
 xfun               0.41      2023-11-01 [1] CRAN (R 4.3.1)
 xml2               1.3.6     2023-12-04 [1] CRAN (R 4.3.1)
 xtable             1.8-4     2019-04-21 [1] CRAN (R 4.3.0)
 XVector          * 0.42.0    2023-10-26 [1] Bioconductor
 yaml               2.3.8     2023-12-11 [1] CRAN (R 4.3.1)
 zip                2.3.0     2023-04-17 [1] CRAN (R 4.3.0)
 zlibbioc           1.48.0    2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────
