---
title: "07-Sex DEG hypergeometric tests for GWAS gene sets"
author: "Bernard Mulvey"
date: "2024-01-13"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(hypeR)
library(BiocParallel)
library(parallelly)


require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

```

```{r}
degs <- readRDS("data/09-Sex DE/01-dreamlet-ranfx_nnsvg10-HmnylmbNA-BS-15-20-31-15VMHARCclpsd.RDS")

un <- fread("raw-data/Genesets from GWAS pubs, TWAS, and disgenet/nonredundant_union_set_genesymbols.txt")

sbp <- MulticoreParam(4)
reslist <- bplapply(degs, BPPARAM=sbp,function(x){
    i <- 1
    ndom <- length(unique(x$assay))
    for (i in c(1:(2*ndom))){
        if (i<=ndom){
            d <- i
            dom <- unique(x$assay)[d]
            domname <- paste0(dom,"_MaleUp")
        }
        else{
            d <- i-ndom
            dom <- unique(x$assay)[d]
            domname <- paste0(dom,"_FemUp")
        }
        
        curde <- x[assay==dom]
        
        if (i<=ndom){
            cursig <- curde[adj.P.Val<0.05&logFC>0,gene_name]
        }
        else {
            cursig <- curde[adj.P.Val<0.05&logFC<0,gene_name]
        }
        
        if(length(cursig)==0){
            next(i)
        }
        

        # subset disease genes to unique, shared genes analyzed for DE in the current domain and format into a list of vectors for each disease
        j<-1
        curun <- un[symbol %in% x[assay==dom,gene_name]]
        tmplists <- list()
        for (j in c(1:length(unique(un$dx)))){
            curdx <- unique(un$dx)[j]
            curdxset <- unique(curun[dx==curdx,symbol])
            tmplists[[j]] <- curdxset
            names(tmplists)[j] <- curdx
            # belt and suspenders clearout
            rm(curdx,curdxset)
        }
        rm(j)
        
        # actual hypergeo tests
        tmpres <- hypeR::hypeR(signature=cursig,genesets=tmplists,background = nrow(curde)-length(cursig),test = "hypergeometric")
        tmpres2 <- as.data.table(tmpres$data)
        tmpres2[,spdomain_sex:=domname]
    
        if (i==1){
            currestab <- copy(tmpres2)
        }
        else{
            currestab <- rbind(copy(tmpres2),currestab)
        }
            rm(tmpres,tmpres2,curun,tmplists)
    }
    return(currestab)
})
rm(i,j)

i<-1
for (i in c(1:length(reslist))){
    if (i==1){
        resdat <- copy(reslist[[i]])
        resdat[,clus_set:=names(reslist)[i]]
    }
    else{
        tmp <- copy(reslist[[i]])
        tmp[,clus_set:=names(reslist)[i]]
        resdat <- rbind(resdat,tmp)
        rm(tmp)
    }
} 
    
fwrite(resdat,"data/11-GSEA/07-hypergeometric enrichment for FDR-sig sex DEGs per k15 cluster in neuropsych GWAS-TWAS-litmined union sets.txt",sep="\t",quote=F)
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] colorout_1.3-0.1    parallelly_1.36.0   BiocParallel_1.36.0
 [4] hypeR_2.0.0         Biostrings_2.70.1   GenomeInfoDb_1.38.5
 [7] XVector_0.42.0      IRanges_2.36.0      S4Vectors_0.40.2   
[10] BiocGenerics_0.48.1 data.table_1.14.10  rlang_1.1.3        

loaded via a namespace (and not attached):
 [1] gtable_0.3.4            babelgene_22.9          xfun_0.41              
 [4] ggplot2_3.4.4           visNetwork_2.1.2        htmlwidgets_1.6.4      
 [7] bitops_1.0-7            vctrs_0.6.5             tools_4.3.2            
[10] generics_0.1.3          parallel_4.3.2          tibble_3.2.1           
[13] fansi_1.0.6             pkgconfig_2.0.3         webshot_0.5.5          
[16] GenomeInfoDbData_1.2.11 lifecycle_1.0.4         compiler_4.3.2         
[19] farver_2.1.1            stringr_1.5.1           munsell_0.5.0          
[22] ggforce_0.4.1           codetools_0.2-19        httpuv_1.6.13          
[25] htmltools_0.5.7         RCurl_1.98-1.14         crayon_1.5.2           
[28] pillar_1.9.0            later_1.3.2             MASS_7.3-60.0.1        
[31] ellipsis_0.3.2          sessioninfo_1.2.2       mime_0.12              
[34] tidyselect_1.2.0        rvest_1.0.3             zip_2.3.0              
[37] digest_0.6.34           stringi_1.8.3           dplyr_1.1.4            
[40] reshape2_1.4.4          purrr_1.0.2             rprojroot_2.0.4        
[43] polyclip_1.10-6         fastmap_1.1.1           grid_4.3.2             
[46] here_1.0.1              colorspace_2.1-0        cli_3.6.2              
[49] magrittr_2.0.3          utf8_1.2.4              withr_3.0.0            
[52] scales_1.3.0            promises_1.2.1          rmarkdown_2.25         
[55] httr_1.4.7              igraph_1.6.0            openxlsx_4.2.5.2       
[58] kableExtra_1.3.4        shiny_1.8.0             evaluate_0.23          
[61] knitr_1.45              viridisLite_0.4.2       msigdbr_7.5.1          
[64] Rcpp_1.0.12             xtable_1.8-4            glue_1.7.0             
[67] reactable_0.4.4         tweenr_2.0.2            xml2_1.3.6             
[70] jsonlite_1.8.8          svglite_2.1.3           rstudioapi_0.15.0      
[73] R6_2.5.1                plyr_1.8.9              zlibbioc_1.48.0        
[76] systemfonts_1.0.5      
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-01-21
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────
 package          * version    date (UTC) lib source
 babelgene          22.9       2022-09-29 [1] CRAN (R 4.3.0)
 BiocGenerics     * 0.48.1     2023-11-02 [1] Bioconductor
 BiocParallel     * 1.36.0     2023-10-26 [1] Bioconductor
 Biostrings       * 2.70.1     2023-10-25 [1] Bioconductor
 bitops             1.0-7      2021-04-24 [1] CRAN (R 4.3.2)
 cli                3.6.2      2023-12-11 [1] CRAN (R 4.3.2)
 codetools          0.2-19     2023-02-01 [1] CRAN (R 4.3.0)
 colorout         * 1.3-0.1    2024-01-11 [1] local
 colorspace         2.1-0      2023-01-23 [1] CRAN (R 4.3.2)
 crayon             1.5.2      2022-09-29 [1] CRAN (R 4.3.0)
 data.table       * 1.14.10    2023-12-08 [1] CRAN (R 4.3.2)
 digest             0.6.34     2024-01-11 [1] CRAN (R 4.3.2)
 dplyr              1.1.4      2023-11-17 [1] CRAN (R 4.3.2)
 ellipsis           0.3.2      2021-04-29 [1] CRAN (R 4.3.2)
 evaluate           0.23       2023-11-01 [1] CRAN (R 4.3.1)
 fansi              1.0.6      2023-12-08 [1] CRAN (R 4.3.2)
 farver             2.1.1      2022-07-06 [1] CRAN (R 4.3.2)
 fastmap            1.1.1      2023-02-24 [1] CRAN (R 4.3.2)
 generics           0.1.3      2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb     * 1.38.5     2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData   1.2.11     2024-01-11 [1] Bioconductor
 ggforce            0.4.1      2022-10-04 [1] CRAN (R 4.3.2)
 ggplot2            3.4.4      2023-10-12 [1] CRAN (R 4.3.1)
 glue               1.7.0      2024-01-09 [1] CRAN (R 4.3.2)
 gtable             0.3.4      2023-08-21 [1] CRAN (R 4.3.0)
 here               1.0.1      2020-12-13 [1] CRAN (R 4.3.0)
 htmltools          0.5.7      2023-11-03 [1] CRAN (R 4.3.2)
 htmlwidgets        1.6.4      2023-12-06 [1] CRAN (R 4.3.1)
 httpuv             1.6.13     2023-12-06 [1] CRAN (R 4.3.1)
 httr               1.4.7      2023-08-15 [1] CRAN (R 4.3.0)
 hypeR            * 2.0.0      2023-10-26 [1] Bioconductor
 igraph             1.6.0      2023-12-11 [1] CRAN (R 4.3.1)
 IRanges          * 2.36.0     2023-10-24 [1] Bioconductor
 jsonlite           1.8.8      2023-12-04 [1] CRAN (R 4.3.2)
 kableExtra         1.3.4      2021-02-20 [1] CRAN (R 4.3.0)
 knitr              1.45       2023-10-30 [1] CRAN (R 4.3.2)
 later              1.3.2      2023-12-06 [1] CRAN (R 4.3.2)
 lifecycle          1.0.4      2023-11-07 [1] CRAN (R 4.3.1)
 magrittr           2.0.3      2022-03-30 [1] CRAN (R 4.3.2)
 MASS               7.3-60.0.1 2024-01-13 [1] CRAN (R 4.3.2)
 mime               0.12       2021-09-28 [1] CRAN (R 4.3.2)
 msigdbr            7.5.1      2022-03-30 [1] CRAN (R 4.3.2)
 munsell            0.5.0      2018-06-12 [1] CRAN (R 4.3.0)
 openxlsx           4.2.5.2    2023-02-06 [1] CRAN (R 4.3.2)
 parallelly       * 1.36.0     2023-05-26 [1] CRAN (R 4.3.0)
 pillar             1.9.0      2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig          2.0.3      2019-09-22 [1] CRAN (R 4.3.0)
 plyr               1.8.9      2023-10-02 [1] CRAN (R 4.3.2)
 polyclip           1.10-6     2023-09-27 [1] CRAN (R 4.3.1)
 promises           1.2.1      2023-08-10 [1] CRAN (R 4.3.2)
 purrr              1.0.2      2023-08-10 [1] CRAN (R 4.3.2)
 R6                 2.5.1      2021-08-19 [1] CRAN (R 4.3.0)
 Rcpp               1.0.12     2024-01-09 [1] CRAN (R 4.3.2)
 RCurl              1.98-1.14  2024-01-09 [1] CRAN (R 4.3.1)
 reactable          0.4.4      2023-03-12 [1] CRAN (R 4.3.0)
 reshape2           1.4.4      2020-04-09 [1] CRAN (R 4.3.2)
 rlang            * 1.1.3      2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown          2.25       2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot          2.0.4      2023-11-05 [1] CRAN (R 4.3.1)
 rstudioapi         0.15.0     2023-07-07 [1] CRAN (R 4.3.0)
 rvest              1.0.3      2022-08-19 [1] CRAN (R 4.3.0)
 S4Vectors        * 0.40.2     2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 scales             1.3.0      2023-11-28 [1] CRAN (R 4.3.2)
 sessioninfo        1.2.2      2021-12-06 [1] CRAN (R 4.3.0)
 shiny              1.8.0      2023-11-17 [1] CRAN (R 4.3.1)
 stringi            1.8.3      2023-12-11 [1] CRAN (R 4.3.1)
 stringr            1.5.1      2023-11-14 [1] CRAN (R 4.3.1)
 svglite            2.1.3      2023-12-08 [1] CRAN (R 4.3.1)
 systemfonts        1.0.5      2023-10-09 [1] CRAN (R 4.3.2)
 tibble             3.2.1      2023-03-20 [1] CRAN (R 4.3.2)
 tidyselect         1.2.0      2022-10-10 [1] CRAN (R 4.3.0)
 tweenr             2.0.2      2022-09-06 [1] CRAN (R 4.3.2)
 utf8               1.2.4      2023-10-22 [1] CRAN (R 4.3.2)
 vctrs              0.6.5      2023-12-01 [1] CRAN (R 4.3.2)
 viridisLite        0.4.2      2023-05-02 [1] CRAN (R 4.3.0)
 visNetwork         2.1.2      2022-09-29 [1] CRAN (R 4.3.0)
 webshot            0.5.5      2023-06-26 [1] CRAN (R 4.3.0)
 withr              3.0.0      2024-01-16 [1] CRAN (R 4.3.1)
 xfun               0.41       2023-11-01 [1] CRAN (R 4.3.2)
 xml2               1.3.6      2023-12-04 [1] CRAN (R 4.3.2)
 xtable             1.8-4      2019-04-21 [1] CRAN (R 4.3.0)
 XVector          * 0.42.0     2023-10-24 [1] Bioconductor
 zip                2.3.0      2023-04-17 [1] CRAN (R 4.3.2)
 zlibbioc           1.48.0     2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

───────────────────────────────────────────────────────────────────────────
