---
title: "12-Sex DE and GSEA supp table outputs"
author: "Bernard Mulvey"
date: "2024-02-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

```

### read in, write out sex DE dreamlet tabs
```{r}
sexde <- readRDS("data/09-Sex DE/01-dreamlet-ranfx_nnsvg10-HmnylmbNA-BS-15-20-31-15VMHARCclpsd.RDS")

ofile <- names(sexde)

mapply(x=sexde,y=ofile,FUN=function(x,y){
    setnames(x,c("Gene ID","Domain","logFC (Male to Female)","Avg Expression","t","P.Value","adj.P.Val","B","z.std","Gene Symbol","Chromosome"))
    x <- x[,c(2,10,3,7,6,5,8,9,4,11,1)]
    fwrite(x,paste0("local_supptab_assembly/sex de sheets/dreamlet_",y,".txt"),sep='\t',quote=F)})
rm(sexde,ofile)
```

### read in msigdb pre and post "collapsePathways" results
```{r}
msigdbfull <- readRDS("data/11-GSEA/03a-Sex DE msigdb_1,2,3,5,6,8 GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")

msigdbcollapsed <- readRDS("data/11-GSEA/03c-Sex DE mSigDB_1,2,3,5,6,8 GSEA clpspath mainPathways-only.RDS")

msigbindpatterns <- c("BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH","BSpace_k15_HARMONYlmbna_nnsvg10","BSpace_k20_HARMONYlmbna_nnsvg10","BSpace_k31_HARMONYlmbna_nnsvg10")

clustres <- list()
i<-1
for (i in c(1:length(msigbindpatterns))){
    curpatt <- msigbindpatterns[i]
    subfull <- msigdbfull[grep(names(msigdbfull),pattern=curpatt)]
    subclps <- msigdbcollapsed[grep(names(msigdbcollapsed),pattern=curpatt)]
    
    j<-1
    for (j in c(1:length(subfull))){
        tmpres <- subfull[[j]]
        tmpres[,spdom:=gsub(spdom,pattern="^.*_(X.*)$",replacement="\\1")]
        
        tmpres[,mainPathway:="No"]
        tmpres[pathway %in% msigdbcollapsed[[names(subfull)[j]]]$pathway,mainPathway:="Yes"]
        
        if(j==1){catres <- copy(tmpres)}
        else{catres <- rbind(catres,tmpres)}
        rm(tmpres)
    }
    clustres[[i]] <- catres
    rm(subfull,subclps,curpatt,j)
}
rm(i,catres)

### fix VMH/ARC cluster labels for the collapsed case which doesn't really work with this (no leading X)
clustres[[1]][,spdom:=gsub(spdom,pattern="^BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_(.*)$",replacement="\\1")]

unique(clustres[[1]]$spdom)

## remove collapsed cluster results from the k=15 single domain results. not sure why i copied them there before. assign X3/5/11/12 the respective VMH and ARC ids previously used
clustres[[2]] <- clustres[[2]][spdom %in% paste0("X",c(1:15))]
clustres[[2]][spdom=="X3",spdom:="ARC.1"]
clustres[[2]][spdom=="X5",spdom:="ARC.2"]
clustres[[2]][spdom=="X11",spdom:="VMH.1"]
clustres[[2]][spdom=="X12",spdom:="VMH.2"]

### add the k20/k31 ARC and VMH cluster identifiers as used elsewhere
clustres[[3]][spdom=="X7",spdom:="VMH20.1"]
clustres[[3]][spdom=="X9",spdom:="VMH20.2"]
clustres[[3]][spdom=="X13",spdom:="VMH20.3"]
clustres[[3]][spdom=="X14",spdom:="ARC20.1"]
clustres[[3]][spdom=="X18",spdom:="ARC20.2"]

clustres[[4]][spdom=="X3",spdom:="VMH31.1"]
clustres[[4]][spdom=="X4",spdom:="VMH31.2"]
clustres[[4]][spdom=="X6",spdom:="VMH31.3"]
clustres[[4]][spdom=="X19",spdom:="VMH31.4"]
clustres[[4]][spdom=="X20",spdom:="VMH31.5"]
clustres[[4]][spdom=="X29",spdom:="VMH31.6"]
clustres[[4]][spdom=="X1",spdom:="ARC31.1"]
clustres[[4]][spdom=="X7",spdom:="ARC31.2"]
clustres[[4]][spdom=="X8",spdom:="ARC31.3"]
clustres[[4]][spdom=="X21",spdom:="ARC31.4"]

### name the lists and columns, rearrange, and save
names(clustres) <- c("BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH","BSpace_k15_HARMONYlmbna_nnsvg10","BSpace_k20_HARMONYlmbna_nnsvg10","BSpace_k31_HARMONYlmbna_nnsvg10")

clustres <- lapply(clustres,function(x){
    setnames(x,c(1:10),c("Pathway","P.Value","Adj.P","log2err","Enrichment Score","Normalized Enrichment Score","Geneset Size","Leading Edge Genes","Visium Domain","Main Pathway after fgsea::collapsePathways"))
    x <- x[,c(9,1,10,3,6,5,7,4,2,8)]
    return(x)
})

saveRDS(clustres,"local_supptab_assembly/Supp Data-Sex DE msigdb GSEA results.RDS")

rm(clustres,msigdbcollapsed,msigdbfull,msigbindpatterns)
```

### write out the table of TF-target sets tested for GSEA
### this is created as an intermediate product of analysis/11-GSEA/00-Join enrichr TF-target sets...rmd
```{r}
library(Biostrings)
library(stringr)

tftargsets <- list.files("data/11-GSEA/00-enrichr TF-target sets from dbs and lit mining DLed 120623",pattern=".txt",full.names=T)

sourcename <- gsub(tftargsets,pattern="^.*DLed 120623/(.*).txt$",replacement="\\1")

# read each file as a single column table of longass lines, split at the first \t to get set name and set members separated, then expand them out to a one-to-one table to feed into split
tftargsets <- mapply(X=tftargsets,Y=sourcename,SIMPLIFY=F,FUN=function(X,Y){
  tmp <- fread(X,sep = '\n')
  tmp[,setmembers:=apply(.SD,MARGIN=1,FUN=function(x){stringi::stri_split_fixed(x,pattern="\t",n=2)[[1]][2]}),.SDcols=1]
  tmp[,setname:=apply(.SD,MARGIN=1,FUN=function(x){stringi::stri_split_fixed(x,pattern="\t",n=2)[[1]][1]}),.SDcols=1]
  tmp[,setmembers:=gsub(setmembers,pattern="\t",replacement=";")]
  tmp[,setmembers:=gsub("^;","",setmembers)]
  tmp[,setname:=gsub(setname,pattern=" ",replacement="_")]
  tmp[,sourceset:=gsub(Y,pattern=" ",replacement="_")]
  return(tmp[,c(3,2,4)])
})

# for the TF-LOF_Expression_from_GEO we need to take an extra step, since these also have scores (DE values?) with them
tftargmanual <- fread("data/11-GSEA/00-enrichr TF-target sets from dbs and lit mining DLed 120623/TF-LOF_Expression_from_GEO.txt",sep="\n")
setnames(tftargmanual,1,"origstr")
tftargmanual[,setname:=apply(.SD,MARGIN=1,FUN=function(x){str_split_fixed(x,pattern="\t",n=2)[1]}),.SDcols=1]
tftargmanual[,setname:=gsub(" ","_",setname)]
tftargmanual[,setmembers:=apply(.SD,MARGIN=1,FUN=function(x){str_split_fixed(x,pattern="\t",n=2)[2]}),.SDcols=1]
tftargmanual[,setmembers2 := gsub("\\d","",setmembers)]
tftargmanual[,setmembers2 := gsub("\\.","",setmembers2)]
tftargmanual[,setmembers2:=gsub(",\t",";",setmembers2)]
tftargmanual[,setmembers2:=gsub("\t","",setmembers2)]
tftargmanual[,setmembers2:=gsub(" ","",setmembers2)]
tftargmanual[,setmembers:=setmembers2]
tftargmanual[,setmembers2:=NULL]
tftargmanual[,sourceset:="TF-LOF_Expression_from_GEO"]

### replace the original TF-LOF_Expression_from_GEO with the new one
tftargsets[[4]] <- copy(tftargmanual)[,c(3,2,4)]

# clean up
rm(tftargmanual)

# concatenate

i<-1
for (i in c(1:length(tftargsets))){
    if(i==1){tftargtab <- copy(tftargsets[[i]])}
    else{tftargtab <- rbind(tftargtab,tftargsets[[i]])}
}

### add a column corresponding to the TF symbol itself. this comes from code in 11-GSEA/08 and 09 ....rmd

#### separate out the TF symbol from the full set name so we can filter the SPE accordingly
y <- tftargtab[!(sourceset=="Rummagene_transcription_factors")]
y[,tf:=apply(.SD,MARGIN = 1,FUN=function(h){strsplit(h,"_",1)[[1]][1]}),.SDcols="setname"]

### rummagene TFs are named differently:
z <- tftargtab[sourceset=="Rummagene_transcription_factors"]
# reverse the output of a strsplit of a strsplit o.0 to get the string immediately preceding "_Rummagene_.."
z[,tf:=apply(.SD,MARGIN=1,FUN=function(p){
    rev.default(strsplit(p,"_",-1)[[1]])[1]
}),.SDcols="setname"]
fix <- rbind(z,y)
# make all upper case
fix[,tf:=toupper(tf)]


### finally, change the TF set names to match those in the result tables by appending the source
fix[,setname:=paste0(setname,"_",sourceset)]

names(fix)
fix <- fix[,c(4,1,3,2)]
setnames(fix,c("TF Symbol","TF Target Set Name","TF Target Set Source","TF Target Symbols"))



fwrite(fix,"local_supptab_assembly/TF_target_sets.txt",sep='\t',quote=F)
 
### change names back to simple ones for merging in this info into results
setnames(fix,c("tf","setname","sourceset","setmembers"))

rm(y,z,tftargsets,tftargtab,tftarg,sourcename,tftargs)
```

### write out sex DE-TF target results; append the TF symbol; append whether that TF is among those that passed the domain-expression threshold
```{r}
tffull <- readRDS("data/11-GSEA/04a-Sex DE TF-target GSEA svg10-hmnyLmdaNA_BS60k-k15-k20-k31-k15clpsd.RDS")

tfcollapsed <- readRDS("data/11-GSEA/04c-Sex DE TF-target GSEA clpspath mainPathways-only.RDS")

tfxprd <- readRDS("data/11-GSEA/09-Sex DE TF-target GSEA results from 04a filtered to TFs expressed in avg 1pct of cluster spots and at least .25pct of cluster spots in 3samps.RDS")

tfbindpatterns <- c("BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH","BSpace_k15_HARMONYlmbna_nnsvg10","BSpace_k20_HARMONYlmbna_nnsvg10","BSpace_k31_HARMONYlmbna_nnsvg10")

clustres <- list()
i<-1
for (i in c(1:length(tfbindpatterns))){
    curpatt <- tfbindpatterns[i]
    subfull <- tffull[grep(names(tffull),pattern=curpatt)]
    subclps <- tfcollapsed[grep(names(tfcollapsed),pattern=curpatt)]
    subxpr <- tfxprd[grep(names(tfcollapsed),pattern=curpatt)]
    j<-1
    for (j in c(1:length(subfull))){
        tmpres <- subfull[[j]]
        tmpres[,spdom:=gsub(spdom,pattern="^.*_(X.*)$",replacement="\\1")]
        
        tmpres[,mainPathway:="No"]
        tmpres[pathway %in% tfcollapsed[[names(subfull)[j]]]$pathway,mainPathway:="Yes"]
        
        tmpres[,Expressed_TF:="No"]
        tmpres[pathway %in% subxpr[[names(subfull)[j]]]$pathway,Expressed_TF:="Yes"]

        
        if(j==1){catres <- copy(tmpres)}
        else{catres <- rbind(catres,tmpres)}
        rm(tmpres)
    }
    clustres[[i]] <- catres
    rm(subfull,subclps,curpatt,j)
}
rm(i,catres)

### fix VMH/ARC cluster labels for the collapsed case which doesn't really work with this (no leading X)
clustres[[1]][,spdom:=gsub(spdom,pattern="^BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH_(.*)$",replacement="\\1")]

unique(clustres[[1]]$spdom)

## remove collapsed cluster results from the k=15 single domain results. not sure why i copied them there before. assign X3/5/11/12 the respective VMH and ARC ids previously used
clustres[[2]] <- clustres[[2]][spdom %in% paste0("X",c(1:15))]
clustres[[2]][spdom=="X3",spdom:="ARC.1"]
clustres[[2]][spdom=="X5",spdom:="ARC.2"]
clustres[[2]][spdom=="X11",spdom:="VMH.1"]
clustres[[2]][spdom=="X12",spdom:="VMH.2"]

### add the k20/k31 ARC and VMH cluster identifiers as used elsewhere
clustres[[3]][spdom=="X7",spdom:="VMH20.1"]
clustres[[3]][spdom=="X9",spdom:="VMH20.2"]
clustres[[3]][spdom=="X13",spdom:="VMH20.3"]
clustres[[3]][spdom=="X14",spdom:="ARC20.1"]
clustres[[3]][spdom=="X18",spdom:="ARC20.2"]

clustres[[4]][spdom=="X3",spdom:="VMH31.1"]
clustres[[4]][spdom=="X4",spdom:="VMH31.2"]
clustres[[4]][spdom=="X6",spdom:="VMH31.3"]
clustres[[4]][spdom=="X19",spdom:="VMH31.4"]
clustres[[4]][spdom=="X20",spdom:="VMH31.5"]
clustres[[4]][spdom=="X29",spdom:="VMH31.6"]
clustres[[4]][spdom=="X1",spdom:="ARC31.1"]
clustres[[4]][spdom=="X7",spdom:="ARC31.2"]
clustres[[4]][spdom=="X8",spdom:="ARC31.3"]
clustres[[4]][spdom=="X21",spdom:="ARC31.4"]

# name lists and save RDS
names(clustres) <- c("BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH","BSpace_k15_HARMONYlmbna_nnsvg10","BSpace_k20_HARMONYlmbna_nnsvg10","BSpace_k31_HARMONYlmbna_nnsvg10")

clustres <- lapply(clustres,function(x){
    z <- merge.data.table(fix[,.(tf,setname)],x,by.x="setname",by.y="pathway",all.y=T)
    setnames(z,c("TF Target Set","TF Symbol","P.Value","Adj.P","log2err","Enrichment Score","Normalized Enrichment Score","Geneset Size","Leading Edge Genes","Visium Domain","Main Geneset after fgsea::collapsePathways","Expressed TF"))
    z <- z[,c(2,10:12,4,7,6,5,1,8,3,9)]
    return(z)
})

saveRDS(clustres,"local_supptab_assembly/Supp Data-sexDE_TFtarget_GSEA_results.RDS")

rm(clustres,tfcollapsed,tffull,tfbindpatterns,fix)
```

### write out sex DE - disease GSEA results
```{r}
sexdisres <- readRDS("data/11-GSEA/06-Common variant disease gene enrichment in sex DEGs.RDS")

unique(sexdisres$spdom)

spdompatt <- c("BSpace_k15_HARMONYlmbna_nnsvg10_clpsdARC_clpsdVMH","BSpace_k15_HARMONYlmbna_nnsvg10_X","BSpace_k20_HARMONYlmbna_nnsvg10","BSpace_k31_HARMONYlmbna_nnsvg10")

i<-1
for (i in c(1:4)){
    
    if(i==2){
        sexdisres[spdom %in% grep(spdom, pattern=spdompatt[i],value=T),Domain:=gsub(spdom,pattern=paste0(spdompatt[i],"(.*)"),replacement=paste0("X","\\1"))]
    }
    
    else{
    sexdisres[spdom %in% grep(spdom, pattern=spdompatt[i],value=T),Domain:=gsub(spdom,pattern=paste0(spdompatt[i],"_(.*)"),replacement="\\1")]
    }
    
    sexdisres[spdom %in% grep(spdom,pattern=spdompatt[i],value=T),`BayesSpace Run`:=gsub(spdom,pattern="BSpace_(k.*)_HARMONY.*",replacement="\\1")]
}

### correct for the collapsed run
sexdisres[spdom %in% grep(spdom,pattern=spdompatt[1],value=T),`BayesSpace Run`:="k15-ARC and VMH Domains Collapsed After"]

sexdisres[,spdom:=NULL]

### add cluster identifiers as above
## remove collapsed cluster results from the k=15 single domain results. not sure why i copied them there before. assign X3/5/11/12 the respective VMH and ARC ids previously used
sexdisres[Domain=="X3"&`BayesSpace Run`=="k15",Domain:="ARC.1"]
sexdisres[Domain=="X5"&`BayesSpace Run`=="k15",Domain:="ARC.2"]
sexdisres[Domain=="X11"&`BayesSpace Run`=="k15",Domain:="VMH.1"]
sexdisres[Domain=="X12"&`BayesSpace Run`=="k15",Domain:="VMH.2"]

### add the k20/k31 ARC and VMH cluster identifiers as used elsewhere
sexdisres[Domain=="X7"&`BayesSpace Run`=="k20",Domain:="VMH20.1"]
sexdisres[Domain=="X9"&`BayesSpace Run`=="k20",Domain:="VMH20.2"]
sexdisres[Domain=="X13"&`BayesSpace Run`=="k20",Domain:="VMH20.3"]
sexdisres[Domain=="X14"&`BayesSpace Run`=="k20",Domain:="ARC20.1"]
sexdisres[Domain=="X18"&`BayesSpace Run`=="k20",Domain:="ARC20.2"]

sexdisres[Domain=="X3"&`BayesSpace Run`=="k31",Domain:="VMH31.1"]
sexdisres[Domain=="X4"&`BayesSpace Run`=="k31",Domain:="VMH31.2"]
sexdisres[Domain=="X6"&`BayesSpace Run`=="k31",Domain:="VMH31.3"]
sexdisres[Domain=="X19"&`BayesSpace Run`=="k31",Domain:="VMH31.4"]
sexdisres[Domain=="X20"&`BayesSpace Run`=="k31",Domain:="VMH31.5"]
sexdisres[Domain=="X29"&`BayesSpace Run`=="k31",Domain:="VMH31.6"]
sexdisres[Domain=="X1"&`BayesSpace Run`=="k31",Domain:="ARC31.1"]
sexdisres[Domain=="X7"&`BayesSpace Run`=="k31",Domain:="ARC31.2"]
sexdisres[Domain=="X8"&`BayesSpace Run`=="k31",Domain:="ARC31.3"]
sexdisres[Domain=="X21"&`BayesSpace Run`=="k31",Domain:="ARC31.4"]

names(sexdisres)
setnames(sexdisres,c("Diagnosis or Trait","P.Value","P.Adj","log2err","Enrichment Score","Normalized Enrichment Score","Geneset Size","Leading Edge Genes","Data or Publication Type from Which Geneset Derives","Domain","BayesSpace Run"))
sexdisres[1,]
sexdisres <- sexdisres[,c(1,9,11,10,3,6,5,4,7,2,8)]
names(sexdisres)

fwrite(sexdisres,"local_supptab_assembly/source mats/sex de sheets/sex-disgene-gsea.txt",sep='\t',quote=F)
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] stringr_1.5.1       Biostrings_2.70.2   GenomeInfoDb_1.38.5
 [4] XVector_0.42.0      IRanges_2.36.0      S4Vectors_0.40.2   
 [7] BiocGenerics_0.48.1 default_1.0.0       colorout_1.3-0.1   
[10] data.table_1.14.10  rlang_1.1.3        

loaded via a namespace (and not attached):
  [1] later_1.3.2                   BiocIO_1.12.0                
  [3] bitops_1.0-7                  filelock_1.0.3               
  [5] fields_15.2                   tibble_3.2.1                 
  [7] XML_3.99-0.16.1               lifecycle_1.0.4              
  [9] edgeR_4.0.14                  doParallel_1.0.17            
 [11] rprojroot_2.0.4               lattice_0.22-5               
 [13] magrittr_2.0.3                limma_3.58.1                 
 [15] plotly_4.10.4                 sass_0.4.8                   
 [17] rmarkdown_2.25                jquerylib_0.1.4              
 [19] yaml_2.3.8                    httpuv_1.6.13                
 [21] spam_2.10-0                   sessioninfo_1.2.2            
 [23] cowplot_1.1.3                 DBI_1.2.1                    
 [25] RColorBrewer_1.1-3            golem_0.4.1                  
 [27] maps_3.4.2                    abind_1.4-5                  
 [29] zlibbioc_1.48.0               GenomicRanges_1.54.1         
 [31] purrr_1.0.2                   RCurl_1.98-1.14              
 [33] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      
 [35] ggrepel_0.9.5                 irlba_2.3.5.1                
 [37] DelayedMatrixStats_1.24.0     codetools_0.2-19             
 [39] DelayedArray_0.28.0           DT_0.31                      
 [41] scuttle_1.12.0                tidyselect_1.2.0             
 [43] ScaledMatrix_1.10.0           viridis_0.6.5                
 [45] shinyWidgets_0.8.1            matrixStats_1.2.0            
 [47] BiocFileCache_2.10.1          GenomicAlignments_1.38.2     
 [49] jsonlite_1.8.8                BiocNeighbors_1.20.2         
 [51] ellipsis_0.3.2                scater_1.30.1                
 [53] iterators_1.0.14              foreach_1.5.2                
 [55] tools_4.3.2                   Rcpp_1.0.12                  
 [57] glue_1.7.0                    gridExtra_2.3                
 [59] SparseArray_1.2.3             xfun_0.41                    
 [61] here_1.0.1                    MatrixGenerics_1.14.0        
 [63] dplyr_1.1.4                   BiocManager_1.30.22          
 [65] fastmap_1.1.1                 fansi_1.0.6                  
 [67] digest_0.6.34                 rsvd_1.0.5                   
 [69] R6_2.5.1                      mime_0.12                    
 [71] colorspace_2.1-0              RSQLite_2.3.5                
 [73] config_0.3.2                  utf8_1.2.4                   
 [75] tidyr_1.3.1                   generics_0.1.3               
 [77] rtracklayer_1.62.0            httr_1.4.7                   
 [79] htmlwidgets_1.6.4             S4Arrays_1.2.0               
 [81] pkgconfig_2.0.3               gtable_0.3.4                 
 [83] blob_1.2.4                    SingleCellExperiment_1.24.0  
 [85] htmltools_0.5.7               dotCall64_1.1-1              
 [87] scales_1.3.0                  Biobase_2.62.0               
 [89] png_0.1-8                     attempt_0.3.1                
 [91] SpatialExperiment_1.12.0      knitr_1.45                   
 [93] rstudioapi_0.15.0             rjson_0.2.21                 
 [95] curl_5.2.0                    cachem_1.0.8                 
 [97] BiocVersion_3.18.1            parallel_4.3.2               
 [99] vipor_0.4.7                   AnnotationDbi_1.64.1         
[101] restfulr_0.0.15               pillar_1.9.0                 
[103] grid_4.3.2                    vctrs_0.6.5                  
[105] promises_1.2.1                BiocSingular_1.18.0          
[107] dbplyr_2.4.0                  beachmat_2.18.0              
[109] xtable_1.8-4                  beeswarm_0.4.0               
[111] paletteer_1.6.0               evaluate_0.23                
[113] magick_2.8.2                  cli_3.6.2                    
[115] locfit_1.5-9.8                compiler_4.3.2               
[117] Rsamtools_2.18.0              crayon_1.5.2                 
[119] rematch2_2.1.2                ggbeeswarm_0.7.2             
[121] stringi_1.8.3                 viridisLite_0.4.2            
[123] BiocParallel_1.36.0           munsell_0.5.0                
[125] lazyeval_0.2.2                Matrix_1.6-5                 
[127] ExperimentHub_2.10.0          benchmarkme_1.0.8            
[129] sparseMatrixStats_1.14.0      bit64_4.0.5                  
[131] ggplot2_3.4.4                 KEGGREST_1.42.0              
[133] statmod_1.5.0                 shiny_1.8.0                  
[135] SummarizedExperiment_1.32.0   interactiveDisplayBase_1.40.0
[137] AnnotationHub_3.10.0          memoise_2.0.1                
[139] bslib_0.6.1                   benchmarkmeData_1.0.4        
[141] bit_4.0.5                     spatialLIBD_1.14.1           
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2024-02-12
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────
 package                * version     date (UTC) lib source
 abind                    1.4-5       2016-07-21 [1] CRAN (R 4.3.0)
 AnnotationDbi            1.64.1      2023-11-02 [1] Bioconductor
 AnnotationHub            3.10.0      2023-10-24 [1] Bioconductor
 attempt                  0.3.1       2020-05-03 [1] CRAN (R 4.3.0)
 beachmat                 2.18.0      2023-10-24 [1] Bioconductor
 beeswarm                 0.4.0       2021-06-01 [1] CRAN (R 4.3.2)
 benchmarkme              1.0.8       2022-06-12 [1] CRAN (R 4.3.0)
 benchmarkmeData          1.0.4       2020-04-23 [1] CRAN (R 4.3.0)
 Biobase                  2.62.0      2023-10-24 [1] Bioconductor
 BiocFileCache            2.10.1      2023-10-26 [1] Bioconductor
 BiocGenerics           * 0.48.1      2023-11-02 [1] Bioconductor
 BiocIO                   1.12.0      2023-10-26 [1] Bioconductor
 BiocManager              1.30.22     2023-08-08 [1] CRAN (R 4.3.0)
 BiocNeighbors            1.20.2      2024-01-07 [1] Bioconductor 3.18 (R 4.3.2)
 BiocParallel             1.36.0      2023-10-26 [1] Bioconductor
 BiocSingular             1.18.0      2023-10-24 [1] Bioconductor
 BiocVersion              3.18.1      2023-11-18 [1] Bioconductor 3.18 (R 4.3.2)
 Biostrings             * 2.70.2      2024-01-28 [1] Bioconductor 3.18 (R 4.3.2)
 bit                      4.0.5       2022-11-15 [1] CRAN (R 4.3.2)
 bit64                    4.0.5       2020-08-30 [1] CRAN (R 4.3.2)
 bitops                   1.0-7       2021-04-24 [1] CRAN (R 4.3.2)
 blob                     1.2.4       2023-03-17 [1] CRAN (R 4.3.0)
 bslib                    0.6.1       2023-11-28 [1] CRAN (R 4.3.1)
 cachem                   1.0.8       2023-05-01 [1] CRAN (R 4.3.2)
 cli                      3.6.2       2023-12-11 [1] CRAN (R 4.3.2)
 codetools                0.2-19      2023-02-01 [1] CRAN (R 4.3.0)
 colorout               * 1.3-0.1     2024-01-11 [1] local
 colorspace               2.1-0       2023-01-23 [1] CRAN (R 4.3.2)
 config                   0.3.2       2023-08-30 [1] CRAN (R 4.3.0)
 cowplot                  1.1.3       2024-01-22 [1] CRAN (R 4.3.1)
 crayon                   1.5.2       2022-09-29 [1] CRAN (R 4.3.0)
 curl                     5.2.0       2023-12-08 [1] CRAN (R 4.3.2)
 data.table             * 1.14.10     2023-12-08 [1] CRAN (R 4.3.2)
 DBI                      1.2.1       2024-01-12 [1] CRAN (R 4.3.2)
 dbplyr                   2.4.0       2023-10-26 [1] CRAN (R 4.3.1)
 default                * 1.0.0       2017-08-07 [1] CRAN (R 4.3.0)
 DelayedArray             0.28.0      2023-10-24 [1] Bioconductor
 DelayedMatrixStats       1.24.0      2023-11-06 [1] Bioconductor
 digest                   0.6.34      2024-01-11 [1] CRAN (R 4.3.2)
 doParallel               1.0.17      2022-02-07 [1] CRAN (R 4.3.0)
 dotCall64                1.1-1       2023-11-28 [1] CRAN (R 4.3.2)
 dplyr                    1.1.4       2023-11-17 [1] CRAN (R 4.3.2)
 DT                       0.31        2023-12-09 [1] CRAN (R 4.3.1)
 edgeR                    4.0.14      2024-01-29 [1] Bioconductor 3.18 (R 4.3.2)
 ellipsis                 0.3.2       2021-04-29 [1] CRAN (R 4.3.2)
 evaluate                 0.23        2023-11-01 [1] CRAN (R 4.3.1)
 ExperimentHub            2.10.0      2023-10-26 [1] Bioconductor
 fansi                    1.0.6       2023-12-08 [1] CRAN (R 4.3.2)
 fastmap                  1.1.1       2023-02-24 [1] CRAN (R 4.3.2)
 fields                   15.2        2023-08-17 [1] CRAN (R 4.3.2)
 filelock                 1.0.3       2023-12-11 [1] CRAN (R 4.3.2)
 foreach                  1.5.2       2022-02-02 [1] CRAN (R 4.3.0)
 generics                 0.1.3       2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb           * 1.38.5      2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData         1.2.11      2024-01-11 [1] Bioconductor
 GenomicAlignments        1.38.2      2024-01-16 [1] Bioconductor 3.18 (R 4.3.2)
 GenomicRanges            1.54.1      2023-10-29 [1] Bioconductor
 ggbeeswarm               0.7.2       2023-04-29 [1] CRAN (R 4.3.0)
 ggplot2                  3.4.4       2023-10-12 [1] CRAN (R 4.3.1)
 ggrepel                  0.9.5       2024-01-10 [1] CRAN (R 4.3.2)
 glue                     1.7.0       2024-01-09 [1] CRAN (R 4.3.2)
 golem                    0.4.1       2023-06-05 [1] CRAN (R 4.3.0)
 gridExtra                2.3         2017-09-09 [1] CRAN (R 4.3.0)
 gtable                   0.3.4       2023-08-21 [1] CRAN (R 4.3.0)
 here                     1.0.1       2020-12-13 [1] CRAN (R 4.3.0)
 htmltools                0.5.7       2023-11-03 [1] CRAN (R 4.3.2)
 htmlwidgets              1.6.4       2023-12-06 [1] CRAN (R 4.3.1)
 httpuv                   1.6.13      2023-12-06 [1] CRAN (R 4.3.1)
 httr                     1.4.7       2023-08-15 [1] CRAN (R 4.3.0)
 interactiveDisplayBase   1.40.0      2023-10-26 [1] Bioconductor
 IRanges                * 2.36.0      2023-10-24 [1] Bioconductor
 irlba                    2.3.5.1     2022-10-03 [1] CRAN (R 4.3.2)
 iterators                1.0.14      2022-02-05 [1] CRAN (R 4.3.0)
 jquerylib                0.1.4       2021-04-26 [1] CRAN (R 4.3.0)
 jsonlite                 1.8.8       2023-12-04 [1] CRAN (R 4.3.2)
 KEGGREST                 1.42.0      2023-10-26 [1] Bioconductor
 knitr                    1.45        2023-10-30 [1] CRAN (R 4.3.2)
 later                    1.3.2       2023-12-06 [1] CRAN (R 4.3.2)
 lattice                  0.22-5      2023-10-24 [1] CRAN (R 4.3.2)
 lazyeval                 0.2.2       2019-03-15 [1] CRAN (R 4.3.2)
 lifecycle                1.0.4       2023-11-07 [1] CRAN (R 4.3.1)
 limma                    3.58.1      2023-10-31 [1] Bioconductor
 locfit                   1.5-9.8     2023-06-11 [1] CRAN (R 4.3.2)
 magick                   2.8.2       2023-12-20 [1] CRAN (R 4.3.1)
 magrittr                 2.0.3       2022-03-30 [1] CRAN (R 4.3.2)
 maps                     3.4.2       2023-12-15 [1] CRAN (R 4.3.1)
 Matrix                   1.6-5       2024-01-11 [1] CRAN (R 4.3.2)
 MatrixGenerics           1.14.0      2023-10-26 [1] Bioconductor
 matrixStats              1.2.0       2023-12-11 [1] CRAN (R 4.3.2)
 memoise                  2.0.1       2021-11-26 [1] CRAN (R 4.3.0)
 mime                     0.12        2021-09-28 [1] CRAN (R 4.3.2)
 munsell                  0.5.0       2018-06-12 [1] CRAN (R 4.3.0)
 paletteer                1.6.0       2024-01-21 [1] CRAN (R 4.3.1)
 pillar                   1.9.0       2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig                2.0.3       2019-09-22 [1] CRAN (R 4.3.0)
 plotly                   4.10.4      2024-01-13 [1] CRAN (R 4.3.1)
 png                      0.1-8       2022-11-29 [1] CRAN (R 4.3.2)
 promises                 1.2.1       2023-08-10 [1] CRAN (R 4.3.2)
 purrr                    1.0.2       2023-08-10 [1] CRAN (R 4.3.2)
 R6                       2.5.1       2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs                 0.3.3       2021-01-31 [1] CRAN (R 4.3.2)
 RColorBrewer             1.1-3       2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp                     1.0.12      2024-01-09 [1] CRAN (R 4.3.2)
 RCurl                    1.98-1.14   2024-01-09 [1] CRAN (R 4.3.1)
 rematch2                 2.1.2       2020-05-01 [1] CRAN (R 4.3.0)
 restfulr                 0.0.15      2022-06-16 [1] CRAN (R 4.3.2)
 rjson                    0.2.21      2022-01-09 [1] CRAN (R 4.3.2)
 rlang                  * 1.1.3       2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown                2.25        2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot                2.0.4       2023-11-05 [1] CRAN (R 4.3.1)
 Rsamtools                2.18.0      2023-10-24 [1] Bioconductor
 RSQLite                  2.3.5       2024-01-21 [1] CRAN (R 4.3.2)
 rstudioapi               0.15.0      2023-07-07 [1] CRAN (R 4.3.0)
 rsvd                     1.0.5       2021-04-16 [1] CRAN (R 4.3.0)
 rtracklayer              1.62.0      2023-10-26 [1] Bioconductor
 S4Arrays                 1.2.0       2023-10-24 [1] Bioconductor
 S4Vectors              * 0.40.2      2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 sass                     0.4.8       2023-12-06 [1] CRAN (R 4.3.1)
 ScaledMatrix             1.10.0      2023-11-06 [1] Bioconductor
 scales                   1.3.0       2023-11-28 [1] CRAN (R 4.3.2)
 scater                   1.30.1      2023-11-16 [1] Bioconductor
 scuttle                  1.12.0      2023-10-24 [1] Bioconductor
 sessioninfo              1.2.2       2021-12-06 [1] CRAN (R 4.3.0)
 shiny                    1.8.0       2023-11-17 [1] CRAN (R 4.3.1)
 shinyWidgets             0.8.1       2024-01-10 [1] CRAN (R 4.3.1)
 SingleCellExperiment     1.24.0      2023-11-06 [1] Bioconductor
 spam                     2.10-0      2023-10-23 [1] CRAN (R 4.3.2)
 SparseArray              1.2.3       2023-12-25 [1] Bioconductor 3.18 (R 4.3.2)
 sparseMatrixStats        1.14.0      2023-10-24 [1] Bioconductor
 SpatialExperiment        1.12.0      2023-10-26 [1] Bioconductor
 spatialLIBD              1.14.1      2023-11-30 [1] Bioconductor 3.18 (R 4.3.2)
 statmod                  1.5.0       2023-01-06 [1] CRAN (R 4.3.2)
 stringi                  1.8.3       2023-12-11 [1] CRAN (R 4.3.1)
 stringr                * 1.5.1       2023-11-14 [1] CRAN (R 4.3.1)
 SummarizedExperiment     1.32.0      2023-11-06 [1] Bioconductor
 tibble                   3.2.1       2023-03-20 [1] CRAN (R 4.3.2)
 tidyr                    1.3.1       2024-01-24 [1] CRAN (R 4.3.1)
 tidyselect               1.2.0       2022-10-10 [1] CRAN (R 4.3.0)
 utf8                     1.2.4       2023-10-22 [1] CRAN (R 4.3.2)
 vctrs                    0.6.5       2023-12-01 [1] CRAN (R 4.3.2)
 vipor                    0.4.7       2023-12-18 [1] CRAN (R 4.3.1)
 viridis                  0.6.5       2024-01-29 [1] CRAN (R 4.3.2)
 viridisLite              0.4.2       2023-05-02 [1] CRAN (R 4.3.0)
 xfun                     0.41        2023-11-01 [1] CRAN (R 4.3.2)
 XML                      3.99-0.16.1 2024-01-22 [1] CRAN (R 4.3.1)
 xtable                   1.8-4       2019-04-21 [1] CRAN (R 4.3.0)
 XVector                * 0.42.0      2023-10-24 [1] Bioconductor
 yaml                     2.3.8       2023-12-11 [1] CRAN (R 4.3.2)
 zlibbioc                 1.48.0      2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────
