---
title: "03-Compare Yao23 single-cell Sex DE to Xenium ARC, VMH types"
author: "Bernard Mulvey"
date: "2024-11-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
require(colorout)
library(orthogene)
library(biomaRt)
library(gtools)
library(BiocParallel)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)
```

load dreamlet results for Visium and Yao 23 mouse scseq
```{r}
ms <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-subclass-sexDE_dreamlet_donormouse-ranfx.txt")
msn <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-ncells-per-subclass.txt")

msst <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-VMHARCsupertypes-sexDE_dreamlet_donormouse-ranfx.txt")
msstn <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-ncells-per-VMHARCsupertype.txt")

mspb <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-VMHARCalltypescollapsed-sexDE_dreamlet_donormouse-ranfx.txt")
mspbn <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-ncells-per-VMHARCcollapsed.txt")

vde <- readRDS("processed-data/09-Sex DE/01-voomLmFit_svg10-svg20-hvg20_Hmnydflt-mnn30-HmnylmbNA_BS-15-20-31-15VMHARCclpsd.RDS")
```

## get 1-1 mapping mouse orthologs of human genes (by ensg id) and filter the two datasets accordingly
```{r}
queryg <- unique(unlist(lapply(vde,FUN=function(x){x$gene_id})))
hggenes <- as.data.table(queryg)
msgenes <- convert_orthologs(gene_df = hggenes,gene_input = "queryg",gene_output = "columns",input_species = "human",output_species = "mouse",non121_strategy = "drop_both_species")

# don't filter yet -- we need to retrieve chromosome info and drop sex chr and MT genes first
```

## drop sex chromosomal and mitochondrial genes from mouse and human results, then calculate pairwise logFCs. we also lose a few other genes here, so update the filtering of vde accordingly
```{r}
## get mouse gene chromosomes
biomart <- useMart("ensembl",dataset="mmusculus_gene_ensembl")
mschrs <- biomaRt::getBM(attributes=c("external_gene_name","chromosome_name"),mart=biomart,filter="external_gene_name",values=unique(msgenes$ortholog_gene))
rm(biomart)

## filter mouse genes to autosomal, non-mitochondrial
mschrs <- as.data.table(mschrs)
mschrs <- mschrs[!(chromosome_name %in% c("X","Y","MT"))]

## filter the ortholog table and merge it into the two datasets for cluster-pair-wise logfc correlations
msgenes <- as.data.table(msgenes)
msgenes <- msgenes[ortholog_gene %in% mschrs$external_gene_name]

## now filter the sex DE result tables
ms <- ms[ID %in% msgenes$ortholog_gene]
msst <- msst[ID %in% msgenes$ortholog_gene]
mspb <- mspb[ID %in% msgenes$ortholog_gene]

vde <- lapply(vde,FUN=function(x){x[gene_id %in% msgenes$input_gene & !(chromosome_name %in% c("chrX","chrY","chrMT"))]})

### merge the gene info so we can merge the DE results downstream for gene logFC correlation calcs
ms <- merge(ms,msgenes,by.x="ID",by.y="ortholog_gene")
msst <- merge(msst,msgenes,by.x="ID",by.y="ortholog_gene")
mspb <- merge(mspb,msgenes,by.x="ID",by.y="ortholog_gene")
vde <- lapply(vde,FUN=function(x){merge(x,msgenes,by.x="gene_id",by.y="input_gene")})
```

## for subclasses, get unique combinations of clusters, join and correlate the two male-vs-female logFCs for each
```{r}
cluspairs <- lapply(vde,FUN=function(x){
    tmp <- as.data.table(expand.grid(unique(ms$assay),unique(x$assay)))
    tmp[,Var1:=as.character(Var1)]
    tmp[,Var2:=as.character(Var2)]
    return(tmp)})

# initialize result table
sbp<-MulticoreParam(workers=8)
register(sbp)

restabs <- bpmapply(x=vde,y=cluspairs,SIMPLIFY = F, BPPARAM=sbp,FUN=function(x,y){
    
    i <- 1
    for (i in c(1:nrow(y))){
        ## result holder for iteration
        tmptab <- as.data.frame(matrix(nrow=9,ncol=5))
        
        msassay <- y[i,Var1]
        hgassay <- y[i,Var2]
        
        # filter to current clusters and merge by human ENSG id
        tmpms <- copy(ms)[assay==msassay][,.(t,input_gene)]
        setnames(tmpms,"t","tms")
        
        tmpx <- x[assay==hgassay][,.(t,gene_id)]
        # get absolute t for sorting purposes only
        tmpx[,abst:=abs(t)]
        
        tmpcompare <- merge.data.table(tmpms,tmpx,by.x="input_gene",by.y="gene_id")
        
        j <- 1
        topn <- c(25,50,75,100,250,500,750,1000,nrow(tmpcompare))
        for (j in c(1:9)){
            ## get the top n genes by abs t. DONT USE THE ABST for correl!
            curtop <- topn[j]
            setorderv(tmpcompare,"abst",-1)
            curcompare <- copy(tmpcompare)[c(1:curtop)]

            ## calculate the correlation
            tmptab[j,1] <- msassay
            tmptab[j,2] <- hgassay
            tmptab[j,3] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="pearson")
            tmptab[j,4] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="spearman")
            tmptab[j,5] <- curtop
            rm(curtop,curcompare)
        }
        
        if(i==1){
            restab <- copy(tmptab)
        }
        else{restab <- rbind(restab,tmptab)}
        rm(tmpms,tmpx,tmpcompare,mgassay,hgassay,tmptab)
    }
    restab <- as.data.table(restab)
    setnames(restab,c("mouse_subclass","human_domain","pcc","spearmancor","top_n_human"))
    return(restab)
})
names(restabs) <- names(vde)


saveRDS(restabs,"processed-data/12-Mouse comparisons/03-Yao23-subclass-MvFsexDE-tstat_correls_to_k15-20-31-15collapsed_bytopNhumanDEGs.RDS")
```


### now do the same for the VMH and ARC supertype analysis results
```{r}
cluspairs <- lapply(vde,FUN=function(x){
    tmp <- as.data.table(expand.grid(unique(msst$assay),unique(x$assay)))
    tmp[,Var1:=as.character(Var1)]
    tmp[,Var2:=as.character(Var2)]
    return(tmp)})

# initialize result table
sbp<-MulticoreParam(workers=8)
register(sbp)

restabs <- bpmapply(x=vde,y=cluspairs,SIMPLIFY = F, BPPARAM=sbp,FUN=function(x,y){
    
    i <- 1
    for (i in c(1:nrow(y))){
        ## result holder for iteration
        tmptab <- as.data.frame(matrix(nrow=9,ncol=5))
        
        msassay <- y[i,Var1]
        hgassay <- y[i,Var2]
        
        # filter to current clusters and merge by human ENSG id
        tmpms <- copy(msst)[assay==msassay][,.(t,input_gene)]
        setnames(tmpms,"t","tms")
        
        tmpx <- x[assay==hgassay][,.(t,gene_id)]
        # get absolute t for sorting purposes only
        tmpx[,abst:=abs(t)]
        
        tmpcompare <- merge.data.table(tmpms,tmpx,by.x="input_gene",by.y="gene_id")
        
        j <- 1
        topn <- c(25,50,75,100,250,500,750,1000,nrow(tmpcompare))
        for (j in c(1:9)){
            ## get the top n genes by abs t. DONT USE THE ABST for correl!
            curtop <- topn[j]
            setorderv(tmpcompare,"abst",-1)
            curcompare <- copy(tmpcompare)[c(1:curtop)]

            ## calculate the correlation
            tmptab[j,1] <- msassay
            tmptab[j,2] <- hgassay
            tmptab[j,3] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="pearson")
            tmptab[j,4] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="spearman")
            tmptab[j,5] <- curtop
            rm(curtop,curcompare)
        }
        
        if(i==1){
            restab <- copy(tmptab)
        }
        else{restab <- rbind(restab,tmptab)}
        rm(tmpms,tmpx,tmpcompare,mgassay,hgassay,tmptab)
    }
    restab <- as.data.table(restab)
    setnames(restab,c("mouse_supertype","human_domain","pcc","spearmancor","top_n_human"))
    return(restab)
})
names(restabs) <- names(vde)


saveRDS(restabs,"processed-data/12-Mouse comparisons/03-Yao23-VMHARCsupertype-MvFsexDE-tstat_correls_to_k15-20-31-15collapsed_bytopNhumanDEGs.RDS")
```


### and once more for pseudobulked mouse VMH/ARH across all subclasses
```{r}
cluspairs <- lapply(vde,FUN=function(x){
    tmp <- as.data.table(expand.grid(unique(mspb$assay),unique(x$assay)))
    tmp[,Var1:=as.character(Var1)]
    tmp[,Var2:=as.character(Var2)]
    return(tmp)})

# initialize result table
sbp<-MulticoreParam(workers=8)
register(sbp)

restabs <- bpmapply(x=vde,y=cluspairs,SIMPLIFY = F, BPPARAM=sbp,FUN=function(x,y){
    
    i <- 1
    for (i in c(1:nrow(y))){
        ## result holder for iteration
        tmptab <- as.data.frame(matrix(nrow=9,ncol=5))
        
        msassay <- y[i,Var1]
        hgassay <- y[i,Var2]
        
        # filter to current clusters and merge by human ENSG id
        tmpms <- copy(mspb)[assay==msassay][,.(t,input_gene)]
        setnames(tmpms,"t","tms")
        
        tmpx <- x[assay==hgassay][,.(t,gene_id)]
        # get absolute t for sorting purposes only
        tmpx[,abst:=abs(t)]
        
        tmpcompare <- merge.data.table(tmpms,tmpx,by.x="input_gene",by.y="gene_id")
        
        j <- 1
        topn <- c(25,50,75,100,250,500,750,1000,nrow(tmpcompare))
        for (j in c(1:9)){
            ## get the top n genes by abs t. DONT USE THE ABST for correl!
            curtop <- topn[j]
            setorderv(tmpcompare,"abst",-1)
            curcompare <- copy(tmpcompare)[c(1:curtop)]

            ## calculate the correlation
            tmptab[j,1] <- msassay
            tmptab[j,2] <- hgassay
            tmptab[j,3] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="pearson")
            tmptab[j,4] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="spearman")
            tmptab[j,5] <- curtop
            rm(curtop,curcompare)
        }
        
        if(i==1){
            restab <- copy(tmptab)
        }
        else{restab <- rbind(restab,tmptab)}
        rm(tmpms,tmpx,tmpcompare,mgassay,hgassay,tmptab)
    }
    restab <- as.data.table(restab)
    setnames(restab,c("mouse_pseudobulk_domain","human_domain","pcc","spearmancor","top_n_human"))
    return(restab)
})

saveRDS(restabs,"processed-data/12-Mouse comparisons/03-Yao23-VMHARCcollapsedSubclasses-MvFsexDE-tstat_correls_to_k15-20-31-15collapsed_bytopNhumanDEGs.RDS")
```


```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] default_1.0.0       BiocParallel_1.38.0 gtools_3.9.5        biomaRt_2.60.1     
[5] orthogene_1.10.0    colorout_1.3-0.2    data.table_1.15.4   rlang_1.1.4        

loaded via a namespace (and not attached):
  [1] bitops_1.0-7              DBI_1.2.3                 grr_0.9.5                
  [4] httr2_1.0.1               magrittr_2.0.3            compiler_4.4.1           
  [7] RSQLite_2.3.7             png_0.1-8                 vctrs_0.6.5              
 [10] gprofiler2_0.2.3          stringr_1.5.1             pkgconfig_2.0.3          
 [13] crayon_1.5.3              fastmap_1.2.0             backports_1.5.0          
 [16] dbplyr_2.5.0              XVector_0.44.0            utf8_1.2.4               
 [19] UCSC.utils_1.0.0          purrr_1.0.2               bit_4.0.5                
 [22] xfun_0.45                 zlibbioc_1.50.0           cachem_1.1.0             
 [25] aplot_0.2.3               GenomeInfoDb_1.40.1       jsonlite_1.8.8           
 [28] progress_1.2.3            blob_1.2.4                broom_1.0.6              
 [31] parallel_4.4.1            prettyunits_1.2.0         R6_2.5.1                 
 [34] stringi_1.8.4             car_3.1-2                 Rcpp_1.0.13              
 [37] knitr_1.48                IRanges_2.38.1            Matrix_1.7-0             
 [40] tidyselect_1.2.1          rstudioapi_0.16.0         abind_1.4-5              
 [43] codetools_0.2-20          curl_5.2.1                lattice_0.22-6           
 [46] tibble_3.2.1              withr_3.0.0               Biobase_2.64.0           
 [49] treeio_1.28.0             KEGGREST_1.44.1           gridGraphics_0.5-1       
 [52] BiocFileCache_2.12.0      xml2_1.3.6                Biostrings_2.72.1        
 [55] filelock_1.0.3            pillar_1.9.0              ggtree_3.12.0            
 [58] ggpubr_0.6.0              carData_3.0-5             stats4_4.4.1             
 [61] ggfun_0.1.5               plotly_4.10.4             generics_0.1.3           
 [64] RCurl_1.98-1.14           rprojroot_2.0.4           S4Vectors_0.42.1         
 [67] hms_1.1.3                 ggplot2_3.5.1             munsell_0.5.1            
 [70] scales_1.3.0              tidytree_0.4.6            glue_1.7.0               
 [73] lazyeval_0.2.2            tools_4.4.1               ggsignif_0.6.4           
 [76] babelgene_22.9            fs_1.6.4                  grid_4.4.1               
 [79] tidyr_1.3.1               ape_5.8                   AnnotationDbi_1.66.0     
 [82] colorspace_2.1-0          nlme_3.1-165              GenomeInfoDbData_1.2.12  
 [85] patchwork_1.2.0           homologene_1.4.68.19.3.27 cli_3.6.3                
 [88] rappdirs_0.3.3            fansi_1.0.6               viridisLite_0.4.2        
 [91] dplyr_1.1.4               gtable_0.3.5              rstatix_0.7.2            
 [94] yulab.utils_0.1.4         digest_0.6.36             BiocGenerics_0.50.0      
 [97] ggplotify_0.1.2           htmlwidgets_1.6.4         memoise_2.0.1            
[100] htmltools_0.5.8.1         lifecycle_1.0.4           httr_1.4.7               
[103] here_1.0.1                bit64_4.0.5              
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-11-10
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   NA

─ Packages ───────────────────────────────────────────────────────────────────────
 ! package          * version        date (UTC) lib source
   abind              1.4-5          2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi      1.66.0         2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   ape                5.8            2024-04-11 [1] CRAN (R 4.4.0)
   aplot              0.2.3          2024-06-17 [1] CRAN (R 4.4.0)
   babelgene          22.9           2022-09-29 [1] CRAN (R 4.4.0)
   backports          1.5.0          2024-05-23 [1] CRAN (R 4.4.0)
   Biobase            2.64.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache      2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics       0.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel     * 1.38.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   biomaRt          * 2.60.1         2024-06-26 [1] Bioconductor 3.19 (R 4.4.1)
   Biostrings         2.72.1         2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                4.0.5          2022-11-15 [1] CRAN (R 4.4.0)
   bit64              4.0.5          2020-08-30 [1] CRAN (R 4.4.0)
   bitops             1.0-7          2021-04-24 [1] CRAN (R 4.4.0)
   blob               1.2.4          2023-03-17 [1] CRAN (R 4.4.0)
   broom              1.0.6          2024-05-17 [1] CRAN (R 4.4.0)
   cachem             1.1.0          2024-05-16 [1] CRAN (R 4.4.0)
   car                3.1-2          2023-03-30 [1] CRAN (R 4.4.0)
   carData            3.0-5          2022-01-06 [1] CRAN (R 4.4.0)
 P cli                3.6.3          2024-06-21 [2] CRAN (R 4.4.0)
   codetools          0.2-20         2024-03-31 [1] CRAN (R 4.4.1)
   colorout         * 1.3-0.2        2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace         2.1-0          2023-01-23 [1] CRAN (R 4.4.0)
   crayon             1.5.3          2024-06-20 [1] CRAN (R 4.4.0)
   curl               5.2.1          2024-03-01 [1] CRAN (R 4.4.0)
   data.table       * 1.15.4         2024-03-30 [2] CRAN (R 4.4.0)
   DBI                1.2.3          2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr             2.5.0          2024-03-19 [1] CRAN (R 4.4.0)
   default          * 1.0.0          2017-08-07 [1] CRAN (R 4.4.0)
   digest             0.6.36         2024-06-23 [1] CRAN (R 4.4.0)
   dplyr              1.1.4          2023-11-17 [1] CRAN (R 4.4.0)
   fansi              1.0.6          2023-12-08 [1] CRAN (R 4.4.0)
   fastmap            1.2.0          2024-05-15 [1] CRAN (R 4.4.0)
   filelock           1.0.3          2023-12-11 [1] CRAN (R 4.4.0)
   fs                 1.6.4          2024-04-25 [1] CRAN (R 4.4.0)
   generics           0.1.3          2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb       1.40.1         2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData   1.2.12         2024-05-01 [1] Bioconductor
   ggfun              0.1.5          2024-05-28 [1] CRAN (R 4.4.0)
   ggplot2            3.5.1          2024-04-23 [1] CRAN (R 4.4.0)
   ggplotify          0.1.2          2023-08-09 [1] CRAN (R 4.4.0)
   ggpubr             0.6.0          2023-02-10 [1] CRAN (R 4.4.0)
   ggsignif           0.6.4          2022-10-13 [1] CRAN (R 4.4.0)
   ggtree             3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   glue               1.7.0          2024-01-09 [1] CRAN (R 4.4.0)
   gprofiler2         0.2.3          2024-02-23 [1] CRAN (R 4.4.0)
   gridGraphics       0.5-1          2020-12-13 [1] CRAN (R 4.4.0)
   grr                0.9.5          2016-08-26 [1] CRAN (R 4.4.0)
   gtable             0.3.5          2024-04-22 [1] CRAN (R 4.4.0)
   gtools           * 3.9.5          2023-11-20 [1] CRAN (R 4.4.0)
   here               1.0.1          2020-12-13 [1] CRAN (R 4.4.0)
   hms                1.1.3          2023-03-21 [1] CRAN (R 4.4.0)
   homologene         1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.4.0)
   htmltools          0.5.8.1        2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets        1.6.4          2023-12-06 [1] CRAN (R 4.4.0)
   httr               1.4.7          2023-08-15 [1] CRAN (R 4.4.0)
   httr2              1.0.1          2024-04-01 [1] CRAN (R 4.4.0)
   IRanges            2.38.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   jsonlite           1.8.8          2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST           1.44.1         2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   knitr              1.48           2024-07-07 [1] CRAN (R 4.4.1)
   lattice            0.22-6         2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval           0.2.2          2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle          1.0.4          2023-11-07 [1] CRAN (R 4.4.0)
   magrittr           2.0.3          2022-03-30 [1] CRAN (R 4.4.0)
   Matrix             1.7-0          2024-04-26 [1] CRAN (R 4.4.1)
   memoise            2.0.1          2021-11-26 [1] CRAN (R 4.4.0)
   munsell            0.5.1          2024-04-01 [1] CRAN (R 4.4.0)
   nlme               3.1-165        2024-06-06 [1] CRAN (R 4.4.0)
   orthogene        * 1.10.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   patchwork          1.2.0          2024-01-08 [1] CRAN (R 4.4.0)
   pillar             1.9.0          2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig          2.0.3          2019-09-22 [1] CRAN (R 4.4.0)
   plotly             4.10.4         2024-01-13 [1] CRAN (R 4.4.0)
   png                0.1-8          2022-11-29 [1] CRAN (R 4.4.0)
   prettyunits        1.2.0          2023-09-24 [1] CRAN (R 4.4.0)
   progress           1.2.3          2023-12-06 [1] CRAN (R 4.4.0)
   purrr              1.0.2          2023-08-10 [1] CRAN (R 4.4.0)
   R6                 2.5.1          2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs           0.3.3          2021-01-31 [1] CRAN (R 4.4.0)
   Rcpp               1.0.13         2024-07-17 [1] CRAN (R 4.4.0)
   RCurl              1.98-1.14      2024-01-09 [1] CRAN (R 4.4.0)
 P rlang            * 1.1.4          2024-06-04 [2] CRAN (R 4.4.1)
   rprojroot          2.0.4          2023-11-05 [1] CRAN (R 4.4.0)
   RSQLite            2.3.7          2024-05-27 [1] CRAN (R 4.4.0)
   rstatix            0.7.2          2023-02-01 [1] CRAN (R 4.4.0)
   rstudioapi         0.16.0         2024-03-24 [1] CRAN (R 4.4.0)
   S4Vectors          0.42.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   scales             1.3.0          2023-11-28 [1] CRAN (R 4.4.0)
   sessioninfo        1.2.2          2021-12-06 [1] CRAN (R 4.4.0)
   stringi            1.8.4          2024-05-06 [1] CRAN (R 4.4.0)
   stringr            1.5.1          2023-11-14 [1] CRAN (R 4.4.0)
   tibble             3.2.1          2023-03-20 [1] CRAN (R 4.4.0)
   tidyr              1.3.1          2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect         1.2.1          2024-03-11 [1] CRAN (R 4.4.0)
   tidytree           0.4.6          2023-12-12 [1] CRAN (R 4.4.0)
   treeio             1.28.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   UCSC.utils         1.0.0          2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   utf8               1.2.4          2023-10-22 [1] CRAN (R 4.4.0)
   vctrs              0.6.5          2023-12-01 [1] CRAN (R 4.4.0)
   viridisLite        0.4.2          2023-05-02 [1] CRAN (R 4.4.0)
   withr              3.0.0          2024-01-16 [1] CRAN (R 4.4.0)
   xfun               0.45           2024-06-16 [1] CRAN (R 4.4.0)
   xml2               1.3.6          2023-12-04 [1] CRAN (R 4.4.0)
   XVector            0.44.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yulab.utils        0.1.4          2024-01-28 [1] CRAN (R 4.4.0)
   zlibbioc           1.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────────
