---
title: "03-Compare Yao23 Subclass Sex DE to Visium"
author: "Bernard Mulvey"
date: "2024-08-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
require(colorout)
library(orthogene)
library(biomaRt)
library(gtools)
library(BiocParallel)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))
```

load dreamlet results for Visium and Yao 23 mouse scseq
```{r}
ms <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-subclass-sexDE_dreamlet_donormouse-ranfx.txt")
msn <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-ncells-per-subclass.txt")

msst <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-VMHARCsupertypes-sexDE_dreamlet_donormouse-ranfx.txt")
msstn <- fread("processed-data/12-Mouse comparisons/02-Yao23-HYP-ncells-per-VMHARCsupertype.txt")


vde <- readRDS("processed-data/09-Sex DE/01-voomLmFit_svg10-svg20-hvg20_Hmnydflt-mnn30-HmnylmbNA_BS-15-20-31-15VMHARCclpsd.RDS")
```

## get 1-1 mapping mouse orthologs of human genes (by ensg id) and filter the two datasets accordingly
```{r}
queryg <- unique(unlist(lapply(vde,FUN=function(x){x$gene_id})))
hggenes <- as.data.table(queryg)
msgenes <- convert_orthologs(gene_df = hggenes,gene_input = "queryg",gene_output = "columns",input_species = "human",output_species = "mouse",non121_strategy = "drop_both_species")

# don't filter yet -- we need to retrieve chromosome info and drop sex chr and MT genes first
```

## drop sex chromosomal and mitochondrial genes from mouse and human results, then calculate pairwise logFCs. we also lose a few other genes here, so update the filtering of vde accordingly
```{r}
## get mouse gene chromosomes
biomart <- useMart("ensembl",dataset="mmusculus_gene_ensembl")
mschrs <- biomaRt::getBM(attributes=c("external_gene_name","chromosome_name"),mart=biomart,filter="external_gene_name",values=unique(msgenes$ortholog_gene))
rm(biomart)

## filter mouse genes to autosomal, non-mitochondrial
mschrs <- as.data.table(mschrs)
mschrs <- mschrs[!(chromosome_name %in% c("X","Y","MT"))]

## filter the ortholog table and merge it into the two datasets for cluster-pair-wise logfc correlations
msgenes <- as.data.table(msgenes)
msgenes <- msgenes[ortholog_gene %in% mschrs$external_gene_name]

## now filter the sex DE result tables
ms <- ms[ID %in% msgenes$ortholog_gene]
msst <- msst[ID %in% msgenes$ortholog_gene]
vde <- lapply(vde,FUN=function(x){x[gene_id %in% msgenes$input_gene & !(chromosome_name %in% c("chrX","chrY","chrMT"))]})

### merge the gene info so we can merge the DE results downstream for gene logFC correlation calcs
ms <- merge(ms,msgenes,by.x="ID",by.y="ortholog_gene")
msst <- merge(msst,msgenes,by.x="ID",by.y="ortholog_gene")
vde <- lapply(vde,FUN=function(x){merge(x,msgenes,by.x="gene_id",by.y="input_gene")})
```

## for subclasses, get unique combinations of clusters, join and correlate the two male-vs-female logFCs for each
```{r}
cluspairs <- lapply(vde,FUN=function(x){
    tmp <- as.data.table(expand.grid(unique(ms$assay),unique(x$assay)))
    tmp[,Var1:=as.character(Var1)]
    tmp[,Var2:=as.character(Var2)]
    return(tmp)})

# initialize result table
sbp<-MulticoreParam(workers=8)
register(sbp)

restabs <- bpmapply(x=vde,y=cluspairs,SIMPLIFY = F, BPPARAM=sbp,FUN=function(x,y){
    
    i <- 1
    for (i in c(1:nrow(y))){
        ## result holder for iteration
        tmptab <- as.data.frame(matrix(nrow=9,ncol=5))
        
        msassay <- y[i,Var1]
        hgassay <- y[i,Var2]
        
        # filter to current clusters and merge by human ENSG id
        tmpms <- copy(ms)[assay==msassay][,.(t,input_gene)]
        setnames(tmpms,"t","tms")
        
        tmpx <- x[assay==hgassay][,.(t,gene_id)]
        # get absolute t for sorting purposes only
        tmpx[,abst:=abs(t)]
        
        tmpcompare <- merge.data.table(tmpms,tmpx,by.x="input_gene",by.y="gene_id")
        
        j <- 1
        topn <- c(25,50,75,100,250,500,750,1000,nrow(tmpcompare))
        for (j in c(1:9)){
            ## get the top n genes by abs t. DONT USE THE ABST for correl!
            curtop <- topn[j]
            setorderv(tmpcompare,"abst",-1)
            curcompare <- copy(tmpcompare)[c(1:curtop)]

            ## calculate the correlation
            tmptab[j,1] <- msassay
            tmptab[j,2] <- hgassay
            tmptab[j,3] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="pearson")
            tmptab[j,4] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="spearman")
            tmptab[j,5] <- curtop
            rm(curtop,curcompare)
        }
        
        if(i==1){
            restab <- copy(tmptab)
        }
        else{restab <- rbind(restab,tmptab)}
        rm(tmpms,tmpx,tmpcompare,mgassay,hgassay,tmptab)
    }
    restab <- as.data.table(restab)
    setnames(restab,c("mouse_subclass","human_domain","pcc","spearmancor","top_n_human"))
    return(restab)
})
names(restabs) <- names(vde)


saveRDS(restabs,"processed-data/12-Mouse comparisons/03-Yao23-subclass-MvFsexDE-tstat_correls_to_k15-20-31-15collapsed_bytopNhumanDEGs.RDS")
```


### now do the same for the VMH and ARC supertype analysis results
```{r}
cluspairs <- lapply(vde,FUN=function(x){
    tmp <- as.data.table(expand.grid(unique(msst$assay),unique(x$assay)))
    tmp[,Var1:=as.character(Var1)]
    tmp[,Var2:=as.character(Var2)]
    return(tmp)})

# initialize result table
sbp<-MulticoreParam(workers=8)
register(sbp)

restabs <- bpmapply(x=vde,y=cluspairs,SIMPLIFY = F, BPPARAM=sbp,FUN=function(x,y){
    
    i <- 1
    for (i in c(1:nrow(y))){
        ## result holder for iteration
        tmptab <- as.data.frame(matrix(nrow=9,ncol=5))
        
        msassay <- y[i,Var1]
        hgassay <- y[i,Var2]
        
        # filter to current clusters and merge by human ENSG id
        tmpms <- copy(msst)[assay==msassay][,.(t,input_gene)]
        setnames(tmpms,"t","tms")
        
        tmpx <- x[assay==hgassay][,.(t,gene_id)]
        # get absolute t for sorting purposes only
        tmpx[,abst:=abs(t)]
        
        tmpcompare <- merge.data.table(tmpms,tmpx,by.x="input_gene",by.y="gene_id")
        
        j <- 1
        topn <- c(25,50,75,100,250,500,750,1000,nrow(tmpcompare))
        for (j in c(1:9)){
            ## get the top n genes by abs t. DONT USE THE ABST for correl!
            curtop <- topn[j]
            setorderv(tmpcompare,"abst",-1)
            curcompare <- copy(tmpcompare)[c(1:curtop)]

            ## calculate the correlation
            tmptab[j,1] <- msassay
            tmptab[j,2] <- hgassay
            tmptab[j,3] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="pearson")
            tmptab[j,4] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="spearman")
            tmptab[j,5] <- curtop
            rm(curtop,curcompare)
        }
        
        if(i==1){
            restab <- copy(tmptab)
        }
        else{restab <- rbind(restab,tmptab)}
        rm(tmpms,tmpx,tmpcompare,mgassay,hgassay,tmptab)
    }
    restab <- as.data.table(restab)
    setnames(restab,c("mouse_supertype","human_domain","pcc","spearmancor","top_n_human"))
    return(restab)
})
names(restabs) <- names(vde)


saveRDS(restabs,"processed-data/12-Mouse comparisons/03-Yao23-VMHARCsupertype-MvFsexDE-tstat_correls_to_k15-20-31-15collapsed_bytopNhumanDEGs.RDS")
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] default_1.0.0               gtools_3.9.5               
 [3] biomaRt_2.60.1              orthogene_1.10.0           
 [5] dreamlet_1.2.1              variancePartition_1.34.0   
 [7] BiocParallel_1.38.0         limma_3.60.3               
 [9] colorout_1.3-0.2            spatialLIBD_1.16.2         
[11] SpatialExperiment_1.14.0    SingleCellExperiment_1.26.0
[13] SummarizedExperiment_1.34.0 Biobase_2.64.0             
[15] GenomicRanges_1.56.1        MatrixGenerics_1.16.0      
[17] matrixStats_1.3.0           gridExtra_2.3              
[19] ggplot2_3.5.1               Biostrings_2.72.1          
[21] GenomeInfoDb_1.40.1         XVector_0.44.0             
[23] IRanges_2.38.1              S4Vectors_0.42.1           
[25] BiocGenerics_0.50.0         data.table_1.15.4          
[27] rlang_1.1.4                

loaded via a namespace (and not attached):
  [1] fs_1.6.4                  bitops_1.0-7              httr_1.4.7               
  [4] RColorBrewer_1.1-3        doParallel_1.0.17         Rgraphviz_2.48.0         
  [7] numDeriv_2016.8-1.1       tools_4.4.1               backports_1.5.0          
 [10] metafor_4.6-0             utf8_1.2.4                R6_2.5.1                 
 [13] DT_0.33                   lazyeval_0.2.2            withr_3.0.0              
 [16] prettyunits_1.2.0         cli_3.6.3                 sass_0.4.9               
 [19] KEGGgraph_1.64.0          SQUAREM_2021.1            mvtnorm_1.2-5            
 [22] mixsqp_0.3-54             yulab.utils_0.1.4         Rsamtools_2.20.0         
 [25] zenith_1.6.0              scater_1.32.0             sessioninfo_1.2.2        
 [28] attempt_0.3.1             invgamma_1.1              maps_3.4.2               
 [31] rstudioapi_0.16.0         RSQLite_2.3.7             gridGraphics_0.5-1       
 [34] generics_0.1.3            BiocIO_1.14.0             car_3.1-2                
 [37] dplyr_1.1.4               homologene_1.4.68.19.3.27 metadat_1.2-0            
 [40] Matrix_1.7-0              ggbeeswarm_0.7.2          fansi_1.0.6              
 [43] abind_1.4-5               lifecycle_1.0.4           yaml_2.3.9               
 [46] edgeR_4.2.0               carData_3.0-5             mathjaxr_1.6-0           
 [49] gplots_3.1.3.1            SparseArray_1.4.8         BiocFileCache_2.12.0     
 [52] paletteer_1.6.0           grid_4.4.1                blob_1.2.4               
 [55] promises_1.3.0            ExperimentHub_2.12.0      crayon_1.5.3             
 [58] lattice_0.22-6            msigdbr_7.5.1             beachmat_2.20.0          
 [61] cowplot_1.1.3             annotate_1.82.0           KEGGREST_1.44.1          
 [64] magick_2.8.3              pillar_1.9.0              knitr_1.48               
 [67] rjson_0.2.21              boot_1.3-30               corpcor_1.6.10           
 [70] codetools_0.2-20          glue_1.7.0                ggfun_0.1.5              
 [73] treeio_1.28.0             vctrs_0.6.5               png_0.1-8                
 [76] spam_2.10-0               Rdpack_2.6                gtable_0.3.5             
 [79] assertthat_0.2.1          rematch2_2.1.2            cachem_1.1.0             
 [82] xfun_0.45                 rbibutils_2.2.16          S4Arrays_1.4.1           
 [85] mime_0.12                 Rfast_2.1.0               iterators_1.0.14         
 [88] fields_16.2               statmod_1.5.0             nlme_3.1-165             
 [91] pbkrtest_0.5.3            ggtree_3.12.0             bit64_4.0.5              
 [94] progress_1.2.3            EnvStats_2.8.1            filelock_1.0.3           
 [97] rprojroot_2.0.4           bslib_0.7.0               irlba_2.3.5.1            
[100] vipor_0.4.7               KernSmooth_2.23-24        rmeta_3.0                
[103] colorspace_2.1-0          DBI_1.2.3                 tidyselect_1.2.1         
[106] bit_4.0.5                 compiler_4.4.1            curl_5.2.1               
[109] httr2_1.0.1               graph_1.82.0              BiocNeighbors_1.22.0     
[112] xml2_1.3.6                DelayedArray_0.30.1       plotly_4.10.4            
[115] rtracklayer_1.64.0        scales_1.3.0              caTools_1.18.2           
[118] remaCor_0.0.18            rappdirs_0.3.3            stringr_1.5.1            
[121] digest_0.6.36             minqa_1.2.7               rmarkdown_2.27           
[124] benchmarkmeData_1.0.4     aod_1.3.3                 RhpcBLASctl_0.23-42      
[127] htmltools_0.5.8.1         pkgconfig_2.0.3           lme4_1.1-35.5            
[130] sparseMatrixStats_1.16.0  mashr_0.2.79              dbplyr_2.5.0             
[133] fastmap_1.2.0             htmlwidgets_1.6.4         UCSC.utils_1.0.0         
[136] shiny_1.8.1.1             DelayedMatrixStats_1.26.0 jquerylib_0.1.4          
[139] jsonlite_1.8.8            config_0.3.2              BiocSingular_1.20.0      
[142] RCurl_1.98-1.14           magrittr_2.0.3            ggplotify_0.1.2          
[145] scuttle_1.14.0            GenomeInfoDbData_1.2.12   dotCall64_1.1-1          
[148] patchwork_1.2.0           munsell_0.5.1             Rcpp_1.0.13              
[151] ape_5.8                   babelgene_22.9            viridis_0.6.5            
[154] RcppZiggurat_0.1.6        EnrichmentBrowser_2.34.1  stringi_1.8.4            
[157] zlibbioc_1.50.0           MASS_7.3-61               AnnotationHub_3.12.0     
[160] plyr_1.8.9                parallel_4.4.1            ggrepel_0.9.5            
[163] splines_4.4.1             hms_1.1.3                 locfit_1.5-9.10          
[166] ggpubr_0.6.0              ggsignif_0.6.4            reshape2_1.4.4           
[169] ScaledMatrix_1.12.0       gprofiler2_0.2.3          BiocVersion_3.19.1       
[172] XML_3.99-0.17             evaluate_0.24.0           golem_0.4.1              
[175] RcppParallel_5.1.8        BiocManager_1.30.23       nloptr_2.1.1             
[178] foreach_1.5.2             httpuv_1.6.15             grr_0.9.5                
[181] tidyr_1.3.1               purrr_1.0.2               scattermore_1.2          
[184] benchmarkme_1.0.8         ashr_2.2-63               rsvd_1.0.5               
[187] broom_1.0.6               xtable_1.8-4              restfulr_0.0.15          
[190] tidytree_0.4.6            fANCOVA_0.6-1             rstatix_0.7.2            
[193] later_1.3.2               viridisLite_0.4.2         truncnorm_1.0-9          
[196] tibble_3.2.1              aplot_0.2.3               lmerTest_3.1-3           
[199] memoise_2.0.1             beeswarm_0.4.0            AnnotationDbi_1.66.0     
[202] GenomicAlignments_1.40.0  shinyWidgets_0.8.6        GSEABase_1.66.0          
[205] here_1.0.1               
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-08-18
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────────────
 ! package              * version        date (UTC) lib source
   abind                  1.4-5          2016-07-21 [1] CRAN (R 4.4.0)
   annotate               1.82.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationDbi          1.66.0         2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub          3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   aod                    1.3.3          2023-12-13 [1] CRAN (R 4.4.0)
   ape                    5.8            2024-04-11 [1] CRAN (R 4.4.0)
   aplot                  0.2.3          2024-06-17 [1] CRAN (R 4.4.0)
   ashr                   2.2-63         2023-08-21 [1] CRAN (R 4.4.0)
   assertthat             0.2.1          2019-03-21 [1] CRAN (R 4.4.0)
   attempt                0.3.1          2020-05-03 [1] CRAN (R 4.4.0)
   babelgene              22.9           2022-09-29 [1] CRAN (R 4.4.0)
   backports              1.5.0          2024-05-23 [1] CRAN (R 4.4.0)
   beachmat               2.20.0         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm               0.4.0          2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme            1.0.8          2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData        1.0.4          2020-04-23 [1] CRAN (R 4.4.0)
   Biobase              * 2.64.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache          2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics         * 0.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                 1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager            1.30.23        2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors          1.22.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel         * 1.38.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocSingular           1.20.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion            3.19.1         2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   biomaRt              * 2.60.1         2024-06-26 [1] Bioconductor 3.19 (R 4.4.1)
   Biostrings           * 2.72.1         2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                    4.0.5          2022-11-15 [1] CRAN (R 4.4.0)
   bit64                  4.0.5          2020-08-30 [1] CRAN (R 4.4.0)
   bitops                 1.0-7          2021-04-24 [1] CRAN (R 4.4.0)
   blob                   1.2.4          2023-03-17 [1] CRAN (R 4.4.0)
   boot                   1.3-30         2024-02-26 [1] CRAN (R 4.4.1)
   broom                  1.0.6          2024-05-17 [1] CRAN (R 4.4.0)
   bslib                  0.7.0          2024-03-29 [1] CRAN (R 4.4.0)
   cachem                 1.1.0          2024-05-16 [1] CRAN (R 4.4.0)
   car                    3.1-2          2023-03-30 [1] CRAN (R 4.4.0)
   carData                3.0-5          2022-01-06 [1] CRAN (R 4.4.0)
   caTools                1.18.2         2021-03-28 [1] CRAN (R 4.4.0)
 P cli                    3.6.3          2024-06-21 [2] CRAN (R 4.4.0)
   codetools              0.2-20         2024-03-31 [1] CRAN (R 4.4.1)
   colorout             * 1.3-0.2        2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace             2.1-0          2023-01-23 [1] CRAN (R 4.4.0)
   config                 0.3.2          2023-08-30 [1] CRAN (R 4.4.0)
   corpcor                1.6.10         2021-09-16 [1] CRAN (R 4.4.0)
   cowplot                1.1.3          2024-01-22 [1] CRAN (R 4.4.0)
   crayon                 1.5.3          2024-06-20 [1] CRAN (R 4.4.0)
   curl                   5.2.1          2024-03-01 [1] CRAN (R 4.4.0)
   data.table           * 1.15.4         2024-03-30 [2] CRAN (R 4.4.0)
   DBI                    1.2.3          2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                 2.5.0          2024-03-19 [1] CRAN (R 4.4.0)
   default              * 1.0.0          2017-08-07 [1] CRAN (R 4.4.0)
   DelayedArray           0.30.1         2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats     1.26.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   digest                 0.6.36         2024-06-23 [1] CRAN (R 4.4.0)
   doParallel             1.0.17         2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64              1.1-1          2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                  1.1.4          2023-11-17 [1] CRAN (R 4.4.0)
   dreamlet             * 1.2.1          2024-06-09 [1] Bioconductor 3.19 (R 4.4.0)
   DT                     0.33           2024-04-04 [1] CRAN (R 4.4.0)
   edgeR                  4.2.0          2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   EnrichmentBrowser      2.34.1         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   EnvStats               2.8.1          2023-08-22 [1] CRAN (R 4.4.0)
   evaluate               0.24.0         2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub          2.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fANCOVA                0.6-1          2020-11-13 [1] CRAN (R 4.4.0)
   fansi                  1.0.6          2023-12-08 [1] CRAN (R 4.4.0)
   fastmap                1.2.0          2024-05-15 [1] CRAN (R 4.4.0)
   fields                 16.2           2024-06-27 [1] CRAN (R 4.4.0)
   filelock               1.0.3          2023-12-11 [1] CRAN (R 4.4.0)
   foreach                1.5.2          2022-02-02 [1] CRAN (R 4.4.0)
   fs                     1.6.4          2024-04-25 [1] CRAN (R 4.4.0)
   generics               0.1.3          2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb         * 1.40.1         2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData       1.2.12         2024-05-01 [1] Bioconductor
   GenomicAlignments      1.40.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges        * 1.56.1         2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm             0.7.2          2023-04-29 [1] CRAN (R 4.4.0)
   ggfun                  0.1.5          2024-05-28 [1] CRAN (R 4.4.0)
   ggplot2              * 3.5.1          2024-04-23 [1] CRAN (R 4.4.0)
   ggplotify              0.1.2          2023-08-09 [1] CRAN (R 4.4.0)
   ggpubr                 0.6.0          2023-02-10 [1] CRAN (R 4.4.0)
   ggrepel                0.9.5          2024-01-10 [1] CRAN (R 4.4.0)
   ggsignif               0.6.4          2022-10-13 [1] CRAN (R 4.4.0)
   ggtree                 3.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   glue                   1.7.0          2024-01-09 [1] CRAN (R 4.4.0)
   golem                  0.4.1          2023-06-05 [1] CRAN (R 4.4.0)
   gplots                 3.1.3.1        2024-02-02 [1] CRAN (R 4.4.0)
   gprofiler2             0.2.3          2024-02-23 [1] CRAN (R 4.4.0)
   graph                  1.82.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   gridExtra            * 2.3            2017-09-09 [1] CRAN (R 4.4.0)
   gridGraphics           0.5-1          2020-12-13 [1] CRAN (R 4.4.0)
   grr                    0.9.5          2016-08-26 [1] CRAN (R 4.4.0)
   GSEABase               1.66.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   gtable                 0.3.5          2024-04-22 [1] CRAN (R 4.4.0)
   gtools               * 3.9.5          2023-11-20 [1] CRAN (R 4.4.0)
   here                   1.0.1          2020-12-13 [1] CRAN (R 4.4.0)
   hms                    1.1.3          2023-03-21 [1] CRAN (R 4.4.0)
   homologene             1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.4.0)
   htmltools              0.5.8.1        2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets            1.6.4          2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                 1.6.15         2024-03-26 [1] CRAN (R 4.4.0)
   httr                   1.4.7          2023-08-15 [1] CRAN (R 4.4.0)
   httr2                  1.0.1          2024-04-01 [1] CRAN (R 4.4.0)
   invgamma               1.1            2017-05-07 [1] CRAN (R 4.4.0)
   IRanges              * 2.38.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                  2.3.5.1        2022-10-03 [1] CRAN (R 4.4.0)
   iterators              1.0.14         2022-02-05 [1] CRAN (R 4.4.0)
   jquerylib              0.1.4          2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite               1.8.8          2023-12-04 [1] CRAN (R 4.4.0)
   KEGGgraph              1.64.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   KEGGREST               1.44.1         2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   KernSmooth             2.23-24        2024-05-17 [1] CRAN (R 4.4.1)
   knitr                  1.48           2024-07-07 [1] CRAN (R 4.4.1)
   later                  1.3.2          2023-12-06 [1] CRAN (R 4.4.0)
   lattice                0.22-6         2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval               0.2.2          2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle              1.0.4          2023-11-07 [1] CRAN (R 4.4.0)
   limma                * 3.60.3         2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   lme4                   1.1-35.5       2024-07-03 [1] CRAN (R 4.4.1)
   lmerTest               3.1-3          2020-10-23 [1] CRAN (R 4.4.0)
   locfit                 1.5-9.10       2024-06-24 [1] CRAN (R 4.4.0)
   magick                 2.8.3          2024-02-18 [1] CRAN (R 4.4.0)
   magrittr               2.0.3          2022-03-30 [1] CRAN (R 4.4.0)
   maps                   3.4.2          2023-12-15 [1] CRAN (R 4.4.0)
   mashr                  0.2.79         2023-10-18 [1] CRAN (R 4.4.0)
   MASS                   7.3-61         2024-06-13 [1] CRAN (R 4.4.0)
   mathjaxr               1.6-0          2022-02-28 [1] CRAN (R 4.4.0)
   Matrix                 1.7-0          2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics       * 1.16.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats          * 1.3.0          2024-04-11 [1] CRAN (R 4.4.0)
   memoise                2.0.1          2021-11-26 [1] CRAN (R 4.4.0)
   metadat                1.2-0          2022-04-06 [1] CRAN (R 4.4.0)
   metafor                4.6-0          2024-03-28 [1] CRAN (R 4.4.0)
   mime                   0.12           2021-09-28 [1] CRAN (R 4.4.0)
   minqa                  1.2.7          2024-05-20 [1] CRAN (R 4.4.0)
   mixsqp                 0.3-54         2023-12-20 [1] CRAN (R 4.4.0)
   msigdbr                7.5.1          2022-03-30 [1] CRAN (R 4.4.0)
   munsell                0.5.1          2024-04-01 [1] CRAN (R 4.4.0)
   mvtnorm                1.2-5          2024-05-21 [1] CRAN (R 4.4.0)
   nlme                   3.1-165        2024-06-06 [1] CRAN (R 4.4.0)
   nloptr                 2.1.1          2024-06-25 [1] CRAN (R 4.4.0)
   numDeriv               2016.8-1.1     2019-06-06 [1] CRAN (R 4.4.0)
   orthogene            * 1.10.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   paletteer              1.6.0          2024-01-21 [1] CRAN (R 4.4.0)
   patchwork              1.2.0          2024-01-08 [1] CRAN (R 4.4.0)
   pbkrtest               0.5.3          2024-06-26 [1] CRAN (R 4.4.0)
   pillar                 1.9.0          2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig              2.0.3          2019-09-22 [1] CRAN (R 4.4.0)
   plotly                 4.10.4         2024-01-13 [1] CRAN (R 4.4.0)
   plyr                   1.8.9          2023-10-02 [1] CRAN (R 4.4.0)
   png                    0.1-8          2022-11-29 [1] CRAN (R 4.4.0)
   prettyunits            1.2.0          2023-09-24 [1] CRAN (R 4.4.0)
   progress               1.2.3          2023-12-06 [1] CRAN (R 4.4.0)
   promises               1.3.0          2024-04-05 [1] CRAN (R 4.4.0)
   purrr                  1.0.2          2023-08-10 [1] CRAN (R 4.4.0)
   R6                     2.5.1          2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs               0.3.3          2021-01-31 [1] CRAN (R 4.4.0)
   rbibutils              2.2.16         2023-10-25 [1] CRAN (R 4.4.0)
   RColorBrewer           1.1-3          2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                   1.0.13         2024-07-17 [1] CRAN (R 4.4.0)
   RcppParallel           5.1.8          2024-07-06 [1] CRAN (R 4.4.0)
   RcppZiggurat           0.1.6          2020-10-20 [1] CRAN (R 4.4.0)
   RCurl                  1.98-1.14      2024-01-09 [1] CRAN (R 4.4.0)
   Rdpack                 2.6            2023-11-08 [1] CRAN (R 4.4.0)
   remaCor                0.0.18         2024-02-08 [1] CRAN (R 4.4.0)
   rematch2               2.1.2          2020-05-01 [1] CRAN (R 4.4.0)
   reshape2               1.4.4          2020-04-09 [1] CRAN (R 4.4.0)
   restfulr               0.0.15         2022-06-16 [1] CRAN (R 4.4.0)
   Rfast                  2.1.0          2023-11-09 [1] CRAN (R 4.4.0)
   Rgraphviz              2.48.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RhpcBLASctl            0.23-42        2023-02-11 [1] CRAN (R 4.4.0)
   rjson                  0.2.21         2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                * 1.1.4          2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown              2.27           2024-05-17 [1] CRAN (R 4.4.0)
   rmeta                  3.0            2018-03-20 [1] CRAN (R 4.4.0)
   rprojroot              2.0.4          2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools              2.20.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                2.3.7          2024-05-27 [1] CRAN (R 4.4.0)
   rstatix                0.7.2          2023-02-01 [1] CRAN (R 4.4.0)
   rstudioapi             0.16.0         2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                   1.0.5          2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer            1.64.0         2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   S4Arrays               1.4.1          2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors            * 0.42.1         2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                   0.4.9          2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix           1.12.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                 1.3.0          2023-11-28 [1] CRAN (R 4.4.0)
   scater                 1.32.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scattermore            1.2            2023-06-12 [1] CRAN (R 4.4.0)
   scuttle                1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo            1.2.2          2021-12-06 [1] CRAN (R 4.4.0)
   shiny                  1.8.1.1        2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets           0.8.6          2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment * 1.26.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spam                   2.10-0         2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray            1.4.8          2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats      1.16.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment    * 1.14.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD          * 1.16.2         2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   SQUAREM                2021.1         2021-01-13 [1] CRAN (R 4.4.0)
   statmod                1.5.0          2023-01-06 [1] CRAN (R 4.4.0)
   stringi                1.8.4          2024-05-06 [1] CRAN (R 4.4.0)
   stringr                1.5.1          2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment * 1.34.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   tibble                 3.2.1          2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                  1.3.1          2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect             1.2.1          2024-03-11 [1] CRAN (R 4.4.0)
   tidytree               0.4.6          2023-12-12 [1] CRAN (R 4.4.0)
   treeio                 1.28.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   truncnorm              1.0-9          2023-03-20 [1] CRAN (R 4.4.0)
   UCSC.utils             1.0.0          2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   utf8                   1.2.4          2023-10-22 [1] CRAN (R 4.4.0)
   variancePartition    * 1.34.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   vctrs                  0.6.5          2023-12-01 [1] CRAN (R 4.4.0)
   vipor                  0.4.7          2023-12-18 [1] CRAN (R 4.4.0)
   viridis                0.6.5          2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite            0.4.2          2023-05-02 [1] CRAN (R 4.4.0)
   withr                  3.0.0          2024-01-16 [1] CRAN (R 4.4.0)
   xfun                   0.45           2024-06-16 [1] CRAN (R 4.4.0)
   XML                    3.99-0.17      2024-06-25 [1] CRAN (R 4.3.3)
   xml2                   1.3.6          2023-12-04 [1] CRAN (R 4.4.0)
   xtable                 1.8-4          2019-04-21 [1] CRAN (R 4.4.0)
   XVector              * 0.44.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                   2.3.9          2024-07-05 [1] CRAN (R 4.4.0)
   yulab.utils            0.1.4          2024-01-28 [1] CRAN (R 4.4.0)
   zenith                 1.6.0          2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   zlibbioc               1.50.0         2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────
