---
title: "03-Compare Yao23 Subclass Sex DE to Visium"
author: "Bernard Mulvey"
date: "2024-01-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(ggplot2)
library(gridExtra)
require(colorout)
library(orthogene)
library(biomaRt)
library(gtools)
library(BiocParallel)
source("~/.R/calc_tau_for_cellOrTissueSpecificity_metric.R")
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))
```

load dreamlet results for Visium and Yao 23 mouse scseq
```{r}
ms <- fread("data/12-Mouse comparisons/02-Yao23-HYP-subclass-sexDE_dreamlet_donormouse-ranfx.txt")
msn <- fread("data/12-Mouse comparisons/02-Yao23-HYP-ncells-per-subclass.txt")

vde <- readRDS("data/09-Sex DE/01-dreamlet-ranfx_nnsvg10-HmnylmbNA-BS-15-20-31-15VMHARCclpsd.RDS")
```

## get 1-1 mapping mouse orthologs of human genes (by ensg id) and filter the two datasets accordingly
```{r}
queryg <- unique(unlist(lapply(vde,FUN=function(x){x$ID})))
hggenes <- as.data.table(queryg)
msgenes <- convert_orthologs(gene_df = hggenes,gene_input = "queryg",gene_output = "columns",input_species = "human",output_species = "mouse",non121_strategy = "drop_both_species")

# don't filter yet -- we need to retrieve chromosome info and drop sex chr and MT genes first
```

## filter the mouse sex DE to cell types with at least 50 cells in the HYP datasets (since these cluster assignments are based on brain-wide clustering, many clusters are represented very little)
```{r}
ms <- ms[assay %in% msn[N>49,V1]]
```

## drop sex chromosomal and mitochondrial genes from mouse and human results, then calculate pairwise logFCs. we also lose a few other genes here, so update the filtering of vde accordingly
```{r}
## get mouse gene chromosomes
biomart <- useMart("ensembl",dataset="mmusculus_gene_ensembl")
mschrs <- biomaRt::getBM(attributes=c("external_gene_name","chromosome_name"),mart=biomart,filter="external_gene_name",values=unique(msgenes$ortholog_gene))
rm(biomart)

## filter mouse genes to autosomal, non-mitochondrial
mschrs <- as.data.table(mschrs)
mschrs <- mschrs[!(chromosome_name %in% c("X","Y","MT"))]

## filter the ortholog table and merge it into the two datasets for cluster-pair-wise logfc correlations
msgenes <- as.data.table(msgenes)
msgenes <- msgenes[ortholog_gene %in% mschrs$external_gene_name]

## now filter the sex DE result tables
ms <- ms[ID %in% msgenes$ortholog_gene]
vde <- lapply(vde,FUN=function(x){x[ID %in% msgenes$input_gene & !(chromosome_name %in% c("chrX","chrY","chrMT"))]})

### merge the gene info so we can merge the DE results downstream for gene logFC correlation calcs
ms <- merge(ms,msgenes,by.x="ID",by.y="ortholog_gene")
vde <- lapply(vde,FUN=function(x){merge(x,msgenes,by.x="ID",by.y="input_gene")})
```

## get unique combinations of clusters, join and correlate the two male-vs-female logFCs for each
```{r}
cluspairs <- lapply(vde,FUN=function(x){
    tmp <- as.data.table(expand.grid(unique(ms$assay),unique(x$assay)))
    tmp[,Var1:=as.character(Var1)]
    tmp[,Var2:=as.character(Var2)]
    return(tmp)})

# initialize result table
sbp<-MulticoreParam(workers=4)
register(sbp)

restabs <- bpmapply(x=vde,y=cluspairs,SIMPLIFY = F, BPPARAM=sbp,FUN=function(x,y){
    
    i <- 1
    for (i in c(1:nrow(y))){
        ## result holder for iteration
        tmptab <- as.data.frame(matrix(nrow=9,ncol=5))
        
        msassay <- y[i,Var1]
        hgassay <- y[i,Var2]
        
        # filter to current clusters and merge by human ENSG id
        tmpms <- copy(ms)[assay==msassay][,.(t,input_gene)]
        setnames(tmpms,"t","tms")
        
        tmpx <- x[assay==hgassay][,.(t,ID)]
        # get absolute t for sorting purposes only
        tmpx[,abst:=abs(t)]
        
        tmpcompare <- merge.data.table(tmpms,tmpx,by.x="input_gene",by.y="ID")
        
        j <- 1
        topn <- c(25,50,75,100,250,500,750,1000,nrow(tmpcompare))
        for (j in c(1:9)){
            ## get the top n genes by abs t. DONT USE THE ABST for correl!
            curtop <- topn[j]
            setorderv(tmpcompare,"abst",-1)
            curcompare <- copy(tmpcompare)[c(1:curtop)]

            ## calculate the correlation
            tmptab[j,1] <- msassay
            tmptab[j,2] <- hgassay
            tmptab[j,3] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="pearson")
            tmptab[j,4] <- cor(curcompare$tms,curcompare$t,use="pairwise.complete.obs",method="spearman")
            tmptab[j,5] <- curtop
            rm(curtop,curcompare)
        }
        
        if(i==1){
            restab <- copy(tmptab)
        }
        else{restab <- rbind(restab,tmptab)}
        rm(tmpms,tmpx,tmpcompare,mgassay,hgassay,tmptab)
    }
    restab <- as.data.table(restab)
    setnames(restab,c("mouse_subclass","human_domain","pcc","spearmancor","top_n_human"))
    return(restab)
})
names(restabs) <- names(vde)


saveRDS(restabs,"data/12-Mouse comparisons/03-Yao23-subclass-MvFsexDE-tstat_correls_to_k15-20-31-15collapsed_bytopNhumanDEGs.RDS")
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] default_1.0.0       BiocParallel_1.36.0 gtools_3.9.5       
 [4] biomaRt_2.58.2      orthogene_1.8.0     colorout_1.3-0.1   
 [7] gridExtra_2.3       ggplot2_3.4.4       Biostrings_2.70.2  
[10] GenomeInfoDb_1.38.5 XVector_0.42.0      IRanges_2.36.0     
[13] S4Vectors_0.40.2    BiocGenerics_0.48.1 data.table_1.14.10 
[16] rlang_1.1.3        

loaded via a namespace (and not attached):
  [1] later_1.3.2                   BiocIO_1.12.0                
  [3] ggplotify_0.1.2               bitops_1.0-7                 
  [5] filelock_1.0.3                fields_15.2                  
  [7] tibble_3.2.1                  XML_3.99-0.16.1              
  [9] lifecycle_1.0.4               rstatix_0.7.2                
 [11] edgeR_4.0.14                  doParallel_1.0.17            
 [13] rprojroot_2.0.4               lattice_0.22-5               
 [15] backports_1.4.1               magrittr_2.0.3               
 [17] limma_3.58.1                  plotly_4.10.4                
 [19] sass_0.4.8                    rmarkdown_2.25               
 [21] jquerylib_0.1.4               yaml_2.3.8                   
 [23] httpuv_1.6.13                 spam_2.10-0                  
 [25] sessioninfo_1.2.2             cowplot_1.1.3                
 [27] DBI_1.2.1                     RColorBrewer_1.1-3           
 [29] golem_0.4.1                   maps_3.4.2                   
 [31] abind_1.4-5                   zlibbioc_1.48.0              
 [33] GenomicRanges_1.54.1          purrr_1.0.2                  
 [35] RCurl_1.98-1.14               yulab.utils_0.1.4            
 [37] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      
 [39] ggrepel_0.9.5                 irlba_2.3.5.1                
 [41] tidytree_0.4.6                pheatmap_1.0.12              
 [43] DelayedMatrixStats_1.24.0     codetools_0.2-19             
 [45] DelayedArray_0.28.0           xml2_1.3.6                   
 [47] DT_0.31                       scuttle_1.12.0               
 [49] tidyselect_1.2.0              aplot_0.2.2                  
 [51] ScaledMatrix_1.10.0           viridis_0.6.5                
 [53] shinyWidgets_0.8.1            matrixStats_1.2.0            
 [55] BiocFileCache_2.10.1          GenomicAlignments_1.38.2     
 [57] jsonlite_1.8.8                BiocNeighbors_1.20.2         
 [59] ellipsis_0.3.2                scater_1.30.1                
 [61] iterators_1.0.14              foreach_1.5.2                
 [63] progress_1.2.3                tools_4.3.2                  
 [65] treeio_1.26.0                 Rcpp_1.0.12                  
 [67] glue_1.7.0                    SparseArray_1.2.3            
 [69] xfun_0.41                     here_1.0.1                   
 [71] MatrixGenerics_1.14.0         dplyr_1.1.4                  
 [73] withr_3.0.0                   BiocManager_1.30.22          
 [75] fastmap_1.1.1                 fansi_1.0.6                  
 [77] digest_0.6.34                 rsvd_1.0.5                   
 [79] gridGraphics_0.5-1            R6_2.5.1                     
 [81] mime_0.12                     colorspace_2.1-0             
 [83] RSQLite_2.3.5                 config_0.3.2                 
 [85] utf8_1.2.4                    tidyr_1.3.1                  
 [87] generics_0.1.3                rtracklayer_1.62.0           
 [89] prettyunits_1.2.0             httr_1.4.7                   
 [91] htmlwidgets_1.6.4             S4Arrays_1.2.0               
 [93] scatterplot3d_0.3-44          pkgconfig_2.0.3              
 [95] gtable_0.3.4                  blob_1.2.4                   
 [97] SingleCellExperiment_1.24.0   htmltools_0.5.7              
 [99] carData_3.0-5                 dotCall64_1.1-1              
[101] scales_1.3.0                  Biobase_2.62.0               
[103] png_0.1-8                     attempt_0.3.1                
[105] SpatialExperiment_1.12.0      ggfun_0.1.4                  
[107] knitr_1.45                    rstudioapi_0.15.0            
[109] rjson_0.2.21                  nlme_3.1-164                 
[111] curl_5.2.0                    cachem_1.0.8                 
[113] stringr_1.5.1                 Polychrome_1.5.1             
[115] BiocVersion_3.18.1            parallel_4.3.2               
[117] vipor_0.4.7                   AnnotationDbi_1.64.1         
[119] restfulr_0.0.15               pillar_1.9.0                 
[121] grid_4.3.2                    vctrs_0.6.5                  
[123] ggpubr_0.6.0                  promises_1.2.1               
[125] car_3.1-2                     BiocSingular_1.18.0          
[127] dbplyr_2.4.0                  beachmat_2.18.0              
[129] xtable_1.8-4                  beeswarm_0.4.0               
[131] paletteer_1.6.0               evaluate_0.23                
[133] magick_2.8.2                  cli_3.6.2                    
[135] locfit_1.5-9.8                compiler_4.3.2               
[137] Rsamtools_2.18.0              crayon_1.5.2                 
[139] grr_0.9.5                     ggsignif_0.6.4               
[141] gprofiler2_0.2.2              rematch2_2.1.2               
[143] fs_1.6.3                      ggbeeswarm_0.7.2             
[145] stringi_1.8.3                 viridisLite_0.4.2            
[147] babelgene_22.9                munsell_0.5.0                
[149] lazyeval_0.2.2                homologene_1.4.68.19.3.27    
[151] Matrix_1.6-5                  ExperimentHub_2.10.0         
[153] benchmarkme_1.0.8             patchwork_1.2.0              
[155] hms_1.1.3                     sparseMatrixStats_1.14.0     
[157] bit64_4.0.5                   KEGGREST_1.42.0              
[159] statmod_1.5.0                 shiny_1.8.0                  
[161] SummarizedExperiment_1.32.0   interactiveDisplayBase_1.40.0
[163] AnnotationHub_3.10.0          broom_1.0.5                  
[165] memoise_2.0.1                 bslib_0.6.1                  
[167] ggtree_3.10.0                 benchmarkmeData_1.0.4        
[169] bit_4.0.5                     spatialLIBD_1.14.1           
[171] ape_5.7-1                    
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2024-02-11
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────
 package                * version        date (UTC) lib source
 abind                    1.4-5          2016-07-21 [1] CRAN (R 4.3.0)
 AnnotationDbi            1.64.1         2023-11-02 [1] Bioconductor
 AnnotationHub            3.10.0         2023-10-24 [1] Bioconductor
 ape                      5.7-1          2023-03-13 [1] CRAN (R 4.3.2)
 aplot                    0.2.2          2023-10-06 [1] CRAN (R 4.3.1)
 attempt                  0.3.1          2020-05-03 [1] CRAN (R 4.3.0)
 babelgene                22.9           2022-09-29 [1] CRAN (R 4.3.0)
 backports                1.4.1          2021-12-13 [1] CRAN (R 4.3.2)
 beachmat                 2.18.0         2023-10-24 [1] Bioconductor
 beeswarm                 0.4.0          2021-06-01 [1] CRAN (R 4.3.2)
 benchmarkme              1.0.8          2022-06-12 [1] CRAN (R 4.3.0)
 benchmarkmeData          1.0.4          2020-04-23 [1] CRAN (R 4.3.0)
 Biobase                  2.62.0         2023-10-24 [1] Bioconductor
 BiocFileCache            2.10.1         2023-10-26 [1] Bioconductor
 BiocGenerics           * 0.48.1         2023-11-02 [1] Bioconductor
 BiocIO                   1.12.0         2023-10-26 [1] Bioconductor
 BiocManager              1.30.22        2023-08-08 [1] CRAN (R 4.3.0)
 BiocNeighbors            1.20.2         2024-01-07 [1] Bioconductor 3.18 (R 4.3.2)
 BiocParallel           * 1.36.0         2023-10-26 [1] Bioconductor
 BiocSingular             1.18.0         2023-10-24 [1] Bioconductor
 BiocVersion              3.18.1         2023-11-18 [1] Bioconductor 3.18 (R 4.3.2)
 biomaRt                * 2.58.2         2024-01-30 [1] Bioconductor 3.18 (R 4.3.2)
 Biostrings             * 2.70.2         2024-01-28 [1] Bioconductor 3.18 (R 4.3.2)
 bit                      4.0.5          2022-11-15 [1] CRAN (R 4.3.2)
 bit64                    4.0.5          2020-08-30 [1] CRAN (R 4.3.2)
 bitops                   1.0-7          2021-04-24 [1] CRAN (R 4.3.2)
 blob                     1.2.4          2023-03-17 [1] CRAN (R 4.3.0)
 broom                    1.0.5          2023-06-09 [1] CRAN (R 4.3.0)
 bslib                    0.6.1          2023-11-28 [1] CRAN (R 4.3.1)
 cachem                   1.0.8          2023-05-01 [1] CRAN (R 4.3.2)
 car                      3.1-2          2023-03-30 [1] CRAN (R 4.3.0)
 carData                  3.0-5          2022-01-06 [1] CRAN (R 4.3.0)
 cli                      3.6.2          2023-12-11 [1] CRAN (R 4.3.2)
 codetools                0.2-19         2023-02-01 [1] CRAN (R 4.3.0)
 colorout               * 1.3-0.1        2024-01-11 [1] local
 colorspace               2.1-0          2023-01-23 [1] CRAN (R 4.3.2)
 config                   0.3.2          2023-08-30 [1] CRAN (R 4.3.0)
 cowplot                  1.1.3          2024-01-22 [1] CRAN (R 4.3.1)
 crayon                   1.5.2          2022-09-29 [1] CRAN (R 4.3.0)
 curl                     5.2.0          2023-12-08 [1] CRAN (R 4.3.2)
 data.table             * 1.14.10        2023-12-08 [1] CRAN (R 4.3.2)
 DBI                      1.2.1          2024-01-12 [1] CRAN (R 4.3.2)
 dbplyr                   2.4.0          2023-10-26 [1] CRAN (R 4.3.1)
 default                * 1.0.0          2017-08-07 [1] CRAN (R 4.3.0)
 DelayedArray             0.28.0         2023-10-24 [1] Bioconductor
 DelayedMatrixStats       1.24.0         2023-11-06 [1] Bioconductor
 digest                   0.6.34         2024-01-11 [1] CRAN (R 4.3.2)
 doParallel               1.0.17         2022-02-07 [1] CRAN (R 4.3.0)
 dotCall64                1.1-1          2023-11-28 [1] CRAN (R 4.3.2)
 dplyr                    1.1.4          2023-11-17 [1] CRAN (R 4.3.2)
 DT                       0.31           2023-12-09 [1] CRAN (R 4.3.1)
 edgeR                    4.0.14         2024-01-29 [1] Bioconductor 3.18 (R 4.3.2)
 ellipsis                 0.3.2          2021-04-29 [1] CRAN (R 4.3.2)
 evaluate                 0.23           2023-11-01 [1] CRAN (R 4.3.1)
 ExperimentHub            2.10.0         2023-10-26 [1] Bioconductor
 fansi                    1.0.6          2023-12-08 [1] CRAN (R 4.3.2)
 fastmap                  1.1.1          2023-02-24 [1] CRAN (R 4.3.2)
 fields                   15.2           2023-08-17 [1] CRAN (R 4.3.2)
 filelock                 1.0.3          2023-12-11 [1] CRAN (R 4.3.2)
 foreach                  1.5.2          2022-02-02 [1] CRAN (R 4.3.0)
 fs                       1.6.3          2023-07-20 [1] CRAN (R 4.3.0)
 generics                 0.1.3          2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb           * 1.38.5         2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData         1.2.11         2024-01-11 [1] Bioconductor
 GenomicAlignments        1.38.2         2024-01-16 [1] Bioconductor 3.18 (R 4.3.2)
 GenomicRanges            1.54.1         2023-10-29 [1] Bioconductor
 ggbeeswarm               0.7.2          2023-04-29 [1] CRAN (R 4.3.0)
 ggfun                    0.1.4          2024-01-19 [1] CRAN (R 4.3.2)
 ggplot2                * 3.4.4          2023-10-12 [1] CRAN (R 4.3.1)
 ggplotify                0.1.2          2023-08-09 [1] CRAN (R 4.3.0)
 ggpubr                   0.6.0          2023-02-10 [1] CRAN (R 4.3.0)
 ggrepel                  0.9.5          2024-01-10 [1] CRAN (R 4.3.2)
 ggsignif                 0.6.4          2022-10-13 [1] CRAN (R 4.3.0)
 ggtree                   3.10.0         2023-11-06 [1] Bioconductor
 glue                     1.7.0          2024-01-09 [1] CRAN (R 4.3.2)
 golem                    0.4.1          2023-06-05 [1] CRAN (R 4.3.0)
 gprofiler2               0.2.2          2023-06-14 [1] CRAN (R 4.3.0)
 gridExtra              * 2.3            2017-09-09 [1] CRAN (R 4.3.0)
 gridGraphics             0.5-1          2020-12-13 [1] CRAN (R 4.3.0)
 grr                      0.9.5          2016-08-26 [1] CRAN (R 4.3.2)
 gtable                   0.3.4          2023-08-21 [1] CRAN (R 4.3.0)
 gtools                 * 3.9.5          2023-11-20 [1] CRAN (R 4.3.2)
 here                     1.0.1          2020-12-13 [1] CRAN (R 4.3.0)
 hms                      1.1.3          2023-03-21 [1] CRAN (R 4.3.0)
 homologene               1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.3.0)
 htmltools                0.5.7          2023-11-03 [1] CRAN (R 4.3.2)
 htmlwidgets              1.6.4          2023-12-06 [1] CRAN (R 4.3.1)
 httpuv                   1.6.13         2023-12-06 [1] CRAN (R 4.3.1)
 httr                     1.4.7          2023-08-15 [1] CRAN (R 4.3.0)
 interactiveDisplayBase   1.40.0         2023-10-26 [1] Bioconductor
 IRanges                * 2.36.0         2023-10-24 [1] Bioconductor
 irlba                    2.3.5.1        2022-10-03 [1] CRAN (R 4.3.2)
 iterators                1.0.14         2022-02-05 [1] CRAN (R 4.3.0)
 jquerylib                0.1.4          2021-04-26 [1] CRAN (R 4.3.0)
 jsonlite                 1.8.8          2023-12-04 [1] CRAN (R 4.3.2)
 KEGGREST                 1.42.0         2023-10-26 [1] Bioconductor
 knitr                    1.45           2023-10-30 [1] CRAN (R 4.3.2)
 later                    1.3.2          2023-12-06 [1] CRAN (R 4.3.2)
 lattice                  0.22-5         2023-10-24 [1] CRAN (R 4.3.2)
 lazyeval                 0.2.2          2019-03-15 [1] CRAN (R 4.3.2)
 lifecycle                1.0.4          2023-11-07 [1] CRAN (R 4.3.1)
 limma                    3.58.1         2023-10-31 [1] Bioconductor
 locfit                   1.5-9.8        2023-06-11 [1] CRAN (R 4.3.2)
 magick                   2.8.2          2023-12-20 [1] CRAN (R 4.3.1)
 magrittr                 2.0.3          2022-03-30 [1] CRAN (R 4.3.2)
 maps                     3.4.2          2023-12-15 [1] CRAN (R 4.3.1)
 Matrix                   1.6-5          2024-01-11 [1] CRAN (R 4.3.2)
 MatrixGenerics           1.14.0         2023-10-26 [1] Bioconductor
 matrixStats              1.2.0          2023-12-11 [1] CRAN (R 4.3.2)
 memoise                  2.0.1          2021-11-26 [1] CRAN (R 4.3.0)
 mime                     0.12           2021-09-28 [1] CRAN (R 4.3.2)
 munsell                  0.5.0          2018-06-12 [1] CRAN (R 4.3.0)
 nlme                     3.1-164        2023-11-27 [1] CRAN (R 4.3.2)
 orthogene              * 1.8.0          2023-11-06 [1] Bioconductor
 paletteer                1.6.0          2024-01-21 [1] CRAN (R 4.3.1)
 patchwork                1.2.0          2024-01-08 [1] CRAN (R 4.3.1)
 pheatmap                 1.0.12         2019-01-04 [1] CRAN (R 4.3.0)
 pillar                   1.9.0          2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig                2.0.3          2019-09-22 [1] CRAN (R 4.3.0)
 plotly                   4.10.4         2024-01-13 [1] CRAN (R 4.3.1)
 png                      0.1-8          2022-11-29 [1] CRAN (R 4.3.2)
 Polychrome               1.5.1          2022-05-03 [1] CRAN (R 4.3.0)
 prettyunits              1.2.0          2023-09-24 [1] CRAN (R 4.3.1)
 progress                 1.2.3          2023-12-06 [1] CRAN (R 4.3.1)
 promises                 1.2.1          2023-08-10 [1] CRAN (R 4.3.2)
 purrr                    1.0.2          2023-08-10 [1] CRAN (R 4.3.2)
 R6                       2.5.1          2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs                 0.3.3          2021-01-31 [1] CRAN (R 4.3.2)
 RColorBrewer             1.1-3          2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp                     1.0.12         2024-01-09 [1] CRAN (R 4.3.2)
 RCurl                    1.98-1.14      2024-01-09 [1] CRAN (R 4.3.1)
 rematch2                 2.1.2          2020-05-01 [1] CRAN (R 4.3.0)
 restfulr                 0.0.15         2022-06-16 [1] CRAN (R 4.3.2)
 rjson                    0.2.21         2022-01-09 [1] CRAN (R 4.3.2)
 rlang                  * 1.1.3          2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown                2.25           2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot                2.0.4          2023-11-05 [1] CRAN (R 4.3.1)
 Rsamtools                2.18.0         2023-10-24 [1] Bioconductor
 RSQLite                  2.3.5          2024-01-21 [1] CRAN (R 4.3.2)
 rstatix                  0.7.2          2023-02-01 [1] CRAN (R 4.3.0)
 rstudioapi               0.15.0         2023-07-07 [1] CRAN (R 4.3.0)
 rsvd                     1.0.5          2021-04-16 [1] CRAN (R 4.3.0)
 rtracklayer              1.62.0         2023-10-26 [1] Bioconductor
 S4Arrays                 1.2.0          2023-10-24 [1] Bioconductor
 S4Vectors              * 0.40.2         2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 sass                     0.4.8          2023-12-06 [1] CRAN (R 4.3.1)
 ScaledMatrix             1.10.0         2023-11-06 [1] Bioconductor
 scales                   1.3.0          2023-11-28 [1] CRAN (R 4.3.2)
 scater                   1.30.1         2023-11-16 [1] Bioconductor
 scatterplot3d            0.3-44         2023-05-05 [1] CRAN (R 4.3.0)
 scuttle                  1.12.0         2023-10-24 [1] Bioconductor
 sessioninfo              1.2.2          2021-12-06 [1] CRAN (R 4.3.0)
 shiny                    1.8.0          2023-11-17 [1] CRAN (R 4.3.1)
 shinyWidgets             0.8.1          2024-01-10 [1] CRAN (R 4.3.1)
 SingleCellExperiment     1.24.0         2023-11-06 [1] Bioconductor
 spam                     2.10-0         2023-10-23 [1] CRAN (R 4.3.2)
 SparseArray              1.2.3          2023-12-25 [1] Bioconductor 3.18 (R 4.3.2)
 sparseMatrixStats        1.14.0         2023-10-24 [1] Bioconductor
 SpatialExperiment        1.12.0         2023-10-26 [1] Bioconductor
 spatialLIBD              1.14.1         2023-11-30 [1] Bioconductor 3.18 (R 4.3.2)
 statmod                  1.5.0          2023-01-06 [1] CRAN (R 4.3.2)
 stringi                  1.8.3          2023-12-11 [1] CRAN (R 4.3.1)
 stringr                  1.5.1          2023-11-14 [1] CRAN (R 4.3.1)
 SummarizedExperiment     1.32.0         2023-11-06 [1] Bioconductor
 tibble                   3.2.1          2023-03-20 [1] CRAN (R 4.3.2)
 tidyr                    1.3.1          2024-01-24 [1] CRAN (R 4.3.1)
 tidyselect               1.2.0          2022-10-10 [1] CRAN (R 4.3.0)
 tidytree                 0.4.6          2023-12-12 [1] CRAN (R 4.3.1)
 treeio                   1.26.0         2023-11-06 [1] Bioconductor
 utf8                     1.2.4          2023-10-22 [1] CRAN (R 4.3.2)
 vctrs                    0.6.5          2023-12-01 [1] CRAN (R 4.3.2)
 vipor                    0.4.7          2023-12-18 [1] CRAN (R 4.3.1)
 viridis                  0.6.5          2024-01-29 [1] CRAN (R 4.3.2)
 viridisLite              0.4.2          2023-05-02 [1] CRAN (R 4.3.0)
 withr                    3.0.0          2024-01-16 [1] CRAN (R 4.3.1)
 xfun                     0.41           2023-11-01 [1] CRAN (R 4.3.2)
 XML                      3.99-0.16.1    2024-01-22 [1] CRAN (R 4.3.1)
 xml2                     1.3.6          2023-12-04 [1] CRAN (R 4.3.2)
 xtable                   1.8-4          2019-04-21 [1] CRAN (R 4.3.0)
 XVector                * 0.42.0         2023-10-24 [1] Bioconductor
 yaml                     2.3.8          2023-12-11 [1] CRAN (R 4.3.2)
 yulab.utils              0.1.4          2024-01-28 [1] CRAN (R 4.3.2)
 zlibbioc                 1.48.0         2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────
