---
title: "06a-Spot correlations of VMH to kim 2019 ant-post smartseq markers"
author: "Bernie Mulvey"
date: "2025-03-22"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(SpatialExperiment)
library(spatialLIBD)
library(gprofiler2)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel)


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 9), axis.title.x = element_text(size = 8), axis.text.y = element_text(size = 8), axis.title.y = element_text(size =9), plot.title = element_markdown(size = 11,hjust=0.5), strip.text = element_text(size=11), legend.text = element_text(size=8), legend.title = element_text(size=9,hjust=0.5),plot.title.position="plot"))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

load smartseq cpm dgcmatrix and metadata (both at https://data.mendeley.com/datasets/ypx3sw2f7c/3 ) --> SCE
```{r}
load("raw-data/KimAnderson19_msVMHprofiles/smartseq_cpm.rda")
# drop undetected genes if any
dropg <- rownames(cpm)[rowSums(cpm)==0]
cpm <- cpm[!rownames(cpm) %in% dropg,]

met <- fread("raw-data/KimAnderson19_msVMHprofiles/SMART-seq_VMH_metadata.csv")
```

there isnt a tabular form of cllusterwise matching between smartseq profiles and the anterior-posterior Seqfish clusters that I can find, so we'll need to make based on Fig 2? 3? panel C (heatmap of smartseq cluster corresp. to the anteroposterior seqfish clusters). the labels they have on the x axis match to those in met$cca_cluster_label. 
```{r}
smart_to_sfish <- fread("raw-data/KimAnderson19_msVMHprofiles/KimAnderson19_smart_seqfishpairs_fromfig3c.txt")

# stopifnot(all(unique(smart_to_sfish$smartseq_clus) %in% met$cca_cluster_label)) 
smart_to_sfish[!(smartseq_clus %in% met$cca_cluster_label),unique(smartseq_clus)] # Esr1_2, esr1_3
grep(unique(met$cca_cluster_label),pattern="Esr1_",value=T) # those two clusters are collapsed as Esr1_2|3 i guess
smart_to_sfish[smartseq_clus %in% c("Esr1_2","Esr1_3"),smartseq_clus:="Esr1_2|3"]
stopifnot(all(unique(smart_to_sfish$smartseq_clus) %in% met$cca_cluster_label)) 

multimapfish <- smart_to_sfish[,.N,by="smartseq_clus"][N>1,smartseq_clus]
multimap <- lapply(multimapfish,function(x){
    tmp <- as.data.table(cbind(unique(smart_to_sfish[smartseq_clus==x,smartseq_clus]),paste0(smart_to_sfish[smartseq_clus==x,seqfish_clus],collapse=";")))
    setnames(tmp,c("smartseq_clus","seqfish_clus"))
    tmp[,APorder_fig3d:=paste0(smart_to_sfish[smartseq_clus==x,APorder_fig3d],collapse=";")]
    return(tmp)
})
multimap <- rbindlist(multimap)

smart_to_sfish <- as.data.table(rbind(smart_to_sfish[!(smartseq_clus %in% multimapfish)],multimap))
stopifnot(all(unique(smart_to_sfish$smartseq_clus) %in% met$cca_cluster_label)) 
```

now that we've added the annotations we want for examining anteroposterior stuff specifically, make the sce
```{r}
# append seqfish-smartseq correspondences where applicable
met <- merge.data.table(met,smart_to_sfish,by.x="cca_cluster_label",by.y="smartseq_clus",all.x=T)

stopifnot(ncol(cpm)==nrow(met))
stopifnot(all(colnames(cpm) %in% met$sample_name))

met <- DataFrame(met,row.names=met$sample_name)[colnames(cpm),]
smrt.sce <- SingleCellExperiment(assays = list(cpm = cpm), colData = met,rowData=DataFrame(cbind(rownames(cpm),rownames(cpm))))

rm(cpm)
gc(full=T)
```

## get marker stats for SCE to run spatial registration from
## to do this right, we need to load the SPE and determine what genes not only have 1-1 orthologs, but are represented in both datasets.

```{r}
# load hyp spe, pare down to VMH, and clear out unneeded ram heavy bits while were here
hyp2 <- readRDS("processed-data/03-QC_filters/hypN10_umi210_gene126_chrm50_spotsweeped_lognorm_rotsNmirrors_072224.RDS")
# did we already have the clusters in here? dont think so but check
# colnames(colData(hyp2))
# yano

clus <- fread("processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")
setnames(clus,2,"cl")
clus[,cl:=paste0("X",cl)]
clus[,lab:="drop"]
clus[cl %in% paste0("X",c(7,12)),lab:="VMH"]
clus <- DataFrame(clus,row.names=clus$rn)[colnames(hyp2),]

colLabels(hyp2) <- clus$lab
hyp2 <- hyp2[,hyp2$label=="VMH"]
imgData(hyp2) <- NULL

## drop genes undetected in VMH
dropg <- rownames(hyp2)[rowSums(counts(hyp2))==0]
hyp2 <- hyp2[!(rownames(hyp2) %in% dropg),]

rm(clus,dropg)
gc(full=T)
```


# get the orthologs using gorth and subset to those 1-1 mapping between human and mouse and further are in the human spe
```{r}
orths <- as.data.table(gorth(query=rownames(smrt.sce),source_organism="mmusculus",target_organism="hsapiens"))
dropms <- orths[,.N,by="input_ensg"][N>1,input_ensg]
drophg <- orths[,.N,by="ortholog_ensg"][N>1,ortholog_ensg]
orths <- orths[!(input_ensg %in% dropms) & !(ortholog_ensg %in% drophg)]
orths <- orths[ortholog_ensg %in% rownames(hyp2)]

stopifnot(nrow(unique(orths[,c("input_ensg","ortholog_ensg")]))==nrow(orths))

## subset both datasets to the orths
smrt.sce <- smrt.sce[rownames(smrt.sce) %in% orths$input,]
hyp2 <- hyp2[rownames(hyp2) %in% orths$ortholog_ensg,]
stopifnot(nrow(hyp2)==nrow(smrt.sce))

## cleanup some misc
rm(dropg,drophg,dropms,multimap,smart_to_sfish,multimapfish)
gc(full=T)
```

## run registration stats for single cell
```{r}
colnames(rowData(smrt.sce)) <- c("gene_id","gene_name")

# fix cluster names spatialLIBD doesn't fix to meet its needs, just bitches about; 
smrt.sce$smart_seq_cluster_label <- gsub(smrt.sce$smart_seq_cluster_label,pattern=" ",replacement="_")
smrt.sce$smart_seq_cluster_label <- gsub(smrt.sce$smart_seq_cluster_label,pattern="\\|",replacement="_")

# ?? it also cant fathom genes starting with a numeral and those are all just 2390457293487297Riks anyhow so drop em
# dropg <- grep(rownames(smrt.sce),pattern="^[[:digit:]]{1}.*$",value=T)
# smrt.sce <- smrt.sce[!rownames(smrt.sce) %in% dropg,]
# model factors can't start with numerals here which could be a problem for the animal nname labels, which are gibberish and mostly strings of numbers, so append a character to the front of those just in case.
smrt.sce$animal_name_label <- paste0("x",smrt.sce$animal_name_label)

# for fucks sake it won't run if there's only one assay, even with the expected name "logcounts". the cpm here are are all the authors provide, and not the now-standard logcounts, so w/e just deceive the package into getting what it wants)
assayNames(smrt.sce)[1] <- "counts"
assays(smrt.sce)[2] <- assays(smrt.sce)[1]
assayNames(smrt.sce)[2] <- "logcounts"


msvmh <- registration_wrapper(smrt.sce,var_registration="smart_seq_cluster_label",var_sample_id="animal_name_label",gene_ensembl="gene_id",gene_name="gene_id",min_ncells = 10)

# extract global marker t-stats, which are what we need
msvmh <- as.data.table(msvmh$enrichment)
keepcols <- c("gene",grep(names(msvmh),pattern="t_stat",value=T))
msvmh <- msvmh[,..keepcols]

## add human ensg names for matching
msvmh[,human_ensg:=orths[match(msvmh$gene,orths$input),ortholog_ensg]]

# don't need the sce now
rm(smrt.sce)
gc(full=T)
```

## to do this at the spot level, we actually don't need to use a registration wrapper at all: we can just spearman correlate the spot's logcounts with the marker stats from the single-nucleus. (If a spot and cell perfectly matched and have a tissue-wide exclusive marker, for example, that marker would have a stratospheric t stat and, assuming it is equally likely to be detected as any other gene--ok that's a big if--they would be the top-ranked entry in both datsets. nonparametrics ftw). plus, there's no need to normalize anything across samples or etc since we're using the normalized data (logcounts), and not calculating any sample-specific anything from the visium.

```{r}
## get spot logcounts
lct <- as.data.table(as.data.frame(as.matrix(logcounts(hyp2))),keep.rownames=T)
## dont need the spe
rm(hyp2)
gc(full=T)

## some genes aren't well represented enough for SCE regist stats so we can drop those from the logcounts
lct <- lct[rn %in% msvmh$human_ensg,]
stopifnot(nrow(lct)==nrow(msvmh))

# divy up the data so we don't have to pass the whole logcount table to each worker
spots <- names(lct)[2:ncol(lct)]
spotsl <- lapply(spots,function(s){
    getcol <- c("rn",s)
    o<- lct[,..getcol]
    return(o)
})
names(spotsl) <- spots
rm(spots,lct)
gc(full=T)

## run it
setDTthreads(threads=2,restore_after_fork = F) # prevent resource overcommiting; restore after fork set to F per their docs on misbehavior on mac
is <- c(1:length(spotsl))
spotcor <- mcmapply(s=spotsl,
                    i=is,
                    n=names(spotsl),
                    mc.cores=4,
                    SIMPLIFY=FALSE,
                    FUN=function(s,i,n){
  setDTthreads(2,restore_after_fork = T) # since DT is not too nice about mac implementation, put it here too, and assuming some forking happens during loop execution because i dont get when it happens. and they provide no context for when restore_after_fork does/doesn't work as expected on mac so taking a wild guess tehre

# message doesn't come back but we
#   if(is.integer((i-1)/1000)){
#     message(paste0("starting gene ",i))
#   }

  tmpd <- merge.data.table(s,msvmh,by.x="rn",by.y="human_ensg")

  ans <- tmpd[,lapply(.SD,FUN=function(x){cor(get(n),x,method="spearman")}),.SDcols=grep(names(tmpd),pattern="t_stat",value=T)]
  ans[,spot:=n] # temporarily, just to rotate around; this is a chonk column so we'll drop it agn
  ans <- melt.data.table(ans,id.vars="spot")
  ans[,spot:=NULL]
  ans[,variable:=gsub(variable,pattern="t_stat_",replacement="")]

  return(ans)
})
# fwiw resource monitoring indicates this worked as desired -- 4 processes spawned each using ~"200"% cpu (i.e., 2 cores)

## put it all together
spotcor <- rbindlist(spotcor)

saveRDS(spotcor,"processed-data/12-Mouse comparisons/06-VMHspot_logct_spearman_to_kim2019smartseq_markerTstats.RDS")


```

repinf
```{r}
sessionInfo()
sessioninfo::session_info()
```

R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats    
[4] graphics  grDevices utils    
[7] methods   base     

other attached packages:
 [1] BiocParallel_1.40.0        
 [2] parallelly_1.42.0          
 [3] colorout_1.3-0.2           
 [4] gprofiler2_0.2.3           
 [5] spatialLIBD_1.18.0         
 [6] SpatialExperiment_1.16.0   
 [7] SingleCellExperiment_1.28.1
 [8] SummarizedExperiment_1.36.0
 [9] Biobase_2.66.0             
[10] GenomicRanges_1.58.0       
[11] GenomeInfoDb_1.42.3        
[12] IRanges_2.40.1             
[13] S4Vectors_0.44.0           
[14] BiocGenerics_0.52.0        
[15] MatrixGenerics_1.18.1      
[16] matrixStats_1.5.0          
[17] gridExtra_2.3              
[18] ggtext_0.1.2               
[19] ggplot2_3.5.1              
[20] data.table_1.17.0          
[21] rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      
  [2] jsonlite_1.9.1          
  [3] magrittr_2.0.3          
  [4] ggbeeswarm_0.7.2        
  [5] magick_2.8.5            
  [6] BiocIO_1.16.0           
  [7] fields_16.3.1           
  [8] zlibbioc_1.52.0         
  [9] vctrs_0.6.5             
 [10] memoise_2.0.1           
 [11] config_0.3.2            
 [12] Rsamtools_2.22.0        
 [13] paletteer_1.6.0         
 [14] RCurl_1.98-1.16         
 [15] benchmarkme_1.0.8       
 [16] htmltools_0.5.8.1       
 [17] S4Arrays_1.6.0          
 [18] AnnotationHub_3.14.0    
 [19] curl_6.2.1              
 [20] BiocNeighbors_2.0.1     
 [21] SparseArray_1.6.2       
 [22] sass_0.4.9              
 [23] bslib_0.9.0             
 [24] htmlwidgets_1.6.4       
 [25] plotly_4.10.4           
 [26] cachem_1.1.0            
 [27] GenomicAlignments_1.42.0
 [28] mime_0.12               
 [29] lifecycle_1.0.4         
 [30] iterators_1.0.14        
 [31] pkgconfig_2.0.3         
 [32] rsvd_1.0.5              
 [33] Matrix_1.7-2            
 [34] R6_2.6.1                
 [35] fastmap_1.2.0           
 [36] GenomeInfoDbData_1.2.13 
 [37] shiny_1.10.0            
 [38] digest_0.6.37           
 [39] colorspace_2.1-1        
 [40] rematch2_2.1.2          
 [41] AnnotationDbi_1.68.0    
 [42] scater_1.34.1           
 [43] rprojroot_2.0.4         
 [44] irlba_2.3.5.1           
 [45] ExperimentHub_2.14.0    
 [46] RSQLite_2.3.9           
 [47] beachmat_2.22.0         
 [48] filelock_1.0.3          
 [49] httr_1.4.7              
 [50] abind_1.4-8             
 [51] compiler_4.4.3          
 [52] here_1.0.1              
 [53] bit64_4.6.0-1           
 [54] withr_3.0.2             
 [55] doParallel_1.0.17       
 [56] attempt_0.3.1           
 [57] viridis_0.6.5           
 [58] DBI_1.2.3               
 [59] maps_3.4.2.1            
 [60] sessioninfo_1.2.3       
 [61] rappdirs_0.3.3          
 [62] DelayedArray_0.32.0     
 [63] rjson_0.2.23            
 [64] tools_4.4.3             
 [65] vipor_0.4.7             
 [66] beeswarm_0.4.0          
 [67] httpuv_1.6.15           
 [68] glue_1.8.0              
 [69] restfulr_0.0.15         
 [70] promises_1.3.2          
 [71] gridtext_0.1.5          
 [72] grid_4.4.3              
 [73] generics_0.1.3          
 [74] gtable_0.3.6            
 [75] tidyr_1.3.1             
 [76] ScaledMatrix_1.14.0     
 [77] BiocSingular_1.22.0     
 [78] xml2_1.3.6              
 [79] XVector_0.46.0          
 [80] ggrepel_0.9.6           
 [81] BiocVersion_3.20.0      
 [82] foreach_1.5.2           
 [83] pillar_1.10.1           
 [84] spam_2.11-1             
 [85] limma_3.62.2            
 [86] later_1.4.1             
 [87] benchmarkmeData_1.0.4   
 [88] dplyr_1.1.4             
 [89] BiocFileCache_2.14.0    
 [90] lattice_0.22-6          
 [91] rtracklayer_1.66.0      
 [92] bit_4.6.0               
 [93] tidyselect_1.2.1        
 [94] locfit_1.5-9.12         
 [95] scuttle_1.16.0          
 [96] Biostrings_2.74.1       
 [97] knitr_1.50              
 [98] edgeR_4.4.2             
 [99] xfun_0.51               
[100] statmod_1.5.0           
[101] DT_0.33                 
[102] UCSC.utils_1.2.0        
[103] lazyeval_0.2.2          
[104] yaml_2.3.10             
[105] shinyWidgets_0.9.0      
[106] evaluate_1.0.3          
[107] codetools_0.2-20        
[108] tibble_3.2.1            
[109] BiocManager_1.30.25     
[110] cli_3.6.3               
[111] xtable_1.8-4            
[112] jquerylib_0.1.4         
[113] munsell_0.5.1           
[114] golem_0.5.1             
[115] Rcpp_1.0.14             
[116] dbplyr_2.5.0            
[117] png_0.1-8               
[118] XML_3.99-0.18           
[119] blob_1.2.4              
[120] dotCall64_1.2           
[121] bitops_1.0-9            
[122] viridisLite_0.4.2       
[123] scales_1.3.0            
[124] purrr_1.0.4             
[125] crayon_1.5.3            
[126] cowplot_1.1.3           
[127] KEGGREST_1.46.0         
> sessioninfo::session_info()
─ Session info ──────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sonoma 14.7.1
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2025-03-23
 pandoc   NA
 quarto   1.6.40 @ /Applications/Positron.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ──────────────────────
 !  package              * version   date (UTC) lib source
    abind                  1.4-8     2024-09-12 [1] CRAN (R 4.4.1)
    AnnotationDbi          1.68.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    AnnotationHub          3.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    attempt                0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
    beachmat               2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    beeswarm               0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
    benchmarkme            1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
    benchmarkmeData        1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocFileCache          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocIO                 1.16.0    2024-12-12 [1] Github (Bioconductor/BiocIO@8e80624)
    BiocManager            1.30.25   2024-08-28 [1] CRAN (R 4.4.1)
    BiocNeighbors          2.0.1     2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
    BiocParallel         * 1.40.0    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    BiocSingular           1.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocVersion            3.20.0    2024-05-04 [1] Bioconductor 3.20 (R 4.4.0)
    Biostrings             2.74.1    2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
    bit                    4.6.0     2025-03-06 [1] CRAN (R 4.4.1)
    bit64                  4.6.0-1   2025-01-16 [1] CRAN (R 4.4.1)
    bitops                 1.0-9     2024-10-03 [1] CRAN (R 4.4.1)
    blob                   1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
    bslib                  0.9.0     2025-01-30 [1] CRAN (R 4.4.1)
    cachem                 1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
 VP cli                    3.6.3     2025-02-13 [2] CRAN (R 4.4.1) (on disk 3.6.4)
    codetools              0.2-20    2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1     2024-07-26 [1] CRAN (R 4.4.0)
    config                 0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
    cowplot                1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
    crayon                 1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
    curl                   6.2.1     2025-02-19 [1] CRAN (R 4.4.1)
    data.table           * 1.17.0    2025-02-22 [1] CRAN (R 4.4.1)
    DBI                    1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
    dbplyr                 2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    digest                 0.6.37    2024-08-19 [1] CRAN (R 4.4.1)
    doParallel             1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
    dotCall64              1.2       2024-10-04 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
    DT                     0.33      2024-04-04 [1] CRAN (R 4.4.0)
    edgeR                  4.4.2     2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    evaluate               1.0.3     2025-01-10 [1] CRAN (R 4.4.1)
    ExperimentHub          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    fastmap                1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
    fields                 16.3.1    2025-03-08 [1] CRAN (R 4.4.1)
    filelock               1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
    foreach                1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
    generics               0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
    GenomeInfoDb         * 1.42.3    2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13    2024-12-12 [1] Bioconductor
    GenomicAlignments      1.42.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    GenomicRanges        * 1.58.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2              * 3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
    ggrepel                0.9.6     2024-09-07 [1] CRAN (R 4.4.1)
    ggtext               * 0.1.2     2022-09-16 [1] CRAN (R 4.4.0)
    glue                   1.8.0     2024-09-30 [1] CRAN (R 4.4.1)
    golem                  0.5.1     2024-08-27 [1] CRAN (R 4.4.1)
    gprofiler2           * 0.2.3     2024-02-23 [1] CRAN (R 4.4.0)
    gridExtra            * 2.3       2017-09-09 [1] CRAN (R 4.4.0)
    gridtext               0.1.5     2022-09-16 [1] CRAN (R 4.4.0)
    gtable                 0.3.6     2024-10-25 [1] CRAN (R 4.4.1)
    here                   1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
    htmlwidgets            1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
    httpuv                 1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
    httr                   1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1    2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
    iterators              1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
    jquerylib              0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
    jsonlite               1.9.1     2025-03-03 [1] CRAN (R 4.4.1)
    KEGGREST               1.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    knitr                  1.50      2025-03-16 [1] CRAN (R 4.4.2)
    later                  1.4.1     2024-11-27 [1] CRAN (R 4.4.1)
    lattice                0.22-6    2024-03-20 [1] CRAN (R 4.4.3)
    lazyeval               0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
    lifecycle              1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
    limma                  3.62.2    2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    locfit                 1.5-9.12  2025-03-05 [1] CRAN (R 4.4.1)
    magick                 2.8.5     2024-09-20 [1] CRAN (R 4.4.1)
    magrittr               2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
    maps                   3.4.2.1   2024-11-10 [1] CRAN (R 4.4.1)
    Matrix                 1.7-2     2025-01-23 [1] CRAN (R 4.4.3)
    MatrixGenerics       * 1.18.1    2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0     2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
    mime                   0.12      2021-09-28 [1] CRAN (R 4.4.0)
    munsell                0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
    paletteer              1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
    parallelly           * 1.42.0    2025-01-30 [1] CRAN (R 4.4.1)
    pillar                 1.10.1    2025-01-07 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
    plotly                 4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
    png                    0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
    promises               1.3.2     2024-11-28 [1] CRAN (R 4.4.1)
    purrr                  1.0.4     2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1     2025-02-15 [1] CRAN (R 4.4.1)
    rappdirs               0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
    RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14    2025-01-12 [1] CRAN (R 4.4.1)
    RCurl                  1.98-1.16 2024-07-11 [1] CRAN (R 4.4.0)
    rematch2               2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
    restfulr               0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
    rjson                  0.2.23    2024-09-16 [1] CRAN (R 4.4.1)
 P  rlang                * 1.1.5     2025-01-17 [2] CRAN (R 4.4.1)
    rprojroot              2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
    Rsamtools              2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    RSQLite                2.3.9     2024-12-03 [1] CRAN (R 4.4.1)
    rsvd                   1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
    rtracklayer            1.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Arrays               1.6.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sass                   0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
    ScaledMatrix           1.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
    scater                 1.34.1    2025-03-03 [1] Bioconductor 3.20 (R 4.4.2)
    scuttle                1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    sessioninfo            1.2.3     2025-02-05 [1] CRAN (R 4.4.1)
    shiny                  1.10.0    2024-12-14 [1] CRAN (R 4.4.1)
    shinyWidgets           0.9.0     2025-02-21 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    spam                   2.11-1    2025-01-20 [1] CRAN (R 4.4.1)
    SparseArray            1.6.2     2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    SpatialExperiment    * 1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    spatialLIBD          * 1.18.0    2024-11-07 [1] Bioconductor 3.20 (R 4.4.1)
    statmod                1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    vctrs                  0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
    viridis                0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
    viridisLite            0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
    withr                  3.0.2     2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.51      2025-02-19 [1] CRAN (R 4.4.1)
    XML                    3.99-0.18 2025-01-01 [1] CRAN (R 4.4.1)
    xml2                   1.3.6     2023-12-04 [1] CRAN (R 4.4.0)
    xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10    2024-07-26 [1] CRAN (R 4.4.0)
    zlibbioc               1.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

─────────────────────────────────