---
title: "06c-Spot correlations of hypomap visium VMH to kim 19 AP smartseq"
author: "Bernie Mulvey"
date: "2025-04-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(SpatialExperiment)
library(spatialLIBD)
library(gprofiler2)
library(Seurat)
library(dplyr)

# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(parallel)

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```


load spatial data, convert to SPE;
obtained file from https://www.repository.cam.ac.uk/bitstreams/a47399a8-2735-4b65-b6cb-3cd1a48b08d7/download
```{r}
hmap <- readRDS("~/Desktop/humanHYPOMAP_spatial.rds")

## convert to SPE using code from https://github.com/drighelli/SpatialExperiment/issues/115 
seurat_to_spe <- function(seu, sample_id, img_id) {
    
    ## Convert to SCE
    sce <- Seurat::as.SingleCellExperiment(seu)
    
    ## Extract spatial coordinates
    spatialCoords <- as.matrix(
        seu@images[[img_id]]@coordinates[, c("imagecol", "imagerow")])
    
    ## Extract and process image data
    img <- SpatialExperiment::SpatialImage(
        x = as.raster(seu@images[[img_id]]@image))
    
    imgData <- DataFrame(
        sample_id = sample_id,
        image_id = img_id,
        data = I(list(img)),
        scaleFactor = seu@images[[img_id]]@scale.factors$lowres)
    
    # Convert to SpatialExperiment
    spe <- SpatialExperiment(
        assays = assays(sce),
        rowData = rowData(sce),
        colData = colData(sce),
        metadata = metadata(sce),
        reducedDims = reducedDims(sce),
        altExps = altExps(sce),
        sample_id = sample_id,
        spatialCoords = spatialCoords,
        imgData = imgData
    )
    # indicate all spots are on the tissue since the data is already processed
    spe$in_tissue <- 1
    spe$sample_id <- sample_id
    # Return Spatial Experiment object
    spe
}


## sample_ids and image_ids are permutations of the following:
## sampleid = hmap$captureArea (b2,a3,...)
## image id is a permutation of this (slice2B, slice3A..)


spat_metas <- list(c("b2","slice2B"),
                   c("a3","slice3A"),
                   c("b4","slice4B"),
                   c("a5","slice5A"),
                   c("b6","slice6B"),
                   c("a7","slice7A"),
                   c("b8","slice8B"),
                   c("b1_b","sliceB1B"),
                   c("c1_a","sliceC1A"))
names(spat_metas) <- unlist(lapply(spat_metas,function(x){x[2]}))

spe_ls <- lapply(spat_metas, function(m) {
    seurat_to_spe(seu = hmap[,hmap$captureArea==m[1]], sample_id = m[1], img_id = m[2])
})

spe <- Reduce(cbind, spe_ls)
rm(hmap,spat_metas,spe_ls)
gc(full=T)
```

## load the smartseq AP marker test results from both approaches used for vVMH (spatial_HYP/code/12/06a,06b)
```{r}
fig2mks <- fread("processed-data/12-Mouse comparisons/06a-Kim19msVMH_Fig2_APclustsOnly_smartseq_markerTstat.txt")

ss.apmk <- fread("processed-data/12-Mouse comparisons/06b-AllKim19_sFISH-matched-smartseq-clusts_mkTstats_by_peakAPloc.txt")
``` 

## subset spe to VMH and shared 1-1 ortholog genes (specifically those we also had for comparing to vVMH--already the only ones in these tables). we'll probably have to do two separate sets since different numbers of genes pass marker analysis minima in 06a and 06b
```{r}
vmhspots <- unique(c(
    grep(spe$regional_clusters_named,pattern="VMH",value=T),grep(spe$regional_clusters_grouped,pattern="VMH",value=T))) 
# oh, they're all just called VMH in both cols.
rm(vmhspots)

spe <- spe[,spe$regional_clusters_named=="VMH"|spe$regional_clusters_grouped=="VMH"]

# the hypomap spatial only has gene symbols, no ENSGs
# get genes that are in at least one of the 2 mouse marker analyses and hypomap spat VMH
orths <- fread("processed-data/12-Mouse comparisons/06a-1to1_orths_in_KimSmartseq_and_vVMH.txt")
keepg <- orths[ortholog_name %in% rownames(spe) & orths$ortholog_ensg %in% c(fig2mks$human_ensg,ss.apmk$human_ensg),.(ortholog_name,ortholog_ensg)]
spe <- spe[keepg$ortholog_name,]

## get spot logcounts for vmh spots only; when we merge in the t stats, nonshared genes will drop out so no need for two separate logcount tables
lct <- as.data.table(as.data.frame(as.matrix(logcounts(spe))),keep.rownames=T)
## lastly, the column names here are just barcode-1_. , which is useless for identifying samples. so before we dump off the SPE, get the colname-sample_id pairs
sampcn <- as.data.table(colData(spe),keep.rownames=T)[,.(rn,sample_id)]
setnames(sampcn,"sample_id","samp")

## dont need the spe
rm(spe)

# append ortholog info so we can match by human ensg mappings for consistency with prior analyses
lct <- merge.data.table(lct,keepg,by.x="rn",by.y="ortholog_name")
lct[,rn:=NULL]
setnames(lct,"ortholog_ensg","rn")
lctreo <- c("rn",names(lct)[!(names(lct)=="rn")])
lct <- lct[,..lctreo]

# divy up the data so we don't have to pass the whole logcount table to each worker
spots <- names(lct)[2:ncol(lct)]
spotsl <- lapply(spots,function(s){
    getcol <- c("rn",s)
    o<- lct[,..getcol]
    return(o)
})
names(spotsl) <- spots
rm(spots,lct,lctreo)
gc(full=T)

## run it
setDTthreads(threads=2,restore_after_fork = F) # prevent resource overcommiting; restore after fork set to F per their docs on misbehavior on mac

spotcor.list <- lapply(list(fig2mks,ss.apmk),FUN=function(m){
    is <- c(1:length(spotsl))
    spotcor <- mcmapply(s=spotsl,
                        i=is,
                        n=names(spotsl),
                        mc.cores=4,
                        SIMPLIFY=FALSE,
                        FUN=function(s,i,n){
      setDTthreads(2,restore_after_fork = T) # since DT is not too nice about mac implementation, put it here too, and assuming some forking happens during loop execution because i dont get when it happens. and they provide no context for when restore_after_fork does/doesn't work as expected on mac so taking a wild guess tehre

      tmpd <- merge.data.table(s,m,by.x="rn",by.y="human_ensg")
      tmpd[,gene:=NULL]
      sdnames <- grep(names(tmpd),pattern="rn|-1",value=T,invert=T)
    
      ans <- tmpd[,lapply(.SD,FUN=function(x){cor(get(n),x,method="spearman")}),.SDcols=sdnames]
      ans[,spot:=n] # temporarily, just to rotate around; this is a chonk column so we'll drop it agn
      ans <- melt.data.table(ans,id.vars="spot")
      ## unlike with our data, we DO need to keep the spot ID to match back to samples so don't toss the column this time
    
      return(ans)
    })
})

## assemble each to a table
spotcor.list <- lapply(spotcor.list,FUN=function(s){rbindlist(s)})
```


## get summary metrics per sample
```{r}
spotcor.list.summary <- lapply(spotcor.list,FUN=function(s){
    s <- merge.data.table(s,sampcn,by.x="spot",by.y="rn")
    s[,spot:=NULL]
    s[,sampMean:=mean(value,na.rm=T),by=c("samp","variable")]
    s[,sampMed:=median(value,na.rm=T),by=c("samp","variable")]
    s[,topquint:=value>quantile(value,0.8),by=c("samp","variable")]
    s[,botquint:=value<quantile(value,0.2),by=c("samp","variable")]
    s[topquint==T,sampMean_topquint:=mean(value,na.rm=T),by=c("samp","variable")]
    s[botquint==T,sampMean_botquint:=mean(value,na.rm=T),by=c("samp","variable")]
    s[,topquint:=NULL]
    s[,botquint:=NULL]
    
    # drop spot values and get the summary metrics
    tmp <- copy(s)
    tmp[,value:=NULL]
    tmp[,rn:=NULL]
    spotcor.b <- unique(tmp[!is.na(sampMean_botquint)])
    spotcor.t <- unique(tmp[!is.na(sampMean_topquint)])
    spotcor.ap2 <- merge.data.table(
        spotcor.b[,.SD,.SDcols=names(spotcor.b)!="sampMean_topquint"],
        spotcor.t[,.SD,.SDcols=c("samp","variable","sampMean_topquint")],
                              by=c("samp","variable"))
    return(spotcor.ap2)

})

names(spotcor.list.summary) <- c("by_smrtseq_clust","by_peakAPloc")
saveRDS(spotcor.list.summary,
        file="processed-data/12-Mouse comparisons/06c-smrtseqClustMkTstat_smrtseqPeakAPgroupMkTstat_spotcorrelations_hypomapVisVMH.RDS")
```

well, story's actually pretty similar here, except the sample-level summary metrics remain positive in the 6th of 7 A->P sections ("ff") in the by-AP peak cell groupings. not the case in our data.

repinf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.7.4

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] dplyr_1.1.4                 Seurat_5.2.1                SeuratObject_5.0.2         
 [4] sp_2.2-0                    parallelly_1.43.0           colorout_1.3-0.2           
 [7] gprofiler2_0.2.3            spatialLIBD_1.18.0          SpatialExperiment_1.16.0   
[10] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0 Biobase_2.66.0             
[13] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3         IRanges_2.40.1             
[16] S4Vectors_0.44.0            BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[19] matrixStats_1.5.0           gridExtra_2.3               ggtext_0.1.2               
[22] ggplot2_3.5.1               data.table_1.17.0           rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.1-0    bitops_1.0-9             httr_1.4.7              
  [4] RColorBrewer_1.1-3       doParallel_1.0.17        tools_4.4.3             
  [7] sctransform_0.4.1        R6_2.6.1                 DT_0.33                 
 [10] uwot_0.2.3               lazyeval_0.2.2           withr_3.0.2             
 [13] progressr_0.15.1         cli_3.6.4                spatstat.explore_3.4-2  
 [16] fastDummies_1.7.5        labeling_0.4.3           sass_0.4.9              
 [19] spatstat.data_3.1-6      ggridges_0.5.6           pbapply_1.7-2           
 [22] Rsamtools_2.22.0         scater_1.34.1            sessioninfo_1.2.3       
 [25] attempt_0.3.1            maps_3.4.2.1             limma_3.62.2            
 [28] rstudioapi_0.17.1        RSQLite_2.3.9            generics_0.1.3          
 [31] BiocIO_1.16.0            spatstat.random_3.3-3    ica_1.0-3               
 [34] Matrix_1.7-3             ggbeeswarm_0.7.2         abind_1.4-8             
 [37] lifecycle_1.0.4          yaml_2.3.10              edgeR_4.4.2             
 [40] SparseArray_1.6.2        BiocFileCache_2.14.0     Rtsne_0.17              
 [43] paletteer_1.6.0          grid_4.4.3               blob_1.2.4              
 [46] promises_1.3.2           ExperimentHub_2.14.0     crayon_1.5.3            
 [49] miniUI_0.1.1.1           lattice_0.22-6           beachmat_2.22.0         
 [52] cowplot_1.1.3            KEGGREST_1.46.0          magick_2.8.6            
 [55] pillar_1.10.1            knitr_1.50               rjson_0.2.23            
 [58] future.apply_1.11.3      codetools_0.2-20         glue_1.8.0              
 [61] spatstat.univar_3.1-2    vctrs_0.6.5              png_0.1-8               
 [64] spam_2.11-1              gtable_0.3.6             rematch2_2.1.2          
 [67] cachem_1.1.0             xfun_0.51                S4Arrays_1.6.0          
 [70] mime_0.13                survival_3.8-3           iterators_1.0.14        
 [73] fields_16.3.1            statmod_1.5.0            fitdistrplus_1.2-2      
 [76] ROCR_1.0-11              nlme_3.1-167             bit64_4.6.0-1           
 [79] filelock_1.0.3           RcppAnnoy_0.0.22         rprojroot_2.0.4         
 [82] bslib_0.9.0              irlba_2.3.5.1            vipor_0.4.7             
 [85] KernSmooth_2.23-26       colorspace_2.1-1         DBI_1.2.3               
 [88] tidyselect_1.2.1         bit_4.6.0                compiler_4.4.3          
 [91] curl_6.2.2               BiocNeighbors_2.0.1      xml2_1.3.8              
 [94] DelayedArray_0.32.0      plotly_4.10.4            rtracklayer_1.66.0      
 [97] scales_1.3.0             lmtest_0.9-40            rappdirs_0.3.3          
[100] goftest_1.2-3            stringr_1.5.1            digest_0.6.37           
[103] spatstat.utils_3.1-3     rmarkdown_2.29           benchmarkmeData_1.0.4   
[106] XVector_0.46.0           htmltools_0.5.8.1        pkgconfig_2.0.3         
[109] dbplyr_2.5.0             fastmap_1.2.0            htmlwidgets_1.6.4       
[112] UCSC.utils_1.2.0         shiny_1.10.0             farver_2.1.2            
[115] jquerylib_0.1.4          zoo_1.8-13               jsonlite_1.9.1          
[118] BiocParallel_1.40.0      config_0.3.2             BiocSingular_1.22.0     
[121] RCurl_1.98-1.17          magrittr_2.0.3           scuttle_1.16.0          
[124] GenomeInfoDbData_1.2.13  dotCall64_1.2            patchwork_1.3.0         
[127] munsell_0.5.1            Rcpp_1.0.14              viridis_0.6.5           
[130] reticulate_1.41.0.1      stringi_1.8.4            zlibbioc_1.52.0         
[133] MASS_7.3-65              AnnotationHub_3.14.0     plyr_1.8.9              
[136] listenv_0.9.1            ggrepel_0.9.6            deldir_2.0-4            
[139] Biostrings_2.74.1        splines_4.4.3            tensor_1.5              
[142] gridtext_0.1.5           locfit_1.5-9.12          igraph_2.1.4            
[145] spatstat.geom_3.3-6      RcppHNSW_0.6.0           pkgload_1.4.0           
[148] reshape2_1.4.4           ScaledMatrix_1.14.0      BiocVersion_3.20.0      
[151] XML_3.99-0.18            evaluate_1.0.3           golem_0.5.1             
[154] BiocManager_1.30.25      foreach_1.5.2            httpuv_1.6.15           
[157] polyclip_1.10-7          RANN_2.6.2               tidyr_1.3.1             
[160] purrr_1.0.4              future_1.34.0            benchmarkme_1.0.8       
[163] scattermore_1.2          rsvd_1.0.5               xtable_1.8-4            
[166] restfulr_0.0.15          RSpectra_0.16-2          later_1.4.1             
[169] viridisLite_0.4.2        tibble_3.2.1             memoise_2.0.1           
[172] beeswarm_0.4.0           AnnotationDbi_1.68.0     GenomicAlignments_1.42.0
[175] cluster_2.1.8.1          shinyWidgets_0.9.0       globals_0.16.3          
[178] here_1.0.1              
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sonoma 14.7.4
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-04-09
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ────────────────────────────────────────────────────────────────────────────
 ! package              * version   date (UTC) lib source
   abind                  1.4-8     2024-09-12 [1] CRAN (R 4.4.1)
   AnnotationDbi          1.68.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   AnnotationHub          3.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   attempt                0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
   beachmat               2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   beeswarm               0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme            1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData        1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
   Biobase              * 2.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocFileCache          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocGenerics         * 0.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocIO                 1.16.0    2024-12-12 [1] Github (Bioconductor/BiocIO@8e80624)
   BiocManager            1.30.25   2024-08-28 [1] CRAN (R 4.4.1)
   BiocNeighbors          2.0.1     2024-11-28 [1] Bioconductor 3.20 (R 4.4.2)
   BiocParallel           1.40.0    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
   BiocSingular           1.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   BiocVersion            3.20.0    2024-05-04 [1] Bioconductor 3.20 (R 4.4.0)
   Biostrings             2.74.1    2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
   bit                    4.6.0     2025-03-06 [1] CRAN (R 4.4.1)
   bit64                  4.6.0-1   2025-01-16 [1] CRAN (R 4.4.1)
   bitops                 1.0-9     2024-10-03 [1] CRAN (R 4.4.1)
   blob                   1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   bslib                  0.9.0     2025-01-30 [1] CRAN (R 4.4.1)
   cachem                 1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
 P cli                    3.6.4     2025-02-13 [2] CRAN (R 4.4.1)
   cluster                2.1.8.1   2025-03-12 [1] CRAN (R 4.4.1)
   codetools              0.2-20    2024-03-31 [1] CRAN (R 4.4.3)
   colorout             * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace             2.1-1     2024-07-26 [1] CRAN (R 4.4.0)
   config                 0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
   crayon                 1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   curl                   6.2.2     2025-03-24 [1] CRAN (R 4.4.1)
   data.table           * 1.17.0    2025-02-22 [1] CRAN (R 4.4.1)
   DBI                    1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                 2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray           0.32.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   deldir                 2.0-4     2024-02-28 [1] CRAN (R 4.4.0)
   digest                 0.6.37    2024-08-19 [1] CRAN (R 4.4.1)
   doParallel             1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64              1.2       2024-10-04 [1] CRAN (R 4.4.1)
   dplyr                * 1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   DT                     0.33      2024-04-04 [1] CRAN (R 4.4.0)
   edgeR                  4.4.2     2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
   evaluate               1.0.3     2025-01-10 [1] CRAN (R 4.4.1)
   ExperimentHub          2.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   farver                 2.1.2     2024-05-13 [1] CRAN (R 4.4.0)
   fastDummies            1.7.5     2025-01-20 [1] CRAN (R 4.4.1)
   fastmap                1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fields                 16.3.1    2025-03-08 [1] CRAN (R 4.4.1)
   filelock               1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   fitdistrplus           1.2-2     2025-01-07 [1] CRAN (R 4.4.1)
   foreach                1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
   future                 1.34.0    2024-07-29 [1] CRAN (R 4.4.0)
   future.apply           1.11.3    2024-10-27 [1] CRAN (R 4.4.1)
   generics               0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb         * 1.42.3    2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
   GenomeInfoDbData       1.2.13    2024-12-12 [1] Bioconductor
   GenomicAlignments      1.42.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   GenomicRanges        * 1.58.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   ggbeeswarm             0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2              * 3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrepel                0.9.6     2024-09-07 [1] CRAN (R 4.4.1)
   ggridges               0.5.6     2024-01-23 [1] CRAN (R 4.4.0)
   ggtext               * 0.1.2     2022-09-16 [1] CRAN (R 4.4.0)
   globals                0.16.3    2024-03-08 [1] CRAN (R 4.4.0)
   glue                   1.8.0     2024-09-30 [1] CRAN (R 4.4.1)
   goftest                1.2-3     2021-10-07 [1] CRAN (R 4.4.0)
   golem                  0.5.1     2024-08-27 [1] CRAN (R 4.4.1)
   gprofiler2           * 0.2.3     2024-02-23 [1] CRAN (R 4.4.0)
   gridExtra            * 2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gridtext               0.1.5     2022-09-16 [1] CRAN (R 4.4.0)
   gtable                 0.3.6     2024-10-25 [1] CRAN (R 4.4.1)
   here                   1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools              0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets            1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                 1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
   httr                   1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   ica                    1.0-3     2022-07-08 [1] CRAN (R 4.4.0)
   igraph                 2.1.4     2025-01-23 [1] CRAN (R 4.4.1)
   IRanges              * 2.40.1    2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
   irlba                  2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   iterators              1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
   jquerylib              0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite               1.9.1     2025-03-03 [1] CRAN (R 4.4.1)
   KEGGREST               1.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   KernSmooth             2.23-26   2025-01-01 [1] CRAN (R 4.4.3)
   knitr                  1.50      2025-03-16 [1] CRAN (R 4.4.2)
   labeling               0.4.3     2023-08-29 [1] CRAN (R 4.4.0)
   later                  1.4.1     2024-11-27 [1] CRAN (R 4.4.1)
   lattice                0.22-6    2024-03-20 [1] CRAN (R 4.4.3)
   lazyeval               0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle              1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                  3.62.2    2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
   listenv                0.9.1     2024-01-29 [1] CRAN (R 4.4.0)
   lmtest                 0.9-40    2022-03-21 [1] CRAN (R 4.4.0)
   locfit                 1.5-9.12  2025-03-05 [1] CRAN (R 4.4.1)
   magick                 2.8.6     2025-03-23 [1] CRAN (R 4.4.1)
   magrittr               2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   maps                   3.4.2.1   2024-11-10 [1] CRAN (R 4.4.1)
   MASS                   7.3-65    2025-02-28 [1] CRAN (R 4.4.1)
   Matrix                 1.7-3     2025-03-11 [1] CRAN (R 4.4.1)
   MatrixGenerics       * 1.18.1    2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
   matrixStats          * 1.5.0     2025-01-07 [1] CRAN (R 4.4.1)
   memoise                2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   mime                   0.13      2025-03-17 [1] CRAN (R 4.4.1)
   miniUI                 0.1.1.1   2018-05-18 [1] CRAN (R 4.4.0)
   munsell                0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   nlme                   3.1-167   2025-01-27 [1] CRAN (R 4.4.3)
   paletteer              1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
   parallelly           * 1.43.0    2025-03-24 [1] CRAN (R 4.4.1)
   patchwork              1.3.0     2024-09-16 [1] CRAN (R 4.4.1)
   pbapply                1.7-2     2023-06-27 [1] CRAN (R 4.4.0)
   pillar                 1.10.1    2025-01-07 [1] CRAN (R 4.4.1)
   pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   pkgload                1.4.0     2024-06-28 [1] CRAN (R 4.4.1)
   plotly                 4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
   plyr                   1.8.9     2023-10-02 [1] CRAN (R 4.4.0)
   png                    0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   polyclip               1.10-7    2024-07-23 [1] CRAN (R 4.4.0)
   progressr              0.15.1    2024-11-22 [1] CRAN (R 4.4.1)
   promises               1.3.2     2024-11-28 [1] CRAN (R 4.4.1)
   purrr                  1.0.4     2025-02-05 [1] CRAN (R 4.4.1)
   R6                     2.6.1     2025-02-15 [1] CRAN (R 4.4.1)
   RANN                   2.6.2     2024-08-25 [1] CRAN (R 4.4.1)
   rappdirs               0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                   1.0.14    2025-01-12 [1] CRAN (R 4.4.1)
   RcppAnnoy              0.0.22    2024-01-23 [1] CRAN (R 4.4.0)
   RcppHNSW               0.6.0     2024-02-04 [1] CRAN (R 4.4.0)
   RCurl                  1.98-1.17 2025-03-22 [1] CRAN (R 4.4.1)
   rematch2               2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
   reshape2               1.4.4     2020-04-09 [1] CRAN (R 4.4.0)
   restfulr               0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
   reticulate             1.41.0.1  2025-03-09 [1] CRAN (R 4.4.1)
   rjson                  0.2.23    2024-09-16 [1] CRAN (R 4.4.1)
 P rlang                * 1.1.5     2025-01-17 [2] CRAN (R 4.4.1)
   rmarkdown              2.29      2024-11-04 [1] CRAN (R 4.4.1)
   ROCR                   1.0-11    2020-05-02 [1] CRAN (R 4.4.0)
   rprojroot              2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools              2.22.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   RSpectra               0.16-2    2024-07-18 [1] CRAN (R 4.4.0)
   RSQLite                2.3.9     2024-12-03 [1] CRAN (R 4.4.1)
   rstudioapi             0.17.1    2024-10-22 [1] CRAN (R 4.4.1)
   rsvd                   1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer            1.66.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   Rtsne                  0.17      2023-12-07 [1] CRAN (R 4.4.0)
   S4Arrays               1.6.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   S4Vectors            * 0.44.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   sass                   0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix           1.14.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   scales                 1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater                 1.34.1    2025-03-03 [1] Bioconductor 3.20 (R 4.4.2)
   scattermore            1.2       2023-06-12 [1] CRAN (R 4.4.0)
   sctransform            0.4.1     2023-10-19 [1] CRAN (R 4.4.0)
   scuttle                1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   sessioninfo            1.2.3     2025-02-05 [1] CRAN (R 4.4.1)
   Seurat               * 5.2.1     2025-01-24 [1] CRAN (R 4.4.1)
   SeuratObject         * 5.0.2     2024-05-08 [1] CRAN (R 4.4.0)
   shiny                  1.10.0    2024-12-14 [1] CRAN (R 4.4.1)
   shinyWidgets           0.9.0     2025-02-21 [1] CRAN (R 4.4.1)
   SingleCellExperiment * 1.28.1    2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
   sp                   * 2.2-0     2025-02-01 [1] CRAN (R 4.4.1)
   spam                   2.11-1    2025-01-20 [1] CRAN (R 4.4.1)
   SparseArray            1.6.2     2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
   SpatialExperiment    * 1.16.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   spatialLIBD          * 1.18.0    2024-11-07 [1] Bioconductor 3.20 (R 4.4.3)
   spatstat.data          3.1-6     2025-03-17 [1] CRAN (R 4.4.2)
   spatstat.explore       3.4-2     2025-03-21 [1] CRAN (R 4.4.1)
   spatstat.geom          3.3-6     2025-03-18 [1] CRAN (R 4.4.1)
   spatstat.random        3.3-3     2025-03-19 [1] CRAN (R 4.4.1)
   spatstat.sparse        3.1-0     2024-06-21 [1] CRAN (R 4.4.0)
   spatstat.univar        3.1-2     2025-03-05 [1] CRAN (R 4.4.1)
   spatstat.utils         3.1-3     2025-03-15 [1] CRAN (R 4.4.1)
   statmod                1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   stringi                1.8.4     2024-05-06 [1] CRAN (R 4.4.0)
   stringr                1.5.1     2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment * 1.36.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   survival               3.8-3     2024-12-17 [1] CRAN (R 4.4.3)
   tensor                 1.5       2012-05-05 [1] CRAN (R 4.4.0)
   tibble                 3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                  1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect             1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   UCSC.utils             1.2.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   uwot                   0.2.3     2025-02-24 [1] CRAN (R 4.4.1)
   vctrs                  0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                  0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite            0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   withr                  3.0.2     2024-10-28 [1] CRAN (R 4.4.1)
   xfun                   0.51      2025-02-19 [1] CRAN (R 4.4.1)
   XML                    3.99-0.18 2025-01-01 [1] CRAN (R 4.4.1)
   xml2                   1.3.8     2025-03-14 [1] CRAN (R 4.4.1)
   xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
   XVector                0.46.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   yaml                   2.3.10    2024-07-26 [1] CRAN (R 4.4.0)
   zlibbioc               1.52.0    2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
   zoo                    1.8-13    2025-02-22 [1] CRAN (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 P ── Loaded and on-disk path mismatch.

───────────────────────────────────────────────────────────────────────────────────────
