---
title: "01-Macaque VMH-DMH snRNAseq"
author: "Bernard Mulvey"
date: "2023-08-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(ggplot2)
library(gridExtra)
library(SpatialExperiment)
library(spatialLIBD)
require(colorout)
library(job)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))
```

```{r}
job({hyp2 <- readRDS("data/processing/hyp_umi600_gene450_chrm35_lognorm_D07-075_B1 crop 070223.RDS")
    job::export(hyp2)})

bscl <- fread("analysis/data/spe_053123/H02-clustering/03-bs_harmony_hvg1020_nnsvg1020_9-15-20-31_out/BSpace_Harmony_nnsvg10pctile_15.txt")

bscl <- bscl[rn %in% colnames(hyp2)]
setnames(bscl,2,"clust")
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]
colLabels(hyp2) <- as.factor(paste0("X",bscl$clust))


### set rownames to gene symbols for compatibility with single cell from different species (different ENS gene IDs)
rownames(hyp2) <- rowData(hyp2)$gene_name
```

Load single cell
```{r}
primate <- DropletUtils::read10xCounts("~/Desktop/Macaque VMH DMH Geo/",col.names = T)
# set rownames to gene symbol for cross species compatibility
rownames(primate) <- rowData(primate)$Symbol

### append further metadata to coldata
pmeta <- fread("~/Desktop/Macaque VMH DMH Geo/metadata.csv")
pmeta <- DataFrame(pmeta)
rownames(pmeta) <- pmeta$ID
# set a different name for the actual sample id
names(pmeta)[2] <- "sample_id"
# drop cells not iin the metadata and append
primate <- primate[,colnames(primate) %in% pmeta$ID]
pmeta <- pmeta[colnames(primate),]
stopifnot(identical(rownames(pmeta),colnames(primate)))
colData(primate) <- cbind(colData(primate),pmeta)

### create a cluster ID that tells us cell cluster, type, neuron cluster if applicable, VMH cluster if applicable, and conserved cluster if applicable
# this notation doesn't work with an S4 component so make data  table. still saves a TON of characters!
# colData(primate)$cl_type_neutype_VMHcl <- as.data.table(colData(primate)) %$% paste(Cell_cluster,Cell_type,Neuron_cluster,VMH_cluster,sep="_")
# colData(primate)$cl_type_neutype_VMHcl <- as.factor(colData(primate)$cl_type_neutype_VMHcl)

### there are multiple rows for genes with the same symbol, so drop those
primate2 <- primate[unique(rownames(primate)),]
nrow(primate2)
## spot check: rownames have updated synchronously
# rownames(rowData(primate2))[100]
# rownames(primate2)[100]
primate3 <- primate2[,!is.na(colData(primate2)$VMH_cluster)]
dim(primate3)
table(primate3$VMH_cluster,primate3$sample_id)
# at least 40 cells per cluster per sample
primate3$VMH_cluster <- as.factor(primate3$VMH_cluster)
```


### run pseudobulk enrichment tasks
```{r}

    hyp_model_res <- registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_name",gene_name = "gene_name")



sce_model_res <- registration_wrapper(primate3,var_registration="VMH_cluster",var_sample_id = "sample_id",gene_ensembl = "Symbol",gene_name = "Symbol",min_ncells = 10)

registration_stats <- sce_model_res$enrichment[,grep(colnames(sce_model_res$enrichment),pattern="^t_stat",value=T)]
colnames(registration_stats) %<>% gsub(.,pattern="^t_stat_",replacement="")


```

### and with collapsed VMH clusters (5+6)
```{r}
hyp3 <- hyp2
hyp3$label <- gsub(hyp3$label,pattern="X5|X6",replacement="VMH")
hyp3$label <- as.factor(hyp3$label)

hyp3_model_res <- registration_wrapper(hyp3,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_name",gene_name = "gene_name")
```
### now interestingly, the primate samples were female, what if we only look at female?
```{r}
demos <- fread("raw-data/demos.txt")
demos[1,]
cdtmp <- as.data.table(colData(hyp3))
cdtmp <- merge.data.table(cdtmp,demos,by.x="brnum",by.y="BrNum")
cdtmp <- DataFrame(cdtmp)
rownames(cdtmp) <- cdtmp$key
cdtmp <- cdtmp[colnames(hyp3),]
colData(hyp3) <- cdtmp

hyp4 <- hyp3[,hyp3$Sex=="F"]
hyp4_model_res <- registration_wrapper(hyp4,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_name",gene_name = "gene_name")

cor_hyp4clus <- spatialLIBD::layer_stat_cor(stats=registration_stats,modeling_results = hyp4_model_res,model_type = "enrichment",top_n =4000)

# test
layer_stat_cor_plot(cor_hyp4clus,max(cor_hyp4clus))
title("Top 4k genes")
dev.off()
```

plot
```{r}
j<-c(25,50,75,100,250,500,750,1000)
i<-1
for (i in c(1:8)){
    curj <- j[i]
    cor_hypclus <- spatialLIBD::layer_stat_cor(stats = registration_stats,modeling_results = hyp_model_res,model_type = "enrichment",top_n = curj)
    pdf(paste0("plots/spe_053123/H06-Spatial registration/","Hyp Harmon nnsvg10 BS15 clust to Macaque VMH neurons top",j[i],".pdf"),height=5,width=6)
    layer_stat_cor_plot(cor_hypclus,max=max(cor_hypclus))
    title(paste0("BS15 clust to Macaque VMH neurons top",j[i],"genes"))
    dev.off()
    rm(cor_hypclus)
    
    cor_hypcol <- spatialLIBD::layer_stat_cor(stats=registration_stats,modeling_results=hyp3_model_res,model_type="enrichment",top_n=curj)
        pdf(paste0("plots/spe_053123/H06-Spatial registration/","Hyp Harmon nnsvg10 BS15 collapsed VMH to Macaque VMH neurons top",j[i],".pdf"),height=5,width=6)
    layer_stat_cor_plot(cor_hypcol,max=max(cor_hypcol))
    title(paste0("BS15 Joint VMH to Macaque VMH neurons top",j[i],"genes"))
    dev.off()
    rm(cor_hypcol)
    
    cor_hypfcol <- spatialLIBD::layer_stat_cor(stats=registration_stats,modeling_results=hyp4_model_res,model_type="enrichment",top_n=curj)
        pdf(paste0("plots/spe_053123/H06-Spatial registration/","Hyp Harmon nnsvg10 BS15 Females collapsed VMH to Macaque VMH neurons top",j[i],".pdf"),height=5,width=6)
    layer_stat_cor_plot(cor_hypfcol,max=max(cor_hypfcol))
    title(paste0("BS15 Females Joint VMH \n to Macaque VMH neurons top",j[i],"genes"))
    dev.off()
    rm(cor_hypfcol)
}
```
