---
title: "03-Compare Yao23 Subclass Sex DE to Visium"
author: "Bernard Mulvey"
date: "2024-01-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(ggplot2)
library(gridExtra)
require(colorout)
library(orthogene)
library(biomaRt)
library(gtools)
source("~/.R/calc_tau_for_cellOrTissueSpecificity_metric.R")
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")

## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

theme_set(theme_bw()+theme(axis.text.x = element_text(size = 14), axis.title.x = element_text(size = 16), axis.text.y = element_text(size = 14), axis.title.y = element_text(size =16), plot.title = element_text(size = 20,hjust=0.5), strip.text = element_text(size=18), legend.text = element_text(size=10), legend.title = element_text(size=11,hjust=0.5)))
```

load dreamlet results for Visium and Yao 23 mouse scseq
```{r}
ms <- fread("data/12-Mouse comparisons/02-Yao23-HYP-subclass-sexDE_dreamlet_donormouse-ranfx.txt")
msn <- fread("data/12-Mouse comparisons/02-Yao23-HYP-ncells-per-subclass.txt")

vde <- fread("data/09-Sex DE/01b-dreamlet-ranfx_nnsvg10-HmnylmbNA-BS-15-singleClusts.txt")
```

## get 1-1 mapping mouse orthologs and filter the two datasets accordingly
```{r}
msgenes <- as.data.table(unique(ms$ID))
msgenes <- convert_orthologs(gene_df = msgenes,gene_input = "V1",gene_output = "columns",input_species = "mouse",output_species = "human",non121_strategy = "drop_both_species")

# don't filter yet -- we need to retrieve chromosome info and drop sex chr and MT genes first
```

## filter the mouse sex DE to cell types with at least 50 cells in the HYP datasets (since these cluster assignments are based on brain-wide clustering, many clusters are represented very little)
```{r}
ms <- ms[assay %in% msn[N>49,V1]]
```

## drop sex chromosomal and mitochondrial genes from mouse and human results, then calculate pairwise logFCs. we also lose a few other genes here, so update the filtering of vde accordingly
```{r}
## get mouse gene chromosomes
biomart <- useMart("ensembl",dataset="mmusculus_gene_ensembl")
mschrs <- biomaRt::getBM(attributes=c("external_gene_name","chromosome_name"),mart=biomart,filter="external_gene_name",values=unique(msgenes$input_gene))
rm(biomart)

## filter mouse genes to autosomal, non-mitochondrial
mschrs <- as.data.table(mschrs)
mschrs <- mschrs[!(chromosome_name %in% c("X","Y","MT"))]

## filter the ortholog table and merge it into the two datasets for cluster-pair-wise logfc correlations
msgenes <- as.data.table(msgenes)
msgenes <- msgenes[input_gene %in% mschrs$external_gene_name]

## now filter the sex DE result tables
ms <- ms[ID %in% msgenes$input_gene]
vde <- vde[gene_name %in% msgenes$ortholog_gene & !(chromosome_name %in% c("chrX","chrY","chrMT"))]

### merge the gene info so we can merge the DE results downstream for gene logFC correlation calcs
ms <- merge(ms,msgenes,by.x="ID",by.y="input_gene")
vde <- merge(vde,msgenes,by.x="gene_name",by.y="ortholog_gene")
```

## get unique combinations of clusters, join and correlate the two male-vs-female logFCs for each
```{r}
cluspairs <- expand.grid(unique(ms$assay),unique(vde$assay))
cluspairs <- as.data.table(cluspairs)
cluspairs[,Var1:=as.character(Var1)]
cluspairs[,Var2:=as.character(Var2)]

# initialize result table
restab <- as.data.frame(matrix(nrow=nrow(cluspairs),ncol=4))

## calculate pairwise logFCs
i <- 1
for (i in c(1:nrow(cluspairs))){
    msassay <- cluspairs[i,Var1]
    hgassay <- cluspairs[i,Var2]
    
    # filter to current clusters and merge by human gene name
    tmpms <- copy(ms)[assay==msassay][,.(logFC,ortholog_gene)]
    tmpvde <- copy(vde)[assay==hgassay][,.(logFC,gene_name)]
    setnames(tmpms,"logFC","logFCms")
    
    tmpcompare <- merge.data.table(tmpms,tmpvde,by.x="ortholog_gene",by.y="gene_name")
    
    restab[i,1] <- msassay
    restab[i,2] <- hgassay
    restab[i,3] <- cor(tmpcompare$logFCms,tmpcompare$logFC,use="pairwise.complete.obs",method="pearson")
     restab[i,4] <- cor(tmpcompare$logFCms,tmpcompare$logFC,use="pairwise.complete.obs",method="spearman")
     rm(tmpms,tmpvde,tmpcompare,mgassay,hgassay)
}
rm(i)

restab <- as.data.table(restab)
setnames(restab,c("mouse_subclass","human_domain","pcc","spearmancor"))

fwrite(restab,"data/12-Mouse comparisons/03-Yao23-subclass-MvFsexDE_logFCcorrels_to_singleSpDom-MvF.txt",sep='\t',quote=F)
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] gtools_3.9.5        biomaRt_2.58.0      default_1.0.0      
 [4] orthogene_1.8.0     colorout_1.3-0      gridExtra_2.3      
 [7] ggplot2_3.4.4       Biostrings_2.70.1   GenomeInfoDb_1.38.5
[10] XVector_0.42.0      IRanges_2.36.0      S4Vectors_0.40.2   
[13] BiocGenerics_0.48.1 data.table_1.14.10  rlang_1.1.3        

loaded via a namespace (and not attached):
 [1] DBI_1.2.0                 bitops_1.0-7             
 [3] grr_0.9.5                 magrittr_2.0.3           
 [5] compiler_4.3.2            RSQLite_2.3.4            
 [7] png_0.1-8                 vctrs_0.6.5              
 [9] gprofiler2_0.2.2          stringr_1.5.1            
[11] pkgconfig_2.0.3           crayon_1.5.2             
[13] fastmap_1.1.1             dbplyr_2.4.0             
[15] backports_1.4.1           utf8_1.2.4               
[17] rmarkdown_2.25            sessioninfo_1.2.2        
[19] purrr_1.0.2               bit_4.0.5                
[21] xfun_0.41                 zlibbioc_1.48.0          
[23] cachem_1.0.8              aplot_0.2.2              
[25] jsonlite_1.8.8            progress_1.2.3           
[27] blob_1.2.4                prettyunits_1.2.0        
[29] broom_1.0.5               parallel_4.3.2           
[31] R6_2.5.1                  stringi_1.8.3            
[33] car_3.1-2                 Rcpp_1.0.12              
[35] knitr_1.45                Matrix_1.6-4             
[37] tidyselect_1.2.0          rstudioapi_0.15.0        
[39] abind_1.4-5               yaml_2.3.8               
[41] curl_5.2.0                lattice_0.22-5           
[43] tibble_3.2.1              Biobase_2.62.0           
[45] treeio_1.26.0             withr_2.5.2              
[47] KEGGREST_1.42.0           evaluate_0.23            
[49] gridGraphics_0.5-1        BiocFileCache_2.10.1     
[51] xml2_1.3.6                filelock_1.0.3           
[53] pillar_1.9.0              ggtree_3.10.0            
[55] ggpubr_0.6.0              carData_3.0-5            
[57] ggfun_0.1.3               plotly_4.10.3            
[59] generics_0.1.3            rprojroot_2.0.4          
[61] RCurl_1.98-1.14           hms_1.1.3                
[63] munsell_0.5.0             scales_1.3.0             
[65] tidytree_0.4.6            glue_1.7.0               
[67] lazyeval_0.2.2            tools_4.3.2              
[69] ggsignif_0.6.4            babelgene_22.9           
[71] fs_1.6.3                  XML_3.99-0.16            
[73] grid_4.3.2                tidyr_1.3.0              
[75] ape_5.7-1                 AnnotationDbi_1.64.1     
[77] colorspace_2.1-0          nlme_3.1-164             
[79] GenomeInfoDbData_1.2.11   patchwork_1.2.0          
[81] homologene_1.4.68.19.3.27 cli_3.6.2                
[83] rappdirs_0.3.3            fansi_1.0.6              
[85] viridisLite_0.4.2         dplyr_1.1.4              
[87] gtable_0.3.4              rstatix_0.7.2            
[89] yulab.utils_0.1.3         digest_0.6.33            
[91] ggplotify_0.1.2           htmlwidgets_1.6.4        
[93] memoise_2.0.1             htmltools_0.5.7          
[95] lifecycle_1.0.4           httr_1.4.7               
[97] here_1.0.1                bit64_4.0.5      

─ Session info ────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-01-10
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────
 package          * version        date (UTC) lib source
 abind              1.4-5          2016-07-21 [1] CRAN (R 4.3.0)
 AnnotationDbi      1.64.1         2023-11-02 [1] Bioconductor
 ape                5.7-1          2023-03-13 [1] CRAN (R 4.3.0)
 aplot              0.2.2          2023-10-06 [1] CRAN (R 4.3.1)
 babelgene          22.9           2022-09-29 [1] CRAN (R 4.3.0)
 backports          1.4.1          2021-12-13 [1] CRAN (R 4.3.0)
 Biobase            2.62.0         2023-10-26 [1] Bioconductor
 BiocFileCache      2.10.1         2023-10-26 [1] Bioconductor
 BiocGenerics     * 0.48.1         2023-11-02 [1] Bioconductor
 biomaRt          * 2.58.0         2023-10-24 [1] Bioconductor
 Biostrings       * 2.70.1         2023-10-26 [1] Bioconductor
 bit                4.0.5          2022-11-15 [1] CRAN (R 4.3.0)
 bit64              4.0.5          2020-08-30 [1] CRAN (R 4.3.0)
 bitops             1.0-7          2021-04-24 [1] CRAN (R 4.3.0)
 blob               1.2.4          2023-03-17 [1] CRAN (R 4.3.0)
 broom              1.0.5          2023-06-09 [1] CRAN (R 4.3.0)
 cachem             1.0.8          2023-05-01 [1] CRAN (R 4.3.0)
 car                3.1-2          2023-03-30 [1] CRAN (R 4.3.0)
 carData            3.0-5          2022-01-06 [1] CRAN (R 4.3.0)
 cli                3.6.2          2023-12-11 [1] CRAN (R 4.3.1)
 colorout         * 1.3-0          2023-11-01 [1] local (/Users/bmulvey/Desktop/colorout)
 colorspace         2.1-0          2023-01-23 [1] CRAN (R 4.3.0)
 crayon             1.5.2          2022-09-29 [1] CRAN (R 4.3.0)
 curl               5.2.0          2023-12-08 [1] CRAN (R 4.3.1)
 data.table       * 1.14.10        2023-12-08 [1] CRAN (R 4.3.2)
 DBI                1.2.0          2023-12-21 [1] CRAN (R 4.3.1)
 dbplyr             2.4.0          2023-10-26 [1] CRAN (R 4.3.1)
 default          * 1.0.0          2017-08-07 [1] CRAN (R 4.3.0)
 digest             0.6.33         2023-07-07 [1] CRAN (R 4.3.0)
 dplyr              1.1.4          2023-11-17 [1] CRAN (R 4.3.1)
 evaluate           0.23           2023-11-01 [1] CRAN (R 4.3.1)
 fansi              1.0.6          2023-12-08 [1] CRAN (R 4.3.1)
 fastmap            1.1.1          2023-02-24 [1] CRAN (R 4.3.0)
 filelock           1.0.3          2023-12-11 [1] CRAN (R 4.3.1)
 fs                 1.6.3          2023-07-20 [1] CRAN (R 4.3.0)
 generics           0.1.3          2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb     * 1.38.5         2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData   1.2.11         2023-11-10 [1] Bioconductor
 ggfun              0.1.3          2023-09-15 [1] CRAN (R 4.3.0)
 ggplot2          * 3.4.4          2023-10-12 [1] CRAN (R 4.3.1)
 ggplotify          0.1.2          2023-08-09 [1] CRAN (R 4.3.0)
 ggpubr             0.6.0          2023-02-10 [1] CRAN (R 4.3.0)
 ggsignif           0.6.4          2022-10-13 [1] CRAN (R 4.3.0)
 ggtree             3.10.0         2023-10-24 [1] Bioconductor
 glue               1.7.0          2024-01-09 [1] CRAN (R 4.3.1)
 gprofiler2         0.2.2          2023-06-14 [1] CRAN (R 4.3.0)
 gridExtra        * 2.3            2017-09-09 [1] CRAN (R 4.3.0)
 gridGraphics       0.5-1          2020-12-13 [1] CRAN (R 4.3.0)
 grr                0.9.5          2016-08-26 [1] CRAN (R 4.3.0)
 gtable             0.3.4          2023-08-21 [1] CRAN (R 4.3.0)
 gtools           * 3.9.5          2023-11-20 [1] CRAN (R 4.3.1)
 here               1.0.1          2020-12-13 [1] CRAN (R 4.3.0)
 hms                1.1.3          2023-03-21 [1] CRAN (R 4.3.0)
 homologene         1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.3.0)
 htmltools          0.5.7          2023-11-03 [1] CRAN (R 4.3.1)
 htmlwidgets        1.6.4          2023-12-06 [1] CRAN (R 4.3.1)
 httr               1.4.7          2023-08-15 [1] CRAN (R 4.3.0)
 IRanges          * 2.36.0         2023-10-26 [1] Bioconductor
 jsonlite           1.8.8          2023-12-04 [1] CRAN (R 4.3.1)
 KEGGREST           1.42.0         2023-10-26 [1] Bioconductor
 knitr              1.45           2023-10-30 [1] CRAN (R 4.3.1)
 lattice            0.22-5         2023-10-24 [1] CRAN (R 4.3.1)
 lazyeval           0.2.2          2019-03-15 [1] CRAN (R 4.3.0)
 lifecycle          1.0.4          2023-11-07 [1] CRAN (R 4.3.1)
 magrittr           2.0.3          2022-03-30 [1] CRAN (R 4.3.0)
 Matrix             1.6-4          2023-11-30 [1] CRAN (R 4.3.1)
 memoise            2.0.1          2021-11-26 [1] CRAN (R 4.3.0)
 munsell            0.5.0          2018-06-12 [1] CRAN (R 4.3.0)
 nlme               3.1-164        2023-11-27 [1] CRAN (R 4.3.1)
 orthogene        * 1.8.0          2023-10-24 [1] Bioconductor
 patchwork          1.2.0          2024-01-08 [1] CRAN (R 4.3.1)
 pillar             1.9.0          2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig          2.0.3          2019-09-22 [1] CRAN (R 4.3.0)
 plotly             4.10.3         2023-10-21 [1] CRAN (R 4.3.1)
 png                0.1-8          2022-11-29 [1] CRAN (R 4.3.0)
 prettyunits        1.2.0          2023-09-24 [1] CRAN (R 4.3.1)
 progress           1.2.3          2023-12-06 [1] CRAN (R 4.3.1)
 purrr              1.0.2          2023-08-10 [1] CRAN (R 4.3.0)
 R6                 2.5.1          2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs           0.3.3          2021-01-31 [1] CRAN (R 4.3.0)
 Rcpp               1.0.12         2024-01-09 [1] CRAN (R 4.3.1)
 RCurl              1.98-1.14      2024-01-09 [1] CRAN (R 4.3.1)
 rlang            * 1.1.3          2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown          2.25           2023-09-18 [1] CRAN (R 4.3.1)
 rprojroot          2.0.4          2023-11-05 [1] CRAN (R 4.3.1)
 RSQLite            2.3.4          2023-12-08 [1] CRAN (R 4.3.1)
 rstatix            0.7.2          2023-02-01 [1] CRAN (R 4.3.0)
 rstudioapi         0.15.0         2023-07-07 [1] CRAN (R 4.3.0)
 S4Vectors        * 0.40.2         2023-11-25 [1] Bioconductor 3.18 (R 4.3.2)
 scales             1.3.0          2023-11-28 [1] CRAN (R 4.3.1)
 sessioninfo        1.2.2          2021-12-06 [1] CRAN (R 4.3.0)
 stringi            1.8.3          2023-12-11 [1] CRAN (R 4.3.1)
 stringr            1.5.1          2023-11-14 [1] CRAN (R 4.3.1)
 tibble             3.2.1          2023-03-20 [1] CRAN (R 4.3.0)
 tidyr              1.3.0          2023-01-24 [1] CRAN (R 4.3.0)
 tidyselect         1.2.0          2022-10-10 [1] CRAN (R 4.3.0)
 tidytree           0.4.6          2023-12-12 [1] CRAN (R 4.3.1)
 treeio             1.26.0         2023-10-24 [1] Bioconductor
 utf8               1.2.4          2023-10-22 [1] CRAN (R 4.3.1)
 vctrs              0.6.5          2023-12-01 [1] CRAN (R 4.3.1)
 viridisLite        0.4.2          2023-05-02 [1] CRAN (R 4.3.0)
 withr              2.5.2          2023-10-30 [1] CRAN (R 4.3.1)
 xfun               0.41           2023-11-01 [1] CRAN (R 4.3.1)
 XML                3.99-0.16      2023-11-29 [1] CRAN (R 4.3.1)
 xml2               1.3.6          2023-12-04 [1] CRAN (R 4.3.1)
 XVector          * 0.42.0         2023-10-26 [1] Bioconductor
 yaml               2.3.8          2023-12-11 [1] CRAN (R 4.3.1)
 yulab.utils        0.1.3          2024-01-08 [1] CRAN (R 4.3.1)
 zlibbioc           1.48.0         2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

───────────────────────────────────────────────────────────────────────────
