---
title: "02-reRun sex DE for k15 svg10 ARC2 to fix low expr cutoff issue"
author: "Bernie Mulvey"
date: "2025-10-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(ggplot2)
library(ggtext)
library(gridExtra)
library(SpatialExperiment)
# library(SpatialFeatureExperiment)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw()+theme(axis.text.x = element_text(size = 8,color = "#000000"), axis.title.x = element_text(size = 9,color = "#000000"), axis.text.y = element_text(size = 8,color = "#000000"), axis.title.y = element_text(size =9,color = "#000000"), plot.title = element_markdown(size = 10,hjust=0.5,color = "#000000"), strip.text = element_text(size=11), legend.text = element_text(size=7,color = "#000000"), legend.title = element_text(size=8,hjust=0.5,color = "#000000"),axis.ticks = element_line(linewidth=0.5),panel.grid.major = element_line(linewidth=0.5),panel.grid.minor=element_line(linewidth=0.25),plot.title.position="plot"))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
setDTthreads(1,restore_after_fork = FALSE)
library(SpatialExperiment)
library(dreamlet)
library(limma)
library(edgeR)

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure.
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel)
```

```{r}
## load hyp spe
hyp <- readRDS("spatial_HYP/processed-data/03-QC_filters/hypN10_umi210_gene126_chrm50_spotsweeped_lognorm_rotsNmirrors_072224.RDS")

### remove components we don't need here: 10x dimreds, img data, logcount mat
hyp2 <- SpatialExperiment(assays=list(counts=counts(hyp)),rowData=rowData(hyp),colData=colData(hyp),sample_id = colData(hyp)$sample_id,spatialCoords = spatialCoords(hyp))

assayNames(hyp2)
# clean up
rm(hyp)
gc(full=T)

## get rowdata to annotate genes from dreamlet later; add sex chromosome info from ensembl98, the build used in the 10x pipeline
tmprd <- as.data.table(rowData(hyp2))

genchr <- fread("raw-data/USEFOR10X_ensg_lut_hg38_fromEns98_111123.txt")
genchr <- unique(genchr[,.(ensembl_gene_id,chromosome_name)])
genchr[,chromosome_name:=paste0("chr",chromosome_name)]
setnames(genchr,1,"gene_id")
tmprd <- merge.data.table(tmprd,genchr,by="gene_id",all.x=T)
tmprd <- DataFrame(tmprd,row.names = tmprd$gene_id)[rownames(hyp2),]
stopifnot(identical(rownames(tmprd),rownames(hyp2)))
rowData(hyp2) <- tmprd

# switch tmprd back to data.table for later use
tmprd <- as.data.table(tmprd)

# clean up
rm(genchr)
```

```{r}
bscl <- list.files("spatial_HYP/processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out",pattern = ".txt",recursive=F,full.names = T)
bscls <- lapply(bscl,FUN=function(x){
    y <- fread(x)
    setnames(y,c("rn","cl"))
    y[,cl:=paste0("X",cl)]
    return(y)
})
names(bscls) <- gsub(bscl,pattern="spatial_HYP/processed-data/06-BayesSpace/01-bayesspace60kiter_k15-20-31_out/(.*)\\.txt$",replacement="\\1")

bscl <- bscls$BSpace_k15_HARMONYlmbna_nnsvg10

bscl <- DataFrame(bscl,row.names=bscl$rn)[colnames(hyp2),]
colLabels(hyp2) <- bscl$cl

arc2 <- hyp2[,hyp2$label=="X6"]
```

run the same code, but just for this one cluster from this one cluster param
```{r}

sbp <- MulticoreParam(6)
register(sbp)
ag <- dreamlet::aggregateToPseudoBulk(arc2,sample_id="sample_id",cluster_id = "label",BPPARAM = sbp)
agcd <- as.data.frame(colData(ag))
ag <- assays(ag)[[1]]

X <- ag[,colSums(ag)>0]
s <- copy(agcd)[colnames(X),]
stopifnot(length(unique(s$Sex))==2)
dge <- DGEList(counts=as.matrix(X),samples=s)
dge <- normLibSizes(dge,method = "TMM")
      
## find the smaller group for this cluster
smolgrp <- table(dge$samples$Sex)[which.min(table(dge$samples$Sex))]
      
## for the smallest group, determine low expression cutoff
rowmn.m <- as.numeric(rowMeans(cpm(dge[,(dge$samples$Sex==names(smolgrp)[1])],log=TRUE)))
      
## get histogram distribution with 10 bins at the low end of the logcpm space
h <- hist(rowmn.m[which(rowmn.m>min(rowmn.m)&rowmn.m<(4+min(rowmn.m)))],breaks=10,plot=FALSE)
## this does not work correctly for k=15 nnsvg10 ARC.2 (X6) for wonky reasons, so get the second minimum
if(min(h$counts)==883){ # hardcoded exception for this case
        minc <- h$breaks[which(h$counts==min(h$counts[h$counts!=883]))]
}

## that should fix it and the rest of the code is as in script 01
## get the corresponding smallest-library count value for this minimum^
smollib <- min(dge$samples[(dge$samples$Sex==names(smolgrp)[1]),"lib.size"])
countthresh <- (minc^2)*(as.numeric(smollib)/10^6)
      
## and convert that smallest count to the logcpm value in said smallest library to use as a low-expression threshold
cpmthresh <- log2(countthresh/(as.numeric(smollib)*10^-6))
      
## filter for genes with at least thresh counts in at least (smaller sex group size) samples
keep <- rowSums(cpm(dge,log=T)>cpmthresh) >= as.numeric(smolgrp) #13985 genes, which is more like the other clusters
      
## recalc lib factors after dropping low-expressed genes
dge <- dge[keep,,keep.lib.sizes=FALSE]
dge <- calcNormFactors(dge,method = "TMM")

des <- model.matrix(~0+Sex, data = dge$samples)
# INITIALIZE the voomwithqualityweights model, equivalent to pseudobulkDGE(method="voom",qualities=T)
sexdif <- makeContrasts(SexMale-SexFemale,levels=des)

# run voomLmFit for the pseudobulked data, referring donor to duplicateCorrelation; using an adaptive span (number of genes, based on the number of genes in the dge) for smoothing the mean-variance trend
v.swt <- voomLmFit(dge,design = des,block = as.factor(dge$samples$brnum),adaptive.span = T,sample.weights = T)
## get the Elist object to also output for plotting purposes later
v.swt.elist <- as.data.table(v.swt$EList$E,keep.rownames=T)

v.swt.fit <- contrasts.fit(v.swt,sexdif)

v.swt.fit.e <- eBayes(v.swt.fit)

v.swt.e.tt <- topTable(v.swt.fit.e,coef = 1,number=Inf,adjust.method = "BH")
v.swt.e.tt <- as.data.table(v.swt.e.tt,keep.rownames = T)
v.swt.e.tt <- merge.data.table(v.swt.e.tt,tmprd[,.(gene_id,gene_name,chromosome_name)],by.x="rn",by.y="gene_id")
setnames(v.swt.e.tt,"rn","gene_id")
reordnames <- c("gene_name","logFC","P.Value","adj.P.Val","t","B","chromosome_name","AveExpr","gene_id")
v.swt.e.tt <- v.swt.e.tt[,..reordnames])

fwrite(v.swt.e.tt,"spatial_HYP/processed-data/09-Sex DE/02-k15_svg10_ARC.2_sexDE_correctedRun_102725.txt",sep='\t',quote=F,row.names=F,col.names=T)
```

## replace previous ARC2 results in the RDS, tables written out previously
```{r}
rm(list=grep(ls(),pattern="v.swt.e.tt",value=T,invert=T))

v.swt.e.tt[,assay:="X6"] # to match existing table formats

sexderds <- readRDS("spatial_HYP/processed-data/09-Sex DE/01-voomLmFit_svg10-svg20-hvg20_Hmnydflt-mnn30-HmnylmbNA_BS-15-20-31-15VMHARCclpsd.RDS")

sexderds$BSpace_k15_HARMONYlmbna_nnsvg10 <- sexderds$BSpace_k15_HARMONYlmbna_nnsvg10[assay!="X6"]
sexderds$BSpace_k15_HARMONYlmbna_nnsvg10 <- rbind(sexderds$BSpace_k15_HARMONYlmbna_nnsvg10,v.swt.e.tt)
saveRDS(sexderds,"spatial_HYP/processed-data/09-Sex DE/01-voomLmFit_svg10-svg20-hvg20_Hmnydflt-mnn30-HmnylmbNA_BS-15-20-31-15VMHARCclpsd_k15ARC2fixed.RDS")
### 


sexdetab <- fread("spatial_HYP/processed-data/09-Sex DE/01b-voomLmFit_nnsvg10-HmnylmbNA-BS-15-singleClusts.txt")

sexdetab <- sexdetab[assay!="X6"]
v.swt.e.tt[,chromosome_name:=NULL]
sexdetab <- rbind(sexdetab,v.swt.e.tt)

fwrite(sexdetab,"spatial_HYP/processed-data/09-Sex DE/01b-voomLmFit_nnsvg10-HmnylmbNA-BS-15-singleClusts_ARC2fixed.txt",sep='\t',quote=F,row.names=F,col.names=T)
```
seshinf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.3.2

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods  
[9] base     

other attached packages:
 [1] parallelly_1.43.0           colorout_1.3-0.2           
 [3] edgeR_4.4.2                 dreamlet_1.4.1             
 [5] variancePartition_1.36.3    BiocParallel_1.40.0        
 [7] limma_3.62.2                ggplot2_3.5.2              
 [9] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1
[11] SummarizedExperiment_1.36.0 Biobase_2.66.0             
[13] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[15] IRanges_2.40.1              S4Vectors_0.44.0           
[17] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[19] matrixStats_1.5.0           data.table_1.17.4          
[21] rlang_1.1.5                

loaded via a namespace (and not attached):
  [1] splines_4.4.3             bitops_1.0-9              tibble_3.2.1             
  [4] graph_1.84.1              XML_3.99-0.18             lifecycle_1.0.4          
  [7] Rdpack_2.6.3              mixsqp_0.3-54             rprojroot_2.0.4          
 [10] lattice_0.22-6            MASS_7.3-65               backports_1.5.0          
 [13] magrittr_2.0.3            metafor_4.8-0             rmarkdown_2.29           
 [16] yaml_2.3.10               sessioninfo_1.2.3         DBI_1.2.3                
 [19] minqa_1.2.8               RColorBrewer_1.1-3        abind_1.4-8              
 [22] zlibbioc_1.52.0           EnvStats_3.0.0            purrr_1.0.4              
 [25] rmeta_3.0                 msigdbr_10.0.1            RCurl_1.98-1.17          
 [28] GenomeInfoDbData_1.2.13   ggrepel_0.9.6             pbkrtest_0.5.3           
 [31] irlba_2.3.5.1             annotate_1.84.0           DelayedMatrixStats_1.28.1
 [34] codetools_0.2-20          DelayedArray_0.32.0       tidyselect_1.2.1         
 [37] UCSC.utils_1.2.0          farver_2.1.2              lme4_1.1-36              
 [40] mathjaxr_1.8-0            jsonlite_2.0.0            iterators_1.0.14         
 [43] tools_4.4.3               progress_1.2.3            Rcpp_1.0.14              
 [46] zenith_1.8.0              glue_1.8.0                SparseArray_1.6.2        
 [49] xfun_0.52                 here_1.0.1                dplyr_1.1.4              
 [52] withr_3.0.2               numDeriv_2016.8-1.1       fastmap_1.2.0            
 [55] boot_1.3-31               caTools_1.18.3            digest_0.6.37            
 [58] truncnorm_1.0-9           R6_2.6.1                  colorspace_2.1-1         
 [61] scattermore_1.2           gtools_3.9.5              dichromat_2.0-0.1        
 [64] RSQLite_2.3.9             RhpcBLASctl_0.23-42       tidyr_1.3.1              
 [67] generics_0.1.4            corpcor_1.6.10            prettyunits_1.2.0        
 [70] httr_1.4.7                S4Arrays_1.6.0            pkgconfig_2.0.3          
 [73] gtable_0.3.6              blob_1.2.4                XVector_0.46.0           
 [76] remaCor_0.0.18            htmltools_0.5.8.1         zigg_0.0.2               
 [79] GSEABase_1.68.0           scales_1.4.0              png_0.1-8                
 [82] fANCOVA_0.6-1             reformulas_0.4.0          ashr_2.2-63              
 [85] knitr_1.50                rstudioapi_0.17.1         reshape2_1.4.4           
 [88] rjson_0.2.23              nlme_3.1-167              nloptr_2.2.1             
 [91] cachem_1.1.0              stringr_1.5.1             KernSmooth_2.23-26       
 [94] vipor_0.4.7               metadat_1.4-0             AnnotationDbi_1.68.0     
 [97] pillar_1.10.2             grid_4.4.3                vctrs_0.6.5              
[100] gplots_3.2.0              mashr_0.2.79              xtable_1.8-4             
[103] beeswarm_0.4.0            Rgraphviz_2.50.0          evaluate_1.0.3           
[106] KEGGgraph_1.66.0          invgamma_1.1              magick_2.8.6             
[109] mvtnorm_1.3-3             cli_3.6.4                 locfit_1.5-9.12          
[112] compiler_4.4.3            crayon_1.5.3              SQUAREM_2021.1           
[115] plyr_1.8.9                ggbeeswarm_0.7.2          stringi_1.8.7            
[118] assertthat_0.2.1          babelgene_22.9            lmerTest_3.1-3           
[121] Biostrings_2.74.1         aod_1.3.3                 Matrix_1.7-3             
[124] hms_1.1.3                 sparseMatrixStats_1.18.0  bit64_4.6.0-1            
[127] KEGGREST_1.46.0           statmod_1.5.0             rbibutils_2.3            
[130] Rfast_2.1.5.1             broom_1.0.7               memoise_2.0.1            
[133] RcppParallel_5.1.10       bit_4.6.0                 EnrichmentBrowser_2.36.0 
> sessioninfo::session_info()
─ Session info ─────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.3 (2025-02-28)
 os       macOS Sequoia 15.3.2
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2025-10-27
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
 quarto   1.4.555 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/quarto

─ Packages ─────────────────────────────────────────────────────────────────────────
 !  package              * version    date (UTC) lib source
    abind                  1.4-8      2024-09-12 [1] CRAN (R 4.4.1)
    annotate               1.84.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    AnnotationDbi          1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    aod                    1.3.3      2023-12-13 [1] CRAN (R 4.4.0)
    ashr                   2.2-63     2023-08-21 [1] CRAN (R 4.4.0)
    assertthat             0.2.1      2019-03-21 [1] CRAN (R 4.4.0)
    babelgene              22.9       2022-09-29 [1] CRAN (R 4.4.0)
    backports              1.5.0      2024-05-23 [1] CRAN (R 4.4.0)
    beeswarm               0.4.0      2021-06-01 [1] CRAN (R 4.4.0)
    Biobase              * 2.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocGenerics         * 0.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    BiocParallel         * 1.40.0     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    Biostrings             2.74.1     2024-12-16 [1] Bioconductor 3.20 (R 4.4.2)
    bit                    4.6.0      2025-03-06 [1] CRAN (R 4.4.1)
    bit64                  4.6.0-1    2025-01-16 [1] CRAN (R 4.4.1)
    bitops                 1.0-9      2024-10-03 [1] CRAN (R 4.4.1)
    blob                   1.2.4      2023-03-17 [1] CRAN (R 4.4.0)
    boot                   1.3-31     2024-08-28 [1] CRAN (R 4.4.3)
    broom                  1.0.7      2024-09-26 [1] CRAN (R 4.4.1)
    cachem                 1.1.0      2024-05-16 [1] CRAN (R 4.4.0)
    caTools                1.18.3     2024-09-04 [1] CRAN (R 4.4.1)
 VP cli                    3.6.4      2025-04-23 [2] CRAN (R 4.4.1) (on disk 3.6.5)
    codetools              0.2-20     2024-03-31 [1] CRAN (R 4.4.3)
    colorout             * 1.3-0.2    2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
    colorspace             2.1-1      2024-07-26 [1] CRAN (R 4.4.0)
    corpcor                1.6.10     2021-09-16 [1] CRAN (R 4.4.0)
    crayon                 1.5.3      2024-06-20 [1] CRAN (R 4.4.0)
    data.table           * 1.17.4     2025-05-26 [1] CRAN (R 4.4.3)
    DBI                    1.2.3      2024-06-02 [1] CRAN (R 4.4.0)
    DelayedArray           0.32.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    DelayedMatrixStats     1.28.1     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    dichromat              2.0-0.1    2022-05-02 [1] CRAN (R 4.4.0)
    digest                 0.6.37     2024-08-19 [1] CRAN (R 4.4.1)
    dplyr                  1.1.4      2023-11-17 [1] CRAN (R 4.4.0)
    dreamlet             * 1.4.1      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    edgeR                * 4.4.2      2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    EnrichmentBrowser      2.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    EnvStats               3.0.0      2024-08-24 [1] CRAN (R 4.4.1)
    evaluate               1.0.3      2025-01-10 [1] CRAN (R 4.4.1)
    fANCOVA                0.6-1      2020-11-13 [1] CRAN (R 4.4.0)
    farver                 2.1.2      2024-05-13 [1] CRAN (R 4.4.0)
    fastmap                1.2.0      2024-05-15 [1] CRAN (R 4.4.0)
    generics               0.1.4      2025-05-09 [1] CRAN (R 4.4.1)
    GenomeInfoDb         * 1.42.3     2025-01-27 [1] Bioconductor 3.20 (R 4.4.2)
    GenomeInfoDbData       1.2.13     2024-12-12 [1] Bioconductor
    GenomicRanges        * 1.58.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    ggbeeswarm             0.7.2      2023-04-29 [1] CRAN (R 4.4.0)
    ggplot2              * 3.5.2      2025-04-09 [1] CRAN (R 4.4.1)
    ggrepel                0.9.6      2024-09-07 [1] CRAN (R 4.4.1)
    glue                   1.8.0      2024-09-30 [1] CRAN (R 4.4.1)
    gplots                 3.2.0      2024-10-05 [1] CRAN (R 4.4.1)
    graph                  1.84.1     2025-01-02 [1] Bioconductor 3.20 (R 4.4.2)
    GSEABase               1.68.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    gtable                 0.3.6      2024-10-25 [1] CRAN (R 4.4.1)
    gtools                 3.9.5      2023-11-20 [1] CRAN (R 4.4.0)
    here                   1.0.1      2020-12-13 [1] CRAN (R 4.4.0)
    hms                    1.1.3      2023-03-21 [1] CRAN (R 4.4.0)
    htmltools              0.5.8.1    2024-04-04 [1] CRAN (R 4.4.0)
    httr                   1.4.7      2023-08-15 [1] CRAN (R 4.4.0)
    invgamma               1.1        2017-05-07 [1] CRAN (R 4.4.0)
    IRanges              * 2.40.1     2024-12-05 [1] Bioconductor 3.20 (R 4.4.2)
    irlba                  2.3.5.1    2022-10-03 [1] CRAN (R 4.4.0)
    iterators              1.0.14     2022-02-05 [1] CRAN (R 4.4.0)
    jsonlite               2.0.0      2025-03-27 [1] CRAN (R 4.4.1)
    KEGGgraph              1.66.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    KEGGREST               1.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    KernSmooth             2.23-26    2025-01-01 [1] CRAN (R 4.4.3)
    knitr                  1.50       2025-03-16 [1] CRAN (R 4.4.2)
    lattice                0.22-6     2024-03-20 [1] CRAN (R 4.4.3)
    lifecycle              1.0.4      2023-11-07 [1] CRAN (R 4.4.0)
    limma                * 3.62.2     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    lme4                   1.1-36     2025-01-11 [1] CRAN (R 4.4.1)
    lmerTest               3.1-3      2020-10-23 [1] CRAN (R 4.4.0)
    locfit                 1.5-9.12   2025-03-05 [1] CRAN (R 4.4.1)
    magick                 2.8.6      2025-03-23 [1] CRAN (R 4.4.1)
    magrittr               2.0.3      2022-03-30 [1] CRAN (R 4.4.0)
    mashr                  0.2.79     2023-10-18 [1] CRAN (R 4.4.0)
    MASS                   7.3-65     2025-02-28 [1] CRAN (R 4.4.1)
    mathjaxr               1.8-0      2025-04-30 [1] CRAN (R 4.4.1)
    Matrix                 1.7-3      2025-03-11 [1] CRAN (R 4.4.1)
    MatrixGenerics       * 1.18.1     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    matrixStats          * 1.5.0      2025-01-07 [1] CRAN (R 4.4.1)
    memoise                2.0.1      2021-11-26 [1] CRAN (R 4.4.0)
    metadat                1.4-0      2025-02-04 [1] CRAN (R 4.4.1)
    metafor                4.8-0      2025-01-28 [1] CRAN (R 4.4.1)
    minqa                  1.2.8      2024-08-17 [1] CRAN (R 4.4.0)
    mixsqp                 0.3-54     2023-12-20 [1] CRAN (R 4.4.0)
    msigdbr                10.0.1     2025-03-19 [1] CRAN (R 4.4.3)
    mvtnorm                1.3-3      2025-01-10 [1] CRAN (R 4.4.1)
    nlme                   3.1-167    2025-01-27 [1] CRAN (R 4.4.3)
    nloptr                 2.2.1      2025-03-17 [1] CRAN (R 4.4.1)
    numDeriv               2016.8-1.1 2019-06-06 [1] CRAN (R 4.4.0)
    parallelly           * 1.43.0     2025-03-24 [1] CRAN (R 4.4.1)
    pbkrtest               0.5.3      2024-06-26 [1] CRAN (R 4.4.0)
    pillar                 1.10.2     2025-04-05 [1] CRAN (R 4.4.1)
    pkgconfig              2.0.3      2019-09-22 [1] CRAN (R 4.4.0)
    plyr                   1.8.9      2023-10-02 [1] CRAN (R 4.4.0)
    png                    0.1-8      2022-11-29 [1] CRAN (R 4.4.0)
    prettyunits            1.2.0      2023-09-24 [1] CRAN (R 4.4.0)
    progress               1.2.3      2023-12-06 [1] CRAN (R 4.4.0)
    purrr                  1.0.4      2025-02-05 [1] CRAN (R 4.4.1)
    R6                     2.6.1      2025-02-15 [1] CRAN (R 4.4.1)
    rbibutils              2.3        2024-10-04 [1] CRAN (R 4.4.1)
    RColorBrewer           1.1-3      2022-04-03 [1] CRAN (R 4.4.0)
    Rcpp                   1.0.14     2025-01-12 [1] CRAN (R 4.4.1)
    RcppParallel           5.1.10     2025-01-24 [1] CRAN (R 4.4.1)
    RCurl                  1.98-1.17  2025-03-22 [1] CRAN (R 4.4.1)
    Rdpack                 2.6.3      2025-03-16 [1] CRAN (R 4.4.2)
    reformulas             0.4.0      2024-11-03 [1] CRAN (R 4.4.1)
    remaCor                0.0.18     2024-02-08 [1] CRAN (R 4.4.0)
    reshape2               1.4.4      2020-04-09 [1] CRAN (R 4.4.0)
    Rfast                  2.1.5.1    2025-03-14 [1] CRAN (R 4.4.1)
    Rgraphviz              2.50.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    RhpcBLASctl            0.23-42    2023-02-11 [1] CRAN (R 4.4.0)
    rjson                  0.2.23     2024-09-16 [1] CRAN (R 4.4.1)
 VP rlang                * 1.1.5      2025-04-11 [2] CRAN (R 4.4.1) (on disk 1.1.6)
    rmarkdown              2.29       2024-11-04 [1] CRAN (R 4.4.1)
    rmeta                  3.0        2018-03-20 [1] CRAN (R 4.4.0)
    rprojroot              2.0.4      2023-11-05 [1] CRAN (R 4.4.0)
    RSQLite                2.3.9      2024-12-03 [1] CRAN (R 4.4.1)
    rstudioapi             0.17.1     2024-10-22 [1] CRAN (R 4.4.1)
    S4Arrays               1.6.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    S4Vectors            * 0.44.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    scales                 1.4.0      2025-04-24 [1] CRAN (R 4.4.1)
    scattermore            1.2        2023-06-12 [1] CRAN (R 4.4.0)
    sessioninfo            1.2.3      2025-02-05 [1] CRAN (R 4.4.1)
    SingleCellExperiment * 1.28.1     2024-11-23 [1] Bioconductor 3.20 (R 4.4.2)
    SparseArray            1.6.2      2025-02-20 [1] Bioconductor 3.20 (R 4.4.2)
    sparseMatrixStats      1.18.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SpatialExperiment    * 1.16.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    SQUAREM                2021.1     2021-01-13 [1] CRAN (R 4.4.0)
    statmod                1.5.0      2023-01-06 [1] CRAN (R 4.4.0)
    stringi                1.8.7      2025-03-27 [1] CRAN (R 4.4.1)
    stringr                1.5.1      2023-11-14 [1] CRAN (R 4.4.0)
    SummarizedExperiment * 1.36.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    tibble                 3.2.1      2023-03-20 [1] CRAN (R 4.4.0)
    tidyr                  1.3.1      2024-01-24 [1] CRAN (R 4.4.0)
    tidyselect             1.2.1      2024-03-11 [1] CRAN (R 4.4.0)
    truncnorm              1.0-9      2023-03-20 [1] CRAN (R 4.4.0)
    UCSC.utils             1.2.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    variancePartition    * 1.36.3     2025-01-09 [1] Bioconductor 3.20 (R 4.4.2)
    vctrs                  0.6.5      2023-12-01 [1] CRAN (R 4.4.0)
    vipor                  0.4.7      2023-12-18 [1] CRAN (R 4.4.0)
    withr                  3.0.2      2024-10-28 [1] CRAN (R 4.4.1)
    xfun                   0.52       2025-04-02 [1] CRAN (R 4.4.1)
    XML                    3.99-0.18  2025-01-01 [1] CRAN (R 4.4.1)
    xtable                 1.8-4      2019-04-21 [1] CRAN (R 4.4.0)
    XVector                0.46.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    yaml                   2.3.10     2024-07-26 [1] CRAN (R 4.4.0)
    zenith                 1.8.0      2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)
    zigg                   0.0.2      2025-02-07 [1] CRAN (R 4.4.1)
    zlibbioc               1.52.0     2024-11-08 [1] Bioconductor 3.20 (R 4.4.1)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 * ── Packages attached to the search path.
 V ── Loaded and on-disk version mismatch.
 P ── Loaded and on-disk path mismatch.

────────────────────────────────────────────────────────────────────────────────────