---
title: "Fig5_Attributing_SexDE_to_celltypes"
output: html_document
date: "2024-10-31"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table) # Preferred data manipulation package
library(ggplot2) # Dependency for several plotting functions
library(ggtext) # more character types in ggplots
library(ggrepel) # text labels in volcano plots
library(ggrastr) # avoid making raster images one can't even open
library(SpatialExperiment) # Visium data framework
library(SpatialFeatureExperiment) # Xenium data framework
library(spatialLIBD) # one option for plotting cluster assignments, but breaks when the in-tissue portion of a visium area is very un-square.
library(escheR) # alternative spotplotting function, at least for visium
library(viridis) # palettes
library(Polychrome) # better palettes
library(ggstance) # for y axis dodge
library(sf) # define the polygonal boundaries of xenium domains
library(ggrepel) # for most of the ARC cell type palette https://github.com/moldach/vapoRwave 

## rstudio GUI tweaks
require(colorout)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")
##

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc. seemed to need this a couple times, but otherwise havent so its here as a preventative measure. part of this is adding the line 
# OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
# to Renviron.site. see e.g. top response on https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall 
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")


## ggplot defaults
theme_set(theme_bw() + theme(axis.text.x = element_text(size = 8), axis.title.x = element_text(size = 9), axis.text.y = element_text(size = 8), axis.title.y = element_text(size = 9), plot.title = element_blank(), strip.text = element_text(size = 10), legend.text = element_text(size = 8), legend.title = element_text(size = 8, hjust = 0.5)))

## last of all, unload the base package datasets, whose data keep getting in the way of autocompletions (e.g., Theoph is priortized over TRUE)
unloadNamespace("datasets")
```

### cluster label setup ###
```{r}
## drop discarded cell clusters
bkcl <- fread("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/01-BanksyClusts_M0lam0_leiden_multisamp.txt")
setnames(bkcl,2,"cl")

bkanno <- fread("xenium_HYP/processed-data/05_Banksy_M0lam0_res2_multisamp/02-M0l0kg6_topClusMarkers_celltypes_annotated.txt")

## drop sample-specific clusters, append cluster annotations 
bkanno <- unique(bkanno[,.(clus,bjm_annot)])
bkanno <- bkanno[bjm_annot!="DISCARD"&bjm_annot!="VMH_4_DISCARD"]

## convert ARC_1..etc to more descriptive cluster IDs defined based on ARC/VMH specific cluster marker analyses
arcvmhanno <- fread("xenium_HYP/processed-data/07_Domains-subdomains_by_knnSmoothing_Banksytypes/04e_ARC_and_VMH_cluster_identities_based_on_cellgroup_1vall.txt")

## simplify these by removing VMH/ARC and underscores; asterisk-ize so the gene names  can be italiczed in ggplot elements; remove vmh and arc from the names but add a temporary column here so we know they're VMH and ARC for mapping onto the de result sets
arcvmhanno[,subclus_domain:=gsub(subclus_annot,pattern="^(VMH)_.*$",replacement="\\1")]
arcvmhanno[,subclus_annot:=gsub(subclus_annot,pattern="VMH_",replacement="")]

arcvmhanno[subclus_domain!="VMH",subclus_domain:=gsub(subclus_annot,pattern="^(ARC)_.*$",replacement="\\1")]
arcvmhanno[,subclus_annot:=gsub(subclus_annot,pattern="ARC_",replacement="")]

arcvmhanno[,subclus_annot:=gsub(subclus_annot,pattern="_",replacement=" ")]
arcvmhanno[subclus_annot=="CRHR2",subclus_annot:="*CRHR2*"]
arcvmhanno[subclus_annot=="LAMP5",subclus_annot:="*LAMP5*"]
arcvmhanno[subclus_annot=="xVMH-Glia-Mixed",subclus_annot:="xVMH-Glia Mixed"]
arcvmhanno[subclus_annot=="Lateral-Border",subclus_annot:="xVMH Lateral Border"]

arcvmhanno[subclus_annot=="AGRP",subclus_annot:="*AGRP*"]
arcvmhanno[subclus_annot=="TAC3-ESR1",subclus_annot:="*TAC3*-*ESR1*"]
arcvmhanno[subclus_annot=="POMC-Oligo-Mixed",subclus_annot:="*POMC*-Oligo Mixed"]
arcvmhanno[subclus_annot=="SLC17A7-Glia-Mixed",subclus_annot:="*SLC17A7*-Glia Mixed"]
arcvmhanno[subclus_annot=="GAL-TRH-GHRH",subclus_annot:="*GAL*-*TRH*-*GHRH*"]

## append these to the other annots and replace bjm_annot with the ARC/VMH cell type labels where applicable
bkanno <- merge.data.table(bkanno,arcvmhanno,by="bjm_annot",all.x=T)
bkanno[!is.na(subclus_annot),bjm_annot:=subclus_annot]

## this drops rows that were assigned one of the discarded types (ie no longer in the cell type annotations)
bkcl <- merge.data.table(bkcl,bkanno,by.x="cl",by.y="clus")

```

### sex DE results
```{r}
### exclude sex chromosomal, mito genes (need to append this info)
genelut <- fread("spatial_HYP/raw-data/USEFOR10X_ensg_lut_hg38_fromEns98_111123.txt")
genelut <- unique(genelut[,.(ensembl_gene_id,external_gene_name,chromosome_name)])
setnames(genelut,c("ens","gene_name","chr"))
genelut <- genelut[chr %in% c(1:22,"X","Y","MT")]




### load by cell type and full-domains xenium results from 4 ARC-cell-type-smoothing domain
xenarcde <- fread("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/02a-4ARCtypesmoothed_withinARC_celltypeSexDE.txt")
xenvmhde <- fread("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/02b-4ARCtypesmoothed_withinVMH_celltypeSexDE.txt")

xendomde <- fread("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/04-4typeARC-and-dualassignVMH_domainwise_sexDE.txt")

## make sure all genes in the xenium de analysis have only one ENSG for their symbol
stopifnot(length(unique(
   genelut[gene_name %in% unique(c(xenvmhde$gene_name,xenarcde$gene_name,xendomde$gene_name)),ens]))==
      length(unique(c(xenvmhde$gene_name,xenarcde$gene_name))))
##

xenarcde <- merge.data.table(xenarcde,genelut[,.(gene_name,chr)],by="gene_name")
xenarcde <- xenarcde[!(chr %in% c("MT","X","Y"))]
xenvmhde <- merge.data.table(xenvmhde,genelut[,.(gene_name,chr)],by="gene_name")
xenvmhde <- xenvmhde[!(chr %in% c("MT","X","Y"))]

## clean up
rm(genelut,vistmp)

# also load pseudobulk ELists used in the DE analysis for plotting DE from pseudobulk results
arcelist <- readRDS("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/02c-pseudobulkARC_voomLmFitELists.RDS")
vmhelist <- readRDS("xenium_HYP/processed-data/08_VMH-ARC cell type sex DE within domains/02d-pseudobulkVMH_voomLmFitELists.RDS")

## subset the pseudobulk tables to the ARC clusters for the arc domain and VMH for vmh
arcelist <- arcelist[bkanno[subclus_domain=="ARC",clus]]
vmhelist <- vmhelist[bkanno[subclus_domain=="VMH",clus]]

## change the pseudobulk entry names to be the more granularly annotated cell types
names(vmhelist) <-  bkanno[match(names(vmhelist),bkanno$clus),bjm_annot]
names(arcelist) <-  bkanno[match(names(arcelist),bkanno$clus),bjm_annot]

## xenium demographics
demos <- fread("xenium_HYP/raw-data/demos_and_xeniumrun_metadata.tsv")
```

palettes
```{r}
pals <- readRDS("manuscript_plot_code/domain_and_xencluster_palettes_CHECKPALNAMESBEFOREUS.RDS")

pal <- pals[["Domains"]]

vmhpal <- pals[["XenVMH"]]
arcpal <- pals[["XenARC"]]
stopifnot(all(names(vmhpal) %in% unique(bkcl$bjm_annot)))
stopifnot(all(names(arcpal) %in% unique(bkcl$bjm_annot)))
```

## panel A: xenium volcano, VMH types in VMH domain
```{r}
xenvmhdeplt <- xenvmhde[cluster_annotated %in% grep(cluster_annotated,pattern="VMH",value=T)]

xenvmhdeplt <- merge.data.table(xenvmhdeplt,arcvmhanno,by.x="cluster_annotated",by.y="bjm_annot")

xenvmhdeplt[gene_name %in% c("MYT1L","TAC1","MC4R","CYP26A1","TMEM233","EGR1","ESR1","PGR")&P.Value<0.025,label:=gene_name]

## only show two cell types here for simplicity
xenvmhdeplt <- xenvmhdeplt[subclus_annot %in% c("*CRHR2*","*LAMP5*")]

xenvmhdeplts <- lapply(unique(xenvmhdeplt$subclus_annot),FUN=function(v){ 
   tmpdat <- xenvmhdeplt[subclus_annot==v]
   p <- ggplot(tmpdat, aes(y = -log10(P.Value), x = logFC_MvF,col=subclus_annot)) +
    geom_point(size = 0.15) +
    scale_color_manual(values=vmhpal) +
    ggtitle(paste0("xVMH ",v))+
    geom_hline(yintercept = -log10(0.05),
        linetype = "dashed",
        color = "black",
        linewidth=0.25) +
    geom_hline(yintercept = -log10(min(tmpdat[FDR>0.05,P.Value])),
        linetype = "dashed",
        color = "red",
        linewidth=0.25) +
    geom_text_repel(aes(y = -log10(P.Value),
            x = logFC_MvF,
            label = ifelse(is.na(label), "", label)),
        min.segment.length = 0,
        size = 0.9,
        show.legend=FALSE,
        nudge_x=1.25*sign(tmpdat$logFC_MvF),
        nudge_y = tmpdat$P.Value-0.1,
        seed=10403,
        max.iter=100000,
        max.overlaps=10000,
        max.time=3,
        segment.size=0.125)+
    #facet_wrap(~subclus_annot)+
    guides(color="none")+
    xlab("logFC (M:F)")+
    ylab("-log10 (P-value)")+
    # labs(color="VMH Cell Type")+
    scale_x_continuous(limits=
       c(min(xenvmhdeplt$logFC_MvF),max(xenvmhdeplt$logFC_MvF)),expand = c(0.01,0.01))+
    scale_y_continuous(limits=c(min(-log10(xenvmhdeplt$P.Value)),max(-log10(xenvmhdeplt$P.Value))))+
    theme(axis.text=element_text(size=7),
          axis.title=element_text(size=7.5),
          # legend.key.height = ggplot2::unit(0.075,"in"),
          # legend.key.width=ggplot2::unit(0.04,"in"),
          # legend.position.inside = c(0.19,0.9),
          # legend.position = "inside",
          # legend.text=element_markdown(size=5,margin=margin(0,0,0,0,"in")),
          # legend.box=element_blank(),
          # legend.background=element_blank(),
          # legend.title = element_text(size=6,vjust=0.5,margin=margin(0,0,0.01,0,"in"))),
          plot.title=element_markdown(size=8,hjust=0.5))
   
   p <- rasterize(p,layers="Points",dpi=450)
   return(p)
})


# save
pdf("manuscript_plots/Fig5/5A-XeniumVMHcelltypes_sexDE.pdf",height=3,width=2.3)
do.call("grid.arrange",xenvmhdeplts)
dev.off()

# clean up
rm(xenvmhdeplt,p)

```

Panel B: ARC sex DE volcano by cell type (AGRP, TAC3 oonly)
```{r}
xenarcdeplt <- xenarcde[cluster_annotated %in% grep(cluster_annotated,pattern="ARC",value=T)]

xenarcdeplt[gene_name %in% c("MYT1L","CYP26A1","EGR1","ESR1","PGR","RXRG","CDH13")&FDR<0.2,label:=gene_name]

## replace names with informative cell type labels for plot, and thus in palette
xenarcdeplt <- merge.data.table(xenarcdeplt,arcvmhanno,by.x="cluster_annotated",by.y="bjm_annot")

## only show AGRP and TAC3 clusters in main fig
xenarcdeplt <- xenarcdeplt[subclus_annot %in% c("*AGRP*","*TAC3*-*ESR1*")]

xenarcdeplts <- lapply(unique(xenarcdeplt$subclus_annot),FUN=function(v){ 
   tmpdat <- xenarcdeplt[subclus_annot==v]
   p <- ggplot(tmpdat, aes(y = -log10(P.Value), x = logFC_MvF,col=subclus_annot)) +
    geom_point(size = 0.15) +
    scale_color_manual(values=arcpal) +
    ggtitle(paste0("xARC ",v))+
    geom_hline(yintercept = -log10(0.05),
        linetype = "dashed",
        color = "black",
        linewidth=0.25) +
    geom_hline(yintercept = -log10(min(tmpdat[FDR>0.05,P.Value])),
        linetype = "dashed",
        color = "red",
        linewidth=0.25) +
    geom_text_repel(aes(y = -log10(P.Value),
            x = logFC_MvF,
            label = ifelse(is.na(label), "", label)),
        min.segment.length = 0,
        size = 0.9,
        show.legend=FALSE,
        nudge_x=1.25*sign(tmpdat$logFC_MvF),
        nudge_y = tmpdat$P.Value-0.1,
        seed=10403,
        max.iter=100000,
        max.overlaps=10000,
        max.time=3,
        segment.size=0.125)+
    #facet_wrap(~subclus_annot)+
    guides(color="none")+
    xlab("logFC (M:F)")+
    ylab("-log10 (P-value)")+
    # labs(color="VMH Cell Type")+
    scale_x_continuous(limits=
       c(min(xenarcdeplt$logFC_MvF),max(xenarcdeplt$logFC_MvF)),expand = c(0.01,0.01))+
    scale_y_continuous(limits=c(min(-log10(xenarcdeplt$P.Value)),max(-log10(xenarcdeplt$P.Value))))+
    theme(axis.text=element_text(size=7),
          axis.title=element_text(size=7.5),
          # legend.key.height = ggplot2::unit(0.075,"in"),
          # legend.key.width=ggplot2::unit(0.04,"in"),
          # legend.position.inside = c(0.19,0.9),
          # legend.position = "inside",
          # legend.text=element_markdown(size=5,margin=margin(0,0,0,0,"in")),
          # legend.box=element_blank(),
          # legend.background=element_blank(),
          # legend.title = element_text(size=6,vjust=0.5,margin=margin(0,0,0.01,0,"in"))),
          plot.title=element_markdown(size=8,hjust=0.5))
   
   p <- rasterize(p,layers="Points",dpi=450)
   return(p)
})

# save
pdf("manuscript_plots/Fig5/5B-XeniumARCcelltypes_sexDE.pdf",height=3,width=2.3)
do.call("grid.arrange",xenarcdeplts)
dev.off()
```


Panel C: Show a DEG per VMH cluster, shown for all VMH xenium clusters in a violin plot datawide (at the level of pseudobulked samples). VMH_1: MC4R, VMH_2 (and 3): TAC1, VMH_3: TMEM233
```{r}
vmhe <- rbindlist(vmhelist,idcol = "celltype",fill=T)
# append cluster ids
vmhe <- merge.data.table(vmhe,bkanno[,.(subclus_annot,clus)],by.x="celltype",by.y="subclus_annot")

# drop lateral border
vmhe <- vmhe[celltype!="xVMH Lateral Border"]

# removing leading "elist." from colnames
setnames(vmhe,gsub(names(vmhe),pattern="^elist\\.",replacement=""))

# melt for genes of interest
vmhe.m <- melt(vmhe[rn %in% c("MC4R","TAC1","TMEM233")],id.vars=c("rn","celltype","clus"))

# append sex info
vmhe.m <- merge.data.table(vmhe.m,unique(demos[,.(sample_name_SFEonly,Sex)]),by.x="variable",by.y="sample_name_SFEonly")
vmhe.m[,Sex:=gsub(Sex,pattern="^(.).*$",replacement="\\1")]

# append DE FDR
vmhe.m <- merge.data.table(vmhe.m,xenvmhde[,.(gene_name,cluster_id,FDR)],by.x=c("rn","clus"),by.y=c("gene_name","cluster_id"))

# add line break to xVMH-glia mixed
vmhe.m[celltype=="xVMH-Glia Mixed",celltype:="VMH-Glia<br>Mixed"]


## gray out nonsignificant clusters for each gene
vmhe.m[,cluscol:=ifelse(FDR<0.05,yes=celltype,no="ns")]

# add the gray to vmhpal, accomodate linebreak for mixed cluster into vmhpal 
vmhpal2 <- c(vmhpal,"#ababab")
names(vmhpal2)[length(vmhpal2)] <- "ns"
names(vmhpal2)[names(vmhpal2)=="xVMH-Glia Mixed"] <- "VMH-Glia<br>Mixed"

## make the gray being used for xVMH glia-mixed a gray-purple blend so that it stands out from the others
vmhpal2["xVMH-Glia<br>Mixed"] <- "#cf9eb8"

## make a fill color that's sex-specific
vmhe.m[Sex=="M",clusfill:=cluscol]
vmhe.m[Sex=="F",clusfill:="*afem"]
vmhpal2["*afem"] <- "#FFFFFF" # makes the female box value always appear first (which the points already do for whatever reason) because alphebetical order
# make the gene names markdown'ed so they appear in italics
vmhe.m[,rn:=paste0("*",rn,"*")]

pdf("manuscript_plots/Fig5/5C-VMH_DEGs_boxplot.pdf",height=3,width=3.5)
ggplot(vmhe.m,aes(x=celltype,y=value))+
       geom_boxplot(outliers = FALSE,linewidth = 0.375,aes(fill=clusfill,col=cluscol,alpha=0.5))+
   geom_point(aes(color=cluscol,shape=Sex),size=1,position=position_jitterdodge(jitter.width = 0.15))+
    guides(shape=guide_legend(override.aes=list(size=3)))+
    # scale_fill_manual(values=c("#FFFFFF","#676767"))+
    scale_fill_manual(values=vmhpal2)+
    scale_color_manual(values=vmhpal2)+
    scale_shape_manual(values=c(1,17))+
    xlab("xVMH Neuron Cluster")+
    guides(color="none",fill="none",alpha="none")+
    ylab("Pseudobulk log CPM")+
    facet_wrap(~rn,scales = "free_y")+
    theme(strip.text = element_markdown(size = 8,margin=margin(0.1,0,0,0,unit="in")),strip.background = element_blank(),axis.text.x=element_markdown(angle=90,size=7,hjust=1,vjust=0.5),axis.title.x=element_text(size=8),axis.title.y=element_text(size=8),axis.text.y=element_text(size=7),plot.margin = margin(-0.065,-0.05,0,0.02,unit="in"),legend.background = element_blank(),legend.box=element_blank(),legend.margin=margin(0,0.1,0,-0.075,unit="in"))
dev.off()


# clean up
rm(vmhe.m,vmhe)
```

## for panel D, we want to do the same for three ARC genes of interest. For ARC_4 we have the pick of the litter--we'll go with CDH13, which has known associations to weight. For ARC_3, RXRG is pretty high up there signif wise. And ARC_1 has signif CYP26A1 DE.  
```{r}
arce <- rbindlist(arcelist,idcol = "celltype",fill=T)

# append cluster ids
arce <- merge.data.table(arce,bkanno[,.(subclus_annot,clus)],by.x="celltype",by.y="subclus_annot")

# removing leading "elist." from colnames
setnames(arce,gsub(names(arce),pattern="^elist\\.",replacement=""))

# shorteb cell types to match palette
arce[celltype=="*SLC17A7*-Glia Mixed",celltype:="*SLC17A7*-Glia<br>Mixed"]
arce[celltype=="*GAL*-*TRH*-*GHRH*",celltype:="*GAL*-*GHRH*"]
arce[celltype=="*POMC*-Oligo Mixed",celltype:="*POMC*-Oligo<br>Mixed"]


# melt for genes of interest
arce.m <- melt(arce[rn %in% c("CDH13","RXRG","CYP26A1")],id.vars=c("rn","celltype","clus"))

arce.m <- merge.data.table(arce.m,unique(demos[,.(sample_name_SFEonly,Sex)]),by.x="variable",by.y="sample_name_SFEonly")
arce.m[,Sex:=gsub(Sex,pattern="^(.).*$",replacement="\\1")]

# append DE FDR
arce.m <- merge.data.table(arce.m,xenarcde[,.(gene_name,cluster_id,FDR)],by.x=c("rn","clus"),by.y=c("gene_name","cluster_id"))

## gray out nonsignificant clusters for each gene
arce.m[,cluscol:=ifelse(FDR<0.05,yes=celltype,no="ns")]

# add the gray to vmhpal, accomodate linebreak for mixed cluster into vmhpal 
arcpal2 <- c(arcpal,"#ababab")
names(arcpal2)[length(arcpal2)] <- "ns"

# make fill color for female always white
arce.m[,fillcol:=cluscol]
arce.m[Sex=="F",fillcol:="*afem"]

arcpal2 <- c(arcpal2,"#FFFFFF")
names(arcpal2)[length(arcpal2)] <- "*afem"

# make the gene names markdown'ed so they appear in italics
arce.m[,rn:=paste0("*",rn,"*")]

# fix other palette color names to match abbreviated celltype names
names(arcpal2)[names(arcpal2)=="*SLC17A7*-Glia Mixed"] <- "*SLC17A7*-Glia<br>Mixed"
names(arcpal2)[names(arcpal2)=="*GAL*-*TRH*-*GHRH*"] <- "*GAL*-*GHRH*"
names(arcpal2)[names(arcpal2)=="*POMC*-Oligo Mixed"] <- "*POMC*-Oligo<br>Mixed"

## order celltype
arce.m[,celltype:=factor(celltype,levels=c("*TAC3*-*ESR1*","*AGRP*","*GAL*-*GHRH*","*POMC*-Oligo<br>Mixed","*SLC17A7*-Glia<br>Mixed"))]

pdf("manuscript_plots/Fig5/5D-ARC_DEGs_boxplot.pdf",height=3,width=3.5)
ggplot(arce.m,aes(x=celltype,y=value,color=cluscol,fill = fillcol))+
    geom_boxplot(outliers = FALSE,linewidth = 0.375,alpha=0.5)+
    geom_point(size=1,position=position_jitterdodge(jitter.width = 0.15),aes(shape=Sex))+
    scale_fill_manual(values=arcpal2)+
    scale_color_manual(values=arcpal2)+
    guides(color="none",fill="none")+
    guides(shape=guide_legend(override.aes=list(size=3)))+
    scale_shape_manual(values=c(1,17))+
    ylab("Pseudobulk log CPM")+
    xlab("xARC Neuron Type (Abbreviated)")+
    facet_wrap(~rn,scales = "free_y")+
    theme(strip.text = element_markdown(size = 8,margin=margin(0.1,0,0,0,unit="in")),strip.background = element_blank(),axis.text.x=element_markdown(angle=90,size=7,hjust=1,vjust=0.5),axis.title.x=element_text(size=8),axis.title.y=element_text(size=8),axis.text.y=element_text(size=7),plot.margin = margin(-0.065,-0.05,0,0.02,unit="in"),legend.background = element_blank(),legend.box=element_blank(),legend.margin=margin(0,0.1,0,-0.075,unit="in"))
dev.off()


# clean up
rm(arce.m,arce)
```

repro inf
```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.4.1 RC (2024-06-06 r86719)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     methods   base     

other attached packages:
 [1] BiocParallel_1.38.0            parallelly_1.38.0             
 [3] colorout_1.3-0.2               sf_1.0-16                     
 [5] ggstance_0.3.7                 Polychrome_1.5.1              
 [7] viridis_0.6.5                  viridisLite_0.4.2             
 [9] escheR_1.4.0                   spatialLIBD_1.16.2            
[11] SpatialFeatureExperiment_1.6.1 SpatialExperiment_1.14.0      
[13] SingleCellExperiment_1.26.0    SummarizedExperiment_1.34.0   
[15] Biobase_2.64.0                 GenomicRanges_1.56.1          
[17] GenomeInfoDb_1.40.1            IRanges_2.38.1                
[19] S4Vectors_0.42.1               BiocGenerics_0.50.0           
[21] MatrixGenerics_1.16.0          matrixStats_1.3.0             
[23] ggrastr_1.0.2                  ggrepel_0.9.5                 
[25] ggtext_0.1.2                   ggplot2_3.5.1                 
[27] data.table_1.15.4              rlang_1.1.4                   

loaded via a namespace (and not attached):
  [1] bitops_1.0-7              EBImage_4.46.0            httr_1.4.7               
  [4] RColorBrewer_1.1-3        doParallel_1.0.17         tools_4.4.1              
  [7] utf8_1.2.4                R6_2.5.1                  DT_0.33                  
 [10] HDF5Array_1.32.0          lazyeval_0.2.2            rhdf5filters_1.16.0      
 [13] withr_3.0.0               sp_2.1-4                  gridExtra_2.3            
 [16] cli_3.6.3                 labeling_0.4.3            sass_0.4.9               
 [19] proxy_0.4-27              Rsamtools_2.20.0          commonmark_1.9.1         
 [22] R.utils_2.12.3            scater_1.32.0             sessioninfo_1.2.2        
 [25] attempt_0.3.1             maps_3.4.2                limma_3.60.3             
 [28] rstudioapi_0.16.0         RSQLite_2.3.7             generics_0.1.3           
 [31] BiocIO_1.14.0             spdep_1.3-5               dplyr_1.1.4              
 [34] Matrix_1.7-0              ggbeeswarm_0.7.2          fansi_1.0.6              
 [37] abind_1.4-5               R.methodsS3_1.8.2         terra_1.7-78             
 [40] lifecycle_1.0.4           scatterplot3d_0.3-44      yaml_2.3.9               
 [43] edgeR_4.2.0               rhdf5_2.48.0              SparseArray_1.4.8        
 [46] BiocFileCache_2.12.0      paletteer_1.6.0           grid_4.4.1               
 [49] blob_1.2.4                promises_1.3.0            dqrng_0.4.1              
 [52] ExperimentHub_2.12.0      crayon_1.5.3              lattice_0.22-6           
 [55] beachmat_2.20.0           cowplot_1.1.3             KEGGREST_1.44.1          
 [58] magick_2.8.3              zeallot_0.1.0             pillar_1.9.0             
 [61] knitr_1.48                rjson_0.2.21              boot_1.3-30              
 [64] codetools_0.2-20          wk_0.9.2                  glue_1.7.0               
 [67] vctrs_0.6.5               png_0.1-8                 spam_2.10-0              
 [70] gtable_0.3.5              rematch2_2.1.2            cachem_1.1.0             
 [73] xfun_0.45                 S4Arrays_1.4.1            mime_0.12                
 [76] DropletUtils_1.24.0       sfheaders_0.4.4           iterators_1.0.14         
 [79] units_0.8-5               fields_16.2               statmod_1.5.0            
 [82] bit64_4.0.5               filelock_1.0.3            rprojroot_2.0.4          
 [85] bslib_0.7.0               irlba_2.3.5.1             vipor_0.4.7              
 [88] KernSmooth_2.23-24        colorspace_2.1-0          spData_2.3.1             
 [91] DBI_1.2.3                 tidyselect_1.2.1          bit_4.0.5                
 [94] compiler_4.4.1            curl_5.2.1                BiocNeighbors_1.22.0     
 [97] xml2_1.3.6                DelayedArray_0.30.1       plotly_4.10.4            
[100] rtracklayer_1.64.0        scales_1.3.0              classInt_0.4-10          
[103] rappdirs_0.3.3            tiff_0.1-12               stringr_1.5.1            
[106] digest_0.6.36             fftwtools_0.9-11          rmarkdown_2.27           
[109] benchmarkmeData_1.0.4     XVector_0.44.0            htmltools_0.5.8.1        
[112] pkgconfig_2.0.3           jpeg_0.1-10               sparseMatrixStats_1.16.0 
[115] dbplyr_2.5.0              fastmap_1.2.0             htmlwidgets_1.6.4        
[118] UCSC.utils_1.0.0          shiny_1.8.1.1             DelayedMatrixStats_1.26.0
[121] farver_2.1.2              jquerylib_0.1.4           jsonlite_1.8.8           
[124] config_0.3.2              R.oo_1.26.0               BiocSingular_1.20.0      
[127] RCurl_1.98-1.14           magrittr_2.0.3            scuttle_1.14.0           
[130] GenomeInfoDbData_1.2.12   s2_1.1.6                  dotCall64_1.1-1          
[133] Rhdf5lib_1.26.0           munsell_0.5.1             Rcpp_1.0.13              
[136] stringi_1.8.4             zlibbioc_1.50.0           AnnotationHub_3.12.0     
[139] parallel_4.4.1            deldir_2.0-4              Biostrings_2.72.1        
[142] gridtext_0.1.5            locfit_1.5-9.10           markdown_1.13            
[145] ScaledMatrix_1.12.0       BiocVersion_3.19.1        XML_3.99-0.17            
[148] evaluate_0.24.0           golem_0.4.1               BiocManager_1.30.23      
[151] foreach_1.5.2             httpuv_1.6.15             tidyr_1.3.1              
[154] purrr_1.0.2               benchmarkme_1.0.8         rsvd_1.0.5               
[157] xtable_1.8-4              restfulr_0.0.15           e1071_1.7-14             
[160] later_1.3.2               class_7.3-22              tibble_3.2.1             
[163] memoise_2.0.1             beeswarm_0.4.0            AnnotationDbi_1.66.0     
[166] GenomicAlignments_1.40.0  shinyWidgets_0.8.6        here_1.0.1               
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 RC (2024-06-06 r86719)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/Chicago
 date     2024-10-27
 rstudio  2024.07.0-daily+219 Cranberry Hibiscus (desktop)
 pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────────
 ! package                  * version   date (UTC) lib source
   abind                      1.4-5     2016-07-21 [1] CRAN (R 4.4.0)
   AnnotationDbi              1.66.0    2024-05-01 [1] Bioconductor 3.19 (R 4.4.0)
   AnnotationHub              3.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   attempt                    0.3.1     2020-05-03 [1] CRAN (R 4.4.0)
   beachmat                   2.20.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   beeswarm                   0.4.0     2021-06-01 [1] CRAN (R 4.4.0)
   benchmarkme                1.0.8     2022-06-12 [1] CRAN (R 4.4.0)
   benchmarkmeData            1.0.4     2020-04-23 [1] CRAN (R 4.4.0)
   Biobase                  * 2.64.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
   BiocFileCache              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocGenerics             * 0.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocIO                     1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocManager                1.30.23   2024-05-04 [1] CRAN (R 4.4.0)
   BiocNeighbors              1.22.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocParallel             * 1.38.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocSingular               1.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   BiocVersion                3.19.1    2024-04-22 [1] Bioconductor 3.19 (R 4.4.0)
   Biostrings                 2.72.1    2024-06-02 [1] Bioconductor 3.19 (R 4.4.0)
   bit                        4.0.5     2022-11-15 [1] CRAN (R 4.4.0)
   bit64                      4.0.5     2020-08-30 [1] CRAN (R 4.4.0)
   bitops                     1.0-7     2021-04-24 [1] CRAN (R 4.4.0)
   blob                       1.2.4     2023-03-17 [1] CRAN (R 4.4.0)
   boot                       1.3-30    2024-02-26 [1] CRAN (R 4.4.1)
   bslib                      0.7.0     2024-03-29 [1] CRAN (R 4.4.0)
   cachem                     1.1.0     2024-05-16 [1] CRAN (R 4.4.0)
   class                      7.3-22    2023-05-03 [1] CRAN (R 4.4.1)
   classInt                   0.4-10    2023-09-05 [1] CRAN (R 4.4.0)
 P cli                        3.6.3     2024-06-21 [2] CRAN (R 4.4.0)
   codetools                  0.2-20    2024-03-31 [1] CRAN (R 4.4.1)
   colorout                 * 1.3-0.2   2024-05-01 [1] Github (jalvesaq/colorout@c6113a2)
   colorspace                 2.1-0     2023-01-23 [1] CRAN (R 4.4.0)
   commonmark                 1.9.1     2024-01-30 [1] CRAN (R 4.4.0)
   config                     0.3.2     2023-08-30 [1] CRAN (R 4.4.0)
   cowplot                    1.1.3     2024-01-22 [1] CRAN (R 4.4.0)
   crayon                     1.5.3     2024-06-20 [1] CRAN (R 4.4.0)
   curl                       5.2.1     2024-03-01 [1] CRAN (R 4.4.0)
   data.table               * 1.15.4    2024-03-30 [2] CRAN (R 4.4.0)
   DBI                        1.2.3     2024-06-02 [1] CRAN (R 4.4.0)
   dbplyr                     2.5.0     2024-03-19 [1] CRAN (R 4.4.0)
   DelayedArray               0.30.1    2024-05-07 [1] Bioconductor 3.19 (R 4.4.0)
   DelayedMatrixStats         1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   deldir                     2.0-4     2024-02-28 [1] CRAN (R 4.4.0)
   digest                     0.6.36    2024-06-23 [1] CRAN (R 4.4.0)
   doParallel                 1.0.17    2022-02-07 [1] CRAN (R 4.4.1)
   dotCall64                  1.1-1     2023-11-28 [1] CRAN (R 4.4.0)
   dplyr                      1.1.4     2023-11-17 [1] CRAN (R 4.4.0)
   dqrng                      0.4.1     2024-05-28 [1] CRAN (R 4.4.0)
   DropletUtils               1.24.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   DT                         0.33      2024-04-04 [1] CRAN (R 4.4.0)
   e1071                      1.7-14    2023-12-06 [1] CRAN (R 4.4.0)
   EBImage                    4.46.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   edgeR                      4.2.0     2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   escheR                   * 1.4.0     2024-05-16 [1] Bioconductor 3.19 (R 4.4.0)
   evaluate                   0.24.0    2024-06-10 [1] CRAN (R 4.4.0)
   ExperimentHub              2.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   fansi                      1.0.6     2023-12-08 [1] CRAN (R 4.4.0)
   farver                     2.1.2     2024-05-13 [1] CRAN (R 4.4.0)
   fastmap                    1.2.0     2024-05-15 [1] CRAN (R 4.4.0)
   fftwtools                  0.9-11    2021-03-01 [1] CRAN (R 4.4.0)
   fields                     16.2      2024-06-27 [1] CRAN (R 4.4.0)
   filelock                   1.0.3     2023-12-11 [1] CRAN (R 4.4.0)
   foreach                    1.5.2     2022-02-02 [1] CRAN (R 4.4.0)
   generics                   0.1.3     2022-07-05 [1] CRAN (R 4.4.0)
   GenomeInfoDb             * 1.40.1    2024-05-24 [1] Bioconductor 3.19 (R 4.4.0)
   GenomeInfoDbData           1.2.12    2024-05-01 [1] Bioconductor
   GenomicAlignments          1.40.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   GenomicRanges            * 1.56.1    2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
   ggbeeswarm                 0.7.2     2023-04-29 [1] CRAN (R 4.4.0)
   ggplot2                  * 3.5.1     2024-04-23 [1] CRAN (R 4.4.0)
   ggrastr                  * 1.0.2     2023-06-01 [1] CRAN (R 4.4.0)
   ggrepel                  * 0.9.5     2024-01-10 [1] CRAN (R 4.4.0)
   ggstance                 * 0.3.7     2024-04-05 [1] CRAN (R 4.4.0)
   ggtext                   * 0.1.2     2022-09-16 [1] CRAN (R 4.4.0)
   glue                       1.7.0     2024-01-09 [1] CRAN (R 4.4.0)
   golem                      0.4.1     2023-06-05 [1] CRAN (R 4.4.0)
   gridExtra                  2.3       2017-09-09 [1] CRAN (R 4.4.0)
   gridtext                   0.1.5     2022-09-16 [1] CRAN (R 4.4.0)
   gtable                     0.3.5     2024-04-22 [1] CRAN (R 4.4.0)
   HDF5Array                  1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   here                       1.0.1     2020-12-13 [1] CRAN (R 4.4.0)
   htmltools                  0.5.8.1   2024-04-04 [1] CRAN (R 4.4.0)
   htmlwidgets                1.6.4     2023-12-06 [1] CRAN (R 4.4.0)
   httpuv                     1.6.15    2024-03-26 [1] CRAN (R 4.4.0)
   httr                       1.4.7     2023-08-15 [1] CRAN (R 4.4.0)
   IRanges                  * 2.38.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   irlba                      2.3.5.1   2022-10-03 [1] CRAN (R 4.4.0)
   iterators                  1.0.14    2022-02-05 [1] CRAN (R 4.4.0)
   jpeg                       0.1-10    2022-11-29 [1] CRAN (R 4.4.0)
   jquerylib                  0.1.4     2021-04-26 [1] CRAN (R 4.4.0)
   jsonlite                   1.8.8     2023-12-04 [1] CRAN (R 4.4.0)
   KEGGREST                   1.44.1    2024-06-19 [1] Bioconductor 3.19 (R 4.4.0)
   KernSmooth                 2.23-24   2024-05-17 [1] CRAN (R 4.4.1)
   knitr                      1.48      2024-07-07 [1] CRAN (R 4.4.1)
   labeling                   0.4.3     2023-08-29 [1] CRAN (R 4.4.0)
   later                      1.3.2     2023-12-06 [1] CRAN (R 4.4.0)
   lattice                    0.22-6    2024-03-20 [1] CRAN (R 4.4.1)
   lazyeval                   0.2.2     2019-03-15 [1] CRAN (R 4.4.0)
   lifecycle                  1.0.4     2023-11-07 [1] CRAN (R 4.4.0)
   limma                      3.60.3    2024-06-16 [1] Bioconductor 3.19 (R 4.4.0)
   locfit                     1.5-9.10  2024-06-24 [1] CRAN (R 4.4.0)
   magick                     2.8.3     2024-02-18 [1] CRAN (R 4.4.0)
   magrittr                   2.0.3     2022-03-30 [1] CRAN (R 4.4.0)
   maps                       3.4.2     2023-12-15 [1] CRAN (R 4.4.0)
   markdown                   1.13      2024-06-04 [1] CRAN (R 4.4.0)
   Matrix                     1.7-0     2024-04-26 [1] CRAN (R 4.4.1)
   MatrixGenerics           * 1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   matrixStats              * 1.3.0     2024-04-11 [1] CRAN (R 4.4.0)
   memoise                    2.0.1     2021-11-26 [1] CRAN (R 4.4.0)
   mime                       0.12      2021-09-28 [1] CRAN (R 4.4.0)
   munsell                    0.5.1     2024-04-01 [1] CRAN (R 4.4.0)
   paletteer                  1.6.0     2024-01-21 [1] CRAN (R 4.4.0)
   parallelly               * 1.38.0    2024-07-27 [1] CRAN (R 4.4.0)
   pillar                     1.9.0     2023-03-22 [1] CRAN (R 4.4.0)
   pkgconfig                  2.0.3     2019-09-22 [1] CRAN (R 4.4.0)
   plotly                     4.10.4    2024-01-13 [1] CRAN (R 4.4.0)
   png                        0.1-8     2022-11-29 [1] CRAN (R 4.4.0)
   Polychrome               * 1.5.1     2022-05-03 [1] CRAN (R 4.4.0)
   promises                   1.3.0     2024-04-05 [1] CRAN (R 4.4.0)
   proxy                      0.4-27    2022-06-09 [1] CRAN (R 4.4.0)
   purrr                      1.0.2     2023-08-10 [1] CRAN (R 4.4.0)
   R.methodsS3                1.8.2     2022-06-13 [1] CRAN (R 4.4.0)
   R.oo                       1.26.0    2024-01-24 [1] CRAN (R 4.4.0)
   R.utils                    2.12.3    2023-11-18 [1] CRAN (R 4.4.0)
   R6                         2.5.1     2021-08-19 [1] CRAN (R 4.4.0)
   rappdirs                   0.3.3     2021-01-31 [1] CRAN (R 4.4.0)
   RColorBrewer               1.1-3     2022-04-03 [1] CRAN (R 4.4.0)
   Rcpp                       1.0.13    2024-07-17 [1] CRAN (R 4.4.0)
   RCurl                      1.98-1.14 2024-01-09 [1] CRAN (R 4.4.0)
   rematch2                   2.1.2     2020-05-01 [1] CRAN (R 4.4.0)
   restfulr                   0.0.15    2022-06-16 [1] CRAN (R 4.4.0)
   rhdf5                      2.48.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rhdf5filters               1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   Rhdf5lib                   1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   rjson                      0.2.21    2022-01-09 [1] CRAN (R 4.4.0)
 P rlang                    * 1.1.4     2024-06-04 [2] CRAN (R 4.4.1)
   rmarkdown                  2.27      2024-05-17 [1] CRAN (R 4.4.0)
   rprojroot                  2.0.4     2023-11-05 [1] CRAN (R 4.4.0)
   Rsamtools                  2.20.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   RSQLite                    2.3.7     2024-05-27 [1] CRAN (R 4.4.0)
   rstudioapi                 0.16.0    2024-03-24 [1] CRAN (R 4.4.0)
   rsvd                       1.0.5     2021-04-16 [1] CRAN (R 4.4.0)
   rtracklayer                1.64.0    2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   s2                         1.1.6     2023-12-19 [1] CRAN (R 4.4.0)
   S4Arrays                   1.4.1     2024-05-20 [1] Bioconductor 3.19 (R 4.4.0)
   S4Vectors                * 0.42.1    2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
   sass                       0.4.9     2024-03-15 [1] CRAN (R 4.4.0)
   ScaledMatrix               1.12.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scales                     1.3.0     2023-11-28 [1] CRAN (R 4.4.0)
   scater                     1.32.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   scatterplot3d              0.3-44    2023-05-05 [1] CRAN (R 4.4.0)
   scuttle                    1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sessioninfo                1.2.2     2021-12-06 [1] CRAN (R 4.4.0)
   sf                       * 1.0-16    2024-03-24 [1] CRAN (R 4.4.0)
   sfheaders                  0.4.4     2024-01-17 [1] CRAN (R 4.4.0)
   shiny                      1.8.1.1   2024-04-02 [1] CRAN (R 4.4.0)
   shinyWidgets               0.8.6     2024-04-24 [1] CRAN (R 4.4.0)
   SingleCellExperiment     * 1.26.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   sp                         2.1-4     2024-04-30 [1] CRAN (R 4.4.0)
   spam                       2.10-0    2023-10-23 [1] CRAN (R 4.4.0)
   SparseArray                1.4.8     2024-05-30 [1] Bioconductor 3.19 (R 4.4.0)
   sparseMatrixStats          1.16.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialExperiment        * 1.14.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   SpatialFeatureExperiment * 1.6.1     2024-05-15 [1] Bioconductor 3.19 (R 4.4.0)
   spatialLIBD              * 1.16.2    2024-05-28 [1] Bioconductor 3.19 (R 4.4.0)
   spData                     2.3.1     2024-05-31 [1] CRAN (R 4.4.0)
   spdep                      1.3-5     2024-06-10 [1] CRAN (R 4.4.0)
   statmod                    1.5.0     2023-01-06 [1] CRAN (R 4.4.0)
   stringi                    1.8.4     2024-05-06 [1] CRAN (R 4.4.0)
   stringr                    1.5.1     2023-11-14 [1] CRAN (R 4.4.0)
   SummarizedExperiment     * 1.34.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   terra                      1.7-78    2024-05-22 [1] CRAN (R 4.4.0)
   tibble                     3.2.1     2023-03-20 [1] CRAN (R 4.4.0)
   tidyr                      1.3.1     2024-01-24 [1] CRAN (R 4.4.0)
   tidyselect                 1.2.1     2024-03-11 [1] CRAN (R 4.4.0)
   tiff                       0.1-12    2023-11-28 [1] CRAN (R 4.4.0)
   UCSC.utils                 1.0.0     2024-05-06 [1] Bioconductor 3.19 (R 4.4.0)
   units                      0.8-5     2023-11-28 [1] CRAN (R 4.4.0)
   utf8                       1.2.4     2023-10-22 [1] CRAN (R 4.4.0)
   vctrs                      0.6.5     2023-12-01 [1] CRAN (R 4.4.0)
   vipor                      0.4.7     2023-12-18 [1] CRAN (R 4.4.0)
   viridis                  * 0.6.5     2024-01-29 [1] CRAN (R 4.4.0)
   viridisLite              * 0.4.2     2023-05-02 [1] CRAN (R 4.4.0)
   withr                      3.0.0     2024-01-16 [1] CRAN (R 4.4.0)
   wk                         0.9.2     2024-07-09 [1] CRAN (R 4.4.0)
   xfun                       0.45      2024-06-16 [1] CRAN (R 4.4.0)
   XML                        3.99-0.17 2024-06-25 [1] CRAN (R 4.3.3)
   xml2                       1.3.6     2023-12-04 [1] CRAN (R 4.4.0)
   xtable                     1.8-4     2019-04-21 [1] CRAN (R 4.4.0)
   XVector                    0.44.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)
   yaml                       2.3.9     2024-07-05 [1] CRAN (R 4.4.0)
   zeallot                    0.1.0     2018-01-28 [1] CRAN (R 4.4.0)
   zlibbioc                   1.50.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.0)

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
 [2] /Users/bmulvey/Library/R/arm64/4.4/library

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────────