---
title: "02-Spatial Registration vs Yao Subclasses w 50plus cells in ms HYP dissections"
author: "Bernard Mulvey"
date: "2024-02-07"
output: html_document
---
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.height = 10,fig.width = 7,include = FALSE)
knitr::opts_chunk$set(fig.width=7,fig.height=10)
#### sets tab autocompletion for directories to begin from the directory containing the .Rproj session, instead of the script's directory if different
knitr::opts_knit$set(root.dir = here::here())

library(data.table)
library(Biostrings)
library(SingleCellExperiment)
library(SpatialExperiment)
library(spatialLIBD)
require(colorout)
library(magrittr)
library(BiocParallel)
library(orthogene)
ColorOut()
options("styler.addins_style_transformer" = "biocthis::bioc_style()")


## change write.table to output a TSV without quotes or rownames by default
library(default)
default(write.table) <- list(row.names=F,col.names=T,sep='\t',quote=F)

## enable forked parallel processing with BiocParallel::multicoreParam, future::, etc
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
```

load SPE, load Yao data in SCE format
```{r}
hyp2 <- readRDS("data/03-QC_filters/hypN9_umi275_gene166_chrm50_lognorm_111723.RDS")

bscl <- fread("data/06-BayesSpace/02-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl=="X3",cl:="ARC.1"]
bscl[cl=="X5",cl:="ARC.2"]
bscl[cl=="X11",cl:="VMH.1"]
bscl[cl=="X12",cl:="VMH.2"]
bscl[cl=="X1",cl:="WM.1"]
bscl[cl=="X9",cl:="WM.2"]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=c("VMH.1","VMH.2","ARC.1","ARC.2","WM.1","WM.2",paste0("X",c(2,4,6:8,10,13:15))))

### load yao SCE
yao <- readRDS("raw-data/ABA_Whole_Mouse_Brain_Yao23_HYP/bothdatasets_raw_asSCE.RDS")


### check metadata for spatialLIBD compatability (no leading digits)
unique(yao$subclass)
unique(yao$library_label.x)
### k, we should be good.
```

spatial registration wrapper: hyp2
```{r}
sphyp.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphyp.reg,"data/10-Spatial Registration/02a-spHYP_registrationwrapper.RDS")

```

registration wrapper: yao data, min 50 cells (pre-filter to avoid dropped levels errors)
```{r}
# get cell types with a dataset-wide total > 50
cellfilt <- as.data.table(colData(yao),keep.rownames=T)
cellfilt1 <- cellfilt[,.N,by="subclass"]
cellfilt1 <- cellfilt1[N>=50]

cellfilt2 <- cellfilt[subclass %in% cellfilt1$subclass][,.N,by=c("library_label.x","subclass")]

## determine the max number of samples possible
length(unique(cellfilt2$library_label.x))
## 47 samples here, so get those where at >=20 samples have >= ? cells
cellfilt3 <- cellfilt2[N>=20,.N,by="subclass"][subclass %in% grep(subclass,pattern="ARH|VMH",value=T)]
### 20 cells would drop an ARC class (17 samples). try 15 cells.
cellfilt3 <- cellfilt2[N>=15,.N,by="subclass"][subclass %in% grep(subclass,pattern="ARH|VMH",value=T)]
### that gets us at least 22 samples for all of the VMH, ARC classes, which'll do. 47 clusters total
cellfilt3 <- cellfilt3 <- cellfilt2[N>=15,.N,by="subclass"][N>20]

# colnames in the SCE to keep:
cellfilt <- cellfilt[subclass %in% cellfilt3$subclass,rn]
# subset SCE
yaofilt <- yao[,cellfilt]
# sanity check, 47 clusters expected:
length(unique(yaofilt$subclass))
# nice


## clean up
rm(cellfilt,cellfilt1,cellfilt2,cellfilt3,sphyp.reg,yao)
gc(full=T)
```

Yao filt spatialwrapper
```{r}
yaofilt.spwrap <- registration_wrapper(yaofilt,var_registration = "subclass",var_sample_id = "library_label.x",gene_ensembl = "gene_symbol",gene_name = "gene_symbol",min_ncells = 15)

saveRDS(yaofilt.spwrap,"data/10-Spatial Registration/02b-yaofilt_registrationwrapper.RDS")

# RAM free-up
rm(yaofilt.spwrap)
gc(full=T)
```

### once more, but with VMH subclasses and ARC subclasses collapsed down to a single "VMH" and "ARC"
```{r}
tmpcd <- as.data.table(colData(yaofilt),keep.rownames=T)
tmpcd[subclass %in% grep(subclass,pattern="VMH",value=T),subclass:="VMHall"]
# careful here--there's also a subclass not labeled ARH but that has a gene symbol containing arh in the name!
unique(grep(tmpcd$subclass,pattern="ARH",value=T))
tmpcd[subclass %in% grep(subclass,pattern="x075|x120|x121|x114",value=T),subclass:="ARHall"]

tmpcd <- DataFrame(tmpcd)
rownames(tmpcd) <- tmpcd$rn
tmpcd$rn <- NULL
tmpcd <- tmpcd[colnames(yaofilt),]
colData(yaofilt) <- tmpcd
yaofilt$subclass <- factor(yaofilt$subclass)

# should now be 43 subclasses (47-4 ARC - 2 VMH + 1 joint vmh + 1 joint arc)
levels(yaofilt$subclass) # good

### now re-run spatial registration..
yaofilt.collapsed.spwrap <- registration_wrapper(yaofilt,var_registration = "subclass",var_sample_id = "library_label.x",gene_ensembl = "gene_symbol",gene_name = "gene_symbol",min_ncells = 15)


saveRDS(yaofilt.collapsed.spwrap,"data/10-Spatial Registration/02c-yaofilt_collapsedVMHandARHsubclasses_registrationwrapper.RDS")

```

### for posterity, using the k20, k31, and k15 collapsed spatial domains:
    arc20=c("X14","X18"),
    vmh20=c("X7","X9","X13"),

```{r}
rm(list=ls())
gc(full=T)

#
hyp2 <- readRDS("data/03-QC_filters/hypN9_umi275_gene166_chrm50_lognorm_111723.RDS")

# k20
bscl <- fread("data/06-BayesSpace/02-bayesspace60kiter_k15-20-31_out/BSpace_k20_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl=="X7",cl:="VMH20.1"]
bscl[cl=="X9",cl:="VMH20.2"]
bscl[cl=="X13",cl:="VMH20.3"]
bscl[cl=="X14",cl:="ARC20.1"]
bscl[cl=="X18",cl:="ARC20.2"]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=c("VMH20.1","VMH20.2","VMH20.3","ARC20.1","ARC20.2",paste0("X",c(1:6,8,10:12,15:17,19:20))))


sphyp20.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphyp20.reg,"data/10-Spatial Registration/02d-sphypk20_regwrapper.RDS")

# k31:
# arc31=c("X1","X7","X8","X21"),
# vmh31=c("X3","X4","X6","X19","X20","X29")
bscl <- fread("data/06-BayesSpace/02-bayesspace60kiter_k15-20-31_out/BSpace_k31_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl=="X3",cl:="VMH31.1"]
bscl[cl=="X4",cl:="VMH31.2"]
bscl[cl=="X6",cl:="VMH31.3"]
bscl[cl=="X19",cl:="VMH31.4"]
bscl[cl=="X20",cl:="VMH31.5"]
bscl[cl=="X29",cl:="VMH31.6"]

bscl[cl=="X1",cl:="ARC31.1"]
bscl[cl=="X7",cl:="ARC31.2"]
bscl[cl=="X8",cl:="ARC31.3"]
bscl[cl=="X21",cl:="ARC31.4"]

bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

others <- as.numeric(gsub(unique(bscl$cl),pattern="^X(.*)",replacement="\\1"))
others <- others[!is.na(others)]

colLabels(hyp2) <- factor(bscl$cl,levels=c("VMH31.1","VMH31.2","VMH31.3","VMH31.4","VMH31.5","VMH31.6","ARC31.1","ARC31.2","ARC31.3","ARC31.4",paste0("X",others)))

sphyp31.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphyp31.reg,"data/10-Spatial Registration/02e-sphypk31_regwrapper.RDS")

# collapsed
bscl <- fread("data/06-BayesSpace/02-bayesspace60kiter_k15-20-31_out/BSpace_k15_HARMONYlmbna_nnsvg10.txt")

setnames(bscl,c("rn","cl"))
bscl[,cl:=paste0("X",cl)]
bscl[cl %in% c("X1","X9"),cl:="WM"]
bscl[cl %in% c("X3","X5"),cl:="ARC"]
bscl[cl %in% c("X11","X12"),cl:="VMH"]
bscl <- DataFrame(bscl)
rownames(bscl) <- bscl$rn
bscl <- bscl[colnames(hyp2),]

colLabels(hyp2) <- factor(bscl$cl,levels=c("WM","ARC","VMH",paste0("X",c(2,4,6:8,10,13:15))))

sphypcollapse.reg <- spatialLIBD::registration_wrapper(hyp2,var_registration = "label",var_sample_id = "sample_id",gene_ensembl = "gene_id",gene_name = "gene_name",min_ncells = 10)

saveRDS(sphypcollapse.reg,"data/10-Spatial Registration/02f-sphypcollapse_regwrapper.RDS")
```

Wipe env and load the outputs from spatialLIBD::registration_wrapper
```{r}
rm(list=ls())
gc(full=T)


abasub <- readRDS("data/10-Spatial Registration/02b-yaofilt_registrationwrapper.RDS")

abasub.singlevmharc <- readRDS("data/10-Spatial Registration/02c-yaofilt_collapsedVMHandARHsubclasses_registrationwrapper.RDS")

sphyp <- readRDS("data/10-Spatial Registration/02a-spHYP_registrationwrapper.RDS")

sphyp20 <- readRDS("data/10-Spatial Registration/02d-sphypk20_regwrapper.RDS")

sphyp31 <- readRDS("data/10-Spatial Registration/02e-sphypk31_regwrapper.RDS")

sphypclps <- readRDS("data/10-Spatial Registration/02f-sphypcollapse_regwrapper.RDS")
# 

orthos <- as.data.table(unique(as.data.table(sphyp$enrichment$ensembl)))
orthos <- convert_orthologs(gene_df = orthos,gene_input = "V1",gene_output = "columns",input_species = "human",output_species = "mouse",non121_strategy = "drop_both_species")

tmplist2 <- list(sphyp,sphyp20,sphyp31,sphypclps)
tmplist2 <- lapply(X=tmplist2,FUN=function(x){
    
    Y <- as.data.table(x$enrichment)
    keepcols <- c("ensembl",grep(names(Y),pattern="^t_stat_",value=T))
    Y <- Y[,..keepcols]
    
    setnames(Y,c(2:ncol(Y)),gsub(names(Y)[c(2:ncol(Y))],pattern="^t_stat_",replacement=""))


    Y <- as.data.frame(Y)
    rownames(Y) <- Y$ensembl
    Y$ensembl <- NULL
    return(as.data.frame(Y))
})

tmplist <- list(abasub,abasub.singlevmharc)
tmplist <- lapply(X=tmplist,function(x){

    Y <- as.data.table(x$enrichment,keep.rownames=T)
    keepcols <- c("rn",grep(names(Y),pattern="^t_stat_",value=T))
    Y <- Y[,..keepcols]
    
    setnames(Y,c(2:ncol(Y)),gsub(names(Y)[c(2:ncol(Y))],pattern="^t_stat_",replacement=""))
    
    # get human ensembl orthologs
    Y <- merge.data.table(Y,orthos,by.x="rn",by.y="ortholog_gene")

    Y <- as.data.frame(Y)
    rownames(Y) <- Y$input_gene
    Y$rn <- NULL
    Y$input_gene <- NULL
    return(as.data.frame(Y))
    
    
})


abas <- list(tmplist[[1]],tmplist[[2]],tmplist[[1]],tmplist[[2]],tmplist[[1]],tmplist[[2]],tmplist[[1]],tmplist[[2]])
names(abas) <- rep(c("subclass","subclass_collapsed"),4)

hyps <- list(tmplist2[[1]],tmplist2[[1]],tmplist2[[2]],tmplist2[[2]],tmplist2[[3]],tmplist2[[3]],tmplist2[[4]],tmplist2[[4]])
names(hyps) <- c(rep("sphypk15",2),rep("sphypk20",2),rep("sphypk31",2),rep("sphypk15_collapsed",2))
```

spatialLIBD layer_stat_cor won't cooperate so just do this manually
```{r}
outfilesuff <- as.list(paste0(names(abas),"_to_",names(hyps)))

sbp <- MulticoreParam(8)

spatregs <- bpmapply(m=hyps,s=abas,z=outfilesuff,SIMPLIFY = F,BPPARAM = sbp,FUN = function(m,s,z){
    j<-c(25,50,75,100,250,500,750,1000)
    i<-1
    outlist <- list()
    for (i in c(1:9)){
        if(i<=8){
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- j[i]
        }
        else{
            # no value for top_n --> all genes used
            cortab<-merge.data.table(as.data.table(m,keep.rownames=T),as.data.table(s,keep.rownames=T),by="rn")
            ngene <- nrow(cortab)
        }
        # initialize cormat
        nclaba <- ncol(s)
        nclhyp <- ncol(m)
        cormat <- as.data.frame(matrix(nrow=nclhyp,ncol=nclaba))
        k <- 1
        for (k in c(1:nclhyp)){
            rownames(cormat)[k] <- colnames(m)[k]
            l <- 1
            for (l in c(1:nclaba)){
                colnames(cormat)[l] <- colnames(s)[l]
                
                curhyp <- colnames(m)[k]
                curaba <- colnames(s)[l]
                setorderv(cortab,curhyp,-1)
                
                cormat[k,l] <- cor(cortab[c(1:ngene),..curhyp],cortab[c(1:ngene),..curaba],use="pairwise.complete.obs",method = "pearson")
            }
        }
        plttab <- as.data.table(cormat,keep.rownames=T)
        setnames(plttab,"rn","sphyp_domain")
        outlist[[i]] <- plttab
        names(outlist)[i] <- paste0("ABA23_HYP_",z,"_top_",ngene,"_sphypdomainmks")
        rm(curj,plttab,cortab)
    }
    return(outlist)
})

saveRDS(spatregs,"data/10-Spatial Registration/02g-sphyp-k15-20-31-15clps_ABAms-47subclass-47withVMHARCclps.RDS")
```

```{r}
sessionInfo()
sessioninfo::session_info()
```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] parallelly_1.36.0           default_1.0.0              
 [3] orthogene_1.8.0             BiocParallel_1.36.0        
 [5] magrittr_2.0.3              colorout_1.3-0.1           
 [7] spatialLIBD_1.14.1          SpatialExperiment_1.12.0   
 [9] SingleCellExperiment_1.24.0 SummarizedExperiment_1.32.0
[11] Biobase_2.62.0              GenomicRanges_1.54.1       
[13] MatrixGenerics_1.14.0       matrixStats_1.2.0          
[15] Biostrings_2.70.2           GenomeInfoDb_1.38.5        
[17] XVector_0.42.0              IRanges_2.36.0             
[19] S4Vectors_0.40.2            BiocGenerics_0.48.1        
[21] data.table_1.14.10          rlang_1.1.3                

loaded via a namespace (and not attached):
  [1] later_1.3.2                   BiocIO_1.12.0                
  [3] ggplotify_0.1.2               bitops_1.0-7                 
  [5] filelock_1.0.3                fields_15.2                  
  [7] tibble_3.2.1                  XML_3.99-0.16.1              
  [9] lifecycle_1.0.4               rstatix_0.7.2                
 [11] edgeR_4.0.14                  doParallel_1.0.17            
 [13] rprojroot_2.0.4               lattice_0.22-5               
 [15] backports_1.4.1               sass_0.4.8                   
 [17] limma_3.58.1                  plotly_4.10.4                
 [19] rmarkdown_2.25                jquerylib_0.1.4              
 [21] yaml_2.3.8                    httpuv_1.6.13                
 [23] spam_2.10-0                   sessioninfo_1.2.2            
 [25] cowplot_1.1.3                 DBI_1.2.1                    
 [27] RColorBrewer_1.1-3            golem_0.4.1                  
 [29] maps_3.4.2                    abind_1.4-5                  
 [31] zlibbioc_1.48.0               purrr_1.0.2                  
 [33] RCurl_1.98-1.14               yulab.utils_0.1.4            
 [35] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      
 [37] ggrepel_0.9.5                 irlba_2.3.5.1                
 [39] tidytree_0.4.6                DelayedMatrixStats_1.24.0    
 [41] codetools_0.2-19              DelayedArray_0.28.0          
 [43] DT_0.31                       scuttle_1.12.0               
 [45] xml2_1.3.6                    tidyselect_1.2.0             
 [47] aplot_0.2.2                   ScaledMatrix_1.10.0          
 [49] viridis_0.6.5                 shinyWidgets_0.8.1           
 [51] BiocFileCache_2.10.1          GenomicAlignments_1.38.2     
 [53] jsonlite_1.8.8                BiocNeighbors_1.20.2         
 [55] ellipsis_0.3.2                scater_1.30.1                
 [57] iterators_1.0.14              foreach_1.5.2                
 [59] tools_4.3.2                   progress_1.2.3               
 [61] treeio_1.26.0                 Rcpp_1.0.12                  
 [63] glue_1.7.0                    gridExtra_2.3                
 [65] SparseArray_1.2.3             xfun_0.41                    
 [67] here_1.0.1                    dplyr_1.1.4                  
 [69] withr_3.0.0                   BiocManager_1.30.22          
 [71] fastmap_1.1.1                 fansi_1.0.6                  
 [73] digest_0.6.34                 rsvd_1.0.5                   
 [75] gridGraphics_0.5-1            R6_2.5.1                     
 [77] mime_0.12                     colorspace_2.1-0             
 [79] biomaRt_2.58.2                RSQLite_2.3.5                
 [81] config_0.3.2                  utf8_1.2.4                   
 [83] tidyr_1.3.1                   generics_0.1.3               
 [85] rtracklayer_1.62.0            prettyunits_1.2.0            
 [87] httr_1.4.7                    htmlwidgets_1.6.4            
 [89] S4Arrays_1.2.0                pkgconfig_2.0.3              
 [91] gtable_0.3.4                  rsconnect_1.2.0              
 [93] blob_1.2.4                    htmltools_0.5.7              
 [95] carData_3.0-5                 dotCall64_1.1-1              
 [97] scales_1.3.0                  png_0.1-8                    
 [99] attempt_0.3.1                 ggfun_0.1.4                  
[101] knitr_1.45                    rstudioapi_0.15.0            
[103] rjson_0.2.21                  nlme_3.1-164                 
[105] curl_5.2.0                    cachem_1.0.8                 
[107] stringr_1.5.1                 BiocVersion_3.18.1           
[109] parallel_4.3.2                vipor_0.4.7                  
[111] AnnotationDbi_1.64.1          restfulr_0.0.15              
[113] pillar_1.9.0                  grid_4.3.2                   
[115] vctrs_0.6.5                   ggpubr_0.6.0                 
[117] promises_1.2.1                car_3.1-2                    
[119] BiocSingular_1.18.0           dbplyr_2.4.0                 
[121] beachmat_2.18.0               xtable_1.8-4                 
[123] beeswarm_0.4.0                paletteer_1.6.0              
[125] evaluate_0.23                 magick_2.8.2                 
[127] cli_3.6.2                     locfit_1.5-9.8               
[129] compiler_4.3.2                Rsamtools_2.18.0             
[131] crayon_1.5.2                  grr_0.9.5                    
[133] ggsignif_0.6.4                gprofiler2_0.2.2             
[135] rematch2_2.1.2                fs_1.6.3                     
[137] ggbeeswarm_0.7.2              stringi_1.8.3                
[139] viridisLite_0.4.2             babelgene_22.9               
[141] munsell_0.5.0                 lazyeval_0.2.2               
[143] homologene_1.4.68.19.3.27     Matrix_1.6-5                 
[145] ExperimentHub_2.10.0          benchmarkme_1.0.8            
[147] patchwork_1.2.0               hms_1.1.3                    
[149] sparseMatrixStats_1.14.0      bit64_4.0.5                  
[151] ggplot2_3.4.4                 KEGGREST_1.42.0              
[153] statmod_1.5.0                 shiny_1.8.0                  
[155] interactiveDisplayBase_1.40.0 AnnotationHub_3.10.0         
[157] broom_1.0.5                   memoise_2.0.1                
[159] bslib_0.6.1                   ggtree_3.10.0                
[161] benchmarkmeData_1.0.4         bit_4.0.5                    
[163] ape_5.7-1                    
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.2 (2023-10-31)
 os       macOS Ventura 13.6
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2024-02-07
 rstudio  2023.09.1+494 Desert Sunflower (desktop)
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────
 package                * version        date (UTC) lib source
 abind                    1.4-5          2016-07-21 [1] CRAN (R 4.3.0)
 AnnotationDbi            1.64.1         2023-11-02 [1] Bioconductor
 AnnotationHub            3.10.0         2023-10-24 [1] Bioconductor
 ape                      5.7-1          2023-03-13 [1] CRAN (R 4.3.2)
 aplot                    0.2.2          2023-10-06 [1] CRAN (R 4.3.1)
 attempt                  0.3.1          2020-05-03 [1] CRAN (R 4.3.0)
 babelgene                22.9           2022-09-29 [1] CRAN (R 4.3.0)
 backports                1.4.1          2021-12-13 [1] CRAN (R 4.3.2)
 beachmat                 2.18.0         2023-10-24 [1] Bioconductor
 beeswarm                 0.4.0          2021-06-01 [1] CRAN (R 4.3.2)
 benchmarkme              1.0.8          2022-06-12 [1] CRAN (R 4.3.0)
 benchmarkmeData          1.0.4          2020-04-23 [1] CRAN (R 4.3.0)
 Biobase                * 2.62.0         2023-10-24 [1] Bioconductor
 BiocFileCache            2.10.1         2023-10-26 [1] Bioconductor
 BiocGenerics           * 0.48.1         2023-11-02 [1] Bioconductor
 BiocIO                   1.12.0         2023-10-26 [1] Bioconductor
 BiocManager              1.30.22        2023-08-08 [1] CRAN (R 4.3.0)
 BiocNeighbors            1.20.2         2024-01-07 [1] Bioconductor 3.18 (R 4.3.2)
 BiocParallel           * 1.36.0         2023-10-26 [1] Bioconductor
 BiocSingular             1.18.0         2023-10-24 [1] Bioconductor
 BiocVersion              3.18.1         2023-11-18 [1] Bioconductor 3.18 (R 4.3.2)
 biomaRt                  2.58.2         2024-01-30 [1] Bioconductor 3.18 (R 4.3.2)
 Biostrings             * 2.70.2         2024-01-28 [1] Bioconductor 3.18 (R 4.3.2)
 bit                      4.0.5          2022-11-15 [1] CRAN (R 4.3.2)
 bit64                    4.0.5          2020-08-30 [1] CRAN (R 4.3.2)
 bitops                   1.0-7          2021-04-24 [1] CRAN (R 4.3.2)
 blob                     1.2.4          2023-03-17 [1] CRAN (R 4.3.0)
 broom                    1.0.5          2023-06-09 [1] CRAN (R 4.3.0)
 bslib                    0.6.1          2023-11-28 [1] CRAN (R 4.3.1)
 cachem                   1.0.8          2023-05-01 [1] CRAN (R 4.3.2)
 car                      3.1-2          2023-03-30 [1] CRAN (R 4.3.0)
 carData                  3.0-5          2022-01-06 [1] CRAN (R 4.3.0)
 cli                      3.6.2          2023-12-11 [1] CRAN (R 4.3.2)
 codetools                0.2-19         2023-02-01 [1] CRAN (R 4.3.0)
 colorout               * 1.3-0.1        2024-01-11 [1] local
 colorspace               2.1-0          2023-01-23 [1] CRAN (R 4.3.2)
 config                   0.3.2          2023-08-30 [1] CRAN (R 4.3.0)
 cowplot                  1.1.3          2024-01-22 [1] CRAN (R 4.3.1)
 crayon                   1.5.2          2022-09-29 [1] CRAN (R 4.3.0)
 curl                     5.2.0          2023-12-08 [1] CRAN (R 4.3.2)
 data.table             * 1.14.10        2023-12-08 [1] CRAN (R 4.3.2)
 DBI                      1.2.1          2024-01-12 [1] CRAN (R 4.3.2)
 dbplyr                   2.4.0          2023-10-26 [1] CRAN (R 4.3.1)
 default                * 1.0.0          2017-08-07 [1] CRAN (R 4.3.0)
 DelayedArray             0.28.0         2023-10-24 [1] Bioconductor
 DelayedMatrixStats       1.24.0         2023-11-06 [1] Bioconductor
 digest                   0.6.34         2024-01-11 [1] CRAN (R 4.3.2)
 doParallel               1.0.17         2022-02-07 [1] CRAN (R 4.3.0)
 dotCall64                1.1-1          2023-11-28 [1] CRAN (R 4.3.2)
 dplyr                    1.1.4          2023-11-17 [1] CRAN (R 4.3.2)
 DT                       0.31           2023-12-09 [1] CRAN (R 4.3.1)
 edgeR                    4.0.14         2024-01-29 [1] Bioconductor 3.18 (R 4.3.2)
 ellipsis                 0.3.2          2021-04-29 [1] CRAN (R 4.3.2)
 evaluate                 0.23           2023-11-01 [1] CRAN (R 4.3.1)
 ExperimentHub            2.10.0         2023-10-26 [1] Bioconductor
 fansi                    1.0.6          2023-12-08 [1] CRAN (R 4.3.2)
 fastmap                  1.1.1          2023-02-24 [1] CRAN (R 4.3.2)
 fields                   15.2           2023-08-17 [1] CRAN (R 4.3.2)
 filelock                 1.0.3          2023-12-11 [1] CRAN (R 4.3.2)
 foreach                  1.5.2          2022-02-02 [1] CRAN (R 4.3.0)
 fs                       1.6.3          2023-07-20 [1] CRAN (R 4.3.0)
 generics                 0.1.3          2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb           * 1.38.5         2023-12-30 [1] Bioconductor 3.18 (R 4.3.2)
 GenomeInfoDbData         1.2.11         2024-01-11 [1] Bioconductor
 GenomicAlignments        1.38.2         2024-01-16 [1] Bioconductor 3.18 (R 4.3.2)
 GenomicRanges          * 1.54.1         2023-10-29 [1] Bioconductor
 ggbeeswarm               0.7.2          2023-04-29 [1] CRAN (R 4.3.0)
 ggfun                    0.1.4          2024-01-19 [1] CRAN (R 4.3.2)
 ggplot2                  3.4.4          2023-10-12 [1] CRAN (R 4.3.1)
 ggplotify                0.1.2          2023-08-09 [1] CRAN (R 4.3.0)
 ggpubr                   0.6.0          2023-02-10 [1] CRAN (R 4.3.0)
 ggrepel                  0.9.5          2024-01-10 [1] CRAN (R 4.3.2)
 ggsignif                 0.6.4          2022-10-13 [1] CRAN (R 4.3.0)
 ggtree                   3.10.0         2023-11-06 [1] Bioconductor
 glue                     1.7.0          2024-01-09 [1] CRAN (R 4.3.2)
 golem                    0.4.1          2023-06-05 [1] CRAN (R 4.3.0)
 gprofiler2               0.2.2          2023-06-14 [1] CRAN (R 4.3.0)
 gridExtra                2.3            2017-09-09 [1] CRAN (R 4.3.0)
 gridGraphics             0.5-1          2020-12-13 [1] CRAN (R 4.3.0)
 grr                      0.9.5          2016-08-26 [1] CRAN (R 4.3.2)
 gtable                   0.3.4          2023-08-21 [1] CRAN (R 4.3.0)
 here                     1.0.1          2020-12-13 [1] CRAN (R 4.3.0)
 hms                      1.1.3          2023-03-21 [1] CRAN (R 4.3.0)
 homologene               1.4.68.19.3.27 2019-03-28 [1] CRAN (R 4.3.0)
 htmltools                0.5.7          2023-11-03 [1] CRAN (R 4.3.2)
 htmlwidgets              1.6.4          2023-12-06 [1] CRAN (R 4.3.1)
 httpuv                   1.6.13         2023-12-06 [1] CRAN (R 4.3.1)
 httr                     1.4.7          2023-08-15 [1] CRAN (R 4.3.0)
 interactiveDisplayBase   1.40.0         2023-10-26 [1] Bioconductor
 IRanges                * 2.36.0         2023-10-24 [1] Bioconductor
 irlba                    2.3.5.1        2022-10-03 [1] CRAN (R 4.3.2)
 iterators                1.0.14         2022-02-05 [1] CRAN (R 4.3.0)
 jquerylib                0.1.4          2021-04-26 [1] CRAN (R 4.3.0)
 jsonlite                 1.8.8          2023-12-04 [1] CRAN (R 4.3.2)
 KEGGREST                 1.42.0         2023-10-26 [1] Bioconductor
 knitr                    1.45           2023-10-30 [1] CRAN (R 4.3.2)
 later                    1.3.2          2023-12-06 [1] CRAN (R 4.3.2)
 lattice                  0.22-5         2023-10-24 [1] CRAN (R 4.3.2)
 lazyeval                 0.2.2          2019-03-15 [1] CRAN (R 4.3.2)
 lifecycle                1.0.4          2023-11-07 [1] CRAN (R 4.3.1)
 limma                    3.58.1         2023-10-31 [1] Bioconductor
 locfit                   1.5-9.8        2023-06-11 [1] CRAN (R 4.3.2)
 magick                   2.8.2          2023-12-20 [1] CRAN (R 4.3.1)
 magrittr               * 2.0.3          2022-03-30 [1] CRAN (R 4.3.2)
 maps                     3.4.2          2023-12-15 [1] CRAN (R 4.3.1)
 Matrix                   1.6-5          2024-01-11 [1] CRAN (R 4.3.2)
 MatrixGenerics         * 1.14.0         2023-10-26 [1] Bioconductor
 matrixStats            * 1.2.0          2023-12-11 [1] CRAN (R 4.3.2)
 memoise                  2.0.1          2021-11-26 [1] CRAN (R 4.3.0)
 mime                     0.12           2021-09-28 [1] CRAN (R 4.3.2)
 munsell                  0.5.0          2018-06-12 [1] CRAN (R 4.3.0)
 nlme                     3.1-164        2023-11-27 [1] CRAN (R 4.3.2)
 orthogene              * 1.8.0          2023-11-06 [1] Bioconductor
 paletteer                1.6.0          2024-01-21 [1] CRAN (R 4.3.1)
 parallelly             * 1.36.0         2023-05-26 [1] CRAN (R 4.3.0)
 patchwork                1.2.0          2024-01-08 [1] CRAN (R 4.3.1)
 pillar                   1.9.0          2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig                2.0.3          2019-09-22 [1] CRAN (R 4.3.0)
 plotly                   4.10.4         2024-01-13 [1] CRAN (R 4.3.1)
 png                      0.1-8          2022-11-29 [1] CRAN (R 4.3.2)
 prettyunits              1.2.0          2023-09-24 [1] CRAN (R 4.3.1)
 progress                 1.2.3          2023-12-06 [1] CRAN (R 4.3.1)
 promises                 1.2.1          2023-08-10 [1] CRAN (R 4.3.2)
 purrr                    1.0.2          2023-08-10 [1] CRAN (R 4.3.2)
 R6                       2.5.1          2021-08-19 [1] CRAN (R 4.3.0)
 rappdirs                 0.3.3          2021-01-31 [1] CRAN (R 4.3.2)
 RColorBrewer             1.1-3          2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp                     1.0.12         2024-01-09 [1] CRAN (R 4.3.2)
 RCurl                    1.98-1.14      2024-01-09 [1] CRAN (R 4.3.1)
 rematch2                 2.1.2          2020-05-01 [1] CRAN (R 4.3.0)
 restfulr                 0.0.15         2022-06-16 [1] CRAN (R 4.3.2)
 rjson                    0.2.21         2022-01-09 [1] CRAN (R 4.3.2)
 rlang                  * 1.1.3          2024-01-10 [1] CRAN (R 4.3.2)
 rmarkdown                2.25           2023-09-18 [1] CRAN (R 4.3.2)
 rprojroot                2.0.4          2023-11-05 [1] CRAN (R 4.3.1)
 Rsamtools                2.18.0         2023-10-24 [1] Bioconductor
 rsconnect                1.2.0          2023-12-15 [1] CRAN (R 4.3.1)
 RSQLite                  2.3.5          2024-01-21 [1] CRAN (R 4.3.2)
 rstatix                  0.7.2          2023-02-01 [1] CRAN (R 4.3.0)
 rstudioapi               0.15.0         2023-07-07 [1] CRAN (R 4.3.0)
 rsvd                     1.0.5          2021-04-16 [1] CRAN (R 4.3.0)
 rtracklayer              1.62.0         2023-10-26 [1] Bioconductor
 S4Arrays                 1.2.0          2023-10-24 [1] Bioconductor
 S4Vectors              * 0.40.2         2023-11-23 [1] Bioconductor 3.18 (R 4.3.2)
 sass                     0.4.8          2023-12-06 [1] CRAN (R 4.3.1)
 ScaledMatrix             1.10.0         2023-11-06 [1] Bioconductor
 scales                   1.3.0          2023-11-28 [1] CRAN (R 4.3.2)
 scater                   1.30.1         2023-11-16 [1] Bioconductor
 scuttle                  1.12.0         2023-10-24 [1] Bioconductor
 sessioninfo              1.2.2          2021-12-06 [1] CRAN (R 4.3.0)
 shiny                    1.8.0          2023-11-17 [1] CRAN (R 4.3.1)
 shinyWidgets             0.8.1          2024-01-10 [1] CRAN (R 4.3.1)
 SingleCellExperiment   * 1.24.0         2023-11-06 [1] Bioconductor
 spam                     2.10-0         2023-10-23 [1] CRAN (R 4.3.2)
 SparseArray              1.2.3          2023-12-25 [1] Bioconductor 3.18 (R 4.3.2)
 sparseMatrixStats        1.14.0         2023-10-24 [1] Bioconductor
 SpatialExperiment      * 1.12.0         2023-10-26 [1] Bioconductor
 spatialLIBD            * 1.14.1         2023-11-30 [1] Bioconductor 3.18 (R 4.3.2)
 statmod                  1.5.0          2023-01-06 [1] CRAN (R 4.3.2)
 stringi                  1.8.3          2023-12-11 [1] CRAN (R 4.3.1)
 stringr                  1.5.1          2023-11-14 [1] CRAN (R 4.3.1)
 SummarizedExperiment   * 1.32.0         2023-11-06 [1] Bioconductor
 tibble                   3.2.1          2023-03-20 [1] CRAN (R 4.3.2)
 tidyr                    1.3.1          2024-01-24 [1] CRAN (R 4.3.1)
 tidyselect               1.2.0          2022-10-10 [1] CRAN (R 4.3.0)
 tidytree                 0.4.6          2023-12-12 [1] CRAN (R 4.3.1)
 treeio                   1.26.0         2023-11-06 [1] Bioconductor
 utf8                     1.2.4          2023-10-22 [1] CRAN (R 4.3.2)
 vctrs                    0.6.5          2023-12-01 [1] CRAN (R 4.3.2)
 vipor                    0.4.7          2023-12-18 [1] CRAN (R 4.3.1)
 viridis                  0.6.5          2024-01-29 [1] CRAN (R 4.3.2)
 viridisLite              0.4.2          2023-05-02 [1] CRAN (R 4.3.0)
 withr                    3.0.0          2024-01-16 [1] CRAN (R 4.3.1)
 xfun                     0.41           2023-11-01 [1] CRAN (R 4.3.2)
 XML                      3.99-0.16.1    2024-01-22 [1] CRAN (R 4.3.1)
 xml2                     1.3.6          2023-12-04 [1] CRAN (R 4.3.2)
 xtable                   1.8-4          2019-04-21 [1] CRAN (R 4.3.0)
 XVector                * 0.42.0         2023-10-24 [1] Bioconductor
 yaml                     2.3.8          2023-12-11 [1] CRAN (R 4.3.2)
 yulab.utils              0.1.4          2024-01-28 [1] CRAN (R 4.3.2)
 zlibbioc                 1.48.0         2023-10-26 [1] Bioconductor

 [1] /Users/bmulvey/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────
